// @oracle/oraclejet-preact/cjs/UNSAFE_ContainerDiagram/themes/DiagramStyles.css.js
/* @oracle/oraclejet-preact: undefined */
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });



var buttonStyle = 'DiagramStyles_buttonStyle__18tpn7n5';
var childContentStyle = 'DiagramStyles_childContentStyle__18tpn7n7';
var dimensionStyle = 'DiagramStyles_dimensionStyle__18tpn7na';
var itemStyles = {nodeStyles:'DiagramStyles_nodeStyles__18tpn7n2',linkStyles:'DiagramStyles_linkStyles__18tpn7n3',dimmedItemStyle:'DiagramStyles_dimmedItemStyle__18tpn7n8'};
var labelStyles = {labelStyle:'DiagramStyles_labelStyle__18tpn7n4',labelBorderStyle:'DiagramStyles_labelBorderStyle__18tpn7n6'};
var layerStyles = 'DiagramStyles_layerStyles__18tpn7n1';
var outerStyles = 'DiagramStyles_outerStyles__18tpn7n0';
var panZoomStyles = 'DiagramStyles_panZoomStyles__18tpn7n9';

exports.buttonStyle = buttonStyle;
exports.childContentStyle = childContentStyle;
exports.dimensionStyle = dimensionStyle;
exports.itemStyles = itemStyles;
exports.labelStyles = labelStyles;
exports.layerStyles = layerStyles;
exports.outerStyles = outerStyles;
exports.panZoomStyles = panZoomStyles;
