// @oracle/oraclejet-preact/cjs/resources/nls/ar-XB/bundle.js
/* @oracle/oraclejet-preact: undefined */
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

// This file is auto-generated by @oracle/oraclejet-icu-l10n from src/resources/icu/bundle.json
// and should not be edited by hand. Update the JSON file and rerun the build to regenerate this content.
const bundle = {
    accAnnounceMovedKey: (p) => '[РђЈРђ«Moved ' + p['itemKey'] + ' ' + p['position'] + ' ' + p['referenceKey'] + '.РђгРђЈ]',
    accButtonProgressComplete: (p) => '[РђЈРђ«' + p['label'] + ' action completedРђгРђЈ]',
    accButtonProgressStatus: (p) => '[РђЈРђ«' + p['label'] + ' action in progressРђгРђЈ]',
    accColorPalette: () => '[РђЈРђ«Color PaletteРђгРђЈ]',
    accColorSwatch: () => '[РђЈРђ«Color SwatchРђгРђЈ]',
    accCommandKey: () => '[РђЈРђ«CommandРђгРђЈ]',
    accContextMenuLabel: () => '[РђЈРђ«actionsРђгРђЈ]',
    accControlKey: () => '[РђЈРђ«CloseРђгРђЈ]',
    accPositionAfter: () => '[РђЈРђ«afterРђгРђЈ]',
    accPositionBefore: () => '[РђЈРђ«beforeРђгРђЈ]',
    accReorderInstructions: (p) => '[РђЈРђ«Press and hold ' +
        p['keyPressed'] +
        ' and shift keys, then use the arrow keys to reorder.РђгРђЈ]',
    calendarDateConverter_parseError: (p) => '[РђЈРђ«Enter a complete date like this: ' + p['dateExample'] + '.РђгРђЈ]',
    calendarMonthConverter_parseError: (p) => '[РђЈРђ«Enter a month and year like this: ' + p['dateExample'] + '.РђгРђЈ]',
    chart_labelDataLabel: () => '[РђЈРђ«LabelРђгРђЈ]',
    chart_labelGroup: () => '[РђЈРђ«GroupРђгРђЈ]',
    chart_labelSeries: () => '[РђЈРђ«SeriesРђгРђЈ]',
    chart_labelValue: () => '[РђЈРђ«ValueРђгРђЈ]',
    chart_labelWidthWeight: () => '[РђЈРђ«ZРђгРђЈ]',
    chart_labelX: () => '[РђЈРђ«XРђгРђЈ]',
    chart_labelY: () => '[РђЈРђ«YРђгРђЈ]',
    checkboxSet_requiredMessageDetail: () => '[РђЈРђ«Select a value.РђгРђЈ]',
    checkbox_helpForCheckbox: () => '[РђЈРђ«Help for checkboxРђгРђЈ]',
    checkbox_requiredMessageDetail: () => '[РђЈРђ«Select a value.РђгРђЈ]',
    checkbox_selected: () => '[РђЈРђ«Checkbox checkedРђгРђЈ]',
    checkbox_unselected: () => '[РђЈРђ«Checkbox not checkedРђгРђЈ]',
    close: () => '[РђЈРђ«CloseРђгРђЈ]',
    collection_apply: () => '[РђЈРђ«ApplyРђгРђЈ]',
    collection_cancel: () => '[РђЈРђ«CancelРђгРђЈ]',
    collection_loaded: () => '[РђЈРђ«LoadedРђгРђЈ]',
    collection_loading: () => '[РђЈРђ«LoadingРђгРђЈ]',
    collection_noData: () => '[РђЈРђ«No data to display.РђгРђЈ]',
    collection_pixels: () => '[РђЈРђ«pxРђгРђЈ]',
    collection_resizeColumnDialog: () => '[РђЈРђ«Resize columnРђгРђЈ]',
    collection_resizeColumnMenu: () => '[РђЈРђ«Resize Column...РђгРђЈ]',
    collection_resizeMessageDetail: (p) => '[РђЈРђ«Enter a number between ' + p['MIN'] + ' and ' + p['MAX'] + '.РђгРђЈ]',
    collection_selectAllRows: () => '[РђЈРђ«Select all rowsРђгРђЈ]',
    collection_selectRow: (p) => '[РђЈРђ«Select row ' + p['ROW_NAME'] + 'РђгРђЈ]',
    collection_sortAscendingSubMenu: () => '[РђЈРђ«AscendingРђгРђЈ]',
    collection_sortColumnMenu: () => '[РђЈРђ«Sort ColumnРђгРђЈ]',
    collection_sortDescendingSubMenu: () => '[РђЈРђ«DescendingРђгРђЈ]',
    collection_sortDisabled: () => '[РђЈРђ«Not sortableРђгРђЈ]',
    collection_sortEnabled: () => '[РђЈРђ«SortableРђгРђЈ]',
    collection_width: () => '[РђЈРђ«WidthРђгРђЈ]',
    colorPicker_label: () => '[РђЈРђ«Color PickerРђгРђЈ]',
    confirmation: () => '[РђЈРђ«ConfirmationРђгРђЈ]',
    datePicker_currentDate: () => '[РђЈРђ«Current dateРђгРђЈ]',
    datePicker_goToToday: () => '[РђЈРђ«Go to todayРђгРђЈ]',
    datePicker_label: () => '[РђЈРђ«Date PickerРђгРђЈ]',
    datePicker_month: () => '[РђЈРђ«MonthРђгРђЈ]',
    datePicker_months: () => '[РђЈРђ«MonthsРђгРђЈ]',
    datePicker_next: () => '[РђЈРђ«NextРђгРђЈ]',
    datePicker_nextDecade: () => '[РђЈРђ«Next decadeРђгРђЈ]',
    datePicker_nextMonth: () => '[РђЈРђ«Next monthРђгРђЈ]',
    datePicker_prev: () => '[РђЈРђ«PreviousРђгРђЈ]',
    datePicker_prevDecade: () => '[РђЈРђ«Previous decadeРђгРђЈ]',
    datePicker_prevMonth: () => '[РђЈРђ«Previous monthРђгРђЈ]',
    datePicker_restricted: () => '[РђЈРђ«RestrictedРђгРђЈ]',
    datePicker_selected: () => '[РђЈРђ«SelectedРђгРђЈ]',
    datePicker_today: () => '[РђЈРђ«TodayРђгРђЈ]',
    datePicker_weekNum: (p) => '[РђЈРђ«Week ' + p['NUM'] + 'РђгРђЈ]',
    datePicker_weekNumColLabel: () => '[РђЈРђ«Week NumberРђгРђЈ]',
    datePicker_year: () => '[РђЈРђ«YearРђгРђЈ]',
    datePicker_years: () => '[РђЈРђ«YearsРђгРђЈ]',
    diagram_promotedLink: () => '[РђЈРђ«1 linkРђгРђЈ]',
    diagram_promotedLinkAriaDesc: () => '[РђЈРђ«IndirectРђгРђЈ]',
    diagram_promotedLinks: (p) => '[РђЈРђ«' + p['num_links'] + ' linksРђгРђЈ]',
    error: () => '[РђЈРђ«ErrorРђгРђЈ]',
    expandableList_expandCollapseInstructionText: () => '[РђЈРђ«Use arrow keys to expand and collapseРђгРђЈ]',
    expandableList_groupCollapse: () => '[РђЈРђ«CollapsedРђгРђЈ]',
    expandableList_groupExpand: () => '[РђЈРђ«ExpandedРђгРђЈ]',
    filePicker_addFiles: () => '[РђЈРђ«Add FilesРђгРђЈ]',
    filePicker_dropzonePrimaryText: () => '[РђЈРђ«Drag and DropРђгРђЈ]',
    filePicker_dropzoneSecondaryText: () => '[РђЈРђ«Select a file or drop one hereРђгРђЈ]',
    filePicker_dropzoneSecondaryTextMultiple: () => '[РђЈРђ«Select or drop files hereРђгРђЈ]',
    filePicker_multipleFileTypeUploadError: (p) => "[РђЈРђ«You can't upload files of type: " + p['fileTypes'] + 'РђгРђЈ]',
    filePicker_singleFileUploadError: () => '[РђЈРђ«Upload one file at a timeРђгРђЈ]',
    filePicker_singleTypeUploadError: (p) => "[РђЈРђ«You can't upload files of type " + p['fileType'] + 'РђгРђЈ]',
    filePicker_unknownFileTypeUploadError: () => '[РђЈРђ«unknownРђгРђЈ]',
    formControl_clear: () => '[РђЈРђ«ClearРђгРђЈ]',
    formControl_day: () => '[РђЈРђ«DayРђгРђЈ]',
    formControl_empty_segment: () => '[РђЈРђ«EmptyРђгРђЈ]',
    formControl_helpAvailable: () => '[РђЈРђ«Help availableРђгРђЈ]',
    formControl_hide: () => '[РђЈРђ«HideРђгРђЈ]',
    formControl_hour: () => '[РђЈРђ«HourРђгРђЈ]',
    formControl_limitReached: (p) => '[РђЈРђ«YouРђЎve reached the ' +
        ({
            one: new Intl.NumberFormat('ar-XB').format(p['CHARACTER_LIMIT']) + '-character'
        }[new Intl.PluralRules('ar-XB').select(p['CHARACTER_LIMIT'])] ||
            new Intl.NumberFormat('ar-XB').format(p['CHARACTER_LIMIT']) + '-character') +
        ' limit.РђгРђЈ]',
    formControl_loading: () => '[РђЈРђ«LoadingРђгРђЈ]',
    formControl_maxLength: (p) => '[РђЈРђ«You can enter up to ' + p['MAX_LENGTH'] + ' characters.РђгРђЈ]',
    formControl_maxLengthExceeded: (p) => '[РђЈРђ«Maximum length ' + p['MAX_LENGTH'] + ' exceeded.РђгРђЈ]',
    formControl_maxLengthRemaining: (p) => '[РђЈРђ«' + p['CHARACTER_COUNT'] + ' characters left.РђгРђЈ]',
    formControl_millisecond: () => '[РђЈРђ«MillisecondРђгРђЈ]',
    formControl_minute: () => '[РђЈРђ«MinuteРђгРђЈ]',
    formControl_month: () => '[РђЈРђ«MonthРђгРђЈ]',
    formControl_readOnly: () => '[РђЈРђ«Read onlyРђгРђЈ]',
    formControl_requiredMessageDetail: () => '[РђЈРђ«Enter a value.РђгРђЈ]',
    formControl_second: () => '[РђЈРђ«SecondРђгРђЈ]',
    formControl_selectionExactMessageDetail: (p) => '[РђЈРђ«Select ' + p['n'] + ' values.РђгРђЈ]',
    formControl_selectionOverflowMessageDetail: (p) => '[РђЈРђ«YouРђЎve reached the ' + p['max'] + '-selection limit.РђгРђЈ]',
    formControl_selectionRangeMessageDetail: (p) => '[РђЈРђ«Select ' + p['min'] + ' to ' + p['max'] + ' values.РђгРђЈ]',
    formControl_selectionUnderflowMessageDetail: (p) => '[РђЈРђ«Select at least ' + p['min'] + ' values.РђгРђЈ]',
    formControl_show: () => '[РђЈРђ«ShowРђгРђЈ]',
    formControl_year: () => '[РђЈРђ«YearРђгРђЈ]',
    gantt_labelDate: () => '[РђЈРђ«DateРђгРђЈ]',
    gantt_labelEnd: () => '[РђЈРђ«EndРђгРђЈ]',
    gantt_labelLabel: () => '[РђЈРђ«LabelРђгРђЈ]',
    gantt_labelRow: () => '[РђЈРђ«RowРђгРђЈ]',
    gantt_labelStart: () => '[РђЈРђ«StartРђгРђЈ]',
    indexer_disabledLabel: (p) => '[РђЈРђ«' + p['SECTION'] + '. No matching section found.РђгРђЈ]',
    indexer_keyboardInstructionText: () => '[РђЈРђ«Press enter to select value.РђгРђЈ]',
    indexer_othersLabel: () => '#',
    indexer_separatorLabel: (p) => '[РђЈРђ«Between ' + p['FROM_SECTION'] + ' and ' + p['TO_SECTION'] + '.РђгРђЈ]',
    indexer_touchInstructionText: () => '[РђЈРђ«Double tap and hold to enter gesture mode, then drag up or down to adjust value.РђгРђЈ]',
    info: () => '[РђЈРђ«InfoРђгРђЈ]',
    inputColor_selectColor: () => '[РђЈРђ«Select a colorРђгРђЈ]',
    inputDateMask_date_cleared: () => '[РђЈРђ«Date clearedРђгРђЈ]',
    inputDateMask_dayPlaceholder: () => '[РђЈРђ«ddРђгРђЈ]',
    inputDateMask_empty_segment: () => '[РђЈРђ«EmptyРђгРђЈ]',
    inputDateMask_monthPlaceholder: () => '[РђЈРђ«mmРђгРђЈ]',
    inputDateMask_yearPlaceholder: () => '[РђЈРђ«yyyyРђгРђЈ]',
    inputDatePicker_dateRangeOverflowMessageDetail: (p) => "[РђЈРђ«Enter a date that's on or before " + p['max'] + '.РђгРђЈ]',
    inputDatePicker_dateRangeUnderflowMessageDetail: (p) => "[РђЈРђ«Enter a date that's on or after " + p['min'] + '.РђгРђЈ]',
    inputDatePicker_dateRestrictionMessageDetail: () => '[РђЈРђ«Enter an available date.РђгРђЈ]',
    inputDatePicker_instruction: () => '[РђЈРђ«Press Tab to access the calendar.РђгРђЈ]',
    inputDatePicker_selectDate: () => '[РђЈРђ«Select a dateРђгРђЈ]',
    inputMonthMask_dateRangeOverflowMessageDetail: (p) => "[РђЈРђ«Enter a month and year that's on or before " + p['max'] + '.РђгРђЈ]',
    inputMonthMask_dateRangeUnderflowMessageDetail: (p) => "[РђЈРђ«Enter a month and year that's on or after " + p['min'] + '.РђгРђЈ]',
    inputNumber_converterParseError: () => '[РђЈРђ«Enter a number.РђгРђЈ]',
    inputNumber_decrease: () => '[РђЈРђ«DecreaseРђгРђЈ]',
    inputNumber_increase: () => '[РђЈРђ«IncreaseРђгРђЈ]',
    inputPassword_hidden: () => '[РђЈРђ«Password HiddenРђгРђЈ]',
    inputPassword_hide: () => '[РђЈРђ«Hide PasswordРђгРђЈ]',
    inputPassword_show: () => '[РђЈРђ«Show PasswordРђгРђЈ]',
    inputSensitiveText_hidden: () => '[РђЈРђ«Sensitive text hiddenРђгРђЈ]',
    inputTimeMask_timeRangeOverflowMessageDetail: (p) => "[РђЈРђ«Enter a time that's " + p['max'] + ' or earlier.РђгРђЈ]',
    inputTimeMask_timeRangeUnderflowMessageDetail: (p) => "[РђЈРђ«Enter a time that's " + p['min'] + ' or later.РђгРђЈ]',
    inputTimeMask_time_cleared: () => '[РђЈРђ«Time clearedРђгРђЈ]',
    labelNone: () => '[РђЈРђ«NoneРђгРђЈ]',
    list_msgFetchCompleted: () => '[РђЈРђ«LoadedРђгРђЈ]',
    list_msgFetchingData: () => '[РђЈРђ«LoadingРђгРђЈ]',
    list_suggestion: () => '[РђЈРђ«SuggestionРђгРђЈ]',
    menu_stateLoading: () => '[РђЈРђ«LoadingРђгРђЈ]',
    messageToast_allMessagesClosed: () => '[РђЈРђ«All toast messages are closed.РђгРђЈ]',
    message_close: () => '[РђЈРђ«CloseРђгРђЈ]',
    message_confirmation: () => '[РђЈРђ«SuccessРђгРђЈ]',
    message_error: () => '[РђЈРђ«ErrorРђгРђЈ]',
    message_info: () => '[РђЈРђ«InfoРђгРђЈ]',
    message_navigationFromMessagesRegion: () => '[РђЈРђ«Entering messages region. Press F6 to navigate.РђгРђЈ]',
    message_navigationToMessagesRegion: () => '[РђЈРђ«Messages region has new messages. Press F6 to navigate.РђгРђЈ]',
    message_navigationToTouch: () => '[РђЈРђ«Messages region has new messages.РђгРђЈ]',
    message_warning: () => '[РђЈРђ«WarningРђгРђЈ]',
    noData_message: () => '[РђЈРђ«No items to displayРђгРђЈ]',
    overflowItemLabel: () => '[РђЈРђ«More TabsРђгРђЈ]',
    plural_separator: () => ', ',
    popup_close_skipLink: () => '[РђЈРђ«Double tap to close the open pop-up.РђгРђЈ]',
    popup_liveRegion_F6_initial_focus_firstFocusable: () => '[РђЈРђ«Entering pop-up. Press F6 to navigate between the pop-up and associated control.РђгРђЈ]',
    popup_liveRegion_F6_initial_focus_none: () => '[РђЈРђ«Pop-up opened. Press F6 to navigate between the pop-up and associated control.РђгРђЈ]',
    progressIndeterminate: () => '[РђЈРђ«In ProgressРђгРђЈ]',
    radioButton_checked: () => '[РђЈРђ«Radio button checkedРђгРђЈ]',
    radioButton_selected: () => '[РђЈРђ«Radio button selectedРђгРђЈ]',
    radioButton_unchecked: () => '[РђЈРђ«Radio button not checkedРђгРђЈ]',
    radioButton_unselected: () => '[РђЈРђ«Radio button not selectedРђгРђЈ]',
    radio_helpForRadio: () => '[РђЈРђ«Help for radio buttonРђгРђЈ]',
    radio_requiredMessageDetail: () => '[РђЈРђ«Select a value.РђгРђЈ]',
    selectMultiple_apply: () => '[РђЈРђ«ApplyРђгРђЈ]',
    selectMultiple_back: () => '[РђЈРђ«BackРђгРђЈ]',
    selectMultiple_countPlus: (p) => p['COUNT'] + '+',
    selectMultiple_highlightSelectedTagsInstructionText: () => '[РђЈРђ«Use left and right arrow keys to highlight selected values.РђгРђЈ]',
    selectMultiple_removeSelectedTagInstructionText: () => '[РђЈРђ«Press Backspace or Delete to remove the selected value.РђгРђЈ]',
    selectMultiple_selectedValues: () => '[РђЈРђ«Selected valuesРђгРђЈ]',
    selectMultiple_showSelectedValues: () => '[РђЈРђ«Show selected values only.РђгРђЈ]',
    selectMultiple_valuesSelected: (p) => '[РђЈРђ«' + p['VALUE_COUNT'] + ' values selected.РђгРђЈ]',
    select_addToList: () => '[РђЈРђ«Add to listРђгРђЈ]',
    select_addToListAvailable: () => '[РђЈРђ«Add to list availableРђгРђЈ]',
    select_closeDropdown: () => '[РђЈРђ«Press Escape to close.РђгРђЈ]',
    select_moreSearchOptions: () => '[РђЈРђ«More search optionsРђгРђЈ]',
    select_moreSearchOptionsAvailable: () => '[РђЈРђ«More search options availableРђгРђЈ]',
    select_noMatchesFound: () => '[РђЈРђ«No matches found.РђгРђЈ]',
    select_oneMatchFound: () => '[РђЈРђ«One match foundРђгРђЈ]',
    select_requiredMessageDetail: () => '[РђЈРђ«Select a value.РђгРђЈ]',
    select_sizeMatchesFound: (p) => '[РђЈРђ«' + p['TOTAL_SIZE'] + ' matches foundРђгРђЈ]',
    select_sizeOrMoreMatchesFound: (p) => '[РђЈРђ«' + p['TOTAL_SIZE'] + ' or more matches foundРђгРђЈ]',
    selectorAll_deselectAll: () => '[РђЈРђ«Deselect allРђгРђЈ]',
    selectorAll_selectAll: () => '[РђЈРђ«Select allРђгРђЈ]',
    selector_selected: () => '[РђЈРђ«Checkbox checkedРђгРђЈ]',
    selector_unselected: () => '[РђЈРђ«Checkbox not checkedРђгРђЈ]',
    sheet_close: () => '[РђЈРђ«CloseРђгРђЈ]',
    switch_off: () => '[РђЈРђ«OffРђгРђЈ]',
    switch_on: () => '[РђЈРђ«OnРђгРђЈ]',
    tabBarNavigationList_removeCueText: () => '[РђЈРђ«RemovableРђгРђЈ]',
    tabBar_tabContainsErrors: () => '[РђЈРђ«Contains errorsРђгРђЈ]',
    tabbar_accAnnounceRemovedKey: (p) => '[РђЈРђ«Removed ' + p['itemKey'] + 'РђгРђЈ]',
    tabbar_labelContextMenu: () => '[РђЈРђ«has context menuРђгРђЈ]',
    tabbar_labelMoveLeft: () => '[РђЈРђ«Move LeftРђгРђЈ]',
    tabbar_labelMoveRight: () => '[РђЈРђ«Move RightРђгРђЈ]',
    tabbar_labelRemove: () => '[РђЈРђ«RemoveРђгРђЈ]',
    timeComponent_tooltipZoomIn: () => '[РђЈРђ«Zoom InРђгРђЈ]',
    timeComponent_tooltipZoomOut: () => '[РђЈРђ«Zoom OutРђгРђЈ]',
    timeConverter_parseError: () => '[РђЈРђ«Enter a complete time.РђгРђЈ]',
    train_current: () => '[РђЈРђ«CurrentРђгРђЈ]',
    train_message_type: () => '[РђЈРђ«Message TypeРђгРђЈ]',
    train_not_visited: () => '[РђЈРђ«Not visitedРђгРђЈ]',
    train_status: (p) => '[РђЈРђ«Step ' + p['currentStep'] + ' out of ' + p['numberOfSteps'] + 'РђгРђЈ]',
    train_visited: () => '[РђЈРђ«VisitedРђгРђЈ]',
    userAssistance_learnMore: () => '[РђЈРђ«Learn moreРђгРђЈ]',
    userAssistance_required: () => '[РђЈРђ«RequiredРђгРђЈ]',
    vis_clearSelection: () => '[РђЈРђ«Clear SelectionРђгРђЈ]',
    vis_drillable: () => '[РђЈРђ«DrillableРђгРђЈ]',
    vis_invalidData: () => '[РђЈРђ«Invalid dataРђгРђЈ]',
    vis_labelAndValue: (p) => p['LABEL'] + ': ' + p['VALUE'],
    vis_labelCountWithTotal: (p) => '[РђЈРђ«' + p['itemCount'] + ' out of ' + p['totalCount'] + 'РђгРђЈ]',
    vis_marqueeSelectTooltip: () => '[РђЈРђ«Marquee selectРђгРђЈ]',
    vis_marqueeZoomTooltip: () => '[РђЈРђ«Marquee zoomРђгРђЈ]',
    vis_noData: () => '[РђЈРђ«No data to displayРђгРђЈ]',
    vis_panTooltip: () => '[РђЈРђ«PanРђгРђЈ]',
    vis_stateCollapsed: () => '[РђЈРђ«CollapsedРђгРђЈ]',
    vis_stateExpanded: () => '[РђЈРђ«ExpandedРђгРђЈ]',
    vis_stateLoaded: () => '[РђЈРђ«LoadedРђгРђЈ]',
    vis_stateLoading: () => '[РђЈРђ«LoadingРђгРђЈ]',
    vis_stateSelected: () => '[РђЈРђ«SelectedРђгРђЈ]',
    vis_stateUnselected: () => '[РђЈРђ«UnselectedРђгРђЈ]',
    warn: () => '[РђЈРђ«WarningРђгРђЈ]'
};

exports["default"] = bundle;
