// @oracle/oraclejet-preact/es/Dropdown-019aba37.js
/* @oracle/oraclejet-preact: undefined */
import { jsx, Fragment } from 'preact/jsx-runtime';
import { useRef, useCallback, useEffect } from 'preact/hooks';
import { F as Floating } from './Floating-4a82db94.js';
import { L as Layer } from './Layer-535577b5.js';
import 'preact/compat';
import './LayerHost-7931b424.js';
import { K as KEYS } from './keyboardUtils-986fe923.js';
import { f as focusWithin } from './tabbableUtils-c8311ec6.js';
import { u as usePopupAnimation } from './usePopupAnimation-96d2bc83.js';
import { g as getTransformOrigin } from './popupAnimationUtils-175151d9.js';
import { C as CollectionInteractionContext } from './CollectionInteractionContext-7cf19abc.js';
import { dropdownStyles } from './UNSAFE_Dropdown/themes/dropdownStyles.css.js';
import { u as useOutsideMousedown } from './useOutsideMousedown-c45c22c0.js';
import { m as mergeAnimationStateObjects, D as DURATION_LARGE, b as SCALE_UP_FULL } from './animationUtils-5f409a6c.js';
import { U as UNSAFE_DOM_ACCESS } from './unsafeDomAccess-453eb74e.js';
import { u as useEffectEvent } from './useEffectEvent-1bb008cb.js';
import { g as getClientHints } from './clientHints-a42b510c.js';

/**
 * @license
 * Copyright (c) %FIRST_YEAR% %CURRENT_YEAR%, Oracle and/or its affiliates.
 * Licensed under The Universal Permissive License (UPL), Version 1.0
 * as shown at https://oss.oracle.com/licenses/upl/
 * @ignore
 */
/**
 * Returns the key from the item element
 */
const getKey = (item) => {
    // Key is set on data-oj-menu-item-key on each menuItem. To get this key from the
    // the actual menuItem we need to do a kebab to camel-case conversion.
    return item.dataset['ojMenuItemKey'];
};
/**
 * A helper function that return the key of the n available item.
 */
const getNVisible = (root, positionIndex) => {
    if (root) {
        const items = root.querySelectorAll('[data-oj-menu-item-key]');
        const positionedItem = items[positionIndex];
        if (positionedItem) {
            const key = getKey(positionedItem);
            if (key) {
                return key;
            }
        }
    }
    return null;
};
/**
 * A helper function that return the key of the last available item.
 */
const getLastVisible = (root) => {
    if (root) {
        const items = root.querySelectorAll('[data-oj-menu-item-key]');
        const lastItem = items[items.length - 1];
        if (lastItem) {
            const key = getKey(lastItem);
            if (key) {
                return key;
            }
        }
    }
    return null;
};
/**
 * A helper function that returns the key of the previous or the next item
 * given the specified current key
 */
const getPrevNext = (root, isPrev, currentKey) => {
    return () => {
        if (root) {
            const items = Array.from(root.querySelectorAll('[data-oj-menu-item-key]'));
            const currentKeyIndex = items.findIndex((item) => getKey(item) === currentKey);
            let nextIndex = currentKeyIndex + (isPrev ? -1 : 1);
            const itemsLastIndex = items.length - 1;
            if (nextIndex < 0) {
                nextIndex = itemsLastIndex;
            }
            if (nextIndex > itemsLastIndex) {
                nextIndex = 0;
            }
            const key = getKey(items[nextIndex]);
            return key ? key : null;
        }
        return null;
    };
};
/**
 * A helper function to get the key from an element
 */
const keyElemExtractor = (element) => {
    const elem = element.closest('[data-oj-menu-item-key]');
    if (elem) {
        const key = getKey(elem);
        return key === undefined ? null : { key, elem };
    }
    return null;
};
/**
 * Function to get the corresponding keyboard behavior to close the menu.
 */
const getKeyboardCloseProps = (onClose, menuLevel) => {
    const onKeyDown = (event) => {
        if (event.key === 'Tab') {
            event.stopPropagation();
            onClose?.({ reason: 'dismissed' });
            event.preventDefault(); //TODO: Revisit this since we should let focus be at its natural flow, quick solution for now to focus correct item in main use case
        }
        else if (event.key === 'Escape' && menuLevel === 'main') {
            event.stopPropagation();
            onClose?.({ reason: 'dismissed' });
        }
    };
    return { onKeyDown };
};
/**
 * Function to get the corresponding specific behavior for closing submenu via keyboard
 */
const getCloseSubmenuEvent = (direction, closeSubmenu) => {
    if (!closeSubmenu) {
        return {};
    }
    const onKeyDown = (e) => {
        if (e.key === 'Escape' ||
            (direction === 'ltr' && e.key === 'ArrowLeft') ||
            (direction === 'rtl' && e.key === 'ArrowRight')) {
            e.stopPropagation();
            closeSubmenu?.();
        }
    };
    return { onKeyDown };
};
/**
 * Function to get the corresponding specific behavior for opening submenu via keyboard
 */
const getOpenSubmenuEvent = (direction, openSubmenu) => {
    if (!openSubmenu) {
        return {};
    }
    const onKeyDown = (e) => {
        if (direction === 'ltr') {
            if (e.key === 'ArrowRight') {
                openSubmenu?.();
            }
        }
        else {
            if (e.key === 'ArrowLeft') {
                openSubmenu?.();
            }
        }
    };
    return { onKeyDown };
};
/**
 * Functions to get type of device
 */
const getIsMobile = () => {
    const { deviceRenderMode } = getClientHints();
    return deviceRenderMode === 'phone';
};
const getIsIosTablet = () => {
    const { deviceRenderMode, platform } = getClientHints();
    return platform === 'ios' && deviceRenderMode === 'tablet';
};
/**
 * Get blur and focus events
 */
const getMenuContainerFocusMethods = (blurFunction, focusFunction) => {
    const onBlur = () => {
        setTimeout(() => {
            blurFunction();
        }, 100);
    };
    const onFocus = () => {
        focusFunction();
    };
    return { onBlur, onFocus };
};
/**
 * Set currentKey based on clicking menuItem
 */
const getMenuItemClickChangeKey = (key, changeKeyMethod) => {
    const onClick = () => {
        changeKeyMethod({ key, reason: 'click' });
    };
    return { onClick };
};
const getMenuContainerMouseLeave = (mouseLeaveFunction) => {
    const onMouseLeave = (e) => {
        mouseLeaveFunction(e);
    };
    return { onMouseLeave };
};
/**
 * Function to get the corresponding specific behavior for clicking/touching down menuContainer
 */
const getMenuPointerDown = () => {
    const onPointerDown = (event) => {
        //We just want to prevent default when element that was
        //pressed down was not a menu item.
        const keyElem = keyElemExtractor(event.target);
        if (!keyElem?.key) {
            event.preventDefault();
        }
    };
    return { onPointerDown };
};
const isMenuChildrenFalsy = (children) => {
    return !children || (Array.isArray(children) && !children.length);
};

const DEFAULT_PLACEMENT = 'bottom-start';
const DEFAULT_OFFSET = { mainAxis: 4, crossAxis: 0 };
const DEFAULT_SIZE_OPTIONS = { isMaxWidthAdjusted: true, isMaxHeightAdjusted: true };
const DEFAULT_FLIP_OPTIONS = { mainAxis: true, crossAxis: true };
const animationStates = {
    opening: (node) => {
        return mergeAnimationStateObjects([
            SCALE_UP_FULL,
            {
                from: { transformOrigin: getTransformOrigin(node.placement, node.direction) },
                to: { transformOrigin: getTransformOrigin(node.placement, node.direction) }
            }
        ], {
            duration: DURATION_LARGE
        });
    }
};
//TODO: Add animation to dropdown. JET-62628
//TODO: Investigate best approach to have different virtual padding depending on the position and implement it.JET-57950
function Dropdown({ isOpen = false, anchorRef, children, id, onClose, minWidth, placement = DEFAULT_PLACEMENT, offset = DEFAULT_OFFSET, onPosition, flipOptions = DEFAULT_FLIP_OPTIONS, sizeOptions = DEFAULT_SIZE_OPTIONS, shiftOptions, initialFocus = 'none' }) {
    const dropdownRef = useRef(null);
    const onTransitionEnd = useCallback((isOpen) => {
        if (!isOpen) {
            return;
        }
        const dropdownElement = dropdownRef.current;
        if (!dropdownElement) {
            return;
        }
        // honor firstFocusable initial focus behavior
        if (initialFocus === 'firstFocusable') {
            focusWithin(dropdownElement);
            return;
        }
        // JET-76442 + JET-76109: When VoiceOver is enabled on iOS tablets, we want the first menu item to receive focus.
        // This is to ensure that VoiceOver announces the menu items correctly and focus does not go to the next tabbable element.
        if (getIsIosTablet()) {
            const firstItem = dropdownElement.querySelector('[data-oj-menu-item-key]');
            if (firstItem) {
                focusWithin(firstItem);
            }
        }
    }, [initialFocus]);
    const { status, setAnimationElementRef, onPosition: onPositionAnimated } = usePopupAnimation({
        isOpen,
        placement,
        isAnimatedOnMount: true,
        animationStates,
        onPosition,
        onTransitionEnd
    });
    const handleMousedownClose = useCallback((event) => {
        onClose?.({ reason: 'outsideMousedown', target: event.target });
    }, [onClose]);
    const handleKeyDown = useCallback((event) => {
        if (event.defaultPrevented) {
            return;
        }
        switch (event.code) {
            case KEYS.ESC:
                onClose?.({ reason: 'dismissed' });
                break;
            case KEYS.TAB:
                onClose?.({ reason: 'dismissed' });
                break;
        }
    }, [onClose]);
    // use the mousedown event to be consistent with legacy JET
    useOutsideMousedown({
        isDisabled: false,
        ref: [dropdownRef],
        handler: handleMousedownClose
    });
    // useEffectEvent ensures we always access the latest anchorRef.current value inside effects,
    // without having to include anchorRef in the dependency array and the eslint-disable-next comment
    const STABLE_getAnchor = useEffectEvent(() => anchorRef?.current);
    // Restore focus to anchor after Dropdown is closed
    useEffect(() => {
        const anchorElement = STABLE_getAnchor();
        if (!isOpen &&
            status === 'unmounted' &&
            anchorElement &&
            anchorElement instanceof HTMLElement) {
            if (document.body.contains(anchorElement)) {
                anchorElement.focus();
            }
        }
    }, [STABLE_getAnchor, isOpen, status]);
    return (jsx(Fragment, { children: status !== 'unmounted' ? (jsx(Layer, { logicalParentRef: toLogicalParentRef(anchorRef), children: jsx(Floating, { anchorRef: anchorRef, ref: setAnimationElementRef, placement: placement, flipOptions: flipOptions, shiftOptions: shiftOptions, sizeOptions: sizeOptions, offsetValue: offset, onPosition: onPositionAnimated, boundaryOptions: { padding: { top: 8, bottom: 8, left: 0, right: 0 } }, children: jsx("div", { ref: dropdownRef, id: id, class: dropdownStyles.base, style: { minWidth: `${minWidth || 0}px` }, onKeyDown: handleKeyDown, children: jsx(CollectionInteractionContext.Provider, { value: 'embedded', children: children }) }) }) })) : null }));
}
function toLogicalParentRef(anchorRef) {
    const current = anchorRef.current;
    if (current) {
        if (current instanceof Element) {
            return anchorRef;
        }
        else if (current[UNSAFE_DOM_ACCESS]) {
            return anchorRef;
        }
    }
    return null;
}

export { Dropdown as D, getNVisible as a, getMenuPointerDown as b, getMenuContainerMouseLeave as c, getMenuContainerFocusMethods as d, getCloseSubmenuEvent as e, getKeyboardCloseProps as f, getPrevNext as g, getLastVisible as h, getIsMobile as i, isMenuChildrenFalsy as j, keyElemExtractor as k, getMenuItemClickChangeKey as l, getOpenSubmenuEvent as m };
