// @oracle/oraclejet-preact/es/GroupLoadingIndicator-19c6545a.js
/* @oracle/oraclejet-preact: undefined */
import { jsx, jsxs } from 'preact/jsx-runtime';
import { useState, useRef, useCallback, useEffect } from 'preact/hooks';
import { c as classNames } from './classNames-4e12b00d.js';
import { c as containsKey } from './keys-49c92dbb.js';
import { u as useId } from './useId-03dbfdf0.js';
import { T as TabbableModeContext } from './TabbableModeContext-7d8ad946.js';
import { u as useComponentTheme } from './useComponentTheme-d2f9e47f.js';
import { u as useInteractionStyle } from './useInteractionStyle-99e32c23.js';
import { m as mergeInterpolations } from './mergeInterpolations-30cd9a69.js';
import { f as flexitemInterpolations } from './flexitem-7b7f7920.js';
import { listGroupHeaderMultiVariantStyles } from './PRIVATE_List/themes/ListStyles.css.js';
import { GroupedListRedwoodTheme } from './UNSAFE_GroupedList/themes/redwood/GroupedListTheme.js';
import { F as Flex } from './Flex-03129963.js';
import { S as Skeleton } from './Skeleton-d6f0969e.js';
import { S as SkeletonContainer } from './List-4636ba9b.js';

// default style class for items in ListView
const GROUP_SELECTOR = '[data-oj-listview-group]';
const STICKY_SELECTOR = '[data-oj-listview-sticky]';
/**
 * The internal component used to render a group header in ExpandableList and GroupedList.
 */
function ListGroupHeader({ ariaDescribedBy, children, itemKey, itemIndex, itemDepth, isFocused, isFocusRingVisible, isActive, isGridlineVisible, isExpandable, expandedKeys, viewportConfig }) {
    const [isStuck, setStuck] = useState(false);
    const { classes: groupedListThemeClasses, styles: groupedListStyles } = useComponentTheme(GroupedListRedwoodTheme);
    const { interactionProps, applyHoverStyle, applyPseudoHoverStyle, applyActiveStyle } = useInteractionStyle();
    const variantClasses = listGroupHeaderMultiVariantStyles({
        expandable: isExpandable ? 'isExpandable' : 'notExpandable',
        needsEventsHover: applyHoverStyle ? 'isNeedsEventsHover' : 'notNeedsEventsHover',
        pseudoHover: applyPseudoHoverStyle ? 'isPseudoHover' : 'notPseudoHover',
        active: applyActiveStyle || isActive ? 'isActive' : 'notActive',
        focusRingVisible: isFocusRingVisible && isFocused ? 'isFocusRingVisible' : 'notFocusRingVisible',
        gridlineTop: itemIndex > 0 ? 'visible' : 'hidden',
        gridlineBottom: isGridlineVisible ? 'visible' : 'hidden'
    });
    const baseClasses = [variantClasses];
    if (isStuck) {
        baseClasses.push(groupedListThemeClasses);
        baseClasses.push(groupedListStyles.stuckHeader);
    }
    const classes = classNames(baseClasses);
    const styleInterpolations = mergeInterpolations([
        ...Object.values(flexitemInterpolations)
    ]);
    const { class: cls, ...cellStyles } = styleInterpolations({
        alignSelf: 'center',
        flex: '1 0 auto'
    });
    // create these refs up front to avoid having to update ref.current values directly during render
    // function or setup a useEffect to update if options/values change in a subsequent render.
    const expandableScrollerRef = useRef(null);
    // rely on provided scrollerRef, not to body on first render.
    const scrollerRef = isExpandable ? expandableScrollerRef : viewportConfig?.scrollerRef ?? null;
    const headerRef = useRef(null);
    // compute stuck state based on scrollTop and element positions; avoids IntersectionObserver edge cases
    const onScroll = useCallback(() => {
        const scroller = scrollerRef ? scrollerRef.current : null;
        const headerEl = headerRef.current;
        if (!scroller || !headerEl)
            return;
        let isStuck = false;
        const nextEl = headerEl.nextElementSibling;
        if (nextEl) {
            const headerRect = headerEl.getBoundingClientRect();
            // apply sticky styling if next element has started to scroll under this header
            isStuck = headerRect.bottom > nextEl.getBoundingClientRect().top;
            if (!isStuck) {
                let prevEl = headerEl.previousElementSibling;
                // check for this header's top value matching the previous header's top value - the case where
                // this header has replaced the previous header, but the next item has not yet scrolled under it
                while (prevEl) {
                    if (prevEl.hasAttribute('data-oj-listview-group')) {
                        isStuck = prevEl.getBoundingClientRect().top === headerRect.top;
                        break;
                    }
                    prevEl = prevEl.previousElementSibling;
                }
            }
        }
        setStuck(isStuck);
    }, [scrollerRef]);
    useEffect(() => {
        const scroller = scrollerRef ? scrollerRef.current : null;
        if (!scroller)
            return;
        onScroll();
        scroller.addEventListener('scroll', onScroll);
        return () => {
            scroller.removeEventListener('scroll', onScroll);
        };
    }, [scrollerRef, onScroll]);
    return (jsx("div", { ref: headerRef, id: useId(), role: "row", "aria-rowindex": itemIndex + 1, "aria-level": isNaN(itemDepth) ? undefined : itemDepth + 1, class: classes, "data-oj-key": itemKey, "data-oj-list-item": "", "data-oj-listview-group": "", ...(isStuck && { 'data-oj-listview-sticky': '' }), ...(typeof itemKey === 'number' && { 'data-oj-key-type': 'number' }), ...interactionProps, children: jsx("div", { id: useId(), role: "gridcell", "aria-describedby": ariaDescribedBy, "aria-colindex": 1, "aria-expanded": containsKey(expandedKeys, itemKey), style: cellStyles, children: jsx(TabbableModeContext.Provider, { value: { isTabbable: false }, children: children }) }) }));
}

const GroupLoadingIndicator = () => (jsx(SkeletonContainer, { minimumCount: 3, children: (index) => {
        const padding = { paddingLeft: index * 10 + 'px' };
        return (jsxs(Flex, { height: "12x", align: "center", children: [jsx("span", { style: padding }), jsx(Skeleton, { height: "4x" })] }));
    } }));

export { GROUP_SELECTOR as G, ListGroupHeader as L, STICKY_SELECTOR as S, GroupLoadingIndicator as a };
