// @oracle/oraclejet-preact/es/InputDateMask-ec27aee7.js
/* @oracle/oraclejet-preact: undefined */
import { jsx, jsxs } from 'preact/jsx-runtime';
import { forwardRef } from 'preact/compat';
import { useRef, useCallback, useEffect, useMemo, useContext, useReducer, useImperativeHandle } from 'preact/hooks';
import { L as Label } from './Label-5d1ae8d7.js';
import { a as isSelectAll, b as isBackspaceOrDelete } from './keyboardUtils-986fe923.js';
import { R as ReadonlyTextField } from './StyledTextField-7605d76a.js';
import './ObfuscatedTextFieldInputStyles.styles.css';
import { R as ReadonlyTextFieldInput } from './ReadonlyTextFieldInput-106ce959.js';
import { T as TextField } from './TextField-7884fcd7.js';
import { u as useTranslationBundle } from './useTranslationBundle-4822502f.js';
import './LiveRegionStyles.styles.css';
import './LayerHost-7931b424.js';
import './UNSAFE_TextField/themes/TextFieldStyles.css.js';
import './SkeletonStyles.styles.css';
import './TextFieldLoadingStyles.styles.css';
import './TextFieldStyles.styles.css';
import './UNSAFE_TextField/themes/redwood/TextFieldBaseTheme.styles.css';
import './UNSAFE_TextField/themes/redwood/TextFieldVariants.css.js';
import './logger-c92f309c.js';
import './TabbableModeContext-7d8ad946.js';
import './UserAssistanceStyles.styles.css';
import './UNSAFE_UserAssistance/themes/redwood/UserAssistanceBaseTheme.styles.css';
import './UNSAFE_UserAssistance/themes/redwood/UserAssistanceVariants.css.js';
import { a as InlineUserAssistance } from './InlineUserAssistance-890293dd.js';
import { C as CompactUserAssistance } from './CompactUserAssistance-6aa8894b.js';
import { m as mergeProps } from './mergeProps-88ea8306.js';
import './TooltipContentStyles.styles.css';
import './hooks/UNSAFE_useTooltip/themes/redwood/TooltipContentBaseTheme.styles.css';
import './hooks/UNSAFE_useTooltip/themes/redwood/TooltipContentVariants.css.js';
import './tooltipUtils-af0abd38.js';
import { createContext } from 'preact';
import './IconStyle.styles.css';
import './Popup-992b6c27.js';
import './LabelStyles.styles.css';
import './UNSAFE_Label/themes/redwood/LabelBaseTheme.styles.css';
import './UNSAFE_Label/themes/redwood/LabelVariants.css.js';
import './InputGroupContext-20bdbd6e.js';
import './HiddenAccessibleStyles.styles.css';
import './UNSAFE_Popup/themes/redwood/PopupBaseTheme.styles.css';
import './UNSAFE_Popup/themes/redwood/PopupVariants.css.js';
import { u as useFocusableTextField } from './useFocusableTextField-8fe41cd4.js';
import './FormContext-55fb8000.js';
import { u as useFormContext } from './useFormContext-bb9dc478.js';
import { F as FormFieldContext } from './FormFieldContext-47ca0c20.js';
import { u as useTextField } from './useTextField-eb9db737.js';
import { u as useUser } from './useUser-06d34d19.js';
import { b as isLeapYear } from './getFormatParse-6d6fdd4d.js';
import { n as getDaysInMonth, r as getMonthsInYear, M as MIN_ISO_YEAR, s as MAX_ISO_YEAR, u as MIN_ISO_MONTH, v as MIN_ISO_DAY, f as formatFullCalendarDate, w as getDatePlaceholderMasks, x as formatShortCalendarDate } from './calendarDateUtils-30611c35.js';
import { i as isNumberOnlyString, s as stringToNumber, n as numberToString, w as withinRange, a as getTodayAsCalendarDate, g as getMonthName, t as translateCustomMasks, c as calendarDatesAreDifferent } from './maskUtils-68d5a6b3.js';
import { l } from './stringUtils-0fe25200.js';
import { u as useComponentTheme } from './useComponentTheme-d2f9e47f.js';
import { u as useFormFieldContext } from './useFormFieldContext-daf7a676.js';
import { F as Flex } from './Flex-03129963.js';
import { H as HiddenAccessible } from './HiddenAccessible-61197e7c.js';
import { L as LiveRegion } from './LiveRegion-cebac18d.js';
import { TextFieldInputRedwoodTheme } from './UNSAFE_TextField/themes/redwood/TextFieldInputTheme.js';
import { i as isEmbeddedVariant } from './formControlUtils-84e2af0e.js';
import { c as classNames } from './classNames-4e12b00d.js';
import { g as getClientHints } from './clientHints-a42b510c.js';
import { segmentStyles } from './UNSAFE_InputDateMask/themes/SegmentStyles.css.js';
import { u as useDoubleTap } from './useDoubleTap-4e290707.js';
import { u as useTabbableMode } from './useTabbableMode-090ed4b5.js';

// Used to disable browser cut and paste, which don't really work with segments (copy is fine).
const preventDefaultHandler = (e) => {
    e.preventDefault();
};
const isIos = () => getClientHints().platform === 'ios';
/**
 * A DateSegment renders a part of a calendar date, e.g. the day, month, or year.
 * It can be used as a spinbutton to step or spin its value.
 */
const DateSegment = ({ 'aria-label': ariaLabel, 'aria-valuemax': ariaValueMax, 'aria-valuemin': ariaValueMin, 'aria-valuenow': ariaValueNow, 'aria-valuetext': ariaValueText, inputRef, isComplete, isDisabled, isFieldFocused, isHidden, isHighlighted, isInvalid, isRequired, isSelected, onChange, orderedSegments, placeholder, text, type }) => {
    const hasDisplayValue = text && text.length > 0;
    const rootClasses = classNames([
        segmentStyles.base,
        isSelected && segmentStyles.selected,
        isHighlighted && segmentStyles.highlighted,
        isFieldFocused && segmentStyles.accessibleTouchTarget
    ]);
    const placeholderClasses = classNames([
        segmentStyles.placeholder,
        isHidden || hasDisplayValue ? segmentStyles.hidden : segmentStyles.notHidden
    ]);
    const translations = useTranslationBundle('@oracle/oraclejet-preact');
    const emptySegmentValueText = translations.inputDateMask_empty_segment();
    // If a ref was passed in, use it, otherwise get a ref.
    const sRef = useRef(null);
    const segmentRef = inputRef ?? sRef;
    const hasLeadingZero = useRef(false);
    const { locale } = useUser();
    const onKeyDown = useCallback((event) => {
        if (isDisabled ||
            event.key === 'Tab' ||
            event.key === 'ArrowLeft' ||
            event.key === 'ArrowRight' ||
            event.key === 'Enter' ||
            // JET-73224: let Escape pass through in order to close a popup or dialog.
            event.key === 'Escape' ||
            isSelectAll(event)) {
            // Let the event pass through.
            return;
        }
        if (isHighlighted) {
            // If date is selected, ignore edits and let backspace/delete pass through to clear.
            if (isBackspaceOrDelete(event) || event.key === 'Backspace') {
                return;
            }
        }
        else if (event.key === '0' && (isComplete || text === undefined)) {
            // Typing '0' is considered a leading zero if the segment is complete, in which case
            // the user is typing a new value, or if there is no current value.
            hasLeadingZero.current = true;
        }
        else if (isNumberOnlyString(event.key)) {
            onChange?.({
                actionType: 'updateSegment',
                data: {
                    type,
                    userInput: event.key,
                    hasLeadingZero: hasLeadingZero.current,
                    locale,
                    orderedSegments,
                    existingText: text,
                    min: ariaValueMin,
                    max: ariaValueMax
                }
            });
            hasLeadingZero.current = false;
        }
        else if (isBackspaceOrDelete(event) || event.key === 'Backspace') {
            // Android sends Backspace as event.key, not event.code, so check for it.
            onChange?.({ actionType: 'clearSegment', data: { type, orderedSegments } });
            hasLeadingZero.current = false;
        }
        else if (event.key === 'ArrowUp') {
            onChange?.({
                actionType: 'step',
                data: {
                    direction: 'increase',
                    locale,
                    type,
                    orderedSegments,
                    min: ariaValueMin,
                    max: ariaValueMax
                }
            });
            hasLeadingZero.current = false;
        }
        else if (event.key === 'ArrowDown') {
            onChange?.({
                actionType: 'step',
                data: {
                    direction: 'decrease',
                    locale,
                    type,
                    orderedSegments,
                    min: ariaValueMin,
                    max: ariaValueMax
                }
            });
            hasLeadingZero.current = false;
        }
        else if (event.key === 'PageUp') {
            onChange?.({
                actionType: 'page',
                data: {
                    direction: 'increase',
                    locale,
                    type,
                    orderedSegments,
                    min: ariaValueMin,
                    max: ariaValueMax
                }
            });
            hasLeadingZero.current = false;
        }
        else if (event.key === 'PageDown') {
            onChange?.({
                actionType: 'page',
                data: {
                    direction: 'decrease',
                    locale,
                    type,
                    orderedSegments,
                    min: ariaValueMin,
                    max: ariaValueMax
                }
            });
            hasLeadingZero.current = false;
        }
        else if (event.key === 'Home') {
            onChange?.({
                actionType: 'goToHome',
                data: { locale, type, orderedSegments, min: ariaValueMin }
            });
            hasLeadingZero.current = false;
        }
        else if (event.key === 'End') {
            onChange?.({
                actionType: 'goToEnd',
                data: { locale, type, orderedSegments, max: ariaValueMax }
            });
            hasLeadingZero.current = false;
        }
        event.preventDefault();
        event.stopPropagation();
    }, [
        ariaValueMax,
        ariaValueMin,
        isComplete,
        isDisabled,
        isHighlighted,
        locale,
        onChange,
        orderedSegments,
        text,
        type
    ]);
    // Used to focus the segment when isSelected is true.
    useEffect(() => {
        if (isSelected) {
            segmentRef.current?.focus();
        }
    }, [isSelected, segmentRef]);
    // This attribute is used to distinguish between segments when clicked (see DateField.tsx).
    const customDataAttribute = { 'data-segment': [type] };
    // If ariaValueText is undefined, substitute a string that indicates the segment is empty.
    // If ariaValueNow and ariaValueText represent the same value, omit the value text; this
    // is the same as what we do in NumberInputText.
    const valueText = ariaValueText === undefined
        ? emptySegmentValueText
        : ariaValueNow?.toString() === ariaValueText
            ? undefined
            : ariaValueText;
    return (
    // Introduce a div wrapper around the contentEditable element to prevent it from grabbing focus.
    // https://stackoverflow.com/questions/34354085/clicking-outside-a-contenteditable-div-stills-give-focus-to-it
    // JET-64874: We change the segment's role to "textbox" on iOS to work around a VoiceOver bug, and when we do that
    // we need to add aria-multiline={false}
    jsx("div", { children: jsxs("div", { ref: segmentRef, role: isIos() ? 'textbox' : 'spinbutton', "aria-disabled": isDisabled, "aria-invalid": isInvalid || undefined, "aria-label": ariaLabel, "aria-multiline": isIos() ? false : undefined, "aria-required": isRequired, "aria-valuemin": isIos() ? undefined : ariaValueMin, "aria-valuemax": isIos() ? undefined : ariaValueMax, "aria-valuenow": isIos() ? undefined : ariaValueNow, "aria-valuetext": isIos() ? undefined : valueText, autocapitalize: "off", autocorrect: "off", class: rootClasses, contentEditable: !isDisabled, ...customDataAttribute, enterkeyhint: isDisabled ? undefined : 'done', inputMode: isDisabled ? 'none' : 'numeric', onCut: preventDefaultHandler, onKeyDown: onKeyDown, onPaste: preventDefaultHandler, spellcheck: false, tabIndex: isSelected ? 0 : -1, children: [jsx("span", { "aria-hidden": "true", class: placeholderClasses, ...customDataAttribute, children: placeholder }), text] }) }));
};

/**
 * LiteralSegment is used to represent a literal separator in a calendar date,
 * such as the '/' in 11/29/2023.
 */
const LiteralSegment = ({ isHidden, isHighlighted, isPlaceholder, text }) => {
    const spanClasses = classNames([
        segmentStyles.literalBase,
        isPlaceholder && segmentStyles.placeholder,
        isHighlighted && segmentStyles.highlighted,
        isHidden ? segmentStyles.hidden : segmentStyles.notHidden
    ]);
    return (jsx("span", { "aria-hidden": "true", class: spanClasses, children: text }));
};

/**
 * @license
 * Copyright (c) %FIRST_YEAR% %CURRENT_YEAR%, Oracle and/or its affiliates.
 * The Universal Permissive License (UPL), Version 1.0
 * as shown at https://oss.oracle.com/licenses/upl/
 * @ignore
 */
/**
 * Reducer function for updating DateFieldState per the specified action.
 * 'aria-valuenow' is the segment's value. In the dateReducer we look at the 'aria-valuenow'
 * for each segment and when a user updates one segment, that new value may affect the others.
 * In the dateReducer we have a currentDate state that is calculated at the end of the dispatch action
 * - it takes each segment's 'aria-valuenow' and creates the currentDate from it.
 * @param state
 * @param action
 */
const dateReducer = (state, action) => {
    // On mount, the reducer is initialized with initial state derived from the DateField's value.
    // If that value is changed programmatically afterwards, the only way to push that update
    // into the reducer is to define a "reset" action.
    if (action.actionType === 'reset') {
        return action.data;
    }
    switch (action.actionType) {
        case 'clearDate': {
            const updatedSegments = selectFirst(clearDate(state.segments), action.data.orderedSegments);
            const newState = createDateFieldStateWithSelectedDate(updatedSegments, false);
            return newState;
        }
        case 'clearSegment': {
            const updatedSegment = clearSegment(findSegment$1(state.segments, action.data.type));
            const newState = createDateFieldState(state, action.data.orderedSegments, action.data.type, updatedSegment);
            return newState;
        }
        case 'commitOnBlur': {
            const updatedSegments = checkFeb29EdgeCase(state.segments);
            const updatedSegmentsClearedAndComplete = clearSelectionAndComplete(updatedSegments);
            const newState = createDateFieldStateWithSelectedDate(updatedSegmentsClearedAndComplete, false);
            return newState;
        }
        case 'commitOnEnter': {
            const updatedSegments = checkFeb29EdgeCase(state.segments);
            const newState = createDateFieldStateWithSelectedDate(updatedSegments, false);
            return newState;
        }
        case 'goToEnd': {
            const segmentState = findSegment$1(state.segments, action.data.type);
            const max = action.data.max;
            const updatedSegment = goToEnd(action.data.type, segmentState, max);
            const newState = createDateFieldState(state, action.data.orderedSegments, action.data.type, updatedSegment, false);
            return newState;
        }
        case 'goToHome': {
            const segmentState = findSegment$1(state.segments, action.data.type);
            const min = action.data.min;
            const updatedSegment = goToHome(action.data.type, segmentState, min);
            const newState = createDateFieldState(state, action.data.orderedSegments, action.data.type, updatedSegment, false);
            return newState;
        }
        case 'page': {
            const direction = action.data.direction;
            const segmentState = findSegment$1(state.segments, action.data.type);
            const min = action.data.min;
            const max = action.data.max;
            const updatedSegment = doStepOrPage(action.data.type, segmentState, 'page', direction, min, max);
            const newState = createDateFieldState(state, action.data.orderedSegments, action.data.type, updatedSegment, false);
            return newState;
        }
        case 'selectDate': {
            const newState = selectDate(state);
            return newState;
        }
        case 'selectFirst': {
            const updatedSegments = selectFirst(state.segments, action.data.orderedSegments);
            const newState = createDateFieldStateWithSelectedDate(updatedSegments, false);
            return newState;
        }
        case 'selectNext': {
            const updatedSegments = selectNext(state.segments, action.data.orderedSegments);
            const newState = createDateFieldStateWithSelectedDate(updatedSegments, false);
            return newState;
        }
        case 'selectPrevious': {
            const updatedSegments = selectPrevious(state.segments, action.data.orderedSegments);
            const newState = createDateFieldStateWithSelectedDate(updatedSegments, false);
            return newState;
        }
        case 'selectSegment': {
            const updatedSegments = selectSegment(state.segments, action.data.type);
            const newState = createDateFieldStateWithSelectedDate(updatedSegments, false);
            return newState;
        }
        case 'step': {
            const direction = action.data.direction;
            const segmentState = findSegment$1(state.segments, action.data.type);
            const min = action.data.min;
            const max = action.data.max;
            const updatedSegment = doStepOrPage(action.data.type, segmentState, 'step', direction, min, max);
            const newState = createDateFieldState(state, action.data.orderedSegments, action.data.type, updatedSegment, false);
            return newState;
        }
        case 'updateSegment': {
            const segmentType = action.data.type;
            const segment = findSegment$1(state.segments, segmentType);
            const minValue = action.data.min;
            const maxValue = action.data.max;
            const userInput = action.data.userInput;
            const hasLeadingZero = action.data.hasLeadingZero;
            const existingText = action.data.existingText;
            const updatedSegment = updateSegment(segmentType, segment, userInput, hasLeadingZero, maxValue, minValue, existingText);
            // This passes in the updatedSegment and the rest of the state and figures out if it needs to update other segments.
            // This changes leap year days, for example. Then it takes the updated segments and creates a date field state from it.
            const newState = createDateFieldState(state, action.data.orderedSegments, action.data.type, updatedSegment);
            return newState;
        }
        default: {
            throw new Error('Unknown action type');
        }
    }
};
const findSegment$1 = (segments, type) => {
    return segments.find((segment) => segment.type === type);
};
// A segment is considered complete if appending a zero to the current value causes it
// to exceed the max. For instance, if you type 2-9 into month or 4-9 into day, the
// segment is complete because you can't type any more digits and have it be a valid value.
const isSegmentComplete = (val, max) => {
    const num = stringToNumber(numberToString(val) + '0');
    return num > max;
};
// A date is considered a partial date if at least one segment has a value.
const isPartialDate = (allSegments) => {
    return allSegments.some((seg) => seg['aria-valuenow'] !== undefined);
};
// Construct a calendar date that reflects the values of each segment.
// If none of the segments have a value, return undefined.
const createCalendarDateFromSegments = (segments) => {
    const hasAnySegmentWithAValue = isPartialDate(segments);
    if (!hasAnySegmentWithAValue) {
        return undefined;
    }
    // This returns something that looks like [ {month: 12}, {day: 25}, {year: 2020} ].
    // Note: we use the aria-valuenow to get the current value as a number (not string).
    const segmentValues = segments.map((seg) => {
        return { [seg.type]: seg['aria-valuenow'] };
    });
    // Iterate over segmentValues and spread each value to get {month: 12, day:25, year: 2020}.
    const iterator = segmentValues.values();
    let cd = {};
    for (const segValue of iterator) {
        cd = { ...cd, ...segValue };
    }
    return cd;
};
// Find the current value for year.
const getYearValue = (segments) => {
    const yearState = findSegment$1(segments, 'year');
    return yearState['aria-valuenow'];
};
// Find the current value for month.
const getMonthValue = (segments) => {
    const monthState = findSegment$1(segments, 'month');
    return monthState['aria-valuenow'];
};
// Replace the updated segment and return a new DateSegmentState[].
const replaceDateSegment = (existingSegments, updatedSegmentType, updatedSegment) => {
    return existingSegments.map((seg) => (seg.type === updatedSegmentType ? updatedSegment : seg));
};
// Return a new DateSegmentState[] that reflects changes across all segments.
// When you change the month this can affect the day, for instance.
const createDateSegmentStates = (existingSegments, orderedSegments, updatedSegmentType, updatedSegment, autoAdvance) => {
    const modifiedSegments = replaceDateSegment(existingSegments, updatedSegmentType, updatedSegment);
    let updatedSegments = modifiedSegments;
    if (updatedSegmentType === 'month' || updatedSegmentType === 'year') {
        // If month was changed, update the day as needed; this includes adjusting the
        // day value if it now exceeds the max number of days in the month.
        // If year was changed, that can also affect the max number of days (if month is February).
        const daySeg = findSegment$1(modifiedSegments, 'day');
        if (daySeg) {
            const monthValue = updatedSegmentType == 'month'
                ? updatedSegment['aria-valuenow']
                : getMonthValue(modifiedSegments);
            const yearValue = updatedSegmentType === 'year'
                ? updatedSegment['aria-valuenow']
                : getYearValue(modifiedSegments);
            // We will leave dayMax for Feb to 29 until the year is a 4-digit year when the user is typing.
            // JET-75124 Leap year date of 02/29 is converted to 28 automatically (as soon as you type the first number of the year)
            const isFebAndLessThanFourDigitYear = monthValue && monthValue === 2 && yearValue && yearValue < 1000;
            // autoAdvance is true when you type and it is false when you spin. If the user is typing into the year,
            // then we want max day to be 29.
            const dayMax = isFebAndLessThanFourDigitYear && updatedSegmentType === 'year' && autoAdvance
                ? 29
                : getDaysInMonth(monthValue ?? '*', yearValue ?? '*');
            const currentDay = daySeg['aria-valuenow'];
            const dayValue = !!currentDay && currentDay > dayMax ? dayMax : currentDay;
            const updatedDaySeg = {
                ...daySeg,
                'aria-valuenow': dayValue
            };
            updatedSegments = replaceDateSegment(modifiedSegments, 'day', updatedDaySeg);
        }
    }
    // If the updatedSegment is complete, automatically select the next segment.
    updatedSegments =
        updatedSegment.isComplete && autoAdvance
            ? selectNext(updatedSegments, orderedSegments)
            : updatedSegments;
    return updatedSegments;
};
// Construct a new DateFieldState given an updated segment, which may affect other segments.
const createDateFieldState = (currentState, orderedSegments, updatedSegmentType, updatedSegment, autoAdvance = true) => {
    const updatedSegments = createDateSegmentStates(currentState.segments, orderedSegments, updatedSegmentType, updatedSegment, autoAdvance);
    return createDateFieldStateWithSelectedDate(updatedSegments, currentState.isDateSelected);
};
// Return a new copy of the segment with the value cleared.
const clearSegment = (segmentState) => {
    return {
        ...segmentState,
        'aria-valuenow': undefined,
        isComplete: false
    };
};
// Return a new copy of segments with all values cleared.
const clearDate = (segments) => {
    return segments.map((segmentState) => {
        const cleared = clearSegment(segmentState);
        return { ...cleared };
    });
};
// Return a new copy of the segment with updated values.
const updateSegment = (segmentType, segmentState, userInput, hasLeadingZero, maxValue, minValue, existingText) => {
    const wasComplete = segmentState.isComplete;
    // If the segment was already complete, replace any existing text with userInput.
    // Otherwise append userInput onto existing text to get the requested string.
    // For instance if the existing text is '1' and userInput (i.e. that the user typed)
    // is '2' then requestedValStr is '12'.
    const requestedValStr = wasComplete || existingText === undefined ? userInput : existingText + userInput;
    // If the requested value is out of range, discard the existing text/value and only
    // keep the new one. Check if the segment is complete.
    const requestedVal = stringToNumber(requestedValStr);
    const inRange = withinRange(requestedVal, minValue, maxValue);
    const newVal = inRange ? requestedVal : stringToNumber(userInput);
    const isComplete = isSegmentComplete(newVal, maxValue) || (!!hasLeadingZero && segmentType !== 'year');
    // Create a new segment state that reflects the updated value.
    return {
        ...segmentState,
        'aria-valuenow': newVal,
        isComplete
    };
};
// Return a new DateSegmentState[] where the specified segment is selected, all others cleared.
const selectSegment = (segments, segmentType) => {
    const updatedSegments = checkFeb29EdgeCase(segments);
    return updatedSegments.map((seg) => {
        return { ...seg, isSelected: seg.type === segmentType ? true : false };
    });
};
// Return a new DateSegmentState[] where the first segment is selected.
const selectFirst = (segments, orderedSegments) => {
    return selectSegment(segments, orderedSegments[0]);
};
// Find the index of the selected segment in the ordered list; return -1 if there is no selection.
const getSelectedSegmentIndex = (segments, orderedSegments) => {
    const sel = segments.find((seg) => seg.isSelected === true);
    return sel === undefined ? -1 : orderedSegments.findIndex((type) => type === sel.type);
};
// Convenience function to mark the selected segment complete.
const markSelectedSegmentComplete = (segments) => {
    return segments.map((seg) => (seg.isSelected ? { ...seg, isComplete: true } : seg));
};
// Select the segment following the one that is currently selected.
const selectNext = (segments, orderedSegments) => {
    const selectedIndex = getSelectedSegmentIndex(segments, orderedSegments);
    const max = orderedSegments.length - 1;
    // If we're already at the last position, do nothing. Note if nothing was currently selected
    // then selectedIndex is -1. The first segment will be selected.
    return selectedIndex === max
        ? segments
        : selectSegment(markSelectedSegmentComplete(segments), orderedSegments[selectedIndex + 1]);
};
// Select the segment prior to the one that is currently selected.
const selectPrevious = (segments, orderedSegments) => {
    const selectedIndex = getSelectedSegmentIndex(segments, orderedSegments);
    return selectedIndex === 0
        ? segments
        : selectedIndex === -1
            ? selectFirst(segments, orderedSegments)
            : selectSegment(markSelectedSegmentComplete(segments), orderedSegments[selectedIndex - 1]);
};
// Return a new DateSegmentState[] where all segments are unselected and any values are
// marked complete. We want to do this when the user commits the DateField value (on blur).
const clearSelectionAndComplete = (segments) => {
    return segments.map((seg) => {
        return { ...seg, isSelected: false, isComplete: !!seg['aria-valuenow'] };
    });
};
// Return a new DateSegmentState[] where all segments are unselected.
const clearSelection = (segments) => {
    return segments.map((seg) => {
        return { ...seg, isSelected: false };
    });
};
// Return a new DateFieldState where the entire date is selected.
const selectDate = (currentState) => {
    return createDateFieldStateWithSelectedDate(clearSelection(currentState.segments), true);
};
// Create a new DateFieldState from the updatedSegments and isDateSelected params.
const createDateFieldStateWithSelectedDate = (updatedSegments, isDateSelected) => {
    return {
        currentDate: createCalendarDateFromSegments(updatedSegments),
        isDateSelected,
        segments: updatedSegments
    };
};
const YEAR_HOME_VALUE = 1900;
const YEAR_END_VALUE = 2100;
// If a segment is empty, use today's date as a starting point for any step or page action.
const getDefaultSegmentValue = (segmentType) => {
    const { year: yearDefault, month: monthDefault, day: dayDefault } = getTodayAsCalendarDate();
    const defaults = { year: yearDefault, month: monthDefault, day: dayDefault };
    return defaults[segmentType];
};
// Get the minimum value for the segment.
const getHomeSegmentValue = (segmentType, min) => {
    return segmentType === 'year' ? YEAR_HOME_VALUE : min;
};
// Get the maximum value for the segment (for year, this is YEAR_END_VALUE).
const getEndSegmentValue = (segmentType, max) => {
    return segmentType === 'year' ? YEAR_END_VALUE : max;
};
// Return the result of stepping currentValue in the requested direction, taking into account min and max.
const getSteppedSegmentValue = (currentValue, direction, min, max) => {
    const newValue = direction === 'increase'
        ? currentValue + 1 > max
            ? min
            : currentValue + 1
        : currentValue - 1 < min
            ? max
            : currentValue - 1;
    return newValue;
};
// Return the result of paging the value up or down, taking into account min and max.
const getPagedSegmentValue = (currentValue, direction, min, max, segmentType) => {
    const step = segmentType === 'month' ? 2 : segmentType === 'year' ? 5 : 7;
    const newValue = direction === 'increase'
        ? currentValue + step > max
            ? (currentValue + step) % max
            : currentValue + step
        : currentValue - step < min
            ? currentValue - step + max
            : currentValue - step;
    return newValue;
};
// Create a new segment with an updated value from a keyboard action such as step, page, Home, or End.
const createSegmentFromKeyboardAction = (newVal, segmentState) => {
    // Create a new segment state that reflects the updated value.
    // After any keyboard action, the value is complete so that if the user types anything,
    // it will replace the stepped value.
    return {
        ...segmentState,
        'aria-valuenow': newVal,
        isComplete: true
    };
};
// Set the segment to its minimum value.
const goToHome = (segmentType, segmentState, min) => {
    const newValue = getHomeSegmentValue(segmentType, min);
    return createSegmentFromKeyboardAction(newValue, segmentState);
};
// Set the segment to its maximum value.
const goToEnd = (segmentType, segmentState, max) => {
    const newValue = getEndSegmentValue(segmentType, max);
    return createSegmentFromKeyboardAction(newValue, segmentState);
};
// Step or page a segment's value.
const doStepOrPage = (segmentType, segmentState, type, direction, min, max) => {
    const valueNow = segmentState['aria-valuenow'];
    const newValue = valueNow === undefined
        ? getDefaultSegmentValue(segmentType)
        : type === 'page'
            ? getPagedSegmentValue(valueNow, direction, min, max, segmentType)
            : getSteppedSegmentValue(valueNow, direction, min, max);
    return createSegmentFromKeyboardAction(newValue, segmentState);
};
// This is for an edge case where the month is February, the day is 29, and the year is < 1000, and it is not a leap year.
// If so, we update the daySegment to 28. This function is called on 'commit'.
// JET-75124 Leap year date of 02/29 is converted to 28 automatically (as soon as you type the first number in the year)
const checkFeb29EdgeCase = (stateSegments) => {
    const monthSegment = stateSegments.find((seg) => seg.type === 'month');
    const daySegment = stateSegments.find((seg) => seg.type === 'day');
    const yearSegment = stateSegments.find((seg) => seg.type === 'year');
    const year = yearSegment?.['aria-valuenow'];
    if (year !== undefined &&
        year < 1000 &&
        !isLeapYear(year) &&
        monthSegment?.['aria-valuenow'] === 2 &&
        daySegment?.['aria-valuenow'] === 29) {
        const updatedDaySeg = {
            ...daySegment,
            'aria-valuenow': 28
        };
        return replaceDateSegment(stateSegments, 'day', updatedDaySeg);
    }
    return stateSegments;
};

// useInputFocusHandler.ts
/**
 * Hook that determines whether to dispatch a 'selectFirst' action when an input field receives focus.
 *
 * @param {Object} options
 * @param {Dispatch<A>} options.dispatch - Dispatch function for actions of type A that must contain {actionType: 'selectFirst'}.
 * @param {boolean} [options.isInputFocused] - Whether the input field is currently focused.
 * @param {EditableDateSegment[]} - the ordered segments
 * @returns {function} - An `onPointerDown` handler.
 */
const useInputDateMaskSelectFirstHandler = ({ dispatch, isInputFocused, orderedSegments }) => {
    // The following is a simplified version of how we detect keyboard vs. mouse focus
    // in the useCollectionFocusRing hook. If the user tabs onto the field, we want to
    // automatically select the first segment.
    const recentPointerRef = useRef(false);
    const pointerDownTimerRef = useRef();
    const onPointerDown = useCallback(() => {
        recentPointerRef.current = true;
        clearTimeout(pointerDownTimerRef.current);
        pointerDownTimerRef.current = setTimeout(() => {
            recentPointerRef.current = false;
        }, 200);
    }, []);
    const lastInputFocusedRef = useRef(false);
    // If we receive focus that is *not* due to the user clicking in the field, i.e. the user
    // tabbed onto the field, then select the first segment. (!recentPointerRef.current means the mouse wasn't used in the component).
    // Note: we cannot check isFocused because it includes the ? icon in compact UAD, and we don't
    // want the text selected when the help icon is focused.
    // Every call to the reducer is always producing a new object and then when we create a new object, this hook gets called again. There are libraries that would fix this.
    if (isInputFocused && !lastInputFocusedRef.current && !recentPointerRef.current) {
        const action = { actionType: 'selectFirst', data: { orderedSegments } };
        dispatch(action); // safe because A must include SelectFirstAction
    }
    // If the inputFocus has changed, then set recentPointerRef.current to false to not wait until the timeout, and update the lastInputFocusedRef.
    // Setting refs does not cause a re-render.
    if (isInputFocused !== lastInputFocusedRef.current) {
        recentPointerRef.current = false;
        lastInputFocusedRef.current = !!isInputFocused;
    }
    return onPointerDown;
};

/**
 * @license
 * Copyright (c) %FIRST_YEAR% %CURRENT_YEAR%, Oracle and/or its affiliates.
 * The Universal Permissive License (UPL), Version 1.0
 * as shown at https://oss.oracle.com/licenses/upl/
 * @ignore
 */
function isMobile() {
    const deviceType = getClientHints().deviceType;
    return deviceType === 'phone' || deviceType === 'tablet';
}
/**
 * A hook that creates various handlers for a DateField.
 * useDateFieldHandler handles user events like click, Ctrl+A/Delete, Arrow Right, Enter etc., and it will
 * dispatch actions based on those events.
 *
 * This hook creates a dateClearedRef. It is set to false by default and whenever the
 * value isn't undefined, including a partial value.
 * When the user's action is to clear the field, this hook dispatches the 'clearDate' action,
 * and sets dateClearedRef to true.
 * dateClearedRef is returned from the hook so that the aria-live DOM can be updated to say Date Cleared
 * or to clear this text.
 */
const useDateFieldHandlers = ({ hasSelectedSegment, value, direction, dispatch, groupRef, isCompleteDate, isDateSelected, isDisabled, isInputFocused, orderedSegments, state, onCommit, onInput }) => {
    const { tabbableModeProps } = useTabbableMode();
    const { tabIndex: tabbableModeIndex } = tabbableModeProps;
    // If any segment is selected, it is contenteditable=true and it has the focus. If you shift-Tab, you will
    // land on the outer div (DateField) first meaning it takes 2 shift-tab keys to get the previous element.
    // To prevent this, if any segment is selected the tabindex on the div is set to -1.
    // We also need to set tabindex to -1 for the disabled case, since we are dealing with a div instead
    // of an intrinsic input and have to handle it directly.
    const updatedTabIndex = tabbableModeIndex === -1 || hasSelectedSegment || isDisabled ? -1 : 0;
    // Keep track whether the date was cleared.
    const dateClearedRef = useRef(false);
    // The dateClearedRef.current is set to true when the user does a clear all.
    // We set it back to false whenever the value is not undefined
    if (value !== undefined && Object.keys(value).length > 0) {
        dateClearedRef.current = false;
    }
    // Handle requested segment changes by calling the reducer dispatch function.
    const onChange = useCallback((action) => {
        // Handle requested segment changes via the reducer's dispatch function.
        // This pattern ensures that rendered state (like segment isSelected) is updated via dispatch,
        // while also allowing us to immediately compute the resulting currentDate by re-reducing the state.
        // We then call onInput with the new value within the same event handler,
        // ensuring onInput is not called within a useLayoutEffect or other lifecycle hook.
        // This allows the parent to receive the currentDate synchronously in response to user input
        // and will re-render InputDateMask with the updated currentDate as needed.
        dispatch(action);
        const updatedState = dateReducer(state, action);
        onInput({ value: updatedState.currentDate });
    }, [dispatch, onInput, state]);
    // If the user clicked on a segment, dispatch a request to select it.
    // If they clicked on the date field itself, select the first segment.
    const onClick = useCallback((e) => {
        const segmentType = e.target.getAttribute('data-segment');
        if (segmentType !== null) {
            onChange({ actionType: 'selectSegment', data: { type: segmentType } });
        }
        else {
            onChange({ actionType: 'selectFirst', data: { orderedSegments } });
        }
    }, [onChange, orderedSegments]);
    // If complete, select the date.
    const selectAll = useCallback((isGroupFocus = true) => {
        if (isCompleteDate) {
            onChange({ actionType: 'selectDate' });
            // Focus can end up in different places depending on if the user selected a segment
            // first (via click or arrow key), and then did select all. To make shift-tabbing
            // consistent in all cases, set focus to the group when selecting the date.
            isGroupFocus && groupRef?.current?.focus();
        }
    }, [groupRef, isCompleteDate, onChange]);
    const onDoubleClick = useCallback(() => {
        selectAll();
    }, [selectAll]);
    const onDoubleTap = useCallback((e) => {
        const segmentType = e.target.getAttribute('data-segment');
        // Only do select all if a segment was double tapped, which means the
        // mobile keyboard will stay up.
        !!segmentType && selectAll(false);
    }, [selectAll]);
    const doubleTapHandlers = useDoubleTap({ onDoubleTap: onDoubleTap, onSingleTap: onClick });
    // Handle arrow key navigation between segments. If no segment is currently selected, then select the first one.
    const onKeyDown = useCallback((event) => {
        if ((direction === 'ltr' && event.key === 'ArrowLeft') ||
            (direction === 'rtl' && event.key === 'ArrowRight')) {
            // Select the previous segment when pressing Left arrow (or Right arrow in rtl).
            onChange({ actionType: 'selectPrevious', data: { orderedSegments } });
            event.stopPropagation();
        }
        else if ((direction === 'ltr' && event.key === 'ArrowRight') ||
            (direction === 'rtl' && event.key === 'ArrowLeft')) {
            // Select the next segment when pressing Right arrow (or Left arrow in rtl).
            onChange({ actionType: 'selectNext', data: { orderedSegments } });
            event.stopPropagation();
        }
        else if ((isBackspaceOrDelete(event) || event.key === 'Backspace') && isDateSelected) {
            // Clear the entire date.
            onChange?.({ actionType: 'clearDate', data: { orderedSegments } });
            dateClearedRef.current = true;
            // Kill the event to prevent weird issue with placeholders getting deleted on iOS only.
            event.preventDefault();
            event.stopPropagation();
            event.stopImmediatePropagation();
        }
        else if (isSelectAll(event)) {
            selectAll();
            event.preventDefault();
            event.stopPropagation();
        }
        else if (event.key === 'Enter') {
            // Hitting Enter commits any changes.
            dispatch({ actionType: 'commitOnEnter' });
            const updatedState = dateReducer(state, { actionType: 'commitOnEnter' });
            onInput?.({ value: updatedState.currentDate });
            onCommit?.({ value: updatedState.currentDate });
            event.preventDefault();
            event.stopPropagation();
        }
    }, [
        direction,
        isDateSelected,
        onChange,
        orderedSegments,
        selectAll,
        dispatch,
        state,
        onInput,
        onCommit
    ]);
    const onPointerDown = useInputDateMaskSelectFirstHandler({
        dispatch,
        isInputFocused,
        orderedSegments
    });
    const handlers = isMobile()
        ? { ...doubleTapHandlers, onKeyDown }
        : { onClick, onDoubleClick, onKeyDown, onPointerDown };
    return {
        dateClearedRef,
        dateFieldHandlers: isDisabled ? {} : handlers,
        segmentHandlers: isDisabled ? {} : { onChange },
        tabIndex: updatedTabIndex
    };
};

/**
 * Uses the properties to derive information that can be used to render the DateField and its segments.
 * As much as possible is derived from the properties, and the rest is merged in from the DateFieldState,
 * which is only the segment's isSelected and isComplete and isDateSelected.
 *
 * @param currentState - The current DateFieldState.
 * @param ariaLabels - An object containing aria labels for the date segments.
 * @param granularity - The date granularity.
 * @param locale - The locale.
 * @param value - The new CalendarDate value.
 * @returns A new object with properties needed to render the DateField and its DateSegments.
 */
const mergeDerivedStateWithReducerState = ({ currentState, ariaLabels, granularity, locale, value }) => {
    const segments = getDerivedDateSegmentRenderProps({
        ariaLabels,
        granularity,
        locale,
        value
    });
    // This is where we are merging in the DateField state with the derived state.
    // We merge in isSelected and isComplete from the currentState. Otherwise we use the derived info.
    const updatedSegments = segments.map((newSegment, index) => {
        const currentSegment = currentState.segments[index];
        return {
            ...newSegment,
            isSelected: currentSegment.isSelected,
            // Since the newSegment's isComplete is true when it has any value,
            // we want to keep the currentSegment.isComplete because it keeps track of what the user is typing.
            isComplete: currentSegment.isComplete
        };
    });
    const isPartialDate = value?.year !== undefined || value?.month !== undefined || value?.day !== undefined;
    const isCompleteDate = value?.year !== undefined &&
        value?.month !== undefined &&
        (granularity === 'month' || value?.day !== undefined);
    return {
        isCompleteDate,
        isDateSelected: currentState.isDateSelected,
        isPartialDate,
        segments: updatedSegments
    };
};
/**
 * Derives the information to render the date segments based on the provided properties
 * that come from the InputDateMask component.
 *
 * @param props - An object containing the properties used to derive the initial state.
 * @param props.ariaLabels - An object containing aria labels for the date segments.
 * @param props.granularity
 * @param props.locale
 * @param props.value The CalendarDate value
 * @returns The array that holds the information about each segment.
 */
const getDerivedDateSegmentRenderProps = ({ ariaLabels, granularity, locale, value: currentValue }) => {
    // Derive some data from currentValue that will be used to render the segments
    const yearValue = currentValue?.year;
    const monthValue = currentValue?.month;
    const dayValue = currentValue?.day;
    const monthMax = getMonthsInYear(yearValue ?? '*');
    const dayMax = getDaysInMonth(monthValue ?? '*', yearValue ?? '*');
    const yearValueStr = yearValue ? numberToString(yearValue) : undefined;
    const monthValueStr = monthValue ? numberToString(monthValue) : undefined;
    const dayValueStr = dayValue ? numberToString(dayValue) : undefined;
    const monthAria = monthValue ? getMonthAria(locale, monthValue, yearValue) : undefined;
    // Create state for each editable segment of a date, i.e. year, month, or day.
    const editableSegments = [
        {
            type: 'year',
            'aria-label': ariaLabels.year,
            'aria-valuemin': MIN_ISO_YEAR,
            'aria-valuemax': MAX_ISO_YEAR,
            'aria-valuenow': yearValue,
            'aria-valuetext': yearValueStr,
            text: yearValueStr
        },
        {
            type: 'month',
            'aria-label': ariaLabels.month,
            'aria-valuemin': MIN_ISO_MONTH,
            'aria-valuemax': monthMax,
            'aria-valuenow': monthValue,
            'aria-valuetext': monthAria,
            text: monthValueStr
        },
        ...(granularity === 'day'
            ? [
                {
                    type: 'day',
                    'aria-label': ariaLabels.day,
                    'aria-valuemin': MIN_ISO_DAY,
                    'aria-valuemax': dayMax,
                    'aria-valuenow': dayValue,
                    'aria-valuetext': dayValueStr,
                    text: dayValueStr
                }
            ]
            : [])
    ];
    return editableSegments;
};
// Determine the aria-valuetext for the given month, which includes the month name.
const getMonthAria = (locale, month, year) => {
    const monthName = getMonthName(locale, month, year ?? '*', 'long');
    return `${month} - ${monthName}`;
};

const findSegment = (segments, type) => {
    return segments.find((segment) => segment.type === type);
};
/**
 * Container for Segment children that represent different parts of a date with no time.
 */
const DateField = ({ 'aria-describedby': describedBy, 'aria-invalid': ariaInvalid, 'aria-label': ariaLabel, 'aria-labelledby': ariaLabelledBy, baseId, dispatch, fieldRef, granularity = 'day', hasInsideLabel, isRequired, masks: dateMasks, onBlur, onCommit, onFocus, onInput, state, textAlign, value, variant = 'default' }) => {
    const translations = useTranslationBundle('@oracle/oraclejet-preact');
    const announceDateCleared = translations.inputDateMask_date_cleared();
    const { direction, locale } = useUser();
    const { isDisabled, isFocused, isInputFocused } = useFormFieldContext();
    const hasValue = value !== undefined;
    // DateField is similar to TextFieldInput for styling. The main difference
    // is TextFieldInput renders a textarea or input and DateField renders a div, so the positioning with
    // padding is different. DateField and TextFieldInput are both the mainContent
    // of a TextField component.
    const { classes } = useComponentTheme(TextFieldInputRedwoodTheme, {
        type: 'notPassword',
        styleVariant: isEmbeddedVariant(variant) ? 'embedded' : 'default',
        textarea: 'notTextArea',
        input: 'notInput',
        prefix: 'noPrefix',
        suffix: 'noSuffix',
        startContent: 'noStartContent',
        endContent: 'noEndContent',
        insideLabel: hasInsideLabel ? 'hasInsideLabel' : 'noInsideLabel',
        value: hasValue ? 'hasValue' : 'noValue',
        focused: isFocused ? 'isFocused' : 'notFocused',
        disabled: isDisabled ? 'isDisabled' : 'notDisabled'
    });
    const compClasses = classNames([classes]);
    const ariaLabels = useMemo(() => ({
        month: translations.formControl_month(),
        day: translations.formControl_day(),
        year: translations.formControl_year()
    }), [translations]);
    // On every render we figure out information we need to render.
    // And we merge in the state we are keeping like what segment is selected.
    const { isCompleteDate, isDateSelected, isPartialDate, segments: editableSegments } = mergeDerivedStateWithReducerState({
        currentState: state,
        ariaLabels,
        granularity,
        locale,
        value
    });
    // Construct an ordered list of the editable segments, i.e. ['month', 'day', 'year'].
    // Figure this out from the dateMasks
    const orderedSegments = useMemo(() => {
        const fdm = dateMasks.filter(({ type }) => type !== 'literal');
        return fdm.map(({ type }) => type);
    }, [dateMasks]);
    const yearSegment = findSegment(editableSegments, 'year');
    const monthSegment = findSegment(editableSegments, 'month');
    const daySegment = granularity === 'day' ? findSegment(editableSegments, 'day') : undefined;
    const segmentsMap = {
        year: yearSegment,
        month: monthSegment,
        day: daySegment
    };
    const groupRef = useRef(null);
    const hasSelectedSegment = editableSegments.some((segment) => !!segment.isSelected);
    const { dateClearedRef, dateFieldHandlers, segmentHandlers, tabIndex } = useDateFieldHandlers({
        hasSelectedSegment,
        value,
        direction,
        dispatch,
        groupRef,
        isCompleteDate,
        isDateSelected,
        isDisabled,
        isInputFocused,
        orderedSegments,
        state,
        onCommit,
        onInput
    });
    // Like other form components, don't show placeholders for inside labels unless focused.
    // This flag causes the segment placeholders to be hidden.
    const isHidden = value === undefined && hasInsideLabel && !isFocused;
    // If the date is complete, create a hidden accessible region with the full formatted date.
    const formattedDateStr = !isCompleteDate
        ? ''
        : formatFullCalendarDate(locale, value, granularity);
    const formattedDateAriaId = `${baseId}-date`;
    const formattedDateAria = isCompleteDate ? (jsx(HiddenAccessible, { id: formattedDateAriaId, children: formattedDateStr })) : undefined;
    const ariaDescribedBy = isCompleteDate ? l([describedBy, formattedDateAriaId]) : describedBy;
    // Add an announcement to the live region if the date was cleared.
    // Note: we deliberately do not reset dateClearedRef because there are multiple
    // rerenders after clearing the date and we want to make sure the announcement
    // has time to be heard.
    const announceText = dateClearedRef.current ? announceDateCleared : '';
    // firstSegmentType is 'day', 'month', or 'year' and is used to determine
    // where to put the fieldRef.
    const firstSegmentType = orderedSegments[0];
    return (jsxs("div", { "aria-describedby": ariaDescribedBy, "aria-label": ariaLabel, "aria-labelledby": ariaLabelledBy, class: compClasses, onFocusIn: onFocus, onFocusOut: onBlur, ref: groupRef, role: "group", tabIndex: tabIndex, ...dateFieldHandlers, children: [jsx(Flex, { direction: "row", wrap: "nowrap", justify: textAlign, height: "100%", align: "center", children: dateMasks.map(({ type, value: placeholder }) => {
                    if (type === 'literal') {
                        return (jsx(LiteralSegment, { isHidden: isHidden, isHighlighted: isDateSelected, isPlaceholder: !isPartialDate, text: placeholder }));
                    }
                    else {
                        const segmentProps = segmentsMap[type];
                        if (segmentProps) {
                            return (jsx(DateSegment, { inputRef: type === firstSegmentType ? fieldRef : undefined, isDisabled: isDisabled, isFieldFocused: isFocused, isInvalid: !!ariaInvalid, isHidden: isHidden, isHighlighted: isDateSelected, isRequired: isRequired, orderedSegments: orderedSegments, placeholder: placeholder, ...segmentHandlers, ...segmentProps }));
                        }
                        return null;
                    }
                }) }), formattedDateAria, jsx(LiveRegion, { children: announceText })] }));
};

/**
 * Context used to indicate when the component is being used in
 * composition within some parent, such as InputDatePicker.
 */
const InputDateMaskContext = createContext({
    isInputDatePickerParent: false,
    isDropdownOpen: false
});
const useInputDateMaskContext = () => useContext(InputDateMaskContext);

/**
 * @license
 * Copyright (c) %FIRST_YEAR% %CURRENT_YEAR%, Oracle and/or its affiliates.
 * The Universal Permissive License (UPL), Version 1.0
 * as shown at https://oss.oracle.com/licenses/upl/
 * @ignore
 */
/**
 * The initial state of the date field can be passed to the dateReducer when the component
 * mounts, and can also be used to reset the state when the value changes programmatically.
 *
 * @param props - An object containing the properties used to derive the initial state.
 * @param props.granularity
 * @param props.value The CalendarDate value
 * @returns The initial state of the date field that can be passed to the dateReducer.
 */
const getDateFieldInitialReducerState = ({ granularity, value: currentValue }) => {
    // Derive some data from currentValue that will become part of our initial state.
    const yearValue = currentValue?.year;
    const monthValue = currentValue?.month;
    const dayValue = currentValue?.day;
    // 'aria-valuenow' is the segment's value. In the dateReducer we look at the 'aria-valuenow'
    // for each segment and when a user updates one segment, that new value may affect the others.
    // In the dateReducer we have a currentDate state that is calculated at the end of the dispatch action
    // - it takes each segment's 'aria-valuenow' and creates the currentDate from it.
    const daySeg = [
        {
            type: 'day',
            'aria-valuenow': dayValue,
            isComplete: !!dayValue,
            isSelected: false
        }
    ];
    // Create state for each editable segment of a date, i.e. year, month, or day.
    const editableSegments = [
        {
            type: 'year',
            'aria-valuenow': yearValue,
            isComplete: !!yearValue,
            isSelected: false
        },
        {
            type: 'month',
            'aria-valuenow': monthValue,
            isComplete: !!monthValue,
            isSelected: false
        },
        ...(granularity === 'day' ? daySeg : [])
    ];
    // Create the initial state to pass to our reducer.
    const initialState = {
        currentDate: currentValue,
        isDateSelected: false,
        segments: editableSegments
    };
    return initialState;
};

/**
 * @license
 * Copyright (c) %FIRST_YEAR% %CURRENT_YEAR%, Oracle and/or its affiliates.
 * The Universal Permissive License (UPL), Version 1.0
 * as shown at https://oss.oracle.com/licenses/upl/
 * @ignore
 */
/**
 * Hook that initializes the date state by calling useReducer.
 * It returns a state and a dispatch method to request changes to that state.
 */
const useDateFieldState = ({ dateMasks, granularity, locale, value: currentValue }) => {
    const init = (props) => getDateFieldInitialReducerState(props);
    const [state, dispatch] = useReducer(dateReducer, {
        dateMasks,
        granularity,
        locale,
        value: currentValue
    }, init);
    return { dispatch, state };
};

/**
 * An InputDateMask allows you to edit the month, day, and year fields of a calendar date.
 */
const InputDateMask = forwardRef(({ 'aria-describedby': ariaDescribedBy, assistiveText, columnSpan, endContent, granularity = 'day', helpSourceLink, helpSourceText, isDisabled: propIsDisabled, isReadonly: propIsReadonly, isRequired, isRequiredShown, label, labelEdge: propLabelEdge, labelStartWidth: propLabelStartWidth, mainFieldRef, messages, masks, testId, textAlign: propTextAlign, userAssistanceDensity: propUserAssistanceDensity, value, variant, onCommit, onInput }, ref) => {
    const { isDisabled: isFormDisabled, isReadonly: isFormReadonly, labelEdge: formLabelEdge, labelStartWidth: formLabelStartWidth, textAlign: formTextAlign, userAssistanceDensity: formUserAssistanceDensity } = useFormContext();
    // default to FormContext values if component properties are not specified
    const isDisabled = propIsDisabled ?? isFormDisabled;
    const isReadonly = propIsReadonly ?? isFormReadonly;
    const labelEdge = propLabelEdge ?? formLabelEdge;
    const labelStartWidth = propLabelStartWidth ?? formLabelStartWidth;
    // JET-72775 - Form control alignment in tables
    // default to 'start' if no text align is set
    const textAlign = propTextAlign ?? formTextAlign ?? 'start';
    const userAssistanceDensity = propUserAssistanceDensity ?? formUserAssistanceDensity;
    const enabledElementRef = useRef(null);
    const readonlyElementRef = useRef(null);
    const aRef = useRef(null);
    const anchorRef = mainFieldRef ?? aRef;
    const { isDropdownOpen, isInputDatePickerParent } = useInputDateMaskContext();
    const { focusProps, isFocused, methods: focusMethods } = useFocusableTextField({
        isDisabled,
        isReadonly,
        enabledElementRef,
        readonlyElementRef
    });
    useImperativeHandle(ref, () => focusMethods, [focusMethods]);
    const isComponentFocused = isFocused || (isInputDatePickerParent && isDropdownOpen);
    const { baseId, formFieldContext, inputProps, labelProps: origLabelProps, textFieldProps, userAssistanceProps } = useTextField({
        ariaDescribedBy,
        helpSourceLink,
        helpSourceText,
        isDisabled,
        isFocused: isComponentFocused,
        isReadonly,
        isRequiredShown,
        labelEdge,
        messages,
        styleVariant: variant,
        userAssistanceDensity,
        value
    });
    // When using a div with role=group (i.e. DateField), you need aria-labelledby instead of a
    // for-id association with a label. Remove forId and id and use aria-labelledby instead.
    const { forId, ...labelProps } = origLabelProps;
    const { id, onBlur: inputPropsOnBlur, ...dateFieldProps } = inputProps;
    const ariaLabelledBy = labelProps.id;
    // Because we don't use for-id, clicking the label does not automatically focus the field.
    // We have to do it manually.
    const onClick = useCallback(() => {
        enabledElementRef.current?.focus();
    }, [enabledElementRef]);
    const labelComp = labelEdge !== 'none' ? (jsx(Label, { ...(!isDisabled && !isReadonly ? { onClick } : {}), ...labelProps, children: label })) : undefined;
    const fieldLabelProps = {
        label: labelEdge !== 'none' ? labelComp : undefined,
        labelEdge: labelEdge !== 'none' ? labelEdge : undefined,
        labelStartWidth: labelEdge !== 'none' ? labelStartWidth : undefined,
        labelText: labelEdge === 'inside' && !isDisabled ? label : undefined
    };
    const ariaLabel = labelEdge === 'none' ? label : undefined;
    const hasInsideLabel = label !== undefined && labelEdge === 'inside';
    const isInlineDensity = userAssistanceDensity === 'efficient' || userAssistanceDensity === 'reflow';
    const inlineUserAssistance = isInlineDensity ? (isDisabled ? (
    // save space for user assistance if density is 'efficient', even though we don't
    // render user assistance for disabled fields
    userAssistanceDensity !== 'efficient' ? undefined : (jsx(InlineUserAssistance, { userAssistanceDensity: userAssistanceDensity, ...userAssistanceProps }))) : isReadonly ? (jsx(InlineUserAssistance, { fieldLabel: label, messages: messages, userAssistanceDensity: userAssistanceDensity, ...userAssistanceProps })) : (jsx(InlineUserAssistance, { assistiveText: assistiveText, fieldLabel: label, helpSourceLink: helpSourceLink, helpSourceText: helpSourceText, messages: messages, isRequiredShown: isRequiredShown, userAssistanceDensity: userAssistanceDensity, ...userAssistanceProps }))) : undefined;
    const compactUserAssistance = userAssistanceDensity === 'compact' ? (jsx(CompactUserAssistance, { anchorRef: anchorRef, messages: messages, assistiveText: assistiveText, ...userAssistanceProps })) : undefined;
    const { locale } = useUser();
    const translations = useTranslationBundle('@oracle/oraclejet-preact');
    const monthPlaceholder = translations.inputDateMask_monthPlaceholder();
    const dayPlaceholder = translations.inputDateMask_dayPlaceholder();
    const yearPlaceholder = translations.inputDateMask_yearPlaceholder();
    // The DatePlaceholders type already ensures the masks contain the right placeholders,
    // for either full calendar dates or month-year dates. This check ensures the masks,
    // if provided, match the granularity. E.g. for granularity=month, you should pass
    // month and year placeholders, and a separator.
    if (masks &&
        ((granularity === 'day' && masks.length !== 5) ||
            (granularity === 'month' && masks.length !== 3))) {
        throw new Error('Specified masks must match the granularity');
    }
    const dateMasks = useMemo(() => {
        // Determine the locale specific placeholder masks and separators in the correct order
        // for the locale, unless a custom set of masks was passed in. Each mask represents a
        // placeholder for a segment of a date (i.e. mm/dd/yyyy). Memoize because we shouldn't
        // need to recreate this data often.
        const placeholderMasks = masks
            ? translateCustomMasks(masks, monthPlaceholder, dayPlaceholder, yearPlaceholder)
            : getDatePlaceholderMasks(locale, granularity, monthPlaceholder, dayPlaceholder, yearPlaceholder);
        return placeholderMasks;
    }, [masks, locale, granularity, dayPlaceholder, monthPlaceholder, yearPlaceholder]);
    const { state, dispatch } = useDateFieldState({
        dateMasks,
        granularity,
        locale,
        value
    });
    // Keep track of the last committed date value sent to onCommit.
    const lastCommittedDateRef = useRef(value);
    // lastDateRef is a reference to the last date value that was changed by the user and sent via onInput.
    // It is used to make sure we do not call onInput with the same value as the previous onInput.
    const lastDateRef = useRef(value);
    const onInputIfDifferent = useCallback((detail) => {
        if (calendarDatesAreDifferent(detail.value, lastDateRef.current)) {
            const frozenDate = detail.value !== undefined ? Object.freeze({ ...detail.value }) : detail.value;
            onInput?.({ value: frozenDate });
            lastDateRef.current = frozenDate;
        }
    }, [onInput]);
    const onCommitIfDifferent = useCallback((detail) => {
        if (calendarDatesAreDifferent(detail.value, lastCommittedDateRef.current)) {
            onCommit?.({ value: detail.value });
            lastCommittedDateRef.current = detail.value;
        }
    }, [onCommit]);
    // The state gets reset when the component's value does not equal the last known value set in onInput.
    if (calendarDatesAreDifferent(lastDateRef.current, value)) {
        const initialState = getDateFieldInitialReducerState({
            granularity,
            locale,
            value
        });
        dispatch({ actionType: 'reset', data: initialState });
        // If the value was programmatically changed, then update the
        // last changed and last committed date refs to stay in sync.
        lastDateRef.current = value;
        lastCommittedDateRef.current = value;
    }
    // JET-73689: When we get onFocusOut on the date field that is not due to
    // interaction with the dropdown, we call onInput and onCommit.
    // We do not want to commit if we are simply moving between segments.
    // We dispatch 'commitOnBlur' to clear any selected segments and mark non-empty segments as
    // complete.
    const onDateFieldFocusOut = useCallback((event) => {
        if ((!isInputDatePickerParent || !isDropdownOpen) &&
            !event.currentTarget.contains(event.relatedTarget)) {
            // This pattern ensures that rendered state (like segment isSelected) is updated via dispatch,
            // while also allowing us to immediately compute the resulting currentDate by re-reducing the state.
            // We then call onInput and onCommit with the new value within the same event handler,
            // ensuring these are not called within a useLayoutEffect or other lifecycle hook.
            // This allows the parent to receive the currentDate synchronously in response to user action
            // and will re-render InputDateMask with the updated currentDate as needed.
            dispatch({ actionType: 'commitOnBlur' });
            const updatedState = dateReducer(state, { actionType: 'commitOnBlur' });
            onInputIfDifferent({ value: updatedState.currentDate });
            onCommitIfDifferent({ value: updatedState.currentDate });
        }
    }, [
        dispatch,
        isDropdownOpen,
        isInputDatePickerParent,
        onCommitIfDifferent,
        onInputIfDifferent,
        state
    ]);
    // merge the date field's onblurs.
    const { onBlur: handleDateFieldBlur } = useMemo(() => {
        return mergeProps({ onBlur: onDateFieldFocusOut }, { onBlur: inputPropsOnBlur });
    }, [inputPropsOnBlur, onDateFieldFocusOut]);
    if (isReadonly) {
        return (jsx(FormFieldContext.Provider, { value: formFieldContext, children: jsx(ReadonlyTextField, { role: "presentation", columnSpan: columnSpan, compactUserAssistance: compactUserAssistance, inlineUserAssistance: inlineUserAssistance, onBlur: focusProps.onFocusOut, onFocus: focusProps.onFocusIn, ref: anchorRef, testId: testId, mainContent: jsx(ReadonlyTextFieldInput, { "aria-describedby": dateFieldProps['aria-describedby'], "aria-label": ariaLabel, "aria-labelledby": labelProps.id, as: "div", elementRef: readonlyElementRef, textAlign: textAlign, value: value === undefined ? '' : formatShortCalendarDate(value, granularity, dateMasks), hasEmptyLabel: label === '' && labelEdge === 'none', hasInsideLabel: hasInsideLabel, onBlur: inputProps.onBlur, onFocus: inputProps.onFocus }), ...fieldLabelProps }) }));
    }
    // The main content inside TextField is a DateField (not TextFieldInput) in this case.
    // We reuse the TextField building blocks and hooks, but recompose them.
    const mainContent = (jsx(DateField, { "aria-label": ariaLabel, "aria-labelledby": ariaLabelledBy, baseId: baseId, dispatch: dispatch, fieldRef: enabledElementRef, granularity: granularity, hasInsideLabel: labelComp !== undefined && labelEdge === 'inside', isRequired: isRequired, masks: dateMasks, onCommit: onCommitIfDifferent, onBlur: handleDateFieldBlur, onInput: onInputIfDifferent, state: state, textAlign: textAlign, value: value, ...dateFieldProps }));
    return (jsx(FormFieldContext.Provider, { value: formFieldContext, children: jsx(TextField, { columnSpan: columnSpan, compactUserAssistance: compactUserAssistance, endContent: endContent, inlineUserAssistance: inlineUserAssistance, mainContent: mainContent, onBlur: focusProps.onFocusOut, onFocus: focusProps.onFocusIn, mainFieldRef: anchorRef, testId: testId, ...textFieldProps, ...fieldLabelProps }) }));
});

export { InputDateMask as I, InputDateMaskContext as a, useInputDateMaskContext as u };
