// @oracle/oraclejet-preact/es/NavigationListLinkItem-e28fd846.js
/* @oracle/oraclejet-preact: undefined */
import { jsxs, Fragment, jsx } from 'preact/jsx-runtime';
import { S as SvgClose } from './Close-5a460764.js';
import { E as EmbeddedIconButton } from './EmbeddedIconButton-46434710.js';
import { useContext, useCallback, useState, useRef } from 'preact/hooks';
import { u as useHover } from './useHover-85077adb.js';
import { u as useTranslationBundle } from './useTranslationBundle-4822502f.js';
import { T as Text } from './Text-834eb35e.js';
import { u as usePress } from './usePress-051f5adb.js';
import { m as mergeProps } from './mergeProps-88ea8306.js';
import { c as classNames } from './classNames-4e12b00d.js';
import { u as useInteractionStyle } from './useInteractionStyle-99e32c23.js';
import { u as useComponentTheme } from './useComponentTheme-d2f9e47f.js';
import { createContext } from 'preact';
import { m as mergeInterpolations } from './mergeInterpolations-30cd9a69.js';
import { f as flexitemInterpolations } from './flexitem-7b7f7920.js';
import { NavigationListItemRedwoodTheme } from './UNSAFE_NavigationListCommon/themes/redwood/NavigationListItemTheme.js';
import { u as useUser } from './useUser-06d34d19.js';
import { F as Flex } from './Flex-03129963.js';
import { B as Badge } from './Badge-a458c469.js';
import { S as SvgWarningS } from './WarningS-3e53d87e.js';
import { S as SvgErrorS } from './ErrorS-3a890bf6.js';
import { S as SvgInformationS } from './InformationS-d53c2a91.js';
import { S as SvgSuccessS } from './SuccessS-b5192d40.js';
import { u as useTooltipControlled } from './useTooltipControlled-f2cfa454.js';

/**
 * @license
 * Copyright (c) %FIRST_YEAR% %CURRENT_YEAR%, Oracle and/or its affiliates.
 * Licensed under The Universal Permissive License (UPL), Version 1.0
 * as shown at https://oss.oracle.com/licenses/upl/
 * @ignore
 */
/**
 * Context used to pass navlist information without having to pass it to navlist children props.
 * We want to communicate information down to any interested navlist item children.
 */
const NavigationListContext = createContext(null);

/**
 * @license
 * Copyright (c) %FIRST_YEAR% %CURRENT_YEAR%, Oracle and/or its affiliates.
 * Licensed under The Universal Permissive License (UPL), Version 1.0
 * as shown at https://oss.oracle.com/licenses/upl/
 * @ignore
 */
/**
 * Utility hook for consuming the NavigationListContext
 *
 * @returns The value of NavigationListContext provider
 */
function useNavigationListContext() {
    return useContext(NavigationListContext);
}

const flexItemStyleInterpolations = mergeInterpolations([
    ...Object.values(flexitemInterpolations)
]);
const { class: labelContainerFlexItemClasses, ...labelContainerFlexItemStyle } = flexItemStyleInterpolations({
    alignSelf: 'center'
});
function useNavigationListItem({ itemKey }) {
    const { onSelectionChange, selection, currentKey, showFocusRing, navigationListItemPrefix, edge, display } = useNavigationListContext();
    //Selection and focus state
    const isCurrent = currentKey === itemKey;
    const isSelected = selection === itemKey;
    const isIconOnly = display === 'icons' ? true : false;
    const isEdgeEnd = edge === 'end' ? true : false;
    //Styling
    const { interactionProps, applyPseudoHoverStyle, applyHoverStyle, applyActiveStyle } = useInteractionStyle();
    const { direction } = useUser();
    const isRtl = direction === 'rtl';
    const { classes: listItemClasses, styles: { labelContainerClasses, navigationItemStackedBadge, navigationItemStackedBadgeRtl, startIconStacked } } = useComponentTheme(NavigationListItemRedwoodTheme, {
        focusRing: showFocusRing ? 'isFocusRing' : 'notFocusRing',
        current: isCurrent ? 'isCurrent' : 'notCurrtent',
        selected: isSelected ? 'isSelected' : 'notSelected',
        direction: isRtl ? 'isRtl' : 'notRtl',
        hybridHover: applyHoverStyle ? 'isHybridHover' : 'notHybridHover',
        pseudoHover: applyPseudoHoverStyle ? 'isPseudoHover' : 'notPseudoHover',
        active: applyActiveStyle ? 'isActive' : 'notActive',
        edgeEnd: edge === 'end' ? 'isEnd' : 'notEnd',
        iconOnly: isIconOnly ? 'isIconOnly' : 'notIconOnly'
    });
    const navigationItemIconOnlyClasses = classNames([startIconStacked]);
    const navigationItemStackedBadgeClasses = classNames([
        navigationItemStackedBadge,
        direction === 'rtl' && navigationItemStackedBadgeRtl
    ]);
    //Item handlers
    const handlePress = useCallback(() => {
        onSelectionChange?.({ value: itemKey, reason: 'pointer' });
    }, [itemKey, onSelectionChange]);
    //Legacy navlist makes selection on keyDown, but since preact tabbar made selection after keyUp, will do the same
    //for preact navlist
    const { pressProps } = usePress(handlePress);
    return {
        itemId: navigationListItemPrefix + itemKey,
        itemHandlers: mergeProps(interactionProps, pressProps /*,onFocusProps*/),
        itemClasses: classNames([listItemClasses]),
        isSelected,
        labelContainerClasses: classNames([labelContainerClasses, labelContainerFlexItemClasses]),
        labelContainerStyle: labelContainerFlexItemStyle,
        isIconOnly,
        isEdgeEnd,
        stackedBadgeClasses: classNames([navigationItemStackedBadgeClasses]),
        iconOnlyClasses: classNames([navigationItemIconOnlyClasses]),
        isCurrent,
        showFocusRing
    };
}

const useTabBarItemTooltip = ({ label, isDisabled, position }) => {
    const [isOpen, setOpen] = useState(false);
    const { tooltipProps, tooltipContent } = useTooltipControlled({
        isOpen: isOpen,
        isDisabled,
        position: position,
        focusStrategy: 'logical',
        text: label,
        onToggle: ({ value }) => {
            setOpen(value);
        }
    });
    const { onLogicalFocus, onLogicalBlur, ...tooltipEventHandlerProps } = tooltipProps;
    return { tooltipContent, tooltipEventHandlerProps, onLogicalFocus, onLogicalBlur };
};

const onClickLink = (event) => {
    event.preventDefault();
};
/*TODO: JET-58535. Add ripple effect when touching navigationListItem. */
function BaseNavigationListItem({ itemKey, label, badge, metadata, severity = 'none', removeIcon, icon, href, 'aria-controls': ariaControls }) {
    const itemRef = useRef(null);
    //Logic of the component handled here
    const { itemId, itemHandlers, itemClasses, isSelected, labelContainerClasses, labelContainerStyle, isIconOnly, isEdgeEnd, stackedBadgeClasses, iconOnlyClasses, isCurrent, showFocusRing } = useNavigationListItem({
        itemKey
    });
    const severityRequirement = !(severity === 'none' || severity === undefined);
    const IconComponent = (severityValue) => {
        switch (severityValue) {
            case 'error':
                return jsx(SvgErrorS, { color: "danger" });
            case 'warning':
                return jsx(SvgWarningS, { color: "warning" });
            case 'info':
                return jsx(SvgInformationS, { color: "info" });
            case 'confirmation':
                return jsx(SvgSuccessS, { color: "success" });
            default:
                return null;
        }
    };
    const { styles: { iconLabelContainer, iconContainer, startIconPadding, endContentPadding, removeIconPaddingWithEndContent, navigationItemMetadata, navigationListItemLink } } = useComponentTheme(NavigationListItemRedwoodTheme);
    const navigationItemMatadata = classNames([navigationItemMetadata, endContentPadding]);
    const navigationItemLinkClass = classNames([navigationListItemLink]);
    const endIconContainerClasses = classNames([
        iconLabelContainer,
        iconContainer,
        endContentPadding
    ]);
    const startIconContainerClasses = classNames([
        !isIconOnly && iconLabelContainer,
        iconContainer,
        //For icon only icon to label padding is not required.
        !isIconOnly && startIconPadding,
        isIconOnly && iconOnlyClasses
    ]);
    const { styles: { navigationlistRemoveIcon } } = useComponentTheme(NavigationListItemRedwoodTheme);
    const isComplexItem = metadata || severityRequirement || badge;
    const removeClass = classNames([
        navigationlistRemoveIcon,
        isComplexItem && removeIconPaddingWithEndContent,
        !isComplexItem && endContentPadding
    ]);
    const isLinkItem = href != null;
    const labelRequirement = label && !(label === ' ' && label === undefined);
    const iconOnlyRequirement = icon && icon != null && isIconOnly;
    const showTooltip = iconOnlyRequirement;
    const { tooltipContent, tooltipEventHandlerProps, onLogicalFocus, onLogicalBlur } = useTabBarItemTooltip({
        label,
        isDisabled: !showTooltip,
        // For edge end the tooltip will appear after icon in the order tooltip, icon, selection when navlist is placed at end of page.
        position: !isEdgeEnd ? 'end' : 'start'
    });
    if (showTooltip) {
        if ((isCurrent && showFocusRing) || isSelected) {
            onLogicalFocus(itemRef.current);
        }
        else if (!isCurrent || !showFocusRing) {
            onLogicalBlur();
        }
    }
    const labelRenderer = () => {
        return (jsxs(Fragment, { children: [icon && (jsx(Flex, { justify: "start", children: jsx("span", { class: startIconContainerClasses, children: icon }) })), !isIconOnly && labelRequirement && (jsx(Flex, { align: "center", justify: "start", flex: "1 1 auto", children: jsx("span", { class: labelContainerClasses, style: labelContainerStyle, children: jsx(Text, { size: "inherit", weight: "inherit", variant: "inherit", children: label }) }) })), removeIcon && !isLinkItem && jsx("span", { class: removeClass, children: removeIcon })] }));
    };
    const complexItemRenderer = () => {
        return (jsxs(Fragment, { children: [iconOnlyRequirement && (jsxs("div", { class: startIconContainerClasses, children: [icon, jsx("span", { class: stackedBadgeClasses, children: badge ? (jsx(Badge, { size: "xs", variant: "neutral", children: badge })) : null })] })), !isIconOnly && icon && (jsx(Flex, { justify: "start", children: jsx("span", { class: startIconContainerClasses, children: icon }) })), (!isIconOnly || /* This covers invalid cases*/ (isIconOnly && !iconOnlyRequirement)) &&
                    labelRequirement && (jsx(Flex, { align: "center", justify: "start", flex: "1 1 auto", children: jsx("span", { class: labelContainerClasses, style: labelContainerStyle, children: jsx(Text, { size: "inherit", weight: "inherit", variant: "inherit", children: label }) }) })), (metadata || severityRequirement || badge || removeIcon) && (jsxs(Flex, { align: "center", justify: "end", children: [' ', (!isIconOnly || /* This covers invalid cases*/ (isIconOnly && !iconOnlyRequirement)) &&
                            badge && (jsx("span", { class: endIconContainerClasses, children: jsx(Badge, { size: "md", variant: "neutralSubtle", children: badge }) })), metadata && (jsx("span", { class: navigationItemMatadata, children: jsx(Text, { children: '( ' + metadata + ' )' }) })), severityRequirement && (jsx("span", { class: endIconContainerClasses, children: IconComponent(severity) })), removeIcon && !isLinkItem && jsx("span", { class: removeClass, children: removeIcon })] }))] }));
    };
    const navlistItemRenderer = () => {
        return (jsxs("div", { role: "tab", ...mergeProps(tooltipEventHandlerProps, itemHandlers), "data-oj-key": itemKey, id: itemId, class: itemClasses, "aria-selected": isSelected, ref: itemRef, "aria-controls": ariaControls, children: [isComplexItem ? complexItemRenderer() : labelRenderer(), isIconOnly && tooltipContent] }));
    };
    const navlistLinkItemRenderer = () => {
        return (jsx("div", { ...itemHandlers, class: itemClasses, children: jsx("a", { tabindex: -1, role: "tab", "data-oj-key": itemKey, id: itemId, class: navigationItemLinkClass, "aria-selected": isSelected, "aria-controls": ariaControls, href: href, onClick: onClickLink, children: isComplexItem ? complexItemRenderer() : labelRenderer() }) }));
    };
    return isLinkItem ? navlistLinkItemRenderer() : navlistItemRenderer();
}

function RemovableNavigationListItem({ itemKey, label, badge, metadata, severity }) {
    const { selection, currentKey, onRemove } = useNavigationListContext();
    const isSelected = selection === itemKey;
    const isCurrent = currentKey === itemKey;
    const handleRemove = useCallback(() => {
        onRemove?.({ value: itemKey });
    }, [onRemove, itemKey]);
    const { hoverProps, isHover } = useHover({});
    const translations = useTranslationBundle('@oracle/oraclejet-preact');
    //instead of specifying flex: '1 0 auto' in cssProps var
    return (jsx("div", { ...hoverProps, children: jsx(BaseNavigationListItem, { itemKey: itemKey, label: label, badge: badge, metadata: metadata, severity: severity, removeIcon: jsx("span", { "data-oj-navigationlist-item-remove-icon": true, children: jsx(EmbeddedIconButton, { "aria-label": translations.tabBarNavigationList_removeCueText(), size: "xs", onAction: handleRemove, children: (isHover || isSelected || isCurrent) && jsx(SvgClose, {}) }) }) }) }));
}

/**
 * NavigationListLinkItem is used to render links in a NavigationList. It allows to view the browser's built in link context menu to open the link in new browser tab.
 */
function NavigationListLinkItem({ itemKey, badge, metadata, severity, href, label, 'aria-controls': ariaControls }) {
    return (jsx(BaseNavigationListItem, { "aria-controls": ariaControls, itemKey: itemKey, label: label, badge: badge, metadata: metadata, severity: severity, href: href }));
}

export { BaseNavigationListItem as B, NavigationListLinkItem as N, RemovableNavigationListItem as R, NavigationListContext as a, useTabBarItemTooltip as u };
