// @oracle/oraclejet-preact/es/PRIVATE_TableList.js
/* @oracle/oraclejet-preact: undefined */
import { jsx, jsxs, Fragment } from 'preact/jsx-runtime';
import { createRef } from 'preact';
import { memo, forwardRef } from 'preact/compat';
import { useState, useRef, useCallback, useMemo, useEffect, useLayoutEffect, useImperativeHandle, useContext } from 'preact/hooks';
import './CollectionInteractionContext-7cf19abc.js';
import { u as useCollectionInteractionContext } from './useCollectionInteractionContext-3895c632.js';
import { u as useComponentTheme } from './useComponentTheme-d2f9e47f.js';
import { u as useId } from './useId-03dbfdf0.js';
import { T as TabbableModeContext } from './TabbableModeContext-7d8ad946.js';
import { u as useTestId } from './useTestId-f24ac480.js';
import { u as useUser } from './useUser-06d34d19.js';
import { c as classNames } from './classNames-4e12b00d.js';
import { H as HiddenAccessible } from './HiddenAccessible-61197e7c.js';
import { F as FocusTracker } from './FocusTracker-70300f1c.js';
import { a as getRowKey, b as getElementBySelector, g as getCellType, c as getRowIndexForRowElement$1, d as getLastRenderedRowKey$1, i as isHTMLScroller, f as getVerticalOverflowDiff, j as getScrollbarHeight, u as useDataManager } from './ScrollUtils-8a50c01e.js';
import { m as mergeProps } from './mergeProps-88ea8306.js';
import { u as useContextMenuGesture } from './useContextMenuGesture-e09f4c69.js';
import { k as keyExtractor, f as findElementByKey, g as getElementContainsFunc, o as handleEnterActionableMode, p as handleWrapActionableFocus, h as dragToScroll } from './collectionUtils-5313629f.js';
import { a as allTabbableElements, g as getActiveElement, c as getBodyElement, i as isTabbableElement } from './tabbableUtils-c8311ec6.js';
import { u as useReorderable, a as useReorderableItem } from './useReorderableItem-857636b7.js';
import './DragAndDropStyles.styles.css';
import './LayerHost-7931b424.js';
import './ReorderableItemStyles.styles.css';
import './ReorderableContext-0daeea89.js';
import { i as isLogicalCtrl, g as getRowActionGestureContext, h as handleRowActionGesture } from './EventUtils-e1b93d3a.js';
import { c as containsKey, a as addKey, r as removeKey, e as isSameKey, i as isKeyDefined } from './keys-49c92dbb.js';
import { g as getIsSelectionPending } from './textSelectionUtils-e87a47ff.js';
import { U as UNSAFE_DOM_ACCESS } from './unsafeDomAccess-453eb74e.js';
import { sparkleStyles, reorderableStyles, scrollerStyles } from './PRIVATE_TableList/themes/List.css.js';
import { ListRedwoodTheme } from './PRIVATE_TableList/themes/redwood/ListTheme.js';
import { g as globalImages } from './ImageVars.css-7b3180c8.js';
import { u as useTheme } from './useTheme-0e7497db.js';
import { L as LOADMORE_STYLE_CLASS, a as LoadMoreCollection } from './LoadMoreCollection-d9f22790.js';
import { u as useTranslationBundle } from './useTranslationBundle-4822502f.js';
import { listRowStyles, listRowStylesHC } from './PRIVATE_TableList/themes/ListRowStyles.css.js';
import { ListNoDataRedwoodTheme } from './PRIVATE_TableList/themes/redwood/ListNoDataTheme.js';
import { u as useInteractionStyle } from './useInteractionStyle-99e32c23.js';
import { S as Selector } from './Selector-23bccbe7.js';
import { textAlignStyles, verticalAlignStyles } from './PRIVATE_TableCommon/themes/Common.css.js';
import { listCellStylesHC, listCellStyles } from './PRIVATE_TableList/themes/ListCellStyles.css.js';
import { listSkeletonStylesHC, listSkeletonStyles } from './PRIVATE_TableList/themes/ListSkeletonStyles.css.js';
import { d as dimensionInterpolations } from './dimensions-e4fbc23d.js';
import { m as mergeInterpolations } from './mergeInterpolations-30cd9a69.js';
import { M as Menu } from './SelectMenuGroupContext-aa1fe822.js';
import './logger-c92f309c.js';
import './TooltipContentStyles.styles.css';
import './hooks/UNSAFE_useTooltip/themes/redwood/TooltipContentBaseTheme.styles.css';
import './hooks/UNSAFE_useTooltip/themes/redwood/TooltipContentVariants.css.js';
import './tooltipUtils-af0abd38.js';
import './IconStyle.styles.css';
import './UNSAFE_Menu/themes/MenuItemStyles.css.js';
import './UNSAFE_Separator/themes/SeparatorStyles.css.js';
import './MenuSeparatorStyles.styles.css';
import './UNSAFE_Menu/themes/redwood/MenuSeparatorBaseTheme.styles.css';
import './UNSAFE_Menu/themes/redwood/MenuSeparatorVariants.css.js';
import './MenuStyles.styles.css';
import './HiddenAccessibleStyles.styles.css';
import './PRIVATE_FocusTracker/themes/FocusTrackerStyles.css.js';
import './FocusTrackerStyles.styles.css';
import './clientHints-a42b510c.js';
import './head-8f89fe66.js';
import './_arity-be492b9e.js';
import './_curry3-577eb45e.js';
import './_curry1-63949a9b.js';
import './_curry2-01c87f61.js';
import './_isArray-694cc52d.js';
import './_isString-675f1de9.js';
import './dndUtils-deb68797.js';
import './hooks/PRIVATE_useReorderable/themes/ReorderableItemStyles.css.js';
import './index-aec1cc7f.js';
import './UNSAFE_Layer/themes/LayerHostStyles.css.js';
import './LayerHostStyles.styles.css';
import './ImageVars.styles.css';
import './Common.styles.css';
import './List.styles.css';
import './PRIVATE_TableList/themes/ListStyles.css.js';
import './ListStyles.styles2.css';
import './PRIVATE_TableList/themes/redwood/ListBaseTheme.css.js';
import './PRIVATE_TableList/themes/redwood/ListBaseTheme.styles.css';
import './PRIVATE_TableList/themes/redwood/ListVariants.css.js';
import './vanilla-extract-recipes-createRuntimeFn.esm-2aaf8c98.js';
import './Collection-e1ef8b8d.js';
import './useViewportIntersect-6ff65310.js';
import './ListRowStyles.styles.css';
import './PRIVATE_TableList/themes/ListNoDataStyles.css.js';
import './ListNoDataStyles.styles.css';
import './PRIVATE_TableList/themes/redwood/ListNoDataVariants.css.js';
import './useHover-85077adb.js';
import './useToggle-8b7fcefe.js';
import './useActive-fb6d4d52.js';
import './BaseSelector-7cbc48bc.js';
import './useFocus-38c95977.js';
import './StyledCheckbox-d6c8f557.js';
import './Check-ac39fe00.js';
import './Icon-4983d06e.js';
import './size-af1e638c.js';
import './utils-b1eb54f9.js';
import './themeContract.css-1ffed3bf.js';
import './colorUtils-b3804ee8.js';
import './useTooltip-94af94d4.js';
import './useTooltipControlled-f2cfa454.js';
import './useTouch-4828df25.js';
import './useAnimation-fe39685e.js';
import './useThemeInterpolations-e144e8ec.js';
import './useColorScheme-c4f6caab.js';
import './useScale-434ad939.js';
import './theme-63551f30.js';
import './Theme-e6dec6db.js';
import './useDensity-1acb0153.js';
import './hooks/UNSAFE_useTooltip/themes/redwood/TooltipContentTheme.js';
import './hooks/UNSAFE_useTooltip/themes/TooltipContentStyles.css.js';
import './hooks/UNSAFE_useTooltip/themes/redwood/TooltipContentBaseTheme.css.js';
import './TooltipContentVariants.styles.css';
import './UNSAFE_Icon/themes/IconStyle.css.js';
import './CheckboxOff-3359a12d.js';
import './CheckboxOn-930b64de.js';
import './CheckboxMixed-c4594250.js';
import './UNSAFE_Checkbox/themes/CheckboxIconContract.css.js';
import './UNSAFE_Checkbox/themes/redwood/CheckboxIconTheme.js';
import './UNSAFE_Checkbox/themes/CheckboxIconStyles.css.js';
import './CheckboxIconStyles.styles.css';
import './UNSAFE_Checkbox/themes/redwood/CheckboxIconBaseTheme.css.js';
import './UNSAFE_Checkbox/themes/redwood/CheckboxIconBaseTheme.styles.css';
import './UNSAFE_Checkbox/themes/redwood/CheckboxIconVariants.css.js';
import './CheckboxIconVariants.styles.css';
import './useTabbableMode-090ed4b5.js';
import './StyledCheckbox.styles.css';
import './PRIVATE_BaseSelector/themes/SelectorStyles.css.js';
import './SelectorStyles.styles.css';
import './ListCellStyles.styles.css';
import './ListSkeletonStyles.styles.css';
import './mergeDeepWithKey-33e3d1be.js';
import './_isObject-1fab0f5b.js';
import './Sheet-f9f8a4a4.js';
import './Modal-c0f730b8.js';
import './useModal-953082f2.js';
import './Layer-535577b5.js';
import './UNSAFE_Layer/themes/LayerStyles.css.js';
import './LayerStyles.styles.css';
import './UNSAFE_Modal/themes/ModalStyles.css.js';
import './ModalStyles.styles.css';
import './popupUtils-51366a29.js';
import './WindowOverlay-7264c44a.js';
import './vanilla-extract-dynamic.esm-c909758c.js';
import './UNSAFE_WindowOverlay/themes/WindowOverlayStyles.css.js';
import './WindowOverlayStyles.styles.css';
import './UNSAFE_WindowOverlay/themes/WindowOverlayContract.css.js';
import './UNSAFE_WindowOverlay/themes/redwood/WindowOverlayTheme.js';
import './UNSAFE_WindowOverlay/themes/redwood/WindowOverlayBaseTheme.css.js';
import './UNSAFE_WindowOverlay/themes/redwood/WindowOverlayBaseTheme.styles.css';
import './UNSAFE_WindowOverlay/themes/redwood/WindowOverlayVariants.css.js';
import './UNSAFE_Sheet/themes/SheetStyles.css.js';
import './SheetStyles.styles.css';
import './refUtils-76275c69.js';
import './useAnimationStatus-b881f39a.js';
import './animationUtils-5f409a6c.js';
import './Dropdown-019aba37.js';
import './Floating-4a82db94.js';
import './useFloating-b9d680f7.js';
import './positionUtils-f26caad9.js';
import './useOutsideClick-b777a43e.js';
import './arrayUtils-35a58161.js';
import './UNSAFE_Floating/themes/redwood/FloatingTheme.js';
import './UNSAFE_Floating/themes/FloatingStyles.css.js';
import './FloatingStyles.styles.css';
import './UNSAFE_Floating/themes/redwood/FloatingBaseTheme.css.js';
import './UNSAFE_Floating/themes/redwood/FloatingBaseTheme.styles.css';
import './UNSAFE_Floating/themes/redwood/FloatingVariants.css.js';
import './UNSAFE_Floating/themes/FloatingContract.css.js';
import './useUnsafeDomElementRef-32290273.js';
import './keyboardUtils-986fe923.js';
import './usePopupAnimation-96d2bc83.js';
import './popupAnimationUtils-175151d9.js';
import './UNSAFE_Dropdown/themes/dropdownStyles.css.js';
import './dropdownStyles.styles.css';
import './useOutsideMousedown-c45c22c0.js';
import './useEffectEvent-1bb008cb.js';
import './UNSAFE_Menu/themes/MenuStyles.css.js';
import './UNSAFE_Menu/themes/DropdownMenuStyles.css.js';
import './DropdownMenuStyles.styles.css';
import './Flex-03129963.js';
import './boxalignment-8980856d.js';
import './boxalignment.styles.css';
import './vanilla-extract-sprinkles-createRuntimeSprinkles.esm-2d655d37.js';
import './flexbox-529f25da.js';
import './flexbox.styles.css';
import './flexitem-7b7f7920.js';
import './flexitem.styles.css';
import './FlexStyles.styles.css';
import './Skeleton-d6f0969e.js';
import './borders-98a63040.js';
import './UNSAFE_Skeleton/themes/redwood/SkeletonTheme.js';
import './UNSAFE_Skeleton/themes/SkeletonStyles.css.js';
import './SkeletonStyles.styles.css';
import './UNSAFE_Skeleton/themes/redwood/SkeletonBaseTheme.css.js';
import './UNSAFE_Skeleton/themes/redwood/SkeletonBaseTheme.styles.css';
import './UNSAFE_Skeleton/themes/redwood/SkeletonVariants.css.js';
import './UNSAFE_Menu/themes/MenuSkeletonStyles.css.js';
import './MenuSkeletonStyles.styles.css';
import './LiveRegion-cebac18d.js';
import './LiveRegionStyles.styles.css';
import './UNSAFE_Menu/themes/redwood/MenuItemTheme.js';
import './UNSAFE_Menu/themes/redwood/MenuItemBaseTheme.css.js';
import './UNSAFE_Menu/themes/redwood/MenuItemBaseTheme.styles.css';
import './UNSAFE_Menu/themes/redwood/MenuItemVariants.css.js';
import './MenuItemStyles.styles.css';
import './MenuItemVariants.styles.css';
import './Text-834eb35e.js';
import './UNSAFE_Text/themes/TextStyles.css.js';
import './TextStyles.styles.css';
import './usePress-051f5adb.js';
import './EnvironmentProvider-6a0a8212.js';
import './LayerManager-52a26c55.js';
import './SeparatorStyles.styles.css';

const LIST_ROW_SELECTOR = 'data-oj-list-row';
const LIST_NODATA_SELECTOR = 'data-oj-list-nodata';
const LIST_LOADING_SELECTOR = 'data-oj-list-loading';
const LIST_LOADMORE_SELECTOR = 'data-oj-list-loadmore';
const LIST_FOCUSABLE_SELECTOR = 'data-oj-list-focusable';
/**
 * Returns whether or not auto focus is enabled on the specified cell
 */
const isAutoFocusCell = (rootElement, listId, activeCell) => {
    const cellElement = getElementFromCell(rootElement, listId, activeCell);
    return cellElement != null && cellElement.dataset['ojListAutoFocus'] === 'enabled';
};
/**
 * Returns the key for the row at the given index
 */
const getRowKeyForRowIndex = (rootElement, listId, index) => {
    const rowElements = rootElement.querySelectorAll(`[${LIST_ROW_SELECTOR}='${listId}']`);
    if (rowElements.length > index) {
        return getRowKey(rowElements[index]);
    }
    return undefined;
};
/**
 * Returns the index of the given row element
 */
const getRowIndexForRowElement = (rootElement, listId, rowElement) => {
    return getRowIndexForRowElement$1(rootElement, listId, rowElement, LIST_ROW_SELECTOR);
};
/**
 * Returns the index of the row for the given key.
 */
function getRowIndexForRowKey(rootElement, listId, rowKey) {
    const rowElements = rootElement.querySelectorAll(`[${LIST_ROW_SELECTOR}='${listId}']`);
    for (let i = 0; i < rowElements.length; i++) {
        if (rowKey === getRowKey(rowElements[i])) {
            return i;
        }
    }
    return undefined;
}
/**
 * Helper method to get the cell key definition from an element
 */
function logicalCellExtractor(element, listId, cellSelector, onlyAncestors) {
    const selectorString = cellSelector != null ? cellSelector : `[${LIST_FOCUSABLE_SELECTOR}='${listId}']`;
    const startingElement = onlyAncestors ? element.parentElement : element;
    if (startingElement != null) {
        const cellElement = startingElement.closest(selectorString);
        if (cellElement != null) {
            const cellType = getCellType(cellElement);
            if (cellType === 'data') {
                return {
                    rowKey: keyExtractor(cellElement, `[${LIST_ROW_SELECTOR}='${listId}']`),
                    type: cellType
                };
            }
            else if (cellType === 'noData') {
                return { type: cellType };
            }
            else if (cellType === 'loading' || cellType === 'loadMore') {
                return { type: cellType };
            }
        }
    }
    return undefined;
}
function rowExtractor(element, listId) {
    const selectorString = `[${LIST_ROW_SELECTOR}='${listId}']`;
    return element.closest(selectorString);
}
/**
 * Helper method to get an element from a cell descriptor
 */
function getElementFromCell(rootElement, listId, cell) {
    if (cell.type === 'data') {
        // data body case
        const rowElements = rootElement.querySelectorAll(`[${LIST_ROW_SELECTOR}='${listId}']`);
        for (const rowElement of rowElements) {
            if (cell.rowKey === getRowKey(rowElement)) {
                const cellElement = rowElement.querySelector(`[${LIST_FOCUSABLE_SELECTOR}='${listId}']`);
                if (cellElement != null) {
                    return cellElement;
                }
            }
        }
    }
    else if (cell.type === 'noData') {
        // no data case
        return getNoDataCellElement(rootElement, listId);
    }
    else if (cell.type === 'loading') {
        // loading skeletons case
        return getLoadingCellElement(rootElement, listId);
    }
    else if (cell.type === 'loadMore') {
        // loadMore skeletons case
        return getLoadMoreCellElement(rootElement, listId);
    }
    return undefined;
}
/**
 * Returns the key of the last focusable row in the List
 */
const getLastRenderedRowKey = (rootElement, listId) => {
    return getLastRenderedRowKey$1(rootElement, listId, LIST_ROW_SELECTOR);
};
/**
 * Returns the cell definition of the first rendered cell in the List
 */
const getFirstFocusableCell = (rootElement, listId) => {
    const firstCellElement = rootElement.querySelector(`[${LIST_FOCUSABLE_SELECTOR}='${listId}']`);
    if (firstCellElement != null) {
        const firstCellType = getCellType(firstCellElement);
        if (firstCellType === 'noData') {
            return { type: 'noData' };
        }
        return {
            rowKey: keyExtractor(firstCellElement, `[${LIST_ROW_SELECTOR}='${listId}']`),
            type: 'data'
        };
    }
    return undefined;
};
/**
 * Helper method to get the 'loading' cell
 */
const getLoadingCellElement = (rootElement, listId) => {
    return getElementBySelector(rootElement, listId, LIST_LOADING_SELECTOR);
};
/**
 * Helper method to get the 'loadMore' cell
 */
const getLoadMoreCellElement = (rootElement, listId) => {
    return getElementBySelector(rootElement, listId, LIST_LOADMORE_SELECTOR);
};
/**
 * Helper method to get the 'noData' cell
 */
const getNoDataCellElement = (rootElement, listId) => {
    return getElementBySelector(rootElement, listId, LIST_NODATA_SELECTOR);
};
/**
 * Helper method to determine if the 'loadMore' cell is present in the List
 */
const hasLoadMoreCell = (rootElement, listId) => {
    return getLoadMoreCellElement(rootElement, listId) != null;
};

/**
 * Helper method to get the top scroller offset of the List
 */
const _getScrollerTopOffset = (viewportConfig) => {
    return viewportConfig.scrollerOffsetTop ?? 0;
};
/**
 * Helper method to get the bottom scroller offset of the List
 */
const _getScrollerBottomOffset = (viewportConfig) => {
    return viewportConfig.scrollerOffsetBottom ?? 0;
};
/**
 * Helper method to return the top position of the scroller element.
 */
const getScrollerTop = (scrollerElement, viewportConfig) => {
    return isHTMLScroller(scrollerElement)
        ? _getScrollerTopOffset(viewportConfig)
        : _getScrollerTopOffset(viewportConfig) + scrollerElement.getBoundingClientRect().top;
};
/**
 * Helper method to return the bottom position of the scroller element.
 */
const getScrollerBottom = (scrollerElement, viewportConfig) => {
    const bottomOffset = _getScrollerBottomOffset(viewportConfig);
    if (isHTMLScroller(scrollerElement)) {
        return scrollerElement.clientHeight - bottomOffset;
    }
    const scrollbarHeight = getScrollbarHeight(scrollerElement);
    return scrollerElement.getBoundingClientRect().bottom - (bottomOffset + scrollbarHeight);
};
/**
 * A helper function to make sure specified elem is visible in the specified container
 */
const _scrollRowIntoView = (rowElement, scrollerElement, viewportInfo, location) => {
    const verticalOverflowDiff = getVerticalOverflowDiff(rowElement, getScrollerTop(scrollerElement, viewportInfo.viewportConfig), getScrollerBottom(scrollerElement, viewportInfo.viewportConfig));
    const topOverflowDiff = verticalOverflowDiff.top;
    const hasTopOverflow = topOverflowDiff > 0;
    const bottomOverflowDiff = verticalOverflowDiff.bottom;
    const hasBottomOverflow = bottomOverflowDiff > 0;
    // don't adjust scroll position if row has overflow in both directions
    if (hasTopOverflow && hasBottomOverflow) {
        return;
    }
    if (location == null || location === 'inView') {
        // if row fits fully in viewport, scroll overflow side into view
        // otherwise, scroll the shortest amount to fill viewport with row
        if (hasBottomOverflow) {
            if (Math.abs(topOverflowDiff) > Math.abs(bottomOverflowDiff)) {
                scrollerElement.scrollTop = _roundScrollValue(scrollerElement.scrollTop + bottomOverflowDiff, true);
            }
            else {
                scrollerElement.scrollTop = _roundScrollValue(scrollerElement.scrollTop - topOverflowDiff, false);
            }
        }
        else if (hasTopOverflow) {
            if (Math.abs(bottomOverflowDiff) > Math.abs(topOverflowDiff)) {
                scrollerElement.scrollTop = _roundScrollValue(scrollerElement.scrollTop - topOverflowDiff, false);
            }
            else {
                scrollerElement.scrollTop = _roundScrollValue(scrollerElement.scrollTop + bottomOverflowDiff, true);
            }
        }
    }
    else if (location === 'top') {
        scrollerElement.scrollTop = _roundScrollValue(scrollerElement.scrollTop - topOverflowDiff, false);
    }
    else {
        scrollerElement.scrollTop = _roundScrollValue(scrollerElement.scrollTop + bottomOverflowDiff, true);
    }
};
/**
 * Helper method to round a desired scrollTop or scrollLeft value to the nearest valid value
 * depending on the client browser.
 */
const _roundScrollValue = (value, isOver) => {
    // use Math.floor() and Math.ceil() to ensure we round to nearest containing valid
    // pixel value in either direction as valid scrollTop values are only integers
    return isOver ? Math.ceil(value) : Math.floor(value);
};
/**
 * Helper to set the vertical scroll position on the List.
 */
function setScrollPosition(rootElement, scrollerElement, viewportInfo, listId, scrollPosition) {
    const y = _getScrollTopFromScrollPosition(rootElement, scrollerElement, viewportInfo, listId, scrollPosition);
    scrollerElement.scrollTop = y ?? 0;
}
/**
 * Helper to determine the 'y' value of a given scroll position.
 */
function _getScrollTopFromScrollPosition(rootElement, scrollerElement, viewportInfo, listId, scrollPosition) {
    if (scrollPosition != null) {
        let rowKey;
        let useRow = false;
        if (scrollPosition.key != null) {
            // rowKey takes precedence over everything else
            useRow = true;
            rowKey = scrollPosition.key;
        }
        else if (scrollPosition.offsetY != null && scrollPosition.y == null) {
            // offsetY can be used on its own if it is the only vertical aspect provided
            useRow = true;
            rowKey = getScrollPosition(rootElement, scrollerElement, viewportInfo, listId).key;
        }
        if (useRow) {
            if (rowKey != null) {
                const scrollerTop = getScrollerTop(scrollerElement, viewportInfo.viewportConfig);
                const offsetY = scrollPosition.offsetY ?? 0;
                const rowElements = rootElement.querySelectorAll(`[${LIST_ROW_SELECTOR}='${listId}']`);
                for (let i = 0; i < rowElements.length; i++) {
                    if (rowKey === getRowKey(rowElements[i])) {
                        const rowTop = rowElements[i].getBoundingClientRect().top;
                        return scrollerElement.scrollTop + rowTop - scrollerTop + offsetY;
                    }
                }
            }
        }
        else if (scrollPosition.y != null) {
            return scrollPosition.y;
        }
    }
    // either nothing was provided, or what was provided was invalid
    return undefined;
}
/**
 * Helper to get the logical scroll position for the given scrollTop value. If no scrollTop value is given,
 * it will use the current scrollTop value of the List.
 */
function getScrollPosition(rootElement, scrollerElement, viewportInfo, listId) {
    // ensure scroll position y value is initialized
    const scrollPosition = {
        y: Math.abs(scrollerElement.scrollTop),
        key: undefined,
        offsetY: undefined
    };
    const scrollerTop = getScrollerTop(scrollerElement, viewportInfo.viewportConfig);
    const rowElements = rootElement.querySelectorAll(`[${LIST_ROW_SELECTOR}='${listId}']`);
    for (let i = 0; i < rowElements.length; i++) {
        const rowRect = rowElements[i].getBoundingClientRect();
        if (rowRect.bottom > scrollerTop || i === rowElements.length - 1) {
            scrollPosition.key = getRowKeyForRowIndex(rootElement, listId, i);
            scrollPosition.offsetY = scrollerTop - rowRect.top;
            break;
        }
    }
    return scrollPosition;
}
function applyScrollDetail(rootElement, scrollerElement, viewportInfo, listId, scrollDetail) {
    if (scrollDetail.cell.type === 'data') {
        const rowElement = findElementByKey(rootElement, scrollDetail.cell.rowKey, `[${LIST_ROW_SELECTOR}='${listId}']`);
        if (rowElement != null && scrollDetail.locationY != null) {
            _scrollRowIntoView(rowElement, scrollerElement, viewportInfo, scrollDetail.locationY);
        }
    }
    else if (scrollDetail.cell.type === 'loadMore' || scrollDetail.cell.type === 'loading') {
        const cellElement = getElementFromCell(rootElement, listId, scrollDetail.cell);
        if (cellElement != null && scrollDetail.locationY != null) {
            _scrollRowIntoView(cellElement, scrollerElement, viewportInfo, scrollDetail.cell.type === 'loading' && scrollDetail.locationY === 'inView'
                ? 'top'
                : scrollDetail.locationY);
        }
    }
}
/**
 * Helper function to reset the scroll position of the scroller element
 */
const resetScrollPosition = (scrollerElement) => {
    scrollerElement.scrollTop = 0;
};
/**
 * Helper function to get the viewport config.
 */
const getViewportConfig = (scrollerRef, config) => {
    return config ?? { scrollerRef };
};

/**
 * Helper function to determine if a given active cell represents a 'transient' state.
 */
function isTransientCell(activeCell) {
    return activeCell != null && (activeCell.type === 'loadMore' || activeCell.type === 'loading');
}
/**
 * Helper function to compare two given cell descriptors.
 */
function cellComparator(cell1, cell2) {
    if (cell1 == null) {
        return cell2 == null;
    }
    if (cell2 == null) {
        return false;
    }
    return cell1.type === cell2.type && cell1.rowKey === cell2.rowKey;
}
function getCellContextMenuContextFromCell(rootElement, listId, listData, cell) {
    if (cell?.type === 'data') {
        const allMenuItems = {};
        const defaultMenuItems = undefined;
        for (const rowData of listData) {
            if (rowData.metadata.key === cell.rowKey) {
                return {
                    allMenuItems,
                    defaultMenuItems,
                    data: rowData.data,
                    metadata: { key: cell.rowKey },
                    index: getRowIndexForRowKey(rootElement, listId, cell.rowKey)
                };
            }
        }
    }
    return undefined;
}
/**
 * Helper function to determine the resulting cell from a cell override.
 */
function _getCellFromCellOverride(cellOverride) {
    let cellType;
    let cellRowKey;
    if (cellOverride?.rowKey != null) {
        cellType = 'data';
        cellRowKey = cellOverride.rowKey;
    }
    else {
        return undefined;
    }
    return { type: cellType, rowKey: cellRowKey };
}
/**
 * Helper function to determine the resulting cell type from a cell override.
 */
function _getCellTypeFromCellOverride(cellOverride) {
    if (cellOverride?.rowKey != null) {
        return 'data';
    }
    return 'data';
}
/**
 * Helper function to apply a current cell override if necessary.
 */
function applyCurrentCellOverride(rootElement, scrollerElement, viewportInfo, listId, currencyStatusRef, onActiveCellChanged, isEmbeddedSelect, isDataLoaded, currentItemOverride) {
    // ensure any new currentItemOverride settings are applied if necessary
    if (currencyStatusRef.current.hasPendingCurrentCellOverride ||
        currencyStatusRef.current.currentItemOverride !== currentItemOverride) {
        if (currentItemOverride != null) {
            // attempt to find the cell the override specifies
            const realizedCell = _getCellFromCellOverride(currentItemOverride);
            if (realizedCell != null) {
                // override cell was determined - apply the override immediately
                onActiveCellChanged({ value: realizedCell }, isAutoFocusCell(rootElement, listId, realizedCell), false, true);
                currencyStatusRef.current.currentItemOverride = currentItemOverride;
                currencyStatusRef.current.hasPendingCurrentCellOverride = false;
                if (isEmbeddedSelect) {
                    // scroll 'current-cell-override' vertically into view immediately for embedded select
                    // otherwise, cell will be scrolled into view once the component or cell receives focus
                    if (getElementFromCell(rootElement, listId, realizedCell) != null) {
                        applyScrollDetail(rootElement, scrollerElement, viewportInfo, listId, {
                            cell: realizedCell,
                            locationY: 'inView'
                        });
                        currencyStatusRef.current.hasPendingCurrentCellOverrideScroll = false;
                    }
                    else if (!isDataLoaded) {
                        // corresponding element was not found - data is not ready for scrolling
                        currencyStatusRef.current.hasPendingCurrentCellOverrideScroll = true;
                    }
                    else {
                        // corresponding element was not found - value was garbage
                        currencyStatusRef.current.hasPendingCurrentCellOverrideScroll = false;
                    }
                }
                else {
                    // not in an embedded select, so no scrolling should be performed
                    currencyStatusRef.current.hasPendingCurrentCellOverrideScroll = false;
                }
            }
            else if (!isDataLoaded) {
                // override cell could not be determined - data is still fetching
                currencyStatusRef.current.hasPendingCurrentCellOverride = true;
            }
            else {
                // override cell could not be determined - value was garbage
                currencyStatusRef.current.currentItemOverride = currentItemOverride;
                currencyStatusRef.current.hasPendingCurrentCellOverride = false;
                currencyStatusRef.current.hasPendingCurrentCellOverrideScroll = false;
            }
        }
        else {
            // override is set to clear the current cell
            onActiveCellChanged({ value: currentItemOverride });
            currencyStatusRef.current.currentItemOverride = currentItemOverride;
            currencyStatusRef.current.hasPendingCurrentCellOverride = false;
            if (isEmbeddedSelect) {
                // scroll back to the top for embedded select
                resetScrollPosition(scrollerElement);
            }
            currencyStatusRef.current.hasPendingCurrentCellOverrideScroll = false;
        }
    }
    else if (currencyStatusRef.current.hasPendingCurrentCellOverrideScroll) {
        // try and apply any pending scroll from a previous cell override at this time
        if (currencyStatusRef.current.currentCell != null) {
            if (getElementFromCell(rootElement, listId, currencyStatusRef.current.currentCell) != null) {
                applyScrollDetail(rootElement, scrollerElement, viewportInfo, listId, {
                    cell: currencyStatusRef.current.currentCell,
                    locationY: 'inView'
                });
                currencyStatusRef.current.hasPendingCurrentCellOverrideScroll = false;
            }
            else if (isDataLoaded) {
                // corresponding element was not found - value was garbage
                currencyStatusRef.current.hasPendingCurrentCellOverrideScroll = false;
            }
        }
        else {
            // there is no longer a current cell to try and scroll to
            currencyStatusRef.current.hasPendingCurrentCellOverrideScroll = false;
        }
    }
}
/**
 * Helper to determine if the given render state includes the cell specified.
 */
function _renderIncludesCell(cell, listData, hasMore) {
    if (listData.length > 0) {
        if (cell.type === 'loadMore') {
            return hasMore;
        }
        if (cell.type === 'data') {
            for (let i = 0; i < listData.length; i++) {
                if (listData[i].metadata.key === cell.rowKey) {
                    return true;
                }
            }
        }
    }
    else {
        return (cell.type === 'loading' && hasMore) || (cell.type === 'noData' && !hasMore);
    }
    return false;
}
/**
 * Helper to get the first cell of a given render.
 */
function _getFirstCell(listData) {
    if (listData.length > 0) {
        return { rowKey: listData[0].metadata.key, type: 'data' };
    }
    return { type: 'loading' };
}
/**
 * Helper to get the cell that should currently be a tab stop for the given render state.
 */
function determineActiveCell(currencyStatusRef, listData, hasMore, isEmbeddedSelect, currentItemOverride) {
    // when embedded select, use most recent non-null current cell value as override base
    const currentCell = isEmbeddedSelect
        ? currencyStatusRef.current.previousEmbeddedSelectCurrentCell
        : currencyStatusRef.current.currentCell;
    if (currencyStatusRef.current.hasPendingCurrentCellOverride ||
        currencyStatusRef.current.currentItemOverride !== currentItemOverride) {
        if (currentItemOverride != null) {
            // attempt to find the cell the override specifies
            const realizedCell = _getCellFromCellOverride(currentItemOverride);
            if (realizedCell != null && _renderIncludesCell(realizedCell, listData, hasMore)) {
                return realizedCell;
            }
            // if here, the resulting cell from the override could not be determined or was not valid
            const overrideCellType = _getCellTypeFromCellOverride(currentItemOverride);
            if (overrideCellType === 'data') {
                if (listData.length === 0 && hasMore) {
                    // pending 'data' override when 'loading' skeletons are shown
                    return { type: 'loading' };
                }
            }
            else if (overrideCellType === 'noData' && listData.length === 0 && hasMore) {
                // pending 'noData' override when 'loading' skeletons are shown
                return { type: 'loading' };
            }
        }
        return _getFirstCell(listData);
    }
    // if here, there is no current cell override that can be applied during this render
    const activeCell = currencyStatusRef.current.activeCell;
    if (activeCell != null) {
        if (activeCell.type === 'loadMore') {
            const initialData = currencyStatusRef.current.loadMoreInitialData;
            // if loadMore skeletons are no longer shown, or the data has changed, shift to
            // the next row in the data set if possible - otherwise just the closest row
            if (listData.length > 0 && listData !== initialData) {
                const rowKey = initialData != null && listData.length > initialData.length
                    ? listData[initialData.length].metadata.key
                    : listData[listData.length - 1].metadata.key;
                const cell = { type: 'data', rowKey: rowKey };
                if (_renderIncludesCell(cell, listData, hasMore)) {
                    return cell;
                }
            }
            else if (listData.length === 0) {
                return hasMore ? { type: 'loading' } : { type: 'noData' };
            }
        }
        if (_renderIncludesCell(activeCell, listData, hasMore)) {
            return activeCell;
        }
        else if (activeCell.type === 'loading') {
            if (currentCell != null && (currentCell.type === 'data' || currentCell.type === 'noData')) {
                // if the current cell is already set to 'data' or 'noData', and is now available, return it
                if (_renderIncludesCell(currentCell, listData, hasMore)) {
                    return currentCell;
                }
            }
            // otherwise, find the first row if available
            if (listData.length > 0) {
                const rowKey = listData[0].metadata.key;
                const cell = { type: 'data', rowKey: rowKey };
                if (_renderIncludesCell(cell, listData, hasMore)) {
                    return cell;
                }
            }
            else {
                return { type: 'noData' };
            }
        }
    }
    // if here, there is no active cell, or the active cell is no longer valid
    if (currentCell != null) {
        if (_renderIncludesCell(currentCell, listData, hasMore)) {
            return currentCell;
        }
        if (currentCell.type === 'data') {
            if (listData.length === 0 && hasMore) {
                // pending 'data' override when 'loading' skeletons are shown
                return { type: 'loading' };
            }
        }
        else if (currentCell.type === 'noData' && listData.length === 0 && hasMore) {
            // pending 'noData' override when 'loading' skeletons are shown
            return { type: 'loading' };
        }
    }
    if (listData.length === 0) {
        return hasMore ? { type: 'loading' } : { type: 'noData' };
    }
    return _getFirstCell(listData);
}

const TOUCH_OFFSET_VALUE = 40;
const menuPropGestureStates = {
    mouse: {
        initialFocus: 'menu',
        placement: 'bottom-start',
        offsetValue: 2
    },
    keyboard: {
        initialFocus: 'firstItem',
        placement: 'bottom-start',
        offsetValue: 0
    },
    touch: {
        initialFocus: 'menu',
        placement: 'end',
        offsetValue: TOUCH_OFFSET_VALUE
    }
};
const _initialMenuProps = {
    isOpen: false,
    initialFocus: 'menu',
    placement: 'bottom-start',
    offsetValue: 0,
    anchorRef: { current: null }
};
const useContextMenu = (rootRef, listId, listData, contextMenuConfig) => {
    const [contextMenuContext, setContextMenuContext] = useState();
    const [menuProps, setMenuProps] = useState(_initialMenuProps);
    const currentCellElementRef = useRef();
    const { triggerProps } = useContextMenuGesture(({ gesture, anchor, target }) => {
        if (rootRef.current) {
            // find the cell element the menu is being launched from
            const cellElement = target.closest(`[${LIST_FOCUSABLE_SELECTOR}='${listId}']`);
            if (cellElement != null) {
                const targetCell = logicalCellExtractor(cellElement, listId);
                if (targetCell != null) {
                    const cellContextMenuContext = getCellContextMenuContextFromCell(rootRef.current, listId, listData, targetCell);
                    if (cellContextMenuContext != null) {
                        currentCellElementRef.current = cellElement;
                        const anchorBasedOnGesture = gesture === 'keyboard' ? cellElement : anchor;
                        setContextMenuContext(cellContextMenuContext);
                        setMenuProps({
                            ...menuPropGestureStates[gesture],
                            anchorRef: { current: anchorBasedOnGesture },
                            isOpen: true
                        });
                    }
                }
            }
        }
    }, { isDisabled: !contextMenuConfig });
    const handleCloseContextMenu = useCallback((detail) => {
        // set focus back on the cell that was determined as the initial target
        if (detail.reason === 'dismissed' || detail.reason === 'itemAction') {
            currentCellElementRef.current?.focus({ preventScroll: true });
        }
        // close the menu
        setMenuProps({ ..._initialMenuProps });
    }, []);
    return {
        menuProps: { ...menuProps, onClose: handleCloseContextMenu },
        contextMenuContext: contextMenuContext,
        contextMenuTriggerProps: triggerProps
    };
};

// constant timeout values for delays following a input gesture until enabling focus rings
const MOUSE_FOCUS_SHIFT_TIMEOUT = 100;
const TOUCH_FOCUS_SHIFT_TIMEOUT = 200;
// constant event.key values that trigger focus ring logic
const NAVIGATION_KEYS = [
    ' ',
    'F2',
    'Enter',
    'Esc', // Firefox 36 and earlier uses 'Esc' instead of 'Escape'
    'Escape',
    'ArrowLeft',
    'ArrowUp',
    'ArrowRight',
    'ArrowDown',
    'Home',
    'End',
    'PageUp',
    'PageDown'
];
// constant event.key values that trigger 'scroll to current' logic
const SCROLL_KEYS = [
    ' ',
    'F2',
    'Enter',
    'Esc', // Firefox 36 and earlier uses 'Esc' instead of 'Escape'
    'Escape'
];
/**
 * Hook that manages focus interactions on the Preact List
 */
function useFocusHandling({ rootRef, viewportInfo, listId, listData, hasMore, selectionExtensionElementRef, currentItemOverride, onPersistCurrentItem, isEmbedded, itemEnterKeyFocusBehavior, isDataLoaded, reorderContext, deferredScrollDetail, deferredScrollDetailRef, updateDeferredScrollDetail }) {
    // tracks whether the most recent 'blur' was caused by focus being lost by the browser window (or iframe)
    // used when determining whether scrolling active into view should be skipped when focus is regained
    const blurFromWindowRef = useRef(false);
    // tracks whether a keydown 'tab' is currently detected - used when handling a keyup 'tab' event to scroll
    // the active cell into view in the event that focus is received from outside of the page (or iframe)
    const isTabDownDetectedRef = useRef(false);
    // tracks focus loss during render cycle for potential focus recapture cases
    const isRenderCycle = useRef(true);
    isRenderCycle.current = true;
    const blurredRenderCycleElementRef = useRef(null);
    // track current 'focus info' state
    const hasFocusRef = useRef(false);
    const [isTabbableMode, setIsTabbableMode] = useState(false);
    const [isShowFocusRing, setIsShowFocusRing] = useState(false);
    // track pointer down information
    const recentPointerRef = useRef(false);
    const pointerTimerRef = useRef(null);
    const isShowFocusRingOverrideRef = useRef(null);
    // track the active and current cell information
    const [activeCell, setActiveCell] = useState(undefined);
    const currencyStatusRef = useRef({
        hasPendingCurrentCellOverride: false,
        hasPendingCurrentCellOverrideScroll: false
    });
    const wasTransient = isTransientCell(currencyStatusRef.current.activeCell);
    const currentActiveCell = determineActiveCell(currencyStatusRef, listData, hasMore, isEmbedded, currentItemOverride);
    // intercept onPersistCurrentItem calls to ensure internal state is updated appropriately
    const onActiveCellChanged = useMemo(() => {
        return (detail, isAutoFocus, isSelectionExtension, skipScrollCellIntoView) => {
            const newActiveCell = detail.value;
            if (!cellComparator(currencyStatusRef.current.activeCell, newActiveCell)) {
                setActiveCell(newActiveCell);
                currencyStatusRef.current.activeCell = newActiveCell;
                if (!isTransientCell(newActiveCell)) {
                    const newCurrentCell = newActiveCell;
                    // only update current cell when 'active' is not one of the transient 'skeleton' regions
                    if (onPersistCurrentItem != null && newCurrentCell.rowKey != null) {
                        onPersistCurrentItem({ value: newCurrentCell.rowKey });
                    }
                    currencyStatusRef.current.currentCell = newCurrentCell;
                    // make sure pending current cell override states are updated on current changed
                    if (newCurrentCell != null) {
                        currencyStatusRef.current.previousEmbeddedSelectCurrentCell = newCurrentCell;
                    }
                    currencyStatusRef.current.currentItemOverride = currentItemOverride;
                    currencyStatusRef.current.hasPendingCurrentCellOverride = false;
                    currencyStatusRef.current.hasPendingCurrentCellOverrideScroll = false;
                }
                else if (newActiveCell?.type === 'loadMore') {
                    currencyStatusRef.current.loadMoreInitialData = listData;
                }
                if (!isSelectionExtension) {
                    selectionExtensionElementRef.current = undefined;
                    if (isAutoFocus && newActiveCell != null) {
                        // attempt to auto-focus cell contents
                        setIsTabbableMode(true);
                        setIsShowFocusRing(false);
                    }
                    else if (isTabbableMode) {
                        // exit tabbable mode when navigating away from auto-focus cell
                        setIsTabbableMode(false);
                    }
                }
            }
            if (newActiveCell != null && !recentPointerRef.current && !skipScrollCellIntoView) {
                updateDeferredScrollDetail({
                    cell: newActiveCell,
                    locationX: 'inView',
                    locationY: 'inView'
                });
            }
        };
    }, [
        currentItemOverride,
        isTabbableMode,
        listData,
        onPersistCurrentItem,
        selectionExtensionElementRef,
        updateDeferredScrollDetail
    ]);
    // Helper function to set the List's focus tracking state to reflect that it is inactive.
    const _setAsInactive = useCallback(() => {
        hasFocusRef.current = false;
        setIsTabbableMode(false);
        setIsShowFocusRing(false);
    }, []);
    // Helper function to set the List's focus tracking state to reflect that one of its cells has focus.
    const _enableNavigationMode = useCallback((cell, skipScrollCellIntoView) => {
        const rootElement = rootRef.current;
        const scrollerElement = viewportInfo.viewportConfig.scrollerRef.current;
        if (rootElement != null && scrollerElement != null) {
            setIsTabbableMode(false);
            const focusRingState = isShowFocusRingOverrideRef.current !== null
                ? isShowFocusRingOverrideRef.current
                : !recentPointerRef.current;
            setIsShowFocusRing(focusRingState);
            onActiveCellChanged({ value: cell }, false, false, skipScrollCellIntoView);
        }
    }, [rootRef, viewportInfo, onActiveCellChanged]);
    // Helper function to set the List's focus tracking state to reflect that it is tabbable.
    const _enableTabbableMode = useCallback((cell, isKeyboard, skipScrollCellIntoView) => {
        const rootElement = rootRef.current;
        const scrollerElement = viewportInfo.viewportConfig.scrollerRef.current;
        if (rootElement != null && scrollerElement != null) {
            setIsTabbableMode(true);
            if (!isKeyboard) {
                setIsShowFocusRing(false);
                onActiveCellChanged({ value: cell }, true, false, skipScrollCellIntoView);
            }
        }
    }, [rootRef, viewportInfo, onActiveCellChanged]);
    // Helper function to set the List's focus tracking state to reflect that a user has
    // exited tabbable mode via the keyboard (esc or 'F2')
    const _disableTabbableMode = useCallback((cell) => {
        const rootElement = rootRef.current;
        if (rootElement != null) {
            setIsTabbableMode(false);
            const cellElement = getElementFromCell(rootElement, listId, cell);
            if (cellElement != null) {
                cellElement.focus({ preventScroll: true });
            }
            else {
                rootElement.focus({ preventScroll: true });
            }
        }
    }, [rootRef, listId]);
    const notifyFocusFromPointer = useCallback((isTouch) => {
        if (pointerTimerRef.current != null) {
            clearTimeout(pointerTimerRef.current);
        }
        recentPointerRef.current = true;
        pointerTimerRef.current = setTimeout(() => {
            recentPointerRef.current = false;
        }, isTouch ? TOUCH_FOCUS_SHIFT_TIMEOUT : MOUSE_FOCUS_SHIFT_TIMEOUT);
    }, []);
    // PointerDown handler for the outer List. Focus transfers that occur following a pointerdown
    // on desktop devices should not result in a focus ring being shown on the focused element
    const _pointerDownHandler = useCallback((event) => {
        if (isShowFocusRing) {
            setIsShowFocusRing(false);
        }
        if (event.pointerType !== 'touch') {
            notifyFocusFromPointer(false);
        }
    }, [isShowFocusRing, notifyFocusFromPointer]);
    // PointerUp handler for the outer List. Focus transfers that occur following a pointerup
    // on any type of device should not result in a focus ring being shown on the focused element
    const _pointerUpHandler = useCallback((event) => {
        notifyFocusFromPointer(event.pointerType === 'touch');
    }, [notifyFocusFromPointer]);
    // Focus handler for the outer List. On focus, a focusable area needs to be 'active'.
    // If an 'active' area already exists, nothing further is needed. Otherwise, the first
    // focusable element should be made 'active'.
    const _onFocusHandler = (event) => {
        hasFocusRef.current = true;
        const rootElement = rootRef.current;
        const scrollerElement = viewportInfo.viewportConfig.scrollerRef.current;
        if (rootElement != null) {
            // don't scroll active into view if previous blur was due to window focus loss
            const skipScrollCellIntoView = blurFromWindowRef.current;
            blurFromWindowRef.current = false;
            const targetElement = event.target;
            if (targetElement === rootElement || targetElement === scrollerElement) {
                if (currentActiveCell != null) {
                    _enableNavigationMode(currentActiveCell, skipScrollCellIntoView);
                }
            }
            else {
                const cell = logicalCellExtractor(targetElement, listId);
                if (cell != null) {
                    if (getCellType(targetElement) != null) {
                        // handle individual focusable regions receiving focus
                        _enableNavigationMode(cell, skipScrollCellIntoView);
                    }
                    else {
                        _enableTabbableMode(cell, false, skipScrollCellIntoView);
                    }
                }
            }
        }
        isShowFocusRingOverrideRef.current = null;
    };
    // Blur handler for the outer List. When a blur is caused by the browser window itself
    // losing focus (changing tabs or leaving an iframe for example), an eventual re-focus due
    // to the window re-gaining focus needs to be handled differently (no auto-scroll mostly).
    const _onBlurHandler = useCallback((event) => {
        const rootElement = rootRef.current;
        if (rootElement != null) {
            if (event.target === document.activeElement) {
                // save focus ring state so it's set once focus comes back
                isShowFocusRingOverrideRef.current = isShowFocusRing;
            }
            const relatedTarget = event.relatedTarget;
            const isUnknownRelatedTarget = relatedTarget == null;
            blurFromWindowRef.current = isUnknownRelatedTarget;
            if (isUnknownRelatedTarget || !getElementContainsFunc(rootElement, true)(relatedTarget)) {
                _setAsInactive();
                if (isRenderCycle.current) {
                    // store element that loses focus during a render cycle for recapturing focus logic
                    blurredRenderCycleElementRef.current = event.target;
                }
                // reset tab down detection when component loses focus
                isTabDownDetectedRef.current = false;
            }
        }
    }, [rootRef, isShowFocusRing, _setAsInactive]);
    // KeyDown handler for the outer List. This specific handler tracks 'focus' transfer keys
    // Esc, F2, Enter, and Tab. It also ensures the active cell is scrolled into view on any key
    const _onKeyDownHandler = useCallback((event) => {
        if (event.key === 'Tab') {
            isTabDownDetectedRef.current = true;
        }
        const rootElement = rootRef.current;
        const scrollerElement = viewportInfo.viewportConfig.scrollerRef.current;
        if (rootElement != null && scrollerElement != null) {
            const isNavigationKey = NAVIGATION_KEYS.indexOf(event.key) > -1;
            if (isNavigationKey) {
                if (pointerTimerRef.current != null) {
                    clearTimeout(pointerTimerRef.current);
                }
                pointerTimerRef.current = null;
                recentPointerRef.current = false;
            }
            if (activeCell != null) {
                if (event.key === 'F2') {
                    if (isTabbableMode) {
                        _disableTabbableMode(activeCell);
                    }
                    else {
                        _enableTabbableMode(activeCell, true);
                    }
                }
                else if (event.key === 'Esc' || event.key === 'Escape') {
                    // Firefox 36 and earlier uses 'Esc' instead of 'Escape'
                    _disableTabbableMode(activeCell);
                }
                else if (event.key === 'Enter' &&
                    !isTabbableMode &&
                    !isEmbedded &&
                    itemEnterKeyFocusBehavior === 'focusWithin') {
                    _enableTabbableMode(activeCell, true);
                }
                if (!isTabbableMode) {
                    if (isNavigationKey) {
                        setIsShowFocusRing(true);
                    }
                    if (SCROLL_KEYS.indexOf(event.key) > -1) {
                        updateDeferredScrollDetail({
                            cell: activeCell,
                            locationX: 'inView',
                            locationY: 'inView'
                        });
                    }
                }
            }
        }
    }, [
        rootRef,
        viewportInfo,
        isTabbableMode,
        isEmbedded,
        activeCell,
        _enableTabbableMode,
        _disableTabbableMode,
        updateDeferredScrollDetail,
        itemEnterKeyFocusBehavior
    ]);
    // KeyUp handler for the outer Table. This specific handler tracks the 'tab' key
    // It ensures the active cell is scrolled into view when tabbing into the component
    const _onKeyUpHandler = useCallback((event) => {
        if (event.key === 'Tab') {
            // if a keyup 'tab' event is detected without a matching keydown 'tab', scroll active
            // cell into view to handle case where focus is gained from outside the page (or iframe)
            if (activeCell != null && !isTabDownDetectedRef.current) {
                updateDeferredScrollDetail({
                    cell: activeCell,
                    locationX: 'inView',
                    locationY: 'inView'
                });
            }
            isTabDownDetectedRef.current = false;
        }
    }, [activeCell, updateDeferredScrollDetail]);
    // Callback for the outer List's FocusTracker usage.
    const onStartFocusTracking = useCallback((detail) => {
        const rootElement = rootRef.current;
        if (rootElement != null && activeCell != null) {
            handleEnterActionableMode(detail.hasFocusWithin && getCellType(detail.activeElement) == null, allTabbableElements(getElementFromCell(rootElement, listId, activeCell)), (value) => {
                setIsTabbableMode(value != undefined);
            });
        }
    }, [rootRef, activeCell, listId]);
    // Callback for the outer List's FocusTracker usage.
    const onFocusStartEdge = useCallback((detail) => {
        handleWrapActionableFocus(true, detail.tabbableElements, (value) => {
            setIsTabbableMode(value != undefined);
        });
    }, []);
    // Callback for the outer List's FocusTracker usage.
    const onFocusEndEdge = useCallback((detail) => {
        handleWrapActionableFocus(false, detail.tabbableElements, (value) => {
            setIsTabbableMode(value != undefined);
        });
    }, []);
    // Ensure focus is properly set at the completion of every render cycle
    // NO DEPENDENCY ARRAY HERE TO ENSURE THIS RUNS ON EVERY RENDER CYCLE
    useEffect(() => {
        // make sure the reorder currentItemKey is always up to date
        if (reorderContext != null) {
            reorderContext.currentItemKey.current = activeCell?.rowKey;
        }
        const rootElement = rootRef.current;
        const scrollerElement = viewportInfo.viewportConfig.scrollerRef.current;
        if (rootElement != null && scrollerElement != null) {
            // internal method responsible for ensuring focus is applied to the appropriate cells or contents
            const _applyInternalFocus = (rootElement, activeElement, isCaptureFocus) => {
                if (currentActiveCell != null) {
                    // ensure real browser focus is on the correct cell element
                    const cellElement = getElementFromCell(rootElement, listId, currentActiveCell);
                    if (cellElement != null) {
                        if (isCaptureFocus) {
                            applyScrollDetail(rootElement, scrollerElement, viewportInfo, listId, {
                                cell: currentActiveCell,
                                locationX: 'inView',
                                locationY: 'inView'
                            });
                        }
                        // cell is valid, so set focus on it if needed
                        if (!isTabbableMode && activeElement !== cellElement) {
                            if (wasTransient &&
                                !isTransientCell(currentActiveCell) &&
                                isAutoFocusCell(rootElement, listId, currentActiveCell)) {
                                _enableTabbableMode(currentActiveCell);
                            }
                            else {
                                cellElement.focus({ preventScroll: true });
                            }
                            return;
                        }
                        else if (isTabbableMode &&
                            !getElementContainsFunc(cellElement, true)(activeElement)) {
                            if (isCaptureFocus) {
                                _disableTabbableMode(currentActiveCell);
                            }
                            else {
                                handleEnterActionableMode(false, allTabbableElements(cellElement), (value) => {
                                    setIsTabbableMode(value != undefined);
                                    if (isCaptureFocus) {
                                        rootElement.focus({ preventScroll: true });
                                    }
                                }, true);
                            }
                            return;
                        }
                        else ;
                    }
                }
                if (isCaptureFocus) {
                    rootElement.focus({ preventScroll: true });
                }
            };
            // ensure the current cell is scrolled into view when necessary
            // only run if the state and ref match - otherwise another render cycle may be pending
            if (deferredScrollDetail != null &&
                deferredScrollDetail === deferredScrollDetailRef.current) {
                applyScrollDetail(rootElement, scrollerElement, viewportInfo, listId, deferredScrollDetail);
                deferredScrollDetailRef.current = undefined;
            }
            // ensure current-cell-override is up to date on each render
            applyCurrentCellOverride(rootElement, scrollerElement, viewportInfo, listId, currencyStatusRef, onActiveCellChanged, isEmbedded, isDataLoaded, currentItemOverride);
            const blurredRenderCycleElement = blurredRenderCycleElementRef.current;
            blurredRenderCycleElementRef.current = null;
            isRenderCycle.current = false;
            const activeElement = getActiveElement(rootElement);
            // if we were tracking focus during the render cycle, ensure we recapture focus if needed
            if (blurredRenderCycleElement != null &&
                activeElement === getBodyElement(rootElement) &&
                (!getElementContainsFunc(rootElement, true)(blurredRenderCycleElement) ||
                    !isTabbableElement(blurredRenderCycleElement, true))) {
                deferredScrollDetailRef.current = undefined;
                _applyInternalFocus(rootElement, activeElement, true);
            }
            else if (hasFocusRef.current &&
                cellComparator(activeCell, currencyStatusRef.current.activeCell)) {
                _applyInternalFocus(rootElement, activeElement);
            }
        }
    });
    return {
        focusHandlingProps: {
            onPointerDown: _pointerDownHandler,
            onPointerUp: _pointerUpHandler,
            onFocus: _onFocusHandler,
            onBlur: _onBlurHandler,
            onKeyDown: _onKeyDownHandler,
            onKeyUp: _onKeyUpHandler
        },
        activeCell,
        onActiveCellChanged,
        currentCell: currencyStatusRef.current.currentCell,
        currentActiveCell,
        isShowFocusRing,
        isTabbableMode,
        onStartFocusTracking,
        onFocusStartEdge,
        onFocusEndEdge,
        updateDeferredScrollDetail
    };
}

const useItemReorder = ({ rootRef, scrollerRef, onReorder, itemSelector }) => {
    const isReorderEnabled = onReorder != undefined;
    const [dragKey, setDragKey] = useState(undefined);
    const { reorderProps, reorderContext, reorderInstructionsId, reorderInstructions } = useReorderable({
        onReorder,
        rootRef,
        itemSelector,
        isDisabled: !isReorderEnabled,
        columns: 1
    });
    const reorderItemProps = useMemo(() => {
        return { dragKey, setDragKey, onReorder, rootRef };
    }, [dragKey, onReorder, rootRef]);
    if (!isReorderEnabled) {
        return {
            reorderProps: {},
            reorderKeyboardProps: reorderProps,
            reorderContext,
            reorderInstructionsId,
            reorderInstructions
        };
    }
    const handleDragOver = (event) => {
        if (dragKey !== undefined) {
            dragToScroll(event, dragKey !== undefined, scrollerRef.current);
        }
    };
    return {
        reorderProps: { onDragOver: handleDragOver },
        reorderKeyboardProps: reorderProps,
        reorderItemProps,
        reorderContext,
        reorderInstructionsId,
        reorderInstructions
    };
};

/**
 * @license
 * Copyright (c) %FIRST_YEAR% %CURRENT_YEAR%, Oracle and/or its affiliates.
 * The Universal Permissive License (UPL), Version 1.0
 * as shown at https://oss.oracle.com/licenses/upl/
 * @ignore
 */
/**
 * Helper function to get the item action context from row information
 */
const getActionContext = (index, key, data) => {
    return { metadata: { key }, data, index };
};

/**
 * Helper function to handle an 'Up' or 'Down' arrow key press.
 * Up - Moves to the same cell in the previous row (including headers and footers). No op if already on the first row.
 * Down - Moves to the same cell in the next row (including headers and footers). No op if already on the last row.
 */
function handleAdjacentRowGesture(rootElement, listId, initialCell, onActiveCellChanged, isPrevious, isExtendSelectionGesture, updateDeferredScrollDetail) {
    let newCell = initialCell;
    const cellType = initialCell.type;
    if (cellType === 'loadMore') {
        if (isPrevious) {
            const rowKey = getLastRenderedRowKey(rootElement, listId);
            if (rowKey != null) {
                newCell = { rowKey: rowKey, type: 'data' };
                onActiveCellChanged({ value: newCell }, isAutoFocusCell(rootElement, listId, newCell));
            }
        }
    }
    else {
        if (cellType === 'data') {
            const currentRowKey = initialCell.rowKey;
            const currentRowElement = findElementByKey(rootElement, currentRowKey, `[${LIST_ROW_SELECTOR}='${listId}']`);
            if (currentRowElement != null) {
                const rowElements = [...rootElement.querySelectorAll(`[${LIST_ROW_SELECTOR}='${listId}']`)];
                const currentRowIndex = rowElements.indexOf(currentRowElement);
                let adjacentRowElement;
                if (isPrevious && currentRowIndex > 0) {
                    adjacentRowElement = rowElements[currentRowIndex - 1];
                }
                else if (!isPrevious && currentRowIndex < rowElements.length - 1) {
                    adjacentRowElement = rowElements[currentRowIndex + 1];
                }
                if (adjacentRowElement != null) {
                    const rowKey = getRowKey(adjacentRowElement);
                    if (rowKey != null) {
                        newCell = { rowKey: rowKey, type: 'data' };
                        onActiveCellChanged({ value: newCell }, isAutoFocusCell(rootElement, listId, newCell), isExtendSelectionGesture);
                    }
                }
                else if (!isExtendSelectionGesture &&
                    !isPrevious &&
                    hasLoadMoreCell(rootElement, listId)) {
                    newCell = { type: 'loadMore' };
                    if (getElementFromCell(rootElement, listId, newCell) != null) {
                        onActiveCellChanged({ value: newCell }, isAutoFocusCell(rootElement, listId, newCell));
                    }
                }
            }
        }
    }
    if (cellComparator(initialCell, newCell)) {
        updateDeferredScrollDetail({ cell: newCell, locationX: 'inView', locationY: 'inView' });
    }
}
/**
 * Helper function to handle a 'PageUp' or 'PageDown' arrow key press.
 * PageUp - Moves to the same cell in the first visible row in the current viewport,
 * and scrolls the List so it becomes the last visible row in the viewport when able.
 * PageDown - Moves to the same cell in the last visible row in the current viewport,
 * and scrolls the List so it becomes the first visible row in the viewport when able.
 */
function handlePageRowGesture(rootElement, scrollerElement, viewportInfo, listId, initialCell, onActiveCellChanged, isPrevious, updateDeferredScrollDetail) {
    if (initialCell.type !== 'noData') {
        const rowElements = rootElement.querySelectorAll(`[${LIST_ROW_SELECTOR}='${listId}']`);
        if (rowElements.length > 0) {
            const scrollPosition = getScrollPosition(rootElement, scrollerElement, viewportInfo, listId);
            const verticalRowIndex = getRowIndexForRowKey(rootElement, listId, scrollPosition.key);
            let targetRowIndex;
            if (isPrevious) {
                if (scrollPosition.offsetY === 0) {
                    targetRowIndex = Math.max(verticalRowIndex - 1, 0);
                }
                else {
                    targetRowIndex = verticalRowIndex;
                }
            }
            else {
                const scrollerBottom = getScrollerBottom(scrollerElement, viewportInfo.viewportConfig);
                targetRowIndex = verticalRowIndex;
                while (rowElements.length > targetRowIndex + 1) {
                    targetRowIndex += 1;
                    const elementBottom = rowElements[targetRowIndex].getBoundingClientRect().bottom;
                    if (elementBottom > scrollerBottom) {
                        break;
                    }
                }
            }
            const newRowKey = getRowKeyForRowIndex(rootElement, listId, targetRowIndex);
            if (newRowKey != null) {
                const newCell = {
                    rowKey: newRowKey,
                    type: 'data'
                };
                if (newRowKey !== initialCell.rowKey) {
                    onActiveCellChanged({ value: newCell }, isAutoFocusCell(rootElement, listId, newCell), false, true);
                }
                updateDeferredScrollDetail({
                    cell: newCell,
                    locationX: 'inView',
                    locationY: isPrevious ? 'bottom' : 'top'
                });
            }
        }
    }
}
function isReorderGesture(event) {
    return event.shiftKey && isLogicalCtrl(event);
}

/**
 * Helper function to handle a selection gesture (click, spacebar, etc...)
 */
function handleSelectionGesture(listId, targetElement, isModifierKey, onSelectionChange, isDropDownSelection, selectionMode, selectedKeys) {
    const emptyKeys = { all: false, keys: new Set() };
    let selectedRows = selectedKeys != null ? selectedKeys : emptyKeys;
    const newRowKey = keyExtractor(targetElement, `[${LIST_ROW_SELECTOR}='${listId}']`);
    if (newRowKey != null) {
        if (selectionMode === 'single' ||
            selectionMode === 'multiple' ||
            selectionMode === 'multipleToggle') {
            const isToggleSelectionMode = isDropDownSelection || selectionMode === 'multipleToggle';
            const isMultiSelectRowGesture = isModifierKey || isToggleSelectionMode;
            if (selectedRows == null) {
                selectedRows = { all: false, keys: new Set([newRowKey]) };
            }
            else if (!containsKey(selectedRows, newRowKey)) {
                if (isMultiSelectRowGesture && selectionMode !== 'single') {
                    selectedRows = addKey(selectedRows, newRowKey);
                }
                else {
                    selectedRows = { all: false, keys: new Set([newRowKey]) };
                }
            }
            else if (isMultiSelectRowGesture) {
                selectedRows = removeKey(selectedRows, newRowKey, false);
            }
            else {
                selectedRows = { all: false, keys: new Set([newRowKey]) };
            }
        }
    }
    if (!isSameKey(selectedRows, selectedKeys)) {
        onSelectionChange({ value: selectedRows, target: targetElement }, false);
    }
}
/**
 * Helper function to calculate and apply a new selection state corresponding to the range provided
 */
function _applyRangeSelection(rootElement, listId, targetElement, startIndex, endIndex, isRowGesture, onSelectionChange, isDropDownSelection, selectionMode, selectedKeys, previousEndIndex) {
    const emptyKeys = { all: false, keys: new Set() };
    let selectedRows = selectedKeys != null ? selectedKeys : emptyKeys;
    if (isRowGesture) {
        if (selectionMode === 'multiple' || selectionMode === 'multipleToggle') {
            const isToggleSelectionMode = isDropDownSelection || selectionMode === 'multipleToggle';
            selectedRows = isToggleSelectionMode ? selectedRows : emptyKeys;
            // selected rows should be added in the order that they are selected
            if (startIndex <= endIndex) {
                for (let i = startIndex; i <= endIndex; i++) {
                    const key = getRowKeyForRowIndex(rootElement, listId, i);
                    selectedRows = addKey(selectedRows, key);
                }
                // toggle selection gesture is additive except for rows added from previous gesture
                if (isToggleSelectionMode && previousEndIndex != null) {
                    if (previousEndIndex > endIndex) {
                        for (let i = previousEndIndex; i > endIndex; i--) {
                            const key = getRowKeyForRowIndex(rootElement, listId, i);
                            selectedRows = removeKey(selectedRows, key, false);
                        }
                    }
                    else if (previousEndIndex < startIndex) {
                        for (let i = previousEndIndex; i < startIndex; i++) {
                            const key = getRowKeyForRowIndex(rootElement, listId, i);
                            selectedRows = removeKey(selectedRows, key, false);
                        }
                    }
                }
            }
            else {
                for (let i = startIndex; i >= endIndex; i--) {
                    const key = getRowKeyForRowIndex(rootElement, listId, i);
                    selectedRows = addKey(selectedRows, key);
                }
                // toggle selection gesture is additive except for rows added from previous gesture
                if (isToggleSelectionMode && previousEndIndex != null) {
                    if (previousEndIndex < endIndex) {
                        for (let i = previousEndIndex; i < endIndex; i++) {
                            const key = getRowKeyForRowIndex(rootElement, listId, i);
                            selectedRows = removeKey(selectedRows, key, false);
                        }
                    }
                    else if (previousEndIndex > startIndex) {
                        for (let i = previousEndIndex; i > startIndex; i--) {
                            const key = getRowKeyForRowIndex(rootElement, listId, i);
                            selectedRows = removeKey(selectedRows, key, false);
                        }
                    }
                }
            }
        }
    }
    if (!isSameKey(selectedRows, selectedKeys)) {
        onSelectionChange({ value: selectedRows, target: targetElement }, true);
    }
}
/**
 * Determines whether a potential range selection gesture is valid based on the given target
 */
function isRangeSelectionGesture(rootElement, listId, targetElement, activeCell, selectionMode) {
    const targetRowElement = rowExtractor(targetElement, listId);
    if ((selectionMode === 'multiple' || selectionMode === 'multipleToggle') &&
        activeCell.type === 'data' &&
        targetRowElement != null) {
        const activeCellElement = getElementFromCell(rootElement, listId, activeCell);
        if (activeCellElement != null && targetRowElement != null) {
            return (getRowIndexForRowElement(rootElement, listId, activeCellElement.parentElement) != null && getRowIndexForRowElement(rootElement, listId, targetRowElement) != null);
        }
    }
    return false;
}
/**
 * Helper function to handle a range selection gesture (shift+click)
 */
function handleRangeSelectionGesture(rootElement, listId, targetElement, activeCell, onActiveCellChanged, selectionExtensionElementRef, onSelectionChange, isDropDownSelection, selectionMode, selectedKeys) {
    const targetRowElement = rowExtractor(targetElement, listId);
    if (targetRowElement != null) {
        const activeCellElement = getElementFromCell(rootElement, listId, activeCell);
        if (activeCellElement != null) {
            if ((selectionMode === 'multiple' || selectionMode === 'multipleToggle') &&
                activeCell.type === 'data') {
                let previousEndIndex;
                const isToggleSelectionMode = isDropDownSelection || selectionMode === 'multipleToggle';
                const endElement = targetRowElement;
                const startElement = isToggleSelectionMode && selectionExtensionElementRef.current != null
                    ? selectionExtensionElementRef.current
                    : activeCellElement.parentElement;
                const startIndex = getRowIndexForRowElement(rootElement, listId, startElement);
                const endIndex = getRowIndexForRowElement(rootElement, listId, endElement);
                const previousElement = isToggleSelectionMode
                    ? activeCellElement.parentElement
                    : selectionExtensionElementRef.current;
                if (previousElement != null) {
                    previousEndIndex = getRowIndexForRowElement(rootElement, listId, previousElement);
                }
                if (startIndex != null && endIndex != null) {
                    if (!isToggleSelectionMode) {
                        selectionExtensionElementRef.current = endElement;
                    }
                    else {
                        if (selectionExtensionElementRef.current == null) {
                            selectionExtensionElementRef.current = activeCellElement.parentElement;
                        }
                        let targetCell = logicalCellExtractor(targetElement, listId);
                        if (targetCell == null) {
                            targetCell = getFirstFocusableCell(targetRowElement, listId);
                        }
                        if (targetCell != null) {
                            onActiveCellChanged({ value: targetCell }, false, true);
                            const targetCellElement = getElementFromCell(rootElement, listId, targetCell);
                            targetCellElement?.focus({ preventScroll: true });
                        }
                    }
                    _applyRangeSelection(rootElement, listId, targetElement, startIndex, endIndex, true, onSelectionChange, isDropDownSelection, selectionMode, selectedKeys, previousEndIndex);
                    return true;
                }
            }
        }
    }
    return false;
}
/**
 * Helper function to handle a selection extension gesture (shift+arrow key)
 */
function handleExtendSelectionGesture(rootElement, listId, targetElement, isPrevious, activeCell, selectionExtensionElementRef, onSelectionChange, isDropDownSelection, updateDeferredScrollDetail, selectionMode, selectedKeys) {
    const activeCellElement = getElementFromCell(rootElement, listId, activeCell);
    if (activeCellElement != null) {
        let startIndex;
        let startElement;
        const isToggleSelectionMode = isDropDownSelection || selectionMode === 'multipleToggle';
        const isRow = activeCell.type === 'data';
        if (isRow) {
            startElement =
                isToggleSelectionMode && selectionExtensionElementRef.current != null
                    ? selectionExtensionElementRef.current
                    : activeCellElement.parentElement;
            startIndex = getRowIndexForRowElement(rootElement, listId, startElement);
        }
        if (startIndex != null) {
            let endIndex;
            let previousEndIndex;
            let extendedElement;
            const cursorElement = isToggleSelectionMode && isRow
                ? activeCellElement.parentElement
                : selectionExtensionElementRef.current;
            if (cursorElement != null) {
                previousEndIndex = getRowIndexForRowElement(rootElement, listId, cursorElement);
                extendedElement = isPrevious
                    ? cursorElement.previousElementSibling
                    : cursorElement.nextElementSibling;
                if (extendedElement != null && (!isRow || (isRow && getRowKey(extendedElement) != null))) {
                    if (!isToggleSelectionMode || !isRow) {
                        selectionExtensionElementRef.current = extendedElement;
                    }
                    else if (selectionExtensionElementRef.current == null) {
                        selectionExtensionElementRef.current = activeCellElement.parentElement;
                    }
                    endIndex = getRowIndexForRowElement(rootElement, listId, extendedElement);
                }
                else {
                    endIndex = getRowIndexForRowElement(rootElement, listId, cursorElement);
                }
            }
            else {
                extendedElement = isPrevious
                    ? startElement?.previousElementSibling
                    : startElement?.nextElementSibling;
                if (extendedElement != null && (!isRow || (isRow && getRowKey(extendedElement) != null))) {
                    selectionExtensionElementRef.current = extendedElement;
                    endIndex = getRowIndexForRowElement(rootElement, listId, extendedElement);
                }
                else {
                    endIndex = startIndex;
                }
            }
            if (endIndex != null) {
                _applyRangeSelection(rootElement, listId, targetElement, startIndex, endIndex, isRow, onSelectionChange, isDropDownSelection, selectionMode, selectedKeys, previousEndIndex);
            }
            // scroll 'extensionElement' into view if focus is not shifted
            if (!isRow || !isToggleSelectionMode) {
                const extensionElement = selectionExtensionElementRef.current != null
                    ? selectionExtensionElementRef.current
                    : activeCellElement;
                let extensionCell;
                if (isRow) {
                    const rowKey = keyExtractor(extensionElement, `[${LIST_ROW_SELECTOR}='${listId}']`);
                    if (rowKey != null) {
                        extensionCell = {
                            type: activeCell.type,
                            rowKey: rowKey
                        };
                    }
                }
                else {
                    extensionCell = logicalCellExtractor(extensionElement, listId);
                }
                if (extensionCell != null) {
                    updateDeferredScrollDetail({
                        cell: extensionCell,
                        locationX: 'inView',
                        locationY: 'inView'
                    });
                }
            }
        }
    }
}

/**
 * Hook that manages keyboard interactions on the Preact TableList
 */
function useKeyboardHandling({ rootRef, viewportInfo, listId, listData, isTabbableMode, selectionExtensionElementRef, activeCell, onActiveCellChanged, isDropDownSelection, selectionMode, selectedKeys, onListSelectionChange, onItemAction, updateDeferredScrollDetail, reorderKeyboardProps, isReorderEnabled }) {
    // state for the key that a space or enter keydown occurrs on for matching keyup handling
    const triggerDownKeyRef = useRef();
    // Keydown handler for the outer List. This handles all internal
    // keyboard navigation for the List.
    const _keyDownHandler = useCallback((event) => {
        const rootElement = rootRef.current;
        const scrollerElement = viewportInfo.viewportConfig.scrollerRef.current;
        if (rootElement != null && scrollerElement != null) {
            const keyHandlers = {
                ' ': function () {
                    if (activeCell != null) {
                        triggerDownKeyRef.current = activeCell.rowKey;
                        if (!isTabbableMode) {
                            // prevent default action on space to avoid scrolling during a selection
                            event.preventDefault();
                        }
                        if (onItemAction != null &&
                            getRowActionGestureContext(listId, listData, event.target, LIST_ROW_SELECTOR, getActionContext, activeCell) != null) {
                            event.stopPropagation();
                        }
                    }
                },
                Enter: function () {
                    if (activeCell != null) {
                        triggerDownKeyRef.current = activeCell.rowKey;
                        if (activeCell.type === 'data') {
                            if (isDropDownSelection && !isTabbableMode) {
                                event.preventDefault();
                            }
                            if (onItemAction != null &&
                                getRowActionGestureContext(listId, listData, event.target, LIST_ROW_SELECTOR, getActionContext, activeCell) != null) {
                                event.stopPropagation();
                            }
                        }
                    }
                },
                ArrowUp: function () {
                    if (isReorderEnabled && !isTabbableMode && isReorderGesture(event)) {
                        if (activeCell?.type === 'data') {
                            reorderKeyboardProps.onKeyDown?.(event);
                            updateDeferredScrollDetail({
                                cell: activeCell,
                                locationX: 'inView',
                                locationY: 'inView'
                            });
                        }
                        return;
                    }
                    if (activeCell != null &&
                        (!isTabbableMode || isAutoFocusCell(rootElement, listId, activeCell))) {
                        let isExtendSelectionGesture = false;
                        if (!isTabbableMode &&
                            activeCell.type === 'data' &&
                            event.shiftKey &&
                            onListSelectionChange != null &&
                            (selectionMode === 'multiple' || selectionMode === 'multipleToggle')) {
                            handleExtendSelectionGesture(rootElement, listId, rootElement, true, activeCell, selectionExtensionElementRef, onListSelectionChange, isDropDownSelection, updateDeferredScrollDetail, selectionMode, selectedKeys);
                            isExtendSelectionGesture = true;
                        }
                        if (!isExtendSelectionGesture ||
                            isDropDownSelection ||
                            selectionMode === 'multipleToggle') {
                            handleAdjacentRowGesture(rootElement, listId, activeCell, onActiveCellChanged, true, isExtendSelectionGesture, updateDeferredScrollDetail);
                        }
                        event.preventDefault();
                    }
                },
                ArrowDown: function () {
                    if (isReorderEnabled && !isTabbableMode && isReorderGesture(event)) {
                        if (activeCell?.type === 'data') {
                            reorderKeyboardProps.onKeyDown?.(event);
                            updateDeferredScrollDetail({
                                cell: activeCell,
                                locationX: 'inView',
                                locationY: 'inView'
                            });
                        }
                        return;
                    }
                    if (activeCell != null &&
                        (!isTabbableMode || isAutoFocusCell(rootElement, listId, activeCell))) {
                        let isExtendSelectionGesture = false;
                        if (!isTabbableMode &&
                            activeCell.type === 'data' &&
                            event.shiftKey &&
                            onListSelectionChange != null &&
                            (selectionMode === 'multiple' || selectionMode === 'multipleToggle')) {
                            handleExtendSelectionGesture(rootElement, listId, rootElement, false, activeCell, selectionExtensionElementRef, onListSelectionChange, isDropDownSelection, updateDeferredScrollDetail, selectionMode, selectedKeys);
                            isExtendSelectionGesture = true;
                        }
                        if (!isExtendSelectionGesture ||
                            isDropDownSelection ||
                            selectionMode === 'multipleToggle') {
                            handleAdjacentRowGesture(rootElement, listId, activeCell, onActiveCellChanged, false, isExtendSelectionGesture, updateDeferredScrollDetail);
                        }
                        event.preventDefault();
                    }
                },
                PageUp: function () {
                    if (activeCell != null &&
                        (!isTabbableMode || isAutoFocusCell(rootElement, listId, activeCell))) {
                        handlePageRowGesture(rootElement, scrollerElement, viewportInfo, listId, activeCell, onActiveCellChanged, true, updateDeferredScrollDetail);
                        event.preventDefault();
                    }
                },
                PageDown: function () {
                    if (activeCell != null &&
                        (!isTabbableMode || isAutoFocusCell(rootElement, listId, activeCell))) {
                        handlePageRowGesture(rootElement, scrollerElement, viewportInfo, listId, activeCell, onActiveCellChanged, false, updateDeferredScrollDetail);
                        event.preventDefault();
                    }
                },
                Home: function () {
                    if (activeCell != null &&
                        (!isTabbableMode || isAutoFocusCell(rootElement, listId, activeCell))) {
                        event.preventDefault();
                    }
                },
                End: function () {
                    if (activeCell != null &&
                        (!isTabbableMode || isAutoFocusCell(rootElement, listId, activeCell))) {
                        event.preventDefault();
                    }
                }
            };
            if (Object.keys(keyHandlers).includes(event.key)) {
                keyHandlers[event.key]();
            }
        }
    }, [
        rootRef,
        viewportInfo,
        listId,
        listData,
        isTabbableMode,
        selectionExtensionElementRef,
        activeCell,
        onActiveCellChanged,
        isDropDownSelection,
        selectionMode,
        selectedKeys,
        onListSelectionChange,
        onItemAction,
        updateDeferredScrollDetail,
        reorderKeyboardProps,
        isReorderEnabled
    ]);
    // Keyup handler for the outer List.
    const _keyUpHandler = useCallback((event) => {
        const rootElement = rootRef.current;
        const scrollerElement = viewportInfo.viewportConfig.scrollerRef.current;
        if (rootElement != null && scrollerElement != null) {
            const keyHandlers = {
                ' ': function () {
                    if (activeCell != null) {
                        if (triggerDownKeyRef.current === activeCell.rowKey) {
                            if (!isTabbableMode) {
                                if (onListSelectionChange != null) {
                                    handleSelectionGesture(listId, getElementFromCell(rootElement, listId, activeCell), true, onListSelectionChange, isDropDownSelection, selectionMode, selectedKeys);
                                }
                            }
                            if (onItemAction != null &&
                                handleRowActionGesture(listId, listData, onItemAction, event.target, LIST_ROW_SELECTOR, getActionContext, activeCell)) {
                                event.stopPropagation();
                            }
                        }
                        triggerDownKeyRef.current = undefined;
                    }
                },
                Enter: function () {
                    if (activeCell != null && triggerDownKeyRef.current === activeCell.rowKey) {
                        if (activeCell.type === 'data') {
                            if (isDropDownSelection && !isTabbableMode) {
                                if (onListSelectionChange != null) {
                                    handleSelectionGesture(listId, getElementFromCell(rootElement, listId, activeCell), true, onListSelectionChange, isDropDownSelection, selectionMode, selectedKeys);
                                }
                            }
                            if (onItemAction != null &&
                                handleRowActionGesture(listId, listData, onItemAction, event.target, LIST_ROW_SELECTOR, getActionContext, activeCell)) {
                                event.stopPropagation();
                            }
                        }
                    }
                    triggerDownKeyRef.current = undefined;
                }
            };
            if (Object.keys(keyHandlers).includes(event.key)) {
                keyHandlers[event.key]();
            }
        }
    }, [
        rootRef,
        viewportInfo,
        listId,
        listData,
        isTabbableMode,
        activeCell,
        isDropDownSelection,
        selectionMode,
        selectedKeys,
        onListSelectionChange,
        onItemAction
    ]);
    return { onKeyDown: _keyDownHandler, onKeyUp: _keyUpHandler };
}

/**
 * Hook that manages pointer interactions on the Preact TableList
 */
function usePointerHandling({ rootRef, listId, listData, selectionExtensionElementRef, activeCell, onActiveCellChanged, isDropDownSelection, selectionMode, selectedKeys, onListSelectionChange, onItemAction }) {
    // tracks whether the most recent 'pointerdown' event was due to a touch gesture
    const selectionTouchRef = useRef(false);
    // tracks the current 'pointer over' element for comparison during 'click' event handling to
    // prevent screen readers from triggering undesired selection gestures from simulated events
    const pointerInfoRef = useRef();
    // Pointer move handler for the outer List.
    const _pointerMoveHandler = useCallback((event) => {
        const rootElement = rootRef.current;
        if (rootElement != null) {
            pointerInfoRef.current = { target: event.target, x: event.clientX, y: event.clientY };
        }
    }, [rootRef]);
    // Pointer over handler for the outer List.
    const _pointerOverHandler = useCallback((event) => {
        const rootElement = rootRef.current;
        if (rootElement != null) {
            pointerInfoRef.current = { target: event.target, x: event.clientX, y: event.clientY };
        }
    }, [rootRef]);
    // Pointer down handler for the outer List. It ensures the area targeted is 'current'.
    // NOTE - This event handling is triggered prior to 'onFocus' handling and ensures the
    // 'pending current' area is set when an initial pointer interaction focuses the List.
    const _pointerDownHandler = useCallback((event) => {
        const rootElement = rootRef.current;
        if (rootElement != null) {
            selectionTouchRef.current = event.pointerType === 'touch';
            const targetElement = event.target;
            if (onListSelectionChange != null &&
                activeCell != null &&
                event.shiftKey &&
                isRangeSelectionGesture(rootElement, listId, targetElement, activeCell, selectionMode)) {
                // prevent focus change on selection extension gesture
                if (!getIsSelectionPending()) {
                    event.preventDefault();
                }
            }
            else {
                const cell = logicalCellExtractor(targetElement, listId);
                if (cell == null ||
                    (cell.type === 'data' &&
                        keyExtractor(targetElement, `[${LIST_ROW_SELECTOR}='${listId}']`) == null)) {
                    // prevent focus shift when interacting with 'non-focusable' regions (ie - scrollbar)
                    event.preventDefault();
                }
            }
        }
    }, [rootRef, listId, activeCell, selectionMode, onListSelectionChange]);
    // Pointer leave handler for the outer List.
    const _pointerLeaveHandler = useCallback(() => {
        pointerInfoRef.current = undefined;
    }, []);
    // Click handler for the outer List. It ensures the List's selection state is
    // updated to reflect the area targeted, and any action events are fired correctly.
    const _clickHandler = useCallback((event) => {
        const rootElement = rootRef.current;
        if (rootElement != null) {
            // only handle click events that match the latest mouse over/move info to prevent screen readers
            // from triggering undesired selections due to simulated click events when transferring focus
            if (selectionTouchRef.current ||
                (pointerInfoRef.current != null &&
                    pointerInfoRef.current.target === event.target &&
                    Math.abs(pointerInfoRef.current.x - event.clientX) <= 1 &&
                    Math.abs(pointerInfoRef.current.y - event.clientY) <= 1)) {
                let isSelectionExtension = false;
                const target = event.target;
                if (onListSelectionChange != null && !getIsSelectionPending()) {
                    if (activeCell != null && event.shiftKey) {
                        isSelectionExtension = handleRangeSelectionGesture(rootElement, listId, target, activeCell, onActiveCellChanged, selectionExtensionElementRef, onListSelectionChange, isDropDownSelection, selectionMode, selectedKeys);
                    }
                    if (!isSelectionExtension) {
                        handleSelectionGesture(listId, target, isLogicalCtrl(event) || selectionTouchRef.current, onListSelectionChange, isDropDownSelection, selectionMode, selectedKeys);
                    }
                }
                if (onItemAction != null &&
                    !getIsSelectionPending() &&
                    handleRowActionGesture(listId, listData, onItemAction, target, LIST_ROW_SELECTOR, getActionContext)) {
                    event.stopPropagation();
                }
                // if click event occurs on the row element itself, focus the first child grid cell
                const isToggleSelectionMode = isDropDownSelection || selectionMode === 'multipleToggle';
                if ((!isSelectionExtension || isToggleSelectionMode) && getRowKey(target) != null) {
                    const logicalCell = getFirstFocusableCell(target, listId);
                    if (logicalCell != null) {
                        onActiveCellChanged({ value: logicalCell }, false, true);
                        getElementFromCell(rootElement, listId, logicalCell)?.focus({ preventScroll: true });
                    }
                }
            }
        }
    }, [
        rootRef,
        listId,
        listData,
        selectionExtensionElementRef,
        activeCell,
        onActiveCellChanged,
        isDropDownSelection,
        selectionMode,
        selectedKeys,
        onListSelectionChange,
        onItemAction
    ]);
    return {
        pointerHandlingProps: {
            onPointerOver: _pointerOverHandler,
            onPointerMove: _pointerMoveHandler,
            onPointerDown: _pointerDownHandler,
            onPointerLeave: _pointerLeaveHandler,
            onClick: _clickHandler
        }
    };
}

/**
 * Hook that manages handling scroll events on the Preact TableList
 */
function useScrollHandling({ rootRef, viewportInfo, listId, isLayoutRequired, scrollPositionOverride }) {
    const scrollPositionOverrideRef = useRef();
    // update horizontal and/or vertical scroll position if a new value is provided
    // NO DEPENDENCY ARRAY AS THIS SHOULD RUN ON EVERY RENDER
    useLayoutEffect(() => {
        const rootElement = rootRef.current;
        const scrollerElement = viewportInfo.viewportConfig.scrollerRef.current;
        if (scrollPositionOverrideRef.current !== scrollPositionOverride &&
            rootElement != null &&
            scrollerElement != null &&
            !isLayoutRequired) {
            setScrollPosition(rootElement, scrollerElement, viewportInfo, listId, scrollPositionOverride);
            scrollPositionOverrideRef.current = scrollPositionOverride;
        }
    });
    return viewportInfo.isExternalScroller ? {} : {};
}

const _defaultDisabledRowInteractionKeys = ['enter'];
/**
 * Hook that manages interactions and events on the Preact TableList
 */
function useInteractionManager({ rootRef, viewportInfo, ref, listId, listData, hasMore, isEmbedded, isLayoutRequired, scrollPositionOverride, currentItemOverride, onPersistCurrentItem, selectionMode, selectedKeys, onSelectionChange, onItemAction, itemEnterKeyFocusBehavior, contextMenuConfig, onReorder }) {
    // whether row selection is triggered on 'enter' keypresses in addition to 'spacebar'
    const disabledRowInteractionKeys = isEmbedded ? undefined : _defaultDisabledRowInteractionKeys;
    // tracks the current selection extension element (if any)
    const selectionExtensionElementRef = useRef();
    // intercept onSelectionChange calls to ensure local selectionExtensionElementRef is cleared
    const onListSelectionChange = useMemo(() => {
        return onSelectionChange != null
            ? (detail, isExtendableSelection) => {
                if (!isExtendableSelection) {
                    selectionExtensionElementRef.current = undefined;
                }
                onSelectionChange(detail);
            }
            : undefined;
    }, [onSelectionChange]);
    // track pending scrolls to minimize 'flash' of focus rings when scrolling
    const [deferredScrollDetail, setDeferredScrollDetail] = useState(undefined);
    const deferredScrollDetailRef = useRef();
    // keep state and ref in-sync to avoid edge cases where a gesture updating the scroll detail is handled
    // inbetween a render cycle that changes the scroll detail and the useEffect that applies the detail
    const updateDeferredScrollDetail = useCallback((scrollDetail) => {
        setDeferredScrollDetail(scrollDetail);
        deferredScrollDetailRef.current = scrollDetail;
    }, []);
    const { reorderProps, reorderKeyboardProps, reorderItemProps, reorderContext, reorderInstructions, reorderInstructionsId } = useItemReorder({
        rootRef,
        scrollerRef: viewportInfo.viewportConfig.scrollerRef,
        itemSelector: `[${LIST_ROW_SELECTOR}='${listId}']`,
        onReorder
    });
    // setup focus tracking using the useFocusHandling hook
    const { focusHandlingProps, activeCell, onActiveCellChanged, currentActiveCell, currentCell, isShowFocusRing, isTabbableMode, onStartFocusTracking, onFocusStartEdge, onFocusEndEdge } = useFocusHandling({
        rootRef,
        viewportInfo,
        listId,
        listData,
        hasMore,
        selectionExtensionElementRef,
        currentItemOverride,
        onPersistCurrentItem,
        isEmbedded,
        itemEnterKeyFocusBehavior,
        isDataLoaded: !hasMore || listData.length > 0,
        reorderContext,
        deferredScrollDetail,
        deferredScrollDetailRef,
        updateDeferredScrollDetail
    });
    const keyboardHandlingProps = useKeyboardHandling({
        rootRef,
        viewportInfo,
        listId,
        listData,
        isTabbableMode,
        selectionExtensionElementRef,
        activeCell,
        onActiveCellChanged,
        isDropDownSelection: isEmbedded,
        selectionMode,
        selectedKeys,
        onListSelectionChange,
        onItemAction,
        updateDeferredScrollDetail,
        reorderKeyboardProps,
        isReorderEnabled: onReorder != null
    });
    const { pointerHandlingProps } = usePointerHandling({
        rootRef,
        listId,
        listData,
        selectionExtensionElementRef,
        activeCell,
        onActiveCellChanged,
        isDropDownSelection: isEmbedded,
        selectionMode,
        selectedKeys,
        onListSelectionChange,
        onItemAction
    });
    const scrollHandlingProps = useScrollHandling({
        rootRef,
        viewportInfo,
        listId,
        isLayoutRequired,
        scrollPositionOverride
    });
    const { contextMenuTriggerProps, menuProps, contextMenuContext } = useContextMenu(rootRef, listId, listData, contextMenuConfig);
    const interactionManagerProps = mergeProps(focusHandlingProps, keyboardHandlingProps, pointerHandlingProps, reorderProps, contextMenuTriggerProps);
    // setup imperative handle implementation
    useImperativeHandle(ref, () => {
        const rootElement = rootRef.current;
        const scrollerElement = viewportInfo.viewportConfig.scrollerRef.current;
        return {
            focus: () => {
                if (rootElement != null) {
                    if (currentActiveCell != null) {
                        const cellElement = getElementFromCell(rootElement, listId, currentActiveCell);
                        if (cellElement != null) {
                            cellElement.focus();
                            return;
                        }
                    }
                    // if no current cell element was found, just focus the root element
                    rootElement.focus();
                }
            },
            getScrollPosition: () => {
                if (rootElement == null || scrollerElement == null || isLayoutRequired) {
                    return {};
                }
                return getScrollPosition(rootElement, scrollerElement, viewportInfo, listId);
            }
        };
    }, [rootRef, viewportInfo, listId, currentActiveCell, isLayoutRequired]);
    return {
        interactionManagerProps,
        scrollHandlingProps,
        onListSelectionChange,
        isShowFocusRing,
        isTabbableMode,
        currentCell,
        currentActiveCell,
        contextMenuContext,
        menuProps,
        onStartFocusTracking,
        onFocusStartEdge,
        onFocusEndEdge,
        disabledRowInteractionKeys,
        reorderItemProps,
        reorderInstructions,
        reorderInstructionsId
    };
}

/**
 * Hook that manages sizing on the Preact TableList
 */
function useSizingManager({ viewportConfig, externalScrollerAccessRef }) {
    // scrollerRef is callback-based to meet useResizeObserver hook's state management requirements
    const [scrollerRef, setScrollerRef] = useState(createRef());
    const scrollerRefCallback = useCallback((node) => {
        setScrollerRef({ current: node });
    }, []);
    useImperativeHandle(externalScrollerAccessRef, () => {
        if (viewportConfig != null) {
            return { [UNSAFE_DOM_ACCESS]: viewportConfig.scrollerRef.current };
        }
        return { [UNSAFE_DOM_ACCESS]: scrollerRef.current };
    }, [scrollerRef, viewportConfig]);
    // current viewport information
    const viewportInfo = useMemo(() => {
        const listViewportConfig = getViewportConfig(scrollerRef, viewportConfig);
        return {
            isExternalScroller: viewportConfig != null,
            viewportConfig: listViewportConfig
        };
    }, [scrollerRef, viewportConfig]);
    return { viewportInfo, scrollerRefCallback };
}

/**
 * A sparkle component for smart suggestion indicator in TableList
 * @param sparkleHeight the height of sparkle
 */
function Sparkle() {
    const sparkleClasses = classNames([sparkleStyles.base]);
    return jsx("div", { class: sparkleClasses });
}

/**
 * The internal component used to render an empty list cell in a ListBody.
 */
function ListNoData({ listId, isShowFocusRing, hasTabIndex, isTabbableMode, ariaRowIndex, renderer }) {
    const translations = useTranslationBundle('@oracle/oraclejet-preact');
    const rowClasses = classNames([listRowStyles.base]);
    const { classes: themeClasses } = useComponentTheme(ListNoDataRedwoodTheme, {
        defaultRenderer: renderer == null ? 'isDefaultRenderer' : 'notDefaultRenderer',
        showFocusRing: isShowFocusRing ? 'isShowFocusRing' : 'notShowFocusRing'
    });
    const cellClasses = classNames([themeClasses]);
    function defaultContentRenderer() {
        return translations.collection_noData();
    }
    const contentRenderer = renderer != null ? renderer : defaultContentRenderer;
    return (jsx("div", { "aria-rowindex": ariaRowIndex, class: rowClasses, style: 'height:100%;width:100%;', role: 'row', children: jsx("div", { "aria-colindex": 1, class: cellClasses, tabIndex: hasTabIndex ? 0 : -1, role: 'gridcell', "data-oj-cell-type": 'noData', "data-oj-list-focusable": listId, "data-oj-list-nodata": listId, children: contentRenderer({ isTabbable: isTabbableMode }) }) }));
}

/**
 * Helper function to generate the class array for a list row.
 */
const getListRowClassArray = ({ isFinalRowIndex, hasTopGridline, hasBottomGridline, isSelected, isPreviousSelected, isShowFocusRing, isSelectionEnabled, isSingleSelection, isActive, isHover, isPseudoHover, isHighContrast, variant, padding }) => {
    const rowStyles = isHighContrast ? listRowStylesHC : listRowStyles;
    const classArray = [rowStyles.base];
    // Padding styling logic
    const { isTopPadding, isBottomPadding, isStartPadding, isEndPadding } = _getPaddingInfo(padding);
    classArray.push(isStartPadding ? rowStyles.startPadding : rowStyles.startNoPadding);
    classArray.push(isEndPadding ? rowStyles.endPadding : rowStyles.endNoPadding);
    _applyFocusStyling(classArray, rowStyles, isShowFocusRing);
    // Horizontal and vertical alignment styling logic
    classArray.push(textAlignStyles['start']);
    classArray.push(verticalAlignStyles['center']);
    // Interaction styling logic
    if (isSelectionEnabled) {
        if (isActive) {
            classArray.push(rowStyles.active);
        }
        else if (isHover) {
            classArray.push(rowStyles.hover);
        }
        else if (isPseudoHover) {
            classArray.push(rowStyles.pseudoHover);
        }
    }
    if (variant === 'highlight') {
        classArray.push(rowStyles.rowHighlight);
    }
    // Selected Rows and Horizontal Gridline styling logic
    let hasBackground = false;
    if (isSelected) {
        hasBackground = true;
        classArray.push(rowStyles.selectedBackground);
    }
    if (isSingleSelection && (isSelected || isPreviousSelected)) {
        classArray.push(rowStyles.horizontalSelectedGridTop);
        classArray.push(isTopPadding ? rowStyles.selectedBorderTopPadding : rowStyles.selectedBorderTopNoPadding);
    }
    else if (hasTopGridline) {
        classArray.push(rowStyles.horizontalGridTop);
        classArray.push(isTopPadding ? rowStyles.borderTopPadding : rowStyles.borderTopNoPadding);
    }
    else {
        classArray.push(isTopPadding ? rowStyles.topSpacerPadding : rowStyles.topSpacerNoPadding);
    }
    if (isFinalRowIndex) {
        classArray.push(rowStyles.lastRow);
        if (isSingleSelection && isSelected) {
            classArray.push(rowStyles.horizontalSelectedGridBottom);
            classArray.push(isBottomPadding
                ? rowStyles.selectedBorderBottomPadding
                : rowStyles.selectedBorderBottomNoPadding);
        }
        else if (hasBottomGridline) {
            classArray.push(rowStyles.horizontalGridBottom);
            classArray.push(isBottomPadding ? rowStyles.borderBottomPadding : rowStyles.borderBottomNoPadding);
        }
        else {
            classArray.push(isBottomPadding ? rowStyles.bottomSpacerPadding : rowStyles.bottomSpacerNoPadding);
        }
    }
    else {
        classArray.push(rowStyles.notLastRow);
        classArray.push(isBottomPadding ? rowStyles.bottomPadding : rowStyles.bottomNoPadding);
    }
    if (!hasBackground) {
        classArray.push(rowStyles.background);
    }
    return classArray;
};
/**
 * Helper function to generate the class array for a list cell.
 */
const getListCellClassArray = ({ isHighContrast }) => {
    const cellStyles = isHighContrast ? listCellStylesHC : listCellStyles;
    const classArray = [cellStyles.base];
    return classArray;
};
const _getPaddingInfo = (padding) => {
    const isPaddingEnabled = padding === 'enabled';
    const isComplexPadding = padding instanceof Object;
    const isBottomPadding = isPaddingEnabled || (isComplexPadding && padding?.bottom === 'enabled');
    const isEndPadding = isPaddingEnabled || (isComplexPadding && padding?.end === 'enabled');
    const isStartPadding = isPaddingEnabled || (isComplexPadding && padding?.start === 'enabled');
    const isTopPadding = isPaddingEnabled || (isComplexPadding && padding?.top === 'enabled');
    return { isTopPadding, isBottomPadding, isStartPadding, isEndPadding };
};
const _applyFocusStyling = (classArray, elementStyles, isShowFocusRing) => {
    if (isShowFocusRing) {
        classArray.push(elementStyles.focused);
    }
};

function _defaultCellRenderer() {
    return undefined;
}
/**
 * The internal component used to render a single cell in a ListRow.
 */
function ListCell({ listId, rowData, rowKey, rowIndex, hasTabIndex, isSelectionEnabled, isSelected, isTabbableMode, renderer = _defaultCellRenderer, selectorRenderer, isHighContrast }) {
    const cellClasses = classNames(getListCellClassArray({ isHighContrast }));
    const cellRendererProps = {
        data: rowData,
        metadata: { key: rowKey },
        selector: selectorRenderer,
        isSelected,
        index: rowIndex,
        isTabbable: isTabbableMode
    };
    return (jsx("div", { "aria-colindex": 1, class: cellClasses, tabIndex: hasTabIndex ? 0 : -1, role: 'gridcell', "data-oj-cell-type": 'data', "data-oj-list-focusable": listId, "data-oj-list-data": listId, ...(isSelectionEnabled
            ? isSelected
                ? { 'aria-selected': true }
                : { 'aria-selected': false }
            : {}), children: renderer(cellRendererProps) }));
}

/**
 * The internal component used to render a single row in List.
 */
function ListRow({ listId, rowKey, rowData, rowIndex, isFinalRow, children, padding, hasTopGridline, hasBottomGridline, isSelected, isPreviousSelected, isSelectionEnabled, isSingleSelectionEnabled, selectedRowKeys, isShowFocusRing, isActive, isTabbableMode, isSuggestion, variant, disabledInteractionKeys, onSelectionChange, isRtl, isHighContrast }) {
    const translations = useTranslationBundle('@oracle/oraclejet-preact');
    const { interactionProps, applyActiveStyle, applyHoverStyle, applyPseudoHoverStyle } = useInteractionStyle({ disabledKeys: disabledInteractionKeys });
    const rowClasses = classNames(getListRowClassArray({
        isFinalRowIndex: isFinalRow,
        hasTopGridline,
        hasBottomGridline,
        isSelected,
        isPreviousSelected,
        isShowFocusRing,
        isSelectionEnabled,
        isSingleSelection: isSingleSelectionEnabled,
        isActive: applyActiveStyle,
        isHover: applyHoverStyle,
        isPseudoHover: applyPseudoHoverStyle,
        isHighContrast,
        variant,
        padding
    }));
    const accessibleRowName = (rowIndex + 2).toString();
    const handleRowSelectorChange = (detail) => {
        if (onSelectionChange != null) {
            onSelectionChange({
                value: detail.value,
                target: detail.target
            }, false);
        }
    };
    const selectorRenderer = isSelectionEnabled && !isSingleSelectionEnabled
        ? () => {
            // conversion from 0-based index and adding the header row as it is always rendered
            return (jsx(Selector, { "aria-label": translations.collection_selectRow({ ROW_NAME: accessibleRowName }), onChange: handleRowSelectorChange, rowKey: rowKey, selectedKeys: selectedRowKeys }));
        }
        : undefined;
    const hasTabIndex = isActive && !isTabbableMode;
    const cellProps = {
        listId,
        rowData,
        rowKey,
        rowIndex,
        isSelectionEnabled,
        isSelected,
        hasTabIndex,
        isTabbableMode,
        renderer: children,
        selectorRenderer,
        isRtl,
        isHighContrast
    };
    return (jsx("div", { "aria-rowindex": rowIndex + 1, class: rowClasses, role: 'row', "data-oj-key": rowKey, "data-oj-list-row": listId, ...(typeof rowKey === 'number' && { 'data-oj-key-type': 'number' }), ...(isSuggestion && { 'data-oj-suggestion': true }), ...(isSelectionEnabled && interactionProps), children: jsx(ListCell, { ...cellProps }) }, rowKey));
}
const MemoizeListRow = memo(ListRow, (prev, next) => {
    if (prev && next) {
        return (prev.listId === next.listId &&
            prev.rowKey === next.rowKey &&
            prev.rowData === next.rowData &&
            prev.rowIndex === next.rowIndex &&
            prev.isFinalRow === next.isFinalRow &&
            prev.children === next.children &&
            prev.padding === next.padding &&
            prev.hasTopGridline === next.hasTopGridline &&
            prev.hasBottomGridline === next.hasBottomGridline &&
            prev.isSelected === next.isSelected &&
            prev.isPreviousSelected === next.isPreviousSelected &&
            prev.isSelectionEnabled === next.isSelectionEnabled &&
            prev.isSingleSelectionEnabled === next.isSingleSelectionEnabled &&
            (prev.isSelectionEnabled && !prev.isSingleSelectionEnabled
                ? isSameKey(prev.selectedRowKeys, next.selectedRowKeys)
                : true) &&
            prev.isShowFocusRing === next.isShowFocusRing &&
            prev.isActive === next.isActive &&
            prev.isTabbableMode === next.isTabbableMode &&
            prev.isSuggestion === next.isSuggestion &&
            prev.variant === next.variant &&
            prev.onSelectionChange === next.onSelectionChange &&
            prev.isRtl === next.isRtl &&
            prev.isHighContrast === next.isHighContrast &&
            prev.disabledInteractionKeys === next.disabledInteractionKeys);
    }
    return false;
});

/**
 * The internal component used to render a single reorderable row in List.
 */
function ListReorderableRow({ children, dragKey, setDragKey, rootRef, onReorder, ...rowProps }) {
    const itemRef = useRef(null);
    const itemStyles = {
        itemOverlayFront: reorderableStyles.reorderableItemOverlayFront,
        itemOverlayRear: reorderableStyles.reorderableItemOverlayRear,
        itemDragger: reorderableStyles.reorderableItemDragger
    };
    const itemSelector = `[${LIST_ROW_SELECTOR}='${rowProps.listId}']`;
    const getItem = () => itemRef?.current?.closest(itemSelector);
    const setTransferData = useCallback((event) => {
        event.dataTransfer?.setData('text/listview-items-key', JSON.stringify(rowProps.rowKey));
    }, [rowProps.rowKey]);
    const { rootProps, isDragged, overlayZoneBeforeProps, overlayZoneAfterProps } = useReorderableItem({
        isDisabled: onReorder ? false : true,
        dragKey,
        setDragKey,
        onReorder,
        rootRef,
        ref: itemRef,
        getItem,
        setTransferData,
        itemStyles,
        itemSelector,
        isVertical: true
    });
    const reorderClassArray = [];
    if (rootProps?.class != null) {
        reorderClassArray.push(rootProps.class);
    }
    reorderClassArray.push(reorderableStyles.reorderableItemDnd);
    const reorderClasses = classNames(reorderClassArray);
    const reorderableRenderer = (context) => {
        return (jsxs("div", { ...rootProps, class: reorderClasses, children: [children(context), !isDragged && isKeyDefined(dragKey) && (jsxs(Fragment, { children: [jsx("div", { ...overlayZoneBeforeProps }), jsx("div", { ...overlayZoneAfterProps })] }))] }));
    };
    return jsx(ListRow, { ...rowProps, children: reorderableRenderer });
}
const MemoizeListReorderableRow = memo(ListReorderableRow, (prev, next) => {
    if (prev && next) {
        return (prev.listId === next.listId &&
            prev.rowKey === next.rowKey &&
            prev.rowData === next.rowData &&
            prev.rowIndex === next.rowIndex &&
            prev.isFinalRow === next.isFinalRow &&
            prev.children === next.children &&
            prev.padding === next.padding &&
            prev.hasTopGridline === next.hasTopGridline &&
            prev.hasBottomGridline === next.hasBottomGridline &&
            prev.isSelected === next.isSelected &&
            prev.isPreviousSelected === next.isPreviousSelected &&
            prev.isSelectionEnabled === next.isSelectionEnabled &&
            prev.isSingleSelectionEnabled === next.isSingleSelectionEnabled &&
            (prev.isSelectionEnabled && !prev.isSingleSelectionEnabled
                ? isSameKey(prev.selectedRowKeys, next.selectedRowKeys)
                : true) &&
            prev.isShowFocusRing === next.isShowFocusRing &&
            prev.isActive === next.isActive &&
            prev.isTabbableMode === next.isTabbableMode &&
            prev.isSuggestion === next.isSuggestion &&
            prev.variant === next.variant &&
            prev.onSelectionChange === next.onSelectionChange &&
            prev.isRtl === next.isRtl &&
            prev.isHighContrast === next.isHighContrast &&
            prev.disabledInteractionKeys === next.disabledInteractionKeys &&
            prev.dragKey === next.dragKey &&
            prev.setDragKey === next.setDragKey &&
            prev.onReorder === next.onReorder &&
            prev.rootRef === next.rootRef);
    }
    return false;
});

const dimensions = ['height', 'width'];
// Create an array [dimensionInterpolations['height'], dimensionInterpolations['width']]
const skeletonDimensionInterpolation = Array.from(dimensions, (x) => dimensionInterpolations[x]);
const interpolations = [...Object.values(skeletonDimensionInterpolation)];
const SkeletonInterpolations = mergeInterpolations(interpolations);
/**
 * ListSkeleton component allows the appropriate skeleton to be rendered based on the
 * property values
 **/
function ListSkeleton({ ...props }) {
    const skeletonStyles = props.isHighContrast ? listSkeletonStylesHC : listSkeletonStyles;
    const classes = classNames([skeletonStyles.bar]);
    const skeletonDimensions = SkeletonInterpolations({ width: '100%', ...props });
    return jsx("div", { style: skeletonDimensions, class: classes });
}

/**
 * ListSkeletonRow renders a set of 'loading' skeletons.
 **/
function ListSkeletonRow({ listId, hasTabIndex, isShowFocusRing, isLoadMore = false, isHighContrast, skeletonRenderer }) {
    const translations = useTranslationBundle('@oracle/oraclejet-preact');
    const skeletonStyles = isHighContrast ? listSkeletonStylesHC : listSkeletonStyles;
    const cellClassArray = [skeletonStyles.cell];
    if (isShowFocusRing) {
        cellClassArray.push(skeletonStyles.focus);
    }
    const cellClasses = classNames(cellClassArray);
    const innerCellClasses = classNames([skeletonStyles.innerCell]);
    const rowClasses = classNames([skeletonStyles.row, skeletonStyles.rowHeight]);
    const firstRowClasses = rowClasses;
    const accCellContent = (jsx("div", { style: 'width:1px;height:0px;overflow:hidden;', children: translations.collection_loading() }));
    return (jsx("div", { class: LOADMORE_STYLE_CLASS, role: 'row', children: jsx("div", { class: cellClasses, role: 'gridcell', tabIndex: hasTabIndex ? 0 : -1, "data-oj-cell-type": isLoadMore ? 'loadMore' : 'loading', ...(isLoadMore ? { 'data-oj-list-loadmore': listId } : { 'data-oj-list-loading': listId }), "data-oj-list-focusable": listId, children: jsxs("div", { class: skeletonRenderer != null ? undefined : innerCellClasses, children: [accCellContent, [...Array(isLoadMore ? 3 : 25)].map((_element, index) => skeletonRenderer != null ? (skeletonRenderer({ loadingStatus: isLoadMore ? 'loadMore' : 'initial', index })) : (jsx("div", { class: index === 0 ? firstRowClasses : rowClasses, children: jsx(ListSkeleton, { height: "4x", isHighContrast: isHighContrast }) })))] }) }) }));
}

/**
 * The internal component used to render the body of a List.
 */
function ListBody({ viewportInfo, listId, listData, children, itemPadding, hasItemGridlines, hasTopGridline, hasBottomGridline, isShowFocusRing, isTabbableMode, hasMore, onLoadMore, activeCell, currentItemKey, currentItemVariant, selectedKeys, isSelectionEnabled, isSingleSelectionEnabled, disabledInteractionKeys, onSelectionChange, promotedSection, noDataRenderer, skeletonRenderer, isRtl, isHighContrast, reorderItemProps }) {
    // returns a ListRowContext based on ItemContext (added key)
    const getRowContext = useCallback((context) => {
        return {
            index: context.index,
            data: context.data.data,
            key: context.data.metadata.key
        };
    }, []);
    let isPreviousSelected = false;
    let isPreviousSuggestion = false;
    const collectionChildrenFunc = (context) => {
        const listRowContext = getRowContext(context);
        const isSuggestion = promotedSection ? promotedSection.count > listRowContext.index : false;
        const rowKey = listRowContext.key;
        const containsActiveCell = activeCell?.rowKey === rowKey;
        const isSelected = containsKey(selectedKeys, rowKey);
        const isFinalRow = listRowContext.index === listData.length - 1;
        const isTopGridlineVisible = listRowContext.index === 0
            ? hasTopGridline
            : hasItemGridlines || (!isSuggestion && isPreviousSuggestion);
        const rowProps = {
            listId,
            children,
            padding: itemPadding,
            isSuggestion,
            hasTopGridline: isTopGridlineVisible,
            hasBottomGridline: isFinalRow && (hasMore ? hasItemGridlines : hasBottomGridline),
            isSelected,
            isPreviousSelected,
            disabledInteractionKeys,
            onSelectionChange,
            isSelectionEnabled,
            isSingleSelectionEnabled,
            variant: currentItemKey === rowKey ? currentItemVariant : 'none',
            isShowFocusRing: isShowFocusRing && containsActiveCell,
            isActive: containsActiveCell,
            isTabbableMode,
            selectedRowKeys: selectedKeys,
            rowIndex: listRowContext.index,
            rowKey,
            rowData: listRowContext.data,
            isFinalRow,
            isRtl,
            isHighContrast
        };
        isPreviousSelected = isSelected;
        isPreviousSuggestion = isSuggestion;
        if (reorderItemProps != null) {
            return jsx(MemoizeListReorderableRow, { ...rowProps, ...reorderItemProps });
        }
        return jsx(MemoizeListRow, { ...rowProps });
    };
    const getNoDataProps = () => {
        const isActive = activeCell?.type === 'noData';
        const hasTabIndex = isActive && !isTabbableMode;
        return {
            listId: listId,
            isShowFocusRing: isShowFocusRing && isActive,
            hasTabIndex: hasTabIndex,
            isTabbableMode: isTabbableMode,
            ariaRowIndex: 1,
            renderer: noDataRenderer
        };
    };
    const isShowEmptyLoading = listData.length === 0 && hasMore;
    const isShowNoData = listData.length === 0 && !hasMore;
    const defaultLoadingIndicator = (jsx(ListSkeletonRow, { listId: listId, hasTabIndex: isShowEmptyLoading ? activeCell?.type === 'loading' : activeCell?.type === 'loadMore', isLoadMore: !isShowEmptyLoading, isShowFocusRing: isShowFocusRing &&
            (isShowEmptyLoading ? activeCell?.type === 'loading' : activeCell?.type === 'loadMore'), isHighContrast: isHighContrast, skeletonRenderer: skeletonRenderer }));
    const { name } = useTheme();
    const sparkleContainerClasses = classNames([
        sparkleStyles.container,
        // TODO: this component should create separate theme definitions to only show the image in redwood
        name === 'redwood' && globalImages
    ]);
    return isShowNoData ? (jsx(ListNoData, { ...getNoDataProps() })) : (jsx(LoadMoreCollection, { data: listData, loadMoreIndicator: defaultLoadingIndicator, suggestions: jsx(Sparkle, {}), suggestionsContainerClasses: sparkleContainerClasses, suggestionCount: promotedSection?.count, hasMore: hasMore, onLoadMore: onLoadMore, viewportConfig: viewportInfo.viewportConfig, children: collectionChildrenFunc }));
}

/**
 * The internal component used to render a custom context menu on the List
 */
function ListContextMenu({ contextMenuConfig, contextMenuContext, ...menuProps }) {
    const translations = useTranslationBundle('@oracle/oraclejet-preact');
    const accessibleLabel = contextMenuConfig.accessibleLabel != null
        ? contextMenuConfig.accessibleLabel
        : translations.accContextMenuLabel();
    return contextMenuConfig.itemsRenderer(contextMenuContext) != null ? (jsx(Menu, { "aria-label": accessibleLabel, ...menuProps, children: menuProps.isOpen && contextMenuConfig.itemsRenderer(contextMenuContext) })) : (jsx(Fragment, {}));
}

const _defaultSelectedKeys = { all: false, keys: new Set() };
/**
 * Component that renders data in a flat list.
 */
const List = forwardRef(function List({ 'aria-label': ariaLabel, 'aria-labelledby': ariaLabelledBy, 'aria-describedby': ariaDescribedBy, children = () => undefined, contextMenuConfig, viewportConfig, data, getRowKey, hasMore = false, onLoadMore = () => { }, currentItemOverride, onPersistCurrentItem, currentItemVariant, itemPadding = 'disabled', itemEnterKeyFocusBehavior = 'focusWithin', gridlines = { item: 'hidden', top: 'hidden', bottom: 'hidden' }, selectionMode, selectedKeys = _defaultSelectedKeys, onSelectionChange, onItemAction, scrollPositionOverride, promotedSection, skeletonRenderer, noDataRenderer, scrollerRef, onReorder, testId }, ref) {
    const listId = useId();
    const testIdProps = useTestId(testId);
    // rootRef is callback-based to meet useResizeObserver hook's state management requirements
    const [rootRef, setRootRef] = useState(createRef());
    const rootRefCallback = useCallback((node) => {
        setRootRef({ current: node });
    }, []);
    const { direction, forcedColors } = useUser();
    const isHighContrast = forcedColors === 'active';
    const isRtl = direction === 'rtl';
    const isEmbedded = useCollectionInteractionContext() === 'embedded';
    const itemVariant = currentItemVariant ?? isEmbedded ? 'highlight' : 'none';
    const { managedData: listData, onManagedLoadMore: onListLoadMore } = useDataManager({
        data,
        getRowKey,
        hasMore,
        onLoadMore
    });
    const { viewportInfo, scrollerRefCallback } = useSizingManager({
        viewportConfig,
        externalScrollerAccessRef: scrollerRef
    });
    const ariaRowCount = hasMore ? -1 : listData.length;
    const { interactionManagerProps, scrollHandlingProps, onListSelectionChange, isShowFocusRing, isTabbableMode, currentCell, currentActiveCell, onStartFocusTracking, onFocusStartEdge, onFocusEndEdge, contextMenuContext, menuProps, disabledRowInteractionKeys, reorderItemProps, reorderInstructions, reorderInstructionsId } = useInteractionManager({
        rootRef,
        viewportInfo,
        ref,
        listId,
        listData,
        hasMore,
        isEmbedded,
        isLayoutRequired: false,
        scrollPositionOverride,
        currentItemOverride,
        onPersistCurrentItem,
        selectionMode,
        selectedKeys,
        onSelectionChange,
        onItemAction,
        itemEnterKeyFocusBehavior,
        contextMenuConfig,
        onReorder
    });
    // initialize base List style classes
    const { classes: themeClasses } = useComponentTheme(ListRedwoodTheme, {
        showFocusRing: isShowFocusRing ? 'isShowFocusRing' : 'notShowFocusRing',
        externalScroller: viewportInfo.isExternalScroller ? 'isExternalScroller' : 'notExternalScroller'
    });
    const scrollerClassArray = [scrollerStyles.base];
    if (viewportInfo.isExternalScroller) {
        scrollerClassArray.push(scrollerStyles.externalScroller);
    }
    else {
        scrollerClassArray.push(scrollerStyles.autoScroll);
    }
    const scrollerClasses = classNames(scrollerClassArray);
    const classes = classNames([themeClasses]);
    const hasItemGridlines = gridlines?.item === 'visible';
    const hasTopGridline = gridlines?.top === 'visible';
    const hasBottomGridline = gridlines?.bottom === 'visible';
    const isSelectionEnabled = selectionMode === 'single' ||
        selectionMode === 'multiple' ||
        selectionMode === 'multipleToggle';
    const supportsMultipleSelection = selectionMode === 'multiple' || selectionMode === 'multipleToggle';
    // honor outer isTabbable context for embedded scenarios
    const { isTabbable } = useContext(TabbableModeContext);
    const getListBodyProps = () => {
        return {
            viewportInfo,
            listId,
            listData,
            children,
            itemPadding,
            hasItemGridlines,
            hasTopGridline,
            hasBottomGridline,
            selectedKeys: selectedKeys,
            isSelectionEnabled,
            isSingleSelectionEnabled: selectionMode === 'single',
            activeCell: currentActiveCell,
            currentItemKey: currentCell?.type === 'data' ? currentCell.rowKey : undefined,
            currentItemVariant: itemVariant,
            isShowFocusRing,
            isTabbableMode: isTabbable && isTabbableMode,
            hasMore,
            onLoadMore: onListLoadMore,
            promotedSection,
            noDataRenderer,
            skeletonRenderer,
            disabledInteractionKeys: disabledRowInteractionKeys,
            onSelectionChange: onListSelectionChange,
            isRtl,
            isHighContrast,
            reorderItemProps
        };
    };
    const isReorderEnabled = onReorder != null;
    const compositeDescription = ariaDescribedBy != null
        ? isReorderEnabled
            ? `${ariaDescribedBy} ${reorderInstructionsId}`
            : ariaDescribedBy
        : isReorderEnabled
            ? reorderInstructionsId
            : undefined;
    const ariaMultiSelectableProps = supportsMultipleSelection
        ? { 'aria-multiselectable': true }
        : isSelectionEnabled ? { 'aria-multiselectable': false } : {};
    return (jsxs(Fragment, { children: [jsx("div", { ...testIdProps, ref: rootRefCallback, "aria-colcount": 1, "aria-label": ariaLabel, "aria-labelledby": ariaLabelledBy, "aria-describedby": compositeDescription, "aria-rowcount": ariaRowCount, ...ariaMultiSelectableProps, class: classes, id: listId, role: 'grid', tabIndex: currentActiveCell != null || !isTabbable ? -1 : 0, ...interactionManagerProps, children: jsx("div", { ref: scrollerRefCallback, class: scrollerClasses, role: 'rowgroup', tabIndex: -1, ...scrollHandlingProps, children: jsx(FocusTracker, { isDisabled: !isTabbableMode, onFocusStartEdge: onFocusStartEdge, onFocusEndEdge: onFocusEndEdge, onStartTracking: onStartFocusTracking, children: jsx(TabbableModeContext.Provider, { value: { isTabbable: isTabbableMode }, children: jsx(ListBody, { ...getListBodyProps() }) }) }) }) }), isReorderEnabled && (jsx("span", { id: reorderInstructionsId, children: jsx(HiddenAccessible, { children: reorderInstructions }) })), contextMenuConfig != null && contextMenuContext != null && (jsx(ListContextMenu, { ...menuProps, contextMenuConfig: contextMenuConfig, contextMenuContext: contextMenuContext }))] }));
});

export { List };
