// @oracle/oraclejet-preact/es/PRIVATE_TableDisplay.js
/* @oracle/oraclejet-preact: undefined */
import { jsx, jsxs, Fragment as Fragment$1 } from 'preact/jsx-runtime';
import { memo, forwardRef } from 'preact/compat';
import { Fragment, createRef } from 'preact';
import { useLayoutEffect, useCallback, useState, useRef, useMemo } from 'preact/hooks';
import { c as classNames } from './classNames-4e12b00d.js';
import { u as useId } from './useId-03dbfdf0.js';
import { u as useUser } from './useUser-06d34d19.js';
import { z as getDataCellClassArray, A as getFormTextAlign, F as TABLE_NODATA_SELECTOR, B as TableSkeletonRow, C as getHeaderFooterCellClassArray, D as useSizingManager, v as useScrollHandling, E as TableColGroup } from './TableColGroup-d74bd4f5.js';
import { a as LoadMoreCollection } from './LoadMoreCollection-d9f22790.js';
import { tableRowStylesHC, tableRowStyles } from './PRIVATE_Table/themes/TableRowStyles.css.js';
import { F as FormContext } from './FormContext-55fb8000.js';
import { g as getClientHints } from './clientHints-a42b510c.js';
import { backgroundStyles, tableHeaderCellTextStyles, tableHeaderSortContainerStyles, tableHeaderRowStyles, tableFooterRowStyles, innerTableStyles, scrollerStyles } from './PRIVATE_Table/themes/Table.css.js';
import { TableNoDataRedwoodTheme } from './PRIVATE_Table/themes/redwood/TableNoDataTheme.js';
import { u as useComponentTheme } from './useComponentTheme-d2f9e47f.js';
import { u as useTranslationBundle } from './useTranslationBundle-4822502f.js';
import { justifyContentStyles } from './PRIVATE_TableCommon/themes/Common.css.js';
import { tableHeaderCellStyles } from './PRIVATE_Table/themes/TableHeaderCellStyles.css.js';
import { S as SortControl } from './SortControl-1044fcd9.js';
import { S as SortableContext } from './useSortableContext-88a0fc5a.js';
import { u as useDataManager } from './ScrollUtils-8a50c01e.js';
import { tableSkeletonStylesHC, tableSkeletonStyles } from './PRIVATE_Table/themes/TableSkeletonStyles.css.js';
import { TableRedwoodTheme } from './PRIVATE_Table/themes/redwood/TableTheme.js';
import { u as useTestId } from './useTestId-f24ac480.js';
import { tableDisplayStyles } from './PRIVATE_TableDisplay/themes/TableDisplayStyles.css.js';
import { d as dimensionInterpolations } from './dimensions-e4fbc23d.js';
import { m as mergeInterpolations } from './mergeInterpolations-30cd9a69.js';
import './LayerHost-7931b424.js';
import './index-aec1cc7f.js';
import './UNSAFE_Layer/themes/LayerHostStyles.css.js';
import './LayerHostStyles.styles.css';
import './useResizeObserver-13641643.js';
import './useAddBusyState-7a6af61f.js';
import './BusyStateContext-9ee9751d.js';
import './useBusyStateContext-4d3c27c0.js';
import './unsafeDomAccess-453eb74e.js';
import './collectionUtils-5313629f.js';
import './keys-49c92dbb.js';
import './Selector-23bccbe7.js';
import './BaseSelector-7cbc48bc.js';
import './useActive-fb6d4d52.js';
import './useToggle-8b7fcefe.js';
import './useFocus-38c95977.js';
import './useHover-85077adb.js';
import './mergeProps-88ea8306.js';
import './StyledCheckbox-d6c8f557.js';
import './Check-ac39fe00.js';
import './Icon-4983d06e.js';
import './size-af1e638c.js';
import './utils-b1eb54f9.js';
import './themeContract.css-1ffed3bf.js';
import './colorUtils-b3804ee8.js';
import './_curry1-63949a9b.js';
import './useTooltip-94af94d4.js';
import './useTooltipControlled-f2cfa454.js';
import './useTouch-4828df25.js';
import './useAnimation-fe39685e.js';
import './useThemeInterpolations-e144e8ec.js';
import './useColorScheme-c4f6caab.js';
import './useScale-434ad939.js';
import './theme-63551f30.js';
import './Theme-e6dec6db.js';
import './useDensity-1acb0153.js';
import './hooks/UNSAFE_useTooltip/themes/redwood/TooltipContentTheme.js';
import './hooks/UNSAFE_useTooltip/themes/TooltipContentStyles.css.js';
import './TooltipContentStyles.styles.css';
import './hooks/UNSAFE_useTooltip/themes/redwood/TooltipContentBaseTheme.css.js';
import './hooks/UNSAFE_useTooltip/themes/redwood/TooltipContentBaseTheme.styles.css';
import './hooks/UNSAFE_useTooltip/themes/redwood/TooltipContentVariants.css.js';
import './TooltipContentVariants.styles.css';
import './vanilla-extract-recipes-createRuntimeFn.esm-2aaf8c98.js';
import './tooltipUtils-af0abd38.js';
import './EnvironmentProvider-6a0a8212.js';
import './LayerManager-52a26c55.js';
import './Floating-4a82db94.js';
import './useFloating-b9d680f7.js';
import './positionUtils-f26caad9.js';
import './refUtils-76275c69.js';
import './useOutsideClick-b777a43e.js';
import './arrayUtils-35a58161.js';
import './Layer-535577b5.js';
import './UNSAFE_Layer/themes/LayerStyles.css.js';
import './LayerStyles.styles.css';
import './useModal-953082f2.js';
import './UNSAFE_Floating/themes/redwood/FloatingTheme.js';
import './UNSAFE_Floating/themes/FloatingStyles.css.js';
import './FloatingStyles.styles.css';
import './UNSAFE_Floating/themes/redwood/FloatingBaseTheme.css.js';
import './UNSAFE_Floating/themes/redwood/FloatingBaseTheme.styles.css';
import './UNSAFE_Floating/themes/redwood/FloatingVariants.css.js';
import './vanilla-extract-dynamic.esm-c909758c.js';
import './UNSAFE_Floating/themes/FloatingContract.css.js';
import './useUnsafeDomElementRef-32290273.js';
import './logger-c92f309c.js';
import './UNSAFE_Icon/themes/IconStyle.css.js';
import './IconStyle.styles.css';
import './CheckboxOff-3359a12d.js';
import './CheckboxOn-930b64de.js';
import './CheckboxMixed-c4594250.js';
import './HiddenAccessible-61197e7c.js';
import './HiddenAccessibleStyles.styles.css';
import './UNSAFE_Checkbox/themes/CheckboxIconContract.css.js';
import './UNSAFE_Checkbox/themes/redwood/CheckboxIconTheme.js';
import './UNSAFE_Checkbox/themes/CheckboxIconStyles.css.js';
import './CheckboxIconStyles.styles.css';
import './UNSAFE_Checkbox/themes/redwood/CheckboxIconBaseTheme.css.js';
import './UNSAFE_Checkbox/themes/redwood/CheckboxIconBaseTheme.styles.css';
import './UNSAFE_Checkbox/themes/redwood/CheckboxIconVariants.css.js';
import './CheckboxIconVariants.styles.css';
import './TabbableModeContext-7d8ad946.js';
import './useTabbableMode-090ed4b5.js';
import './StyledCheckbox.styles.css';
import './PRIVATE_BaseSelector/themes/SelectorStyles.css.js';
import './SelectorStyles.styles.css';
import './CollectionInteractionContext-7cf19abc.js';
import './useCollectionInteractionContext-3895c632.js';
import './PRIVATE_Table/themes/TableCellStyles.css.js';
import './Common.styles.css';
import './Table.styles.css';
import './TableRowStyles.styles.css';
import './TableCellStyles.styles.css';
import './PRIVATE_Table/themes/TableFooterCellStyles.css.js';
import './TableFooterCellStyles.styles.css';
import './Collection-e1ef8b8d.js';
import './useViewportIntersect-6ff65310.js';
import './PRIVATE_Table/themes/TableNoDataStyles.css.js';
import './TableNoDataStyles.styles.css';
import './PRIVATE_Table/themes/redwood/TableNoDataVariants.css.js';
import './TableHeaderCellStyles.styles.css';
import './UNSAFE_SortControl/themes/SortControl.css.js';
import './SortControl.styles.css';
import './IconButton-7f91481c.js';
import './BaseButton-41068c85.js';
import './UNSAFE_BaseButton/themes/redwood/BaseButtonTheme.js';
import './UNSAFE_BaseButton/themes/BaseButtonStyles.css.js';
import './BaseButtonStyles.styles.css';
import './UNSAFE_BaseButton/themes/redwood/BaseButtonVariants.css.js';
import './UNSAFE_BaseButton/themes/redwood/BaseButtonBaseTheme.styles.css';
import './BareButton-1e3df7cb.js';
import './useBareButton-ff4bd21e.js';
import './usePress-051f5adb.js';
import './usePressClick-945cc43b.js';
import './buttonUtils-55fb1df7.js';
import './id-83adac50.js';
import './ButtonLabelLayout-1631c997.js';
import './Text-834eb35e.js';
import './UNSAFE_Text/themes/TextStyles.css.js';
import './TextStyles.styles.css';
import './useButtonLabelLayoutTheme-448e671d.js';
import './UNSAFE_ButtonLabelLayout/themes/redwood/ButtonLabelLayoutTheme.js';
import './UNSAFE_ButtonLabelLayout/themes/ButtonLabelLayoutStyles.css.js';
import './ButtonLabelLayoutStyles.styles.css';
import './UNSAFE_ButtonLabelLayout/themes/redwood/ButtonLabelLayoutBaseTheme.css.js';
import './UNSAFE_ButtonLabelLayout/themes/redwood/ButtonLabelLayoutBaseTheme.styles.css';
import './UNSAFE_ButtonLabelLayout/themes/redwood/ButtonLabelLayoutVariants.css.js';
import './ButtonLabelLayoutVariants.styles.css';
import './EmbeddedIconButton-46434710.js';
import './UpParent-2712ea65.js';
import './DownParent-b696d9d0.js';
import './TableSkeletonStyles.styles.css';
import './PRIVATE_Table/themes/TableStyles.css.js';
import './TableStyles.styles.css';
import './PRIVATE_Table/themes/redwood/TableBaseTheme.css.js';
import './PRIVATE_Table/themes/redwood/TableBaseTheme.styles.css';
import './PRIVATE_Table/themes/redwood/TableVariants.css.js';
import './TableDisplayStyles.styles.css';
import './mergeDeepWithKey-33e3d1be.js';
import './_curry3-577eb45e.js';
import './_curry2-01c87f61.js';
import './_isObject-1fab0f5b.js';

function _defaultCellRenderer({ rowData, field }) {
    return field != null ? String(rowData[field]) : undefined;
}
/**
 * The internal component used to render a single cell in a TableDisplayRow.
 */
function TableDisplayCell({ tableId, rowData, field, columnKey, columnIndex, rowKey, rowIndex, isFinalRowIndex, isFinalColumnIndex, isFirstLogicalColumnIndex, isFinalLogicalColumnIndex, hasVerticalGridlines, hasHorizontalGridlines, hasFooters, hasVerticalUnderflow, isSticky, isAccRowHeader, renderer = _defaultCellRenderer, isStickyStartEdge, isStickyEndEdge, hasStartEdge, hasEndEdge, isRtl, isHighContrast, padding, horizontalAlignment = 'start', verticalAlignment = 'center' }) {
    const rowContext = { key: rowKey, data: rowData };
    const columnPadding = typeof padding === 'function' ? padding(rowContext) : padding;
    const hasLastHorizontalGridline = hasHorizontalGridlines && (hasVerticalUnderflow || !hasFooters);
    const cellClasses = classNames(getDataCellClassArray({
        isFirstColumnIndex: columnIndex === 0 || isFirstLogicalColumnIndex,
        isFinalColumnIndex: isFinalColumnIndex || isFinalLogicalColumnIndex,
        isFirstRowIndex: rowIndex === 0,
        isFinalRowIndex,
        hasVerticalGridlines,
        hasHorizontalGridlines,
        hasLastHorizontalGridline,
        isSticky,
        hasStartEdge,
        hasEndEdge,
        isStickyStartEdge,
        isStickyEndEdge,
        isRtl,
        isHighContrast,
        horizontalAlignment,
        verticalAlignment,
        padding: columnPadding
    }));
    const cellRendererProps = {
        rowData: rowData,
        rowKey: rowKey,
        field: field,
        columnKey: columnKey
    };
    return (jsx("td", { "aria-colindex": columnIndex + 1, class: cellClasses, role: isAccRowHeader ? 'rowheader' : 'cell', "data-oj-cell-type": 'data', "data-oj-column-key": columnKey, "data-oj-table-focusable": tableId, "data-oj-table-data": tableId, children: jsx(FormContext.Provider, { value: { textAlign: getFormTextAlign(horizontalAlignment, isRtl) }, children: renderer(cellRendererProps) }) }));
}

/**
 * The internal component used to render a single row in Table.
 */
function TableDisplayRow({ tableId, rowKey, rowData, rowIndex, isFinalRow, columnsArray, rowHeaderColumnKeys, hasVerticalGridlines, hasHorizontalGridlines, hasFooters, hasVerticalUnderflow, startStickyEdge, endStickyEdge, firstStickyColumnIndex, finalStickyColumnIndex, columnBorderEdges, isRtl, isHighContrast }) {
    const rowStyles = isHighContrast ? tableRowStylesHC : tableRowStyles;
    const classArray = [rowStyles.base, isFinalRow ? rowStyles.lastRowHeight : rowStyles.rowHeight];
    const classes = classNames(classArray);
    return (jsx("tr", { "aria-rowindex": rowIndex + 2, class: classes, role: 'row', "data-oj-key": rowKey, "data-oj-table-row": tableId, ...(typeof rowKey === 'number' && { 'data-oj-key-type': 'number' }), children: columnsArray.map((column, columnIndex) => {
            const isSticky = column.value.sticky === 'enabled';
            let isFirstLogicalColumnIndex = false;
            let isFinalLogicalColumnIndex = false;
            if (isSticky) {
                if (startStickyEdge != null && firstStickyColumnIndex === columnIndex) {
                    isFirstLogicalColumnIndex = true;
                }
                if (endStickyEdge != null && finalStickyColumnIndex === columnIndex) {
                    isFinalLogicalColumnIndex = true;
                }
            }
            const cellProps = {
                tableId: tableId,
                rowData: rowData,
                field: column.value.field,
                stickyColumn: column.value.sticky,
                columnKey: column.key,
                columnIndex: columnIndex,
                rowKey: rowKey,
                rowIndex: rowIndex,
                isFinalRowIndex: isFinalRow,
                isFinalColumnIndex: columnIndex === columnsArray.length - 1,
                isFirstLogicalColumnIndex: isFirstLogicalColumnIndex,
                isFinalLogicalColumnIndex: isFinalLogicalColumnIndex,
                hasVerticalGridlines: hasVerticalGridlines,
                hasHorizontalGridlines: hasHorizontalGridlines,
                hasFooters: hasFooters,
                hasVerticalUnderflow: hasVerticalUnderflow,
                isSticky: isSticky,
                isAccRowHeader: rowHeaderColumnKeys.has(column.key),
                renderer: column.value.renderer,
                isStickyStartEdge: columnIndex === startStickyEdge,
                isStickyEndEdge: columnIndex === endStickyEdge,
                hasStartEdge: columnBorderEdges[columnIndex].hasStart,
                hasEndEdge: columnBorderEdges[columnIndex].hasEnd,
                isRtl: isRtl,
                isHighContrast: isHighContrast,
                padding: column.value.padding,
                horizontalAlignment: column.value.alignment?.horizontal != null
                    ? column.value.alignment.horizontal
                    : column.value.horizontalAlignment,
                verticalAlignment: column.value.alignment?.vertical
            };
            return jsx(TableDisplayCell, { ...cellProps });
        }) }));
}
const _areStringSetsEquivalent = (set1, set2) => {
    if (set1 == null) {
        return set2 == null;
    }
    if (set2 == null) {
        return false;
    }
    return set1.size === set2.size && [...set1].every((x) => set2.has(x));
};
const _areColumnsArraysEquivalent = (columnsArray1, columnsArray2) => {
    if (columnsArray1.length === columnsArray2.length) {
        for (let i = 0; i < columnsArray1.length; i++) {
            if (columnsArray1[i].key !== columnsArray2[i].key ||
                columnsArray1[i].value !== columnsArray2[i].value) {
                return false;
            }
        }
        return true;
    }
    return false;
};
const MemoizeTableDisplayRow = memo(TableDisplayRow, (prev, next) => {
    if (prev && next) {
        return (prev.tableId === next.tableId &&
            prev.rowKey === next.rowKey &&
            prev.rowData === next.rowData &&
            prev.rowIndex === next.rowIndex &&
            prev.isFinalRow === next.isFinalRow &&
            _areColumnsArraysEquivalent(prev.columnsArray, next.columnsArray) &&
            prev.hasHorizontalGridlines === next.hasHorizontalGridlines &&
            prev.hasVerticalGridlines === next.hasVerticalGridlines &&
            (!next.isFinalRow ||
                (prev.hasFooters === next.hasFooters &&
                    prev.hasVerticalUnderflow === next.hasVerticalUnderflow)) &&
            _areStringSetsEquivalent(prev.rowHeaderColumnKeys, next.rowHeaderColumnKeys) &&
            prev.startStickyEdge === next.startStickyEdge &&
            prev.endStickyEdge === next.endStickyEdge &&
            prev.firstStickyColumnIndex === next.firstStickyColumnIndex &&
            prev.finalStickyColumnIndex === next.finalStickyColumnIndex &&
            prev.columnBorderEdges === next.columnBorderEdges &&
            prev.isRtl === next.isRtl &&
            prev.isHighContrast === next.isHighContrast);
    }
    return false;
});

/**
 * The internal component used to render an empty table cell in a TableDisplayBody.
 */
function TableDisplayNoData({ rootRef, tableId, columnsCount, ariaRowIndex, renderer }) {
    const browser = getClientHints().browser;
    const translations = useTranslationBundle('@oracle/oraclejet-preact');
    const rowClasses = classNames([tableRowStyles.base]);
    const { classes: themeClasses } = useComponentTheme(TableNoDataRedwoodTheme, {
        defaultRenderer: renderer == null ? 'isDefaultRenderer' : 'notDefaultRenderer',
        showFocusRing: 'notShowFocusRing'
    });
    const cellClasses = classNames([themeClasses]);
    function defaultContentRenderer() {
        return translations.collection_noData();
    }
    const contentRenderer = renderer != null ? renderer : defaultContentRenderer;
    // in firefox, table element heights are not stretched unless explicitly set in some cases
    useLayoutEffect(() => {
        if (browser === 'firefox') {
            const rootElement = rootRef.current;
            if (rootElement != null) {
                const tableBodyElement = rootElement.querySelector(`[data-oj-table-body='${tableId}']`);
                const noDataElement = rootElement.querySelector(`[${TABLE_NODATA_SELECTOR}='${tableId}']`);
                if (tableBodyElement != null && noDataElement != null) {
                    noDataElement.style.height = '';
                    noDataElement.style.height = `${tableBodyElement.getBoundingClientRect().height}px`;
                }
            }
        }
    });
    return (jsx("tr", { "aria-rowindex": ariaRowIndex, class: rowClasses, role: 'row', children: jsx("td", { "aria-colindex": 1, class: cellClasses, style: browser !== 'firefox' ? 'height: 100%;' : undefined, role: 'cell', colSpan: columnsCount, "data-oj-cell-type": 'noData', "data-oj-table-focusable": tableId, "data-oj-table-nodata": tableId, children: contentRenderer() }) }));
}

/**
 * The internal component used to render a the tbody of a Table.
 */
function TableDisplayBody({ rootRef, viewportInfo, tableId, isValidRender, getAccessibleRowHeaders, tableData, columnsArray, hasVerticalGridlines, hasHorizontalGridlines, hasFooters, hasVerticalUnderflow, hasMore, onLoadMore, noDataRenderer, startStickyEdge, endStickyEdge, firstStickyColumnIndex, finalStickyColumnIndex, columnBorderEdges, isRtl, isHighContrast }) {
    const browser = getClientHints().browser;
    // returns a TableRowContext based on ItemContext (added key)
    const getRowContext = useCallback((context) => {
        return {
            index: context.index,
            data: context.data.data,
            key: context.data.metadata.key
        };
    }, []);
    const collectionChildrenFunc = (context) => {
        const tableRowContext = getRowContext(context);
        const rowKey = tableRowContext.key;
        const rowProps = {
            tableId: tableId,
            columnsArray: columnsArray,
            rowHeaderColumnKeys: getAccessibleRowHeaders(tableRowContext),
            hasVerticalGridlines: hasVerticalGridlines,
            hasHorizontalGridlines: hasHorizontalGridlines,
            hasFooters: hasFooters,
            hasVerticalUnderflow: hasVerticalUnderflow,
            hasMore: hasMore,
            rowIndex: tableRowContext.index,
            rowKey: rowKey,
            rowData: tableRowContext.data,
            isFinalRow: tableRowContext.index === tableData.length - 1,
            startStickyEdge: startStickyEdge,
            endStickyEdge: endStickyEdge,
            firstStickyColumnIndex: firstStickyColumnIndex,
            finalStickyColumnIndex: finalStickyColumnIndex,
            columnBorderEdges: columnBorderEdges,
            isRtl: isRtl,
            isHighContrast: isHighContrast
        };
        return jsx(MemoizeTableDisplayRow, { ...rowProps });
    };
    const getNoDataProps = () => {
        return {
            rootRef: rootRef,
            tableId: tableId,
            columnsCount: columnsArray.length,
            ariaRowIndex: isValidRender ? 2 : 1,
            renderer: isValidRender ? noDataRenderer : undefined
        };
    };
    const isShowEmptyLoading = tableData.length === 0 && hasMore;
    const isShowNoData = tableData.length === 0 && !hasMore;
    return (jsx("tbody", { class: backgroundStyles.inherit, role: 'presentation', "data-oj-table-body": tableId, children: isShowNoData || !isValidRender ? (jsx(TableDisplayNoData, { ...getNoDataProps() })) : (jsxs(Fragment, { children: [jsx(LoadMoreCollection, { data: tableData, loadMoreIndicator: jsx(TableSkeletonRow, { tableId: tableId, hasTabIndex: false, colspan: columnsArray.length, isLoadMore: !isShowEmptyLoading, isShowFocusRing: false, isEmbedded: false, isHighContrast: isHighContrast }), hasMore: hasMore, onLoadMore: onLoadMore, viewportConfig: viewportInfo.viewportConfig, children: collectionChildrenFunc }), jsx("tr", { style: browser !== 'firefox' ? 'height: 100%;' : undefined, role: 'presentation', "data-oj-table-spacer": tableId })] })) }));
}

/**
 * The internal component used to render a single cell in a TableDisplayHeaderRow.
 */
function TableDisplayHeaderCell({ tableId, columnKey, columnIndex, isFinalColumnIndex, isFirstLogicalColumnIndex, isFinalLogicalColumnIndex, hasVerticalGridlines, isSticky, headerText, stickyTopOffset = 0, renderer, sortable = 'disabled', sortDirection = 'unsorted', onSortCriterionChange, isStickyStartEdge, isStickyEndEdge, hasStartEdge, hasEndEdge, isRtl, isAccColumnHeader, isHighContrast, padding, horizontalAlignment = 'start', verticalAlignment = 'center' }) {
    const cellClasses = classNames(getHeaderFooterCellClassArray({
        isHeader: true,
        isFirstColumnIndex: columnIndex === 0 || isFirstLogicalColumnIndex,
        isFinalColumnIndex: isFinalColumnIndex || isFinalLogicalColumnIndex,
        hasVerticalGridlines,
        isSticky,
        isStickyStartEdge,
        isStickyEndEdge,
        hasStartEdge,
        hasEndEdge,
        isRtl,
        horizontalAlignment,
        verticalAlignment,
        isHighContrast,
        padding
    }));
    const handleSortGesture = useCallback(() => {
        if (onSortCriterionChange != null) {
            const newDirection = sortDirection === 'ascending' ? 'descending' : 'ascending';
            onSortCriterionChange([{ key: columnKey, sortDirection: newDirection }]);
        }
    }, [columnKey, sortDirection, onSortCriterionChange]);
    const headerCellRendererProps = {
        key: columnKey,
        headerText: headerText,
        sortable: sortable,
        sortDirection: sortDirection
    };
    const getHeaderCellRenderer = useCallback(() => {
        if (renderer != null) {
            return renderer;
        }
        return ({ headerText, sortable, sortDirection }) => {
            const text = headerText != null ? (jsx("div", { class: tableHeaderCellTextStyles.base, children: jsx("span", { class: tableHeaderCellStyles.defaultRendererSpan, children: headerText }) })) : undefined;
            if (sortable === 'enabled' || sortDirection !== 'unsorted') {
                const isIconFirst = horizontalAlignment === 'end' ||
                    (!isRtl && horizontalAlignment === 'right') ||
                    (isRtl && horizontalAlignment === 'left');
                const containerClasses = classNames([
                    tableHeaderSortContainerStyles.base,
                    justifyContentStyles[horizontalAlignment]
                ]);
                return (jsxs("div", { class: containerClasses, children: [isIconFirst && jsx(SortControl, {}), text, !isIconFirst && jsx(SortControl, {})] }));
            }
            return text;
        };
    }, [renderer, horizontalAlignment, isRtl]);
    const getAriaSort = () => {
        if (sortDirection === 'ascending' || sortDirection === 'descending') {
            return { 'aria-sort': sortDirection };
        }
        return undefined;
    };
    return (jsx(SortableContext.Provider, { value: {
            isRendered: sortable === 'enabled',
            sortable,
            sortDirection,
            onSort: handleSortGesture,
            isTabbable: true
        }, children: jsx("th", { "aria-colindex": columnIndex + 1, ...getAriaSort(), class: cellClasses, style: `top:${stickyTopOffset}px;`, role: isAccColumnHeader ? 'columnheader' : 'cell', "data-oj-cell-type": 'header', "data-oj-column-key": columnKey, "data-oj-table-focusable": tableId, "data-oj-table-header": tableId, children: jsx(FormContext.Provider, { value: { textAlign: getFormTextAlign(horizontalAlignment, isRtl) }, children: getHeaderCellRenderer()(headerCellRendererProps) }) }) }));
}

/**
 * The internal component used to render a header region in Table.
 */
function TableDisplayHeader({ tableId, isRendered, columnsArray, hasVerticalGridlines, sortCriterion, onSortCriterionChange, stickyTopOffset, startStickyEdge, endStickyEdge, firstStickyColumnIndex, finalStickyColumnIndex, columnBorderEdges, isRtl, isHighContrast }) {
    const rowClasses = classNames([tableHeaderRowStyles.base]);
    return isRendered ? (jsx("thead", { class: backgroundStyles.inherit, role: 'presentation', children: jsx("tr", { "aria-rowindex": 1, class: rowClasses, role: 'row', "data-oj-table-header-row": tableId, children: columnsArray.map((column, columnIndex) => {
                const columnKey = column.key;
                let sortDirection;
                if (sortCriterion != null) {
                    for (const criterion of sortCriterion) {
                        if (criterion.key === columnKey) {
                            sortDirection = criterion.sortDirection;
                            break;
                        }
                    }
                }
                const isSticky = column.value.sticky === 'enabled';
                let isFirstLogicalColumnIndex = false;
                let isFinalLogicalColumnIndex = false;
                if (isSticky) {
                    if (startStickyEdge != null && firstStickyColumnIndex === columnIndex) {
                        isFirstLogicalColumnIndex = true;
                    }
                    if (endStickyEdge != null && finalStickyColumnIndex === columnIndex) {
                        isFinalLogicalColumnIndex = true;
                    }
                }
                const headerCellProps = {
                    tableId: tableId,
                    hasVerticalGridlines: hasVerticalGridlines,
                    columnKey: columnKey,
                    columnIndex: columnIndex,
                    isFinalColumnIndex: columnIndex === columnsArray.length - 1,
                    isFirstLogicalColumnIndex: isFirstLogicalColumnIndex,
                    isFinalLogicalColumnIndex: isFinalLogicalColumnIndex,
                    stickyTopOffset: stickyTopOffset,
                    isSticky: isSticky,
                    hasStartEdge: columnBorderEdges[columnIndex].hasStart,
                    hasEndEdge: columnBorderEdges[columnIndex].hasEnd,
                    headerText: column.value.headerText,
                    renderer: column.value.headerRenderer,
                    sortable: column.value.sortable,
                    sortDirection: sortDirection,
                    onSortCriterionChange: onSortCriterionChange,
                    isStickyStartEdge: columnIndex === startStickyEdge,
                    isStickyEndEdge: columnIndex === endStickyEdge,
                    isRtl: isRtl,
                    isAccColumnHeader: column.value.accessibleColumnHeader !== 'disabled',
                    isHighContrast: isHighContrast,
                    padding: column.value.headerPadding,
                    horizontalAlignment: column.value.headerAlignment?.horizontal != null
                        ? column.value.headerAlignment.horizontal
                        : column.value.horizontalAlignment,
                    verticalAlignment: column.value.headerAlignment?.vertical
                };
                return jsx(TableDisplayHeaderCell, { ...headerCellProps });
            }) }) })) : null;
}

function _defaultFooterCellRenderer({ footerText }) {
    return footerText != null ? footerText : undefined;
}
/**
 * The internal component used to render a single cell in a TableDisplayFooterRow.
 */
function TableDisplayFooterCell({ tableId, columnKey, columnIndex, isFinalColumnIndex, isFirstLogicalColumnIndex, isFinalLogicalColumnIndex, hasVerticalGridlines, isSticky, stickyBottomOffset = 0, footerText, renderer = _defaultFooterCellRenderer, isStickyStartEdge, isStickyEndEdge, hasStartEdge, hasEndEdge, isRtl, isHighContrast, padding, horizontalAlignment = 'start', verticalAlignment = 'center' }) {
    const cellClasses = classNames(getHeaderFooterCellClassArray({
        isHeader: false,
        isFirstColumnIndex: columnIndex === 0 || isFirstLogicalColumnIndex,
        isFinalColumnIndex: isFinalColumnIndex || isFinalLogicalColumnIndex,
        hasVerticalGridlines,
        isSticky,
        hasStartEdge,
        hasEndEdge,
        isStickyStartEdge,
        isStickyEndEdge,
        isRtl,
        horizontalAlignment,
        verticalAlignment,
        isHighContrast,
        padding
    }));
    const footerCellRendererProps = {
        key: columnKey,
        footerText: footerText
    };
    return (jsx("td", { "aria-colindex": columnIndex + 1, class: cellClasses, style: `bottom:${stickyBottomOffset}px;`, role: 'cell', "data-oj-cell-type": 'footer', "data-oj-column-key": columnKey, "data-oj-table-focusable": tableId, "data-oj-table-footer": tableId, children: jsx(FormContext.Provider, { value: { textAlign: getFormTextAlign(horizontalAlignment, isRtl) }, children: renderer(footerCellRendererProps) }) }));
}

/**
 * The internal component used to render a footer region in Table.
 */
function TableDisplayFooter({ tableId, isRendered, columnsArray, hasVerticalGridlines, stickyBottomOffset, startStickyEdge, endStickyEdge, firstStickyColumnIndex, finalStickyColumnIndex, columnBorderEdges, dataRowCount, isRtl, isHighContrast }) {
    const rowClasses = classNames([tableFooterRowStyles.base]);
    return isRendered ? (jsx("tfoot", { class: backgroundStyles.inherit, role: 'presentation', children: jsx("tr", { "aria-rowindex": Math.max(dataRowCount, 1) + 2, class: rowClasses, role: 'row', "data-oj-table-footer-row": tableId, children: columnsArray.map((column, columnIndex) => {
                const columnKey = column.key;
                const isSticky = column.value.sticky === 'enabled';
                let isFirstLogicalColumnIndex = false;
                let isFinalLogicalColumnIndex = false;
                if (isSticky) {
                    if (startStickyEdge != null && firstStickyColumnIndex === columnIndex) {
                        isFirstLogicalColumnIndex = true;
                    }
                    if (endStickyEdge != null && finalStickyColumnIndex === columnIndex) {
                        isFinalLogicalColumnIndex = true;
                    }
                }
                const footerCellProps = {
                    tableId: tableId,
                    columnKey: columnKey,
                    columnIndex: columnIndex,
                    isFinalColumnIndex: columnIndex === columnsArray.length - 1,
                    isFirstLogicalColumnIndex: isFirstLogicalColumnIndex,
                    isFinalLogicalColumnIndex: isFinalLogicalColumnIndex,
                    hasVerticalGridlines: hasVerticalGridlines,
                    stickyBottomOffset: stickyBottomOffset,
                    isSticky: isSticky,
                    hasStartEdge: columnBorderEdges[columnIndex].hasStart,
                    hasEndEdge: columnBorderEdges[columnIndex].hasEnd,
                    footerText: column.value.footerText,
                    renderer: column.value.footerRenderer,
                    isStickyStartEdge: columnIndex === startStickyEdge,
                    isStickyEndEdge: columnIndex === endStickyEdge,
                    isRtl: isRtl,
                    isHighContrast: isHighContrast,
                    padding: column.value.footerPadding,
                    horizontalAlignment: column.value.footerAlignment?.horizontal != null
                        ? column.value.footerAlignment.horizontal
                        : column.value.horizontalAlignment,
                    verticalAlignment: column.value.footerAlignment?.vertical
                };
                return jsx(TableDisplayFooterCell, { ...footerCellProps });
            }) }) })) : null;
}

/**
 * Component that renders data in a flat table.
 */
const TableDisplayBase = forwardRef(function TableDisplayBase({ 'aria-label': ariaLabel, 'aria-labelledby': ariaLabelledBy, 'aria-describedby': ariaDescribedBy, getAccessibleRowHeaders, columns, columnOrder, columnWidths, viewportConfig, layout = 'contents', data, getRowKey, gridlines, sortCriterion, onSortCriterionChange, horizontalScrollPositionOverride, verticalScrollPositionOverride, noDataRenderer, scrollerRef, testId }) {
    const hasMore = false;
    const onLoadMore = () => { };
    const tableId = useId();
    const testIdProps = useTestId(testId);
    // rootRef is callback-based to meet useResizeObserver hook's state management requirements
    const [rootRef, setRootRef] = useState(createRef());
    const rootRefCallback = useCallback((node) => {
        setRootRef({ current: node });
    }, []);
    const innerTableRef = useRef(null);
    const { direction, forcedColors } = useUser();
    const isHighContrast = forcedColors === 'active';
    const isRtl = direction === 'rtl';
    const { managedData: tableData, onManagedLoadMore: onTableLoadMore } = useDataManager({
        data,
        getRowKey,
        hasMore,
        onLoadMore
    });
    const { columnBorderEdges, isLayoutRequired, viewportInfo, appliedColumnWidths, sizingInfoRef, overflowInfoRef, hasVerticalUnderflow, stickyEdgesState, setStickyEdgesState, stickyColumnInfo, scrollerRefCallback } = useSizingManager({
        rootRef,
        innerTableRef,
        tableId,
        tableData,
        viewportConfig,
        externalScrollerAccessRef: scrollerRef,
        layout,
        columns,
        columnOrder,
        columnWidths,
        isRtl,
        isHighContrast
    });
    // memoize realized columns array on 'columns' and 'columnOrder'
    const columnsArray = useMemo(() => {
        const newColumnsArray = [];
        if (columnOrder != null) {
            for (const key of columnOrder) {
                if (columns[key] != null) {
                    newColumnsArray.push({
                        key: key,
                        value: columns[key]
                    });
                }
            }
        }
        else {
            for (const [key, value] of Object.entries(columns)) {
                newColumnsArray.push({
                    key: key,
                    value: value
                });
            }
        }
        return newColumnsArray;
    }, [columnOrder, columns]);
    const isPendingLayout = layout === 'pending';
    const scrollHandlingProps = useScrollHandling({
        rootRef,
        viewportInfo,
        tableId,
        isLayoutRequired,
        sizingInfoRef,
        overflowInfoRef,
        columnsArray,
        stickyColumnInfo,
        stickyEdgesState,
        setStickyEdgesState,
        fullColumnWidths: appliedColumnWidths?.fullColumnWidths,
        horizontalScrollPositionOverride,
        verticalScrollPositionOverride,
        isPendingLayout,
        isRtl
    });
    // aria-rowcount includes header and footer rows
    const hasFooters = columnsArray.some((column) => {
        return column.value.footerRenderer != null || column.value.footerText != null;
    });
    const ariaRowCount = isPendingLayout ? -1 : tableData.length + (hasFooters ? 2 : 1);
    const ariaColumnCount = Math.max(columnsArray.length, 1);
    // initialize base Table style classes
    const { classes: themeClasses } = useComponentTheme(TableRedwoodTheme, {
        pendingLayout: isPendingLayout ? 'isPendingLayout' : 'notPendingLayout',
        showFocusRing: 'notShowFocusRing',
        externalScroller: viewportInfo.isExternalScroller ? 'isExternalScroller' : 'notExternalScroller'
    });
    let scrollerStyle;
    let scrollerPadding = '';
    const scrollerClassArray = [scrollerStyles.base];
    if (viewportInfo.isExternalScroller) {
        scrollerClassArray.push(scrollerStyles.externalScroller);
    }
    else if (!sizingInfoRef.current.isScrollbarSizeKnown) {
        scrollerClassArray.push(scrollerStyles.forceScroll);
    }
    else {
        if (overflowInfoRef.current.isHorizontal) {
            scrollerClassArray.push(scrollerStyles.horizontalScroll);
        }
        else {
            scrollerClassArray.push(scrollerStyles.noHorizontalScroll);
        }
        if (overflowInfoRef.current.isVertical) {
            scrollerClassArray.push(scrollerStyles.verticalScroll);
        }
        else {
            scrollerClassArray.push(scrollerStyles.noVerticalScroll);
            scrollerPadding = `padding-inline-end: ${sizingInfoRef.current.defaultScrollbarWidth}px;`;
        }
    }
    // hide the table contents when layout is required and otherwise valid
    if (isLayoutRequired && columnsArray.length > 0) {
        scrollerStyle = `${scrollerPadding}opacity:0.001;`;
    }
    else {
        scrollerStyle = scrollerPadding;
    }
    const scrollerClasses = classNames(scrollerClassArray);
    const classes = classNames([themeClasses]);
    const tableClasses = classNames([innerTableStyles.base]);
    const startStickyEdge = stickyEdgesState.start;
    const endStickyEdge = stickyEdgesState.end;
    const hasVerticalGridlines = gridlines?.vertical === 'visible';
    const hasHorizontalGridlines = gridlines?.horizontal !== 'hidden';
    const getTableBodyProps = () => {
        return {
            rootRef: rootRef,
            viewportInfo: viewportInfo,
            tableId: tableId,
            isValidRender: columnsArray.length > 0,
            tableData: tableData,
            columnsArray: columnsArray,
            getAccessibleRowHeaders: getAccessibleRowHeaders,
            hasVerticalGridlines: hasVerticalGridlines,
            hasHorizontalGridlines: hasHorizontalGridlines,
            hasFooters: hasFooters,
            hasVerticalUnderflow: hasVerticalUnderflow,
            hasMore: hasMore,
            onLoadMore: onTableLoadMore,
            noDataRenderer: noDataRenderer,
            startStickyEdge: startStickyEdge,
            endStickyEdge: endStickyEdge,
            firstStickyColumnIndex: stickyColumnInfo.firstStickyColumnIndex,
            finalStickyColumnIndex: stickyColumnInfo.finalStickyColumnIndex,
            columnBorderEdges: columnBorderEdges,
            isRtl: isRtl,
            isHighContrast: isHighContrast
        };
    };
    const getTableColGroupProps = () => {
        let columnWidthsArray;
        if (!isLayoutRequired && appliedColumnWidths?.realColumnWidths != null) {
            columnWidthsArray = appliedColumnWidths.realColumnWidths;
        } /* else if (layout === 'fixed') {
          // default to 100px for each column when layout is not yet available
          columnWidthsArray = [];
          for (let i = 0; i < columnsArray.length; i++) {
            columnWidthsArray[i] = 100;
          }
        } */
        else {
            columnWidthsArray = [];
            for (let i = 0; i < columnsArray.length; i++) {
                columnWidthsArray[i] = undefined;
            }
        }
        return {
            tableId: tableId,
            isRendered: columnsArray.length > 0,
            columnWidthsArray: columnWidthsArray
        };
    };
    const getTableHeaderProps = () => {
        return {
            tableId: tableId,
            isRendered: columnsArray.length > 0,
            columnsArray: columnsArray,
            hasVerticalGridlines: hasVerticalGridlines,
            sortCriterion: sortCriterion,
            onSortCriterionChange: onSortCriterionChange,
            stickyTopOffset: viewportInfo.viewportConfig.scrollerOffsetTop,
            startStickyEdge: startStickyEdge,
            endStickyEdge: endStickyEdge,
            firstStickyColumnIndex: stickyColumnInfo.firstStickyColumnIndex,
            finalStickyColumnIndex: stickyColumnInfo.finalStickyColumnIndex,
            columnBorderEdges: columnBorderEdges,
            isRtl: isRtl,
            isHighContrast: isHighContrast
        };
    };
    const getTableFooterProps = () => {
        return {
            tableId: tableId,
            isRendered: hasFooters,
            columnsArray: columnsArray,
            hasVerticalGridlines: hasVerticalGridlines,
            stickyBottomOffset: viewportInfo.viewportConfig.scrollerOffsetBottom,
            startStickyEdge: startStickyEdge,
            endStickyEdge: endStickyEdge,
            firstStickyColumnIndex: stickyColumnInfo.firstStickyColumnIndex,
            finalStickyColumnIndex: stickyColumnInfo.finalStickyColumnIndex,
            columnBorderEdges: columnBorderEdges,
            dataRowCount: tableData.length,
            isRtl: isRtl,
            isHighContrast: isHighContrast
        };
    };
    const getTableStyle = () => {
        if (!isLayoutRequired && appliedColumnWidths?.realColumnWidths != null) {
            let totalWidth = 0;
            for (let i = 0; i < appliedColumnWidths.realColumnWidths.length; i++) {
                totalWidth += appliedColumnWidths.realColumnWidths[i];
            }
            return `width:${totalWidth}px;table-layout:fixed;`;
        }
        else if (columnsArray.length > 0) {
            /* if (layout === 'fixed') {
              // default to 100px for each column when layout is not yet available
              return `width:${columnsArray.length * 100}px;table-layout:fixed;`;
            } else { */
            // shrink table down as much as possible when a 'contents' layout is required, otherwise
            // the sampled column widths will not reflect the true weight sizes with line-clamping
            return 'width:max-content;will-change:width;';
            // }
        }
        // no columns are present
        return 'width:100%;';
    };
    const skeletonStyles = isHighContrast ? tableSkeletonStylesHC : tableSkeletonStyles;
    const getPendingLayoutSkeletons = () => {
        return isPendingLayout ? (jsx("div", { class: skeletonStyles.pendingLayout, children: jsx("table", { role: 'presentation', style: 'width:100%;table-layout:fixed;', children: jsx("tbody", { role: 'presentation', children: jsx(TableSkeletonRow, { tableId: tableId, hasTabIndex: false, colspan: ariaColumnCount, isPendingLayout: true, isShowFocusRing: false, isEmbedded: false, isHighContrast: isHighContrast }) }) }) })) : undefined;
    };
    return (jsx(Fragment$1, { children: jsxs("div", { ...testIdProps, ref: rootRefCallback, class: classes, id: tableId, children: [jsx("div", { ref: scrollerRefCallback, "aria-colcount": ariaColumnCount, "aria-label": ariaLabel, "aria-labelledby": ariaLabelledBy, "aria-describedby": ariaDescribedBy, "aria-rowcount": ariaRowCount, class: scrollerClasses, style: scrollerStyle, role: 'table', tabIndex: 0, ...scrollHandlingProps, children: jsxs("table", { ref: innerTableRef, "aria-hidden": isPendingLayout ? true : undefined, class: tableClasses, role: 'presentation', style: getTableStyle(), "data-oj-table-inner-table": tableId, children: [jsx(TableColGroup, { ...getTableColGroupProps() }), jsx(TableDisplayHeader, { ...getTableHeaderProps() }), jsx(TableDisplayBody, { ...getTableBodyProps() }), jsx(TableDisplayFooter, { ...getTableFooterProps() })] }) }), getPendingLayoutSkeletons()] }) }));
});

const mergedInterpolations = mergeInterpolations([
    ...Object.values(dimensionInterpolations)
]);
/**
 * Component that renders data in rows of columns.
 * This component is non-virtualized and renders all rows provided in the data prop.
 */
const TableDisplay = forwardRef(({ 'aria-label': ariaLabel, 'aria-labelledby': ariaLabelledBy, 'aria-describedby': ariaDescribedBy, getAccessibleRowHeaders, children, columns, columnOrder, columnWidths, viewportConfig, layout = 'contents', data, getRowKey, gridlines = { horizontal: 'visible', vertical: 'hidden' }, sortCriterion = [], onSortCriterionChange, horizontalScrollPositionOverride, verticalScrollPositionOverride, noDataRenderer, scrollerRef, testId, ...dimensions }, ref) => {
    const { ...mergedStyles } = mergedInterpolations(dimensions);
    const isExternalScroller = viewportConfig != null;
    const classArray = [tableDisplayStyles.base];
    classArray.push(isExternalScroller ? tableDisplayStyles.externalScroller : tableDisplayStyles.internalScroller);
    const tableDisplayClasses = classNames(classArray);
    return (jsx("div", { class: tableDisplayClasses, style: mergedStyles, children: jsx(TableDisplayBase, { "aria-label": ariaLabel, "aria-labelledby": ariaLabelledBy, "aria-describedby": ariaDescribedBy, getAccessibleRowHeaders: getAccessibleRowHeaders, children: children, columns: columns, columnOrder: columnOrder, columnWidths: columnWidths, viewportConfig: viewportConfig, layout: layout, data: data, getRowKey: getRowKey, gridlines: gridlines, horizontalScrollPositionOverride: horizontalScrollPositionOverride, verticalScrollPositionOverride: verticalScrollPositionOverride, sortCriterion: sortCriterion, onSortCriterionChange: onSortCriterionChange, noDataRenderer: noDataRenderer, scrollerRef: scrollerRef, testId: testId, ref: ref }) }));
});

export { TableDisplay };
