// @oracle/oraclejet-preact/es/RadioItem-94909632.js
/* @oracle/oraclejet-preact: undefined */
import { jsx, jsxs } from 'preact/jsx-runtime';
import { createRef, useRef, useEffect, useCallback } from 'preact/compat';
import { u as useComponentTheme } from './useComponentTheme-d2f9e47f.js';
import { u as useFocus } from './useFocus-38c95977.js';
import './FormContext-55fb8000.js';
import { u as useFormContext } from './useFormContext-bb9dc478.js';
import './FormFieldContext-47ca0c20.js';
import { u as useFormFieldContext } from './useFormFieldContext-daf7a676.js';
import { u as useId } from './useId-03dbfdf0.js';
import './TabbableModeContext-7d8ad946.js';
import { u as useTabbableMode } from './useTabbableMode-090ed4b5.js';
import { u as useInteractionStyle } from './useInteractionStyle-99e32c23.js';
import { u as useTranslationBundle } from './useTranslationBundle-4822502f.js';
import { u as useCheckboxRadioContext } from './CheckboxRadioContext-fe425383.js';
import { S as SvgCheck } from './Check-ac39fe00.js';
import { S as SvgRadioButtonOff } from './RadioButtonOff-e005ab12.js';
import { S as SvgRadioButtonOn } from './RadioButtonOn-ca8fdd4a.js';
import { H as HiddenAccessible } from './HiddenAccessible-61197e7c.js';
import { u as useRadioSetContext, a as useRadioItemContext } from './RadioItemContext-fe419b75.js';
import './UserAssistanceStyles.styles.css';
import './UNSAFE_UserAssistance/themes/redwood/UserAssistanceBaseTheme.styles.css';
import './UNSAFE_UserAssistance/themes/redwood/UserAssistanceVariants.css.js';
import 'preact/hooks';
import './Flex-03129963.js';
import './MessageBannerStyles.styles.css';
import './UNSAFE_MessageBanner/themes/redwood/MessageBannerBaseTheme.styles.css';
import './UNSAFE_MessageBanner/themes/redwood/MessageBannerVariants.css.js';
import 'preact';
import './LayerHost-7931b424.js';
import './TooltipContentStyles.styles.css';
import './hooks/UNSAFE_useTooltip/themes/redwood/TooltipContentBaseTheme.styles.css';
import './hooks/UNSAFE_useTooltip/themes/redwood/TooltipContentVariants.css.js';
import './tooltipUtils-af0abd38.js';
import './IconStyle.styles.css';
import { g as getClientHints } from './clientHints-a42b510c.js';
import './IconButton-7f91481c.js';
import './MessageStyles.styles.css';
import './MessageFormattingUtils-cdb25115.js';
import './logger-c92f309c.js';
import './TransitionGroup-5fd80dc9.js';
import './MessagesContext-76544845.js';
import './ComponentMessageStyles.styles.css';
import './UNSAFE_ComponentMessage/themes/redwood/ComponentMessageBaseTheme.styles.css';
import './UNSAFE_ComponentMessage/themes/redwood/ComponentMessageVariants.css.js';
import './InputGroupContext-20bdbd6e.js';
import './Popup-992b6c27.js';
import './UNSAFE_Separator/themes/SeparatorStyles.css.js';
import './LabelStyles.styles.css';
import './UNSAFE_Label/themes/redwood/LabelBaseTheme.styles.css';
import './UNSAFE_Label/themes/redwood/LabelVariants.css.js';
import { I as IconUserAssistance } from './IconUserAssistance-61978211.js';
import { radioIconVars } from './UNSAFE_RadioItem/themes/RadioIconContract.css.js';
import { RadioIconRedwoodTheme } from './UNSAFE_RadioItem/themes/redwood/RadioIconTheme.js';
import { RadioRedwoodTheme } from './UNSAFE_RadioItem/themes/redwood/RadioTheme.js';

/**
 * Returns the accessibility text for a radio button's selection state.
 *
 * The text is determined based on the client's platform (Mac or iOS/PC/Android) and the radio button's selection state.
 * Mac VoiceOver announces "selected" while others use "checked".
 *
 * @param {BundleType} translations The translation bundle containing the necessary text.
 * @param {boolean} isSelected Whether the radio button is currently selected.
 * @returns {string} The accessibility text for the radio button's selection state.
 */
const getRadioSelectionText = (translations, isSelected) => {
    const isMac = getClientHints().platform === 'mac';
    const translationKey = isMac
        ? isSelected
            ? 'radioButton_selected'
            : 'radioButton_unselected'
        : isSelected
            ? 'radioButton_checked'
            : 'radioButton_unchecked';
    return translations[translationKey]();
};
/**
 * RadioItem is an input element of type radio.
 */
const RadioItem = ({ value, children, ...assistanceProps }) => {
    const { isReadonly: isFormReadonly, isDisabled: isFormDisabled } = useFormContext();
    const { isDisabled: isFieldDisabled, isReadonly: isFieldReadonly } = useFormFieldContext();
    const { name, value: groupValue, onCommit, setFocusIndex } = useRadioSetContext();
    const { isFocusRingShown } = useCheckboxRadioContext();
    const { index } = useRadioItemContext();
    const { isTabbable } = useTabbableMode();
    const id = useId();
    const { isFocus, focusProps } = useFocus();
    const labelRef = createRef();
    const isSelected = groupValue === value;
    const isDisabled = isFieldDisabled ?? isFormDisabled;
    const isReadonly = isFieldReadonly ?? isFormReadonly;
    const { interactionProps, applyActiveStyle } = useInteractionStyle();
    const inputRef = useRef(null);
    // JET-70100 - ensure radio button is scrolled into view when focused and focus ring is shown
    useEffect(() => {
        const isActiveElement = document.activeElement === inputRef?.current;
        if (!isActiveElement && isFocus && isFocusRingShown) {
            labelRef?.current?.scrollIntoView({ block: 'nearest' });
        }
    }, [isFocus, isFocusRingShown, labelRef]);
    const { assistiveText, helpSourceLink } = assistanceProps;
    const hasUserAssistive = !isDisabled && !isReadonly && !!(assistiveText || helpSourceLink);
    const helpAvailableId = hasUserAssistive ? `${id}-help-available` : undefined;
    const uaId = hasUserAssistive ? `${id}-icon-ua` : undefined;
    const translations = useTranslationBundle('@oracle/oraclejet-preact');
    const helpAvailableString = translations.formControl_helpAvailable();
    const IconComp = isReadonly
        ? isSelected
            ? SvgCheck
            : SvgRadioButtonOff
        : isSelected
            ? SvgRadioButtonOn
            : SvgRadioButtonOff;
    const tabIndex = isTabbable !== false && ((index === 0 && !groupValue) || isSelected) ? 0 : -1;
    const accessibilityText = `${children || ''} ${translations.formControl_readOnly()} ${getRadioSelectionText(translations, isSelected)}`;
    const handleFocus = useCallback(() => {
        setFocusIndex?.(index);
    }, [setFocusIndex, index]);
    const handleLabelClick = useCallback(() => {
        inputRef.current?.focus();
        setFocusIndex?.(index);
    }, [inputRef, setFocusIndex, index]);
    const handleChange = useCallback((e) => {
        e.stopPropagation();
        onCommit?.({ previousValue: groupValue, value });
    }, [onCommit, groupValue, value]);
    /**
     * JET-76970
     * Safari's `:focus-visible` is applied only during keyboard focus (e.g., Tab).
     * However, when a child of a card receives focus via arrow navigation, Safari
     * ignores `:focus-visible` even if it would match, and it only shows a focus
     * outline at navigation endpoints.
     *
     * For consistency:
     * - For Safari, add a class that forces the ring whenever focus is inside the
     *   card during keyboard navigation.
     * - For other browsers or readonly items, rely on native `:focus-visible` selectors.
     *
     * Known issue:
     * - Because of the timeout in useCollectionFocusRing, a long press on a item can
     *   sometimes trigger a focus ring, whereas a normal click does not.
     */
    const isSafari = getClientHints().browser === 'safari';
    const focusWithRing = isFocus && isFocusRingShown;
    const isFocusRingSelector = isReadonly || (!isSafari && focusWithRing);
    const isForcedFocusRing = isSafari && focusWithRing;
    const { classes: radioClasses, styles: { valueLabelStyle } } = useComponentTheme(RadioRedwoodTheme, {
        disabled: isDisabled ? 'isDisabled' : 'notDisabled',
        focusRing: isForcedFocusRing
            ? 'isForcedFocusRing'
            : isFocusRingSelector
                ? 'isFocusRingSelector'
                : undefined
    });
    const { classes: radioIconClasses } = useComponentTheme(RadioIconRedwoodTheme, {
        disabled: isDisabled ? 'isDisabled' : 'notDisabled',
        selected: isSelected ? 'isSelected' : 'notSelected',
        readonly: isReadonly ? 'isReadonly' : 'notReadonly',
        active: applyActiveStyle ? 'isActive' : 'notActive'
    });
    return (jsx("div", { children: jsxs("label", { class: radioClasses, ref: labelRef, ...(isReadonly
                ? {
                    onFocus: handleFocus,
                    onClick: handleLabelClick
                }
                : {}), children: [jsxs(HiddenAccessible, { children: [isReadonly ? (jsx("div", { "aria-describedby": helpAvailableId, id: id, ref: inputRef, tabIndex: tabIndex, "data-checked": isSelected, ...focusProps, children: accessibilityText })) : (jsx("input", { type: "radio", name: name, id: id, ref: inputRef, value: value, ...(!isDisabled && {
                                onChange: handleChange
                            }), checked: isSelected, disabled: isDisabled, "aria-describedby": helpAvailableId, tabIndex: tabIndex, ...focusProps })), hasUserAssistive && (jsx(HiddenAccessible, { id: helpAvailableId, isHidden: true, children: helpAvailableString }))] }), jsx("span", { "aria-hidden": "true", class: radioIconClasses, ...interactionProps, children: jsx(IconComp, { size: radioIconVars.iconSize, color: "currentColor" }) }), jsxs("span", { class: valueLabelStyle, children: [jsx("span", { ...interactionProps, children: children }), hasUserAssistive && (jsx(IconUserAssistance, { ...assistanceProps, isTabbable: isTabbable && isFocus, id: uaId, variant: "radio" }))] })] }) }));
};

export { RadioItem as R, getRadioSelectionText as g };
