// @oracle/oraclejet-preact/es/StyledSwitch-dbf29d95.js
/* @oracle/oraclejet-preact: undefined */
import { jsxs, jsx } from 'preact/jsx-runtime';
import { forwardRef } from 'preact/compat';
import { useState, useCallback } from 'preact/hooks';
import { u as useTestId } from './useTestId-f24ac480.js';
import { styles } from './PRIVATE_StyledSwitch/themes/StyledSwitchStyles.css.js';
import './TabbableModeContext-7d8ad946.js';
import { u as useTabbableMode } from './useTabbableMode-090ed4b5.js';
import { u as useComponentTheme } from './useComponentTheme-d2f9e47f.js';
import { StyledSwitchRedwoodTheme } from './PRIVATE_StyledSwitch/themes/redwood/StyledSwitchTheme.js';
import { c as classNames } from './classNames-4e12b00d.js';
import { H as HiddenAccessible } from './HiddenAccessible-61197e7c.js';
import { u as useTranslationBundle } from './useTranslationBundle-4822502f.js';
import { u as useId } from './useId-03dbfdf0.js';
import { u as useActive } from './useActive-fb6d4d52.js';
import { m as mergeProps } from './mergeProps-88ea8306.js';
import { l } from './stringUtils-0fe25200.js';

/**
 * Provides standardized event handlers for switch UI components (such as StyledSwitch).
 *
 * - Handles mouse click and keyboard activation (Space/Enter) on <kbd>keyup</kbd> (legacy behavior).
 * - Implements "gating" for Enter and Space: onChange will only trigger on keyup if that key's keydown
 *   happened on the component, preventing accidental activation when focus shifts mid-keypress.
 *   (e.g. Enter down elsewhere, focus moves to switch, Enter up: switch does NOT toggle)
 * - Ignores events if `isDisabled` or `isReadonly` are true.
 * - Internally uses isEnterPressed and isSpacePressed state to track legitimate activation origin.
 * - Forwards `onChange` when value should toggle.
 *
 * @param {SwitchHandlersParams} params - Props controlling switch behavior and callbacks.
 * @returns {{
 *   onClick: (e: MouseEvent) => void;
 *   onKeyDown: (e: KeyboardEvent) => void;
 *   onKeyUp: (e: KeyboardEvent) => void;
 * }} Event handlers to be spread on the root switch element.
 *
 * @example
 *   const { onClick, onKeyDown, onKeyUp } = useSwitchHandlers(...);
 *   <div tabIndex={0} onClick={onClick} onKeyDown={onKeyDown} onKeyUp={onKeyUp} ... />
 */
function useSwitchHandlers({ isDisabled, isReadonly, onChange }) {
    const [isEnterPressed, setIsEnterPressed] = useState(false);
    const [isSpacePressed, setIsSpacePressed] = useState(false);
    const handleClick = useCallback((e) => {
        if (isDisabled || isReadonly)
            return;
        e.preventDefault();
        e.stopPropagation();
        onChange?.(e);
    }, [isDisabled, isReadonly, onChange]);
    /**
     * Gating logic for Enter and Space:
     * Only "arms" activation on the *first* keydown (when event.repeat is false).
     * This means that if a key is held down elsewhere and focus moves to the switch
     * (e.g. user presses and holds Enter on a button, button sets focus to switch
     * on activation, then more auto-repeated keydowns will NOT mistakenly arm the switch),
     * only an Enter or Space press that originated with the switch actually focused
     * will allow a subsequent keyup to activate.
     */
    const handleKeyDown = useCallback((e) => {
        if (isDisabled || isReadonly)
            return;
        if (e.code === 'Enter' && !e.repeat) {
            setIsEnterPressed(true);
            e.preventDefault();
        }
        if (e.code === 'Space' && !e.repeat) {
            setIsSpacePressed(true);
            e.preventDefault();
        }
    }, [isDisabled, isReadonly]);
    const handleKeyUp = useCallback((e) => {
        if (isDisabled || isReadonly)
            return;
        if (e.code === 'Enter' && isEnterPressed) {
            setIsEnterPressed(false);
            e.preventDefault();
            e.stopPropagation();
            onChange?.(e);
        }
        if (e.code === 'Space' && isSpacePressed) {
            setIsSpacePressed(false);
            e.preventDefault();
            e.stopPropagation();
            onChange?.(e);
        }
    }, [isDisabled, isReadonly, isEnterPressed, isSpacePressed, onChange]);
    return {
        onClick: handleClick,
        onKeyDown: handleKeyDown,
        onKeyUp: handleKeyUp,
        onKeyPress: handleKeyPress
    };
}
/**
 * Prevent default on Space otherwise holding down the Space will cause the page to scroll to the bottom.
 */
function handleKeyPress(e) {
    if (e.code === 'Space') {
        e.preventDefault();
    }
}

/**
 * StyledSwitch is a controlled component that displays a stylized binary switch (on/off).
 * It supports accessibility attributes and allows parent components to control its checked and disabled states.
 *
 * This component is implemented with `forwardRef` to allow parent components to obtain a ref
 * to the interactive switch element (track). This is important for programmatically managing focus.
 *
 * @example
 *   <StyledSwitch isChecked={checked} onChange={handleChange} ref={switchRef} />
 */
const StyledSwitch = forwardRef(({ 'aria-describedby': ariaDescribedBy, 'aria-invalid': ariaInvalid, 'aria-label': ariaLabel, 'aria-labelledby': ariaLabelledBy, baseId, isChecked = false, isDisabled, isReadonly, readonlyOnText, readonlyOffText, onChange, testId }, ref = null) => {
    const generatedBaseId = useId();
    const readonlyBaseId = baseId ?? generatedBaseId;
    const readonlyId = `${readonlyBaseId}-readonly`;
    // Prefer isDisabled over isReadonly:
    const resolvedDisabled = !!isDisabled;
    const resolvedReadonly = !resolvedDisabled && !!isReadonly;
    const mergedAriaDescribedBy = resolvedReadonly
        ? l([ariaDescribedBy, readonlyId])
        : ariaDescribedBy;
    const ariaChecked = !!isChecked;
    const { tabbableModeProps } = useTabbableMode(resolvedDisabled);
    const { isActive, activeProps } = useActive({ isDisabled: isDisabled || isReadonly });
    const { classes: themeContainerClass } = useComponentTheme(StyledSwitchRedwoodTheme, {
        target: 'container',
        checked: isChecked ? 'isChecked' : 'notChecked',
        disabled: resolvedDisabled ? 'isDisabled' : 'notDisabled',
        readonly: resolvedReadonly ? 'isReadonly' : 'notReadonly'
    });
    const containerClass = classNames([styles.container, themeContainerClass]);
    const { classes: themeTrackClass } = useComponentTheme(StyledSwitchRedwoodTheme, {
        target: 'track',
        checked: isChecked ? 'isChecked' : 'notChecked',
        disabled: resolvedDisabled ? 'isDisabled' : 'notDisabled',
        readonly: resolvedReadonly ? 'isReadonly' : 'notReadonly',
        active: isActive ? 'isActive' : 'notActive'
    });
    const trackClass = classNames([styles.track, themeTrackClass]);
    const { classes: themeThumbClass } = useComponentTheme(StyledSwitchRedwoodTheme, {
        target: 'thumb',
        checked: isChecked ? 'isChecked' : 'notChecked',
        disabled: resolvedDisabled ? 'isDisabled' : 'notDisabled',
        readonly: resolvedReadonly ? 'isReadonly' : 'notReadonly'
    });
    const thumbClass = classNames([styles.thumb, themeThumbClass]);
    // Use the new hook for events
    const { onClick, onKeyDown, onKeyUp, onKeyPress } = useSwitchHandlers({
        isDisabled: resolvedDisabled,
        isReadonly: resolvedReadonly,
        onChange
    });
    let readonlyText = undefined;
    if (resolvedReadonly) {
        if (isChecked) {
            readonlyText = readonlyOnText;
        }
        else {
            readonlyText = readonlyOffText;
        }
    }
    const rootTestIdProps = useTestId(testId);
    const thumbTestIdProps = useTestId(testId ? testId + '_thumb' : undefined);
    const translations = useTranslationBundle('@oracle/oraclejet-preact');
    const readonlyLabel = translations.formControl_readOnly();
    return (jsxs("div", { class: styles.root, ...rootTestIdProps, children: [jsx("div", { class: containerClass, children: jsx("div", { role: "switch", ...tabbableModeProps, class: trackClass, "aria-checked": ariaChecked, "aria-readonly": resolvedReadonly || undefined, "aria-disabled": resolvedDisabled || undefined, "aria-invalid": ariaInvalid, "aria-label": ariaLabel, "aria-labelledby": ariaLabelledBy, "aria-describedby": mergedAriaDescribedBy, ref: ref, ...mergeProps(activeProps, { onKeyUp, onKeyDown, onClick, onKeyPress }), children: jsx("div", { class: thumbClass, ...thumbTestIdProps, children: readonlyText }) }) }), resolvedReadonly && (jsx(HiddenAccessible, { id: readonlyId, isHidden: true, children: readonlyLabel }))] }));
});

export { StyledSwitch as S };
