// @oracle/oraclejet-preact/es/TruncatingLink-298c6343.js
/* @oracle/oraclejet-preact: undefined */
import { jsxs, Fragment, jsx } from 'preact/jsx-runtime';
import { c as classNames } from './classNames-4e12b00d.js';
import { u as usePressClick } from './usePressClick-945cc43b.js';
import { u as useComponentTheme } from './useComponentTheme-d2f9e47f.js';
import { u as useInteractionStyle } from './useInteractionStyle-99e32c23.js';
import { LinkRedwoodTheme } from './UNSAFE_Link/themes/redwood/LinkTheme.js';
import './TabbableModeContext-7d8ad946.js';
import { u as useTabbableMode } from './useTabbableMode-090ed4b5.js';
import { u as useTestId } from './useTestId-f24ac480.js';
import { u as useDetectTruncation } from './useDetectTruncation-071767c7.js';
import { u as useTooltip } from './useTooltip-94af94d4.js';
import { useState } from 'preact/hooks';
import { m as mergeProps } from './mergeProps-88ea8306.js';
import './LayerHost-7931b424.js';
import 'preact/compat';
import './TooltipContentStyles.styles.css';
import './hooks/UNSAFE_useTooltip/themes/redwood/TooltipContentBaseTheme.styles.css';
import './hooks/UNSAFE_useTooltip/themes/redwood/TooltipContentVariants.css.js';
import './tooltipUtils-af0abd38.js';
import { g as getClientHints } from './clientHints-a42b510c.js';
import { T as Text } from './Text-834eb35e.js';

const noop = () => { };
/**
 * TruncatingLink component creates a hyperlink that shows a tooltip and ellipsis at the end of text when truncated.
 */
function TruncatingLink({ href, variant, placement = 'standalone', target, testId, 'aria-describedby': ariaDescribedBy, 'aria-label': accessibleLabel, 'aria-labelledby': ariaLabelledBy, onClick, isDisabled, lineClamp, children }) {
    const { detectTruncationProps, isTextTruncated } = useDetectTruncation();
    const { interactionProps, applyPseudoHoverStyle, applyHoverStyle } = useInteractionStyle();
    const clientHints = getClientHints();
    const { classes } = useComponentTheme(LinkRedwoodTheme, {
        needsEventsHover: applyHoverStyle ? 'isNeedsEventsHover' : 'notNeedsEventsHover',
        pseudoHover: applyPseudoHoverStyle ? 'isPseudoHover' : 'notPseudoHover',
        variant,
        placement,
        disabled: isDisabled ? 'isDisabled' : 'notDisabled',
        truncatingLinkStyles: placement !== 'embedded' ? 'styles' : 'none',
        truncatingLinkWhiteSpace: placement !== 'embedded' && !lineClamp ? 'nowrap' : 'none',
        safariStyles: clientHints.browser === 'safari' ? 'isSafari' : 'notSafari'
    });
    const { isTabbable } = useTabbableMode();
    const linkClasses = classNames([classes]);
    const { pressProps } = usePressClick(onClick ?? noop);
    const testIdProps = useTestId(testId);
    const [shownText] = useState(children || href);
    const { tooltipContent, tooltipProps } = useTooltip({
        text: shownText,
        isDisabled: !isTextTruncated || isDisabled
    });
    const { 'aria-describedby': notNeeded, ...validTooltipProps } = tooltipProps; //remove aria-describedby to prevent screen readers reading the same text multiple times
    return (jsxs(Fragment, { children: [jsx("a", { ...mergeProps(validTooltipProps, detectTruncationProps, interactionProps, testIdProps, !isDisabled ? pressProps : {}), tabIndex: isTabbable && !isDisabled ? 0 : -1, href: !isDisabled ? href : undefined, "aria-describedby": ariaDescribedBy, "aria-label": accessibleLabel, "aria-labelledby": ariaLabelledBy, target: target, class: linkClasses, children: jsx(Text, { lineClamp: lineClamp, size: "md", variant: "inherit", children: shownText }) }), tooltipContent] }));
}

export { TruncatingLink as T };
