// @oracle/oraclejet-preact/es/UNSAFE_Dialog.js
/* @oracle/oraclejet-preact: undefined */
export { D as Dialog } from './Dialog-83368f52.js';
import { useState, useRef } from 'preact/hooks';
import 'preact/jsx-runtime';
import 'preact/compat';
import './WindowOverlay-7264c44a.js';
import './mergeInterpolations-30cd9a69.js';
import './classNames-4e12b00d.js';
import './mergeDeepWithKey-33e3d1be.js';
import './_curry3-577eb45e.js';
import './_curry1-63949a9b.js';
import './_curry2-01c87f61.js';
import './_isObject-1fab0f5b.js';
import './vanilla-extract-dynamic.esm-c909758c.js';
import './UNSAFE_WindowOverlay/themes/WindowOverlayStyles.css.js';
import './WindowOverlayStyles.styles.css';
import './UNSAFE_WindowOverlay/themes/WindowOverlayContract.css.js';
import './useUser-06d34d19.js';
import './LayerHost-7931b424.js';
import './index-aec1cc7f.js';
import 'preact';
import './UNSAFE_Layer/themes/LayerHostStyles.css.js';
import './LayerHostStyles.styles.css';
import './useComponentTheme-d2f9e47f.js';
import './logger-c92f309c.js';
import './UNSAFE_WindowOverlay/themes/redwood/WindowOverlayTheme.js';
import './UNSAFE_WindowOverlay/themes/redwood/WindowOverlayBaseTheme.css.js';
import './UNSAFE_WindowOverlay/themes/redwood/WindowOverlayBaseTheme.styles.css';
import './UNSAFE_WindowOverlay/themes/redwood/WindowOverlayVariants.css.js';
import './vanilla-extract-recipes-createRuntimeFn.esm-2aaf8c98.js';
import './useTestId-f24ac480.js';
import './useAnimation-fe39685e.js';
import './refUtils-76275c69.js';
import './UNSAFE_Dialog/themes/redwood/DialogTheme.js';
import './UNSAFE_Dialog/themes/DialogStyles.css.js';
import './DialogStyles.styles.css';
import './UNSAFE_Dialog/themes/redwood/DialogBaseTheme.css.js';
import './ImageVars.styles.css';
import './UNSAFE_Dialog/themes/redwood/DialogBaseTheme.styles.css';
import './UNSAFE_Dialog/themes/redwood/DialogVariants.css.js';
import './FocusTrap-b59d5752.js';
import './tabbableUtils-c8311ec6.js';
import './clientHints-a42b510c.js';
import './head-8f89fe66.js';
import './_arity-be492b9e.js';
import './_isArray-694cc52d.js';
import './_isString-675f1de9.js';
import './FocusTracker-70300f1c.js';
import './PRIVATE_FocusTracker/themes/FocusTrackerStyles.css.js';
import './FocusTrackerStyles.styles.css';
import './Layer-535577b5.js';
import './UNSAFE_Layer/themes/LayerStyles.css.js';
import './LayerStyles.styles.css';
import './useThemeInterpolations-e144e8ec.js';
import './useColorScheme-c4f6caab.js';
import './useScale-434ad939.js';
import './theme-63551f30.js';
import './Theme-e6dec6db.js';
import './useDensity-1acb0153.js';
import './unsafeDomAccess-453eb74e.js';
import './Modal-c0f730b8.js';
import './useModal-953082f2.js';
import './UNSAFE_Modal/themes/ModalStyles.css.js';
import './ModalStyles.styles.css';
import './popupUtils-51366a29.js';
import './ImageVars.css-7b3180c8.js';
import './useId-03dbfdf0.js';
import './dimensions-e4fbc23d.js';
import './size-af1e638c.js';
import './utils-b1eb54f9.js';
import './themeContract.css-1ffed3bf.js';
import './colorUtils-b3804ee8.js';
import './useResizeObserver-13641643.js';
import './useAddBusyState-7a6af61f.js';
import './BusyStateContext-9ee9751d.js';
import './useBusyStateContext-4d3c27c0.js';
import './useDraggable-e1aaa8f3.js';
import './positionUtils-f26caad9.js';
import './Floating-4a82db94.js';
import './useFloating-b9d680f7.js';
import './useOutsideClick-b777a43e.js';
import './arrayUtils-35a58161.js';
import './UNSAFE_Floating/themes/redwood/FloatingTheme.js';
import './UNSAFE_Floating/themes/FloatingStyles.css.js';
import './FloatingStyles.styles.css';
import './UNSAFE_Floating/themes/redwood/FloatingBaseTheme.css.js';
import './UNSAFE_Floating/themes/redwood/FloatingBaseTheme.styles.css';
import './UNSAFE_Floating/themes/redwood/FloatingVariants.css.js';
import './UNSAFE_Floating/themes/FloatingContract.css.js';
import './useUnsafeDomElementRef-32290273.js';
import './IconButton-7f91481c.js';
import './BaseButton-41068c85.js';
import './useHover-85077adb.js';
import './useToggle-8b7fcefe.js';
import './useActive-fb6d4d52.js';
import './mergeProps-88ea8306.js';
import './UNSAFE_BaseButton/themes/redwood/BaseButtonTheme.js';
import './UNSAFE_BaseButton/themes/BaseButtonStyles.css.js';
import './BaseButtonStyles.styles.css';
import './UNSAFE_BaseButton/themes/redwood/BaseButtonVariants.css.js';
import './UNSAFE_BaseButton/themes/redwood/BaseButtonBaseTheme.styles.css';
import './BareButton-1e3df7cb.js';
import './useBareButton-ff4bd21e.js';
import './usePress-051f5adb.js';
import './usePressClick-945cc43b.js';
import './TabbableModeContext-7d8ad946.js';
import './useTabbableMode-090ed4b5.js';
import './buttonUtils-55fb1df7.js';
import './id-83adac50.js';
import './ButtonLabelLayout-1631c997.js';
import './Text-834eb35e.js';
import './UNSAFE_Text/themes/TextStyles.css.js';
import './TextStyles.styles.css';
import './useButtonLabelLayoutTheme-448e671d.js';
import './UNSAFE_ButtonLabelLayout/themes/redwood/ButtonLabelLayoutTheme.js';
import './UNSAFE_ButtonLabelLayout/themes/ButtonLabelLayoutStyles.css.js';
import './ButtonLabelLayoutStyles.styles.css';
import './UNSAFE_ButtonLabelLayout/themes/redwood/ButtonLabelLayoutBaseTheme.css.js';
import './UNSAFE_ButtonLabelLayout/themes/redwood/ButtonLabelLayoutBaseTheme.styles.css';
import './UNSAFE_ButtonLabelLayout/themes/redwood/ButtonLabelLayoutVariants.css.js';
import './ButtonLabelLayoutVariants.styles.css';
import './useTooltip-94af94d4.js';
import './useTooltipControlled-f2cfa454.js';
import './useFocus-38c95977.js';
import './useTouch-4828df25.js';
import './hooks/UNSAFE_useTooltip/themes/redwood/TooltipContentTheme.js';
import './hooks/UNSAFE_useTooltip/themes/TooltipContentStyles.css.js';
import './TooltipContentStyles.styles.css';
import './hooks/UNSAFE_useTooltip/themes/redwood/TooltipContentBaseTheme.css.js';
import './hooks/UNSAFE_useTooltip/themes/redwood/TooltipContentBaseTheme.styles.css';
import './hooks/UNSAFE_useTooltip/themes/redwood/TooltipContentVariants.css.js';
import './TooltipContentVariants.styles.css';
import './tooltipUtils-af0abd38.js';
import './EnvironmentProvider-6a0a8212.js';
import './LayerManager-52a26c55.js';
import './Close-5a460764.js';
import './Icon-4983d06e.js';
import './UNSAFE_Icon/themes/IconStyle.css.js';
import './IconStyle.styles.css';
import './Resizable-b88279bc.js';
import './PRIVATE_Resizable/themes/ResizableStyles.css.js';
import './ResizableStyles.styles.css';
import './useBreakpoints-99efa183.js';
import './useTranslationBundle-4822502f.js';
import './animationUtils-5f409a6c.js';

/**
 * A custom hook that adds drag and/or resize capabilities to a component (e. g. Dialog).
 * The hook maintains the original and current position and size and returns handlers and
 * props to be spread on the component to make it "flexible". In addition, allows its consumer
 * to be notified when the drag or resize life-cycle events occur.
 */
const useFlexibleComponent = ({ isDisabled = false, isDraggable = false, isResizable = false, onDragStart: onDragStartHandler, onDragMove: onDragMoveHandler, onDragEnd: onDragEndHandler, onResizeStart: onResizeStartHandler, onResize: onResizeHandler, onResizeEnd: onResizeEndHandler }) => {
    const [coords, _setCoords] = useState();
    const [dimensions, setDimensions] = useState();
    const coordsRef = useRef(null);
    const origPosRef = useRef(null);
    const origSizeRef = useRef(null);
    const setCoords = (coords) => {
        _setCoords({ x: coords.x, y: coords.y });
        coordsRef.current = { x: coords.x, y: coords.y };
    };
    if (isDisabled || (!isDraggable && !isResizable)) {
        return {};
    }
    const onDragStart = (pos) => {
        setCoords({ x: pos.x, y: pos.y });
        origPosRef.current = { x: pos.x, y: pos.y };
        onDragStartHandler?.({
            originalPosition: { x: origPosRef.current.x, y: origPosRef.current.y },
            position: { x: pos.x, y: pos.y }
        });
    };
    const onDragMove = (pos) => {
        setCoords({ x: pos.x, y: pos.y });
        onDragMoveHandler?.({
            originalPosition: { x: origPosRef.current.x, y: origPosRef.current.y },
            position: { x: pos.x, y: pos.y }
        });
    };
    const onDragEnd = (pos) => {
        setCoords({ x: pos.x, y: pos.y });
        onDragEndHandler?.({
            originalPosition: { x: origPosRef.current.x, y: origPosRef.current.y },
            position: { x: pos.x, y: pos.y }
        });
        origPosRef.current = null;
    };
    const onResizeStart = (rect) => {
        setCoords({ x: rect.x, y: rect.y });
        origPosRef.current = { x: rect.x, y: rect.y };
        setDimensions({ width: (rect.width + 'px'), height: (rect.height + 'px') });
        origSizeRef.current = { width: rect.width, height: rect.height };
        onResizeStartHandler?.({
            originalPosition: { x: origPosRef.current.x, y: origPosRef.current.y },
            originalSize: { width: origSizeRef.current.width, height: origSizeRef.current.height },
            position: { x: rect.x, y: rect.y },
            size: { width: rect.width, height: rect.height }
        });
    };
    const onResize = (rect) => {
        setCoords({ x: rect.x, y: rect.y });
        setDimensions({ width: (rect.width + 'px'), height: (rect.height + 'px') });
        onResizeHandler?.({
            originalPosition: { x: origPosRef.current.x, y: origPosRef.current.y },
            originalSize: { width: origSizeRef.current.width, height: origSizeRef.current.height },
            position: { x: rect.x, y: rect.y },
            size: { width: rect.width, height: rect.height }
        });
    };
    const onResizeEnd = (rect) => {
        setCoords({ x: rect.x, y: rect.y });
        setDimensions({ width: (rect.width + 'px'), height: (rect.height + 'px') });
        onResizeEndHandler?.({
            originalPosition: { x: origPosRef.current.x, y: origPosRef.current.y },
            originalSize: { width: origSizeRef.current.width, height: origSizeRef.current.height },
            position: { x: rect.x, y: rect.y },
            size: { width: rect.width, height: rect.height }
        });
        origPosRef.current = null;
        origSizeRef.current = null;
    };
    return {
        ...(isDraggable && { onDragStart, onDragMove, onDragEnd }),
        ...(isResizable && { onResizeStart, onResize, onResizeEnd }),
        ...(coords && { anchorRef: coordsRef }),
        ...(dimensions && { width: dimensions.width, height: dimensions.height })
    };
};

export { useFlexibleComponent };
