// @oracle/oraclejet-preact/es/UNSAFE_LabelledLink.js
/* @oracle/oraclejet-preact: undefined */
import { jsx, jsxs } from 'preact/jsx-runtime';
import { useRef, useState, useCallback } from 'preact/hooks';
import { d as detectTruncation } from './truncationUtils-e37f95ee.js';
import { L as Label } from './Label-5d1ae8d7.js';
import './keyboardUtils-986fe923.js';
import 'preact/compat';
import { R as ReadonlyTextField } from './StyledTextField-7605d76a.js';
import './ObfuscatedTextFieldInputStyles.styles.css';
import { u as useAccessibleContext } from './useAccessibleContext-5744de8b.js';
import './logger-c92f309c.js';
import './FormContext-55fb8000.js';
import { u as useFormContext } from './useFormContext-bb9dc478.js';
import { F as FormFieldContext } from './FormFieldContext-47ca0c20.js';
import './TabbableModeContext-7d8ad946.js';
import './LayerHost-7931b424.js';
import { t as textInterpolations } from './textAlign-ad252afa.js';
import './FormControlUtilsStyles.styles.css';
import './ReadonlyTextFieldInputStyles.styles.css';
import './UNSAFE_TextField/themes/redwood/ReadonlyTextFieldInputVariants.css.js';
import './LabelValueLayoutStyles.styles.css';
import './UNSAFE_TextField/themes/TextFieldStyles.css.js';
import './SkeletonStyles.styles.css';
import './TextFieldLoadingStyles.styles.css';
import './TextFieldStyles.styles.css';
import './UNSAFE_TextField/themes/redwood/TextFieldBaseTheme.styles.css';
import './UNSAFE_TextField/themes/redwood/TextFieldVariants.css.js';
import './LabelStyles.styles.css';
import './UNSAFE_Label/themes/redwood/LabelBaseTheme.styles.css';
import './UNSAFE_Label/themes/redwood/LabelVariants.css.js';
import './UNSAFE_TextField/themes/FormLayoutStyles.css.js';
import './UNSAFE_TextField/themes/redwood/FormLayoutBaseTheme.styles.css';
import './Flex-03129963.js';
import './LayoutStyles.styles.css';
import './InputGroupContext-20bdbd6e.js';
import './TooltipContentStyles.styles.css';
import './hooks/UNSAFE_useTooltip/themes/redwood/TooltipContentBaseTheme.styles.css';
import './hooks/UNSAFE_useTooltip/themes/redwood/TooltipContentVariants.css.js';
import './tooltipUtils-af0abd38.js';
import './LiveRegionStyles.styles.css';
import './UserAssistanceStyles.styles.css';
import './UNSAFE_UserAssistance/themes/redwood/UserAssistanceBaseTheme.styles.css';
import './UNSAFE_UserAssistance/themes/redwood/UserAssistanceVariants.css.js';
import { a as InlineUserAssistance } from './InlineUserAssistance-890293dd.js';
import './MessageBannerStyles.styles.css';
import './UNSAFE_MessageBanner/themes/redwood/MessageBannerBaseTheme.styles.css';
import './UNSAFE_MessageBanner/themes/redwood/MessageBannerVariants.css.js';
import 'preact';
import './IconStyle.styles.css';
import './IconButton-7f91481c.js';
import './MessageStyles.styles.css';
import './MessageFormattingUtils-cdb25115.js';
import './TransitionGroup-5fd80dc9.js';
import './MessagesContext-76544845.js';
import './HiddenAccessibleStyles.styles.css';
import './ComponentMessageStyles.styles.css';
import './UNSAFE_ComponentMessage/themes/redwood/ComponentMessageBaseTheme.styles.css';
import './UNSAFE_ComponentMessage/themes/redwood/ComponentMessageVariants.css.js';
import './Popup-992b6c27.js';
import './UNSAFE_Separator/themes/SeparatorStyles.css.js';
import { u as useFocusWithin } from './useFocusWithin-30b1e2d8.js';
import './UNSAFE_Popup/themes/redwood/PopupBaseTheme.styles.css';
import './UNSAFE_Popup/themes/redwood/PopupVariants.css.js';
import { V as View } from './View-5bdd9b3c.js';
import { u as useFormFieldContextProps } from './useFormFieldContextProps-31bbc12f.js';
import { u as useTextField } from './useTextField-eb9db737.js';
import { H as HiddenAccessible } from './HiddenAccessible-61197e7c.js';
import { c as classNames } from './classNames-4e12b00d.js';
import { u as useComponentTheme } from './useComponentTheme-d2f9e47f.js';
import { m as mergeInterpolations } from './mergeInterpolations-30cd9a69.js';
import { LabelledLinkRedwoodTheme } from './UNSAFE_LabelledLink/themes/redwood/LabelledLinkTheme.js';
import { L as Link } from './Link-8503aab2.js';
import { T as TruncatingLink } from './TruncatingLink-298c6343.js';
import './useFormFieldContext-daf7a676.js';
import './useTooltip-94af94d4.js';
import './useTooltipControlled-f2cfa454.js';
import './useId-03dbfdf0.js';
import './useHover-85077adb.js';
import './useToggle-8b7fcefe.js';
import './useFocus-38c95977.js';
import './useTouch-4828df25.js';
import './mergeProps-88ea8306.js';
import './useAnimation-fe39685e.js';
import './useTestId-f24ac480.js';
import './useThemeInterpolations-e144e8ec.js';
import './useColorScheme-c4f6caab.js';
import './useScale-434ad939.js';
import './theme-63551f30.js';
import './Theme-e6dec6db.js';
import './useDensity-1acb0153.js';
import './unsafeDomAccess-453eb74e.js';
import './hooks/UNSAFE_useTooltip/themes/redwood/TooltipContentTheme.js';
import './hooks/UNSAFE_useTooltip/themes/TooltipContentStyles.css.js';
import './hooks/UNSAFE_useTooltip/themes/redwood/TooltipContentBaseTheme.css.js';
import './TooltipContentVariants.styles.css';
import './vanilla-extract-recipes-createRuntimeFn.esm-2aaf8c98.js';
import './clientHints-a42b510c.js';
import './CompactLabelAssistance-e2fcfcab.js';
import './useTranslationBundle-4822502f.js';
import './useTabbableMode-090ed4b5.js';
import './Help-32a9689a.js';
import './Icon-4983d06e.js';
import './size-af1e638c.js';
import './utils-b1eb54f9.js';
import './themeContract.css-1ffed3bf.js';
import './colorUtils-b3804ee8.js';
import './_curry1-63949a9b.js';
import './UNSAFE_Icon/themes/IconStyle.css.js';
import './UNSAFE_UserAssistance/themes/UserAssistanceStyles.css.js';
import './UNSAFE_Label/themes/redwood/LabelTheme.js';
import './UNSAFE_Label/themes/LabelStyles.css.js';
import './UNSAFE_Label/themes/redwood/LabelBaseTheme.css.js';
import './UNSAFE_TextField/themes/redwood/TextFieldTheme.js';
import './UNSAFE_TextField/themes/redwood/TextFieldBaseTheme.css.js';
import './TextFieldVariants.styles.css';
import './useTextFieldInputHandlers-8122048f.js';
import './formControlUtils-84e2af0e.js';
import './TextFieldUtils-8232bca7.js';
import './UNSAFE_TextField/themes/FormControlUtilsStyles.css.js';
import './UNSAFE_TextField/themes/redwood/TextFieldInputTheme.js';
import './UNSAFE_TextField/themes/TextFieldInputStyles.css.js';
import './TextFieldInputStyles.styles.css';
import './UNSAFE_TextField/themes/redwood/TextFieldInputVariants.css.js';
import './TextFieldInputVariants.styles.css';
import './LabelValueLayout-c6b9945a.js';
import './StyledLabelValueLayout-bb05f474.js';
import './UNSAFE_TextField/themes/redwood/FormLayoutTheme.js';
import './UNSAFE_TextField/themes/redwood/FormLayoutBaseTheme.css.js';
import './formControlUtils-6fc3a497.js';
import './LayoutStyles.css-6b02421a.js';
import './UNSAFE_Skeleton/themes/redwood/SkeletonTheme.js';
import './UNSAFE_Skeleton/themes/SkeletonStyles.css.js';
import './UNSAFE_Skeleton/themes/redwood/SkeletonBaseTheme.css.js';
import './UNSAFE_Skeleton/themes/redwood/SkeletonBaseTheme.styles.css';
import './UNSAFE_Skeleton/themes/redwood/SkeletonVariants.css.js';
import './useInputGroupContext-1d19411e.js';
import './useUser-06d34d19.js';
import './UNSAFE_TextField/themes/redwood/ReadonlyTextFieldTheme.js';
import './UNSAFE_TextField/themes/ReadonlyTextFieldStyles.css.js';
import './ReadonlyTextFieldStyles.styles.css';
import './UNSAFE_TextField/themes/redwood/ReadonlyTextFieldVariants.css.js';
import './ReadonlyTextFieldVariants.styles.css';
import './index-aec1cc7f.js';
import './UNSAFE_Layer/themes/LayerHostStyles.css.js';
import './LayerHostStyles.styles.css';
import './text.styles.css';
import './vanilla-extract-sprinkles-createRuntimeSprinkles.esm-2d655d37.js';
import './boxalignment-8980856d.js';
import './arrayUtils-35a58161.js';
import './boxalignment.styles.css';
import './dimensions-e4fbc23d.js';
import './flexbox-529f25da.js';
import './flexbox.styles.css';
import './flexitem-7b7f7920.js';
import './flexitem.styles.css';
import './FlexStyles.styles.css';
import './EnvironmentProvider-6a0a8212.js';
import './LayerManager-52a26c55.js';
import './Floating-4a82db94.js';
import './useFloating-b9d680f7.js';
import './positionUtils-f26caad9.js';
import './refUtils-76275c69.js';
import './useOutsideClick-b777a43e.js';
import './Layer-535577b5.js';
import './UNSAFE_Layer/themes/LayerStyles.css.js';
import './LayerStyles.styles.css';
import './useModal-953082f2.js';
import './UNSAFE_Floating/themes/redwood/FloatingTheme.js';
import './UNSAFE_Floating/themes/FloatingStyles.css.js';
import './FloatingStyles.styles.css';
import './UNSAFE_Floating/themes/redwood/FloatingBaseTheme.css.js';
import './UNSAFE_Floating/themes/redwood/FloatingBaseTheme.styles.css';
import './UNSAFE_Floating/themes/redwood/FloatingVariants.css.js';
import './vanilla-extract-dynamic.esm-c909758c.js';
import './UNSAFE_Floating/themes/FloatingContract.css.js';
import './useUnsafeDomElementRef-32290273.js';
import './ComponentMessageContainer-d16c8b1f.js';
import './MessagesManager-47000438.js';
import './SuccessS-b5192d40.js';
import './ErrorS-3a890bf6.js';
import './InformationS-d53c2a91.js';
import './WarningS-3e53d87e.js';
import './PRIVATE_Message/themes/MessageStyles.css.js';
import './soundUtils-88264cb5.js';
import './useMessagesContext-9e1dbe91.js';
import './Transition-f9501682.js';
import './UNSAFE_ComponentMessage/themes/redwood/ComponentMessageTheme.js';
import './UNSAFE_ComponentMessage/themes/ComponentMessageStyles.css.js';
import './UNSAFE_ComponentMessage/themes/redwood/ComponentMessageBaseTheme.css.js';
import './UNSAFE_UserAssistance/themes/redwood/UserAssistanceTheme.js';
import './UNSAFE_UserAssistance/themes/redwood/UserAssistanceBaseTheme.css.js';
import './InlineHelp-9eaac91c.js';
import './MessageBannerVariants.styles.css';
import './BaseButton-41068c85.js';
import './useActive-fb6d4d52.js';
import './UNSAFE_BaseButton/themes/redwood/BaseButtonTheme.js';
import './UNSAFE_BaseButton/themes/BaseButtonStyles.css.js';
import './BaseButtonStyles.styles.css';
import './UNSAFE_BaseButton/themes/redwood/BaseButtonVariants.css.js';
import './UNSAFE_BaseButton/themes/redwood/BaseButtonBaseTheme.styles.css';
import './BareButton-1e3df7cb.js';
import './useBareButton-ff4bd21e.js';
import './usePress-051f5adb.js';
import './usePressClick-945cc43b.js';
import './buttonUtils-55fb1df7.js';
import './id-83adac50.js';
import './ButtonLabelLayout-1631c997.js';
import './Text-834eb35e.js';
import './UNSAFE_Text/themes/TextStyles.css.js';
import './TextStyles.styles.css';
import './useButtonLabelLayoutTheme-448e671d.js';
import './UNSAFE_ButtonLabelLayout/themes/redwood/ButtonLabelLayoutTheme.js';
import './UNSAFE_ButtonLabelLayout/themes/ButtonLabelLayoutStyles.css.js';
import './ButtonLabelLayoutStyles.styles.css';
import './UNSAFE_ButtonLabelLayout/themes/redwood/ButtonLabelLayoutBaseTheme.css.js';
import './UNSAFE_ButtonLabelLayout/themes/redwood/ButtonLabelLayoutBaseTheme.styles.css';
import './UNSAFE_ButtonLabelLayout/themes/redwood/ButtonLabelLayoutVariants.css.js';
import './ButtonLabelLayoutVariants.styles.css';
import './getLocale-1c41e86a.js';
import './stringUtils-0fe25200.js';
import './tabbableUtils-c8311ec6.js';
import './head-8f89fe66.js';
import './_arity-be492b9e.js';
import './_curry3-577eb45e.js';
import './_curry2-01c87f61.js';
import './_isArray-694cc52d.js';
import './_isString-675f1de9.js';
import './FocusTrap-b59d5752.js';
import './FocusTracker-70300f1c.js';
import './PRIVATE_FocusTracker/themes/FocusTrackerStyles.css.js';
import './FocusTrackerStyles.styles.css';
import './Modal-c0f730b8.js';
import './UNSAFE_Modal/themes/ModalStyles.css.js';
import './ModalStyles.styles.css';
import './popupUtils-51366a29.js';
import './WindowOverlay-7264c44a.js';
import './UNSAFE_WindowOverlay/themes/WindowOverlayStyles.css.js';
import './WindowOverlayStyles.styles.css';
import './UNSAFE_WindowOverlay/themes/WindowOverlayContract.css.js';
import './UNSAFE_WindowOverlay/themes/redwood/WindowOverlayTheme.js';
import './UNSAFE_WindowOverlay/themes/redwood/WindowOverlayBaseTheme.css.js';
import './UNSAFE_WindowOverlay/themes/redwood/WindowOverlayBaseTheme.styles.css';
import './UNSAFE_WindowOverlay/themes/redwood/WindowOverlayVariants.css.js';
import './usePopupAnimation-96d2bc83.js';
import './useAnimationStatus-b881f39a.js';
import './popupAnimationUtils-175151d9.js';
import './LiveRegion-cebac18d.js';
import './UNSAFE_Popup/themes/PopupContract.css.js';
import './UNSAFE_Popup/themes/redwood/PopupTheme.js';
import './UNSAFE_Popup/themes/redwood/PopupBaseTheme.css.js';
import './PopupStyles.styles.css';
import './colors-03807160.js';
import './SeparatorStyles.styles.css';
import './borders-98a63040.js';
import './padding-90265467.js';
import './aria-efa0582c.js';
import './equals-625352e8.js';
import './_isObject-1fab0f5b.js';
import './useMessageSeverity-65295e8f.js';
import './mergeDeepWithKey-33e3d1be.js';
import './UNSAFE_LabelledLink/themes/LabelledLinkStyles.css.js';
import './LabelledLinkStyles.styles.css';
import './UNSAFE_LabelledLink/themes/redwood/LabelledLinkVariants.css.js';
import './LabelledLinkVariants.styles.css';
import './useInteractionStyle-99e32c23.js';
import './UNSAFE_Link/themes/redwood/LinkTheme.js';
import './UNSAFE_Link/themes/LinkStyles.css.js';
import './LinkStyles.styles.css';
import './UNSAFE_Link/themes/redwood/LinkBaseTheme.css.js';
import './UNSAFE_Link/themes/redwood/LinkBaseTheme.styles.css';
import './UNSAFE_Link/themes/redwood/LinkVariants.css.js';
import './LinkVariants.styles.css';
import './useDetectTruncation-071767c7.js';
import './useResizeObserver-13641643.js';
import './useAddBusyState-7a6af61f.js';
import './BusyStateContext-9ee9751d.js';
import './useBusyStateContext-4d3c27c0.js';

const HiddenLabel = ({ children, id }) => (jsx(HiddenAccessible, { isHidden: true, children: jsx(Label, { id: id, children: children }) }));

function useLabelledLinkAria({ 'aria-describedby': ariaDescribedBy, baseId, hasEmptyLabel, hasHiddenLabel, labelId: propLabelId }) {
    const containerId = `${baseId}--container`;
    const labelId = propLabelId ?? `${baseId}--hidden-label`;
    // JET-52089: add unsafe API for specifying an external label which is added to ariaLabelledBy on the input or div,
    // only when labelEdge is 'none' and label is ''. Since ariaLabelledBy takes precedence over all other
    // kinds of labels, this helps to ensure we don't override a meaningful label.
    const { UNSAFE_ariaLabelledBy } = useAccessibleContext();
    const ariaLabelledBy = classNames([
        labelId,
        hasEmptyLabel && hasHiddenLabel && UNSAFE_ariaLabelledBy,
        containerId
    ]);
    return {
        containerAria: {
            id: containerId
        },
        hiddenLabelAria: hasHiddenLabel
            ? {
                id: labelId
            }
            : {},
        linkAria: {
            'aria-describedby': ariaDescribedBy,
            'aria-labelledby': ariaLabelledBy
        }
    };
}

function useLabelledLinkTheme({ hasInsideLabel, isFormLayout, isFormReadonly, textAlign }) {
    // Style interpolations for text align
    const interpolations = [...Object.values(textInterpolations)];
    const styleInterpolations = mergeInterpolations(interpolations);
    // Calculate the style interpolation classes
    const { class: styleInterpolationClasses } = styleInterpolations({ textAlign });
    // Calculate the theme classes
    const { classes: themeClasses } = useComponentTheme(LabelledLinkRedwoodTheme, {
        formLayout: isFormLayout ? 'isFormLayout' : 'notFormLayout',
        readonlyForm: isFormReadonly ? 'isReadonlyForm' : 'notReadonlyForm',
        insideLabel: hasInsideLabel ? 'hasInsideLabel' : 'noInsideLabel'
    });
    // merge all the classes
    const classes = classNames([styleInterpolationClasses, themeClasses]);
    return { classes };
}

/**
 * LabelledLink component can be used to render a readonly form control field that
 * has a link for its content.
 */
const LabelledLink = ({ 'aria-describedby': ariaDescribedBy, children, columnSpan, href, label, labelEdge: propLabelEdge, labelStartWidth: propLabelStartWidth, lineClamp, onClick: propOnClick, target, textAlign: propTextAlign, userAssistanceDensity: propUserAssistanceDensity }) => {
    // Consume the form context to get the default values
    const { isFormLayout, isReadonly: isFormReadonly, labelEdge: formLabelEdge, labelStartWidth: formLabelStartWidth, textAlign: formTextAlign, userAssistanceDensity: formUserAssistanceDensity } = useFormContext();
    // default to FormContext values if component properties are not specified
    const labelEdge = propLabelEdge ?? formLabelEdge;
    const labelStartWidth = propLabelStartWidth ?? formLabelStartWidth;
    // JET-72775 - Form control alignment in tables
    // We do not have an embedded mode, so always default to 'start' if no text align is set
    const textAlign = propTextAlign ?? formTextAlign ?? 'start';
    const userAssistanceDensity = propUserAssistanceDensity ?? formUserAssistanceDensity;
    // We will use the URL as text if the link text is not provided as a child
    const linkText = children ?? href;
    // Calculate the text field related props
    const { baseId, formFieldContext: textFieldFormFieldContext, labelProps, inputProps, userAssistanceProps } = useTextField({
        ariaDescribedBy,
        isReadonly: true,
        labelEdge,
        value: linkText
    });
    // Track pointer down so we know if we gained focus from keyboard or mouse.
    const recentPointerRef = useRef(false);
    const pointerDownTimerRef = useRef();
    const onPointerDown = () => {
        recentPointerRef.current = true;
        clearTimeout(pointerDownTimerRef.current);
        pointerDownTimerRef.current = setTimeout(() => {
            recentPointerRef.current = false;
        }, 300);
    };
    // Truncated tooltips act differently for LabelledLink, so handle it here. We don't want to
    // show the tooltip when the link is clicked (thereby focusing the field), and if the user
    // tabbed onto the link and activated it with the keyboard, we want to close the tooltip.
    const labelRef = labelProps['labelRef'];
    const linkWasActivatedRef = useRef(false);
    const [showTruncatedTooltip, setShowTruncatedTooltip] = useState(false);
    const onBlurWithin = useCallback(() => {
        setShowTruncatedTooltip(false);
    }, []);
    const onFocusWithin = useCallback(() => {
        // Ignore the focus event the first time we see it after the link was activated.
        // For instance if the link was clicked and opened a new browser tab, we don't want
        // to show the tooltip again when they return to the current tab.
        if (linkWasActivatedRef.current) {
            linkWasActivatedRef.current = false;
            return;
        }
        // We only calculate isTruncated if focus was gained through keyboard interaction.
        const isTruncated = labelEdge === 'inside' && !recentPointerRef.current ? detectTruncation(labelRef) : false;
        setShowTruncatedTooltip(isTruncated);
    }, [labelEdge, labelRef]);
    const { focusProps: { onFocusIn: onFocus, onFocusOut: onBlur } } = useFocusWithin({
        isDisabled: false,
        onBlurWithin,
        onFocusWithin
    });
    const formFieldContext = useFormFieldContextProps({
        ...textFieldFormFieldContext,
        hasTruncatedLabel: showTruncatedTooltip
    });
    const onClick = useCallback((event) => {
        setShowTruncatedTooltip(false);
        linkWasActivatedRef.current = true;
        propOnClick?.(event);
    }, [propOnClick]);
    // Create the label component and the fieldLabelProps
    const labelComp = jsx(Label, { ...labelProps, children: label });
    const fieldLabelProps = {
        label: labelEdge !== 'none' ? labelComp : undefined,
        labelEdge: labelEdge !== 'none' ? labelEdge : undefined,
        labelStartWidth: labelEdge !== 'none' ? labelStartWidth : undefined,
        labelText: labelEdge === 'inside' ? label : undefined
    };
    // save space for user assistance if density is 'efficient', even though we don't
    // render user assistance for readonly fields for now.
    const inlineUserAssistance = userAssistanceDensity === 'efficient' ? (jsx(InlineUserAssistance, { userAssistanceDensity: userAssistanceDensity, ...userAssistanceProps })) : undefined;
    // Get the theme classes
    const { classes } = useLabelledLinkTheme({
        hasInsideLabel: labelEdge === 'inside',
        isFormLayout,
        isFormReadonly,
        textAlign
    });
    // Get the aria props
    const hasEmptyLabel = label === '';
    const hasHiddenLabel = labelEdge === 'none';
    const { containerAria, hiddenLabelAria, linkAria } = useLabelledLinkAria({
        'aria-describedby': inputProps['aria-describedby'],
        baseId,
        hasEmptyLabel,
        hasHiddenLabel,
        labelId: labelProps.id
    });
    return (jsx(FormFieldContext.Provider, { value: formFieldContext, children: jsx(ReadonlyTextField, { columnSpan: columnSpan, role: "presentation", inlineUserAssistance: inlineUserAssistance, ...fieldLabelProps, mainContent: jsxs("div", { onBlur: onBlur, onFocus: onFocus, onPointerDown: onPointerDown, class: classes, ...containerAria, children: [hasHiddenLabel && jsx(HiddenLabel, { ...hiddenLabelAria, children: label }), jsx(View, { width: "100%", children: lineClamp ? (jsx(TruncatingLink, { lineClamp: lineClamp, href: href, target: target, onClick: onClick, ...linkAria, children: linkText })) : (jsx(Link, { href: href, target: target, onClick: onClick, ...linkAria, children: linkText })) })] }) }) }));
};

export { LabelledLink };
