// @oracle/oraclejet-preact/es/UNSAFE_InputColor.js
/* @oracle/oraclejet-preact: undefined */
import { jsx, jsxs, Fragment } from 'preact/jsx-runtime';
import { forwardRef } from 'preact/compat';
import { useState, useCallback, useReducer, useRef, useEffect, useImperativeHandle } from 'preact/hooks';
import { a as colorsAreDifferent, b as isValidColorValue } from './swatchData-64959ad2.js';
import { g as getClientHints } from './clientHints-a42b510c.js';
import { F as FormFieldContext } from './FormFieldContext-47ca0c20.js';
import './logger-c92f309c.js';
import './TabbableModeContext-7d8ad946.js';
import { u as useTranslationBundle } from './useTranslationBundle-4822502f.js';
import './UserAssistanceStyles.styles.css';
import './UNSAFE_UserAssistance/themes/redwood/UserAssistanceBaseTheme.styles.css';
import './UNSAFE_UserAssistance/themes/redwood/UserAssistanceVariants.css.js';
import { a as InlineUserAssistance } from './InlineUserAssistance-890293dd.js';
import './Flex-03129963.js';
import './MessageBannerStyles.styles.css';
import './UNSAFE_MessageBanner/themes/redwood/MessageBannerBaseTheme.styles.css';
import './UNSAFE_MessageBanner/themes/redwood/MessageBannerVariants.css.js';
import 'preact';
import './LayerHost-7931b424.js';
import './TooltipContentStyles.styles.css';
import './hooks/UNSAFE_useTooltip/themes/redwood/TooltipContentBaseTheme.styles.css';
import './hooks/UNSAFE_useTooltip/themes/redwood/TooltipContentVariants.css.js';
import './tooltipUtils-af0abd38.js';
import './IconStyle.styles.css';
import './IconButton-7f91481c.js';
import './MessageStyles.styles.css';
import './MessageFormattingUtils-cdb25115.js';
import './TransitionGroup-5fd80dc9.js';
import './MessagesContext-76544845.js';
import './HiddenAccessibleStyles.styles.css';
import './ComponentMessageStyles.styles.css';
import './UNSAFE_ComponentMessage/themes/redwood/ComponentMessageBaseTheme.styles.css';
import './UNSAFE_ComponentMessage/themes/redwood/ComponentMessageVariants.css.js';
import './Popup-992b6c27.js';
import './UNSAFE_Separator/themes/SeparatorStyles.css.js';
import './LabelStyles.styles.css';
import './UNSAFE_Label/themes/redwood/LabelBaseTheme.styles.css';
import './UNSAFE_Label/themes/redwood/LabelVariants.css.js';
import './InputGroupContext-20bdbd6e.js';
import './UNSAFE_Popup/themes/redwood/PopupBaseTheme.styles.css';
import './UNSAFE_Popup/themes/redwood/PopupVariants.css.js';
import { userAssistanceStyles, dropdownStyles } from './UNSAFE_InputColor/themes/InputColorPickerDropdownStyles.css.js';
import { S as Sheet } from './Sheet-f9f8a4a4.js';
import { C as ColorPicker } from './ColorPicker-ae648c4c.js';
import { a as SectionedContent, S as StickyPositioningDropdown } from './SectionedContent-30e5bf98.js';
import { C as ColorSwatch } from './ColorSwatch-8339ee2f.js';
import { u as useDownToggle } from './useDownToggle-10a7ad88.js';
import { u as useClearIcon } from './useClearIcon-b50e406d.js';
import { u as useFocusableTextField } from './useFocusableTextField-8fe41cd4.js';
import './FormContext-55fb8000.js';
import { u as useFormContext } from './useFormContext-bb9dc478.js';
import { u as useHover } from './useHover-85077adb.js';
import { u as useTextField } from './useTextField-eb9db737.js';
import { L as Label } from './Label-5d1ae8d7.js';
import './keyboardUtils-986fe923.js';
import { T as TextFieldInput, R as ReadonlyTextField } from './StyledTextField-7605d76a.js';
import './ObfuscatedTextFieldInputStyles.styles.css';
import { R as ReadonlyTextFieldInput } from './ReadonlyTextFieldInput-106ce959.js';
import { T as TextField } from './TextField-7884fcd7.js';
import './LiveRegionStyles.styles.css';
import './UNSAFE_TextField/themes/TextFieldStyles.css.js';
import './SkeletonStyles.styles.css';
import './TextFieldLoadingStyles.styles.css';
import './TextFieldStyles.styles.css';
import './UNSAFE_TextField/themes/redwood/TextFieldBaseTheme.styles.css';
import './UNSAFE_TextField/themes/redwood/TextFieldVariants.css.js';
import { C as CompactUserAssistance } from './CompactUserAssistance-6aa8894b.js';
import { m as mergeProps } from './mergeProps-88ea8306.js';
import { C as ClearIcon } from './ClearIcon-f85ea340.js';
import { u as useCurrentValueReducer } from './useCurrentValueReducer-bdf796fa.js';
import { u as useSelectableTextField } from './useSelectableTextField-949184a5.js';
import './colorUtils-b3804ee8.js';
import './_curry1-63949a9b.js';
import './vanilla-extract-recipes-createRuntimeFn.esm-2aaf8c98.js';
import './ComponentMessageContainer-d16c8b1f.js';
import './useTestId-f24ac480.js';
import './MessagesManager-47000438.js';
import './classNames-4e12b00d.js';
import './SuccessS-b5192d40.js';
import './Icon-4983d06e.js';
import './size-af1e638c.js';
import './utils-b1eb54f9.js';
import './themeContract.css-1ffed3bf.js';
import './useTooltip-94af94d4.js';
import './useTooltipControlled-f2cfa454.js';
import './useId-03dbfdf0.js';
import './useFocus-38c95977.js';
import './useToggle-8b7fcefe.js';
import './useTouch-4828df25.js';
import './useAnimation-fe39685e.js';
import './useComponentTheme-d2f9e47f.js';
import './useThemeInterpolations-e144e8ec.js';
import './useColorScheme-c4f6caab.js';
import './useScale-434ad939.js';
import './theme-63551f30.js';
import './Theme-e6dec6db.js';
import './mergeInterpolations-30cd9a69.js';
import './mergeDeepWithKey-33e3d1be.js';
import './_curry3-577eb45e.js';
import './_curry2-01c87f61.js';
import './_isObject-1fab0f5b.js';
import './useDensity-1acb0153.js';
import './unsafeDomAccess-453eb74e.js';
import './hooks/UNSAFE_useTooltip/themes/redwood/TooltipContentTheme.js';
import './hooks/UNSAFE_useTooltip/themes/TooltipContentStyles.css.js';
import './hooks/UNSAFE_useTooltip/themes/redwood/TooltipContentBaseTheme.css.js';
import './TooltipContentVariants.styles.css';
import './UNSAFE_Icon/themes/IconStyle.css.js';
import './ErrorS-3a890bf6.js';
import './InformationS-d53c2a91.js';
import './WarningS-3e53d87e.js';
import './PRIVATE_Message/themes/MessageStyles.css.js';
import './soundUtils-88264cb5.js';
import './useMessagesContext-9e1dbe91.js';
import './Transition-f9501682.js';
import './HiddenAccessible-61197e7c.js';
import './UNSAFE_ComponentMessage/themes/redwood/ComponentMessageTheme.js';
import './UNSAFE_ComponentMessage/themes/ComponentMessageStyles.css.js';
import './UNSAFE_ComponentMessage/themes/redwood/ComponentMessageBaseTheme.css.js';
import './UNSAFE_UserAssistance/themes/redwood/UserAssistanceTheme.js';
import './UNSAFE_UserAssistance/themes/UserAssistanceStyles.css.js';
import './UNSAFE_UserAssistance/themes/redwood/UserAssistanceBaseTheme.css.js';
import './useFormFieldContext-daf7a676.js';
import './InlineHelp-9eaac91c.js';
import './useTabbableMode-090ed4b5.js';
import './useInputGroupContext-1d19411e.js';
import './boxalignment-8980856d.js';
import './arrayUtils-35a58161.js';
import './boxalignment.styles.css';
import './vanilla-extract-sprinkles-createRuntimeSprinkles.esm-2d655d37.js';
import './dimensions-e4fbc23d.js';
import './flexbox-529f25da.js';
import './flexbox.styles.css';
import './flexitem-7b7f7920.js';
import './flexitem.styles.css';
import './FlexStyles.styles.css';
import './MessageBannerVariants.styles.css';
import './index-aec1cc7f.js';
import './UNSAFE_Layer/themes/LayerHostStyles.css.js';
import './LayerHostStyles.styles.css';
import './EnvironmentProvider-6a0a8212.js';
import './LayerManager-52a26c55.js';
import './Floating-4a82db94.js';
import './useFloating-b9d680f7.js';
import './useUser-06d34d19.js';
import './positionUtils-f26caad9.js';
import './refUtils-76275c69.js';
import './useOutsideClick-b777a43e.js';
import './Layer-535577b5.js';
import './UNSAFE_Layer/themes/LayerStyles.css.js';
import './LayerStyles.styles.css';
import './useModal-953082f2.js';
import './UNSAFE_Floating/themes/redwood/FloatingTheme.js';
import './UNSAFE_Floating/themes/FloatingStyles.css.js';
import './FloatingStyles.styles.css';
import './UNSAFE_Floating/themes/redwood/FloatingBaseTheme.css.js';
import './UNSAFE_Floating/themes/redwood/FloatingBaseTheme.styles.css';
import './UNSAFE_Floating/themes/redwood/FloatingVariants.css.js';
import './vanilla-extract-dynamic.esm-c909758c.js';
import './UNSAFE_Floating/themes/FloatingContract.css.js';
import './useUnsafeDomElementRef-32290273.js';
import './BaseButton-41068c85.js';
import './useActive-fb6d4d52.js';
import './UNSAFE_BaseButton/themes/redwood/BaseButtonTheme.js';
import './UNSAFE_BaseButton/themes/BaseButtonStyles.css.js';
import './BaseButtonStyles.styles.css';
import './UNSAFE_BaseButton/themes/redwood/BaseButtonVariants.css.js';
import './UNSAFE_BaseButton/themes/redwood/BaseButtonBaseTheme.styles.css';
import './BareButton-1e3df7cb.js';
import './useBareButton-ff4bd21e.js';
import './usePress-051f5adb.js';
import './usePressClick-945cc43b.js';
import './buttonUtils-55fb1df7.js';
import './id-83adac50.js';
import './ButtonLabelLayout-1631c997.js';
import './Text-834eb35e.js';
import './UNSAFE_Text/themes/TextStyles.css.js';
import './TextStyles.styles.css';
import './useButtonLabelLayoutTheme-448e671d.js';
import './UNSAFE_ButtonLabelLayout/themes/redwood/ButtonLabelLayoutTheme.js';
import './UNSAFE_ButtonLabelLayout/themes/ButtonLabelLayoutStyles.css.js';
import './ButtonLabelLayoutStyles.styles.css';
import './UNSAFE_ButtonLabelLayout/themes/redwood/ButtonLabelLayoutBaseTheme.css.js';
import './UNSAFE_ButtonLabelLayout/themes/redwood/ButtonLabelLayoutBaseTheme.styles.css';
import './UNSAFE_ButtonLabelLayout/themes/redwood/ButtonLabelLayoutVariants.css.js';
import './ButtonLabelLayoutVariants.styles.css';
import './getLocale-1c41e86a.js';
import './stringUtils-0fe25200.js';
import './tabbableUtils-c8311ec6.js';
import './head-8f89fe66.js';
import './_arity-be492b9e.js';
import './_isArray-694cc52d.js';
import './_isString-675f1de9.js';
import './FocusTrap-b59d5752.js';
import './FocusTracker-70300f1c.js';
import './PRIVATE_FocusTracker/themes/FocusTrackerStyles.css.js';
import './FocusTrackerStyles.styles.css';
import './Modal-c0f730b8.js';
import './UNSAFE_Modal/themes/ModalStyles.css.js';
import './ModalStyles.styles.css';
import './popupUtils-51366a29.js';
import './WindowOverlay-7264c44a.js';
import './UNSAFE_WindowOverlay/themes/WindowOverlayStyles.css.js';
import './WindowOverlayStyles.styles.css';
import './UNSAFE_WindowOverlay/themes/WindowOverlayContract.css.js';
import './UNSAFE_WindowOverlay/themes/redwood/WindowOverlayTheme.js';
import './UNSAFE_WindowOverlay/themes/redwood/WindowOverlayBaseTheme.css.js';
import './UNSAFE_WindowOverlay/themes/redwood/WindowOverlayBaseTheme.styles.css';
import './UNSAFE_WindowOverlay/themes/redwood/WindowOverlayVariants.css.js';
import './usePopupAnimation-96d2bc83.js';
import './useAnimationStatus-b881f39a.js';
import './popupAnimationUtils-175151d9.js';
import './LiveRegion-cebac18d.js';
import './UNSAFE_Popup/themes/PopupContract.css.js';
import './UNSAFE_Popup/themes/redwood/PopupTheme.js';
import './UNSAFE_Popup/themes/redwood/PopupBaseTheme.css.js';
import './PopupStyles.styles.css';
import './colors-03807160.js';
import './SeparatorStyles.styles.css';
import './InputColorPickerDropdownStyles.styles.css';
import './UNSAFE_Sheet/themes/SheetStyles.css.js';
import './SheetStyles.styles.css';
import './animationUtils-5f409a6c.js';
import './UNSAFE_ColorPicker/themes/ColorPickerContract.css.js';
import './UNSAFE_ColorPicker/themes/redwood/ColorPickerTheme.js';
import './UNSAFE_ColorPicker/themes/redwood/ColorPickerVariants.css.js';
import './ColorPickerVariants.styles.css';
import './UNSAFE_ColorPicker/themes/redwood/ColorPickerBaseTheme.css.js';
import './UNSAFE_ColorPicker/themes/redwood/ColorPickerBaseTheme.styles.css';
import './clientUtils-71f2dbaf.js';
import './ColorPalette-950c6fc6.js';
import './padding-90265467.js';
import './aria-efa0582c.js';
import './resources/nls/en/bundle.js';
import './SwatchFlexView-7777d7df.js';
import './useResizeObserver-13641643.js';
import './useAddBusyState-7a6af61f.js';
import './BusyStateContext-9ee9751d.js';
import './useBusyStateContext-4d3c27c0.js';
import './BaseSwatchView-2e1e6bd2.js';
import './keys-49c92dbb.js';
import './collectionUtils-5313629f.js';
import './Selector-23bccbe7.js';
import './BaseSelector-7cbc48bc.js';
import './StyledCheckbox-d6c8f557.js';
import './Check-ac39fe00.js';
import './CheckboxOff-3359a12d.js';
import './CheckboxOn-930b64de.js';
import './CheckboxMixed-c4594250.js';
import './UNSAFE_Checkbox/themes/CheckboxIconContract.css.js';
import './UNSAFE_Checkbox/themes/redwood/CheckboxIconTheme.js';
import './UNSAFE_Checkbox/themes/CheckboxIconStyles.css.js';
import './CheckboxIconStyles.styles.css';
import './UNSAFE_Checkbox/themes/redwood/CheckboxIconBaseTheme.css.js';
import './UNSAFE_Checkbox/themes/redwood/CheckboxIconBaseTheme.styles.css';
import './UNSAFE_Checkbox/themes/redwood/CheckboxIconVariants.css.js';
import './CheckboxIconVariants.styles.css';
import './StyledCheckbox.styles.css';
import './PRIVATE_BaseSelector/themes/SelectorStyles.css.js';
import './SelectorStyles.styles.css';
import './CollectionInteractionContext-7cf19abc.js';
import './useCollectionInteractionContext-3895c632.js';
import './useCurrentKey-5f865bbd.js';
import './useCollectionFocusRing-23ac7f97.js';
import './useTabbableModeSet-159b0279.js';
import './useSwatchSelection-1757379d.js';
import './textSelectionUtils-e87a47ff.js';
import './LoadMoreCollection-d9f22790.js';
import './Collection-e1ef8b8d.js';
import './useViewportIntersect-6ff65310.js';
import './PRIVATE_BaseSwatchView/themes/BaseSwatchViewStyles.css.js';
import './BaseSwatchViewStyles.styles.css';
import './Dropdown-019aba37.js';
import './UNSAFE_Dropdown/themes/dropdownStyles.css.js';
import './dropdownStyles.styles.css';
import './useOutsideMousedown-c45c22c0.js';
import './useEffectEvent-1bb008cb.js';
import './View-5bdd9b3c.js';
import './borders-98a63040.js';
import './equals-625352e8.js';
import './useToggleAction-fc0f5399.js';
import './UNSAFE_ColorSwatch/themes/redwood/ColorSwatchTheme.js';
import './UNSAFE_ColorSwatch/themes/ColorSwatchStyles.css.js';
import './ColorSwatchStyles.styles.css';
import './UNSAFE_ColorSwatch/themes/redwood/ColorSwatchBaseTheme.css.js';
import './UNSAFE_ColorSwatch/themes/redwood/ColorSwatchBaseTheme.styles.css';
import './UNSAFE_ColorSwatch/themes/redwood/ColorSwatchVariants.css.js';
import './CheckboxCrossed-7794974a.js';
import './useFocusWithin-30b1e2d8.js';
import './truncationUtils-e37f95ee.js';
import './useMessageSeverity-65295e8f.js';
import './useFormFieldContextProps-31bbc12f.js';
import './CompactLabelAssistance-e2fcfcab.js';
import './Help-32a9689a.js';
import './UNSAFE_Label/themes/redwood/LabelTheme.js';
import './UNSAFE_Label/themes/LabelStyles.css.js';
import './UNSAFE_Label/themes/redwood/LabelBaseTheme.css.js';
import './UNSAFE_TextField/themes/redwood/TextFieldTheme.js';
import './UNSAFE_TextField/themes/redwood/TextFieldBaseTheme.css.js';
import './TextFieldVariants.styles.css';
import './useAccessibleContext-5744de8b.js';
import './useTextFieldInputHandlers-8122048f.js';
import './formControlUtils-84e2af0e.js';
import './textAlign-ad252afa.js';
import './text.styles.css';
import './TextFieldUtils-8232bca7.js';
import './UNSAFE_TextField/themes/FormControlUtilsStyles.css.js';
import './FormControlUtilsStyles.styles.css';
import './UNSAFE_TextField/themes/redwood/TextFieldInputTheme.js';
import './UNSAFE_TextField/themes/TextFieldInputStyles.css.js';
import './TextFieldInputStyles.styles.css';
import './UNSAFE_TextField/themes/redwood/TextFieldInputVariants.css.js';
import './TextFieldInputVariants.styles.css';
import './LabelValueLayout-c6b9945a.js';
import './StyledLabelValueLayout-bb05f474.js';
import './LabelValueLayoutStyles.styles.css';
import './UNSAFE_TextField/themes/redwood/FormLayoutTheme.js';
import './UNSAFE_TextField/themes/FormLayoutStyles.css.js';
import './UNSAFE_TextField/themes/redwood/FormLayoutBaseTheme.css.js';
import './UNSAFE_TextField/themes/redwood/FormLayoutBaseTheme.styles.css';
import './formControlUtils-6fc3a497.js';
import './LayoutStyles.css-6b02421a.js';
import './LayoutStyles.styles.css';
import './UNSAFE_Skeleton/themes/redwood/SkeletonTheme.js';
import './UNSAFE_Skeleton/themes/SkeletonStyles.css.js';
import './UNSAFE_Skeleton/themes/redwood/SkeletonBaseTheme.css.js';
import './UNSAFE_Skeleton/themes/redwood/SkeletonBaseTheme.styles.css';
import './UNSAFE_Skeleton/themes/redwood/SkeletonVariants.css.js';
import './UNSAFE_TextField/themes/redwood/ReadonlyTextFieldTheme.js';
import './UNSAFE_TextField/themes/ReadonlyTextFieldStyles.css.js';
import './ReadonlyTextFieldStyles.styles.css';
import './UNSAFE_TextField/themes/redwood/ReadonlyTextFieldVariants.css.js';
import './ReadonlyTextFieldVariants.styles.css';
import './UNSAFE_TextField/themes/redwood/ReadonlyTextFieldInputTheme.js';
import './UNSAFE_TextField/themes/ReadonlyTextFieldInputStyles.css.js';
import './ReadonlyTextFieldInputStyles.styles.css';
import './UNSAFE_TextField/themes/redwood/ReadonlyTextFieldInputVariants.css.js';
import './Separator-93c70737.js';
import './DeleteCircleS-f45c1c20.js';

/**
 * Component that renders ColorPicker either inside a dropdown or sheet; in mobile,
 * it renders inside a sheet. Additionally, when a dropdown is being shown beneath
 * the component, we also show any user assistance text or messages below the picker
 * because otherwise that information would be obscured from view. When a sheet is being
 * shown, it will always display UA and messages at the bottom of the sheet.
 */
const InputColorPickerDropdown = ({ anchorRef, assistiveText, swatchSize, palette, helpSourceLink, helpSourceText, isDisabled, isOpen, isReadonly, label, onClose, onCommit, messages, testId, userAssistanceDensity, value }) => {
    // If the dropdown opens above the component, we don't want to show UA or messages.
    const [isDropdownAbove, setDropdownAbove] = useState(false);
    const [prevIsOpen, setPrevIsOpen] = useState(isOpen);
    const handleDropdownPosition = useCallback((data) => {
        setDropdownAbove(data.placement.startsWith('top'));
    }, []);
    // Reset the flag when the dropdown is closed.
    if (prevIsOpen !== isOpen) {
        setPrevIsOpen(isOpen);
        if (!isOpen) {
            setDropdownAbove(false);
        }
    }
    const sizeOptions = {
        isMaxWidthAdjusted: false,
        isMaxHeightAdjusted: true
    };
    // On mobile phones, the dropdown becomes a sheet.
    const shouldShowSheet = isPhone();
    const { colorPicker_label } = useTranslationBundle('@oracle/oraclejet-preact');
    const colorPickerAriaLabel = colorPicker_label();
    const colorPickerMainContent = (jsx(ColorPicker, { height: !shouldShowSheet ? 'auto' : undefined, onCommit: onCommit, 
        // for sheet, we want the width to grow to the width of the sheet, otherwise it will use the theme default.
        maxWidth: shouldShowSheet ? '100%' : undefined, width: "100%", swatchSize: swatchSize, palette: palette, testId: testId, value: value }));
    const inlineUserAssistanceFooter = isOpen &&
        (!isDropdownAbove || shouldShowSheet) &&
        !isDisabled &&
        !isReadonly &&
        (assistiveText || helpSourceLink || (messages && messages.length > 0)) &&
        (userAssistanceDensity === 'efficient' || userAssistanceDensity === 'reflow') ? (jsx(FormFieldContext.Provider, { value: { isFocused: true }, children: jsx("div", { class: userAssistanceStyles, children: jsx(InlineUserAssistance, { assistiveText: assistiveText, fieldLabel: label, helpSourceLink: helpSourceLink, helpSourceText: helpSourceText, messages: messages, userAssistanceDensity: userAssistanceDensity }) }) })) : undefined;
    const contentTestId = testId ? testId + (shouldShowSheet ? '_sheet' : '_dropdown') : undefined;
    // The same content is displayed either in a dropdown or sheet: color picker and inline UA.
    // Add styling to the dropdown to enforce a fixed width.
    const dropdownContent = (jsx("div", { class: dropdownStyles, children: jsx(SectionedContent, { paddingBlockEnd: "3.5x", paddingBlockStart: "3.5x", paddingInlineEnd: "3.5x", paddingInlineStart: "3.5x", "aria-label": colorPickerAriaLabel, footer: inlineUserAssistanceFooter, hasFocusTrap: true, main: colorPickerMainContent, role: "dialog", testId: contentTestId }) }));
    const dropdown = shouldShowSheet ? (jsx(Sheet, { isOpen: isOpen, onClose: onClose, children: jsx(SectionedContent, { "aria-label": colorPickerAriaLabel, footer: inlineUserAssistanceFooter, hasFocusTrap: true, main: colorPickerMainContent, role: "dialog", testId: contentTestId }) })) : (jsx(StickyPositioningDropdown, { anchorRef: anchorRef, isOpen: isOpen, onClose: onClose, onPosition: handleDropdownPosition, sizeOptions: sizeOptions, children: dropdownContent }));
    return dropdown;
};
const isPhone = () => {
    return getClientHints().deviceType === 'phone';
};

/**
 * @license
 * Copyright (c) %FIRST_YEAR% %CURRENT_YEAR%, Oracle and/or its affiliates.
 * The Universal Permissive License (UPL), Version 1.0
 * as shown at https://oss.oracle.com/licenses/upl/
 * @ignore
 */
// Use a reducer function to manage InputColorState.
const inputColorReducer = (state, action) => {
    switch (action.type) {
        case 'colorChanged': {
            return {
                ...state,
                colorValue: action.data
            };
        }
        case 'colorSelected': {
            return { ...state, colorValue: action.data };
        }
        case 'reset': {
            return {
                colorValue: action.data
            };
        }
        default:
            return state;
    }
};
/**
 * useInputColorState calls useReducer to determine the initial state. Any changes to that state
 * are managed by dispatching actions to the reducer function.
 */
const useInputColorState = ({ value }) => {
    const [state, dispatch] = useReducer(inputColorReducer, {
        colorValue: value
    });
    return { state, dispatch };
};

/**
 * @license
 * Copyright (c) %FIRST_YEAR% %CURRENT_YEAR%, Oracle and/or its affiliates.
 * The Universal Permissive License (UPL), Version 1.0
 * as shown at https://oss.oracle.com/licenses/upl/
 * @ignore
 */
/**
 * Hook that provides handlers and other functionality for InputColor.
 */
const useInputColor = ({ onCommit, onInput, value: currentValue }) => {
    // Whether the dropdown is open.
    const [isOpen, setIsOpen] = useState(false);
    // Ref to the swatch icon button.
    const iconRef = useRef(null);
    // Ref to the text input which holds color value
    const colorRef = useRef(null);
    // Keep track of whether the dropdown was just dismissed from an outside mousedown event.
    const outsideMousedownRef = useRef(false);
    const outsideMousedownTimerRef = useRef();
    // Keep track of the last color values passed to onInput and onCommit.
    const lastCommittedColorRef = useRef(currentValue);
    const lastColorRef = useRef(currentValue);
    // Keep track of the last controlled value passed from the parent, so we can detect a programmatic change.
    const lastValueFromParentRef = useRef(currentValue);
    // Keep track of whether a color was selected in the color picker.
    const colorSelectedRef = useRef(false);
    const { state, dispatch } = useInputColorState({ value: currentValue });
    // If a new value is being pushed from the parent, and that value differs from
    // the current value of the reducer, then this is a programmatic update
    // and we need to call 'reset' to update the reducer state. Otherwise, the only
    // time the reducer state is updated from the value is on mount.
    if (colorsAreDifferent(lastValueFromParentRef.current, currentValue)) {
        lastValueFromParentRef.current = currentValue;
        if (colorsAreDifferent(currentValue, state.colorValue)) {
            dispatch({ type: 'reset', data: currentValue });
            lastColorRef.current = currentValue;
            lastCommittedColorRef.current = currentValue;
        }
    }
    // Reducer updates are async, so wait until the state updates and then call our callbacks.
    useEffect(() => {
        if (colorsAreDifferent(state.colorValue, lastColorRef.current)) {
            onInput({ previousValue: lastColorRef.current, value: state.colorValue });
            lastColorRef.current = state.colorValue;
        }
        if (colorSelectedRef.current &&
            colorsAreDifferent(state.colorValue, lastCommittedColorRef.current)) {
            onCommit?.({ previousValue: lastCommittedColorRef.current, value: state.colorValue });
            lastCommittedColorRef.current = state.colorValue;
            colorSelectedRef.current = false;
        }
    }, [onCommit, onInput, state.colorValue]);
    // Handle onInput from InputColor as values are entered in the input field.
    // We wait for the state to update before calling onInput.
    const colorInputHandler = useCallback((detail) => {
        dispatch({ type: 'colorChanged', data: detail.value });
    }, [dispatch]);
    // Handle onCommit from InputColor, which is called when the user presses Enter or the field is blurred.
    const colorCommitHandler = useCallback(() => {
        // If the current color value differs from the last committed value, then call onCommit.
        if (colorsAreDifferent(lastCommittedColorRef.current, state.colorValue)) {
            onCommit?.({ value: state.colorValue, previousValue: lastCommittedColorRef.current });
            lastCommittedColorRef.current = state.colorValue;
        }
    }, [onCommit, state.colorValue]);
    // Respond to onAction event from the swatch icon button. Toggles the color picker dropdown.
    const colorIconActionHandler = useCallback(() => {
        // Only toggle the value of isOpen if we didn't just dismiss the dropdown.
        if (!outsideMousedownRef.current) {
            setIsOpen((prevIsOpen) => !prevIsOpen);
        }
        outsideMousedownRef.current = false;
    }, []);
    const dropdownCloseHandler = useCallback((detail) => {
        // Keep track of when the dropdown is closed via outside mousedown, including when the swatch icon
        // is clicked. We need this in order to toggle visibility of the picker and dropdown.
        if (detail.reason === 'outsideMousedown') {
            outsideMousedownRef.current = true;
            clearTimeout(outsideMousedownTimerRef.current);
            // Clear the flag after a short delay to ensure the dropdown can be opened again.
            outsideMousedownTimerRef.current = setTimeout(() => {
                outsideMousedownRef.current = false;
            }, 200);
        }
        else if (detail.reason === 'dismissed' || detail.reason === 'keyboardDismissed') {
            // Focus the swatch icon when the dropdown is closed.
            iconRef.current?.focus();
        }
        setIsOpen(false);
    }, []);
    // Handle color selection in the picker.
    //   - Updates the value in state.
    //   - Closes the dropdown.
    //   - Sets focus to the color field.
    const colorPickerCommitHandler = useCallback((detail) => {
        // Set focus to the swatch when a color is selected.
        iconRef.current?.focus();
        // Set a ref indicating a color was selected.
        colorSelectedRef.current = true;
        dispatch({ type: 'colorSelected', data: detail.value });
        setIsOpen(false);
    }, [dispatch]);
    // The Redwood UX design calls for the dropdown to open when the down arrow is pressed.
    // Handle opening the dropdown when the down arrow is pressed in the text edit field.
    const colorPickerDownHandler = useCallback(() => {
        setIsOpen(true);
    }, []);
    const { triggerProps: containerDropdownProps } = useDownToggle({
        isOpen,
        onToggle: colorPickerDownHandler
    });
    return {
        colorIconActionHandler,
        colorPickerCommitHandler,
        dropdownCloseHandler,
        containerDropdownProps,
        colorValue: state.colorValue,
        iconRef,
        isOpen,
        colorCommitHandler,
        colorInputHandler,
        colorRef
    };
};

/**
 * An Input Color is a form field that allows users to select or input a color value,
 * using a color picker or by enterihg a HEX code.
 */
const InputColor = forwardRef(({ 'aria-describedby': ariaDescribedBy, swatchSize, palette, value, isDisabled: propIsDisabled, isReadonly: propIsReadonly, labelEdge: propLabelEdge, labelStartWidth: propLabelStartWidth, isRequiredShown = false, isRequired = false, placeholder, hasClearIcon, columnSpan, textAlign: propTextAlign, autoComplete, messages, assistiveText, helpSourceLink, helpSourceText, label, onCommit, onInput, testId, variant, userAssistanceDensity: propUserAssistanceDensity }, ref) => {
    const { colorIconActionHandler, colorPickerCommitHandler, dropdownCloseHandler, colorValue, containerDropdownProps, iconRef, isOpen, colorCommitHandler, colorInputHandler, colorRef: enabledElementRef } = useInputColor({
        onCommit,
        onInput,
        value
    });
    const mainFieldRef = useRef(null);
    const { inputColor_selectColor } = useTranslationBundle('@oracle/oraclejet-preact');
    const colorIconAriaLabel = inputColor_selectColor();
    const { onKeyDown, ...inputContainerAriaProps } = containerDropdownProps;
    const { currentCommitValue, dispatch } = useCurrentValueReducer({ value: colorValue });
    const onInputAndDispatch = useCallback((detail) => {
        // Should dispatch happen first? This will queue up a re-render, ordering should not cause issues (this is async)
        dispatch({ type: 'input', payload: detail.value });
        colorInputHandler?.(detail);
    }, [colorInputHandler, dispatch]);
    const onCommitAndDispatch = useCallback((detail) => {
        // Should dispatch happen first? This will queue up a re-render, ordering should not cause issues (this is async)
        dispatch({ type: 'commit', payload: detail.value });
        colorCommitHandler?.(detail);
    }, [colorCommitHandler, dispatch]);
    const startContent = () => (jsx(ColorSwatch, { size: "xs", "aria-haspopup": palette ? 'grid' : undefined, "aria-label": palette ? colorIconAriaLabel : undefined, isDisabled: isDisabled, isButton: true, isTooltipDisabled: isOpen, ref: iconRef, testId: testId ? testId + '_colorIcon' : undefined, value: currentCommitValue, onToggle: isDisabled || !palette ? undefined : colorIconActionHandler }));
    const { isDisabled: isFormDisabled, isReadonly: isFormReadonly, labelEdge: formLabelEdge, labelStartWidth: formLabelStartWidth, textAlign: formTextAlign, userAssistanceDensity: formUserAssistanceDensity } = useFormContext();
    // default to FormContext values if component properties are not specified
    const isDisabled = propIsDisabled ?? isFormDisabled;
    const isReadonly = propIsReadonly ?? isFormReadonly;
    const labelEdge = propLabelEdge ?? formLabelEdge;
    const labelStartWidth = propLabelStartWidth ?? formLabelStartWidth;
    const textAlign = propTextAlign ?? formTextAlign;
    const userAssistanceDensity = propUserAssistanceDensity ?? formUserAssistanceDensity;
    const readonlyElementRef = useRef(null);
    const { focusProps, isFocused, methods: focusMethods } = useFocusableTextField({
        isDisabled,
        isReadonly,
        enabledElementRef,
        readonlyElementRef
    });
    const { methods: selectMethods } = useSelectableTextField(enabledElementRef);
    useImperativeHandle(ref, () => mergeProps(focusMethods, selectMethods), [
        focusMethods,
        selectMethods
    ]);
    const { hoverProps, isHover } = useHover({ isDisabled: isReadonly || isDisabled || false });
    const { formFieldContext, inputProps, labelProps, textFieldProps, userAssistanceProps } = useTextField({
        ariaDescribedBy,
        helpSourceLink,
        helpSourceText,
        isDisabled,
        isFocused: isFocused || isOpen,
        isReadonly,
        isRequiredShown,
        labelEdge,
        messages,
        styleVariant: variant,
        userAssistanceDensity,
        value: colorValue
    });
    const onClickClearIcon = useCallback(() => {
        // Clicking the clear icon should put the focus on the input field
        enabledElementRef.current?.focus();
        // Send an event to clear the field's value if the value is not already cleared
        const nextValue = undefined;
        if (colorValue === nextValue)
            return;
        onInputAndDispatch?.({ previousValue: colorValue, value: nextValue });
    }, [onInputAndDispatch, colorValue, enabledElementRef]);
    const maybeClearIcon = useClearIcon({
        clearIcon: jsx(ClearIcon, { onClick: onClickClearIcon, testId: `${testId}_clearicon` }),
        display: hasClearIcon,
        hasValue: formFieldContext.hasValue,
        isFocused,
        isEnabled: !isReadonly && !isDisabled,
        isHover
    });
    const hasInsideLabel = label !== undefined && labelEdge === 'inside';
    const labelComp = labelEdge !== 'none' ? jsx(Label, { ...labelProps, children: label }) : undefined;
    const fieldLabelProps = {
        label: labelEdge !== 'none' ? labelComp : undefined,
        labelEdge: labelEdge !== 'none' ? labelEdge : undefined,
        labelStartWidth: labelEdge !== 'none' ? labelStartWidth : undefined
    };
    const ariaLabel = labelEdge === 'none' ? label : undefined;
    const isInlineDensity = userAssistanceDensity === 'efficient' || userAssistanceDensity === 'reflow';
    const inlineUserAssistance = isInlineDensity ? (isDisabled || isReadonly ? (
    // save space for user assistance if density is 'efficient', even though we don't
    // render user assistance for disabled or readonly fields
    userAssistanceDensity !== 'efficient' ? undefined : (jsx(InlineUserAssistance, { userAssistanceDensity: userAssistanceDensity, ...userAssistanceProps }))) : (jsx(InlineUserAssistance, { assistiveText: assistiveText, fieldLabel: label, helpSourceLink: helpSourceLink, helpSourceText: helpSourceText, messages: messages, isRequiredShown: isRequiredShown, userAssistanceDensity: userAssistanceDensity, ...userAssistanceProps }))) : undefined;
    const aRef = useRef(null);
    const anchorRef = mainFieldRef ?? aRef;
    const compactUserAssistance = userAssistanceDensity === 'compact' ? (jsx(CompactUserAssistance, { anchorRef: anchorRef, messages: messages, assistiveText: assistiveText, ...userAssistanceProps })) : undefined;
    const mainContent = isReadonly ? (jsx(ReadonlyTextFieldInput, { "aria-describedby": inputProps['aria-describedby'], "aria-label": ariaLabel, "aria-labelledby": labelProps.id, as: "div", elementRef: readonlyElementRef, textAlign: textAlign, value: colorValue, hasEmptyLabel: label === '' && labelEdge === 'none', hasInsideLabel: hasInsideLabel })) : (jsx(TextFieldInput, { "aria-labelledby": labelProps.id, "aria-label": ariaLabel, autoComplete: autoComplete, currentCommitValue: currentCommitValue, hasEmptyLabel: label === '' && labelEdge === 'none', hasEndContent: false, hasInsideLabel: labelComp !== undefined && labelEdge === 'inside', hasStartContent: true, inputRef: enabledElementRef, isRequired: isRequired, placeholder: placeholder, textAlign: textAlign, type: isMobile() ? 'text' : undefined, value: colorValue, onCommit: onCommitAndDispatch, onInput: onInputAndDispatch, onKeyDown: onKeyDown, ...inputProps }));
    return (jsxs(Fragment, { children: [jsx(FormFieldContext.Provider, { value: formFieldContext, children: isReadonly ? (jsx(ReadonlyTextField, { role: "presentation", columnSpan: columnSpan, compactUserAssistance: compactUserAssistance, inlineUserAssistance: inlineUserAssistance, onBlur: focusProps.onFocusOut, onFocus: focusProps.onFocusIn, ref: anchorRef, mainContent: mainContent, testId: testId, ...fieldLabelProps })) : (jsx(TextField, { endContent: maybeClearIcon, columnSpan: columnSpan, compactUserAssistance: compactUserAssistance, inlineUserAssistance: inlineUserAssistance, mainContent: mainContent, onBlur: focusProps.onFocusOut, onFocus: focusProps.onFocusIn, startContent: startContent(), mainFieldRef: anchorRef, hasZeroStartMargin: true, testId: testId, ...textFieldProps, ...fieldLabelProps, ...hoverProps, ...inputContainerAriaProps })) }), isOpen && !isReadonly && !isDisabled && (jsx(InputColorPickerDropdown, { anchorRef: mainFieldRef, assistiveText: assistiveText, helpSourceLink: helpSourceLink, helpSourceText: helpSourceText, isDisabled: isDisabled, isReadonly: isReadonly, isOpen: isOpen, label: label, onClose: dropdownCloseHandler, onCommit: colorPickerCommitHandler, messages: messages, testId: testId ? testId + '_colorPicker' : undefined, userAssistanceDensity: userAssistanceDensity, 
                // memoize this for performance
                value: colorValue && isValidColorValue(colorValue) ? colorValue : undefined, palette: palette, swatchSize: swatchSize }))] }));
});
/**
 * Helper function to determine whether the current device is a mobile device
 * @returns true if running on a mobile device, false otherwise
 */
function isMobile() {
    const deviceType = getClientHints().deviceType;
    return deviceType === 'phone' || deviceType === 'tablet';
}
InputColor.displayName = 'InputColor';

export { InputColor };
