// @oracle/oraclejet-preact/es/UNSAFE_Switch.js
/* @oracle/oraclejet-preact: undefined */
import { jsx, jsxs } from 'preact/jsx-runtime';
import { forwardRef } from 'preact/compat';
import { useRef, useCallback } from 'preact/hooks';
import { F as FormFieldContext } from './FormFieldContext-47ca0c20.js';
import { c as classNames } from './classNames-4e12b00d.js';
import './FormContext-55fb8000.js';
import { u as useFormContext } from './useFormContext-bb9dc478.js';
import { u as useFormFieldContext } from './useFormFieldContext-daf7a676.js';
import { u as useComponentTheme } from './useComponentTheme-d2f9e47f.js';
import { LabelRedwoodTheme } from './UNSAFE_Label/themes/redwood/LabelTheme.js';
import { FormLayoutRedwoodTheme } from './UNSAFE_TextField/themes/redwood/FormLayoutTheme.js';
import { l as labelSlotStyles, a as labelInnerStyles, S as StyledLabelValueLayout } from './StyledLabelValueLayout-bb05f474.js';
import { c as computeStartLabelWidth } from './formControlUtils-6fc3a497.js';
import './SwitchLabelValueLayoutStyles.styles.css';
import { S as StyledSwitch } from './StyledSwitch-dbf29d95.js';
import { u as useTestId } from './useTestId-f24ac480.js';
import { l } from './stringUtils-0fe25200.js';
import { u as useId } from './useId-03dbfdf0.js';
import { u as useFormFieldContextProps } from './useFormFieldContextProps-31bbc12f.js';
import { u as useMessageSeverity } from './useMessageSeverity-65295e8f.js';
import { L as Label } from './Label-5d1ae8d7.js';
import './logger-c92f309c.js';
import './TabbableModeContext-7d8ad946.js';
import { u as useTranslationBundle } from './useTranslationBundle-4822502f.js';
import './UserAssistanceStyles.styles.css';
import './UNSAFE_UserAssistance/themes/redwood/UserAssistanceBaseTheme.styles.css';
import './UNSAFE_UserAssistance/themes/redwood/UserAssistanceVariants.css.js';
import { a as InlineUserAssistance } from './InlineUserAssistance-890293dd.js';
import './Flex-03129963.js';
import './MessageBannerStyles.styles.css';
import './UNSAFE_MessageBanner/themes/redwood/MessageBannerBaseTheme.styles.css';
import './UNSAFE_MessageBanner/themes/redwood/MessageBannerVariants.css.js';
import 'preact';
import './LayerHost-7931b424.js';
import './TooltipContentStyles.styles.css';
import './hooks/UNSAFE_useTooltip/themes/redwood/TooltipContentBaseTheme.styles.css';
import './hooks/UNSAFE_useTooltip/themes/redwood/TooltipContentVariants.css.js';
import './tooltipUtils-af0abd38.js';
import './IconStyle.styles.css';
import './IconButton-7f91481c.js';
import './MessageStyles.styles.css';
import './MessageFormattingUtils-cdb25115.js';
import './TransitionGroup-5fd80dc9.js';
import './MessagesContext-76544845.js';
import './HiddenAccessibleStyles.styles.css';
import './ComponentMessageStyles.styles.css';
import './UNSAFE_ComponentMessage/themes/redwood/ComponentMessageBaseTheme.styles.css';
import './UNSAFE_ComponentMessage/themes/redwood/ComponentMessageVariants.css.js';
import './Popup-992b6c27.js';
import './UNSAFE_Separator/themes/SeparatorStyles.css.js';
import './LabelStyles.styles.css';
import './UNSAFE_Label/themes/redwood/LabelBaseTheme.styles.css';
import './UNSAFE_Label/themes/redwood/LabelVariants.css.js';
import './InputGroupContext-20bdbd6e.js';
import './UNSAFE_Popup/themes/redwood/PopupBaseTheme.styles.css';
import './UNSAFE_Popup/themes/redwood/PopupVariants.css.js';
import { SwitchRedwoodTheme } from './UNSAFE_Switch/themes/redwood/SwitchTheme.js';
import './UNSAFE_Label/themes/LabelStyles.css.js';
import './UNSAFE_Label/themes/redwood/LabelBaseTheme.css.js';
import './vanilla-extract-recipes-createRuntimeFn.esm-2aaf8c98.js';
import './UNSAFE_TextField/themes/FormLayoutStyles.css.js';
import './UNSAFE_TextField/themes/redwood/FormLayoutBaseTheme.css.js';
import './UNSAFE_TextField/themes/redwood/FormLayoutBaseTheme.styles.css';
import './LabelValueLayoutStyles.styles.css';
import './size-af1e638c.js';
import './utils-b1eb54f9.js';
import './themeContract.css-1ffed3bf.js';
import './colorUtils-b3804ee8.js';
import './_curry1-63949a9b.js';
import './LayoutStyles.css-6b02421a.js';
import './LayoutStyles.styles.css';
import './PRIVATE_StyledSwitch/themes/StyledSwitchStyles.css.js';
import './StyledSwitchStyles.styles.css';
import './useTabbableMode-090ed4b5.js';
import './PRIVATE_StyledSwitch/themes/redwood/StyledSwitchTheme.js';
import './PRIVATE_StyledSwitch/themes/redwood/StyledSwitchVariants.css.js';
import './StyledSwitchVariants.styles.css';
import './HiddenAccessible-61197e7c.js';
import './useActive-fb6d4d52.js';
import './useToggle-8b7fcefe.js';
import './mergeProps-88ea8306.js';
import './useTooltip-94af94d4.js';
import './useTooltipControlled-f2cfa454.js';
import './useHover-85077adb.js';
import './useFocus-38c95977.js';
import './useTouch-4828df25.js';
import './useAnimation-fe39685e.js';
import './useThemeInterpolations-e144e8ec.js';
import './useColorScheme-c4f6caab.js';
import './useScale-434ad939.js';
import './theme-63551f30.js';
import './Theme-e6dec6db.js';
import './mergeInterpolations-30cd9a69.js';
import './mergeDeepWithKey-33e3d1be.js';
import './_curry3-577eb45e.js';
import './_curry2-01c87f61.js';
import './_isObject-1fab0f5b.js';
import './useDensity-1acb0153.js';
import './unsafeDomAccess-453eb74e.js';
import './hooks/UNSAFE_useTooltip/themes/redwood/TooltipContentTheme.js';
import './hooks/UNSAFE_useTooltip/themes/TooltipContentStyles.css.js';
import './hooks/UNSAFE_useTooltip/themes/redwood/TooltipContentBaseTheme.css.js';
import './TooltipContentVariants.styles.css';
import './clientHints-a42b510c.js';
import './CompactLabelAssistance-e2fcfcab.js';
import './Help-32a9689a.js';
import './Icon-4983d06e.js';
import './UNSAFE_Icon/themes/IconStyle.css.js';
import './useFocusWithin-30b1e2d8.js';
import './UNSAFE_UserAssistance/themes/UserAssistanceStyles.css.js';
import './truncationUtils-e37f95ee.js';
import './UNSAFE_TextField/themes/redwood/TextFieldTheme.js';
import './UNSAFE_TextField/themes/TextFieldStyles.css.js';
import './SkeletonStyles.styles.css';
import './TextFieldLoadingStyles.styles.css';
import './TextFieldStyles.styles.css';
import './UNSAFE_TextField/themes/redwood/TextFieldBaseTheme.css.js';
import './UNSAFE_TextField/themes/redwood/TextFieldBaseTheme.styles.css';
import './UNSAFE_TextField/themes/redwood/TextFieldVariants.css.js';
import './TextFieldVariants.styles.css';
import './ComponentMessageContainer-d16c8b1f.js';
import './MessagesManager-47000438.js';
import './SuccessS-b5192d40.js';
import './ErrorS-3a890bf6.js';
import './InformationS-d53c2a91.js';
import './WarningS-3e53d87e.js';
import './PRIVATE_Message/themes/MessageStyles.css.js';
import './soundUtils-88264cb5.js';
import './useMessagesContext-9e1dbe91.js';
import './Transition-f9501682.js';
import './UNSAFE_ComponentMessage/themes/redwood/ComponentMessageTheme.js';
import './UNSAFE_ComponentMessage/themes/ComponentMessageStyles.css.js';
import './UNSAFE_ComponentMessage/themes/redwood/ComponentMessageBaseTheme.css.js';
import './UNSAFE_UserAssistance/themes/redwood/UserAssistanceTheme.js';
import './UNSAFE_UserAssistance/themes/redwood/UserAssistanceBaseTheme.css.js';
import './InlineHelp-9eaac91c.js';
import './useInputGroupContext-1d19411e.js';
import './boxalignment-8980856d.js';
import './arrayUtils-35a58161.js';
import './boxalignment.styles.css';
import './vanilla-extract-sprinkles-createRuntimeSprinkles.esm-2d655d37.js';
import './dimensions-e4fbc23d.js';
import './flexbox-529f25da.js';
import './flexbox.styles.css';
import './flexitem-7b7f7920.js';
import './flexitem.styles.css';
import './FlexStyles.styles.css';
import './MessageBannerVariants.styles.css';
import './index-aec1cc7f.js';
import './UNSAFE_Layer/themes/LayerHostStyles.css.js';
import './LayerHostStyles.styles.css';
import './EnvironmentProvider-6a0a8212.js';
import './LayerManager-52a26c55.js';
import './Floating-4a82db94.js';
import './useFloating-b9d680f7.js';
import './useUser-06d34d19.js';
import './positionUtils-f26caad9.js';
import './refUtils-76275c69.js';
import './useOutsideClick-b777a43e.js';
import './Layer-535577b5.js';
import './UNSAFE_Layer/themes/LayerStyles.css.js';
import './LayerStyles.styles.css';
import './useModal-953082f2.js';
import './UNSAFE_Floating/themes/redwood/FloatingTheme.js';
import './UNSAFE_Floating/themes/FloatingStyles.css.js';
import './FloatingStyles.styles.css';
import './UNSAFE_Floating/themes/redwood/FloatingBaseTheme.css.js';
import './UNSAFE_Floating/themes/redwood/FloatingBaseTheme.styles.css';
import './UNSAFE_Floating/themes/redwood/FloatingVariants.css.js';
import './vanilla-extract-dynamic.esm-c909758c.js';
import './UNSAFE_Floating/themes/FloatingContract.css.js';
import './useUnsafeDomElementRef-32290273.js';
import './BaseButton-41068c85.js';
import './UNSAFE_BaseButton/themes/redwood/BaseButtonTheme.js';
import './UNSAFE_BaseButton/themes/BaseButtonStyles.css.js';
import './BaseButtonStyles.styles.css';
import './UNSAFE_BaseButton/themes/redwood/BaseButtonVariants.css.js';
import './UNSAFE_BaseButton/themes/redwood/BaseButtonBaseTheme.styles.css';
import './BareButton-1e3df7cb.js';
import './useBareButton-ff4bd21e.js';
import './usePress-051f5adb.js';
import './usePressClick-945cc43b.js';
import './buttonUtils-55fb1df7.js';
import './id-83adac50.js';
import './ButtonLabelLayout-1631c997.js';
import './Text-834eb35e.js';
import './UNSAFE_Text/themes/TextStyles.css.js';
import './TextStyles.styles.css';
import './useButtonLabelLayoutTheme-448e671d.js';
import './UNSAFE_ButtonLabelLayout/themes/redwood/ButtonLabelLayoutTheme.js';
import './UNSAFE_ButtonLabelLayout/themes/ButtonLabelLayoutStyles.css.js';
import './ButtonLabelLayoutStyles.styles.css';
import './UNSAFE_ButtonLabelLayout/themes/redwood/ButtonLabelLayoutBaseTheme.css.js';
import './UNSAFE_ButtonLabelLayout/themes/redwood/ButtonLabelLayoutBaseTheme.styles.css';
import './UNSAFE_ButtonLabelLayout/themes/redwood/ButtonLabelLayoutVariants.css.js';
import './ButtonLabelLayoutVariants.styles.css';
import './getLocale-1c41e86a.js';
import './tabbableUtils-c8311ec6.js';
import './head-8f89fe66.js';
import './_arity-be492b9e.js';
import './_isArray-694cc52d.js';
import './_isString-675f1de9.js';
import './FocusTrap-b59d5752.js';
import './FocusTracker-70300f1c.js';
import './PRIVATE_FocusTracker/themes/FocusTrackerStyles.css.js';
import './FocusTrackerStyles.styles.css';
import './Modal-c0f730b8.js';
import './UNSAFE_Modal/themes/ModalStyles.css.js';
import './ModalStyles.styles.css';
import './popupUtils-51366a29.js';
import './WindowOverlay-7264c44a.js';
import './UNSAFE_WindowOverlay/themes/WindowOverlayStyles.css.js';
import './WindowOverlayStyles.styles.css';
import './UNSAFE_WindowOverlay/themes/WindowOverlayContract.css.js';
import './UNSAFE_WindowOverlay/themes/redwood/WindowOverlayTheme.js';
import './UNSAFE_WindowOverlay/themes/redwood/WindowOverlayBaseTheme.css.js';
import './UNSAFE_WindowOverlay/themes/redwood/WindowOverlayBaseTheme.styles.css';
import './UNSAFE_WindowOverlay/themes/redwood/WindowOverlayVariants.css.js';
import './usePopupAnimation-96d2bc83.js';
import './useAnimationStatus-b881f39a.js';
import './popupAnimationUtils-175151d9.js';
import './LiveRegion-cebac18d.js';
import './LiveRegionStyles.styles.css';
import './UNSAFE_Popup/themes/PopupContract.css.js';
import './UNSAFE_Popup/themes/redwood/PopupTheme.js';
import './UNSAFE_Popup/themes/redwood/PopupBaseTheme.css.js';
import './PopupStyles.styles.css';
import './colors-03807160.js';
import './SeparatorStyles.styles.css';
import './UNSAFE_Switch/themes/redwood/SwitchBaseTheme.css.js';
import './UNSAFE_Switch/themes/redwood/SwitchBaseTheme.styles.css';

var switchLabelInnerStyles = {startSwitch:'SwitchLabelValueLayoutStyles_switchLabelInnerStyles_startSwitch__a7cetq0',insideSwitch:'SwitchLabelValueLayoutStyles_switchLabelInnerStyles_insideSwitch__a7cetq1'};

const SwitchLabelValueLayout = ({ columnSpan = 1, label, labelEdge = 'start', children, labelStartWidth = '33%' }) => {
    const { baseTheme: labelTheme } = useComponentTheme(LabelRedwoodTheme);
    const { baseTheme: formLayoutTheme } = useComponentTheme(FormLayoutRedwoodTheme);
    const isStart = labelEdge === 'start';
    const isTopOrInside = labelEdge === 'top' || labelEdge === 'inside';
    // labelStartWidth needs to take columnSpan into account
    const width = computeStartLabelWidth(labelStartWidth, columnSpan);
    const valueWidth = `calc(100% - ${width})`;
    const { isFormLayout, isReadonly: isFormReadonly } = useFormContext();
    const { isReadonly } = useFormFieldContext();
    const isPureReadonly = (isFormLayout && isFormReadonly) || (!isFormLayout && isReadonly);
    const labelStyleClasses = classNames([
        formLayoutTheme,
        labelSlotStyles.base,
        isStart && labelSlotStyles.start,
        isTopOrInside && labelSlotStyles.top,
        isTopOrInside && isPureReadonly && labelSlotStyles.topPureReadonly
    ]);
    const labelInnerClasses = classNames([
        labelTheme,
        formLayoutTheme,
        labelInnerStyles.base,
        isStart && labelInnerStyles.start,
        isStart && switchLabelInnerStyles.startSwitch
    ]);
    const labelStyles = isStart ? { flexBasis: width, width, maxWidth: width } : {};
    const valueStyles = isStart
        ? { flexBasis: valueWidth, width: valueWidth, maxWidth: valueWidth }
        : {};
    // Outer container and inner wrapper match CheckboxRadioLabelValueLayout's order exactly:
    return (jsx(StyledLabelValueLayout, { flexAlign: isReadonly && isStart ? 'stretch' : 'start', labelStyleClasses: labelStyleClasses, labelStyles: labelStyles, labelInnerClasses: labelInnerClasses, label: label, valueStyles: valueStyles, children: children }));
};

const SwitchField = ({ columnSpan, inlineUserAssistance, label, labelEdge, labelStartWidth, mainContent }) => {
    // Use SwitchLabelValueLayout for labelEdge 'start', 'top', 'inside'
    if (labelEdge && labelEdge !== 'none') {
        return (jsxs(SwitchLabelValueLayout, { columnSpan: columnSpan, label: label, labelEdge: labelEdge, labelStartWidth: labelStartWidth, children: [mainContent, inlineUserAssistance] }));
    }
    // Otherwise, for 'none', render children/assistance inline
    return (jsxs("div", { style: { display: 'flex', flexDirection: 'column' }, children: [mainContent, inlineUserAssistance] }));
};

/**
 * A custom hook to determine the props for a component that renders
 * a Switch. This is similar to useTextField, but unlike TextField
 * Switch does not render truncated labels nor does it render required.
 */
function useSwitch(props) {
    const { ariaDescribedBy, helpSourceLink, helpSourceText, isDisabled, isReadonly, label, labelEdge, messages, userAssistanceDensity, value } = props;
    const baseId = useId();
    const labelId = labelEdge !== 'none' ? `${baseId}-label` : undefined;
    const uaId = !isDisabled ? `${baseId}-ua` : undefined;
    const hasValue = value ?? false;
    // Create the form field context (no truncation/required logic needed)
    const formFieldContext = useFormFieldContextProps({
        hasValue,
        isDisabled,
        isReadonly
    });
    // Determine the highest severity from the messages
    const messageSeverity = useMessageSeverity(messages);
    // Determine the label variant
    const labelVariant = labelEdge !== 'none' ? labelEdge : undefined;
    const ariaLabel = label && label.length > 0 && labelEdge === 'none' ? label : undefined;
    const ariaLabelledBy = ariaLabel ? undefined : labelId;
    // According to UX, Switch's compact user assistance density behaves like efficient.
    const resolvedUserAssistanceDensity = userAssistanceDensity === 'compact' ? 'efficient' : userAssistanceDensity;
    return {
        baseId,
        formFieldContext,
        styledSwitchProps: {
            'aria-describedby': l([uaId, ariaDescribedBy]),
            'aria-invalid': messageSeverity === 'error'
                ? 'true'
                : undefined,
            ...(ariaLabelledBy && { 'aria-labelledby': ariaLabelledBy }),
            ...(ariaLabel && { 'aria-label': ariaLabel })
        },
        labelProps: {
            helpSourceLink,
            helpSourceText,
            id: labelId,
            userAssistanceDensity: resolvedUserAssistanceDensity,
            variant: labelVariant
        },
        userAssistanceProps: {
            id: uaId
        }
    };
}

/**
 * The Switch component toggles between two mutually exclusive states, such as on and off.
 *
 */
const Switch = forwardRef(({ 'aria-describedby': ariaDescribedBy, assistiveText, columnSpan, isDisabled: propIsDisabled, isReadonly: propIsReadonly, label, labelEdge: propLabelEdge, labelStartWidth: propLabelStartWidth, messages, onCommit, helpSourceLink, helpSourceText, testId, userAssistanceDensity: propUserAssistanceDensity, value: propValue }, ref) => {
    const { isDisabled: isFormDisabled, isReadonly: isFormReadonly, labelEdge: formLabelEdge, labelStartWidth: formLabelStartWidth, userAssistanceDensity: formUserAssistanceDensity } = useFormContext();
    // default to FormContext values if component properties are not specified
    const isDisabled = propIsDisabled ?? isFormDisabled;
    const isReadonly = propIsReadonly ?? isFormReadonly;
    const labelEdge = propLabelEdge ?? formLabelEdge;
    const labelStartWidth = propLabelStartWidth ?? formLabelStartWidth;
    const userAssistanceDensity = propUserAssistanceDensity ?? formUserAssistanceDensity;
    const { baseId, formFieldContext, styledSwitchProps, labelProps, userAssistanceProps } = useSwitch({
        ariaDescribedBy,
        helpSourceLink,
        helpSourceText,
        isDisabled,
        isReadonly,
        label,
        labelEdge,
        messages,
        userAssistanceDensity,
        value: propValue
    });
    // This ref is used for both imperative focus and enabling SwitchField to "click label to focus"
    const switchElementRef = useRef(null);
    // setRefs is passed as the ref prop to <StyledSwitch> below.
    // It synchronizes both:
    // - switchElementRef (internal, for label click/focus logic)
    // - the parent/test ref (for imperative .focus()/.blur() calls)
    // so that both internal and external focus logic work with the same DOM node.
    function setRefs(node) {
        switchElementRef.current = node;
        if (typeof ref === 'function') {
            ref(node);
        }
        else if (ref) {
            ref.current = node;
        }
    }
    // Label click focus logic
    const onClick = useCallback(() => {
        if (switchElementRef) {
            switchElementRef.current?.focus();
        }
    }, [switchElementRef]);
    const labelComp = labelEdge !== 'none' ? (jsx(Label, { ...labelProps, ...(!isDisabled && !isReadonly ? { onClick } : {}), parentComponentVariant: "switch", children: label })) : undefined;
    const fieldLabelProps = {
        label: labelEdge !== 'none' ? labelComp : undefined,
        labelEdge: labelEdge !== 'none' ? labelEdge : undefined,
        labelStartWidth: labelEdge !== 'none' ? labelStartWidth : undefined
    };
    // According to UX, Switch's compact user assistance density behaves like efficient.
    const resolvedUserAssistanceDensity = userAssistanceDensity === 'compact' ? 'efficient' : userAssistanceDensity;
    const isInlineDensity = resolvedUserAssistanceDensity === 'efficient' || resolvedUserAssistanceDensity === 'reflow';
    const inlineUserAssistance = isInlineDensity ? (isDisabled ? (
    // save space for user assistance if density is 'efficient', even though we don't
    // render user assistance for disabled fields.
    // showInlineAssistanceOn is 'always' because for Switch we always show the user assistance.
    resolvedUserAssistanceDensity !== 'efficient' ? undefined : (jsx(InlineUserAssistance, { showInlineAssistanceOn: "always", userAssistanceDensity: resolvedUserAssistanceDensity, ...userAssistanceProps }))) : isReadonly ? (jsx(InlineUserAssistance, { fieldLabel: label, showInlineAssistanceOn: "always", messages: messages, userAssistanceDensity: resolvedUserAssistanceDensity, ...userAssistanceProps })) : (jsx(InlineUserAssistance, { assistiveText: assistiveText, fieldLabel: label, helpSourceLink: helpSourceLink, helpSourceText: helpSourceText, showInlineAssistanceOn: "always", messages: messages, userAssistanceDensity: resolvedUserAssistanceDensity, ...userAssistanceProps }))) : undefined;
    // toggle the value when we get the StyledSwitch's onChange handler is activated
    const onChangeHandler = useCallback(() => {
        const previousValue = !!propValue;
        const value = !previousValue;
        onCommit?.({
            previousValue,
            value
        });
    }, [onCommit, propValue]);
    const translations = useTranslationBundle('@oracle/oraclejet-preact');
    const switchOn = translations.switch_on();
    const switchOff = translations.switch_off();
    const testIdProps = useTestId(testId);
    const styledSwitchTestId = testId && `${testId}-styled-switch`;
    const styledSwitchComp = (jsx(StyledSwitch, { ...styledSwitchProps, baseId: baseId, isChecked: propValue, isDisabled: isDisabled, isReadonly: isReadonly, readonlyOnText: switchOn, readonlyOffText: switchOff, ref: setRefs, testId: styledSwitchTestId, onChange: onChangeHandler }));
    return (jsx("div", { ...testIdProps, className: SwitchRedwoodTheme.baseTheme, children: jsx(FormFieldContext.Provider, { value: formFieldContext, children: jsx(SwitchField, { columnSpan: columnSpan, inlineUserAssistance: inlineUserAssistance, mainContent: styledSwitchComp, ...fieldLabelProps }) }) }));
});

export { Switch };
