// @oracle/oraclejet-preact/es/UNSAFE_ButtonSetMultiple.js
/* @oracle/oraclejet-preact: undefined */
import { jsx } from 'preact/jsx-runtime';
import { forwardRef } from 'preact/compat';
import { B as ButtonSetContext } from './useButtonSetContext-6a483d7f.js';
import 'preact/hooks';
import { B as ButtonSet } from './ButtonSet-b7464e9a.js';
import { p as processGlobals } from './buttonUtils-55fb1df7.js';
import { u as useId } from './useId-03dbfdf0.js';
import 'preact';
import './dimensions-e4fbc23d.js';
import './size-af1e638c.js';
import './utils-b1eb54f9.js';
import './themeContract.css-1ffed3bf.js';
import './colorUtils-b3804ee8.js';
import './_curry1-63949a9b.js';
import './mergeInterpolations-30cd9a69.js';
import './classNames-4e12b00d.js';
import './mergeDeepWithKey-33e3d1be.js';
import './_curry3-577eb45e.js';
import './_curry2-01c87f61.js';
import './_isObject-1fab0f5b.js';
import './UNSAFE_ButtonSet/themes/ButtonSetStyles.css.js';
import './ButtonSetStyles.styles.css';
import './ButtonSetPositionContext-5730d367.js';
import './toggleButtonUtils-75e3d781.js';
import './useTestId-f24ac480.js';
import './LayerHost-7931b424.js';
import './index-aec1cc7f.js';
import './UNSAFE_Layer/themes/LayerHostStyles.css.js';
import './LayerHostStyles.styles.css';
import './useRovingTabIndexContainer-5992e107.js';
import './useUser-06d34d19.js';
import './clientHints-a42b510c.js';
import './TabbableModeContext-7d8ad946.js';
import './useTabbableMode-090ed4b5.js';
import './refUtils-76275c69.js';
import './id-83adac50.js';

/**
 * ButtonSet allow users to select the state of one or more related options.
 */
const ButtonSetMultiple = forwardRef(({ children, value: buttonSetItemValue, onCommit, display = 'all', testId, size = 'md', isDisabled = false, variant = 'outlined', layoutWidth = 'auto', 'aria-label': accessibleLabel, 'aria-labelledby': ariaLabelledBy, 'aria-controls': ariaControls, 'aria-describedby': ariaDescribedBy, width, ...props }, ref) => {
    const { globalEventProps } = processGlobals(props);
    return (jsx(ButtonSetContext.Provider, { value: {
            variant: variant,
            size: size,
            isDisabled: isDisabled,
            layoutWidth: layoutWidth,
            inputType: 'checkbox',
            inputName: useId(),
            display: display,
            width: width,
            buttonSetValue: buttonSetItemValue,
            onCommit: onCommit
        }, children: jsx(ButtonSet, { ref: ref, width: width, testId: testId, "aria-label": accessibleLabel, "aria-labelledby": ariaLabelledBy, "aria-describedby": ariaDescribedBy, "aria-controls": ariaControls, ...globalEventProps, children: children }) }));
});
ButtonSetMultiple.displayName = 'ButtonSetMultiple';

export { ButtonSetMultiple };
