// @oracle/oraclejet-preact/es/useChartDatatip-8dd22b71.js
/* @oracle/oraclejet-preact: undefined */
import { u as useDatatip } from './useDatatip-4895a07a.js';
import { c as calculateOffset } from './layoutUtils-ec5be703.js';
import { jsx } from 'preact/jsx-runtime';
import { V as VisTabularDatatip } from './VisTabularDatatip-85b00c1b.js';
import { a as getElementData } from './eventsUtils-14cb0a49.js';
import { g as generateAriaLabel } from './accUtils-9d5183c4.js';
import { g as getColorRamp } from './colorUtils-77b8874f.js';
import { u as useUser } from './useUser-06d34d19.js';

/**
 * Returns the data required for the VisTabularDatatip to render.
 * @param activeInfo active item info.
 * @param series The array of chart series.
 * @param groups The array of chart groups.
 * @param valueFormats The array of chart series.
 * @param getDataItem function which returns chartItem
 */
function getChartTabularData(data, seriesValue, groupValue, translations, valueFormats, isTimeAxis, defaultXFormatter, defaultYFormatter, isSeriesAssociatedToY2, hasXYValues) {
    let itemValue, itemX, itemY, itemWidthWeight, itemY2;
    const seriesLabel = translations.chart_labelSeries();
    const groupLabel = translations.chart_labelGroup();
    const valueLabel = translations.chart_labelValue();
    const xLabel = translations.chart_labelX();
    const yLabel = translations.chart_labelY();
    const widthWeightLabel = translations.chart_labelWidthWeight();
    const dataLabel = translations.chart_labelDataLabel();
    if (hasXYValues && 'y' in data) {
        const x = data.x;
        itemX = valueFormats?.x?.format ? valueFormats?.x?.format(x) : defaultXFormatter?.(x);
        const y = data.y;
        itemY = valueFormats?.y?.format ? valueFormats?.y?.format(y) : defaultYFormatter?.(y);
    }
    else if ('value' in data && data.value !== undefined) {
        const value = data.value;
        const formattedValue = valueFormats?.value?.format
            ? valueFormats.value.format(value)
            : defaultYFormatter?.(value);
        if (isSeriesAssociatedToY2) {
            itemY2 = valueFormats?.y2?.format ? valueFormats.y2.format(value) : formattedValue;
        }
        else {
            itemValue = formattedValue;
        }
    }
    if ('widthWeight' in data && data.widthWeight !== undefined) {
        const widthWeight = data.widthWeight;
        itemWidthWeight = valueFormats?.widthWeight?.format
            ? valueFormats?.widthWeight?.format?.(widthWeight)
            : defaultYFormatter?.(widthWeight);
    }
    const tableData = [];
    if (valueFormats?.series?.isDisplayed != false) {
        tableData.push({
            key: valueFormats?.series?.label || seriesLabel,
            value: seriesValue
        });
    }
    if (valueFormats?.group?.isDisplayed != false) {
        const isArray = Array.isArray(valueFormats?.group?.label);
        if (isTimeAxis) {
            const group = Array.isArray(groupValue) ? groupValue[0] : groupValue;
            tableData.push({
                key: (isArray ? valueFormats?.group?.label?.[0] : valueFormats?.group?.label) ||
                    'Date',
                value: defaultXFormatter ? defaultXFormatter?.(new Date(group).getTime()) : group
            });
        }
        else {
            const groupsLabel = (isArray ? valueFormats?.group?.label : [valueFormats?.group?.label || groupLabel]);
            const groups = Array.isArray(groupValue) ? groupValue : [groupValue];
            groups.forEach((value, index) => {
                tableData.push({
                    key: groupsLabel[index],
                    value: value
                });
            });
        }
    }
    if (valueFormats?.value?.isDisplayed != false) {
        if (hasXYValues) {
            tableData.push({
                key: valueFormats?.value?.label || valueFormats?.x?.label || xLabel,
                value: itemX
            });
            tableData.push({
                key: valueFormats?.value?.label || valueFormats?.y?.label || yLabel,
                value: itemY
            });
        }
        else {
            if (isSeriesAssociatedToY2) {
                tableData.push({
                    key: valueFormats?.value?.label || valueFormats?.y2?.label || valueLabel,
                    value: itemY2
                });
            }
            else {
                tableData.push({
                    key: valueFormats?.value?.label || valueLabel,
                    value: itemValue
                });
            }
        }
    }
    if (valueFormats?.widthWeight?.isDisplayed != false && itemWidthWeight) {
        tableData.push({
            key: valueFormats?.widthWeight?.label || widthWeightLabel,
            value: itemWidthWeight
        });
    }
    if (valueFormats?.label?.isDisplayed != false && data.label) {
        tableData.push({
            key: valueFormats?.label?.label || dataLabel,
            value: data.label
        });
    }
    return tableData;
}
function getDefaultDatatipRenderer(options) {
    const isSeriesAssociatedToY2 = options.series.associatedYAxis === 'y2';
    const isMixedFrequency = options.timeAxisType === 'mixedFrequency';
    const isRegularTimeAxis = options.timeAxisType === 'enabled' || options.timeAxisType === 'skipGaps';
    return (context) => {
        const groupValue = options.groups;
        const timeValue = (isMixedFrequency ? context.data.x : isRegularTimeAxis ? groupValue : undefined);
        return {
            content: (jsx(VisTabularDatatip, { tableData: getChartTabularData(context.data, options.series.name || options.series.id.toString(), timeValue || options.groups, options.translations, options.valueFormats, options.timeAxisType != null, options.defaultXFormatter, isSeriesAssociatedToY2 ? options.defaultY2Formatter : options.defaultYFormatter, isSeriesAssociatedToY2, options.hasXYValues) })),
            borderColor: context.data.color || options.series.color,
            defaultContainer: 'enabled'
        };
    };
}
function getAriaLabelGenerator(series, translations, valueFormats, leafGroupsInfo, xAxis, isDataXAxis, getGroupFromPath, xDefaultLabelFormatter, yDefaultLabelFormatter, y2DefaultLabelFormatter) {
    const getItemAriaLabel = (data, seriesIndex, groupIndex, isSelectionEnabled, isSelected, isDrillable) => {
        let accLabel = data.accessibleLabel;
        if (!accLabel) {
            const _series = series[seriesIndex];
            const isAssociatedToY2 = _series.associatedYAxis === 'y2';
            const groupValue = getGroupFromPath(leafGroupsInfo[groupIndex].path);
            const timeValue = (xAxis?.timeAxisType === 'mixedFrequency'
                ? data.x
                : xAxis?.timeAxisType === 'enabled'
                    ? groupValue
                    : undefined);
            const tababularData = getChartTabularData(data, _series.name || _series.id.toString(), timeValue || groupValue, translations, valueFormats, xAxis?.timeAxisType != null, xDefaultLabelFormatter, isAssociatedToY2 ? y2DefaultLabelFormatter : yDefaultLabelFormatter, isDataXAxis);
            accLabel = tababularData
                .map(({ key, value }) => `${key ? key + ':' : ''} ${value}`)
                .join(', ');
        }
        return generateAriaLabel(translations, accLabel, {
            isSelected: !isSelectionEnabled ? undefined : isSelected,
            isDrillable
        });
    };
    return getItemAriaLabel;
}
const getTextAndBoundsUtil = (target, rootRef, xAxisTitle, yAxisTitle, y2AxisTitle, xTitleDimention, yTitleDimention, y2TitleDimention, xAxisTickLabels, focusedGroupInfo) => {
    if (target != rootRef?.current) {
        const dataSet = getElementData(rootRef, target);
        const text = dataSet?.['ojPrivateText'];
        if (!text) {
            return;
        }
        if (text === xAxisTitle) {
            return {
                text,
                bounds: xTitleDimention
            };
        }
        else if (text === yAxisTitle) {
            return {
                text,
                bounds: yTitleDimention
            };
        }
        else if (text === y2AxisTitle) {
            return {
                text,
                bounds: y2TitleDimention
            };
        }
        else if (dataSet?.['ojPrivateGroupIndex'] && dataSet?.['ojPrivateLevelIndex']) {
            return {
                // @ts-ignore
                bounds: xAxisTickLabels.find(({ dataProps }) => dataProps['data-oj-private-level-index'] === dataSet['ojPrivateLevelIndex'] &&
                    dataProps['data-oj-private-group-index'] === dataSet['ojPrivateGroupIndex']).dims,
                text: text
            };
        }
    }
    if (target === rootRef?.current && focusedGroupInfo?.isFocusVisible) {
        const dimsAndText = getTextAndDimsInfo(xAxisTickLabels, focusedGroupInfo.groupIndex, focusedGroupInfo.levelIndex);
        return {
            bounds: dimsAndText?.dims,
            text: dimsAndText?.text
        };
    }
    return;
};
const getTextAndDimsInfo = (tickLabels, groupIndex, levelIndex) => {
    return tickLabels?.find((tickLabel) => tickLabel?.dataProps?.['data-oj-private-group-index'] == groupIndex &&
        tickLabel?.dataProps?.['data-oj-private-level-index'] == levelIndex &&
        tickLabel?.isTruncated);
};

/**
 * @license
 * Copyright (c) %FIRST_YEAR% %CURRENT_YEAR%, Oracle and/or its affiliates.
 * Licensed under The Universal Permissive License (UPL), Version 1.0
 * as shown at https://oss.oracle.com/licenses/upl/
 * @ignore
 */
const DATA_CURSOR_OFFSET = 10;
const useChartDatatip = ({ isDataCursorEnabled, dataCursorPosition, isHoriz, datatip, xScale, getDataItem, getRefItem, getDataItemPos, getRefObjectBounds, getRefAxis, hoveredItemInfo, focusedItemInfo, series, valueFormats, leafGroupsInfo, width, height, findNearest, timeAxisType, touchResponse, isDisabled, defaultXFormatter, defaultYFormatter, defaultY2Formatter, hasXYValues, translations, getGroupFromPath }) => {
    const hasData = leafGroupsInfo.length > 0 && series.length > 0;
    let activeInfo = !hasData
        ? undefined
        : !isDisabled
            ? focusedItemInfo?.isCurrent
                ? focusedItemInfo
                : hoveredItemInfo?.isCurrent &&
                    ((hoveredItemInfo.seriesIndex !== undefined && hoveredItemInfo.groupIndex !== undefined) ||
                        hoveredItemInfo?.refObjId !== undefined)
                    ? hoveredItemInfo
                    : undefined
            : undefined;
    const anchor = focusedItemInfo?.isCurrent || isDataCursorEnabled ? 'element' : 'pointer';
    const elementBounds = focusedItemInfo?.isCurrent
        ? focusedItemInfo.refObjId != null
            ? getRefObjectBounds(focusedItemInfo.refObjId)
            : getDataItemPos(focusedItemInfo.seriesIndex, focusedItemInfo.groupIndex)
        : undefined;
    const { direction } = useUser();
    const isRtl = direction === 'rtl';
    let placement = 'top-start';
    let offset = calculateOffset(isRtl, width, elementBounds);
    if (isDataCursorEnabled &&
        dataCursorPosition?.x != null &&
        (dataCursorPosition?.y != null || dataCursorPosition?.y2 != null)) {
        const neareast = findNearest(dataCursorPosition.x, dataCursorPosition.y, dataCursorPosition.y2);
        activeInfo = activeInfo || neareast;
        // TOD0: The offset for data cursor should be calculated according to
        // datatip content dimensions as well.
        const xPos = xScale.transform(dataCursorPosition.x);
        offset = { mainAxis: -1, crossAxis: -1 };
        if (isHoriz) {
            placement = 'end';
            offset.mainAxis = -1 * DATA_CURSOR_OFFSET;
            offset.crossAxis = xPos - height / 2;
        }
        else {
            placement = 'top';
            offset.crossAxis = isRtl ? width / 2 - xPos : xPos - width / 2;
            offset.mainAxis = -1 * DATA_CURSOR_OFFSET;
        }
    }
    let dtContent;
    let dtBorderColor;
    let dtDefaultContainer;
    let item;
    if (activeInfo?.refObjId != undefined) {
        item = getRefItem(activeInfo?.refObjId);
        const axis = item ? getRefAxis(activeInfo.refObjId) : undefined;
        const datatipContext = item && axis
            ? item.type === 'line'
                ? { type: 'referenceLine', axis, data: item }
                : { type: 'referenceArea', axis, data: item }
            : undefined;
        if (datatipContext) {
            const customDatatip = datatip?.(datatipContext);
            dtContent = customDatatip?.content;
            dtBorderColor = customDatatip?.borderColor;
            dtDefaultContainer = customDatatip?.defaultContainer;
            if (!dtContent) {
                dtContent = datatipContext.data.accessibleLabel;
            }
        }
    }
    else if (activeInfo?.seriesIndex != null && activeInfo?.groupIndex != null) {
        item = getDataItem(activeInfo.seriesIndex, activeInfo.groupIndex, leafGroupsInfo[activeInfo.groupIndex].path);
        const datatipContext = item
            ? {
                type: 'item',
                data: item,
                seriesIndex: activeInfo.seriesIndex,
                groupIndex: activeInfo.groupIndex,
                id: item.id,
                groupPath: leafGroupsInfo[activeInfo.groupIndex].path
            }
            : undefined;
        if (datatipContext) {
            const customDatatip = datatip?.(datatipContext);
            dtContent = customDatatip?.content;
            dtBorderColor = customDatatip?.borderColor;
            dtDefaultContainer = customDatatip?.defaultContainer;
            if (dtContent !== '') {
                const defaultDatatip = getDefaultDatatipRenderer({
                    series: series[activeInfo.seriesIndex],
                    groups: getGroupFromPath(leafGroupsInfo[activeInfo.groupIndex].path),
                    translations,
                    valueFormats,
                    timeAxisType,
                    defaultXFormatter,
                    defaultYFormatter,
                    defaultY2Formatter,
                    hasXYValues
                })(datatipContext);
                dtContent = dtContent || defaultDatatip.content;
                dtBorderColor = dtBorderColor || defaultDatatip.borderColor;
                dtDefaultContainer = dtDefaultContainer || defaultDatatip.defaultContainer;
            }
        }
    }
    const colors = getColorRamp();
    const borderColor = dtBorderColor ||
        (activeInfo &&
            (series[activeInfo.seriesIndex]?.color ||
                series[activeInfo.seriesIndex]?.lineColor ||
                series[activeInfo.seriesIndex]?.areaColor)) ||
        (activeInfo && colors[activeInfo.seriesIndex % colors.length]);
    const { datatipContent, datatipProps } = useDatatip({
        content: dtContent,
        anchor,
        placement,
        defaultContainer: dtDefaultContainer,
        offset,
        borderColor,
        touchResponse,
        isDisabled
    });
    return { datatipContent, datatipProps };
};

export { getAriaLabelGenerator as a, getTextAndBoundsUtil as b, getTextAndDimsInfo as g, useChartDatatip as u };
