// @oracle/oraclejet-preact/cjs/UNSAFE_ButtonLabelLayout/themes/redwood/ButtonLabelLayoutTheme.sm.tokens.js
/* @oracle/oraclejet-preact: undefined */
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var utils = require('../../../utils-c0efcd5e.js');
require('../../../themeContract.css-f3d431c6.js');
require('../../../colorUtils-01f3caa2.js');
require('../../../_curry1-e8f0d7ea.js');

const scaleSchemeSm = {
    iconSize: '',
    sizes: {
        xs: {
            iconSize: 'max(' + utils.xUnits(5) + ', 20px)'
        },
        sm: {
            iconSize: utils.xUnits(5)
        },
        md: {
            iconSize: utils.xUnits(5.333)
        },
        lg: {
            iconSize: utils.xUnits(5.333)
        },
        embeddedxs: {
            iconSize: 'max(' + utils.xUnits(5) + ', 20px)'
        },
        embeddedsm: {
            iconSize: utils.xUnits(5)
        },
        embeddedmd: {
            iconSize: utils.xUnits(5.333)
        },
        embeddedlg: {
            iconSize: utils.xUnits(5.333)
        }
    }
};

exports.scaleSchemeSm = scaleSchemeSm;
