// @oracle/oraclejet-preact/cjs/utils/UNSAFE_visUtils.js
/* @oracle/oraclejet-preact: undefined */
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var eventsUtils = require('../eventsUtils-b87ea91f.js');
var textUtils = require('../textUtils-e779c845.js');
var colorUtils = require('../colorUtils-b2204778.js');
var accUtils = require('../accUtils-4a359187.js');
var utils = require('../utils-da07b35e.js');
require('../utils-c0efcd5e.js');
require('../themeContract.css-f3d431c6.js');
require('../colorUtils-01f3caa2.js');
require('../_curry1-e8f0d7ea.js');
require('../clientHints-c40c5250.js');



exports.cancelEvent = eventsUtils.cancelEvent;
exports.getElementData = eventsUtils.getElementData;
exports.getInfo = eventsUtils.getInfo;
exports.isHeightOverlap = textUtils.isHeightOverlap;
exports.isWidthOverlap = textUtils.isWidthOverlap;
exports.getColorRamp = colorUtils.getColorRamp;
exports.generateAriaLabel = accUtils.generateAriaLabel;
exports.supportsMobileScreenReader = accUtils.supportsMobileScreenReader;
exports.getShapeRamp = utils.getShapeRamp;
exports.inSpace = utils.inSpace;
