// @oracle/oraclejet-preact/cjs/utils/UNSAFE_calendarDateUtils.js
/* @oracle/oraclejet-preact: undefined */
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var calendarDateUtils = require('../calendarDateUtils-c572f2d3.js');
require('../getFormatParse-b70abd1a.js');
require('../types-55573a46.js');
require('../logger-2b636482.js');



exports.MAX_ISO_YEAR = calendarDateUtils.MAX_ISO_YEAR;
exports.MIN_ISO_DAY = calendarDateUtils.MIN_ISO_DAY;
exports.MIN_ISO_MONTH = calendarDateUtils.MIN_ISO_MONTH;
exports.MIN_ISO_YEAR = calendarDateUtils.MIN_ISO_YEAR;
exports.addToCalendarDate = calendarDateUtils.addToCalendarDate;
exports.addToMonth = calendarDateUtils.addToMonth;
exports.addToYear = calendarDateUtils.addToYear;
exports.checkIsValidIsoYear = calendarDateUtils.checkIsValidIsoYear;
exports.checkNoCalendarExtension = calendarDateUtils.checkNoCalendarExtension;
exports.compareDates = calendarDateUtils.compareDates;
exports.constrainCalendarDate = calendarDateUtils.constrainCalendarDate;
exports.formatCalendarDateAsString = calendarDateUtils.formatCalendarDateAsString;
exports.formatCalendarMonthRequiredAsExample = calendarDateUtils.formatCalendarMonthRequiredAsExample;
exports.formatFullCalendarDate = calendarDateUtils.formatFullCalendarDate;
exports.formatIsoDateStrAsExample = calendarDateUtils.formatIsoDateStrAsExample;
exports.formatShortCalendarDate = calendarDateUtils.formatShortCalendarDate;
exports.getCalendarDateFromIso = calendarDateUtils.getCalendarDateFromIso;
exports.getCalendarDateToday = calendarDateUtils.getCalendarDateToday;
exports.getClosestDecade = calendarDateUtils.getClosestDecade;
exports.getDatePlaceholderMasks = calendarDateUtils.getDatePlaceholderMasks;
exports.getDatePlaceholdersFromPattern = calendarDateUtils.getDatePlaceholdersFromPattern;
exports.getDaysInMonth = calendarDateUtils.getDaysInMonth;
exports.getFormattedYearMonthDay = calendarDateUtils.getFormattedYearMonthDay;
exports.getIsoDateStr = calendarDateUtils.getIsoDateStr;
exports.getLongFormattedYearMonth = calendarDateUtils.getLongFormattedYearMonth;
exports.getMonthYearPlaceholdersFromDatePlaceholders = calendarDateUtils.getMonthYearPlaceholdersFromDatePlaceholders;
exports.getMonthsInYear = calendarDateUtils.getMonthsInYear;
exports.getNumericFormattedYear = calendarDateUtils.getNumericFormattedYear;
exports.isCalendarDateAfter = calendarDateUtils.isCalendarDateAfter;
exports.isCalendarDateBefore = calendarDateUtils.isCalendarDateBefore;
exports.isCompleteCalendarDate = calendarDateUtils.isCompleteCalendarDate;
exports.isCompleteCalendarMonth = calendarDateUtils.isCompleteCalendarMonth;
exports.isDateOutOfRange = calendarDateUtils.isDateOutOfRange;
exports.isValidCalendarMonthRequired = calendarDateUtils.isValidCalendarMonthRequired;
exports.removeExcessiveLiteralsFromPlaceholders = calendarDateUtils.removeExcessiveLiteralsFromPlaceholders;
exports.toIsoCalendarDate = calendarDateUtils.toIsoCalendarDate;
exports.toJSDate = calendarDateUtils.toJSDate;
exports.toJSDateUTC = calendarDateUtils.toJSDateUTC;
