// @oracle/oraclejet-preact/cjs/UNSAFE_RadioItem/themes/RadioStyles.css.js
/* @oracle/oraclejet-preact: undefined */
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });



var radioBaseStyle = 'RadioStyles_radioBaseStyle__o9of5r0';
var radioDefaultColor = 'RadioStyles_radioDefaultColor__o9of5r3';
var radioDisabledColor = 'RadioStyles_radioDisabledColor__o9of5r4';
var styles = {radioBaseStyle:'RadioStyles_radioBaseStyle__o9of5r0',valueLabelStyle:'RadioStyles_valueLabelStyle__o9of5r2'};
var valueLabelStyle = 'RadioStyles_valueLabelStyle__o9of5r2';
var variants = {disabled:{isDisabled:'RadioStyles_radioDisabledColor__o9of5r4',notDisabled:'RadioStyles_radioDefaultColor__o9of5r3'},focusRing:{isFocusRingSelector:'RadioStyles_focusRingSelector__o9of5r1',isForcedFocusRing:{outlineStyle:'dotted',outlineWidth:'var(--outlineWidth__11tjnon0)',outlineColor:'var(--oj-c-border-keyboard-focus)'}}};

exports.radioBaseStyle = radioBaseStyle;
exports.radioDefaultColor = radioDefaultColor;
exports.radioDisabledColor = radioDisabledColor;
exports.styles = styles;
exports.valueLabelStyle = valueLabelStyle;
exports.variants = variants;
