// @oracle/oraclejet-preact/cjs/Dropdown-1ba432a8.js
/* @oracle/oraclejet-preact: undefined */
'use strict';

var jsxRuntime = require('preact/jsx-runtime');
var hooks = require('preact/hooks');
var Floating = require('./Floating-79cbe5ad.js');
var Layer = require('./Layer-b60deb1b.js');
require('preact/compat');
require('./LayerHost-3cd73603.js');
var keyboardUtils = require('./keyboardUtils-b42fe7d5.js');
var tabbableUtils = require('./tabbableUtils-22a50472.js');
var usePopupAnimation = require('./usePopupAnimation-f3b34413.js');
var popupAnimationUtils = require('./popupAnimationUtils-dcace427.js');
var CollectionInteractionContext = require('./CollectionInteractionContext-f84ff7a1.js');
var UNSAFE_Dropdown_themes_dropdownStyles_css = require('./UNSAFE_Dropdown/themes/dropdownStyles.css.js');
var useOutsideMousedown = require('./useOutsideMousedown-70f36cca.js');
var animationUtils = require('./animationUtils-54852d03.js');
var unsafeDomAccess = require('./unsafeDomAccess-c3a492f0.js');
var useEffectEvent = require('./useEffectEvent-832ffba5.js');
var clientHints = require('./clientHints-c40c5250.js');

/**
 * @license
 * Copyright (c) %FIRST_YEAR% %CURRENT_YEAR%, Oracle and/or its affiliates.
 * Licensed under The Universal Permissive License (UPL), Version 1.0
 * as shown at https://oss.oracle.com/licenses/upl/
 * @ignore
 */
/**
 * Returns the key from the item element
 */
const getKey = (item) => {
    // Key is set on data-oj-menu-item-key on each menuItem. To get this key from the
    // the actual menuItem we need to do a kebab to camel-case conversion.
    return item.dataset['ojMenuItemKey'];
};
/**
 * A helper function that return the key of the n available item.
 */
const getNVisible = (root, positionIndex) => {
    if (root) {
        const items = root.querySelectorAll('[data-oj-menu-item-key]');
        const positionedItem = items[positionIndex];
        if (positionedItem) {
            const key = getKey(positionedItem);
            if (key) {
                return key;
            }
        }
    }
    return null;
};
/**
 * A helper function that return the key of the last available item.
 */
const getLastVisible = (root) => {
    if (root) {
        const items = root.querySelectorAll('[data-oj-menu-item-key]');
        const lastItem = items[items.length - 1];
        if (lastItem) {
            const key = getKey(lastItem);
            if (key) {
                return key;
            }
        }
    }
    return null;
};
/**
 * A helper function that returns the key of the previous or the next item
 * given the specified current key
 */
const getPrevNext = (root, isPrev, currentKey) => {
    return () => {
        if (root) {
            const items = Array.from(root.querySelectorAll('[data-oj-menu-item-key]'));
            const currentKeyIndex = items.findIndex((item) => getKey(item) === currentKey);
            let nextIndex = currentKeyIndex + (isPrev ? -1 : 1);
            const itemsLastIndex = items.length - 1;
            if (nextIndex < 0) {
                nextIndex = itemsLastIndex;
            }
            if (nextIndex > itemsLastIndex) {
                nextIndex = 0;
            }
            const key = getKey(items[nextIndex]);
            return key ? key : null;
        }
        return null;
    };
};
/**
 * A helper function to get the key from an element
 */
const keyElemExtractor = (element) => {
    const elem = element.closest('[data-oj-menu-item-key]');
    if (elem) {
        const key = getKey(elem);
        return key === undefined ? null : { key, elem };
    }
    return null;
};
/**
 * Function to get the corresponding keyboard behavior to close the menu.
 */
const getKeyboardCloseProps = (onClose, menuLevel) => {
    const onKeyDown = (event) => {
        if (event.key === 'Tab') {
            event.stopPropagation();
            onClose?.({ reason: 'dismissed' });
            event.preventDefault(); //TODO: Revisit this since we should let focus be at its natural flow, quick solution for now to focus correct item in main use case
        }
        else if (event.key === 'Escape' && menuLevel === 'main') {
            event.stopPropagation();
            onClose?.({ reason: 'dismissed' });
        }
    };
    return { onKeyDown };
};
/**
 * Function to get the corresponding specific behavior for closing submenu via keyboard
 */
const getCloseSubmenuEvent = (direction, closeSubmenu) => {
    if (!closeSubmenu) {
        return {};
    }
    const onKeyDown = (e) => {
        if (e.key === 'Escape' ||
            (direction === 'ltr' && e.key === 'ArrowLeft') ||
            (direction === 'rtl' && e.key === 'ArrowRight')) {
            e.stopPropagation();
            closeSubmenu?.();
        }
    };
    return { onKeyDown };
};
/**
 * Function to get the corresponding specific behavior for opening submenu via keyboard
 */
const getOpenSubmenuEvent = (direction, openSubmenu) => {
    if (!openSubmenu) {
        return {};
    }
    const onKeyDown = (e) => {
        if (direction === 'ltr') {
            if (e.key === 'ArrowRight') {
                openSubmenu?.();
            }
        }
        else {
            if (e.key === 'ArrowLeft') {
                openSubmenu?.();
            }
        }
    };
    return { onKeyDown };
};
/**
 * Functions to get type of device
 */
const getIsMobile = () => {
    const { deviceRenderMode } = clientHints.getClientHints();
    return deviceRenderMode === 'phone';
};
const getIsIosTablet = () => {
    const { deviceRenderMode, platform } = clientHints.getClientHints();
    return platform === 'ios' && deviceRenderMode === 'tablet';
};
/**
 * Get blur and focus events
 */
const getMenuContainerFocusMethods = (blurFunction, focusFunction) => {
    const onBlur = () => {
        setTimeout(() => {
            blurFunction();
        }, 100);
    };
    const onFocus = () => {
        focusFunction();
    };
    return { onBlur, onFocus };
};
/**
 * Set currentKey based on clicking menuItem
 */
const getMenuItemClickChangeKey = (key, changeKeyMethod) => {
    const onClick = () => {
        changeKeyMethod({ key, reason: 'click' });
    };
    return { onClick };
};
const getMenuContainerMouseLeave = (mouseLeaveFunction) => {
    const onMouseLeave = (e) => {
        mouseLeaveFunction(e);
    };
    return { onMouseLeave };
};
/**
 * Function to get the corresponding specific behavior for clicking/touching down menuContainer
 */
const getMenuPointerDown = () => {
    const onPointerDown = (event) => {
        //We just want to prevent default when element that was
        //pressed down was not a menu item.
        const keyElem = keyElemExtractor(event.target);
        if (!keyElem?.key) {
            event.preventDefault();
        }
    };
    return { onPointerDown };
};
const isMenuChildrenFalsy = (children) => {
    return !children || (Array.isArray(children) && !children.length);
};

const DEFAULT_PLACEMENT = 'bottom-start';
const DEFAULT_OFFSET = { mainAxis: 4, crossAxis: 0 };
const DEFAULT_SIZE_OPTIONS = { isMaxWidthAdjusted: true, isMaxHeightAdjusted: true };
const DEFAULT_FLIP_OPTIONS = { mainAxis: true, crossAxis: true };
const animationStates = {
    opening: (node) => {
        return animationUtils.mergeAnimationStateObjects([
            animationUtils.SCALE_UP_FULL,
            {
                from: { transformOrigin: popupAnimationUtils.getTransformOrigin(node.placement, node.direction) },
                to: { transformOrigin: popupAnimationUtils.getTransformOrigin(node.placement, node.direction) }
            }
        ], {
            duration: animationUtils.DURATION_LARGE
        });
    }
};
//TODO: Add animation to dropdown. JET-62628
//TODO: Investigate best approach to have different virtual padding depending on the position and implement it.JET-57950
function Dropdown({ isOpen = false, anchorRef, children, id, onClose, minWidth, placement = DEFAULT_PLACEMENT, offset = DEFAULT_OFFSET, onPosition, flipOptions = DEFAULT_FLIP_OPTIONS, sizeOptions = DEFAULT_SIZE_OPTIONS, shiftOptions, initialFocus = 'none' }) {
    const dropdownRef = hooks.useRef(null);
    const onTransitionEnd = hooks.useCallback((isOpen) => {
        if (!isOpen) {
            return;
        }
        const dropdownElement = dropdownRef.current;
        if (!dropdownElement) {
            return;
        }
        // honor firstFocusable initial focus behavior
        if (initialFocus === 'firstFocusable') {
            tabbableUtils.focusWithin(dropdownElement);
            return;
        }
        // JET-76442 + JET-76109: When VoiceOver is enabled on iOS tablets, we want the first menu item to receive focus.
        // This is to ensure that VoiceOver announces the menu items correctly and focus does not go to the next tabbable element.
        if (getIsIosTablet()) {
            const firstItem = dropdownElement.querySelector('[data-oj-menu-item-key]');
            if (firstItem) {
                tabbableUtils.focusWithin(firstItem);
            }
        }
    }, [initialFocus]);
    const { status, setAnimationElementRef, onPosition: onPositionAnimated } = usePopupAnimation.usePopupAnimation({
        isOpen,
        placement,
        isAnimatedOnMount: true,
        animationStates,
        onPosition,
        onTransitionEnd
    });
    const handleMousedownClose = hooks.useCallback((event) => {
        onClose?.({ reason: 'outsideMousedown', target: event.target });
    }, [onClose]);
    const handleKeyDown = hooks.useCallback((event) => {
        if (event.defaultPrevented) {
            return;
        }
        switch (event.code) {
            case keyboardUtils.KEYS.ESC:
                onClose?.({ reason: 'dismissed' });
                break;
            case keyboardUtils.KEYS.TAB:
                onClose?.({ reason: 'dismissed' });
                break;
        }
    }, [onClose]);
    // use the mousedown event to be consistent with legacy JET
    useOutsideMousedown.useOutsideMousedown({
        isDisabled: false,
        ref: [dropdownRef],
        handler: handleMousedownClose
    });
    // useEffectEvent ensures we always access the latest anchorRef.current value inside effects,
    // without having to include anchorRef in the dependency array and the eslint-disable-next comment
    const STABLE_getAnchor = useEffectEvent.useEffectEvent(() => anchorRef?.current);
    // Restore focus to anchor after Dropdown is closed
    hooks.useEffect(() => {
        const anchorElement = STABLE_getAnchor();
        if (!isOpen &&
            status === 'unmounted' &&
            anchorElement &&
            anchorElement instanceof HTMLElement) {
            if (document.body.contains(anchorElement)) {
                anchorElement.focus();
            }
        }
    }, [STABLE_getAnchor, isOpen, status]);
    return (jsxRuntime.jsx(jsxRuntime.Fragment, { children: status !== 'unmounted' ? (jsxRuntime.jsx(Layer.Layer, { logicalParentRef: toLogicalParentRef(anchorRef), children: jsxRuntime.jsx(Floating.Floating, { anchorRef: anchorRef, ref: setAnimationElementRef, placement: placement, flipOptions: flipOptions, shiftOptions: shiftOptions, sizeOptions: sizeOptions, offsetValue: offset, onPosition: onPositionAnimated, boundaryOptions: { padding: { top: 8, bottom: 8, left: 0, right: 0 } }, children: jsxRuntime.jsx("div", { ref: dropdownRef, id: id, class: UNSAFE_Dropdown_themes_dropdownStyles_css.dropdownStyles.base, style: { minWidth: `${minWidth || 0}px` }, onKeyDown: handleKeyDown, children: jsxRuntime.jsx(CollectionInteractionContext.CollectionInteractionContext.Provider, { value: 'embedded', children: children }) }) }) })) : null }));
}
function toLogicalParentRef(anchorRef) {
    const current = anchorRef.current;
    if (current) {
        if (current instanceof Element) {
            return anchorRef;
        }
        else if (current[unsafeDomAccess.UNSAFE_DOM_ACCESS]) {
            return anchorRef;
        }
    }
    return null;
}

exports.Dropdown = Dropdown;
exports.getCloseSubmenuEvent = getCloseSubmenuEvent;
exports.getIsMobile = getIsMobile;
exports.getKeyboardCloseProps = getKeyboardCloseProps;
exports.getLastVisible = getLastVisible;
exports.getMenuContainerFocusMethods = getMenuContainerFocusMethods;
exports.getMenuContainerMouseLeave = getMenuContainerMouseLeave;
exports.getMenuItemClickChangeKey = getMenuItemClickChangeKey;
exports.getMenuPointerDown = getMenuPointerDown;
exports.getNVisible = getNVisible;
exports.getOpenSubmenuEvent = getOpenSubmenuEvent;
exports.getPrevNext = getPrevNext;
exports.isMenuChildrenFalsy = isMenuChildrenFalsy;
exports.keyElemExtractor = keyElemExtractor;
