// @oracle/oraclejet-preact/cjs/GroupLoadingIndicator-7e035ba2.js
/* @oracle/oraclejet-preact: undefined */
'use strict';

var jsxRuntime = require('preact/jsx-runtime');
var hooks = require('preact/hooks');
var classNames = require('./classNames-c14c6ef3.js');
var keys = require('./keys-25710458.js');
var useId = require('./useId-6c0eeb27.js');
var TabbableModeContext = require('./TabbableModeContext-e99d527e.js');
var useComponentTheme = require('./useComponentTheme-082fc8e4.js');
var useInteractionStyle = require('./useInteractionStyle-c203a8a0.js');
var mergeInterpolations = require('./mergeInterpolations-6727b536.js');
var flexitem = require('./flexitem-fee13e26.js');
var PRIVATE_List_themes_ListStyles_css = require('./PRIVATE_List/themes/ListStyles.css.js');
var UNSAFE_GroupedList_themes_redwood_GroupedListTheme = require('./UNSAFE_GroupedList/themes/redwood/GroupedListTheme.js');
var Flex = require('./Flex-3a257bc3.js');
var Skeleton = require('./Skeleton-8e35c7d1.js');
var List = require('./List-afcfa338.js');

// default style class for items in ListView
const GROUP_SELECTOR = '[data-oj-listview-group]';
const STICKY_SELECTOR = '[data-oj-listview-sticky]';
/**
 * The internal component used to render a group header in ExpandableList and GroupedList.
 */
function ListGroupHeader({ ariaDescribedBy, children, itemKey, itemIndex, itemDepth, isFocused, isFocusRingVisible, isActive, isGridlineVisible, isExpandable, expandedKeys, viewportConfig }) {
    const [isStuck, setStuck] = hooks.useState(false);
    const { classes: groupedListThemeClasses, styles: groupedListStyles } = useComponentTheme.useComponentTheme(UNSAFE_GroupedList_themes_redwood_GroupedListTheme.GroupedListRedwoodTheme);
    const { interactionProps, applyHoverStyle, applyPseudoHoverStyle, applyActiveStyle } = useInteractionStyle.useInteractionStyle();
    const variantClasses = PRIVATE_List_themes_ListStyles_css.listGroupHeaderMultiVariantStyles({
        expandable: isExpandable ? 'isExpandable' : 'notExpandable',
        needsEventsHover: applyHoverStyle ? 'isNeedsEventsHover' : 'notNeedsEventsHover',
        pseudoHover: applyPseudoHoverStyle ? 'isPseudoHover' : 'notPseudoHover',
        active: applyActiveStyle || isActive ? 'isActive' : 'notActive',
        focusRingVisible: isFocusRingVisible && isFocused ? 'isFocusRingVisible' : 'notFocusRingVisible',
        gridlineTop: itemIndex > 0 ? 'visible' : 'hidden',
        gridlineBottom: isGridlineVisible ? 'visible' : 'hidden'
    });
    const baseClasses = [variantClasses];
    if (isStuck) {
        baseClasses.push(groupedListThemeClasses);
        baseClasses.push(groupedListStyles.stuckHeader);
    }
    const classes = classNames.classNames(baseClasses);
    const styleInterpolations = mergeInterpolations.mergeInterpolations([
        ...Object.values(flexitem.flexitemInterpolations)
    ]);
    const { class: cls, ...cellStyles } = styleInterpolations({
        alignSelf: 'center',
        flex: '1 0 auto'
    });
    // create these refs up front to avoid having to update ref.current values directly during render
    // function or setup a useEffect to update if options/values change in a subsequent render.
    const expandableScrollerRef = hooks.useRef(null);
    // rely on provided scrollerRef, not to body on first render.
    const scrollerRef = isExpandable ? expandableScrollerRef : viewportConfig?.scrollerRef ?? null;
    const headerRef = hooks.useRef(null);
    // compute stuck state based on scrollTop and element positions; avoids IntersectionObserver edge cases
    const onScroll = hooks.useCallback(() => {
        const scroller = scrollerRef ? scrollerRef.current : null;
        const headerEl = headerRef.current;
        if (!scroller || !headerEl)
            return;
        let isStuck = false;
        const nextEl = headerEl.nextElementSibling;
        if (nextEl) {
            const headerRect = headerEl.getBoundingClientRect();
            // apply sticky styling if next element has started to scroll under this header
            isStuck = headerRect.bottom > nextEl.getBoundingClientRect().top;
            if (!isStuck) {
                let prevEl = headerEl.previousElementSibling;
                // check for this header's top value matching the previous header's top value - the case where
                // this header has replaced the previous header, but the next item has not yet scrolled under it
                while (prevEl) {
                    if (prevEl.hasAttribute('data-oj-listview-group')) {
                        isStuck = prevEl.getBoundingClientRect().top === headerRect.top;
                        break;
                    }
                    prevEl = prevEl.previousElementSibling;
                }
            }
        }
        setStuck(isStuck);
    }, [scrollerRef]);
    hooks.useEffect(() => {
        const scroller = scrollerRef ? scrollerRef.current : null;
        if (!scroller)
            return;
        onScroll();
        scroller.addEventListener('scroll', onScroll);
        return () => {
            scroller.removeEventListener('scroll', onScroll);
        };
    }, [scrollerRef, onScroll]);
    return (jsxRuntime.jsx("div", { ref: headerRef, id: useId.useId(), role: "row", "aria-rowindex": itemIndex + 1, "aria-level": isNaN(itemDepth) ? undefined : itemDepth + 1, class: classes, "data-oj-key": itemKey, "data-oj-list-item": "", "data-oj-listview-group": "", ...(isStuck && { 'data-oj-listview-sticky': '' }), ...(typeof itemKey === 'number' && { 'data-oj-key-type': 'number' }), ...interactionProps, children: jsxRuntime.jsx("div", { id: useId.useId(), role: "gridcell", "aria-describedby": ariaDescribedBy, "aria-colindex": 1, "aria-expanded": keys.containsKey(expandedKeys, itemKey), style: cellStyles, children: jsxRuntime.jsx(TabbableModeContext.TabbableModeContext.Provider, { value: { isTabbable: false }, children: children }) }) }));
}

const GroupLoadingIndicator = () => (jsxRuntime.jsx(List.SkeletonContainer, { minimumCount: 3, children: (index) => {
        const padding = { paddingLeft: index * 10 + 'px' };
        return (jsxRuntime.jsxs(Flex.Flex, { height: "12x", align: "center", children: [jsxRuntime.jsx("span", { style: padding }), jsxRuntime.jsx(Skeleton.Skeleton, { height: "4x" })] }));
    } }));

exports.GROUP_SELECTOR = GROUP_SELECTOR;
exports.GroupLoadingIndicator = GroupLoadingIndicator;
exports.ListGroupHeader = ListGroupHeader;
exports.STICKY_SELECTOR = STICKY_SELECTOR;
