// @oracle/oraclejet-preact/cjs/PRIVATE_CollapsibleNavigationList.js
/* @oracle/oraclejet-preact: undefined */
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var jsxRuntime = require('preact/jsx-runtime');
var useCurrentKey = require('./useCurrentKey-1b9e3382.js');
var collectionUtils = require('./collectionUtils-f1d3183f.js');
var hooks = require('preact/hooks');
var preact = require('preact');
var Expander = require('./Expander-f4c574e5.js');
var Flex = require('./Flex-3a257bc3.js');
var mergeProps = require('./mergeProps-e3da7237.js');
var usePress = require('./usePress-886180e4.js');
var Text = require('./Text-056f1ec5.js');
var Badge = require('./Badge-0e522f12.js');
var WarningS = require('./WarningS-cdb0a7ad.js');
var ErrorS = require('./ErrorS-b027f7ad.js');
var InformationS = require('./InformationS-b0bb102b.js');
var SuccessS = require('./SuccessS-b6757c64.js');
var useComponentTheme = require('./useComponentTheme-082fc8e4.js');
var PRIVATE_CollapsibleNavigationList_themes_CollapsibleNavigationListItemContract_css = require('./PRIVATE_CollapsibleNavigationList/themes/CollapsibleNavigationListItemContract.css.js');
var PRIVATE_CollapsibleNavigationList_themes_redwood_CollapsibleNavigationListItemTheme = require('./PRIVATE_CollapsibleNavigationList/themes/redwood/CollapsibleNavigationListItemTheme.js');
var useInteractionStyle = require('./useInteractionStyle-c203a8a0.js');
var useUser = require('./useUser-1295b5cb.js');
var classNames = require('./classNames-c14c6ef3.js');
var mergeInterpolations = require('./mergeInterpolations-6727b536.js');
var flexitem = require('./flexitem-fee13e26.js');
var useCollectionFocusRing = require('./useCollectionFocusRing-bed0aa16.js');
var PRIVATE_CollapsibleNavigationList_themes_CollapsibleNavigationListStyles_css = require('./PRIVATE_CollapsibleNavigationList/themes/CollapsibleNavigationListStyles.css.js');
require('./keys-25710458.js');
require('./clientHints-c40c5250.js');
require('./Selector-6550b8e0.js');
require('./BaseSelector-2dc69917.js');
require('./useActive-6770f917.js');
require('./useToggle-3ebba7d8.js');
require('./useFocus-1b288fb9.js');
require('./useTestId-4f53d561.js');
require('./LayerHost-3cd73603.js');
require('./index-0aa99b92.js');
require('preact/compat');
require('./UNSAFE_Layer/themes/LayerHostStyles.css.js');

require('./useHover-49b0430d.js');
require('./StyledCheckbox-9047ed7d.js');
require('./Check-1272bb86.js');
require('./Icon-56789beb.js');
require('./size-1914356b.js');
require('./utils-c0efcd5e.js');
require('./themeContract.css-f3d431c6.js');
require('./colorUtils-01f3caa2.js');
require('./_curry1-e8f0d7ea.js');
require('./useTooltip-34accecc.js');
require('./useTooltipControlled-0a95856c.js');
require('./useId-6c0eeb27.js');
require('./useTouch-4dec8729.js');
require('./useAnimation-65c0e455.js');
require('./useThemeInterpolations-a37b37d9.js');
require('./useColorScheme-f8b1447d.js');
require('./useScale-db16c2b8.js');
require('./theme-dfd6a9fe.js');
require('./Theme-f06687af.js');
require('./useDensity-0b9fd9ca.js');
require('./unsafeDomAccess-c3a492f0.js');
require('./hooks/UNSAFE_useTooltip/themes/redwood/TooltipContentTheme.js');
require('./hooks/UNSAFE_useTooltip/themes/TooltipContentStyles.css.js');

require('./hooks/UNSAFE_useTooltip/themes/redwood/TooltipContentBaseTheme.css.js');

require('./hooks/UNSAFE_useTooltip/themes/redwood/TooltipContentVariants.css.js');

require('./vanilla-extract-recipes-createRuntimeFn.esm-d1301b2d.js');
require('./tooltipUtils-2bc15136.js');
require('./EnvironmentProvider-5f2435e6.js');
require('./LayerManager-c535d963.js');
require('./Floating-79cbe5ad.js');
require('./useFloating-c753a7f9.js');
require('./positionUtils-a780137e.js');
require('./refUtils-a9872e75.js');
require('./useOutsideClick-dbf3d604.js');
require('./arrayUtils-7d8dcfc3.js');
require('./Layer-b60deb1b.js');
require('./UNSAFE_Layer/themes/LayerStyles.css.js');

require('./useModal-0739a528.js');
require('./UNSAFE_Floating/themes/redwood/FloatingTheme.js');
require('./UNSAFE_Floating/themes/FloatingStyles.css.js');

require('./UNSAFE_Floating/themes/redwood/FloatingBaseTheme.css.js');

require('./UNSAFE_Floating/themes/redwood/FloatingVariants.css.js');
require('./vanilla-extract-dynamic.esm-2e1b7c25.js');
require('./UNSAFE_Floating/themes/FloatingContract.css.js');
require('./useUnsafeDomElementRef-f22a2f0c.js');
require('./logger-2b636482.js');
require('./UNSAFE_Icon/themes/IconStyle.css.js');

require('./CheckboxOff-a3ed334d.js');
require('./CheckboxOn-b5dda28b.js');
require('./CheckboxMixed-a17c8298.js');
require('./HiddenAccessible-087eb49a.js');

require('./UNSAFE_Checkbox/themes/CheckboxIconContract.css.js');
require('./UNSAFE_Checkbox/themes/redwood/CheckboxIconTheme.js');
require('./UNSAFE_Checkbox/themes/CheckboxIconStyles.css.js');

require('./UNSAFE_Checkbox/themes/redwood/CheckboxIconBaseTheme.css.js');

require('./UNSAFE_Checkbox/themes/redwood/CheckboxIconVariants.css.js');

require('./TabbableModeContext-e99d527e.js');
require('./useTabbableMode-dbd254e8.js');

require('./PRIVATE_BaseSelector/themes/SelectorStyles.css.js');

require('./CollectionInteractionContext-f84ff7a1.js');
require('./useCollectionInteractionContext-6fc0200e.js');
require('./EmbeddedIconButton-cc159a01.js');
require('./BaseButton-d8f96bfa.js');
require('./dimensions-6101ca1f.js');
require('./UNSAFE_BaseButton/themes/redwood/BaseButtonTheme.js');
require('./UNSAFE_BaseButton/themes/BaseButtonStyles.css.js');

require('./UNSAFE_BaseButton/themes/redwood/BaseButtonVariants.css.js');

require('./BareButton-e8a030d7.js');
require('./useBareButton-463210c9.js');
require('./usePressClick-7a34624a.js');
require('./buttonUtils-dff2ea3e.js');
require('./id-86356250.js');
require('./ButtonLabelLayout-c876923f.js');
require('./useButtonLabelLayoutTheme-d09db0c9.js');
require('./UNSAFE_ButtonLabelLayout/themes/redwood/ButtonLabelLayoutTheme.js');
require('./UNSAFE_ButtonLabelLayout/themes/ButtonLabelLayoutStyles.css.js');

require('./UNSAFE_ButtonLabelLayout/themes/redwood/ButtonLabelLayoutBaseTheme.css.js');

require('./UNSAFE_ButtonLabelLayout/themes/redwood/ButtonLabelLayoutVariants.css.js');

require('./ChevronDown-55b2c0d4.js');
require('./ChevronRight-d1c46e18.js');
require('./boxalignment-e287d370.js');

require('./vanilla-extract-sprinkles-createRuntimeSprinkles.esm-d68f3e0f.js');
require('./flexbox-2cae9a01.js');


require('./UNSAFE_Text/themes/TextStyles.css.js');

require('./UNSAFE_Badge/themes/redwood/BadgeTheme.js');
require('./UNSAFE_Badge/themes/BadgeStyles.css.js');

require('./UNSAFE_Badge/themes/redwood/BadgeThemeBaseTheme.css.js');

require('./UNSAFE_Badge/themes/redwood/BadgeVariants.css.js');

require('./PRIVATE_CollapsibleNavigationList/themes/NavigationListItemStyles.css.js');

require('./PRIVATE_CollapsibleNavigationList/themes/redwood/CollapsibleNavigationListItemBaseTheme.css.js');

require('./PRIVATE_CollapsibleNavigationList/themes/redwood/CollapsibleNavigationListItemVariants.css.js');

require('./mergeDeepWithKey-210b024d.js');
require('./_curry3-18677bca.js');
require('./_curry2-c15d89cd.js');
require('./_isObject-28636267.js');



/**
 * @license
 * Copyright (c) %FIRST_YEAR% %CURRENT_YEAR%, Oracle and/or its affiliates.
 * Licensed under The Universal Permissive License (UPL), Version 1.0
 * as shown at https://oss.oracle.com/licenses/upl/
 * @ignore
 */
/**
 * Context used to pass TabBar information without having to pass it to TabBar children props.
 * We want to communicate information down to any interested TabBar item children.
 */
const CollapsibleNavigationListContext = preact.createContext(null);

/**
 * @license
 * Copyright (c) %FIRST_YEAR% %CURRENT_YEAR%, Oracle and/or its affiliates.
 * Licensed under The Universal Permissive License (UPL), Version 1.0
 * as shown at https://oss.oracle.com/licenses/upl/
 * @ignore
 */
/**
 * Utility hook for consuming the CollapsibleNavigationListContext
 *
 * @returns The value of CollapsibleNavigationListContext provider
 */
function useCollapsibleNavigationListContext() {
    return hooks.useContext(CollapsibleNavigationListContext);
}

const CollapsibleNavigationListItem = ({ item, level }) => {
    const isGroup = Array.isArray(item.children) && item.children.length > 0;
    const itemKey = item.itemKey;
    const label = item.label;
    const { selection, onToggle, onSelect, expandedKeys, showFocusRing, currentKey, navigationListItemPrefix } = useCollapsibleNavigationListContext();
    const isExpanded = isGroup && expandedKeys?.has(item.itemKey);
    const handlePress = hooks.useCallback(() => {
        onSelect?.({ value: item.itemKey });
        if (isGroup) {
            onToggle?.({ value: item.itemKey });
        }
    }, [itemKey, onSelect, onToggle, isGroup]);
    const { pressProps } = usePress.usePress(handlePress);
    const isCurrent = currentKey === itemKey;
    const isSelected = selection === itemKey;
    const { interactionProps, applyPseudoHoverStyle, applyHoverStyle, applyActiveStyle } = useInteractionStyle.useInteractionStyle();
    const itemHandlers = { ...mergeProps.mergeProps(pressProps, interactionProps) };
    const itemId = navigationListItemPrefix + itemKey;
    const severityRequirement = !(item.severity === 'none' || item.severity === undefined);
    const isComplexItem = item.metadata || severityRequirement || item.badge;
    //Styling
    const flexItemStyleInterpolations = mergeInterpolations.mergeInterpolations([
        ...Object.values(flexitem.flexitemInterpolations)
    ]);
    const { class: labelContainerFlexItemClasses, ...labelContainerFlexItemStyle } = flexItemStyleInterpolations({
        alignSelf: 'center'
    });
    const { direction } = useUser.useUser();
    const isRtl = direction === 'rtl';
    const { classes: listItemClasses, styles: { labelContainerClasses, iconLabelContainer, iconContainer, navigationItemMetadata }, baseTheme } = useComponentTheme.useComponentTheme(PRIVATE_CollapsibleNavigationList_themes_redwood_CollapsibleNavigationListItemTheme.NavigationListItemRedwoodTheme, {
        focusRing: showFocusRing ? 'isFocusRing' : 'notFocusRing',
        current: isCurrent ? 'isCurrent' : 'notCurrent',
        selected: isSelected ? 'isSelected' : 'notSelected',
        direction: isRtl ? 'isRtl' : 'notRtl',
        hybridHover: applyHoverStyle ? 'isHybridHover' : 'notHybridHover',
        pseudoHover: applyPseudoHoverStyle ? 'isPseudoHover' : 'notPseudoHover',
        active: applyActiveStyle ? 'isActive' : 'notActive'
    });
    const labelContainerStyle = labelContainerFlexItemStyle;
    const itemClasses = classNames.classNames([listItemClasses]);
    const navigationItemMatadata = classNames.classNames([navigationItemMetadata]);
    // const navigationItemLinkClass = classNames([navigationListItemLink]);
    const endIconContainerClasses = classNames.classNames([iconLabelContainer, iconContainer]);
    const indentation = calcIndentation(level);
    const IconComponent = (severityValue) => {
        switch (severityValue) {
            case 'error':
                return jsxRuntime.jsx(ErrorS.SvgErrorS, { color: "danger" });
            case 'warning':
                return jsxRuntime.jsx(WarningS.SvgWarningS, { color: "warning" });
            case 'info':
                return jsxRuntime.jsx(InformationS.SvgInformationS, { color: "info" });
            case 'confirmation':
                return jsxRuntime.jsx(SuccessS.SvgSuccessS, { color: "success" });
            default:
                return null;
        }
    };
    const labelRenderer = () => {
        return (jsxRuntime.jsx("span", { class: labelContainerClasses, style: labelContainerStyle, children: jsxRuntime.jsx(Text.Text, { size: "inherit", weight: "inherit", variant: "inherit", children: label }) }));
    };
    const complexItemRenderer = () => {
        return (jsxRuntime.jsxs(jsxRuntime.Fragment, { children: [label && (jsxRuntime.jsx("span", { class: labelContainerClasses, style: labelContainerStyle, children: jsxRuntime.jsx(Text.Text, { size: "inherit", weight: "inherit", variant: "inherit", children: label }) })), (item.metadata || severityRequirement || item.badge) && (jsxRuntime.jsxs(Flex.Flex, { gap: "2x", align: "center", justify: "end", flex: "1 1 auto", children: [' ', item.badge && (jsxRuntime.jsx("span", { class: endIconContainerClasses, children: jsxRuntime.jsx(Badge.Badge, { size: "md", variant: "neutralSubtle", children: item.badge }) })), item.metadata && (jsxRuntime.jsx("span", { class: navigationItemMatadata, children: jsxRuntime.jsx(Text.Text, { children: '( ' + item.metadata + ' )' }) })), item.severityRequirement && (jsxRuntime.jsx("span", { class: endIconContainerClasses, children: IconComponent(item.severity) }))] }))] }));
    };
    const navlistItemRenderer = () => {
        return (jsxRuntime.jsx("div", { role: "treeitem", ...itemHandlers, "aria-expanded": isGroup ? isExpanded : undefined, "data-oj-key": itemKey, id: itemId, class: itemClasses, "aria-selected": isSelected, children: jsxRuntime.jsx("div", { style: { marginInlineStart: indentation, display: 'flex', alignItems: 'center' }, children: jsxRuntime.jsxs(Flex.Flex, { direction: "row", gap: "2x", align: "center", justify: "center", children: [isComplexItem ? complexItemRenderer() : labelRenderer(), isGroup && jsxRuntime.jsx(Expander.Expander, { itemKey: itemKey, expanded: isExpanded, onToggle: onToggle })] }) }) }));
    };
    return (jsxRuntime.jsxs("div", { class: baseTheme, children: [navlistItemRenderer(), isGroup && isExpanded && (jsxRuntime.jsx("div", { role: "group", children: item.children.map((child) => (jsxRuntime.jsx(CollapsibleNavigationListItem, { itemKey: child.itemKey, label: child.label, item: child, expandedKeys: expandedKeys, level: level + 1, onToggle: onToggle }))) }))] }));
};
const calcIndentation = (depth) => {
    return `calc(${Math.max(0, depth)} * ${PRIVATE_CollapsibleNavigationList_themes_CollapsibleNavigationListItemContract_css.childItemVars.childItemPadding} )`;
};

const CollapsibleNavigationList = ({ items, selection, expandedKeys, edge, onToggle, onSelect }) => {
    const groupKeys = hooks.useMemo(() => getGroupKeys(items), [items]);
    const rootRef = hooks.useRef(null);
    const [showFocusRing, focusRingProps] = useCollectionFocusRing.useCollectionFocusRing(rootRef, [
        'ArrowUp',
        'ArrowDown',
        'Home',
        'End'
    ]);
    const onFocus = () => {
        if (rootRef.current && currentKey === undefined) {
            const key = collectionUtils.getFirstVisibleKey(rootRef.current, ITEM_SELECTOR);
            if (key != null) {
                setCurrentKey(key);
            }
        }
    };
    const onKeyDown = (event) => {
        if (rootRef.current && currentKey != null) {
            if (event.key === 'Home' || event.key === 'End') {
                const tabBarItemKey = collectionUtils.getItemKeysArray(rootRef.current, ITEM_SELECTOR);
                event.preventDefault();
                setCurrentKey?.(tabBarItemKey[event.key === 'Home' ? 0 : tabBarItemKey.length - 1]);
            }
            //Enter or space expands a collapsed node, collapses an expanded node and also triggers onSelect
            if (event.key === 'Enter' || event.key === ' ') {
                event.preventDefault();
                onSelect?.({ value: currentKey });
                if (groupKeys.has(currentKey)) {
                    onToggle?.({ value: currentKey });
                }
            }
            //ArrowRight only expands, if detected on a previously expanded node it remains expanded
            if (event.key === 'ArrowRight') {
                if (groupKeys.has(currentKey) &&
                    !(expandedKeys && expandedKeys.has(currentKey))) {
                    onToggle?.({ value: currentKey });
                }
            }
            //ArrowLeft only collapses, if detected on a previously collapsed node it remains collapsed
            if (event.key === 'ArrowLeft') {
                if (groupKeys.has(currentKey) && expandedKeys && expandedKeys.has(currentKey)) {
                    onToggle?.({ value: currentKey });
                }
            }
        }
    };
    const [currentKey, setCurrentKey] = hooks.useState(selection);
    const ITEM_SELECTOR = '[role="treeitem"]';
    const { currentKeyProps } = useCurrentKey.useCurrentKey((element) => {
        const key = collectionUtils.keyExtractor(element, ITEM_SELECTOR);
        return key;
    }, false, false, collectionUtils.getPrevNextKeyUsingRef(rootRef, currentKey, true, ITEM_SELECTOR), collectionUtils.getPrevNextKeyUsingRef(rootRef, currentKey, false, ITEM_SELECTOR), undefined, undefined, currentKey, (detail) => {
        const key = detail.value;
        setCurrentKey(key);
    });
    const navListId = hooks.useId();
    const navigationListItemPrefix = navListId + '_';
    const collapsibleNavigationListContext = {
        selection: selection,
        onSelect: onSelect,
        onToggle: onToggle,
        expandedKeys,
        edge,
        showFocusRing,
        currentKey: currentKey,
        navigationListItemPrefix
    };
    hooks.useEffect(() => {
        if (rootRef.current) {
            const itemKeys = collectionUtils.getItemKeysArray(rootRef.current, ITEM_SELECTOR);
            if (currentKey == null && itemKeys) {
                if (selection != null && itemKeys.includes(selection)) {
                    setCurrentKey(selection);
                }
                else if (itemKeys.length > 0) {
                    setCurrentKey(itemKeys[0]);
                }
            }
        }
    }, [items, currentKey, rootRef, selection]);
    const eventProps = { onKeyDown, onFocus };
    //Styling
    const navigationListClass = classNames.classNames([PRIVATE_CollapsibleNavigationList_themes_CollapsibleNavigationListStyles_css.navigationListStyles.uListStyle]);
    return (jsxRuntime.jsx(CollapsibleNavigationListContext.Provider, { value: collapsibleNavigationListContext, children: jsxRuntime.jsx("div", { role: "tree", class: navigationListClass, ref: rootRef, ...mergeProps.mergeProps(currentKeyProps, focusRingProps, eventProps), tabindex: 0, "aria-activedescendant": currentKey ? navigationListItemPrefix : '', children: items.map((item) => (jsxRuntime.jsx(CollapsibleNavigationListItem, { itemKey: item.itemKey, item: item, expandedKeys: expandedKeys, level: 0, label: item.label, onToggle: onToggle }))) }) }));
};
// Helper to extract all group item keys from items tree
const getGroupKeys = (items) => {
    const groupKeys = new Set();
    const visited = new Set();
    const keysArray = [...items];
    while (keysArray.length) {
        const item = keysArray.pop();
        if (visited.has(item?.itemKey))
            continue;
        visited.add(item?.itemKey);
        if (Array.isArray(item?.children) && item?.children.length > 0) {
            groupKeys.add(item.itemKey);
            keysArray.push(...item.children);
        }
    }
    return groupKeys;
};

exports.CollapsibleNavigationList = CollapsibleNavigationList;
