// @oracle/oraclejet-preact/cjs/PRIVATE_TableDisplay.js
/* @oracle/oraclejet-preact: undefined */
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var jsxRuntime = require('preact/jsx-runtime');
var compat = require('preact/compat');
var preact = require('preact');
var hooks = require('preact/hooks');
var classNames = require('./classNames-c14c6ef3.js');
var useId = require('./useId-6c0eeb27.js');
var useUser = require('./useUser-1295b5cb.js');
var TableColGroup = require('./TableColGroup-b7e1e296.js');
var LoadMoreCollection = require('./LoadMoreCollection-eb1f2603.js');
var PRIVATE_Table_themes_TableRowStyles_css = require('./PRIVATE_Table/themes/TableRowStyles.css.js');
var FormContext = require('./FormContext-d1d05967.js');
var clientHints = require('./clientHints-c40c5250.js');
var PRIVATE_Table_themes_Table_css = require('./PRIVATE_Table/themes/Table.css.js');
var PRIVATE_Table_themes_redwood_TableNoDataTheme = require('./PRIVATE_Table/themes/redwood/TableNoDataTheme.js');
var useComponentTheme = require('./useComponentTheme-082fc8e4.js');
var useTranslationBundle = require('./useTranslationBundle-12dd087d.js');
var PRIVATE_TableCommon_themes_Common_css = require('./PRIVATE_TableCommon/themes/Common.css.js');
var PRIVATE_Table_themes_TableHeaderCellStyles_css = require('./PRIVATE_Table/themes/TableHeaderCellStyles.css.js');
var SortControl = require('./SortControl-8c15705b.js');
var useSortableContext = require('./useSortableContext-0510fd8e.js');
var ScrollUtils = require('./ScrollUtils-e8076071.js');
var PRIVATE_Table_themes_TableSkeletonStyles_css = require('./PRIVATE_Table/themes/TableSkeletonStyles.css.js');
var PRIVATE_Table_themes_redwood_TableTheme = require('./PRIVATE_Table/themes/redwood/TableTheme.js');
var useTestId = require('./useTestId-4f53d561.js');
var PRIVATE_TableDisplay_themes_TableDisplayStyles_css = require('./PRIVATE_TableDisplay/themes/TableDisplayStyles.css.js');
var dimensions = require('./dimensions-6101ca1f.js');
var mergeInterpolations = require('./mergeInterpolations-6727b536.js');
require('./LayerHost-3cd73603.js');
require('./index-0aa99b92.js');
require('./UNSAFE_Layer/themes/LayerHostStyles.css.js');

require('./useResizeObserver-a9e97180.js');
require('./useAddBusyState-d19ae1fa.js');
require('./BusyStateContext-86f40d3c.js');
require('./useBusyStateContext-362eee2a.js');
require('./unsafeDomAccess-c3a492f0.js');
require('./collectionUtils-f1d3183f.js');
require('./keys-25710458.js');
require('./Selector-6550b8e0.js');
require('./BaseSelector-2dc69917.js');
require('./useActive-6770f917.js');
require('./useToggle-3ebba7d8.js');
require('./useFocus-1b288fb9.js');
require('./useHover-49b0430d.js');
require('./mergeProps-e3da7237.js');
require('./StyledCheckbox-9047ed7d.js');
require('./Check-1272bb86.js');
require('./Icon-56789beb.js');
require('./size-1914356b.js');
require('./utils-c0efcd5e.js');
require('./themeContract.css-f3d431c6.js');
require('./colorUtils-01f3caa2.js');
require('./_curry1-e8f0d7ea.js');
require('./useTooltip-34accecc.js');
require('./useTooltipControlled-0a95856c.js');
require('./useTouch-4dec8729.js');
require('./useAnimation-65c0e455.js');
require('./useThemeInterpolations-a37b37d9.js');
require('./useColorScheme-f8b1447d.js');
require('./useScale-db16c2b8.js');
require('./theme-dfd6a9fe.js');
require('./Theme-f06687af.js');
require('./useDensity-0b9fd9ca.js');
require('./hooks/UNSAFE_useTooltip/themes/redwood/TooltipContentTheme.js');
require('./hooks/UNSAFE_useTooltip/themes/TooltipContentStyles.css.js');

require('./hooks/UNSAFE_useTooltip/themes/redwood/TooltipContentBaseTheme.css.js');

require('./hooks/UNSAFE_useTooltip/themes/redwood/TooltipContentVariants.css.js');

require('./vanilla-extract-recipes-createRuntimeFn.esm-d1301b2d.js');
require('./tooltipUtils-2bc15136.js');
require('./EnvironmentProvider-5f2435e6.js');
require('./LayerManager-c535d963.js');
require('./Floating-79cbe5ad.js');
require('./useFloating-c753a7f9.js');
require('./positionUtils-a780137e.js');
require('./refUtils-a9872e75.js');
require('./useOutsideClick-dbf3d604.js');
require('./arrayUtils-7d8dcfc3.js');
require('./Layer-b60deb1b.js');
require('./UNSAFE_Layer/themes/LayerStyles.css.js');

require('./useModal-0739a528.js');
require('./UNSAFE_Floating/themes/redwood/FloatingTheme.js');
require('./UNSAFE_Floating/themes/FloatingStyles.css.js');

require('./UNSAFE_Floating/themes/redwood/FloatingBaseTheme.css.js');

require('./UNSAFE_Floating/themes/redwood/FloatingVariants.css.js');
require('./vanilla-extract-dynamic.esm-2e1b7c25.js');
require('./UNSAFE_Floating/themes/FloatingContract.css.js');
require('./useUnsafeDomElementRef-f22a2f0c.js');
require('./logger-2b636482.js');
require('./UNSAFE_Icon/themes/IconStyle.css.js');

require('./CheckboxOff-a3ed334d.js');
require('./CheckboxOn-b5dda28b.js');
require('./CheckboxMixed-a17c8298.js');
require('./HiddenAccessible-087eb49a.js');

require('./UNSAFE_Checkbox/themes/CheckboxIconContract.css.js');
require('./UNSAFE_Checkbox/themes/redwood/CheckboxIconTheme.js');
require('./UNSAFE_Checkbox/themes/CheckboxIconStyles.css.js');

require('./UNSAFE_Checkbox/themes/redwood/CheckboxIconBaseTheme.css.js');

require('./UNSAFE_Checkbox/themes/redwood/CheckboxIconVariants.css.js');

require('./TabbableModeContext-e99d527e.js');
require('./useTabbableMode-dbd254e8.js');

require('./PRIVATE_BaseSelector/themes/SelectorStyles.css.js');

require('./CollectionInteractionContext-f84ff7a1.js');
require('./useCollectionInteractionContext-6fc0200e.js');
require('./PRIVATE_Table/themes/TableCellStyles.css.js');




require('./PRIVATE_Table/themes/TableFooterCellStyles.css.js');

require('./Collection-e27077ff.js');
require('./useViewportIntersect-2e802bac.js');
require('./PRIVATE_Table/themes/TableNoDataStyles.css.js');

require('./PRIVATE_Table/themes/redwood/TableNoDataVariants.css.js');

require('./UNSAFE_SortControl/themes/SortControl.css.js');

require('./IconButton-7503293b.js');
require('./BaseButton-d8f96bfa.js');
require('./UNSAFE_BaseButton/themes/redwood/BaseButtonTheme.js');
require('./UNSAFE_BaseButton/themes/BaseButtonStyles.css.js');

require('./UNSAFE_BaseButton/themes/redwood/BaseButtonVariants.css.js');

require('./BareButton-e8a030d7.js');
require('./useBareButton-463210c9.js');
require('./usePress-886180e4.js');
require('./usePressClick-7a34624a.js');
require('./buttonUtils-dff2ea3e.js');
require('./id-86356250.js');
require('./ButtonLabelLayout-c876923f.js');
require('./Text-056f1ec5.js');
require('./UNSAFE_Text/themes/TextStyles.css.js');

require('./useButtonLabelLayoutTheme-d09db0c9.js');
require('./UNSAFE_ButtonLabelLayout/themes/redwood/ButtonLabelLayoutTheme.js');
require('./UNSAFE_ButtonLabelLayout/themes/ButtonLabelLayoutStyles.css.js');

require('./UNSAFE_ButtonLabelLayout/themes/redwood/ButtonLabelLayoutBaseTheme.css.js');

require('./UNSAFE_ButtonLabelLayout/themes/redwood/ButtonLabelLayoutVariants.css.js');

require('./EmbeddedIconButton-cc159a01.js');
require('./UpParent-58bbd538.js');
require('./DownParent-104857eb.js');

require('./PRIVATE_Table/themes/TableStyles.css.js');

require('./PRIVATE_Table/themes/redwood/TableBaseTheme.css.js');

require('./PRIVATE_Table/themes/redwood/TableVariants.css.js');

require('./mergeDeepWithKey-210b024d.js');
require('./_curry3-18677bca.js');
require('./_curry2-c15d89cd.js');
require('./_isObject-28636267.js');

function _defaultCellRenderer({ rowData, field }) {
    return field != null ? String(rowData[field]) : undefined;
}
/**
 * The internal component used to render a single cell in a TableDisplayRow.
 */
function TableDisplayCell({ tableId, rowData, field, columnKey, columnIndex, rowKey, rowIndex, isFinalRowIndex, isFinalColumnIndex, isFirstLogicalColumnIndex, isFinalLogicalColumnIndex, hasVerticalGridlines, hasHorizontalGridlines, hasFooters, hasVerticalUnderflow, isSticky, isAccRowHeader, renderer = _defaultCellRenderer, isStickyStartEdge, isStickyEndEdge, hasStartEdge, hasEndEdge, isRtl, isHighContrast, padding, horizontalAlignment = 'start', verticalAlignment = 'center' }) {
    const rowContext = { key: rowKey, data: rowData };
    const columnPadding = typeof padding === 'function' ? padding(rowContext) : padding;
    const hasLastHorizontalGridline = hasHorizontalGridlines && (hasVerticalUnderflow || !hasFooters);
    const cellClasses = classNames.classNames(TableColGroup.getDataCellClassArray({
        isFirstColumnIndex: columnIndex === 0 || isFirstLogicalColumnIndex,
        isFinalColumnIndex: isFinalColumnIndex || isFinalLogicalColumnIndex,
        isFirstRowIndex: rowIndex === 0,
        isFinalRowIndex,
        hasVerticalGridlines,
        hasHorizontalGridlines,
        hasLastHorizontalGridline,
        isSticky,
        hasStartEdge,
        hasEndEdge,
        isStickyStartEdge,
        isStickyEndEdge,
        isRtl,
        isHighContrast,
        horizontalAlignment,
        verticalAlignment,
        padding: columnPadding
    }));
    const cellRendererProps = {
        rowData: rowData,
        rowKey: rowKey,
        field: field,
        columnKey: columnKey
    };
    return (jsxRuntime.jsx("td", { "aria-colindex": columnIndex + 1, class: cellClasses, role: isAccRowHeader ? 'rowheader' : 'cell', "data-oj-cell-type": 'data', "data-oj-column-key": columnKey, "data-oj-table-focusable": tableId, "data-oj-table-data": tableId, children: jsxRuntime.jsx(FormContext.FormContext.Provider, { value: { textAlign: TableColGroup.getFormTextAlign(horizontalAlignment, isRtl) }, children: renderer(cellRendererProps) }) }));
}

/**
 * The internal component used to render a single row in Table.
 */
function TableDisplayRow({ tableId, rowKey, rowData, rowIndex, isFinalRow, columnsArray, rowHeaderColumnKeys, hasVerticalGridlines, hasHorizontalGridlines, hasFooters, hasVerticalUnderflow, startStickyEdge, endStickyEdge, firstStickyColumnIndex, finalStickyColumnIndex, columnBorderEdges, isRtl, isHighContrast }) {
    const rowStyles = isHighContrast ? PRIVATE_Table_themes_TableRowStyles_css.tableRowStylesHC : PRIVATE_Table_themes_TableRowStyles_css.tableRowStyles;
    const classArray = [rowStyles.base, isFinalRow ? rowStyles.lastRowHeight : rowStyles.rowHeight];
    const classes = classNames.classNames(classArray);
    return (jsxRuntime.jsx("tr", { "aria-rowindex": rowIndex + 2, class: classes, role: 'row', "data-oj-key": rowKey, "data-oj-table-row": tableId, ...(typeof rowKey === 'number' && { 'data-oj-key-type': 'number' }), children: columnsArray.map((column, columnIndex) => {
            const isSticky = column.value.sticky === 'enabled';
            let isFirstLogicalColumnIndex = false;
            let isFinalLogicalColumnIndex = false;
            if (isSticky) {
                if (startStickyEdge != null && firstStickyColumnIndex === columnIndex) {
                    isFirstLogicalColumnIndex = true;
                }
                if (endStickyEdge != null && finalStickyColumnIndex === columnIndex) {
                    isFinalLogicalColumnIndex = true;
                }
            }
            const cellProps = {
                tableId: tableId,
                rowData: rowData,
                field: column.value.field,
                stickyColumn: column.value.sticky,
                columnKey: column.key,
                columnIndex: columnIndex,
                rowKey: rowKey,
                rowIndex: rowIndex,
                isFinalRowIndex: isFinalRow,
                isFinalColumnIndex: columnIndex === columnsArray.length - 1,
                isFirstLogicalColumnIndex: isFirstLogicalColumnIndex,
                isFinalLogicalColumnIndex: isFinalLogicalColumnIndex,
                hasVerticalGridlines: hasVerticalGridlines,
                hasHorizontalGridlines: hasHorizontalGridlines,
                hasFooters: hasFooters,
                hasVerticalUnderflow: hasVerticalUnderflow,
                isSticky: isSticky,
                isAccRowHeader: rowHeaderColumnKeys.has(column.key),
                renderer: column.value.renderer,
                isStickyStartEdge: columnIndex === startStickyEdge,
                isStickyEndEdge: columnIndex === endStickyEdge,
                hasStartEdge: columnBorderEdges[columnIndex].hasStart,
                hasEndEdge: columnBorderEdges[columnIndex].hasEnd,
                isRtl: isRtl,
                isHighContrast: isHighContrast,
                padding: column.value.padding,
                horizontalAlignment: column.value.alignment?.horizontal != null
                    ? column.value.alignment.horizontal
                    : column.value.horizontalAlignment,
                verticalAlignment: column.value.alignment?.vertical
            };
            return jsxRuntime.jsx(TableDisplayCell, { ...cellProps });
        }) }));
}
const _areStringSetsEquivalent = (set1, set2) => {
    if (set1 == null) {
        return set2 == null;
    }
    if (set2 == null) {
        return false;
    }
    return set1.size === set2.size && [...set1].every((x) => set2.has(x));
};
const _areColumnsArraysEquivalent = (columnsArray1, columnsArray2) => {
    if (columnsArray1.length === columnsArray2.length) {
        for (let i = 0; i < columnsArray1.length; i++) {
            if (columnsArray1[i].key !== columnsArray2[i].key ||
                columnsArray1[i].value !== columnsArray2[i].value) {
                return false;
            }
        }
        return true;
    }
    return false;
};
const MemoizeTableDisplayRow = compat.memo(TableDisplayRow, (prev, next) => {
    if (prev && next) {
        return (prev.tableId === next.tableId &&
            prev.rowKey === next.rowKey &&
            prev.rowData === next.rowData &&
            prev.rowIndex === next.rowIndex &&
            prev.isFinalRow === next.isFinalRow &&
            _areColumnsArraysEquivalent(prev.columnsArray, next.columnsArray) &&
            prev.hasHorizontalGridlines === next.hasHorizontalGridlines &&
            prev.hasVerticalGridlines === next.hasVerticalGridlines &&
            (!next.isFinalRow ||
                (prev.hasFooters === next.hasFooters &&
                    prev.hasVerticalUnderflow === next.hasVerticalUnderflow)) &&
            _areStringSetsEquivalent(prev.rowHeaderColumnKeys, next.rowHeaderColumnKeys) &&
            prev.startStickyEdge === next.startStickyEdge &&
            prev.endStickyEdge === next.endStickyEdge &&
            prev.firstStickyColumnIndex === next.firstStickyColumnIndex &&
            prev.finalStickyColumnIndex === next.finalStickyColumnIndex &&
            prev.columnBorderEdges === next.columnBorderEdges &&
            prev.isRtl === next.isRtl &&
            prev.isHighContrast === next.isHighContrast);
    }
    return false;
});

/**
 * The internal component used to render an empty table cell in a TableDisplayBody.
 */
function TableDisplayNoData({ rootRef, tableId, columnsCount, ariaRowIndex, renderer }) {
    const browser = clientHints.getClientHints().browser;
    const translations = useTranslationBundle.useTranslationBundle('@oracle/oraclejet-preact');
    const rowClasses = classNames.classNames([PRIVATE_Table_themes_TableRowStyles_css.tableRowStyles.base]);
    const { classes: themeClasses } = useComponentTheme.useComponentTheme(PRIVATE_Table_themes_redwood_TableNoDataTheme.TableNoDataRedwoodTheme, {
        defaultRenderer: renderer == null ? 'isDefaultRenderer' : 'notDefaultRenderer',
        showFocusRing: 'notShowFocusRing'
    });
    const cellClasses = classNames.classNames([themeClasses]);
    function defaultContentRenderer() {
        return translations.collection_noData();
    }
    const contentRenderer = renderer != null ? renderer : defaultContentRenderer;
    // in firefox, table element heights are not stretched unless explicitly set in some cases
    hooks.useLayoutEffect(() => {
        if (browser === 'firefox') {
            const rootElement = rootRef.current;
            if (rootElement != null) {
                const tableBodyElement = rootElement.querySelector(`[data-oj-table-body='${tableId}']`);
                const noDataElement = rootElement.querySelector(`[${TableColGroup.TABLE_NODATA_SELECTOR}='${tableId}']`);
                if (tableBodyElement != null && noDataElement != null) {
                    noDataElement.style.height = '';
                    noDataElement.style.height = `${tableBodyElement.getBoundingClientRect().height}px`;
                }
            }
        }
    });
    return (jsxRuntime.jsx("tr", { "aria-rowindex": ariaRowIndex, class: rowClasses, role: 'row', children: jsxRuntime.jsx("td", { "aria-colindex": 1, class: cellClasses, style: browser !== 'firefox' ? 'height: 100%;' : undefined, role: 'cell', colSpan: columnsCount, "data-oj-cell-type": 'noData', "data-oj-table-focusable": tableId, "data-oj-table-nodata": tableId, children: contentRenderer() }) }));
}

/**
 * The internal component used to render a the tbody of a Table.
 */
function TableDisplayBody({ rootRef, viewportInfo, tableId, isValidRender, getAccessibleRowHeaders, tableData, columnsArray, hasVerticalGridlines, hasHorizontalGridlines, hasFooters, hasVerticalUnderflow, hasMore, onLoadMore, noDataRenderer, startStickyEdge, endStickyEdge, firstStickyColumnIndex, finalStickyColumnIndex, columnBorderEdges, isRtl, isHighContrast }) {
    const browser = clientHints.getClientHints().browser;
    // returns a TableRowContext based on ItemContext (added key)
    const getRowContext = hooks.useCallback((context) => {
        return {
            index: context.index,
            data: context.data.data,
            key: context.data.metadata.key
        };
    }, []);
    const collectionChildrenFunc = (context) => {
        const tableRowContext = getRowContext(context);
        const rowKey = tableRowContext.key;
        const rowProps = {
            tableId: tableId,
            columnsArray: columnsArray,
            rowHeaderColumnKeys: getAccessibleRowHeaders(tableRowContext),
            hasVerticalGridlines: hasVerticalGridlines,
            hasHorizontalGridlines: hasHorizontalGridlines,
            hasFooters: hasFooters,
            hasVerticalUnderflow: hasVerticalUnderflow,
            hasMore: hasMore,
            rowIndex: tableRowContext.index,
            rowKey: rowKey,
            rowData: tableRowContext.data,
            isFinalRow: tableRowContext.index === tableData.length - 1,
            startStickyEdge: startStickyEdge,
            endStickyEdge: endStickyEdge,
            firstStickyColumnIndex: firstStickyColumnIndex,
            finalStickyColumnIndex: finalStickyColumnIndex,
            columnBorderEdges: columnBorderEdges,
            isRtl: isRtl,
            isHighContrast: isHighContrast
        };
        return jsxRuntime.jsx(MemoizeTableDisplayRow, { ...rowProps });
    };
    const getNoDataProps = () => {
        return {
            rootRef: rootRef,
            tableId: tableId,
            columnsCount: columnsArray.length,
            ariaRowIndex: isValidRender ? 2 : 1,
            renderer: isValidRender ? noDataRenderer : undefined
        };
    };
    const isShowEmptyLoading = tableData.length === 0 && hasMore;
    const isShowNoData = tableData.length === 0 && !hasMore;
    return (jsxRuntime.jsx("tbody", { class: PRIVATE_Table_themes_Table_css.backgroundStyles.inherit, role: 'presentation', "data-oj-table-body": tableId, children: isShowNoData || !isValidRender ? (jsxRuntime.jsx(TableDisplayNoData, { ...getNoDataProps() })) : (jsxRuntime.jsxs(preact.Fragment, { children: [jsxRuntime.jsx(LoadMoreCollection.LoadMoreCollection, { data: tableData, loadMoreIndicator: jsxRuntime.jsx(TableColGroup.TableSkeletonRow, { tableId: tableId, hasTabIndex: false, colspan: columnsArray.length, isLoadMore: !isShowEmptyLoading, isShowFocusRing: false, isEmbedded: false, isHighContrast: isHighContrast }), hasMore: hasMore, onLoadMore: onLoadMore, viewportConfig: viewportInfo.viewportConfig, children: collectionChildrenFunc }), jsxRuntime.jsx("tr", { style: browser !== 'firefox' ? 'height: 100%;' : undefined, role: 'presentation', "data-oj-table-spacer": tableId })] })) }));
}

/**
 * The internal component used to render a single cell in a TableDisplayHeaderRow.
 */
function TableDisplayHeaderCell({ tableId, columnKey, columnIndex, isFinalColumnIndex, isFirstLogicalColumnIndex, isFinalLogicalColumnIndex, hasVerticalGridlines, isSticky, headerText, stickyTopOffset = 0, renderer, sortable = 'disabled', sortDirection = 'unsorted', onSortCriterionChange, isStickyStartEdge, isStickyEndEdge, hasStartEdge, hasEndEdge, isRtl, isAccColumnHeader, isHighContrast, padding, horizontalAlignment = 'start', verticalAlignment = 'center' }) {
    const cellClasses = classNames.classNames(TableColGroup.getHeaderFooterCellClassArray({
        isHeader: true,
        isFirstColumnIndex: columnIndex === 0 || isFirstLogicalColumnIndex,
        isFinalColumnIndex: isFinalColumnIndex || isFinalLogicalColumnIndex,
        hasVerticalGridlines,
        isSticky,
        isStickyStartEdge,
        isStickyEndEdge,
        hasStartEdge,
        hasEndEdge,
        isRtl,
        horizontalAlignment,
        verticalAlignment,
        isHighContrast,
        padding
    }));
    const handleSortGesture = hooks.useCallback(() => {
        if (onSortCriterionChange != null) {
            const newDirection = sortDirection === 'ascending' ? 'descending' : 'ascending';
            onSortCriterionChange([{ key: columnKey, sortDirection: newDirection }]);
        }
    }, [columnKey, sortDirection, onSortCriterionChange]);
    const headerCellRendererProps = {
        key: columnKey,
        headerText: headerText,
        sortable: sortable,
        sortDirection: sortDirection
    };
    const getHeaderCellRenderer = hooks.useCallback(() => {
        if (renderer != null) {
            return renderer;
        }
        return ({ headerText, sortable, sortDirection }) => {
            const text = headerText != null ? (jsxRuntime.jsx("div", { class: PRIVATE_Table_themes_Table_css.tableHeaderCellTextStyles.base, children: jsxRuntime.jsx("span", { class: PRIVATE_Table_themes_TableHeaderCellStyles_css.tableHeaderCellStyles.defaultRendererSpan, children: headerText }) })) : undefined;
            if (sortable === 'enabled' || sortDirection !== 'unsorted') {
                const isIconFirst = horizontalAlignment === 'end' ||
                    (!isRtl && horizontalAlignment === 'right') ||
                    (isRtl && horizontalAlignment === 'left');
                const containerClasses = classNames.classNames([
                    PRIVATE_Table_themes_Table_css.tableHeaderSortContainerStyles.base,
                    PRIVATE_TableCommon_themes_Common_css.justifyContentStyles[horizontalAlignment]
                ]);
                return (jsxRuntime.jsxs("div", { class: containerClasses, children: [isIconFirst && jsxRuntime.jsx(SortControl.SortControl, {}), text, !isIconFirst && jsxRuntime.jsx(SortControl.SortControl, {})] }));
            }
            return text;
        };
    }, [renderer, horizontalAlignment, isRtl]);
    const getAriaSort = () => {
        if (sortDirection === 'ascending' || sortDirection === 'descending') {
            return { 'aria-sort': sortDirection };
        }
        return undefined;
    };
    return (jsxRuntime.jsx(useSortableContext.SortableContext.Provider, { value: {
            isRendered: sortable === 'enabled',
            sortable,
            sortDirection,
            onSort: handleSortGesture,
            isTabbable: true
        }, children: jsxRuntime.jsx("th", { "aria-colindex": columnIndex + 1, ...getAriaSort(), class: cellClasses, style: `top:${stickyTopOffset}px;`, role: isAccColumnHeader ? 'columnheader' : 'cell', "data-oj-cell-type": 'header', "data-oj-column-key": columnKey, "data-oj-table-focusable": tableId, "data-oj-table-header": tableId, children: jsxRuntime.jsx(FormContext.FormContext.Provider, { value: { textAlign: TableColGroup.getFormTextAlign(horizontalAlignment, isRtl) }, children: getHeaderCellRenderer()(headerCellRendererProps) }) }) }));
}

/**
 * The internal component used to render a header region in Table.
 */
function TableDisplayHeader({ tableId, isRendered, columnsArray, hasVerticalGridlines, sortCriterion, onSortCriterionChange, stickyTopOffset, startStickyEdge, endStickyEdge, firstStickyColumnIndex, finalStickyColumnIndex, columnBorderEdges, isRtl, isHighContrast }) {
    const rowClasses = classNames.classNames([PRIVATE_Table_themes_Table_css.tableHeaderRowStyles.base]);
    return isRendered ? (jsxRuntime.jsx("thead", { class: PRIVATE_Table_themes_Table_css.backgroundStyles.inherit, role: 'presentation', children: jsxRuntime.jsx("tr", { "aria-rowindex": 1, class: rowClasses, role: 'row', "data-oj-table-header-row": tableId, children: columnsArray.map((column, columnIndex) => {
                const columnKey = column.key;
                let sortDirection;
                if (sortCriterion != null) {
                    for (const criterion of sortCriterion) {
                        if (criterion.key === columnKey) {
                            sortDirection = criterion.sortDirection;
                            break;
                        }
                    }
                }
                const isSticky = column.value.sticky === 'enabled';
                let isFirstLogicalColumnIndex = false;
                let isFinalLogicalColumnIndex = false;
                if (isSticky) {
                    if (startStickyEdge != null && firstStickyColumnIndex === columnIndex) {
                        isFirstLogicalColumnIndex = true;
                    }
                    if (endStickyEdge != null && finalStickyColumnIndex === columnIndex) {
                        isFinalLogicalColumnIndex = true;
                    }
                }
                const headerCellProps = {
                    tableId: tableId,
                    hasVerticalGridlines: hasVerticalGridlines,
                    columnKey: columnKey,
                    columnIndex: columnIndex,
                    isFinalColumnIndex: columnIndex === columnsArray.length - 1,
                    isFirstLogicalColumnIndex: isFirstLogicalColumnIndex,
                    isFinalLogicalColumnIndex: isFinalLogicalColumnIndex,
                    stickyTopOffset: stickyTopOffset,
                    isSticky: isSticky,
                    hasStartEdge: columnBorderEdges[columnIndex].hasStart,
                    hasEndEdge: columnBorderEdges[columnIndex].hasEnd,
                    headerText: column.value.headerText,
                    renderer: column.value.headerRenderer,
                    sortable: column.value.sortable,
                    sortDirection: sortDirection,
                    onSortCriterionChange: onSortCriterionChange,
                    isStickyStartEdge: columnIndex === startStickyEdge,
                    isStickyEndEdge: columnIndex === endStickyEdge,
                    isRtl: isRtl,
                    isAccColumnHeader: column.value.accessibleColumnHeader !== 'disabled',
                    isHighContrast: isHighContrast,
                    padding: column.value.headerPadding,
                    horizontalAlignment: column.value.headerAlignment?.horizontal != null
                        ? column.value.headerAlignment.horizontal
                        : column.value.horizontalAlignment,
                    verticalAlignment: column.value.headerAlignment?.vertical
                };
                return jsxRuntime.jsx(TableDisplayHeaderCell, { ...headerCellProps });
            }) }) })) : null;
}

function _defaultFooterCellRenderer({ footerText }) {
    return footerText != null ? footerText : undefined;
}
/**
 * The internal component used to render a single cell in a TableDisplayFooterRow.
 */
function TableDisplayFooterCell({ tableId, columnKey, columnIndex, isFinalColumnIndex, isFirstLogicalColumnIndex, isFinalLogicalColumnIndex, hasVerticalGridlines, isSticky, stickyBottomOffset = 0, footerText, renderer = _defaultFooterCellRenderer, isStickyStartEdge, isStickyEndEdge, hasStartEdge, hasEndEdge, isRtl, isHighContrast, padding, horizontalAlignment = 'start', verticalAlignment = 'center' }) {
    const cellClasses = classNames.classNames(TableColGroup.getHeaderFooterCellClassArray({
        isHeader: false,
        isFirstColumnIndex: columnIndex === 0 || isFirstLogicalColumnIndex,
        isFinalColumnIndex: isFinalColumnIndex || isFinalLogicalColumnIndex,
        hasVerticalGridlines,
        isSticky,
        hasStartEdge,
        hasEndEdge,
        isStickyStartEdge,
        isStickyEndEdge,
        isRtl,
        horizontalAlignment,
        verticalAlignment,
        isHighContrast,
        padding
    }));
    const footerCellRendererProps = {
        key: columnKey,
        footerText: footerText
    };
    return (jsxRuntime.jsx("td", { "aria-colindex": columnIndex + 1, class: cellClasses, style: `bottom:${stickyBottomOffset}px;`, role: 'cell', "data-oj-cell-type": 'footer', "data-oj-column-key": columnKey, "data-oj-table-focusable": tableId, "data-oj-table-footer": tableId, children: jsxRuntime.jsx(FormContext.FormContext.Provider, { value: { textAlign: TableColGroup.getFormTextAlign(horizontalAlignment, isRtl) }, children: renderer(footerCellRendererProps) }) }));
}

/**
 * The internal component used to render a footer region in Table.
 */
function TableDisplayFooter({ tableId, isRendered, columnsArray, hasVerticalGridlines, stickyBottomOffset, startStickyEdge, endStickyEdge, firstStickyColumnIndex, finalStickyColumnIndex, columnBorderEdges, dataRowCount, isRtl, isHighContrast }) {
    const rowClasses = classNames.classNames([PRIVATE_Table_themes_Table_css.tableFooterRowStyles.base]);
    return isRendered ? (jsxRuntime.jsx("tfoot", { class: PRIVATE_Table_themes_Table_css.backgroundStyles.inherit, role: 'presentation', children: jsxRuntime.jsx("tr", { "aria-rowindex": Math.max(dataRowCount, 1) + 2, class: rowClasses, role: 'row', "data-oj-table-footer-row": tableId, children: columnsArray.map((column, columnIndex) => {
                const columnKey = column.key;
                const isSticky = column.value.sticky === 'enabled';
                let isFirstLogicalColumnIndex = false;
                let isFinalLogicalColumnIndex = false;
                if (isSticky) {
                    if (startStickyEdge != null && firstStickyColumnIndex === columnIndex) {
                        isFirstLogicalColumnIndex = true;
                    }
                    if (endStickyEdge != null && finalStickyColumnIndex === columnIndex) {
                        isFinalLogicalColumnIndex = true;
                    }
                }
                const footerCellProps = {
                    tableId: tableId,
                    columnKey: columnKey,
                    columnIndex: columnIndex,
                    isFinalColumnIndex: columnIndex === columnsArray.length - 1,
                    isFirstLogicalColumnIndex: isFirstLogicalColumnIndex,
                    isFinalLogicalColumnIndex: isFinalLogicalColumnIndex,
                    hasVerticalGridlines: hasVerticalGridlines,
                    stickyBottomOffset: stickyBottomOffset,
                    isSticky: isSticky,
                    hasStartEdge: columnBorderEdges[columnIndex].hasStart,
                    hasEndEdge: columnBorderEdges[columnIndex].hasEnd,
                    footerText: column.value.footerText,
                    renderer: column.value.footerRenderer,
                    isStickyStartEdge: columnIndex === startStickyEdge,
                    isStickyEndEdge: columnIndex === endStickyEdge,
                    isRtl: isRtl,
                    isHighContrast: isHighContrast,
                    padding: column.value.footerPadding,
                    horizontalAlignment: column.value.footerAlignment?.horizontal != null
                        ? column.value.footerAlignment.horizontal
                        : column.value.horizontalAlignment,
                    verticalAlignment: column.value.footerAlignment?.vertical
                };
                return jsxRuntime.jsx(TableDisplayFooterCell, { ...footerCellProps });
            }) }) })) : null;
}

/**
 * Component that renders data in a flat table.
 */
const TableDisplayBase = compat.forwardRef(function TableDisplayBase({ 'aria-label': ariaLabel, 'aria-labelledby': ariaLabelledBy, 'aria-describedby': ariaDescribedBy, getAccessibleRowHeaders, columns, columnOrder, columnWidths, viewportConfig, layout = 'contents', data, getRowKey, gridlines, sortCriterion, onSortCriterionChange, horizontalScrollPositionOverride, verticalScrollPositionOverride, noDataRenderer, scrollerRef, testId }) {
    const hasMore = false;
    const onLoadMore = () => { };
    const tableId = useId.useId();
    const testIdProps = useTestId.useTestId(testId);
    // rootRef is callback-based to meet useResizeObserver hook's state management requirements
    const [rootRef, setRootRef] = hooks.useState(preact.createRef());
    const rootRefCallback = hooks.useCallback((node) => {
        setRootRef({ current: node });
    }, []);
    const innerTableRef = hooks.useRef(null);
    const { direction, forcedColors } = useUser.useUser();
    const isHighContrast = forcedColors === 'active';
    const isRtl = direction === 'rtl';
    const { managedData: tableData, onManagedLoadMore: onTableLoadMore } = ScrollUtils.useDataManager({
        data,
        getRowKey,
        hasMore,
        onLoadMore
    });
    const { columnBorderEdges, isLayoutRequired, viewportInfo, appliedColumnWidths, sizingInfoRef, overflowInfoRef, hasVerticalUnderflow, stickyEdgesState, setStickyEdgesState, stickyColumnInfo, scrollerRefCallback } = TableColGroup.useSizingManager({
        rootRef,
        innerTableRef,
        tableId,
        tableData,
        viewportConfig,
        externalScrollerAccessRef: scrollerRef,
        layout,
        columns,
        columnOrder,
        columnWidths,
        isRtl,
        isHighContrast
    });
    // memoize realized columns array on 'columns' and 'columnOrder'
    const columnsArray = hooks.useMemo(() => {
        const newColumnsArray = [];
        if (columnOrder != null) {
            for (const key of columnOrder) {
                if (columns[key] != null) {
                    newColumnsArray.push({
                        key: key,
                        value: columns[key]
                    });
                }
            }
        }
        else {
            for (const [key, value] of Object.entries(columns)) {
                newColumnsArray.push({
                    key: key,
                    value: value
                });
            }
        }
        return newColumnsArray;
    }, [columnOrder, columns]);
    const isPendingLayout = layout === 'pending';
    const scrollHandlingProps = TableColGroup.useScrollHandling({
        rootRef,
        viewportInfo,
        tableId,
        isLayoutRequired,
        sizingInfoRef,
        overflowInfoRef,
        columnsArray,
        stickyColumnInfo,
        stickyEdgesState,
        setStickyEdgesState,
        fullColumnWidths: appliedColumnWidths?.fullColumnWidths,
        horizontalScrollPositionOverride,
        verticalScrollPositionOverride,
        isPendingLayout,
        isRtl
    });
    // aria-rowcount includes header and footer rows
    const hasFooters = columnsArray.some((column) => {
        return column.value.footerRenderer != null || column.value.footerText != null;
    });
    const ariaRowCount = isPendingLayout ? -1 : tableData.length + (hasFooters ? 2 : 1);
    const ariaColumnCount = Math.max(columnsArray.length, 1);
    // initialize base Table style classes
    const { classes: themeClasses } = useComponentTheme.useComponentTheme(PRIVATE_Table_themes_redwood_TableTheme.TableRedwoodTheme, {
        pendingLayout: isPendingLayout ? 'isPendingLayout' : 'notPendingLayout',
        showFocusRing: 'notShowFocusRing',
        externalScroller: viewportInfo.isExternalScroller ? 'isExternalScroller' : 'notExternalScroller'
    });
    let scrollerStyle;
    let scrollerPadding = '';
    const scrollerClassArray = [PRIVATE_Table_themes_Table_css.scrollerStyles.base];
    if (viewportInfo.isExternalScroller) {
        scrollerClassArray.push(PRIVATE_Table_themes_Table_css.scrollerStyles.externalScroller);
    }
    else if (!sizingInfoRef.current.isScrollbarSizeKnown) {
        scrollerClassArray.push(PRIVATE_Table_themes_Table_css.scrollerStyles.forceScroll);
    }
    else {
        if (overflowInfoRef.current.isHorizontal) {
            scrollerClassArray.push(PRIVATE_Table_themes_Table_css.scrollerStyles.horizontalScroll);
        }
        else {
            scrollerClassArray.push(PRIVATE_Table_themes_Table_css.scrollerStyles.noHorizontalScroll);
        }
        if (overflowInfoRef.current.isVertical) {
            scrollerClassArray.push(PRIVATE_Table_themes_Table_css.scrollerStyles.verticalScroll);
        }
        else {
            scrollerClassArray.push(PRIVATE_Table_themes_Table_css.scrollerStyles.noVerticalScroll);
            scrollerPadding = `padding-inline-end: ${sizingInfoRef.current.defaultScrollbarWidth}px;`;
        }
    }
    // hide the table contents when layout is required and otherwise valid
    if (isLayoutRequired && columnsArray.length > 0) {
        scrollerStyle = `${scrollerPadding}opacity:0.001;`;
    }
    else {
        scrollerStyle = scrollerPadding;
    }
    const scrollerClasses = classNames.classNames(scrollerClassArray);
    const classes = classNames.classNames([themeClasses]);
    const tableClasses = classNames.classNames([PRIVATE_Table_themes_Table_css.innerTableStyles.base]);
    const startStickyEdge = stickyEdgesState.start;
    const endStickyEdge = stickyEdgesState.end;
    const hasVerticalGridlines = gridlines?.vertical === 'visible';
    const hasHorizontalGridlines = gridlines?.horizontal !== 'hidden';
    const getTableBodyProps = () => {
        return {
            rootRef: rootRef,
            viewportInfo: viewportInfo,
            tableId: tableId,
            isValidRender: columnsArray.length > 0,
            tableData: tableData,
            columnsArray: columnsArray,
            getAccessibleRowHeaders: getAccessibleRowHeaders,
            hasVerticalGridlines: hasVerticalGridlines,
            hasHorizontalGridlines: hasHorizontalGridlines,
            hasFooters: hasFooters,
            hasVerticalUnderflow: hasVerticalUnderflow,
            hasMore: hasMore,
            onLoadMore: onTableLoadMore,
            noDataRenderer: noDataRenderer,
            startStickyEdge: startStickyEdge,
            endStickyEdge: endStickyEdge,
            firstStickyColumnIndex: stickyColumnInfo.firstStickyColumnIndex,
            finalStickyColumnIndex: stickyColumnInfo.finalStickyColumnIndex,
            columnBorderEdges: columnBorderEdges,
            isRtl: isRtl,
            isHighContrast: isHighContrast
        };
    };
    const getTableColGroupProps = () => {
        let columnWidthsArray;
        if (!isLayoutRequired && appliedColumnWidths?.realColumnWidths != null) {
            columnWidthsArray = appliedColumnWidths.realColumnWidths;
        } /* else if (layout === 'fixed') {
          // default to 100px for each column when layout is not yet available
          columnWidthsArray = [];
          for (let i = 0; i < columnsArray.length; i++) {
            columnWidthsArray[i] = 100;
          }
        } */
        else {
            columnWidthsArray = [];
            for (let i = 0; i < columnsArray.length; i++) {
                columnWidthsArray[i] = undefined;
            }
        }
        return {
            tableId: tableId,
            isRendered: columnsArray.length > 0,
            columnWidthsArray: columnWidthsArray
        };
    };
    const getTableHeaderProps = () => {
        return {
            tableId: tableId,
            isRendered: columnsArray.length > 0,
            columnsArray: columnsArray,
            hasVerticalGridlines: hasVerticalGridlines,
            sortCriterion: sortCriterion,
            onSortCriterionChange: onSortCriterionChange,
            stickyTopOffset: viewportInfo.viewportConfig.scrollerOffsetTop,
            startStickyEdge: startStickyEdge,
            endStickyEdge: endStickyEdge,
            firstStickyColumnIndex: stickyColumnInfo.firstStickyColumnIndex,
            finalStickyColumnIndex: stickyColumnInfo.finalStickyColumnIndex,
            columnBorderEdges: columnBorderEdges,
            isRtl: isRtl,
            isHighContrast: isHighContrast
        };
    };
    const getTableFooterProps = () => {
        return {
            tableId: tableId,
            isRendered: hasFooters,
            columnsArray: columnsArray,
            hasVerticalGridlines: hasVerticalGridlines,
            stickyBottomOffset: viewportInfo.viewportConfig.scrollerOffsetBottom,
            startStickyEdge: startStickyEdge,
            endStickyEdge: endStickyEdge,
            firstStickyColumnIndex: stickyColumnInfo.firstStickyColumnIndex,
            finalStickyColumnIndex: stickyColumnInfo.finalStickyColumnIndex,
            columnBorderEdges: columnBorderEdges,
            dataRowCount: tableData.length,
            isRtl: isRtl,
            isHighContrast: isHighContrast
        };
    };
    const getTableStyle = () => {
        if (!isLayoutRequired && appliedColumnWidths?.realColumnWidths != null) {
            let totalWidth = 0;
            for (let i = 0; i < appliedColumnWidths.realColumnWidths.length; i++) {
                totalWidth += appliedColumnWidths.realColumnWidths[i];
            }
            return `width:${totalWidth}px;table-layout:fixed;`;
        }
        else if (columnsArray.length > 0) {
            /* if (layout === 'fixed') {
              // default to 100px for each column when layout is not yet available
              return `width:${columnsArray.length * 100}px;table-layout:fixed;`;
            } else { */
            // shrink table down as much as possible when a 'contents' layout is required, otherwise
            // the sampled column widths will not reflect the true weight sizes with line-clamping
            return 'width:max-content;will-change:width;';
            // }
        }
        // no columns are present
        return 'width:100%;';
    };
    const skeletonStyles = isHighContrast ? PRIVATE_Table_themes_TableSkeletonStyles_css.tableSkeletonStylesHC : PRIVATE_Table_themes_TableSkeletonStyles_css.tableSkeletonStyles;
    const getPendingLayoutSkeletons = () => {
        return isPendingLayout ? (jsxRuntime.jsx("div", { class: skeletonStyles.pendingLayout, children: jsxRuntime.jsx("table", { role: 'presentation', style: 'width:100%;table-layout:fixed;', children: jsxRuntime.jsx("tbody", { role: 'presentation', children: jsxRuntime.jsx(TableColGroup.TableSkeletonRow, { tableId: tableId, hasTabIndex: false, colspan: ariaColumnCount, isPendingLayout: true, isShowFocusRing: false, isEmbedded: false, isHighContrast: isHighContrast }) }) }) })) : undefined;
    };
    return (jsxRuntime.jsx(jsxRuntime.Fragment, { children: jsxRuntime.jsxs("div", { ...testIdProps, ref: rootRefCallback, class: classes, id: tableId, children: [jsxRuntime.jsx("div", { ref: scrollerRefCallback, "aria-colcount": ariaColumnCount, "aria-label": ariaLabel, "aria-labelledby": ariaLabelledBy, "aria-describedby": ariaDescribedBy, "aria-rowcount": ariaRowCount, class: scrollerClasses, style: scrollerStyle, role: 'table', tabIndex: 0, ...scrollHandlingProps, children: jsxRuntime.jsxs("table", { ref: innerTableRef, "aria-hidden": isPendingLayout ? true : undefined, class: tableClasses, role: 'presentation', style: getTableStyle(), "data-oj-table-inner-table": tableId, children: [jsxRuntime.jsx(TableColGroup.TableColGroup, { ...getTableColGroupProps() }), jsxRuntime.jsx(TableDisplayHeader, { ...getTableHeaderProps() }), jsxRuntime.jsx(TableDisplayBody, { ...getTableBodyProps() }), jsxRuntime.jsx(TableDisplayFooter, { ...getTableFooterProps() })] }) }), getPendingLayoutSkeletons()] }) }));
});

const mergedInterpolations = mergeInterpolations.mergeInterpolations([
    ...Object.values(dimensions.dimensionInterpolations)
]);
/**
 * Component that renders data in rows of columns.
 * This component is non-virtualized and renders all rows provided in the data prop.
 */
const TableDisplay = compat.forwardRef(({ 'aria-label': ariaLabel, 'aria-labelledby': ariaLabelledBy, 'aria-describedby': ariaDescribedBy, getAccessibleRowHeaders, children, columns, columnOrder, columnWidths, viewportConfig, layout = 'contents', data, getRowKey, gridlines = { horizontal: 'visible', vertical: 'hidden' }, sortCriterion = [], onSortCriterionChange, horizontalScrollPositionOverride, verticalScrollPositionOverride, noDataRenderer, scrollerRef, testId, ...dimensions }, ref) => {
    const { ...mergedStyles } = mergedInterpolations(dimensions);
    const isExternalScroller = viewportConfig != null;
    const classArray = [PRIVATE_TableDisplay_themes_TableDisplayStyles_css.tableDisplayStyles.base];
    classArray.push(isExternalScroller ? PRIVATE_TableDisplay_themes_TableDisplayStyles_css.tableDisplayStyles.externalScroller : PRIVATE_TableDisplay_themes_TableDisplayStyles_css.tableDisplayStyles.internalScroller);
    const tableDisplayClasses = classNames.classNames(classArray);
    return (jsxRuntime.jsx("div", { class: tableDisplayClasses, style: mergedStyles, children: jsxRuntime.jsx(TableDisplayBase, { "aria-label": ariaLabel, "aria-labelledby": ariaLabelledBy, "aria-describedby": ariaDescribedBy, getAccessibleRowHeaders: getAccessibleRowHeaders, children: children, columns: columns, columnOrder: columnOrder, columnWidths: columnWidths, viewportConfig: viewportConfig, layout: layout, data: data, getRowKey: getRowKey, gridlines: gridlines, horizontalScrollPositionOverride: horizontalScrollPositionOverride, verticalScrollPositionOverride: verticalScrollPositionOverride, sortCriterion: sortCriterion, onSortCriterionChange: onSortCriterionChange, noDataRenderer: noDataRenderer, scrollerRef: scrollerRef, testId: testId, ref: ref }) }));
});

exports.TableDisplay = TableDisplay;
