// @oracle/oraclejet-preact/cjs/RadioItem-c74ac827.js
/* @oracle/oraclejet-preact: undefined */
'use strict';

var jsxRuntime = require('preact/jsx-runtime');
var compat = require('preact/compat');
var useComponentTheme = require('./useComponentTheme-082fc8e4.js');
var useFocus = require('./useFocus-1b288fb9.js');
require('./FormContext-d1d05967.js');
var useFormContext = require('./useFormContext-a794f7b6.js');
require('./FormFieldContext-69fe83d2.js');
var useFormFieldContext = require('./useFormFieldContext-4632eb73.js');
var useId = require('./useId-6c0eeb27.js');
require('./TabbableModeContext-e99d527e.js');
var useTabbableMode = require('./useTabbableMode-dbd254e8.js');
var useInteractionStyle = require('./useInteractionStyle-c203a8a0.js');
var useTranslationBundle = require('./useTranslationBundle-12dd087d.js');
var CheckboxRadioContext = require('./CheckboxRadioContext-3b134bd1.js');
var Check = require('./Check-1272bb86.js');
var RadioButtonOff = require('./RadioButtonOff-0b9700f9.js');
var RadioButtonOn = require('./RadioButtonOn-8264d513.js');
var HiddenAccessible = require('./HiddenAccessible-087eb49a.js');
var RadioItemContext = require('./RadioItemContext-3a86c17d.js');


require('./UNSAFE_UserAssistance/themes/redwood/UserAssistanceVariants.css.js');
require('preact/hooks');
require('./Flex-3a257bc3.js');


require('./UNSAFE_MessageBanner/themes/redwood/MessageBannerVariants.css.js');
require('preact');
require('./LayerHost-3cd73603.js');


require('./hooks/UNSAFE_useTooltip/themes/redwood/TooltipContentVariants.css.js');
require('./tooltipUtils-2bc15136.js');

var clientHints = require('./clientHints-c40c5250.js');
require('./IconButton-7503293b.js');

require('./MessageFormattingUtils-539d75b0.js');
require('./logger-2b636482.js');
require('./TransitionGroup-b239d98f.js');
require('./MessagesContext-4e939750.js');


require('./UNSAFE_ComponentMessage/themes/redwood/ComponentMessageVariants.css.js');
require('./InputGroupContext-05f2a46f.js');
require('./Popup-de16d915.js');
require('./UNSAFE_Separator/themes/SeparatorStyles.css.js');


require('./UNSAFE_Label/themes/redwood/LabelVariants.css.js');
var IconUserAssistance = require('./IconUserAssistance-89913ad1.js');
var UNSAFE_RadioItem_themes_RadioIconContract_css = require('./UNSAFE_RadioItem/themes/RadioIconContract.css.js');
var UNSAFE_RadioItem_themes_redwood_RadioIconTheme = require('./UNSAFE_RadioItem/themes/redwood/RadioIconTheme.js');
var UNSAFE_RadioItem_themes_redwood_RadioTheme = require('./UNSAFE_RadioItem/themes/redwood/RadioTheme.js');

/**
 * Returns the accessibility text for a radio button's selection state.
 *
 * The text is determined based on the client's platform (Mac or iOS/PC/Android) and the radio button's selection state.
 * Mac VoiceOver announces "selected" while others use "checked".
 *
 * @param {BundleType} translations The translation bundle containing the necessary text.
 * @param {boolean} isSelected Whether the radio button is currently selected.
 * @returns {string} The accessibility text for the radio button's selection state.
 */
const getRadioSelectionText = (translations, isSelected) => {
    const isMac = clientHints.getClientHints().platform === 'mac';
    const translationKey = isMac
        ? isSelected
            ? 'radioButton_selected'
            : 'radioButton_unselected'
        : isSelected
            ? 'radioButton_checked'
            : 'radioButton_unchecked';
    return translations[translationKey]();
};
/**
 * RadioItem is an input element of type radio.
 */
const RadioItem = ({ value, children, ...assistanceProps }) => {
    const { isReadonly: isFormReadonly, isDisabled: isFormDisabled } = useFormContext.useFormContext();
    const { isDisabled: isFieldDisabled, isReadonly: isFieldReadonly } = useFormFieldContext.useFormFieldContext();
    const { name, value: groupValue, onCommit, setFocusIndex } = RadioItemContext.useRadioSetContext();
    const { isFocusRingShown } = CheckboxRadioContext.useCheckboxRadioContext();
    const { index } = RadioItemContext.useRadioItemContext();
    const { isTabbable } = useTabbableMode.useTabbableMode();
    const id = useId.useId();
    const { isFocus, focusProps } = useFocus.useFocus();
    const labelRef = compat.createRef();
    const isSelected = groupValue === value;
    const isDisabled = isFieldDisabled ?? isFormDisabled;
    const isReadonly = isFieldReadonly ?? isFormReadonly;
    const { interactionProps, applyActiveStyle } = useInteractionStyle.useInteractionStyle();
    const inputRef = compat.useRef(null);
    // JET-70100 - ensure radio button is scrolled into view when focused and focus ring is shown
    compat.useEffect(() => {
        const isActiveElement = document.activeElement === inputRef?.current;
        if (!isActiveElement && isFocus && isFocusRingShown) {
            labelRef?.current?.scrollIntoView({ block: 'nearest' });
        }
    }, [isFocus, isFocusRingShown, labelRef]);
    const { assistiveText, helpSourceLink } = assistanceProps;
    const hasUserAssistive = !isDisabled && !isReadonly && !!(assistiveText || helpSourceLink);
    const helpAvailableId = hasUserAssistive ? `${id}-help-available` : undefined;
    const uaId = hasUserAssistive ? `${id}-icon-ua` : undefined;
    const translations = useTranslationBundle.useTranslationBundle('@oracle/oraclejet-preact');
    const helpAvailableString = translations.formControl_helpAvailable();
    const IconComp = isReadonly
        ? isSelected
            ? Check.SvgCheck
            : RadioButtonOff.SvgRadioButtonOff
        : isSelected
            ? RadioButtonOn.SvgRadioButtonOn
            : RadioButtonOff.SvgRadioButtonOff;
    const tabIndex = isTabbable !== false && ((index === 0 && !groupValue) || isSelected) ? 0 : -1;
    const accessibilityText = `${children || ''} ${translations.formControl_readOnly()} ${getRadioSelectionText(translations, isSelected)}`;
    const handleFocus = compat.useCallback(() => {
        setFocusIndex?.(index);
    }, [setFocusIndex, index]);
    const handleLabelClick = compat.useCallback(() => {
        inputRef.current?.focus();
        setFocusIndex?.(index);
    }, [inputRef, setFocusIndex, index]);
    const handleChange = compat.useCallback((e) => {
        e.stopPropagation();
        onCommit?.({ previousValue: groupValue, value });
    }, [onCommit, groupValue, value]);
    /**
     * JET-76970
     * Safari's `:focus-visible` is applied only during keyboard focus (e.g., Tab).
     * However, when a child of a card receives focus via arrow navigation, Safari
     * ignores `:focus-visible` even if it would match, and it only shows a focus
     * outline at navigation endpoints.
     *
     * For consistency:
     * - For Safari, add a class that forces the ring whenever focus is inside the
     *   card during keyboard navigation.
     * - For other browsers or readonly items, rely on native `:focus-visible` selectors.
     *
     * Known issue:
     * - Because of the timeout in useCollectionFocusRing, a long press on a item can
     *   sometimes trigger a focus ring, whereas a normal click does not.
     */
    const isSafari = clientHints.getClientHints().browser === 'safari';
    const focusWithRing = isFocus && isFocusRingShown;
    const isFocusRingSelector = isReadonly || (!isSafari && focusWithRing);
    const isForcedFocusRing = isSafari && focusWithRing;
    const { classes: radioClasses, styles: { valueLabelStyle } } = useComponentTheme.useComponentTheme(UNSAFE_RadioItem_themes_redwood_RadioTheme.RadioRedwoodTheme, {
        disabled: isDisabled ? 'isDisabled' : 'notDisabled',
        focusRing: isForcedFocusRing
            ? 'isForcedFocusRing'
            : isFocusRingSelector
                ? 'isFocusRingSelector'
                : undefined
    });
    const { classes: radioIconClasses } = useComponentTheme.useComponentTheme(UNSAFE_RadioItem_themes_redwood_RadioIconTheme.RadioIconRedwoodTheme, {
        disabled: isDisabled ? 'isDisabled' : 'notDisabled',
        selected: isSelected ? 'isSelected' : 'notSelected',
        readonly: isReadonly ? 'isReadonly' : 'notReadonly',
        active: applyActiveStyle ? 'isActive' : 'notActive'
    });
    return (jsxRuntime.jsx("div", { children: jsxRuntime.jsxs("label", { class: radioClasses, ref: labelRef, ...(isReadonly
                ? {
                    onFocus: handleFocus,
                    onClick: handleLabelClick
                }
                : {}), children: [jsxRuntime.jsxs(HiddenAccessible.HiddenAccessible, { children: [isReadonly ? (jsxRuntime.jsx("div", { "aria-describedby": helpAvailableId, id: id, ref: inputRef, tabIndex: tabIndex, "data-checked": isSelected, ...focusProps, children: accessibilityText })) : (jsxRuntime.jsx("input", { type: "radio", name: name, id: id, ref: inputRef, value: value, ...(!isDisabled && {
                                onChange: handleChange
                            }), checked: isSelected, disabled: isDisabled, "aria-describedby": helpAvailableId, tabIndex: tabIndex, ...focusProps })), hasUserAssistive && (jsxRuntime.jsx(HiddenAccessible.HiddenAccessible, { id: helpAvailableId, isHidden: true, children: helpAvailableString }))] }), jsxRuntime.jsx("span", { "aria-hidden": "true", class: radioIconClasses, ...interactionProps, children: jsxRuntime.jsx(IconComp, { size: UNSAFE_RadioItem_themes_RadioIconContract_css.radioIconVars.iconSize, color: "currentColor" }) }), jsxRuntime.jsxs("span", { class: valueLabelStyle, children: [jsxRuntime.jsx("span", { ...interactionProps, children: children }), hasUserAssistive && (jsxRuntime.jsx(IconUserAssistance.IconUserAssistance, { ...assistanceProps, isTabbable: isTabbable && isFocus, id: uaId, variant: "radio" }))] })] }) }));
};

exports.RadioItem = RadioItem;
exports.getRadioSelectionText = getRadioSelectionText;
