// @oracle/oraclejet-preact/cjs/ScrollUtils-e8076071.js
/* @oracle/oraclejet-preact: undefined */
'use strict';

var hooks = require('preact/hooks');

/**
 * Hook that manages data on collection components
 */
function useDataManager({ data, getRowKey, hasMore, onLoadMore }) {
    const managedData = hooks.useMemo(() => {
        return data != null ? data.map((d) => ({ data: d, metadata: { key: getRowKey(d) } })) : [];
    }, [data, getRowKey]);
    // debounce loadMore calls to prevent state changes from triggering multiple calls
    const loadMoreDataRef = hooks.useRef();
    if (!hasMore) {
        // clear loadMoreData cache once there is no more data available
        loadMoreDataRef.current = undefined;
    }
    const onManagedLoadMore = hooks.useCallback(() => {
        if (data === loadMoreDataRef.current) {
            return;
        }
        loadMoreDataRef.current = data;
        return onLoadMore();
    }, [data, onLoadMore]);
    return { managedData, onManagedLoadMore };
}

/**
 * Returns the cell type for the given element
 */
const getCellType = (element) => {
    return element.dataset['ojCellType'];
};
/**
 * Returns the row key for the given element
 */
const getRowKey = (element) => {
    return (element.dataset['ojKeyType'] === 'number'
        ? Number(element.dataset['ojKey'])
        : element.dataset['ojKey']);
};
/**
 * Returns the index of the given row element
 */
const getRowIndexForRowElement = (rootElement, rootId, rowElement, rowSelector) => {
    const rowElements = rootElement.querySelectorAll(`[${rowSelector}='${rootId}']`);
    const index = Array.prototype.indexOf.call(rowElements, rowElement);
    return index > -1 ? index : undefined;
};
/**
 * Returns the key of the first rendered row in the component
 */
const getFirstRenderedRowKey = (rootElement, rootId, rowSelector) => {
    const firstRowElement = rootElement.querySelector(`[${rowSelector}='${rootId}']`);
    return firstRowElement != null ? getRowKey(firstRowElement) : undefined;
};
/**
 * Returns the key of the last rendered row in the component
 */
const getLastRenderedRowKey = (rootElement, rootId, rowSelector) => {
    const rowElements = rootElement.querySelectorAll(`[${rowSelector}='${rootId}']`);
    const rowCount = rowElements.length;
    return rowCount > 0 ? getRowKey(rowElements[rowCount - 1]) : undefined;
};
/**
 * Helper method to get the 'no data' element
 */
const getElementBySelector = (rootElement, rootId, elementSelector) => {
    const element = rootElement.querySelector(`[${elementSelector}='${rootId}']`);
    return element != null ? element : undefined;
};

/**
 * @license
 * Copyright (c) %FIRST_YEAR% %CURRENT_YEAR%, Oracle and/or its affiliates.
 * The Universal Permissive License (UPL), Version 1.0
 * as shown at https://oss.oracle.com/licenses/upl/
 * @ignore
 */
/**
 * Helper method to determine whether the scroller is the document body or document element.
 */
const isHTMLScroller = (scrollerElement) => {
    return scrollerElement === document.body || scrollerElement === document.documentElement;
};
/**
 * Helper method to determine if the vertical scrollbar is on the left side in RTL languages.
 */
const isLeftHTMLScrollbar = (scrollerHTMLElement) => {
    return scrollerHTMLElement.getBoundingClientRect().left + scrollerHTMLElement.scrollLeft > 0;
};
/**
 * Helper method to return the current scrollbar height of the scroller element.
 */
const getScrollbarHeight = (scrollerElement) => {
    return isHTMLScroller(scrollerElement)
        ? window.innerHeight - scrollerElement.clientHeight
        : scrollerElement.offsetHeight - scrollerElement.clientHeight;
};
/**
 * Helper method to return the current scrollbar width of the scroller element.
 */
const getScrollbarWidth = (scrollerElement) => {
    return isHTMLScroller(scrollerElement)
        ? window.innerWidth - scrollerElement.clientWidth
        : scrollerElement.offsetWidth - scrollerElement.clientWidth;
};
/**
 * Helper method to get the horizontal overflow diff of the scroller vs the given element.
 */
const getHorizontalOverflowDiff = (element, scrollerStart, scrollerEnd, isRtl) => {
    const elementRect = element.getBoundingClientRect();
    if (isRtl) {
        return {
            left: scrollerEnd - elementRect.left,
            right: elementRect.right - scrollerStart
        };
    }
    return {
        left: scrollerStart - elementRect.left,
        right: elementRect.right - scrollerEnd
    };
};
/**
 * Helper method to get the vertical overflow diff of the scroller vs the given element.
 */
const getVerticalOverflowDiff = (element, scrollerTop, scrollerBottom) => {
    const elementRect = element.getBoundingClientRect();
    return {
        top: scrollerTop - elementRect.top,
        bottom: elementRect.bottom - scrollerBottom
    };
};

exports.getCellType = getCellType;
exports.getElementBySelector = getElementBySelector;
exports.getFirstRenderedRowKey = getFirstRenderedRowKey;
exports.getHorizontalOverflowDiff = getHorizontalOverflowDiff;
exports.getLastRenderedRowKey = getLastRenderedRowKey;
exports.getRowIndexForRowElement = getRowIndexForRowElement;
exports.getRowKey = getRowKey;
exports.getScrollbarHeight = getScrollbarHeight;
exports.getScrollbarWidth = getScrollbarWidth;
exports.getVerticalOverflowDiff = getVerticalOverflowDiff;
exports.isHTMLScroller = isHTMLScroller;
exports.isLeftHTMLScrollbar = isLeftHTMLScrollbar;
exports.useDataManager = useDataManager;
