// @oracle/oraclejet-preact/cjs/StyledSwitch-17127a03.js
/* @oracle/oraclejet-preact: undefined */
'use strict';

var jsxRuntime = require('preact/jsx-runtime');
var compat = require('preact/compat');
var hooks = require('preact/hooks');
var useTestId = require('./useTestId-4f53d561.js');
var PRIVATE_StyledSwitch_themes_StyledSwitchStyles_css = require('./PRIVATE_StyledSwitch/themes/StyledSwitchStyles.css.js');
require('./TabbableModeContext-e99d527e.js');
var useTabbableMode = require('./useTabbableMode-dbd254e8.js');
var useComponentTheme = require('./useComponentTheme-082fc8e4.js');
var PRIVATE_StyledSwitch_themes_redwood_StyledSwitchTheme = require('./PRIVATE_StyledSwitch/themes/redwood/StyledSwitchTheme.js');
var classNames = require('./classNames-c14c6ef3.js');
var HiddenAccessible = require('./HiddenAccessible-087eb49a.js');
var useTranslationBundle = require('./useTranslationBundle-12dd087d.js');
var useId = require('./useId-6c0eeb27.js');
var useActive = require('./useActive-6770f917.js');
var mergeProps = require('./mergeProps-e3da7237.js');
var stringUtils = require('./stringUtils-3e19c8af.js');

/**
 * Provides standardized event handlers for switch UI components (such as StyledSwitch).
 *
 * - Handles mouse click and keyboard activation (Space/Enter) on <kbd>keyup</kbd> (legacy behavior).
 * - Implements "gating" for Enter and Space: onChange will only trigger on keyup if that key's keydown
 *   happened on the component, preventing accidental activation when focus shifts mid-keypress.
 *   (e.g. Enter down elsewhere, focus moves to switch, Enter up: switch does NOT toggle)
 * - Ignores events if `isDisabled` or `isReadonly` are true.
 * - Internally uses isEnterPressed and isSpacePressed state to track legitimate activation origin.
 * - Forwards `onChange` when value should toggle.
 *
 * @param {SwitchHandlersParams} params - Props controlling switch behavior and callbacks.
 * @returns {{
 *   onClick: (e: MouseEvent) => void;
 *   onKeyDown: (e: KeyboardEvent) => void;
 *   onKeyUp: (e: KeyboardEvent) => void;
 * }} Event handlers to be spread on the root switch element.
 *
 * @example
 *   const { onClick, onKeyDown, onKeyUp } = useSwitchHandlers(...);
 *   <div tabIndex={0} onClick={onClick} onKeyDown={onKeyDown} onKeyUp={onKeyUp} ... />
 */
function useSwitchHandlers({ isDisabled, isReadonly, onChange }) {
    const [isEnterPressed, setIsEnterPressed] = hooks.useState(false);
    const [isSpacePressed, setIsSpacePressed] = hooks.useState(false);
    const handleClick = hooks.useCallback((e) => {
        if (isDisabled || isReadonly)
            return;
        e.preventDefault();
        e.stopPropagation();
        onChange?.(e);
    }, [isDisabled, isReadonly, onChange]);
    /**
     * Gating logic for Enter and Space:
     * Only "arms" activation on the *first* keydown (when event.repeat is false).
     * This means that if a key is held down elsewhere and focus moves to the switch
     * (e.g. user presses and holds Enter on a button, button sets focus to switch
     * on activation, then more auto-repeated keydowns will NOT mistakenly arm the switch),
     * only an Enter or Space press that originated with the switch actually focused
     * will allow a subsequent keyup to activate.
     */
    const handleKeyDown = hooks.useCallback((e) => {
        if (isDisabled || isReadonly)
            return;
        if (e.code === 'Enter' && !e.repeat) {
            setIsEnterPressed(true);
            e.preventDefault();
        }
        if (e.code === 'Space' && !e.repeat) {
            setIsSpacePressed(true);
            e.preventDefault();
        }
    }, [isDisabled, isReadonly]);
    const handleKeyUp = hooks.useCallback((e) => {
        if (isDisabled || isReadonly)
            return;
        if (e.code === 'Enter' && isEnterPressed) {
            setIsEnterPressed(false);
            e.preventDefault();
            e.stopPropagation();
            onChange?.(e);
        }
        if (e.code === 'Space' && isSpacePressed) {
            setIsSpacePressed(false);
            e.preventDefault();
            e.stopPropagation();
            onChange?.(e);
        }
    }, [isDisabled, isReadonly, isEnterPressed, isSpacePressed, onChange]);
    return {
        onClick: handleClick,
        onKeyDown: handleKeyDown,
        onKeyUp: handleKeyUp,
        onKeyPress: handleKeyPress
    };
}
/**
 * Prevent default on Space otherwise holding down the Space will cause the page to scroll to the bottom.
 */
function handleKeyPress(e) {
    if (e.code === 'Space') {
        e.preventDefault();
    }
}

/**
 * StyledSwitch is a controlled component that displays a stylized binary switch (on/off).
 * It supports accessibility attributes and allows parent components to control its checked and disabled states.
 *
 * This component is implemented with `forwardRef` to allow parent components to obtain a ref
 * to the interactive switch element (track). This is important for programmatically managing focus.
 *
 * @example
 *   <StyledSwitch isChecked={checked} onChange={handleChange} ref={switchRef} />
 */
const StyledSwitch = compat.forwardRef(({ 'aria-describedby': ariaDescribedBy, 'aria-invalid': ariaInvalid, 'aria-label': ariaLabel, 'aria-labelledby': ariaLabelledBy, baseId, isChecked = false, isDisabled, isReadonly, readonlyOnText, readonlyOffText, onChange, testId }, ref = null) => {
    const generatedBaseId = useId.useId();
    const readonlyBaseId = baseId ?? generatedBaseId;
    const readonlyId = `${readonlyBaseId}-readonly`;
    // Prefer isDisabled over isReadonly:
    const resolvedDisabled = !!isDisabled;
    const resolvedReadonly = !resolvedDisabled && !!isReadonly;
    const mergedAriaDescribedBy = resolvedReadonly
        ? stringUtils.l([ariaDescribedBy, readonlyId])
        : ariaDescribedBy;
    const ariaChecked = !!isChecked;
    const { tabbableModeProps } = useTabbableMode.useTabbableMode(resolvedDisabled);
    const { isActive, activeProps } = useActive.useActive({ isDisabled: isDisabled || isReadonly });
    const { classes: themeContainerClass } = useComponentTheme.useComponentTheme(PRIVATE_StyledSwitch_themes_redwood_StyledSwitchTheme.StyledSwitchRedwoodTheme, {
        target: 'container',
        checked: isChecked ? 'isChecked' : 'notChecked',
        disabled: resolvedDisabled ? 'isDisabled' : 'notDisabled',
        readonly: resolvedReadonly ? 'isReadonly' : 'notReadonly'
    });
    const containerClass = classNames.classNames([PRIVATE_StyledSwitch_themes_StyledSwitchStyles_css.styles.container, themeContainerClass]);
    const { classes: themeTrackClass } = useComponentTheme.useComponentTheme(PRIVATE_StyledSwitch_themes_redwood_StyledSwitchTheme.StyledSwitchRedwoodTheme, {
        target: 'track',
        checked: isChecked ? 'isChecked' : 'notChecked',
        disabled: resolvedDisabled ? 'isDisabled' : 'notDisabled',
        readonly: resolvedReadonly ? 'isReadonly' : 'notReadonly',
        active: isActive ? 'isActive' : 'notActive'
    });
    const trackClass = classNames.classNames([PRIVATE_StyledSwitch_themes_StyledSwitchStyles_css.styles.track, themeTrackClass]);
    const { classes: themeThumbClass } = useComponentTheme.useComponentTheme(PRIVATE_StyledSwitch_themes_redwood_StyledSwitchTheme.StyledSwitchRedwoodTheme, {
        target: 'thumb',
        checked: isChecked ? 'isChecked' : 'notChecked',
        disabled: resolvedDisabled ? 'isDisabled' : 'notDisabled',
        readonly: resolvedReadonly ? 'isReadonly' : 'notReadonly'
    });
    const thumbClass = classNames.classNames([PRIVATE_StyledSwitch_themes_StyledSwitchStyles_css.styles.thumb, themeThumbClass]);
    // Use the new hook for events
    const { onClick, onKeyDown, onKeyUp, onKeyPress } = useSwitchHandlers({
        isDisabled: resolvedDisabled,
        isReadonly: resolvedReadonly,
        onChange
    });
    let readonlyText = undefined;
    if (resolvedReadonly) {
        if (isChecked) {
            readonlyText = readonlyOnText;
        }
        else {
            readonlyText = readonlyOffText;
        }
    }
    const rootTestIdProps = useTestId.useTestId(testId);
    const thumbTestIdProps = useTestId.useTestId(testId ? testId + '_thumb' : undefined);
    const translations = useTranslationBundle.useTranslationBundle('@oracle/oraclejet-preact');
    const readonlyLabel = translations.formControl_readOnly();
    return (jsxRuntime.jsxs("div", { class: PRIVATE_StyledSwitch_themes_StyledSwitchStyles_css.styles.root, ...rootTestIdProps, children: [jsxRuntime.jsx("div", { class: containerClass, children: jsxRuntime.jsx("div", { role: "switch", ...tabbableModeProps, class: trackClass, "aria-checked": ariaChecked, "aria-readonly": resolvedReadonly || undefined, "aria-disabled": resolvedDisabled || undefined, "aria-invalid": ariaInvalid, "aria-label": ariaLabel, "aria-labelledby": ariaLabelledBy, "aria-describedby": mergedAriaDescribedBy, ref: ref, ...mergeProps.mergeProps(activeProps, { onKeyUp, onKeyDown, onClick, onKeyPress }), children: jsxRuntime.jsx("div", { class: thumbClass, ...thumbTestIdProps, children: readonlyText }) }) }), resolvedReadonly && (jsxRuntime.jsx(HiddenAccessible.HiddenAccessible, { id: readonlyId, isHidden: true, children: readonlyLabel }))] }));
});

exports.StyledSwitch = StyledSwitch;
