// @oracle/oraclejet-preact/cjs/TableColGroup-b7e1e296.js
/* @oracle/oraclejet-preact: undefined */
'use strict';

var hooks = require('preact/hooks');
var ScrollUtils = require('./ScrollUtils-e8076071.js');
var preact = require('preact');
var useResizeObserver = require('./useResizeObserver-a9e97180.js');
var unsafeDomAccess = require('./unsafeDomAccess-c3a492f0.js');
var useAddBusyState = require('./useAddBusyState-d19ae1fa.js');
var collectionUtils = require('./collectionUtils-f1d3183f.js');
var PRIVATE_Table_themes_Table_css = require('./PRIVATE_Table/themes/Table.css.js');
var PRIVATE_TableCommon_themes_Common_css = require('./PRIVATE_TableCommon/themes/Common.css.js');
var PRIVATE_Table_themes_TableCellStyles_css = require('./PRIVATE_Table/themes/TableCellStyles.css.js');
var PRIVATE_Table_themes_TableFooterCellStyles_css = require('./PRIVATE_Table/themes/TableFooterCellStyles.css.js');
var PRIVATE_Table_themes_TableHeaderCellStyles_css = require('./PRIVATE_Table/themes/TableHeaderCellStyles.css.js');
var jsxRuntime = require('preact/jsx-runtime');
var useTranslationBundle = require('./useTranslationBundle-12dd087d.js');
var classNames = require('./classNames-c14c6ef3.js');
var LoadMoreCollection = require('./LoadMoreCollection-eb1f2603.js');
var PRIVATE_Table_themes_TableSkeletonStyles_css = require('./PRIVATE_Table/themes/TableSkeletonStyles.css.js');
var dimensions$1 = require('./dimensions-6101ca1f.js');
var mergeInterpolations = require('./mergeInterpolations-6727b536.js');

const TABLE_PENDING_SELECTOR = 'data-oj-table-pending';
const TABLE_BODY_SELECTOR = 'data-oj-table-body';
const TABLE_NODATA_SELECTOR = 'data-oj-table-nodata';
const TABLE_SPACER_SELECTOR = 'data-oj-table-spacer';
const TABLE_LOADING_SELECTOR = 'data-oj-table-loading';
const TABLE_LOADMORE_SELECTOR = 'data-oj-table-loadmore';
const TABLE_FOCUSABLE_SELECTOR = 'data-oj-table-focusable';
const TABLE_DATA_SELECTOR = 'data-oj-table-data';
const TABLE_ROW_SELECTOR = 'data-oj-table-row';
const TABLE_HEADER_SELECTOR = 'data-oj-table-header';
const TABLE_HEADER_ROW_SELECTOR = 'data-oj-table-header-row';
const TABLE_FOOTER_SELECTOR = 'data-oj-table-footer';
const TABLE_FOOTER_ROW_SELECTOR = 'data-oj-table-footer-row';
/**
 * Returns whether or not auto focus is enabled on the specified cell
 */
const isAutoFocusCell = (rootElement, tableId, activeCell) => {
    const cellElement = getElementFromCell(rootElement, tableId, activeCell);
    return cellElement?.dataset['ojTableAutoFocus'] === 'enabled';
};
/**
 * Returns the key for the row at the given index
 */
const getRowKeyForRowIndex = (rootElement, tableId, index) => {
    const rowElements = rootElement.querySelectorAll(`[${TABLE_ROW_SELECTOR}='${tableId}']`);
    return rowElements.length > index
        ? ScrollUtils.getRowKey(rowElements[index])
        : undefined;
};
/**
 * Returns the index of the given row element
 */
const getRowIndexForRowElement = (rootElement, tableId, rowElement) => {
    return ScrollUtils.getRowIndexForRowElement(rootElement, tableId, rowElement, TABLE_ROW_SELECTOR);
};
/**
 * Returns the index of the row for the given key.
 */
function getRowIndexForRowKey(rootElement, tableId, rowKey) {
    const rowElements = rootElement.querySelectorAll(`[${TABLE_ROW_SELECTOR}='${tableId}']`);
    for (let i = 0; i < rowElements.length; i++) {
        if (rowKey === ScrollUtils.getRowKey(rowElements[i])) {
            return i;
        }
    }
    return undefined;
}
/**
 * Returns the column key for the given cell element
 */
const getColumnKey = (element) => {
    return element.dataset['ojColumnKey'];
};
/**
 * Returns the index of the given column header element or footer element
 */
const getColumnIndexForHeaderFooterElement = (rootElement, tableId, cellElement, isHeader) => {
    const columnElements = isHeader
        ? rootElement.querySelectorAll(`[${TABLE_HEADER_SELECTOR}='${tableId}']`)
        : rootElement.querySelectorAll(`[${TABLE_FOOTER_SELECTOR}='${tableId}']`);
    const index = Array.prototype.indexOf.call(columnElements, cellElement);
    return index > -1 ? index : undefined;
};
/**
 * Helper method to get the cell key definition from an element
 */
function logicalCellExtractor(element, tableId, columnsArray, cellSelector, onlyAncestors) {
    const selectorString = cellSelector != null ? cellSelector : `[${TABLE_FOCUSABLE_SELECTOR}='${tableId}']`;
    const startingElement = onlyAncestors ? element.parentElement : element;
    if (startingElement != null) {
        const cellElement = startingElement.closest(selectorString);
        if (cellElement != null) {
            const cellType = ScrollUtils.getCellType(cellElement);
            if (cellType === 'data') {
                return {
                    rowKey: collectionUtils.keyExtractor(cellElement, `[${TABLE_ROW_SELECTOR}='${tableId}']`),
                    columnKey: getColumnKey(cellElement),
                    type: cellType
                };
            }
            else if (cellType === 'header' || cellType === 'footer') {
                return { columnKey: getColumnKey(cellElement), type: cellType };
            }
            else if (cellType === 'noData' || cellType === 'pending') {
                return { type: cellType };
            }
            else if (cellType === 'loading' || cellType === 'loadMore') {
                const columnKey = getBoundaryColumnKey(columnsArray, true);
                if (columnKey != null) {
                    return { columnKey: columnKey, type: cellType };
                }
            }
        }
    }
    return undefined;
}
/**
 * Helper method to get an element from a cell descriptor
 */
function getElementFromCell(rootElement, tableId, cell) {
    if (cell.type === 'header') {
        // header case
        const headerElements = rootElement.querySelectorAll(`[${TABLE_HEADER_SELECTOR}='${tableId}']`);
        for (const headerElement of headerElements) {
            if (cell.columnKey === getColumnKey(headerElement)) {
                return headerElement;
            }
        }
    }
    else if (cell.type === 'footer') {
        // footer case
        const footerElements = rootElement.querySelectorAll(`[${TABLE_FOOTER_SELECTOR}='${tableId}']`);
        for (const footerElement of footerElements) {
            if (cell.columnKey === getColumnKey(footerElement)) {
                return footerElement;
            }
        }
    }
    else if (cell.type === 'data') {
        // data body case
        const rowElements = rootElement.querySelectorAll(`[${TABLE_ROW_SELECTOR}='${tableId}']`);
        for (const rowElement of rowElements) {
            if (cell.rowKey === ScrollUtils.getRowKey(rowElement)) {
                const cellElements = rowElement.querySelectorAll(`[${TABLE_FOCUSABLE_SELECTOR}='${tableId}']`);
                for (const cellElement of cellElements) {
                    if (cell.columnKey === getColumnKey(cellElement)) {
                        return cellElement;
                    }
                }
            }
        }
    }
    else if (cell.type === 'noData') {
        // no data case
        return getNoDataCellElement(rootElement, tableId);
    }
    else if (cell.type === 'pending') {
        // pending skeletons case
        return getPendingLayoutCellElement(rootElement, tableId);
    }
    else if (cell.type === 'loading') {
        // loading skeletons case
        return getLoadingCellElement(rootElement, tableId);
    }
    else if (cell.type === 'loadMore') {
        // loadMore skeletons case
        return getLoadMoreCellElement(rootElement, tableId);
    }
    return undefined;
}
/**
 * Returns the key of the first focusable row in the Table
 */
const getFirstRenderedRowKey = (rootElement, tableId) => {
    return ScrollUtils.getFirstRenderedRowKey(rootElement, tableId, TABLE_ROW_SELECTOR);
};
/**
 * Returns the key of the last focusable row in the Table
 */
const getLastRenderedRowKey = (rootElement, tableId) => {
    return ScrollUtils.getLastRenderedRowKey(rootElement, tableId, TABLE_ROW_SELECTOR);
};
/**
 * Helper function to get the first or last column key.
 */
const getBoundaryColumnKey = (columnsArray, isFirst) => {
    if (columnsArray.length > 0) {
        return isFirst ? columnsArray[0].key : columnsArray[columnsArray.length - 1].key;
    }
    return undefined;
};
/**
 * Helper method to get the 'pending layout' cell
 */
const getPendingLayoutCellElement = (rootElement, tableId) => {
    return ScrollUtils.getElementBySelector(rootElement, tableId, TABLE_PENDING_SELECTOR);
};
/**
 * Helper method to get the 'loading' cell
 */
const getLoadingCellElement = (rootElement, tableId) => {
    return ScrollUtils.getElementBySelector(rootElement, tableId, TABLE_LOADING_SELECTOR);
};
/**
 * Helper method to get the 'loadMore' cell
 */
const getLoadMoreCellElement = (rootElement, tableId) => {
    return ScrollUtils.getElementBySelector(rootElement, tableId, TABLE_LOADMORE_SELECTOR);
};
/**
 * Helper method to get the table body
 */
const getTableBodyElement = (rootElement, tableId) => {
    return ScrollUtils.getElementBySelector(rootElement, tableId, TABLE_BODY_SELECTOR);
};
/**
 * Helper method to get the 'noData' cell
 */
const getNoDataCellElement = (rootElement, tableId) => {
    return ScrollUtils.getElementBySelector(rootElement, tableId, TABLE_NODATA_SELECTOR);
};
/**
 * Helper method to get the 'spacer' row
 */
const getSpacerRowElement = (rootElement, tableId) => {
    return ScrollUtils.getElementBySelector(rootElement, tableId, TABLE_SPACER_SELECTOR);
};
/**
 * Helper method to get the footer row
 */
const getFooterRowElement = (rootElement, tableId) => {
    return ScrollUtils.getElementBySelector(rootElement, tableId, TABLE_FOOTER_ROW_SELECTOR);
};
/**
 * Helper method to get the header row
 */
const getHeaderRowElement = (rootElement, tableId) => {
    return ScrollUtils.getElementBySelector(rootElement, tableId, TABLE_HEADER_ROW_SELECTOR);
};
/**
 * Helper method to determine if the 'loadMore' cell is present in the Table
 */
const hasLoadMoreCell = (rootElement, tableId) => {
    return getLoadMoreCellElement(rootElement, tableId) != null;
};
/**
 * Helper method to determine if the 'noData' cell is present in the Table
 */
const hasNoDataCell = (rootElement, tableId) => {
    return getNoDataCellElement(rootElement, tableId) != null;
};
/**
 * Helper method to determine if footer cells are present in the Table
 */
const hasFooterCells = (rootElement, tableId) => {
    return getFooterRowElement(rootElement, tableId) != null;
};

/**
 * Helper method to return the current scrollbar height of the scroller element.
 */
const _getScrollbarHeight = (scrollerElement, sizingInfo) => {
    return sizingInfo.isInitialRender
        ? ScrollUtils.getScrollbarHeight(scrollerElement)
        : sizingInfo.boxHeight - sizingInfo.contentHeight;
};
/**
 * Helper method to return the current scrollbar width of the scroller element.
 */
const _getScrollbarWidth = (scrollerElement, viewportInfo, sizingInfo, overflowInfo) => {
    if (viewportInfo.isExternalScroller || overflowInfo.isVertical) {
        return sizingInfo.isInitialRender
            ? ScrollUtils.getScrollbarWidth(scrollerElement)
            : sizingInfo.boxWidth - sizingInfo.contentWidth;
    }
    return 0;
};
/**
 * Helper method to get the top scroller offset of the Table
 */
const _getScrollerTopOffset = (rootElement, viewportConfig, tableId) => {
    const headerRowElement = getHeaderRowElement(rootElement, tableId);
    const top = viewportConfig.scrollerOffsetTop ?? 0;
    return top + (headerRowElement != null ? headerRowElement.getBoundingClientRect().height : 0);
};
/**
 * Helper method to get the bottom scroller offset of the Table
 */
const _getScrollerBottomOffset = (rootElement, viewportConfig, tableId) => {
    const footerRowElement = getFooterRowElement(rootElement, tableId);
    const bottom = viewportConfig.scrollerOffsetBottom ?? 0;
    return bottom + (footerRowElement != null ? footerRowElement.getBoundingClientRect().height : 0);
};
/**
 * Helper method to get the start scroller offset of the Table
 */
const _getScrollerStartOffset = (viewportConfig) => {
    return viewportConfig.scrollerOffsetStart ?? 0;
};
/**
 * Helper method to get the end scroller offset of the Table
 */
const _getScrollerEndOffset = (viewportConfig) => {
    return viewportConfig.scrollerOffsetEnd ?? 0;
};
/**
 * Helper method to return the top position of the scroller element.
 */
const getScrollerTop = (rootElement, scrollerElement, viewportConfig, tableId) => {
    const scrollerTopOffset = _getScrollerTopOffset(rootElement, viewportConfig, tableId);
    return ScrollUtils.isHTMLScroller(scrollerElement)
        ? scrollerTopOffset
        : scrollerTopOffset + scrollerElement.getBoundingClientRect().top;
};
/**
 * Helper method to return the bottom position of the scroller element.
 */
const getScrollerBottom = (rootElement, scrollerElement, viewportConfig, sizingInfo, tableId) => {
    const bottomOffset = _getScrollerBottomOffset(rootElement, viewportConfig, tableId);
    if (ScrollUtils.isHTMLScroller(scrollerElement)) {
        return scrollerElement.clientHeight - bottomOffset;
    }
    const scrollbarHeight = _getScrollbarHeight(scrollerElement, sizingInfo);
    return scrollerElement.getBoundingClientRect().bottom - bottomOffset - scrollbarHeight;
};
/**
 * Helper method to return the start position of the scroller element.
 */
const getScrollerStart = (scrollerElement, viewportInfo, sizingInfo, overflowInfo, stickyOffset, isRtl) => {
    const startOffset = stickyOffset + _getScrollerStartOffset(viewportInfo.viewportConfig);
    if (isRtl) {
        if (ScrollUtils.isHTMLScroller(scrollerElement)) {
            const scrollbarOffset = ScrollUtils.isLeftHTMLScrollbar(scrollerElement)
                ? _getScrollbarWidth(scrollerElement, viewportInfo, sizingInfo, overflowInfo)
                : 0;
            return scrollerElement.clientWidth - startOffset + scrollbarOffset;
        }
        return scrollerElement.getBoundingClientRect().right - startOffset;
    }
    return ScrollUtils.isHTMLScroller(scrollerElement)
        ? startOffset
        : scrollerElement.getBoundingClientRect().left + startOffset;
};
/**
 * Helper method to return the end position of the scroller element.
 */
const getScrollerEnd = (scrollerElement, viewportInfo, sizingInfo, overflowInfo, stickyOffset, isRtl) => {
    const endOffset = stickyOffset + _getScrollerEndOffset(viewportInfo.viewportConfig);
    if (ScrollUtils.isHTMLScroller(scrollerElement)) {
        if (isRtl) {
            const scrollbarOffset = ScrollUtils.isLeftHTMLScrollbar(scrollerElement)
                ? _getScrollbarWidth(scrollerElement, viewportInfo, sizingInfo, overflowInfo)
                : 0;
            return endOffset + scrollbarOffset;
        }
        return scrollerElement.clientWidth - endOffset;
    }
    const scrollbarOffset = _getScrollbarWidth(scrollerElement, viewportInfo, sizingInfo, overflowInfo);
    return isRtl
        ? scrollerElement.getBoundingClientRect().left + endOffset + scrollbarOffset
        : scrollerElement.getBoundingClientRect().right - endOffset - scrollbarOffset;
};
/**
 * A helper function to make sure specified elem is visible in the specified container
 */
function _scrollColumnIntoView(columnElement, scrollerElement, viewportInfo, columnsArray, stickyColumnInfo, sizingInfo, overflowInfo, isRtl, appliedColumnWidths, location) {
    const columnKey = getColumnKey(columnElement);
    let columnIndex;
    for (let i = 0; i < columnsArray.length; i++) {
        if (columnsArray[i].key === columnKey) {
            columnIndex = i;
            break;
        }
    }
    if (columnIndex != null && appliedColumnWidths != null) {
        let startOffset = 0;
        let endOffset = 0;
        const borderOffset = appliedColumnWidths.borderOffset;
        const stickyColumns = stickyColumnInfo.stickyColumns;
        for (let i = 0; i < stickyColumns.length; i++) {
            const stickyIndex = stickyColumns[i];
            if (stickyIndex < columnIndex) {
                startOffset += appliedColumnWidths.realColumnWidths[stickyIndex];
            }
            else {
                break;
            }
        }
        // include the logical 'start' border of this column if owned by the previous column
        // if this column is sticky - otherwise that border should already be visible
        if (!appliedColumnWidths.borderEdges[columnIndex].hasStart &&
            !stickyColumns.includes(columnIndex)) {
            startOffset += borderOffset;
        }
        for (let i = stickyColumns.length - 1; i > -1; i--) {
            const stickyIndex = stickyColumns[i];
            if (stickyIndex > columnIndex) {
                endOffset += appliedColumnWidths.realColumnWidths[stickyIndex];
            }
            else {
                break;
            }
        }
        // include the logical 'end' border of this column if owned by the next column if
        // this column is sticky - otherwise that border should already be visible
        if (!appliedColumnWidths.borderEdges[columnIndex].hasEnd &&
            !stickyColumns.includes(columnIndex)) {
            endOffset += borderOffset;
        }
        const horizontalOverflowDiff = ScrollUtils.getHorizontalOverflowDiff(columnElement, getScrollerStart(scrollerElement, viewportInfo, sizingInfo, overflowInfo, startOffset, isRtl), getScrollerEnd(scrollerElement, viewportInfo, sizingInfo, overflowInfo, endOffset, isRtl), isRtl);
        const leftOverflowDiff = horizontalOverflowDiff.left;
        const hasLeftOverflow = leftOverflowDiff > 0;
        const rightOverflowDiff = horizontalOverflowDiff.right;
        const hasRightOverflow = rightOverflowDiff > 0;
        // don't adjust scroll position if row has overflow in both directions
        if (hasLeftOverflow && hasRightOverflow) {
            return;
        }
        if (location == null || location === 'inView') {
            // if column fits fully in viewport, scroll overflow side into view
            // otherwise, scroll the shortest amount to fill viewport with column
            if (hasLeftOverflow) {
                if (Math.abs(rightOverflowDiff) > Math.abs(leftOverflowDiff)) {
                    scrollerElement.scrollLeft = _roundScrollValue(scrollerElement.scrollLeft - leftOverflowDiff, false);
                }
                else {
                    scrollerElement.scrollLeft = _roundScrollValue(scrollerElement.scrollLeft + rightOverflowDiff, true);
                }
            }
            else if (hasRightOverflow) {
                if (Math.abs(leftOverflowDiff) > Math.abs(rightOverflowDiff)) {
                    scrollerElement.scrollLeft = _roundScrollValue(scrollerElement.scrollLeft + rightOverflowDiff, true);
                }
                else {
                    scrollerElement.scrollLeft = _roundScrollValue(scrollerElement.scrollLeft - leftOverflowDiff, false);
                }
            }
        }
        else if (location === 'start') {
            scrollerElement.scrollLeft = _roundScrollValue(scrollerElement.scrollLeft - leftOverflowDiff, false);
        }
        else {
            scrollerElement.scrollLeft = _roundScrollValue(scrollerElement.scrollLeft + rightOverflowDiff, true);
        }
    }
}
/**
 * A helper function to make sure specified elem is visible in the specified container
 */
const _scrollRowIntoView = (rootElement, scrollerElement, viewportInfo, sizingInfo, tableId, rowElement, location) => {
    const verticalOverflowDiff = ScrollUtils.getVerticalOverflowDiff(rowElement, getScrollerTop(rootElement, scrollerElement, viewportInfo.viewportConfig, tableId), getScrollerBottom(rootElement, scrollerElement, viewportInfo.viewportConfig, sizingInfo, tableId));
    const topOverflowDiff = verticalOverflowDiff.top;
    const hasTopOverflow = topOverflowDiff > 0;
    const bottomOverflowDiff = verticalOverflowDiff.bottom;
    const hasBottomOverflow = bottomOverflowDiff > 0;
    // don't adjust scroll position if row has overflow in both directions
    if (hasTopOverflow && hasBottomOverflow) {
        return;
    }
    if (location == null || location === 'inView') {
        // if row fits fully in viewport, scroll overflow side into view
        // otherwise, scroll the shortest amount to fill viewport with row
        if (hasBottomOverflow) {
            if (Math.abs(topOverflowDiff) > Math.abs(bottomOverflowDiff)) {
                scrollerElement.scrollTop = _roundScrollValue(scrollerElement.scrollTop + bottomOverflowDiff, true);
            }
            else {
                scrollerElement.scrollTop = _roundScrollValue(scrollerElement.scrollTop - topOverflowDiff, false);
            }
        }
        else if (hasTopOverflow) {
            if (Math.abs(bottomOverflowDiff) > Math.abs(topOverflowDiff)) {
                scrollerElement.scrollTop = _roundScrollValue(scrollerElement.scrollTop - topOverflowDiff, false);
            }
            else {
                scrollerElement.scrollTop = _roundScrollValue(scrollerElement.scrollTop + bottomOverflowDiff, true);
            }
        }
    }
    else if (location === 'top') {
        scrollerElement.scrollTop = _roundScrollValue(scrollerElement.scrollTop - topOverflowDiff, false);
    }
    else {
        scrollerElement.scrollTop = _roundScrollValue(scrollerElement.scrollTop + bottomOverflowDiff, true);
    }
};
/**
 * Helper method to round a desired scrollTop or scrollLeft value to the nearest valid value
 * depending on the client browser.
 */
const _roundScrollValue = (value, isOver) => {
    // use Math.floor() and Math.ceil() to ensure we round to nearest containing valid
    // pixel value in either direction as valid scrollTop values are only integers
    return isOver ? Math.ceil(value) : Math.floor(value);
};
/**
 * Helper to set the horizontal scroll position on the Table.
 */
function setHorizontalScrollPosition(rootElement, scrollerElement, viewportInfo, sizingInfo, overflowInfo, columnsArray, fullColumnWidths, isRtl, horizontalScrollPosition) {
    const x = _getScrollLeftFromScrollPosition(rootElement, scrollerElement, viewportInfo, sizingInfo, overflowInfo, columnsArray, fullColumnWidths, isRtl, horizontalScrollPosition);
    scrollerElement.scrollLeft = x ?? 0;
}
/**
 * Helper to set the vertical scroll position on the Table.
 */
function setVerticalScrollPosition(rootElement, scrollerElement, viewportInfo, tableId, verticalScrollPosition) {
    const y = _getScrollTopFromScrollPosition(rootElement, scrollerElement, viewportInfo, tableId, verticalScrollPosition);
    scrollerElement.scrollTop = y ?? 0;
}
/**
 * Helper to determine the 'x' value of a given scroll position.
 */
function _getScrollLeftFromScrollPosition(rootElement, scrollerElement, viewportInfo, sizingInfo, overflowInfo, columnsArray, fullColumnWidths, isRtl, scrollPosition) {
    if (scrollPosition != null) {
        let columnKey;
        let useColumn = false;
        if (scrollPosition.columnKey != null) {
            // columnKey takes precedence over everything else
            useColumn = true;
            columnKey = scrollPosition.columnKey;
        }
        else if (scrollPosition.offsetX != null && scrollPosition.x == null) {
            // offsetX can be used on its own if it is the only horizontal aspect provided
            useColumn = true;
            const currentHorizontalScrollPosition = getHorizontalScrollPosition(rootElement, scrollerElement, viewportInfo, sizingInfo, overflowInfo, columnsArray, fullColumnWidths, isRtl);
            columnKey = currentHorizontalScrollPosition.columnKey;
        }
        if (useColumn) {
            if (columnKey != null) {
                let offsetX;
                let columnEnd;
                const scrollerStart = getScrollerStart(scrollerElement, viewportInfo, sizingInfo, overflowInfo, 0, isRtl);
                if (isRtl) {
                    offsetX = -(scrollPosition.offsetX ?? 0);
                    columnEnd = rootElement.getBoundingClientRect().right;
                }
                else {
                    offsetX = scrollPosition.offsetX ?? 0;
                    columnEnd = rootElement.getBoundingClientRect().left;
                }
                for (let i = 0; i < columnsArray.length; i++) {
                    const column = columnsArray[i];
                    if (column.key === columnKey) {
                        return scrollerElement.scrollLeft - scrollerStart + columnEnd + offsetX;
                    }
                    if (column.value.sticky !== 'enabled') {
                        columnEnd += isRtl ? -fullColumnWidths[i] : fullColumnWidths[i];
                    }
                }
            }
        }
        else if (scrollPosition.x != null) {
            return scrollPosition.x;
        }
    }
    // either nothing was provided, or what was provided was invalid
    return undefined;
}
/**
 * Helper to determine the 'y' value of a given scroll position.
 */
function _getScrollTopFromScrollPosition(rootElement, scrollerElement, viewportInfo, tableId, scrollPosition) {
    if (scrollPosition != null) {
        let rowKey;
        let useRow = false;
        if (scrollPosition.rowKey != null) {
            // rowKey takes precedence over everything else
            useRow = true;
            rowKey = scrollPosition.rowKey;
        }
        else if (scrollPosition.offsetY != null && scrollPosition.y == null) {
            // offsetY can be used on its own if it is the only vertical aspect provided
            useRow = true;
            rowKey = getVerticalScrollPosition(rootElement, scrollerElement, viewportInfo, tableId).rowKey;
        }
        if (useRow) {
            if (rowKey != null) {
                const scrollerTop = getScrollerTop(rootElement, scrollerElement, viewportInfo.viewportConfig, tableId);
                const offsetY = scrollPosition.offsetY ?? 0;
                const rowElements = rootElement.querySelectorAll(`[${TABLE_ROW_SELECTOR}='${tableId}']`);
                for (let i = 0; i < rowElements.length; i++) {
                    if (rowKey === ScrollUtils.getRowKey(rowElements[i])) {
                        const rowTop = rowElements[i].getBoundingClientRect().top;
                        return scrollerElement.scrollTop + rowTop - scrollerTop + offsetY;
                    }
                }
            }
        }
        else if (scrollPosition.y != null) {
            return scrollPosition.y;
        }
    }
    // either nothing was provided, or what was provided was invalid
    return undefined;
}
/**
 * Helper to get the logical scroll position for the given scrollLeft value. If no scrollLeft value is given,
 * it will use the current scrollLeft value of the Table.
 */
function getHorizontalScrollPosition(rootElement, scrollerElement, viewportInfo, sizingInfo, overflowInfo, columnsArray, fullColumnWidths, isRtl) {
    // ensure scroll position x value is initialized
    const scrollPosition = {
        x: Math.abs(scrollerElement.scrollLeft),
        columnKey: undefined,
        offsetX: undefined
    };
    let value1;
    let value2;
    if (viewportInfo.isExternalScroller) {
        const scrollerStart = getScrollerStart(scrollerElement, viewportInfo, sizingInfo, overflowInfo, 0, isRtl);
        if (isRtl) {
            value1 = rootElement.getBoundingClientRect().right;
            value2 = scrollerStart;
        }
        else {
            value1 = scrollerStart;
            value2 = rootElement.getBoundingClientRect().left;
        }
    }
    else if (!isRtl) {
        value1 = scrollerElement.scrollLeft;
        value2 = 0;
    }
    else {
        value1 = 0;
        value2 = scrollerElement.scrollLeft;
    }
    for (let i = 0; i < columnsArray.length; i++) {
        const columnWidth = fullColumnWidths[i];
        value1 -= columnWidth;
        const column = columnsArray[i];
        const isSticky = column.value.sticky === 'enabled';
        if ((!isSticky && value2 > value1) ||
            (isSticky && value2 >= value1 + columnWidth) ||
            i === columnsArray.length - 1) {
            scrollPosition.columnKey = column.key;
            scrollPosition.offsetX = value1 - value2 + columnWidth;
            break;
        }
        if (isSticky) {
            value2 -= columnWidth;
        }
    }
    return scrollPosition;
}
/**
 * Helper to get the logical scroll position for the given scrollTop value. If no scrollTop value is given,
 * it will use the current scrollTop value of the Table.
 */
function getVerticalScrollPosition(rootElement, scrollerElement, viewportInfo, tableId) {
    // ensure scroll position y value is initialized
    const scrollPosition = {
        y: Math.abs(scrollerElement.scrollTop),
        rowKey: undefined,
        offsetY: undefined
    };
    const scrollerTop = getScrollerTop(rootElement, scrollerElement, viewportInfo.viewportConfig, tableId);
    const rowElements = rootElement.querySelectorAll(`[${TABLE_ROW_SELECTOR}='${tableId}']`);
    for (let i = 0; i < rowElements.length; i++) {
        const rowRect = rowElements[i].getBoundingClientRect();
        if (rowRect.bottom > scrollerTop || i === rowElements.length - 1) {
            scrollPosition.rowKey = getRowKeyForRowIndex(rootElement, tableId, i);
            scrollPosition.offsetY = scrollerTop - rowRect.top;
            break;
        }
    }
    return scrollPosition;
}
function applyScrollDetail(rootElement, scrollerElement, viewportInfo, tableId, columnsArray, stickyColumnInfo, sizingInfo, overflowInfo, scrollDetail, isRtl, appliedColumnWidths) {
    if (!sizingInfo.isInitialRender) {
        if (scrollDetail.cell.type === 'data') {
            const rowElement = collectionUtils.findElementByKey(rootElement, scrollDetail.cell.rowKey, `[${TABLE_ROW_SELECTOR}='${tableId}']`);
            if (rowElement != null && scrollDetail.locationY != null) {
                _scrollRowIntoView(rootElement, scrollerElement, viewportInfo, sizingInfo, tableId, rowElement, scrollDetail.locationY);
            }
        }
        if (scrollDetail.cell.type === 'data' ||
            scrollDetail.cell.type === 'header' ||
            scrollDetail.cell.type === 'footer') {
            const cellElement = getElementFromCell(rootElement, tableId, scrollDetail.cell);
            if (cellElement != null && scrollDetail.locationX != null) {
                _scrollColumnIntoView(cellElement, scrollerElement, viewportInfo, columnsArray, stickyColumnInfo, sizingInfo, overflowInfo, isRtl, appliedColumnWidths, scrollDetail.locationX);
            }
        }
        else if (scrollDetail.cell.type === 'loadMore' || scrollDetail.cell.type === 'loading') {
            const cellElement = getElementFromCell(rootElement, tableId, scrollDetail.cell);
            if (cellElement != null && scrollDetail.locationY != null) {
                _scrollRowIntoView(rootElement, scrollerElement, viewportInfo, sizingInfo, tableId, cellElement, scrollDetail.cell.type === 'loading' && scrollDetail.locationY === 'inView'
                    ? 'top'
                    : scrollDetail.locationY);
            }
        }
    }
}
/**
 * Helper function to update the 'left' and 'right' values of all sticky columns to correspond
 * to the given sizing information.
 */
function updateStickyColumns(rootElement, scrollerElement, viewportInfo, tableId, isRtl, columnsArray, stickyColumnInfo, sizingInfo, overflowInfo, fullColumnWidths, stickyEdgesState, setStickyEdgesState, isHorizontalScroll) {
    let stickyIndex;
    let stickyStartOffset = 0;
    let stickyEndOffset = 0;
    if (viewportInfo.isExternalScroller) {
        const scrollerOffsetStart = viewportInfo.viewportConfig.scrollerOffsetStart;
        if (scrollerOffsetStart != null) {
            stickyStartOffset = scrollerOffsetStart;
        }
        const scrollerOffsetEnd = viewportInfo.viewportConfig.scrollerOffsetEnd;
        if (scrollerOffsetEnd != null) {
            stickyEndOffset = scrollerOffsetEnd;
        }
    }
    const stickyIndicies = stickyColumnInfo.stickyColumns;
    for (let i = 0; i < stickyIndicies.length; i++) {
        stickyIndex = stickyIndicies[i];
        _applyStickyColumnOffset(rootElement, tableId, isRtl, stickyIndex, stickyStartOffset, true);
        stickyStartOffset += fullColumnWidths[stickyIndex];
    }
    for (let i = stickyIndicies.length - 1; i > -1; i--) {
        stickyIndex = stickyIndicies[i];
        _applyStickyColumnOffset(rootElement, tableId, isRtl, stickyIndex, stickyEndOffset, false);
        stickyEndOffset += fullColumnWidths[stickyIndex];
    }
    // only apply sticky edge styling if horizontal overflow is present
    if (isHorizontalScroll) {
        updateStickyEdges(rootElement, scrollerElement, viewportInfo, columnsArray, stickyColumnInfo, sizingInfo, overflowInfo, fullColumnWidths, isRtl, stickyEdgesState, setStickyEdgesState);
    }
    else if (stickyEdgesState.start != null || stickyEdgesState.end != null) {
        setStickyEdgesState({});
    }
}
/**
 * Helper function to get the sticky column indicies.
 */
const getStickyColumnIndicies = (columnsArray) => {
    const stickyColumns = [];
    for (let i = 0; i < columnsArray.length; i++) {
        if (columnsArray[i].value.sticky === 'enabled') {
            stickyColumns.push(i);
        }
    }
    return stickyColumns;
};
/**
 * Helper function to update the 'left' and 'right' values of a specific column.
 */
const _applyStickyColumnOffset = (rootElement, tableId, isRtl, columnIndex, stickyOffset, isStart) => {
    const styleProperty = (isStart && !isRtl) || (!isStart && isRtl) ? 'left' : 'right';
    const styleValue = `${stickyOffset}px`;
    const headerElements = rootElement.querySelectorAll(`[${TABLE_HEADER_SELECTOR}='${tableId}']`);
    const headerElement = headerElements[columnIndex];
    headerElement.style[styleProperty] = styleValue;
    const rowElements = rootElement.querySelectorAll(`[${TABLE_ROW_SELECTOR}='${tableId}']`);
    for (let j = 0; j < rowElements.length; j++) {
        const rowElement = rowElements[j];
        const cellElements = rowElement.querySelectorAll(`[${TABLE_DATA_SELECTOR}='${tableId}']`);
        const cellElement = cellElements[columnIndex];
        cellElement.style[styleProperty] = styleValue;
    }
    const footerElements = rootElement.querySelectorAll(`[${TABLE_FOOTER_SELECTOR}='${tableId}']`);
    if (footerElements.length > 0) {
        const footerElement = footerElements[columnIndex];
        footerElement.style[styleProperty] = styleValue;
    }
};
/**
 * Helper function to update the sticky edges based on the given scrollLeft value.
 */
function updateStickyEdges(rootElement, scrollerElement, viewportInfo, columnsArray, stickyColumnInfo, sizingInfo, overflowInfo, fullColumnWidths, isRtl, stickyEdgesState, setStickyEdgesState) {
    if (columnsArray.length > 0) {
        let i;
        let currentScrollIndex;
        const newScrollPosition = getHorizontalScrollPosition(rootElement, scrollerElement, viewportInfo, sizingInfo, overflowInfo, columnsArray, fullColumnWidths, isRtl);
        for (i = 0; i < columnsArray.length; i++) {
            if (columnsArray[i].key === newScrollPosition.columnKey) {
                currentScrollIndex = i;
                break;
            }
        }
        let startIndex;
        const stickyIndicies = stickyColumnInfo.stickyColumns;
        // browser zoom levels cause rounding issues where the start scroll
        // position may never be reached, but it should always be within 1
        if (newScrollPosition.x >= 1) {
            for (i = 0; i < stickyIndicies.length; i++) {
                const currentIndex = stickyIndicies[i];
                if (currentIndex < currentScrollIndex) {
                    startIndex = currentIndex;
                }
                else {
                    break;
                }
            }
        }
        let endIndex;
        let currentWidth = 0;
        const maxScrollPos = scrollerElement.scrollWidth - scrollerElement.clientWidth;
        const endOverflow = maxScrollPos - newScrollPosition.x;
        // browser zoom levels cause rounding issues where the max scroll
        // position may never be reached, but it should always be within 1
        if (endOverflow >= 1) {
            for (let i = columnsArray.length - 1; i > -1; i--) {
                if (stickyIndicies.indexOf(i) !== -1) {
                    endIndex = i;
                }
                else {
                    currentWidth += fullColumnWidths[i];
                    if (currentWidth > endOverflow) {
                        break;
                    }
                }
            }
        }
        if (stickyEdgesState.start != startIndex || stickyEdgesState.end != endIndex) {
            setStickyEdgesState({ start: startIndex, end: endIndex });
        }
    }
}
/**
 * Helper function to hide a currently shown tooltip once the anchor cell scrolls out of view.
 */
function hideOutOfViewTooltip(rootElement, scrollerElement, viewportInfo, tableId, stickyColumnInfo, sizingInfo, overflowInfo, fullColumnWidths, isRtl, tooltipAnchorRef, hideTooltip) {
    if (tooltipAnchorRef?.current != null && hideTooltip != null) {
        const verticalOverflowDiff = ScrollUtils.getVerticalOverflowDiff(tooltipAnchorRef.current, getScrollerTop(rootElement, scrollerElement, viewportInfo.viewportConfig, tableId), getScrollerBottom(rootElement, scrollerElement, viewportInfo.viewportConfig, sizingInfo, tableId));
        const topOverflowDiff = verticalOverflowDiff.top;
        const hasTopOverflow = topOverflowDiff >= 0;
        const bottomOverflowDiff = verticalOverflowDiff.bottom;
        const hasBottomOverflow = bottomOverflowDiff >= 0;
        if ((hasTopOverflow && Math.abs(topOverflowDiff) > tooltipAnchorRef.current.offsetHeight) ||
            (hasBottomOverflow && Math.abs(bottomOverflowDiff) > tooltipAnchorRef.current.offsetHeight)) {
            hideTooltip(true);
            return;
        }
        let startOffset = 0;
        let endOffset = 0;
        const stickyColumns = stickyColumnInfo.stickyColumns;
        for (let i = 0; i < stickyColumns.length; i++) {
            const stickyIndex = stickyColumns[i];
            startOffset += fullColumnWidths[stickyIndex];
        }
        for (let i = stickyColumns.length - 1; i > -1; i--) {
            const stickyIndex = stickyColumns[i];
            endOffset += fullColumnWidths[stickyIndex];
        }
        const horizontalOverflowDiff = ScrollUtils.getHorizontalOverflowDiff(tooltipAnchorRef.current, getScrollerStart(scrollerElement, viewportInfo, sizingInfo, overflowInfo, startOffset, isRtl), getScrollerEnd(scrollerElement, viewportInfo, sizingInfo, overflowInfo, endOffset, isRtl), isRtl);
        const leftOverflowDiff = horizontalOverflowDiff.left;
        const hasLeftOverflow = leftOverflowDiff >= 0;
        const rightOverflowDiff = horizontalOverflowDiff.right;
        const hasRightOverflow = rightOverflowDiff >= 0;
        if ((hasLeftOverflow && Math.abs(leftOverflowDiff) > tooltipAnchorRef.current.offsetWidth) ||
            (hasRightOverflow && Math.abs(rightOverflowDiff) > tooltipAnchorRef.current.offsetWidth)) {
            hideTooltip(true);
        }
    }
}
/**
 * Helper function to reset the scroll position of the scroller element
 */
const resetVerticalScrollPosition = (scrollerElement) => {
    scrollerElement.scrollTop = 0;
};
/**
 * Helper function to get the viewport config.
 */
const getViewportConfig = (scrollerRef, config) => {
    return config ?? { scrollerRef };
};
/**
 * Helper function to update the scroller element's styling based on overflow status.
 */
const applyOverflowStatus = (rootElement, scrollerElement, tableId, contentWidth, availableWidth, contentHeight, availableHeight, defaultScrollbarWidth, overflowInfoRef, setHasVerticalUnderflow) => {
    // check for at least 0.5px difference to account for browser rounding issues
    const hasHorizontalOverflow = contentWidth - availableWidth > 0.5;
    const hasVerticalOverflow = contentHeight - availableHeight > 0.5;
    if (hasHorizontalOverflow) {
        scrollerElement.classList.remove(PRIVATE_Table_themes_Table_css.scrollerStyles.noHorizontalScroll);
        scrollerElement.classList.add(PRIVATE_Table_themes_Table_css.scrollerStyles.horizontalScroll);
    }
    else {
        scrollerElement.classList.remove(PRIVATE_Table_themes_Table_css.scrollerStyles.horizontalScroll);
        scrollerElement.classList.add(PRIVATE_Table_themes_Table_css.scrollerStyles.noHorizontalScroll);
    }
    if (hasVerticalOverflow) {
        scrollerElement.classList.remove(PRIVATE_Table_themes_Table_css.scrollerStyles.noVerticalScroll);
        scrollerElement.classList.add(PRIVATE_Table_themes_Table_css.scrollerStyles.verticalScroll);
        scrollerElement.style.paddingInlineEnd = '';
    }
    else {
        scrollerElement.classList.remove(PRIVATE_Table_themes_Table_css.scrollerStyles.verticalScroll);
        scrollerElement.classList.add(PRIVATE_Table_themes_Table_css.scrollerStyles.noVerticalScroll);
        scrollerElement.style.paddingInlineEnd = `${defaultScrollbarWidth}px`;
    }
    overflowInfoRef.current.isHorizontal = hasHorizontalOverflow;
    overflowInfoRef.current.isVertical = hasVerticalOverflow;
    applyUnderflowStatus(rootElement, tableId, setHasVerticalUnderflow);
};
/**
 * Helper function to update the undeflow status of the Table based on the spacer element sizing.
 */
const applyUnderflowStatus = (rootElement, tableId, setHasVerticalUnderflow) => {
    const spacerRowElement = getSpacerRowElement(rootElement, tableId);
    if (spacerRowElement != null) {
        setHasVerticalUnderflow(spacerRowElement.offsetHeight > 0);
    }
};

/**
 * Hook that manages handling scroll events on the Preact Table
 */
function useScrollHandling({ rootRef, viewportInfo, tableId, isLayoutRequired, sizingInfoRef, overflowInfoRef, columnsArray, stickyColumnInfo, stickyEdgesState, setStickyEdgesState, fullColumnWidths, horizontalScrollPositionOverride, verticalScrollPositionOverride, tooltipAnchorRef, hideTooltip, isPendingLayout, isRtl }) {
    const horizontalScrollPositionOverrideRef = hooks.useRef();
    const verticalScrollPositionOverrideRef = hooks.useRef();
    // update horizontal and/or vertical scroll position if a new value is provided
    // NO DEPENDENCY ARRAY AS THIS SHOULD RUN ON EVERY RENDER
    hooks.useLayoutEffect(() => {
        const rootElement = rootRef.current;
        const scrollerElement = viewportInfo.viewportConfig.scrollerRef.current;
        if (rootElement != null &&
            scrollerElement != null &&
            !isPendingLayout &&
            !isLayoutRequired &&
            fullColumnWidths != null) {
            if (horizontalScrollPositionOverrideRef.current !== horizontalScrollPositionOverride) {
                setHorizontalScrollPosition(rootElement, scrollerElement, viewportInfo, sizingInfoRef.current, overflowInfoRef.current, columnsArray, fullColumnWidths, isRtl, horizontalScrollPositionOverride);
                horizontalScrollPositionOverrideRef.current = horizontalScrollPositionOverride;
            }
            if (verticalScrollPositionOverrideRef.current !== verticalScrollPositionOverride) {
                setVerticalScrollPosition(rootElement, scrollerElement, viewportInfo, tableId, verticalScrollPositionOverride);
                verticalScrollPositionOverrideRef.current = verticalScrollPositionOverride;
            }
        }
    });
    // Scroll hander for the outer Table. It ensures the Table's sticky edges are updated
    // and any out of view tooltips are hidden.
    const _scrollHandler = hooks.useCallback(() => {
        const rootElement = rootRef.current;
        const scrollerElement = viewportInfo.viewportConfig.scrollerRef.current;
        if (rootElement != null && scrollerElement != null && fullColumnWidths != null) {
            updateStickyEdges(rootElement, scrollerElement, viewportInfo, columnsArray, stickyColumnInfo, sizingInfoRef.current, overflowInfoRef.current, fullColumnWidths, isRtl, stickyEdgesState, setStickyEdgesState);
            hideOutOfViewTooltip(rootElement, scrollerElement, viewportInfo, tableId, stickyColumnInfo, sizingInfoRef.current, overflowInfoRef.current, fullColumnWidths, isRtl, tooltipAnchorRef, hideTooltip);
        }
    }, [
        rootRef,
        viewportInfo,
        tableId,
        sizingInfoRef,
        overflowInfoRef,
        columnsArray,
        stickyColumnInfo,
        stickyEdgesState,
        setStickyEdgesState,
        fullColumnWidths,
        tooltipAnchorRef,
        hideTooltip,
        isRtl
    ]);
    hooks.useLayoutEffect(() => {
        const isExternalScroller = viewportInfo.isExternalScroller;
        const scrollerElement = viewportInfo.viewportConfig.scrollerRef.current;
        let scrollerEventElement;
        if (isExternalScroller && scrollerElement != null) {
            if (ScrollUtils.isHTMLScroller(scrollerElement)) {
                scrollerEventElement = window;
            }
            else {
                scrollerEventElement = scrollerElement;
            }
            scrollerEventElement.addEventListener('scroll', _scrollHandler);
        }
        return () => {
            if (isExternalScroller && scrollerEventElement != null) {
                scrollerEventElement.removeEventListener('scroll', _scrollHandler);
            }
        };
    }, [viewportInfo, _scrollHandler]);
    return viewportInfo.isExternalScroller ? {} : { onScroll: _scrollHandler };
}

/**
 * Helper method to extract the individual column widths of the Table
 * depending on its layout. With a 'fixed' layout, an initial 'availableWidth'
 * is required, while with a 'contents' layout, -1 can be provided as the
 * 'availableWidth' initially. This will trigger the layout logic to ensure
 * a non-contstrained Table width is supported, and is needed to match
 * current Redwood behaviors with a 'contents' layout.
 */
function applyLayoutColumnWidths(scrollerElement, viewportInfo, innerTableElement, widthOffset, tableId, isFixed, sampledColumnNetWidthsRef, availableRealWidth, columnsArray, columnBorderEdges, borderOffset, staticColumnWidths) {
    let newNetColumnWidths;
    let totalRealWidth = 0;
    let realWidthsArray = [];
    const tableCols = innerTableElement.querySelectorAll(`[data-oj-table-col='${tableId}']`);
    let totalBorderWidth = 0;
    for (const borderEdge of columnBorderEdges) {
        totalBorderWidth += borderEdge.width;
    }
    const availableNetWidth = availableRealWidth !== -1 ? availableRealWidth - totalBorderWidth : -1;
    if (isFixed) {
        newNetColumnWidths = _getFixedLayoutColumnNetWidths(availableNetWidth, columnsArray, staticColumnWidths);
        for (let i = 0; i < columnsArray.length; i++) {
            const width = newNetColumnWidths[i] + columnBorderEdges[i].width;
            realWidthsArray.push(width);
            totalRealWidth += width;
        }
        _updateTableSizing(innerTableElement, tableCols, realWidthsArray, totalRealWidth);
        return _createColumnWidths(columnsArray, newNetColumnWidths, realWidthsArray, columnBorderEdges, borderOffset, totalBorderWidth);
    }
    const headerElements = innerTableElement.querySelectorAll(`[${TABLE_HEADER_SELECTOR}='${tableId}']`);
    newNetColumnWidths = _getContentsLayoutColumnNetWidths(innerTableElement, headerElements, sampledColumnNetWidthsRef, availableNetWidth, columnsArray, columnBorderEdges, staticColumnWidths);
    for (let i = 0; i < columnsArray.length; i++) {
        const realWidth = newNetColumnWidths[i] + columnBorderEdges[i].width;
        realWidthsArray.push(realWidth);
        totalRealWidth += realWidth;
    }
    _updateTableSizing(innerTableElement, tableCols, realWidthsArray, totalRealWidth);
    if (availableRealWidth === -1) {
        // for content sizing, check to see if all space is actually utilized
        // this supports no width constraint scenarios
        if (!viewportInfo.isExternalScroller) {
            availableRealWidth = scrollerElement.getBoundingClientRect().width - widthOffset;
        }
        else {
            availableRealWidth = scrollerElement.clientWidth - widthOffset;
        }
        if (Math.abs(totalRealWidth - availableRealWidth) > 0.005) {
            newNetColumnWidths = _getContentsLayoutColumnNetWidths(innerTableElement, headerElements, sampledColumnNetWidthsRef, availableRealWidth - totalBorderWidth, columnsArray, columnBorderEdges, staticColumnWidths);
            totalRealWidth = 0;
            realWidthsArray = [];
            for (let i = 0; i < columnsArray.length; i++) {
                const width = newNetColumnWidths[i] + columnBorderEdges[i].width;
                realWidthsArray.push(width);
                totalRealWidth += width;
            }
            _updateTableSizing(innerTableElement, tableCols, realWidthsArray, totalRealWidth);
        }
    }
    return _createColumnWidths(columnsArray, newNetColumnWidths, realWidthsArray, columnBorderEdges, borderOffset, totalBorderWidth);
}
/**
 * Helper method to extract the individual column widths of the Table
 * for a 'fixed' layout table that already has its root width applied.
 */
function _getFixedLayoutColumnNetWidths(availableNetWidth, columnsArray, staticColumnWidths) {
    const DEFAULT_COLUMN_WEIGHT = 1;
    const DEFAULT_COLUMN_MIN_WIDTH = 100;
    const DEFAULT_COLUMN_MAX_WIDTH = undefined;
    let requiresActiveSizing = false;
    let totalWorkingWeight = 0;
    let forcedTotalNetWidth = 0;
    let pendingTotalNetWidth = 0;
    const columnMaxWidthsArray = [];
    const columnWeightsArray = [];
    const forcedColumnNetWidthsArray = [];
    const newColumnNetWidthsArray = [];
    const appliedColumnNetWidthsArray = [];
    for (let i = 0; i < columnsArray.length; i++) {
        const column = columnsArray[i];
        let columnMinWidth = column.value.minWidth;
        if (columnMinWidth == null || columnMinWidth <= 0) {
            columnMinWidth = DEFAULT_COLUMN_MIN_WIDTH;
        }
        let columnMaxWidth = column.value.maxWidth;
        if (columnMaxWidth == null || columnMaxWidth < columnMinWidth) {
            columnMaxWidth = DEFAULT_COLUMN_MAX_WIDTH;
        }
        columnMaxWidthsArray.push(columnMaxWidth);
        let columnWeight = column.value.weight;
        if (columnWeight == null || columnWeight < 1) {
            columnWeight = DEFAULT_COLUMN_WEIGHT;
        }
        columnWeightsArray.push(columnWeight);
        const columnWidth = staticColumnWidths != null ? staticColumnWidths[column.key] : undefined;
        if (columnWidth != null) {
            forcedTotalNetWidth += columnWidth;
            pendingTotalNetWidth += columnWidth;
            forcedColumnNetWidthsArray.push(columnWidth);
            newColumnNetWidthsArray.push(columnWidth);
            appliedColumnNetWidthsArray.push(columnWidth);
        }
        else {
            requiresActiveSizing = true;
            pendingTotalNetWidth += columnMinWidth;
            totalWorkingWeight += columnWeight;
            forcedColumnNetWidthsArray.push(undefined);
            newColumnNetWidthsArray.push(columnMinWidth);
            appliedColumnNetWidthsArray.push(columnMinWidth);
        }
    }
    // enforce min and max rules if active sizing is required and available space remains
    if (requiresActiveSizing && pendingTotalNetWidth < availableNetWidth) {
        _enforceFixedMinMaxRules(columnsArray, availableNetWidth, totalWorkingWeight, forcedTotalNetWidth, forcedColumnNetWidthsArray, columnWeightsArray, columnMaxWidthsArray, newColumnNetWidthsArray, appliedColumnNetWidthsArray);
    }
    return newColumnNetWidthsArray;
}
/**
 * Helper function to enforce all of the layout sizing rules on the Table's columns.
 */
function _enforceFixedMinMaxRules(columnsArray, availableWidth, totalWorkingWeight, forcedTotalWidth, forcedColumnWidthsArray, columnWeightsArray, columnMaxWidthsArray, newColumnWidthsArray, appliedColumnWidthsArray) {
    // try and honor the weights of each column combined with the min widths, and then check that against
    // any max widths once completed. if the final result doesn't fill the space AND at least 1 max width
    // rule was enforced, repeat the process until no max rules are enforced OR the space is filled.
    let isMaxRuleApplied = true;
    while (isMaxRuleApplied) {
        isMaxRuleApplied = false;
        _enforceWeightedMinRules(columnsArray, availableWidth, totalWorkingWeight, forcedTotalWidth, forcedColumnWidthsArray, columnWeightsArray, newColumnWidthsArray, appliedColumnWidthsArray);
        let pendingTotalWidth = 0;
        for (let i = 0; i < columnsArray.length; i++) {
            if (forcedColumnWidthsArray[i] == null) {
                const maxWidth = columnMaxWidthsArray[i];
                if (maxWidth != null && newColumnWidthsArray[i] > maxWidth) {
                    isMaxRuleApplied = true;
                    forcedColumnWidthsArray[i] = maxWidth;
                    forcedTotalWidth += maxWidth;
                    totalWorkingWeight -= columnWeightsArray[i];
                    newColumnWidthsArray[i] = maxWidth;
                    appliedColumnWidthsArray[i] = maxWidth;
                }
            }
            pendingTotalWidth += newColumnWidthsArray[i];
        }
        isMaxRuleApplied = isMaxRuleApplied && pendingTotalWidth < availableWidth;
    }
}
/**
 * Helper method to extract the individual column widths of the Table
 * for a 'contents' layout table that already has its root width applied.
 */
function _getContentsLayoutColumnNetWidths(innerTableElement, headerElements, sampledColumnNetWidthsRef, availableNetWidth, columnsArray, columnBorderEdges, staticColumnWidths) {
    const DEFAULT_COLUMN_MIN_WIDTH = undefined;
    const DEFAULT_COLUMN_MAX_WIDTH = undefined;
    let requiresActiveSizing = false;
    const columnMinWidthsArray = [];
    const columnMaxWidthsArray = [];
    const forcedColumnNetWidthsArray = [];
    const newColumnNetWidthsArray = [];
    for (let i = 0; i < columnsArray.length; i++) {
        const column = columnsArray[i];
        let columnMinWidth = column.value.minWidth;
        if (columnMinWidth == null || columnMinWidth <= 0) {
            columnMinWidth = DEFAULT_COLUMN_MIN_WIDTH;
        }
        columnMinWidthsArray.push(columnMinWidth);
        let columnMaxWidth = column.value.maxWidth;
        if (columnMaxWidth == null || columnMaxWidth < (columnMinWidth != null ? columnMinWidth : 0)) {
            columnMaxWidth = DEFAULT_COLUMN_MAX_WIDTH;
        }
        columnMaxWidthsArray.push(columnMaxWidth);
        const columnWidth = staticColumnWidths != null ? staticColumnWidths[column.key] : undefined;
        if (columnWidth != null) {
            forcedColumnNetWidthsArray.push(columnWidth);
            newColumnNetWidthsArray[i] = columnWidth;
        }
        else {
            requiresActiveSizing = true;
            forcedColumnNetWidthsArray.push(undefined);
        }
    }
    // short-circuit normal contents sizing if each column specifies a 'width' value
    if (requiresActiveSizing) {
        // if active sizing is required, update min width values to reflect actual widths
        // performance note - this should be the only logic that requires DOM measurements
        let forcedTotalNetWidth = 0;
        let pendingTotalNetWidth = 0;
        let totalPreferredNetWidth = 0;
        let totalWorkingWeight = 0;
        let totalRemainingWeight = 0;
        const sampledNetWidths = sampledColumnNetWidthsRef.current;
        const sampledMinNetWidthsArray = [];
        const sampledWeightNetWidthsArray = [];
        const appliedColumnNetWidthsArray = [];
        if (sampledNetWidths != null) {
            for (let i = 0; i < columnsArray.length; i++) {
                sampledMinNetWidthsArray.push(sampledNetWidths.minWidths[columnsArray[i].key]);
                sampledWeightNetWidthsArray.push(sampledNetWidths.weightWidths[columnsArray[i].key]);
            }
        }
        else {
            sampledColumnNetWidthsRef.current = {
                minWidths: {},
                weightWidths: {}
            };
            // when sampledColumnNetWidths are not populated, table has 'max-content' width applied,
            // which allows for measurements to represent the column weights
            for (let i = 0; i < headerElements.length; i++) {
                const sampledWeightNetWidth = headerElements[i].getBoundingClientRect().width -
                    columnBorderEdges[i].width;
                sampledWeightNetWidthsArray.push(sampledWeightNetWidth);
                sampledColumnNetWidthsRef.current.weightWidths[columnsArray[i].key] = sampledWeightNetWidth;
            }
            // specify 'min-content' width, allowing measurements to represent minimum auto-resizable column widths
            innerTableElement.style.width = 'min-content';
            for (let i = 0; i < headerElements.length; i++) {
                const sampledMinNetWidth = headerElements[i].getBoundingClientRect().width -
                    columnBorderEdges[i].width;
                sampledMinNetWidthsArray.push(sampledMinNetWidth);
                sampledColumnNetWidthsRef.current.minWidths[columnsArray[i].key] = sampledMinNetWidth;
            }
        }
        for (let i = 0; i < sampledMinNetWidthsArray.length; i++) {
            const forcedWidth = forcedColumnNetWidthsArray[i];
            if (forcedWidth == null) {
                let columnNetWidth;
                const sampledMinNetWidth = sampledMinNetWidthsArray[i];
                const minWidth = columnMinWidthsArray[i];
                const maxWidth = columnMaxWidthsArray[i];
                if (minWidth != null && minWidth >= sampledMinNetWidth) {
                    columnNetWidth = minWidth;
                }
                else if (maxWidth != null && sampledMinNetWidth >= maxWidth) {
                    columnNetWidth = maxWidth;
                }
                else {
                    columnNetWidth = sampledMinNetWidth;
                }
                pendingTotalNetWidth += columnNetWidth;
                // replace sampled weight with min width if needed
                const columnWeight = minWidth != null && minWidth >= sampledWeightNetWidthsArray[i]
                    ? minWidth
                    : sampledWeightNetWidthsArray[i];
                sampledWeightNetWidthsArray[i] = columnWeight;
                totalPreferredNetWidth +=
                    maxWidth != null && maxWidth <= columnWeight ? maxWidth : columnWeight;
                totalWorkingWeight += columnWeight;
                totalRemainingWeight += columnWeight - columnNetWidth;
                // newColumnWidthsArray already has content, so update in place rather than 'push'
                newColumnNetWidthsArray[i] = columnNetWidth;
                appliedColumnNetWidthsArray.push(columnNetWidth);
            }
            else {
                forcedTotalNetWidth += forcedWidth;
                pendingTotalNetWidth += forcedWidth;
                totalPreferredNetWidth += forcedWidth;
                appliedColumnNetWidthsArray.push(forcedWidth);
            }
        }
        // special case for initial rendering where we don't know if overall width is constrained
        if (availableNetWidth === -1) {
            availableNetWidth = totalPreferredNetWidth;
        }
        if (pendingTotalNetWidth < availableNetWidth) {
            _enforceContentsMinMaxRules(columnsArray, availableNetWidth, totalPreferredNetWidth, totalRemainingWeight, totalWorkingWeight, pendingTotalNetWidth, forcedTotalNetWidth, forcedColumnNetWidthsArray, sampledWeightNetWidthsArray, columnMaxWidthsArray, newColumnNetWidthsArray, appliedColumnNetWidthsArray);
        }
    }
    return newColumnNetWidthsArray;
}
/**
 * Helper function to enforce all of the layout sizing rules on the Table's columns.
 */
function _enforceContentsMinMaxRules(columnsArray, availableWidth, totalPreferredWidth, totalRemainingWeight, totalWorkingWeight, appliedPendingTotalWidth, forcedTotalWidth, forcedColumnWidthsArray, columnWeightsArray, columnMaxWidthsArray, newColumnWidthsArray, appliedColumnWidthsArray) {
    // try and honor the weights of each column combined with the min widths, and then check that against
    // any max widths once completed. if the final result doesn't fill the space AND at least 1 max width
    // rule was enforced, repeat the process until no max rules are enforced OR the space is filled.
    let isMaxRuleApplied = true;
    while (isMaxRuleApplied) {
        isMaxRuleApplied = false;
        if (totalPreferredWidth > availableWidth) {
            _enforceConstrainedWeightRules(columnsArray, availableWidth - appliedPendingTotalWidth, totalRemainingWeight, columnWeightsArray, forcedColumnWidthsArray, newColumnWidthsArray, appliedColumnWidthsArray);
        }
        else {
            _enforceWeightedMinRules(columnsArray, availableWidth, totalWorkingWeight, forcedTotalWidth, forcedColumnWidthsArray, columnWeightsArray, newColumnWidthsArray, appliedColumnWidthsArray);
        }
        let pendingTotalWidth = 0;
        for (let i = 0; i < columnsArray.length; i++) {
            if (forcedColumnWidthsArray[i] == null) {
                const maxWidth = columnMaxWidthsArray[i];
                if (maxWidth != null && newColumnWidthsArray[i] > maxWidth) {
                    isMaxRuleApplied = true;
                    forcedColumnWidthsArray[i] = maxWidth;
                    forcedTotalWidth += maxWidth;
                    appliedPendingTotalWidth += maxWidth - appliedColumnWidthsArray[i];
                    totalWorkingWeight -= columnWeightsArray[i];
                    newColumnWidthsArray[i] = maxWidth;
                    appliedColumnWidthsArray[i] = maxWidth;
                }
            }
            pendingTotalWidth += newColumnWidthsArray[i];
        }
        isMaxRuleApplied = isMaxRuleApplied && pendingTotalWidth < availableWidth;
    }
}
/**
 * Helper function to enforce constrained weights / min width rules on the Table's columns.
 */
function _enforceConstrainedWeightRules(columnsArray, initialExtraWidth, totalRemainingWeight, columnWeightsArray, forcedColumnWidthsArray, newColumnWidthsArray, appliedColumnWidthsArray) {
    if (initialExtraWidth > 0) {
        let availableWidth = initialExtraWidth;
        let currentWeightTotal = totalRemainingWeight;
        for (let i = 0; i < columnsArray.length; i++) {
            if (currentWeightTotal > 0 && forcedColumnWidthsArray[i] == null) {
                const currentWeight = columnWeightsArray[i] - appliedColumnWidthsArray[i];
                const weightWidth = (currentWeight / currentWeightTotal) * availableWidth;
                newColumnWidthsArray[i] = weightWidth + appliedColumnWidthsArray[i];
                availableWidth -= weightWidth;
                currentWeightTotal -= currentWeight;
            }
        }
    }
}
/**
 * Helper function to enforce the weights / min width rules on the Table's columns.
 */
function _enforceWeightedMinRules(columnsArray, availableWidth, totalWeight, forcedTotalWidth, forcedColumnWidthsArray, columnWeightsArray, newColumnWidthsArray, appliedColumnWidthsArray) {
    const pendingForcedColumnWidths = [];
    let isMinRuleApplied = true;
    while (isMinRuleApplied) {
        isMinRuleApplied = false;
        let currentWeightTotal = totalWeight;
        let widthDiff = availableWidth - forcedTotalWidth;
        if (widthDiff > 0) {
            for (let i = 0; i < columnsArray.length; i++) {
                if (currentWeightTotal > 0 &&
                    forcedColumnWidthsArray[i] == null &&
                    pendingForcedColumnWidths[i] == null) {
                    const currentWeight = columnWeightsArray[i];
                    const newWidth = (currentWeight / currentWeightTotal) * widthDiff;
                    // if calculated width breaks min width requirement, force min width as needed
                    const minWidth = appliedColumnWidthsArray[i];
                    if (newWidth < minWidth) {
                        // the preferred weighted size conflicts with the min size, so set pending to min size
                        pendingForcedColumnWidths[i] = minWidth;
                        forcedTotalWidth += minWidth;
                        newColumnWidthsArray[i] = minWidth;
                        isMinRuleApplied = true;
                        totalWeight -= currentWeight;
                        break;
                    }
                    newColumnWidthsArray[i] = newWidth;
                    widthDiff -= newWidth;
                    currentWeightTotal -= currentWeight;
                }
            }
        }
    }
}
/**
 * Helper function to create a ColumnWidths object from a TableColumn array and a widths array.
 */
function _createColumnWidths(columnsArray, netWidthsArray, realWidthsArray, borderEdges, borderOffset, totalBorderOffset) {
    const logicalWidthsArray = [];
    const columnWidths = {};
    for (let i = 0; i < columnsArray.length; i++) {
        columnWidths[columnsArray[i].key] = netWidthsArray[i];
        logicalWidthsArray.push(netWidthsArray[i] + borderEdges[i].offset);
    }
    return {
        netColumnWidths: columnWidths,
        fullColumnWidths: logicalWidthsArray,
        realColumnWidths: realWidthsArray,
        borderEdges,
        borderOffset,
        totalBorderOffset
    };
}
/**
 * Helper method to apply the given sizing object to the currently rendered Table
 * and column elements.
 */
function _updateTableSizing(innerTableElement, tableCols, realWidthsArray, totalRealWidth) {
    for (let i = 0; i < tableCols.length; i++) {
        const tableCol = tableCols[i];
        tableCol.style.width = `${realWidthsArray[i]}px`;
    }
    innerTableElement.style.width = `${totalRealWidth}px`;
    innerTableElement.style.tableLayout = 'fixed';
}
/**
 * Helper function to compare two ColumnWidths instances for equality.
 */
const columnWidthsComparator = (columnWidths1, columnWidths2) => {
    const keys1 = Object.keys(columnWidths1);
    const keys2 = Object.keys(columnWidths2);
    if (keys1.length === keys2.length) {
        for (const key of keys1) {
            const columnKey = key;
            if (columnWidths1[columnKey] !== columnWidths2[columnKey]) {
                return false;
            }
        }
        return true;
    }
    return false;
};

/**
 * Hook that manages column sizing on the Preact Table
 */
function useSizingManager({ rootRef, innerTableRef, tableId, tableData, viewportConfig, externalScrollerAccessRef, layout, columns, columnOrder, columnWidths, isRtl, isHighContrast }) {
    // scrollerRef is callback-based to meet useResizeObserver hook's state management requirements
    const [scrollerRef, setScrollerRef] = hooks.useState(preact.createRef());
    const scrollerRefCallback = hooks.useCallback((node) => {
        setScrollerRef({ current: node });
    }, []);
    hooks.useImperativeHandle(externalScrollerAccessRef, () => {
        if (viewportConfig != null) {
            return { [unsafeDomAccess.UNSAFE_DOM_ACCESS]: viewportConfig.scrollerRef.current };
        }
        return { [unsafeDomAccess.UNSAFE_DOM_ACCESS]: scrollerRef.current };
    }, [scrollerRef, viewportConfig]);
    // setup busy start and busy end callbacks for the 2 resize observer regions
    const { onBusyStart: onResizeScrollerBusyStart, onBusyEnd: onResizeScrollerBusyEnd } = useAddBusyState.useAddBusyState('handling scroller resize');
    const { onBusyStart: onResizeTableBusyStart, onBusyEnd: onResizeTableBusyEnd } = useAddBusyState.useAddBusyState('handling table resize');
    // state to force a rerender following the initial render where scrollbars are forced
    const [_isScrollbarSizeKnown, setIsScrollbarSizeKnown] = hooks.useState(false);
    // column width state information
    const sampledColumnNetWidthsRef = hooks.useRef();
    const [appliedColumnWidths, setAppliedColumnWidths] = hooks.useState();
    const [innerTableHeight, setInnerTableHeight] = hooks.useState();
    const [hasVerticalUnderflow, setHasVerticalUnderflow] = hooks.useState(false);
    // general table sizing state information
    const sizingInfoRef = hooks.useRef({
        isInitialRender: true,
        isScrollbarSizeKnown: false
    });
    const pendingResizeEntriesRef = hooks.useRef({});
    // scrollbar overflow state
    const overflowInfoRef = hooks.useRef({
        isHorizontal: false,
        isVertical: false
    });
    // sticky edge state management
    const [stickyEdgesState, setStickyEdgesState] = hooks.useState({});
    // current viewport information
    const viewportInfo = hooks.useMemo(() => {
        const tableViewportConfig = getViewportConfig(scrollerRef, viewportConfig);
        return {
            isExternalScroller: viewportConfig != null,
            viewportConfig: tableViewportConfig
        };
    }, [scrollerRef, viewportConfig]);
    // memoize realized columns array on 'columns' and 'columnOrder'
    const columnsArray = hooks.useMemo(() => {
        const newColumnsArray = [];
        if (columnOrder != null) {
            for (const key of columnOrder) {
                if (columns[key] != null) {
                    newColumnsArray.push({
                        key: key,
                        value: columns[key]
                    });
                }
            }
        }
        else {
            for (const [key, value] of Object.entries(columns)) {
                newColumnsArray.push({
                    key: key,
                    value: value
                });
            }
        }
        return newColumnsArray;
    }, [columnOrder, columns]);
    const stickyColumnInfo = hooks.useMemo(() => {
        const stickyColumns = getStickyColumnIndicies(columnsArray);
        let firstStickyColumnIndex;
        let finalStickyColumnIndex;
        const stickyColumnsCount = stickyColumns.length;
        if (stickyColumns.length > 0) {
            firstStickyColumnIndex = stickyColumns[0];
            finalStickyColumnIndex = stickyColumns[stickyColumnsCount - 1];
        }
        return { stickyColumns, firstStickyColumnIndex, finalStickyColumnIndex };
    }, [columnsArray]);
    const borderOffset = isHighContrast ? 3 : 1;
    const columnBorderEdges = hooks.useMemo(() => {
        const newColumnBorderEdges = [];
        let isPreviousColumnHasEnd = false;
        for (let i = 0; i < columnsArray.length; i++) {
            const hasStart = !isPreviousColumnHasEnd;
            const isLastColumn = i === columnsArray.length - 1;
            const isColumnSticky = stickyColumnInfo.stickyColumns.includes(i);
            const isLastStickyColumn = isColumnSticky && stickyColumnInfo.finalStickyColumnIndex === i;
            const hasEnd = isLastColumn || isLastStickyColumn || stickyEdgesState?.start === i;
            const width = (hasStart ? borderOffset : 0) + (hasEnd ? borderOffset : 0);
            let offset;
            if (isLastStickyColumn ||
                (isLastColumn && stickyColumnInfo.finalStickyColumnIndex !== i - 1)) {
                offset = 2 * borderOffset;
            }
            else {
                offset =
                    !isLastColumn && stickyColumnInfo.finalStickyColumnIndex === i - 1 ? 0 : borderOffset;
            }
            newColumnBorderEdges.push({ hasStart, hasEnd, width, offset });
            isPreviousColumnHasEnd = hasEnd;
        }
        return newColumnBorderEdges;
    }, [borderOffset, columnsArray, stickyEdgesState, stickyColumnInfo]);
    // determine if a new layout is required
    let isLayoutRequired = false;
    if (appliedColumnWidths != null) {
        for (let i = 0; i < columnsArray.length; i++) {
            const key = columnsArray[i].key;
            const netWidth = appliedColumnWidths.netColumnWidths[key];
            const staticWidth = columnWidths != null ? columnWidths[key] : undefined;
            if (netWidth == null || (staticWidth != null && staticWidth !== netWidth)) {
                isLayoutRequired = true;
                break;
            }
        }
    }
    else {
        isLayoutRequired = true;
    }
    // ensure sampled column sizes are cleared if 'columns', 'columnOrder', or 'layout' is updated
    hooks.useLayoutEffect(() => {
        sampledColumnNetWidthsRef.current = undefined;
        setAppliedColumnWidths(undefined);
    }, [columns, columnOrder, layout]);
    // ensure applied column widths are cleared if 'columnWidths' is updated
    hooks.useLayoutEffect(() => {
        setAppliedColumnWidths(undefined);
    }, [columnWidths, stickyEdgesState]);
    // setup sizing calculation updates for subsequent renders
    hooks.useLayoutEffect(() => {
        const scrollerElement = viewportInfo.viewportConfig.scrollerRef.current;
        const innerTableElement = innerTableRef.current;
        if (layout !== 'pending' &&
            scrollerElement != null &&
            innerTableElement != null &&
            !sizingInfoRef.current.isInitialRender &&
            (viewportInfo.isExternalScroller || sizingInfoRef.current.isScrollbarSizeKnown) &&
            isLayoutRequired &&
            columnsArray.length > 0) {
            let widthOffset = 0;
            if (viewportInfo.isExternalScroller) {
                const scrollerOffsetStart = viewportInfo.viewportConfig.scrollerOffsetStart;
                if (scrollerOffsetStart != null) {
                    widthOffset += scrollerOffsetStart;
                }
                const scrollerOffsetEnd = viewportInfo.viewportConfig.scrollerOffsetEnd;
                if (scrollerOffsetEnd != null) {
                    widthOffset += scrollerOffsetEnd;
                }
            }
            else if (sizingInfoRef.current.isScrollbarSizeKnown) {
                widthOffset = sizingInfoRef.current.defaultScrollbarWidth;
            }
            const totalWidth = layout === 'fixed' ? sizingInfoRef.current.contentWidth - widthOffset : -1;
            const newColumnWidths = applyLayoutColumnWidths(scrollerElement, viewportInfo, innerTableElement, widthOffset, tableId, layout === 'fixed', sampledColumnNetWidthsRef, totalWidth, columnsArray, columnBorderEdges, borderOffset, columnWidths);
            setAppliedColumnWidths(newColumnWidths);
        }
    }, [
        viewportInfo,
        innerTableRef,
        tableId,
        columnsArray,
        columnBorderEdges,
        borderOffset,
        columnWidths,
        layout,
        isLayoutRequired,
        isHighContrast
    ]);
    hooks.useLayoutEffect(() => {
        const rootElement = rootRef.current;
        const scrollerElement = viewportInfo.viewportConfig.scrollerRef.current;
        if (layout !== 'pending' &&
            rootElement != null &&
            scrollerElement != null &&
            appliedColumnWidths != null) {
            updateStickyColumns(rootElement, scrollerElement, viewportInfo, tableId, isRtl, columnsArray, stickyColumnInfo, sizingInfoRef.current, overflowInfoRef.current, appliedColumnWidths.fullColumnWidths, stickyEdgesState, setStickyEdgesState, viewportInfo.isExternalScroller || overflowInfoRef.current.isHorizontal);
        }
    }, [
        tableData,
        rootRef,
        viewportInfo,
        tableId,
        appliedColumnWidths,
        columnsArray,
        stickyColumnInfo,
        layout,
        stickyEdgesState,
        isRtl
    ]);
    // helper method to update the sticky edge styling due to resize events
    const _updateStickyEdgesFromResize = hooks.useCallback((rootElement, scrollerElement) => {
        if (appliedColumnWidths != null) {
            const sizingInfo = sizingInfoRef.current;
            const overflowInfo = overflowInfoRef.current;
            if (viewportInfo.isExternalScroller || overflowInfo.isHorizontal) {
                // ensure sticky edges are applied correctly if horizontal overflow is present
                updateStickyEdges(rootElement, scrollerElement, viewportInfo, columnsArray, stickyColumnInfo, sizingInfo, overflowInfo, appliedColumnWidths.fullColumnWidths, isRtl, stickyEdgesState, setStickyEdgesState);
            }
            else if (stickyEdgesState.start != null || stickyEdgesState.end != null) {
                // otherwise clear sticky edges if necessary
                setStickyEdgesState({});
            }
        }
    }, [appliedColumnWidths, columnsArray, stickyColumnInfo, stickyEdgesState, viewportInfo, isRtl]);
    // callback function to handle resizes of the scroller and inner table elements
    const handleResizeCallback = hooks.useCallback(() => {
        let currentEntry = pendingResizeEntriesRef.current.scroller;
        const rootElement = rootRef.current;
        const scrollerElement = viewportInfo.viewportConfig.scrollerRef.current;
        const innerTableElement = innerTableRef.current;
        let isOveralSizeHandled = false;
        // handle scroller element resize notification
        if (currentEntry != null &&
            rootElement != null &&
            scrollerElement != null &&
            innerTableElement != null) {
            const newContentWidth = currentEntry.contentBoxSize[0].inlineSize;
            const newContentHeight = currentEntry.contentBoxSize[0].blockSize;
            let newBoxWidth;
            let newBoxHeight;
            // for HTML scroller, innerWidth / innerHeight to ensure scrollbar size is included
            if (ScrollUtils.isHTMLScroller(scrollerElement)) {
                newBoxWidth = window.innerWidth;
                newBoxHeight = window.innerHeight;
            }
            else {
                newBoxWidth = currentEntry.borderBoxSize[0].inlineSize;
                newBoxHeight = currentEntry.borderBoxSize[0].blockSize;
            }
            if (newBoxWidth !== 0 || newBoxHeight !== 0) {
                isOveralSizeHandled = true;
                let availableWidth;
                let availableWidthOffset = 0;
                let defaultScrollbarWidth = 0;
                if (viewportInfo.isExternalScroller) {
                    const scrollerOffsetStart = viewportInfo.viewportConfig.scrollerOffsetStart;
                    if (scrollerOffsetStart != null) {
                        availableWidthOffset += scrollerOffsetStart;
                    }
                    const scrollerOffsetEnd = viewportInfo.viewportConfig.scrollerOffsetEnd;
                    if (scrollerOffsetEnd != null) {
                        availableWidthOffset += scrollerOffsetEnd;
                    }
                    if (sizingInfoRef.current.isScrollbarSizeKnown) {
                        sizingInfoRef.current = {
                            isInitialRender: false,
                            contentWidth: newContentWidth,
                            boxWidth: newBoxWidth,
                            contentHeight: newContentHeight,
                            boxHeight: newBoxHeight,
                            isScrollbarSizeKnown: true,
                            defaultScrollbarWidth: sizingInfoRef.current.defaultScrollbarWidth,
                            defaultScrollbarHeight: sizingInfoRef.current.defaultScrollbarHeight
                        };
                    }
                    else {
                        sizingInfoRef.current = {
                            isInitialRender: false,
                            contentWidth: newContentWidth,
                            boxWidth: newBoxWidth,
                            contentHeight: newContentHeight,
                            boxHeight: newBoxHeight,
                            isScrollbarSizeKnown: false
                        };
                    }
                    availableWidth = newContentWidth - availableWidthOffset;
                }
                else {
                    const isInitialRender = sizingInfoRef.current.isInitialRender;
                    if (sizingInfoRef.current.isScrollbarSizeKnown) {
                        availableWidthOffset = sizingInfoRef.current.defaultScrollbarWidth;
                        sizingInfoRef.current = {
                            isInitialRender: false,
                            contentWidth: newContentWidth,
                            boxWidth: newBoxWidth,
                            contentHeight: newContentHeight,
                            boxHeight: newBoxHeight,
                            isScrollbarSizeKnown: true,
                            defaultScrollbarWidth: sizingInfoRef.current.defaultScrollbarWidth,
                            defaultScrollbarHeight: sizingInfoRef.current.defaultScrollbarHeight
                        };
                    }
                    else {
                        // on the first render that does not specify an external scroller, horizontal and
                        // vertical scrollbars are forced in order to determine default browser sizes
                        availableWidthOffset = newBoxWidth - newContentWidth;
                        sizingInfoRef.current = {
                            isInitialRender: false,
                            contentWidth: newContentWidth,
                            boxWidth: newBoxWidth,
                            contentHeight: newContentHeight,
                            boxHeight: newBoxHeight,
                            isScrollbarSizeKnown: true,
                            defaultScrollbarWidth: availableWidthOffset,
                            defaultScrollbarHeight: newBoxHeight - newContentHeight
                        };
                        setIsScrollbarSizeKnown(true);
                    }
                    defaultScrollbarWidth = sizingInfoRef.current.defaultScrollbarWidth;
                    availableWidth = !isInitialRender || layout === 'fixed' ? newContentWidth : -1;
                }
                if (columnsArray.length > 0) {
                    let newColumnWidths;
                    if (layout !== 'pending') {
                        newColumnWidths = applyLayoutColumnWidths(scrollerElement, viewportInfo, innerTableElement, availableWidthOffset, tableId, layout === 'fixed', sampledColumnNetWidthsRef, availableWidth, columnsArray, columnBorderEdges, borderOffset, columnWidths);
                        if (appliedColumnWidths == null ||
                            !columnWidthsComparator(appliedColumnWidths.netColumnWidths, newColumnWidths.netColumnWidths)) {
                            setAppliedColumnWidths(newColumnWidths);
                        }
                        else {
                            newColumnWidths = undefined;
                        }
                    }
                    // ensure scrollbars are applied correctly - a scrollbar could be needed
                    if (viewportInfo.isExternalScroller) {
                        applyUnderflowStatus(rootElement, tableId, setHasVerticalUnderflow);
                    }
                    else {
                        const innerTableRect = innerTableElement.getBoundingClientRect();
                        applyOverflowStatus(rootElement, scrollerElement, tableId, innerTableRect.width, newContentWidth, innerTableRect.height, newContentHeight, defaultScrollbarWidth, overflowInfoRef, setHasVerticalUnderflow);
                    }
                    if (layout !== 'pending') {
                        if (newColumnWidths != null) {
                            updateStickyColumns(rootElement, scrollerElement, viewportInfo, tableId, isRtl, columnsArray, stickyColumnInfo, sizingInfoRef.current, overflowInfoRef.current, newColumnWidths.fullColumnWidths, stickyEdgesState, setStickyEdgesState, viewportInfo.isExternalScroller || overflowInfoRef.current.isHorizontal);
                        }
                        else {
                            _updateStickyEdgesFromResize(rootElement, scrollerElement);
                        }
                    }
                }
            }
        }
        currentEntry = pendingResizeEntriesRef.current.innerTable;
        if (currentEntry != null) {
            const newBoxWidth = currentEntry.borderBoxSize[0].inlineSize;
            const newBoxHeight = currentEntry.borderBoxSize[0].blockSize;
            // if only an inner table resize occurred, ensure scrollbars are applied correctly
            if (!isOveralSizeHandled &&
                rootElement != null &&
                scrollerElement != null &&
                columnsArray.length > 0 &&
                (newBoxWidth !== 0 || newBoxHeight !== 0)) {
                if (viewportInfo.isExternalScroller) {
                    applyUnderflowStatus(rootElement, tableId, setHasVerticalUnderflow);
                }
                else if (sizingInfoRef.current.isScrollbarSizeKnown) {
                    applyOverflowStatus(rootElement, scrollerElement, tableId, newBoxWidth, sizingInfoRef.current.contentWidth, newBoxHeight, sizingInfoRef.current.contentHeight, sizingInfoRef.current.defaultScrollbarWidth, overflowInfoRef, setHasVerticalUnderflow);
                }
                if (layout !== 'pending') {
                    _updateStickyEdgesFromResize(rootElement, scrollerElement);
                }
            }
            setInnerTableHeight(newBoxHeight);
        }
        pendingResizeEntriesRef.current = {};
    }, [
        rootRef,
        viewportInfo,
        innerTableRef,
        tableId,
        layout,
        columnsArray,
        stickyColumnInfo,
        columnBorderEdges,
        borderOffset,
        columnWidths,
        appliedColumnWidths,
        stickyEdgesState,
        isRtl,
        _updateStickyEdgesFromResize
    ]);
    // ensure resize observers have access to the latest state information
    const handleResizeCallbackRef = hooks.useRef(handleResizeCallback);
    handleResizeCallbackRef.current = handleResizeCallback;
    // resize observer callback functions are static to avoid hook re-subscribing each render
    const staticHandleScrollerResize = hooks.useCallback((entry) => {
        // only queue up an animation frame if there isn't one already pending
        if (pendingResizeEntriesRef.current.scroller == null ||
            pendingResizeEntriesRef.current.innerTable == null) {
            onResizeScrollerBusyStart();
            window.requestAnimationFrame(() => {
                handleResizeCallbackRef.current();
                onResizeScrollerBusyEnd();
            });
        }
        pendingResizeEntriesRef.current.scroller = entry;
    }, [onResizeScrollerBusyStart, onResizeScrollerBusyEnd]);
    const staticHandleInnerTableResize = hooks.useCallback((entry) => {
        // only queue up an animation frame if there isn't one already pending
        if (pendingResizeEntriesRef.current.scroller == null ||
            pendingResizeEntriesRef.current.innerTable == null) {
            onResizeTableBusyStart();
            window.requestAnimationFrame(() => {
                handleResizeCallbackRef.current();
                onResizeTableBusyEnd();
            });
        }
        pendingResizeEntriesRef.current.innerTable = entry;
    }, [onResizeTableBusyStart, onResizeTableBusyEnd]);
    useResizeObserver.useResizeObserver(viewportInfo.viewportConfig.scrollerRef, staticHandleScrollerResize);
    useResizeObserver.useResizeObserver(innerTableRef, staticHandleInnerTableResize);
    return {
        columnBorderEdges,
        isLayoutRequired,
        viewportInfo,
        columnsArray,
        appliedColumnWidths,
        sizingInfoRef,
        innerTableHeight,
        hasVerticalUnderflow,
        stickyEdgesState,
        setStickyEdgesState,
        stickyColumnInfo,
        overflowInfoRef,
        scrollerRefCallback
    };
}

/**
 * Helper function to generate the class array for a data cell.
 */
const getDataCellClassArray = ({ isFirstColumnIndex, isFinalColumnIndex, isFirstRowIndex, isFinalRowIndex, hasVerticalGridlines, hasHorizontalGridlines, hasLastHorizontalGridline, isRowSelected, isPreviousRowSelected, isColumnSelected, isNextColumnSelected, isPreviousColumnSelected, isShowFocusRing, isSticky, isRowSelectionEnabled, isRowSingleSelection, isActive, isHover, isPseudoHover, isStickyStartEdge, isStickyEndEdge, hasStartEdge, hasEndEdge, isRtl, isHighContrast, horizontalAlignment, verticalAlignment, variant, padding }) => {
    const cellStyles = isHighContrast ? PRIVATE_Table_themes_TableCellStyles_css.tableCellStylesHC : PRIVATE_Table_themes_TableCellStyles_css.tableCellStyles;
    const classArray = [cellStyles.base];
    // Padding styling logic
    const { isTopPadding, isBottomPadding, isStartPadding, isEndPadding } = _getPaddingInfo(padding);
    // Horizontal and vertical alignment styling logic
    _applyAlignmentStyling(classArray, horizontalAlignment, verticalAlignment);
    // Sticky styling logic
    _applyStickyStyling(classArray, cellStyles, isSticky, isStickyStartEdge, isStickyEndEdge, isRtl);
    // Interaction styling logic
    if (isRowSelectionEnabled) {
        if (isActive) {
            classArray.push(cellStyles.active);
        }
        else if (isHover) {
            classArray.push(cellStyles.hover);
        }
        else if (isPseudoHover) {
            classArray.push(cellStyles.pseudoHover);
        }
    }
    if (variant === 'highlight') {
        classArray.push(cellStyles.rowHighlight);
    }
    _applyFocusStyling(classArray, cellStyles, isShowFocusRing);
    // Selected Rows and Horizontal Gridline styling logic
    let hasBackground = false;
    if (isRowSelected) {
        hasBackground = true;
        classArray.push(cellStyles.selectedBackground);
    }
    if (isRowSingleSelection && (isRowSelected || isPreviousRowSelected)) {
        classArray.push(cellStyles.horizontalSelectedGridTop);
        classArray.push(isTopPadding ? cellStyles.selectedBorderTopPadding : cellStyles.selectedBorderTopNoPadding);
    }
    else if (!isFirstRowIndex && hasHorizontalGridlines) {
        classArray.push(cellStyles.horizontalGridTop);
        classArray.push(isTopPadding ? cellStyles.borderTopPadding : cellStyles.borderTopNoPadding);
    }
    else {
        classArray.push(isTopPadding ? cellStyles.topSpacerPadding : cellStyles.topSpacerNoPadding);
    }
    if (isFinalRowIndex) {
        classArray.push(cellStyles.lastRow);
        if (isRowSingleSelection && isRowSelected) {
            classArray.push(cellStyles.horizontalSelectedGridBottom);
            classArray.push(isBottomPadding
                ? cellStyles.selectedBorderBottomPadding
                : cellStyles.selectedBorderBottomNoPadding);
        }
        else if (hasLastHorizontalGridline) {
            classArray.push(cellStyles.horizontalGridBottom);
            classArray.push(isBottomPadding ? cellStyles.borderBottomPadding : cellStyles.borderBottomNoPadding);
        }
        else {
            classArray.push(isBottomPadding ? cellStyles.bottomSpacerPadding : cellStyles.bottomSpacerNoPadding);
        }
    }
    else {
        classArray.push(cellStyles.notLastRow);
        classArray.push(isBottomPadding ? cellStyles.bottomPadding : cellStyles.bottomNoPadding);
    }
    // Selected Columns and Vertical Gridline styling logic
    if (isColumnSelected) {
        if (!hasBackground) {
            classArray.push(cellStyles.selectedBackground);
            hasBackground = true;
        }
        _applySelectedColumnStyling(classArray, cellStyles, isStartPadding, isEndPadding, hasStartEdge, hasEndEdge);
    }
    else {
        if (!hasBackground) {
            classArray.push(cellStyles.background);
        }
        _applyColumnStyling(classArray, cellStyles, isStartPadding, isEndPadding, isFirstColumnIndex, isFinalColumnIndex, hasStartEdge, hasEndEdge, hasVerticalGridlines, isNextColumnSelected, isPreviousColumnSelected);
    }
    return classArray;
};
/**
 * Helper function to generate the class array for a header or footer cell.
 */
const getHeaderFooterCellClassArray = ({ isHeader, isActiveStyle, isHoverStyle, isPreviousCellActiveStyle, isPreviousCellHoverStyle, isFirstColumnIndex, isFinalColumnIndex, hasVerticalGridlines, isSelected, isNextColumnSelected, isPreviousColumnSelected, isSticky, isShowFocusRing, isStickyStartEdge, isStickyEndEdge, hasStartEdge, hasEndEdge, isRtl, horizontalAlignment, verticalAlignment, isHighContrast, padding }) => {
    let cellStyles;
    if (isHeader) {
        cellStyles = isHighContrast ? PRIVATE_Table_themes_TableHeaderCellStyles_css.tableHeaderCellStylesHC : PRIVATE_Table_themes_TableHeaderCellStyles_css.tableHeaderCellStyles;
    }
    else {
        cellStyles = isHighContrast ? PRIVATE_Table_themes_TableFooterCellStyles_css.tableFooterCellStylesHC : PRIVATE_Table_themes_TableFooterCellStyles_css.tableFooterCellStyles;
    }
    const classArray = [cellStyles.base];
    // Padding styling logic
    const { isTopPadding, isBottomPadding, isStartPadding, isEndPadding } = _getPaddingInfo(padding);
    classArray.push(isBottomPadding ? cellStyles.bottomPadding : cellStyles.bottomNoPadding);
    classArray.push(isTopPadding ? cellStyles.topPadding : cellStyles.topNoPadding);
    // Horizontal and vertical alignment styling logic
    _applyAlignmentStyling(classArray, horizontalAlignment, verticalAlignment);
    // Sticky styling logic
    _applyStickyStyling(classArray, cellStyles, isSticky, isStickyStartEdge, isStickyEndEdge, isRtl);
    // Interaction styling logic
    _applyFocusStyling(classArray, cellStyles, isShowFocusRing);
    // Selected and Vertical Gridline styling logic
    if (isSelected) {
        if (isActiveStyle) {
            classArray.push(cellStyles.activeSelected);
        }
        else if (isHoverStyle) {
            classArray.push(cellStyles.hoverSelected);
        }
        classArray.push(cellStyles.selectedBackground);
        _applySelectedColumnStyling(classArray, cellStyles, isStartPadding, isEndPadding, hasStartEdge, hasEndEdge);
    }
    else {
        if (isActiveStyle) {
            classArray.push(cellStyles.active);
        }
        else if (isHoverStyle) {
            classArray.push(cellStyles.hover);
        }
        classArray.push(cellStyles.background);
        _applyColumnStyling(classArray, cellStyles, isStartPadding, isEndPadding, isFirstColumnIndex, isFinalColumnIndex, hasStartEdge, hasEndEdge, hasVerticalGridlines, isNextColumnSelected, isPreviousColumnSelected, isActiveStyle, isHoverStyle, isPreviousCellActiveStyle, isPreviousCellHoverStyle);
    }
    return classArray;
};
const _getPaddingInfo = (padding) => {
    const isPaddingDisabled = padding === 'disabled';
    const isComplexPadding = padding instanceof Object;
    const isBottomPadding = !(isPaddingDisabled ||
        (isComplexPadding && padding?.bottom === 'disabled'));
    const isEndPadding = !(isPaddingDisabled || (isComplexPadding && padding?.end === 'disabled'));
    const isStartPadding = !(isPaddingDisabled ||
        (isComplexPadding && padding?.start === 'disabled'));
    const isTopPadding = !(isPaddingDisabled || (isComplexPadding && padding?.top === 'disabled'));
    return { isTopPadding, isBottomPadding, isStartPadding, isEndPadding };
};
const _applyAlignmentStyling = (classArray, horizontalAlignment, verticalAlignment) => {
    classArray.push(PRIVATE_TableCommon_themes_Common_css.textAlignStyles[horizontalAlignment]);
    classArray.push(PRIVATE_TableCommon_themes_Common_css.verticalAlignStyles[verticalAlignment]);
};
const _applyFocusStyling = (classArray, cellStyles, isShowFocusRing) => {
    if (isShowFocusRing) {
        classArray.push(cellStyles.focused);
    }
};
const _applyStickyStyling = (classArray, cellStyles, isSticky, isStickyStartEdge, isStickyEndEdge, isRtl) => {
    if (isSticky) {
        classArray.push(cellStyles.stickyColumn);
    }
    if (isStickyStartEdge) {
        classArray.push(cellStyles.stickyEdge);
        classArray.push(isRtl ? cellStyles.stickyEdgeLeft : cellStyles.stickyEdgeRight);
    }
    else if (isStickyEndEdge) {
        classArray.push(cellStyles.stickyEdge);
        classArray.push(isRtl ? cellStyles.stickyEdgeRight : cellStyles.stickyEdgeLeft);
    }
};
const _applyColumnStyling = (classArray, cellStyles, isStartPadding, isEndPadding, isFirstColumnIndex, isFinalColumnIndex, hasStartEdge, hasEndEdge, hasVerticalGridlines, isNextColumnSelected, isPreviousColumnSelected, isActiveStyle, isHoverStyle, isPreviousCellActiveStyle, isPreviousCellHoverStyle) => {
    if (hasStartEdge) {
        if (isPreviousColumnSelected) {
            classArray.push(cellStyles.verticalSelectedGridStart);
            classArray.push(isStartPadding
                ? cellStyles.selectedBorderStartPadding
                : cellStyles.selectedBorderStartNoPadding);
        }
        else if (isActiveStyle ||
            isHoverStyle ||
            isPreviousCellActiveStyle ||
            isPreviousCellHoverStyle ||
            (!isFirstColumnIndex && hasVerticalGridlines)) {
            classArray.push(cellStyles.verticalGridStart);
            classArray.push(isStartPadding ? cellStyles.borderStartPadding : cellStyles.borderStartNoPadding);
        }
        else if (isStartPadding) {
            classArray.push(cellStyles.startSpacerPadding);
        }
        else {
            classArray.push(cellStyles.startSpacerNoPadding);
        }
    }
    else if (isStartPadding) {
        classArray.push(cellStyles.startPadding);
    }
    else {
        classArray.push(cellStyles.startNoPadding);
    }
    if (hasEndEdge) {
        if (isNextColumnSelected) {
            classArray.push(cellStyles.verticalSelectedGridEnd);
            classArray.push(isEndPadding ? cellStyles.selectedBorderEndPadding : cellStyles.selectedBorderEndNoPadding);
        }
        else if (isActiveStyle || isHoverStyle || (!isFinalColumnIndex && hasVerticalGridlines)) {
            classArray.push(cellStyles.verticalGridEnd);
            classArray.push(isEndPadding ? cellStyles.borderEndPadding : cellStyles.borderEndNoPadding);
        }
        else if (isEndPadding) {
            classArray.push(cellStyles.endSpacerPadding);
        }
        else {
            classArray.push(cellStyles.endSpacerNoPadding);
        }
    }
    else if (isEndPadding) {
        classArray.push(cellStyles.endPadding);
    }
    else {
        classArray.push(cellStyles.endNoPadding);
    }
};
const _applySelectedColumnStyling = (classArray, cellStyles, isStartPadding, isEndPadding, hasStartEdge, hasEndEdge) => {
    if (hasStartEdge) {
        classArray.push(cellStyles.verticalSelectedGridStart);
        classArray.push(isStartPadding
            ? cellStyles.selectedBorderStartPadding
            : cellStyles.selectedBorderStartNoPadding);
    }
    else {
        classArray.push(isStartPadding ? cellStyles.startPadding : cellStyles.startNoPadding);
    }
    if (hasEndEdge) {
        classArray.push(cellStyles.verticalSelectedGridEnd);
        classArray.push(isEndPadding ? cellStyles.selectedBorderEndPadding : cellStyles.selectedBorderEndNoPadding);
    }
    else {
        classArray.push(isEndPadding ? cellStyles.endPadding : cellStyles.endNoPadding);
    }
};
/**
 * Helper function to convert a column's horizontal alignment value into a form control's value.
 */
const getFormTextAlign = (horizontalAlignment, isRtl) => {
    switch (horizontalAlignment) {
        case 'left':
            return isRtl ? 'end' : 'start';
        case 'center':
            return undefined;
        default:
            return horizontalAlignment;
    }
};

const dimensions = ['height', 'width'];
// Create an array [dimensionInterpolations['height'], dimensionInterpolations['width']]
const skeletonDimensionInterpolation = Array.from(dimensions, (x) => dimensions$1.dimensionInterpolations[x]);
const interpolations = [...Object.values(skeletonDimensionInterpolation)];
const SkeletonInterpolations = mergeInterpolations.mergeInterpolations(interpolations);
/**
 * TableSkeleton component allows the appropriate skeleton to be rendered based on the
 * property values
 **/
function TableSkeleton({ ...props }) {
    const skeletonStyles = props.isHighContrast ? PRIVATE_Table_themes_TableSkeletonStyles_css.tableSkeletonStylesHC : PRIVATE_Table_themes_TableSkeletonStyles_css.tableSkeletonStyles;
    const classes = classNames.classNames([skeletonStyles.bar]);
    const skeletonDimensions = SkeletonInterpolations({ width: '100%', ...props });
    return jsxRuntime.jsx("div", { style: skeletonDimensions, class: classes });
}

/**
 * TableSkeletonRow renders a set of 'loading' skeletons.
 **/
function TableSkeletonRow({ tableId, colspan, hasTabIndex, isShowFocusRing, isEmbedded, isPendingLayout = false, isLoadMore = false, isHighContrast }) {
    const translations = useTranslationBundle.useTranslationBundle('@oracle/oraclejet-preact');
    const skeletonStyles = isHighContrast ? PRIVATE_Table_themes_TableSkeletonStyles_css.tableSkeletonStylesHC : PRIVATE_Table_themes_TableSkeletonStyles_css.tableSkeletonStyles;
    const cellClassArray = [skeletonStyles.cell];
    if (isShowFocusRing) {
        cellClassArray.push(skeletonStyles.focus);
    }
    const cellClasses = classNames.classNames(cellClassArray);
    const innerCellClasses = classNames.classNames([skeletonStyles.innerCell]);
    const rowClasses = classNames.classNames([skeletonStyles.row, skeletonStyles.rowHeight]);
    const firstRowClasses = isPendingLayout
        ? classNames.classNames([skeletonStyles.row, skeletonStyles.headerHeight])
        : rowClasses;
    const accCellContent = (jsxRuntime.jsx("div", { style: 'width:1px;height:0px;overflow:hidden;', children: translations.collection_loading() }));
    // JET-75866 - dropdown should only show three skeletons on initial display with collectionTemplate
    // The number of skeleton rows will be determined as follows:
    // 1. if we are rendering load more rows -> 3 skeletons
    // 2. if we are rendering initial loading in embedded mode -> 2 skeletons (specs need 3 rows - 1 header + 2 skeletons)
    // 3. if we are rendering initial loading in non-embedded mode -> 25 skeletons
    const skeletonsCount = isLoadMore ? 3 : isEmbedded ? 2 : 25;
    return (jsxRuntime.jsx("tr", { class: LoadMoreCollection.LOADMORE_STYLE_CLASS, role: 'row', children: jsxRuntime.jsx("td", { class: cellClasses, colSpan: colspan, role: 'gridcell', tabIndex: hasTabIndex ? 0 : -1, "data-oj-cell-type": isPendingLayout ? 'pending' : isLoadMore ? 'loadMore' : 'loading', ...(isPendingLayout
                ? { 'data-oj-table-pending': tableId }
                : isLoadMore
                    ? { 'data-oj-table-loadmore': tableId }
                    : { 'data-oj-table-loading': tableId }), "data-oj-table-focusable": tableId, children: jsxRuntime.jsxs("div", { class: innerCellClasses, children: [accCellContent, [...Array(skeletonsCount)].map((_element, index) => (jsxRuntime.jsx("div", { class: index === 0 ? firstRowClasses : rowClasses, children: jsxRuntime.jsx(TableSkeleton, { height: "4x", isHighContrast: isHighContrast }) })))] }) }) }));
}

/**
 * The internal component used to render a single colgroup in Table.
 */
function TableColGroup({ tableId, isRendered, columnWidthsArray }) {
    const getColStyle = (columnWidth) => {
        return columnWidth == null ? '' : `width:${columnWidth}px;`;
    };
    return isRendered ? (jsxRuntime.jsx("colgroup", { children: columnWidthsArray.map((columnWidth) => {
            return jsxRuntime.jsx("col", { style: getColStyle(columnWidth), "data-oj-table-col": tableId });
        }) })) : null;
}

exports.TABLE_FOCUSABLE_SELECTOR = TABLE_FOCUSABLE_SELECTOR;
exports.TABLE_HEADER_SELECTOR = TABLE_HEADER_SELECTOR;
exports.TABLE_NODATA_SELECTOR = TABLE_NODATA_SELECTOR;
exports.TABLE_ROW_SELECTOR = TABLE_ROW_SELECTOR;
exports.TableColGroup = TableColGroup;
exports.TableSkeletonRow = TableSkeletonRow;
exports.applyScrollDetail = applyScrollDetail;
exports.getBoundaryColumnKey = getBoundaryColumnKey;
exports.getColumnIndexForHeaderFooterElement = getColumnIndexForHeaderFooterElement;
exports.getColumnKey = getColumnKey;
exports.getDataCellClassArray = getDataCellClassArray;
exports.getElementFromCell = getElementFromCell;
exports.getFirstRenderedRowKey = getFirstRenderedRowKey;
exports.getFormTextAlign = getFormTextAlign;
exports.getHeaderFooterCellClassArray = getHeaderFooterCellClassArray;
exports.getHeaderRowElement = getHeaderRowElement;
exports.getHorizontalScrollPosition = getHorizontalScrollPosition;
exports.getLastRenderedRowKey = getLastRenderedRowKey;
exports.getNoDataCellElement = getNoDataCellElement;
exports.getRowIndexForRowElement = getRowIndexForRowElement;
exports.getRowIndexForRowKey = getRowIndexForRowKey;
exports.getRowKeyForRowIndex = getRowKeyForRowIndex;
exports.getScrollerBottom = getScrollerBottom;
exports.getTableBodyElement = getTableBodyElement;
exports.getVerticalScrollPosition = getVerticalScrollPosition;
exports.hasFooterCells = hasFooterCells;
exports.hasLoadMoreCell = hasLoadMoreCell;
exports.hasNoDataCell = hasNoDataCell;
exports.isAutoFocusCell = isAutoFocusCell;
exports.logicalCellExtractor = logicalCellExtractor;
exports.resetVerticalScrollPosition = resetVerticalScrollPosition;
exports.useScrollHandling = useScrollHandling;
exports.useSizingManager = useSizingManager;
