// @oracle/oraclejet-preact/cjs/TruncatingLink-867da10e.js
/* @oracle/oraclejet-preact: undefined */
'use strict';

var jsxRuntime = require('preact/jsx-runtime');
var classNames = require('./classNames-c14c6ef3.js');
var usePressClick = require('./usePressClick-7a34624a.js');
var useComponentTheme = require('./useComponentTheme-082fc8e4.js');
var useInteractionStyle = require('./useInteractionStyle-c203a8a0.js');
var UNSAFE_Link_themes_redwood_LinkTheme = require('./UNSAFE_Link/themes/redwood/LinkTheme.js');
require('./TabbableModeContext-e99d527e.js');
var useTabbableMode = require('./useTabbableMode-dbd254e8.js');
var useTestId = require('./useTestId-4f53d561.js');
var useDetectTruncation = require('./useDetectTruncation-a5b9dd16.js');
var useTooltip = require('./useTooltip-34accecc.js');
var hooks = require('preact/hooks');
var mergeProps = require('./mergeProps-e3da7237.js');
require('./LayerHost-3cd73603.js');
require('preact/compat');


require('./hooks/UNSAFE_useTooltip/themes/redwood/TooltipContentVariants.css.js');
require('./tooltipUtils-2bc15136.js');
var clientHints = require('./clientHints-c40c5250.js');
var Text = require('./Text-056f1ec5.js');

const noop = () => { };
/**
 * TruncatingLink component creates a hyperlink that shows a tooltip and ellipsis at the end of text when truncated.
 */
function TruncatingLink({ href, variant, placement = 'standalone', target, testId, 'aria-describedby': ariaDescribedBy, 'aria-label': accessibleLabel, 'aria-labelledby': ariaLabelledBy, onClick, isDisabled, lineClamp, children }) {
    const { detectTruncationProps, isTextTruncated } = useDetectTruncation.useDetectTruncation();
    const { interactionProps, applyPseudoHoverStyle, applyHoverStyle } = useInteractionStyle.useInteractionStyle();
    const clientHints$1 = clientHints.getClientHints();
    const { classes } = useComponentTheme.useComponentTheme(UNSAFE_Link_themes_redwood_LinkTheme.LinkRedwoodTheme, {
        needsEventsHover: applyHoverStyle ? 'isNeedsEventsHover' : 'notNeedsEventsHover',
        pseudoHover: applyPseudoHoverStyle ? 'isPseudoHover' : 'notPseudoHover',
        variant,
        placement,
        disabled: isDisabled ? 'isDisabled' : 'notDisabled',
        truncatingLinkStyles: placement !== 'embedded' ? 'styles' : 'none',
        truncatingLinkWhiteSpace: placement !== 'embedded' && !lineClamp ? 'nowrap' : 'none',
        safariStyles: clientHints$1.browser === 'safari' ? 'isSafari' : 'notSafari'
    });
    const { isTabbable } = useTabbableMode.useTabbableMode();
    const linkClasses = classNames.classNames([classes]);
    const { pressProps } = usePressClick.usePressClick(onClick ?? noop);
    const testIdProps = useTestId.useTestId(testId);
    const [shownText] = hooks.useState(children || href);
    const { tooltipContent, tooltipProps } = useTooltip.useTooltip({
        text: shownText,
        isDisabled: !isTextTruncated || isDisabled
    });
    const { 'aria-describedby': notNeeded, ...validTooltipProps } = tooltipProps; //remove aria-describedby to prevent screen readers reading the same text multiple times
    return (jsxRuntime.jsxs(jsxRuntime.Fragment, { children: [jsxRuntime.jsx("a", { ...mergeProps.mergeProps(validTooltipProps, detectTruncationProps, interactionProps, testIdProps, !isDisabled ? pressProps : {}), tabIndex: isTabbable && !isDisabled ? 0 : -1, href: !isDisabled ? href : undefined, "aria-describedby": ariaDescribedBy, "aria-label": accessibleLabel, "aria-labelledby": ariaLabelledBy, target: target, class: linkClasses, children: jsxRuntime.jsx(Text.Text, { lineClamp: lineClamp, size: "md", variant: "inherit", children: shownText }) }), tooltipContent] }));
}

exports.TruncatingLink = TruncatingLink;
