// @oracle/oraclejet-preact/cjs/UNSAFE_FlatTreeView.js
/* @oracle/oraclejet-preact: undefined */
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var jsxRuntime = require('preact/jsx-runtime');
var hooks = require('preact/hooks');
var mergeProps = require('./mergeProps-e3da7237.js');
var collectionUtils = require('./collectionUtils-f1d3183f.js');
var LoadMoreCollection = require('./LoadMoreCollection-eb1f2603.js');
var classNames = require('./classNames-c14c6ef3.js');
var TabbableModeContext = require('./TabbableModeContext-e99d527e.js');
var useId = require('./useId-6c0eeb27.js');
var View = require('./View-ed116d78.js');
var UNSAFE_FlatTreeView_themes_TreeViewStyles_css = require('./UNSAFE_FlatTreeView/themes/TreeViewStyles.css.js');
var UNSAFE_FlatTreeView_themes_TreeViewContract_css = require('./UNSAFE_FlatTreeView/themes/TreeViewContract.css.js');
var Expander = require('./Expander-f4c574e5.js');
var Flex = require('./Flex-3a257bc3.js');
var Skeleton = require('./Skeleton-8e35c7d1.js');
var useHover = require('./useHover-49b0430d.js');
var clientHints = require('./clientHints-c40c5250.js');
var UNSAFE_FlatTreeView_themes_redwood_TreeViewTheme_css = require('./UNSAFE_FlatTreeView/themes/redwood/TreeViewTheme.css.js');
var useSelection = require('./useSelection-37939f09.js');
var useAdditiveSelection = require('./useAdditiveSelection-bf845689.js');
require('./ReorderableContext-16396e17.js');
var keys = require('./keys-25710458.js');
require('./LayerHost-3cd73603.js');
require('preact/compat');
var useExpandCollapse = require('./useExpandCollapse-ef990199.js');
var useComponentTheme = require('./useComponentTheme-082fc8e4.js');
var useUser = require('./useUser-1295b5cb.js');
var useNavigation = require('./useNavigation-fa3f5a50.js');
var Selector = require('./Selector-6550b8e0.js');
var LiveRegion = require('./LiveRegion-62ce9ac7.js');
var HiddenAccessible = require('./HiddenAccessible-087eb49a.js');
var useTranslationBundle = require('./useTranslationBundle-12dd087d.js');
require('preact');
require('./Collection-e27077ff.js');
require('./useViewportIntersect-2e802bac.js');
require('./vanilla-extract-dynamic.esm-2e1b7c25.js');
require('./borders-4b8488cb.js');
require('./dimensions-6101ca1f.js');
require('./size-1914356b.js');
require('./utils-c0efcd5e.js');
require('./themeContract.css-f3d431c6.js');
require('./colorUtils-01f3caa2.js');
require('./_curry1-e8f0d7ea.js');
require('./flexitem-fee13e26.js');

require('./vanilla-extract-sprinkles-createRuntimeSprinkles.esm-d68f3e0f.js');
require('./colors-f3ce979a.js');
require('./padding-ba041f6a.js');
require('./arrayUtils-7d8dcfc3.js');
require('./aria-889aedd0.js');
require('./theme-dfd6a9fe.js');
require('./Theme-f06687af.js');
require('./mergeInterpolations-6727b536.js');
require('./mergeDeepWithKey-210b024d.js');
require('./_curry3-18677bca.js');
require('./_curry2-c15d89cd.js');
require('./_isObject-28636267.js');
require('./EnvironmentProvider-5f2435e6.js');
require('./LayerManager-c535d963.js');
require('./useDensity-0b9fd9ca.js');
require('./useTestId-4f53d561.js');
require('./equals-098f7781.js');
require('./_isArray-73160ad5.js');
require('./_isString-f4443c9e.js');

require('./EmbeddedIconButton-cc159a01.js');
require('./BaseButton-d8f96bfa.js');
require('./useActive-6770f917.js');
require('./useToggle-3ebba7d8.js');
require('./useColorScheme-f8b1447d.js');
require('./UNSAFE_BaseButton/themes/redwood/BaseButtonTheme.js');
require('./UNSAFE_BaseButton/themes/BaseButtonStyles.css.js');

require('./UNSAFE_BaseButton/themes/redwood/BaseButtonVariants.css.js');
require('./vanilla-extract-recipes-createRuntimeFn.esm-d1301b2d.js');

require('./BareButton-e8a030d7.js');
require('./useBareButton-463210c9.js');
require('./usePress-886180e4.js');
require('./usePressClick-7a34624a.js');
require('./useTabbableMode-dbd254e8.js');
require('./buttonUtils-dff2ea3e.js');
require('./id-86356250.js');
require('./unsafeDomAccess-c3a492f0.js');
require('./ButtonLabelLayout-c876923f.js');
require('./Text-056f1ec5.js');
require('./UNSAFE_Text/themes/TextStyles.css.js');

require('./useButtonLabelLayoutTheme-d09db0c9.js');
require('./UNSAFE_ButtonLabelLayout/themes/redwood/ButtonLabelLayoutTheme.js');
require('./UNSAFE_ButtonLabelLayout/themes/ButtonLabelLayoutStyles.css.js');

require('./UNSAFE_ButtonLabelLayout/themes/redwood/ButtonLabelLayoutBaseTheme.css.js');

require('./UNSAFE_ButtonLabelLayout/themes/redwood/ButtonLabelLayoutVariants.css.js');

require('./ChevronDown-55b2c0d4.js');
require('./Icon-56789beb.js');
require('./useTooltip-34accecc.js');
require('./useTooltipControlled-0a95856c.js');
require('./useFocus-1b288fb9.js');
require('./useTouch-4dec8729.js');
require('./useAnimation-65c0e455.js');
require('./useThemeInterpolations-a37b37d9.js');
require('./useScale-db16c2b8.js');
require('./hooks/UNSAFE_useTooltip/themes/redwood/TooltipContentTheme.js');
require('./hooks/UNSAFE_useTooltip/themes/TooltipContentStyles.css.js');

require('./hooks/UNSAFE_useTooltip/themes/redwood/TooltipContentBaseTheme.css.js');

require('./hooks/UNSAFE_useTooltip/themes/redwood/TooltipContentVariants.css.js');

require('./tooltipUtils-2bc15136.js');
require('./Floating-79cbe5ad.js');
require('./useFloating-c753a7f9.js');
require('./positionUtils-a780137e.js');
require('./refUtils-a9872e75.js');
require('./useOutsideClick-dbf3d604.js');
require('./Layer-b60deb1b.js');
require('./UNSAFE_Layer/themes/LayerStyles.css.js');

require('./useModal-0739a528.js');
require('./UNSAFE_Floating/themes/redwood/FloatingTheme.js');
require('./UNSAFE_Floating/themes/FloatingStyles.css.js');

require('./UNSAFE_Floating/themes/redwood/FloatingBaseTheme.css.js');

require('./UNSAFE_Floating/themes/redwood/FloatingVariants.css.js');
require('./UNSAFE_Floating/themes/FloatingContract.css.js');
require('./useUnsafeDomElementRef-f22a2f0c.js');
require('./logger-2b636482.js');
require('./UNSAFE_Icon/themes/IconStyle.css.js');

require('./ChevronRight-d1c46e18.js');
require('./boxalignment-e287d370.js');

require('./flexbox-2cae9a01.js');


require('./UNSAFE_Skeleton/themes/redwood/SkeletonTheme.js');
require('./UNSAFE_Skeleton/themes/SkeletonStyles.css.js');

require('./UNSAFE_Skeleton/themes/redwood/SkeletonBaseTheme.css.js');

require('./UNSAFE_Skeleton/themes/redwood/SkeletonVariants.css.js');

require('./textSelectionUtils-57082239.js');
require('./index-0aa99b92.js');
require('./UNSAFE_Layer/themes/LayerHostStyles.css.js');

require('./useCurrentKey-1b9e3382.js');
require('./useReorderableContext-92514102.js');
require('./useCollectionFocusRing-bed0aa16.js');
require('./useTabbableModeSet-5bce852a.js');
require('./BaseSelector-2dc69917.js');
require('./StyledCheckbox-9047ed7d.js');
require('./Check-1272bb86.js');
require('./CheckboxOff-a3ed334d.js');
require('./CheckboxOn-b5dda28b.js');
require('./CheckboxMixed-a17c8298.js');
require('./UNSAFE_Checkbox/themes/CheckboxIconContract.css.js');
require('./UNSAFE_Checkbox/themes/redwood/CheckboxIconTheme.js');
require('./UNSAFE_Checkbox/themes/CheckboxIconStyles.css.js');

require('./UNSAFE_Checkbox/themes/redwood/CheckboxIconBaseTheme.css.js');

require('./UNSAFE_Checkbox/themes/redwood/CheckboxIconVariants.css.js');


require('./PRIVATE_BaseSelector/themes/SelectorStyles.css.js');

require('./CollectionInteractionContext-f84ff7a1.js');
require('./useCollectionInteractionContext-6fc0200e.js');



/**
 * Allows to specify the time delay for rendering the component
 **/
const timerValue$1 = 50;
/**
 * Represents the skeleton for a tree item in expanding state
 **/
function NodeSkeletonContainer({ children, indentation = 0 }) {
    const [isVisible, setIsVisible] = hooks.useState(false);
    hooks.useEffect(() => {
        setTimeout(() => {
            setIsVisible(true);
        }, timerValue$1);
    }, []);
    return isVisible && children ? (jsxRuntime.jsx("div", { role: "presentation", children: jsxRuntime.jsx(View.View, { paddingInlineStart: indentation, children: [...Array(1)].map((_element, index) => children(index)) }) })) : null;
}

function TreeItemSkeleton({ indentation = 0, isLeaf = false }) {
    return (jsxRuntime.jsx(View.View, { paddingInlineStart: calcIndentation(indentation, isLeaf ? undefined : true), children: jsxRuntime.jsx(View.View, { padding: [0, '4px'], children: jsxRuntime.jsxs(Flex.Flex, { height: "9x", align: "center", gap: "2x", children: [!isLeaf && jsxRuntime.jsx(Skeleton.Skeleton, { width: "6x", height: "6x", borderRadius: "4px" }), jsxRuntime.jsx(Skeleton.Skeleton, { height: "5x", borderRadius: "4px" })] }) }) }));
}
function TreeItemSkeletons() {
    return (jsxRuntime.jsxs(jsxRuntime.Fragment, { children: [jsxRuntime.jsx(TreeItemSkeleton, {}), jsxRuntime.jsx(TreeItemSkeleton, { indentation: 1 }), jsxRuntime.jsx(TreeItemSkeleton, { indentation: 2, isLeaf: true })] }));
}

function TreeItem({ itemKey, depth, context, children, isExpanded, onToggle, selected, isSelectable, isCheckboxless, isTabbable, isLoading, isFocusRingVisible }) {
    const rootRef = hooks.useRef(null);
    const indentation = calcIndentation(depth, isExpanded);
    const skeletonIndentation = calcIndentation(depth + 1, isExpanded);
    const { isHover, hoverProps } = useHover.useHover();
    const classes = classNames.classNames([
        UNSAFE_FlatTreeView_themes_TreeViewStyles_css.styles.treeItemStyles.base,
        isFocusRingVisible && UNSAFE_FlatTreeView_themes_TreeViewStyles_css.styles.treeItemStyles.focused,
        selected === true && UNSAFE_FlatTreeView_themes_TreeViewStyles_css.styles.treeItemStyles.selected,
        selected === true && isCheckboxless && UNSAFE_FlatTreeView_themes_TreeViewStyles_css.styles.treeItemStyles.checkboxlessSelected,
        isSelectable && !selected && isHover && UNSAFE_FlatTreeView_themes_TreeViewStyles_css.styles.treeItemStyles.hovered
    ]);
    return (jsxRuntime.jsxs(jsxRuntime.Fragment, { children: [jsxRuntime.jsx("div", { ref: rootRef, id: useId.useId(), tabIndex: isMobile() ? -1 : undefined, "data-oj-key": itemKey, "data-oj-tree-item": "", class: classes, role: "treeitem", "aria-level": depth + 1, "aria-selected": selected ? 'true' : 'false', ...(isExpanded !== undefined && { 'aria-expanded': isExpanded }), ...(typeof itemKey === 'number' && { 'data-oj-key-type': 'number' }), ...hoverProps, children: jsxRuntime.jsxs(jsxRuntime.Fragment, { children: [selected === true && isHover && jsxRuntime.jsx("div", { class: UNSAFE_FlatTreeView_themes_TreeViewStyles_css.styles.selectedHoveredOverlay }), jsxRuntime.jsx(TabbableModeContext.TabbableModeContext.Provider, { value: { isTabbable }, children: jsxRuntime.jsx(View.View, { paddingInlineStart: indentation, children: jsxRuntime.jsxs(Flex.Flex, { align: "center", children: [isExpanded !== undefined && (jsxRuntime.jsx(Expander.Expander, { itemKey: itemKey, expanded: isExpanded, onToggle: onToggle })), children(context)] }) }) })] }) }), isLoading ? (jsxRuntime.jsx(NodeSkeletonContainer, { indentation: skeletonIndentation, children: () => {
                    return jsxRuntime.jsx(TreeItemSkeletons, {});
                } })) : null] }));
}
const isMobile = () => {
    const { platform } = clientHints.getClientHints();
    return platform === 'ios' || platform === 'android';
};
const calcIndentation = (depth, isExpanded) => {
    const expanderPlaceholder = isExpanded === undefined ? '40px' : '0px';
    return `calc(${Math.max(0, depth)} * ${UNSAFE_FlatTreeView_themes_TreeViewContract_css.treeViewVars.indentWidth} + ${expanderPlaceholder})`;
};

/**
 * Allows to specify the time delay for rendering the component
 **/
const timerValue = 50;
/**
 * SkeletonContainer renders 'minimumCount' number of skeletons
 * of the variant specified from its child element - Skeleton's prop after
 * 'timerValue' ms delay
 **/
function SkeletonContainer({ children, minimumCount = 1, onSkeletonVisible }) {
    const [isVisible, setIsVisible] = hooks.useState(false);
    hooks.useEffect(() => {
        setTimeout(() => {
            setIsVisible(true);
            onSkeletonVisible?.();
        }, timerValue);
    }, [onSkeletonVisible]);
    const containerClasses = classNames.classNames([UNSAFE_FlatTreeView_themes_TreeViewStyles_css.styles.skeletonStyles.container]);
    return isVisible && children ? (jsxRuntime.jsx("div", { class: containerClasses, role: "presentation", children: [...Array(minimumCount)].map((_element, index) => children(index)) })) : null;
}

/**
 * Component that receives the flattened data and renders the hierarchy betweens items based on the shaped metadata.
 * In order to maximize performance, only items that are visible in the viewport are rendered.
 * This is an internal component that is used by FlatTreeView.
 */
function LoadMoreTreeCollection({ data, children, selectionMode, onToggle = () => { }, viewportConfig, hasMore = false, onLoadMore = () => { }, loadMoreIndicator = defaultLoadMoreIndicator, updateContext }) {
    const createTreeItem = (context) => {
        const key = context.data.metadata.key;
        const treeItemContext = {
            index: context.index,
            data: context.data.data,
            metadata: { key }
        };
        // allow container to inject additional things to context
        if (updateContext) {
            updateContext(treeItemContext);
        }
        const expanded = context.data.metadata.expanded;
        return (jsxRuntime.jsx(TreeItem, { itemKey: treeItemContext.metadata.key, context: treeItemContext, depth: context.data.metadata.treeDepth, isExpanded: expanded === 'ongoing' ? true : expanded, onToggle: onToggle, isSelectable: selectionMode !== 'none', isCheckboxless: selectionMode === 'single', selected: context.data.metadata.selected, isLoading: expanded === 'ongoing', isTabbable: false, isFocusRingVisible: context.data.metadata.hasFocus, children: children }, treeItemContext.metadata.key));
    };
    return (jsxRuntime.jsx(jsxRuntime.Fragment, { children: data === null ? null : (jsxRuntime.jsx(LoadMoreCollection.LoadMoreCollection, { data: data, viewportConfig: viewportConfig, hasMore: hasMore, onLoadMore: onLoadMore, loadMoreIndicator: loadMoreIndicator, children: createTreeItem })) }));
}
const defaultLoadMoreIndicator = (jsxRuntime.jsx("div", { class: LoadMoreCollection.LOADMORE_STYLE_CLASS, children: jsxRuntime.jsx(SkeletonContainer, { minimumCount: 1, children: () => {
            return jsxRuntime.jsx(TreeItemSkeletons, {});
        } }) }));

/**
 * Item selector for tree item
 */
const TREE_ITEM_SELECTOR = '[data-oj-tree-item]';

function useTreeNavigation(data, rootRef, isExpanded, idExtractor) {
    const { direction } = useUser.useUser();
    const arrowKeyAndHandlers = {
        ArrowUp: (key) => collectionUtils.getPrevNextKey(rootRef.current, key, true, TREE_ITEM_SELECTOR),
        ArrowDown: (key) => collectionUtils.getPrevNextKey(rootRef.current, key, false, TREE_ITEM_SELECTOR)
    };
    const { navigationProps, currentKey, onCurrentKeyChange, showFocusRing } = useNavigation.useNavigation(data, rootRef, idExtractor, TREE_ITEM_SELECTOR, arrowKeyAndHandlers, undefined, // currentItemOverride,
    undefined, //onPersistCurrentItem,
    undefined, undefined, true, // should always allow shift key
    false, // TODO: update allowTabbableMode when F2 mode is added
    true, false);
    const itemKeyExtractor = hooks.useCallback((element) => {
        if (element.getAttribute('role') === 'tree') {
            return currentKey === undefined ? null : currentKey;
        }
        return collectionUtils.keyExtractor(element, TREE_ITEM_SELECTOR);
    }, [currentKey, collectionUtils.keyExtractor]);
    const onKeyUp = hooks.useCallback((event) => {
        if (event.key === 'ArrowLeft' || event.key === 'ArrowRight') {
            const isRtl = direction === 'rtl';
            const isLeftArrow = event.key === 'ArrowLeft' && !isRtl;
            const itemKey = itemKeyExtractor(event.target);
            if (itemKey) {
                const contains = isExpanded(itemKey);
                // arrow left on a collapsed node, go to its parent
                if (isLeftArrow && !contains && rootRef.current) {
                    const item = collectionUtils.findElementByKey(rootRef.current, itemKey, TREE_ITEM_SELECTOR);
                    if (item) {
                        const treeDepth = item.getAttribute('aria-level');
                        const parent = treeDepth !== null &&
                            findParentElementByKey(rootRef.current, itemKey, Number(treeDepth), TREE_ITEM_SELECTOR);
                        const newKey = parent && itemKeyExtractor(parent);
                        if (keys.isKeyDefined(newKey)) {
                            onCurrentKeyChange?.({ value: newKey });
                        }
                    }
                }
                else if (!isLeftArrow && contains) {
                    // arrow right on an expanded node, go to its first child
                    const newKey = collectionUtils.getPrevNextKey(rootRef.current, itemKey, false, TREE_ITEM_SELECTOR)();
                    if (keys.isKeyDefined(newKey)) {
                        onCurrentKeyChange?.({ value: newKey });
                    }
                }
            }
        }
    }, [direction, isExpanded, onCurrentKeyChange, itemKeyExtractor]);
    const treeNavigationProps = mergeProps.mergeProps(navigationProps, { onKeyUp });
    return { treeNavigationProps, currentKey, showFocusRing };
}
const findParentElementByKey = (root, key, treeDepth, itemSelector) => {
    if (root && key !== null) {
        const items = Array.from(root.querySelectorAll(itemSelector));
        const index = items.findIndex((i) => i.dataset['ojKey'] == key);
        for (let i = index; i >= 0; i--) {
            const level = items[i].getAttribute('aria-level');
            if (level !== null && Number(level) == treeDepth - 1) {
                return items[i];
            }
        }
    }
    return null;
};

/**
 * A FlatTreeView displays flattened data as a tree with highly interactive features.
 */
function FlatTreeView({ 'aria-label': ariaLabel, 'aria-labelledby': ariaLabelledBy, 'aria-describedby': ariaDescribedBy, data, selectionMode = 'none', selectedKeys = { all: false, keys: new Set() }, onSelectionChange, onToggle, viewportConfig, ...rest }) {
    const rootRef = hooks.useRef(null);
    const [liveRegion, setLiveRegion] = hooks.useState();
    const { collection_loading, collection_loaded } = useTranslationBundle.useTranslationBundle('@oracle/oraclejet-preact');
    const loadingDescId = useId.useId();
    const isExpanded = hooks.useCallback((key) => {
        if (rootRef.current) {
            const elem = collectionUtils.findElementByKey(rootRef.current, key, TREE_ITEM_SELECTOR);
            return elem?.getAttribute('aria-expanded') === 'true';
        }
        return false;
    }, []);
    const { treeNavigationProps, currentKey, showFocusRing } = useTreeNavigation(data, rootRef, isExpanded, idExtractor);
    const items = hooks.useMemo(() => data?.map((value) => {
        const itemKey = value.metadata.key;
        const newValue = { ...value };
        newValue.metadata.selected = keys.containsKey(selectedKeys, itemKey);
        newValue.metadata.hasFocus = currentKey === itemKey && showFocusRing;
        return newValue;
    }) || [], [data, selectedKeys, currentKey, showFocusRing]);
    const expandCollapseProps = useExpandCollapse.useExpandCollapse((element) => {
        if (element.getAttribute('role') === 'tree') {
            return currentKey === undefined ? null : currentKey;
        }
        return collectionUtils.keyExtractor(element, TREE_ITEM_SELECTOR);
    }, isExpanded, onToggle, true);
    // tracking the anchor key which is used for shift+click selection
    const anchorKey = hooks.useRef();
    const handleSelectionChange = hooks.useCallback((detail) => {
        if (onSelectionChange) {
            if (detail.value.all === false && detail.value.keys.size > 0) {
                // during regular selection, the last key selected by a user gesture is the anchor key
                anchorKey.current = Array.from(detail.value.keys.values()).pop();
            }
            onSelectionChange(detail);
        }
    }, [anchorKey, onSelectionChange]);
    const handleSelectionRangeChange = hooks.useCallback((detail) => {
        if (data && onSelectionChange) {
            const value = handleSelectionRange(detail, data);
            if (Array.isArray(value)) {
                if (detail.eventType === 'shiftSpace') {
                    // during range selection by shift+space, the first key selected by a user gesture is the anchor key
                    anchorKey.current = detail.value.start;
                }
                else {
                    // during range selection by shift+click/arrow keys, the last key selected by a user gesture is the anchor key
                    anchorKey.current = detail.value.end;
                }
                if (selectionMode === 'multipleToggle') {
                    const keys$1 = keys.addKeys(selectedKeys, value);
                    onSelectionChange({ value: keys$1, target: null });
                }
                else {
                    onSelectionChange({ value: { all: false, keys: new Set(value) }, target: null });
                }
            }
        }
    }, [data, selectedKeys, onSelectionChange]);
    // single or multiple selection mode
    const { selectionProps } = useSelection.useSelection((element) => {
        if (element === rootRef.current) {
            return currentKey === undefined ? null : currentKey;
        }
        return collectionUtils.keyExtractor(element, TREE_ITEM_SELECTOR);
    }, selectedKeys, selectionMode === 'multipleToggle' ? 'none' : selectionMode, false, 'replace', handleSelectionChange, anchorKey.current, currentKey, (currentKey, isPrev) => collectionUtils.getPrevNextKey(rootRef.current, currentKey, isPrev, TREE_ITEM_SELECTOR), undefined, (key) => {
        if (rootRef.current) {
            const elem = collectionUtils.findElementByKey(rootRef.current, key, TREE_ITEM_SELECTOR);
            if (elem) {
                elem.scrollIntoView({ block: 'nearest' });
            }
        }
    }, handleSelectionRangeChange);
    // multipleToggle selection mode
    const { eventProps: multipleToggleSelectionProps } = useAdditiveSelection.useAdditiveSelection((element) => {
        if (element === rootRef.current) {
            return currentKey === undefined ? null : currentKey;
        }
        return collectionUtils.keyExtractor(element, TREE_ITEM_SELECTOR);
    }, selectedKeys, selectionMode === 'multipleToggle' ? 'multiple' : 'none', false, (key) => {
        if (data) {
            return data.findIndex((value) => value.metadata.key === key);
        }
        return -1;
    }, (startIndex, endIndex) => {
        if (data) {
            const keys = data.map((value) => value.metadata.key);
            return keys.slice(startIndex, endIndex);
        }
        return [];
    }, (currentKey, isPrev) => collectionUtils.getPrevNextKey(rootRef.current, currentKey, isPrev, TREE_ITEM_SELECTOR), currentKey, handleSelectionChange, handleSelectionRangeChange);
    //Note: onKeyUp doesn't work with the cmd key
    const selectAllWithKeyboard = hooks.useCallback((event) => {
        if (selectionMode === 'multiple' || selectionMode === 'multipleToggle') {
            if (onSelectionChange && (event.metaKey || event.ctrlKey) && event.key === 'a') {
                event.preventDefault();
                onSelectionChange({
                    value: { all: true, deletedKeys: new Set() },
                    target: event.target
                });
            }
        }
    }, [onSelectionChange, selectionMode]);
    const { classes } = useComponentTheme.useComponentTheme(UNSAFE_FlatTreeView_themes_redwood_TreeViewTheme_css.TreeViewRedwoodTheme);
    viewportConfig = collectionUtils.getViewportConfig(rootRef, viewportConfig);
    const updateContextCallback = hooks.useCallback((context) => {
        const key = context.metadata.key;
        context.selector =
            selectionMode === 'none' || selectionMode === 'single'
                ? undefined
                : (ariaLabel) => (jsxRuntime.jsx("div", { "aria-hidden": isMobile() ? undefined : true, children: jsxRuntime.jsx(Selector.Selector, { "aria-label": isMobile() ? ariaLabel : undefined, rowKey: key, selectedKeys: selectedKeys, onChange: onSelectionChange }) }));
    }, [selectionMode, selectedKeys, onSelectionChange]);
    const updateLiveRegion = hooks.useCallback(() => {
        setLiveRegion(collection_loading());
    }, [collection_loading]);
    return (jsxRuntime.jsxs(jsxRuntime.Fragment, { children: [jsxRuntime.jsx("div", { "aria-label": ariaLabel, "aria-labelledby": ariaLabelledBy, "aria-describedby": data == null
                    ? ariaDescribedBy
                        ? [...new Set(ariaDescribedBy.split(/\s+/).concat(loadingDescId))].join(' ')
                        : loadingDescId
                    : ariaDescribedBy, ref: rootRef, role: "tree", "aria-multiselectable": selectionMode === 'multiple' ? true : false, tabIndex: 0, class: classes, ...mergeProps.mergeProps(treeNavigationProps, selectionProps, multipleToggleSelectionProps, expandCollapseProps, { onKeyDown: selectAllWithKeyboard }), children: data == null ? (defaultLoadingIndicator(updateLiveRegion)) : (jsxRuntime.jsx(LoadMoreTreeCollection, { data: items, selectionMode: selectionMode, onToggle: onToggle, viewportConfig: viewportConfig, updateContext: updateContextCallback, ...rest })) }), jsxRuntime.jsx(HiddenAccessible.HiddenAccessible, { id: loadingDescId, children: collection_loading() }), jsxRuntime.jsx(LiveRegion.LiveRegion, { children: data == null ? liveRegion : liveRegion === collection_loading() ? collection_loaded() : '' })] }));
}
const defaultLoadingIndicator = (callback) => {
    return (jsxRuntime.jsx(SkeletonContainer, { minimumCount: 8, onSkeletonVisible: callback, children: () => {
            return jsxRuntime.jsx(TreeItemSkeletons, {});
        } }));
};
const idExtractor = (elem) => elem.id;
/**
 * A helper function to handle range selection
 */
const handleSelectionRange = (detail, data) => {
    const keys = data.map((value) => {
        return value.metadata.key;
    });
    const startIndex = keys.indexOf(detail.value.start);
    const endIndex = keys.indexOf(detail.value.end);
    const minIndex = Math.min(startIndex, endIndex);
    const maxIndex = Math.max(startIndex, endIndex);
    if (minIndex === -1) {
        // return a range to fetch so we can find all the keys, maxIndex should not be -1
        return { offset: 0, count: -1 };
    }
    else {
        // range of keys are in the current viewport, return them
        return keys.slice(minIndex, maxIndex + 1);
    }
};

exports.FlatTreeView = FlatTreeView;
