// @oracle/oraclejet-preact/cjs/UNSAFE_RichSelectionItem.js
/* @oracle/oraclejet-preact: undefined */
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var jsxRuntime = require('preact/jsx-runtime');
var hooks = require('preact/hooks');
require('./FormFieldContext-69fe83d2.js');
var useFormFieldContext = require('./useFormFieldContext-4632eb73.js');
var RichSelectionStartContext = require('./RichSelectionStartContext-3d10f9ed.js');
var RadioItemContext = require('./RadioItemContext-3a86c17d.js');
var TruncatingText = require('./TruncatingText-ac1f6793.js');
var classNames = require('./classNames-c14c6ef3.js');
var UNSAFE_RichSelectionItem_themes_RichSelectionItemStyles_css = require('./UNSAFE_RichSelectionItem/themes/RichSelectionItemStyles.css.js');
var View = require('./View-ed116d78.js');
var useTestId = require('./useTestId-4f53d561.js');
var AspectRatio = require('./AspectRatio-026ab1d8.js');
var Avatar = require('./Avatar-c84a09ae.js');
var Center = require('./Center-9a0ee4f3.js');
var Flex = require('./Flex-3a257bc3.js');
var useId = require('./useId-6c0eeb27.js');
require('./FormContext-d1d05967.js');
var useFormContext = require('./useFormContext-a794f7b6.js');
var usePress = require('./usePress-886180e4.js');
var useComponentTheme = require('./useComponentTheme-082fc8e4.js');
var useInteractionStyle = require('./useInteractionStyle-c203a8a0.js');
require('./TabbableModeContext-e99d527e.js');
var useTabbableMode = require('./useTabbableMode-dbd254e8.js');
var StyledCheckbox = require('./StyledCheckbox-9047ed7d.js');
var Check = require('./Check-1272bb86.js');
var RadioButtonOff = require('./RadioButtonOff-0b9700f9.js');
var RadioButtonOn = require('./RadioButtonOn-8264d513.js');
var HiddenAccessible = require('./HiddenAccessible-087eb49a.js');
var UNSAFE_RadioItem_themes_RadioIconContract_css = require('./UNSAFE_RadioItem/themes/RadioIconContract.css.js');
var UNSAFE_RadioItem_themes_RadioStyles_css = require('./UNSAFE_RadioItem/themes/RadioStyles.css.js');
var UNSAFE_RadioItem_themes_redwood_RadioIconTheme = require('./UNSAFE_RadioItem/themes/redwood/RadioIconTheme.js');
var useTranslationBundle = require('./useTranslationBundle-12dd087d.js');
var RadioItem = require('./RadioItem-c74ac827.js');
var CheckboxSetContext = require('./CheckboxSetContext-c62db26d.js');
var useFocusableTextField = require('./useFocusableTextField-2c36840d.js');
var StyledCard = require('./StyledCard-13c323e4.js');
require('preact');
require('./useTooltipOnTruncation-8867b10e.js');
require('./useDetectTruncation-a5b9dd16.js');
require('./truncationUtils-2d44b04b.js');
require('./useResizeObserver-a9e97180.js');
require('./useAddBusyState-d19ae1fa.js');
require('./BusyStateContext-86f40d3c.js');
require('./useBusyStateContext-362eee2a.js');
require('./unsafeDomAccess-c3a492f0.js');
require('./useTooltip-34accecc.js');
require('./useTooltipControlled-0a95856c.js');
require('./useHover-49b0430d.js');
require('./useToggle-3ebba7d8.js');
require('./useFocus-1b288fb9.js');
require('./useTouch-4dec8729.js');
require('./mergeProps-e3da7237.js');
require('./useAnimation-65c0e455.js');
require('./useThemeInterpolations-a37b37d9.js');
require('./useColorScheme-f8b1447d.js');
require('./LayerHost-3cd73603.js');
require('./index-0aa99b92.js');
require('preact/compat');
require('./UNSAFE_Layer/themes/LayerHostStyles.css.js');

require('./useScale-db16c2b8.js');
require('./theme-dfd6a9fe.js');
require('./Theme-f06687af.js');
require('./mergeInterpolations-6727b536.js');
require('./mergeDeepWithKey-210b024d.js');
require('./_curry3-18677bca.js');
require('./_curry1-e8f0d7ea.js');
require('./_curry2-c15d89cd.js');
require('./_isObject-28636267.js');
require('./useDensity-0b9fd9ca.js');
require('./hooks/UNSAFE_useTooltip/themes/redwood/TooltipContentTheme.js');
require('./hooks/UNSAFE_useTooltip/themes/TooltipContentStyles.css.js');

require('./hooks/UNSAFE_useTooltip/themes/redwood/TooltipContentBaseTheme.css.js');

require('./hooks/UNSAFE_useTooltip/themes/redwood/TooltipContentVariants.css.js');

require('./vanilla-extract-recipes-createRuntimeFn.esm-d1301b2d.js');
require('./tooltipUtils-2bc15136.js');
require('./EnvironmentProvider-5f2435e6.js');
require('./LayerManager-c535d963.js');
require('./Floating-79cbe5ad.js');
require('./useFloating-c753a7f9.js');
require('./useUser-1295b5cb.js');
require('./positionUtils-a780137e.js');
require('./refUtils-a9872e75.js');
require('./useOutsideClick-dbf3d604.js');
require('./arrayUtils-7d8dcfc3.js');
require('./Layer-b60deb1b.js');
require('./UNSAFE_Layer/themes/LayerStyles.css.js');

require('./useModal-0739a528.js');
require('./utils-c0efcd5e.js');
require('./themeContract.css-f3d431c6.js');
require('./colorUtils-01f3caa2.js');
require('./UNSAFE_Floating/themes/redwood/FloatingTheme.js');
require('./UNSAFE_Floating/themes/FloatingStyles.css.js');

require('./UNSAFE_Floating/themes/redwood/FloatingBaseTheme.css.js');

require('./UNSAFE_Floating/themes/redwood/FloatingVariants.css.js');
require('./vanilla-extract-dynamic.esm-2e1b7c25.js');
require('./UNSAFE_Floating/themes/FloatingContract.css.js');
require('./useUnsafeDomElementRef-f22a2f0c.js');
require('./logger-2b636482.js');
require('./UNSAFE_Text/themes/TextStyles.css.js');

require('./clientHints-c40c5250.js');

require('./borders-4b8488cb.js');
require('./dimensions-6101ca1f.js');
require('./size-1914356b.js');
require('./flexitem-fee13e26.js');

require('./vanilla-extract-sprinkles-createRuntimeSprinkles.esm-d68f3e0f.js');
require('./colors-f3ce979a.js');
require('./padding-ba041f6a.js');
require('./aria-889aedd0.js');
require('./equals-098f7781.js');
require('./_isArray-73160ad5.js');
require('./_isString-f4443c9e.js');

require('./UNSAFE_Avatar/themes/redwood/AvatarTheme.js');
require('./UNSAFE_Avatar/themes/AvatarStyles.css.js');

require('./UNSAFE_Avatar/themes/redwood/AvatarBaseTheme.css.js');


require('./UNSAFE_Avatar/themes/redwood/AvatarVariants.css.js');

require('./ImageVars.css-3fbb1c0b.js');
require('./boxalignment-e287d370.js');

require('./flexbox-2cae9a01.js');


require('./useActive-6770f917.js');
require('./CheckboxOff-a3ed334d.js');
require('./Icon-56789beb.js');
require('./UNSAFE_Icon/themes/IconStyle.css.js');

require('./CheckboxOn-b5dda28b.js');
require('./CheckboxMixed-a17c8298.js');
require('./UNSAFE_Checkbox/themes/CheckboxIconContract.css.js');
require('./UNSAFE_Checkbox/themes/redwood/CheckboxIconTheme.js');
require('./UNSAFE_Checkbox/themes/CheckboxIconStyles.css.js');

require('./UNSAFE_Checkbox/themes/redwood/CheckboxIconBaseTheme.css.js');

require('./UNSAFE_Checkbox/themes/redwood/CheckboxIconVariants.css.js');




require('./UNSAFE_RadioItem/themes/RadioIconStyles.css.js');

require('./UNSAFE_RadioItem/themes/redwood/RadioIconBaseTheme.css.js');

require('./UNSAFE_RadioItem/themes/redwood/RadioIconVariants.css.js');

require('./CheckboxRadioContext-3b134bd1.js');


require('./UNSAFE_UserAssistance/themes/redwood/UserAssistanceVariants.css.js');


require('./UNSAFE_MessageBanner/themes/redwood/MessageBannerVariants.css.js');

require('./IconButton-7503293b.js');
require('./BaseButton-d8f96bfa.js');
require('./UNSAFE_BaseButton/themes/redwood/BaseButtonTheme.js');
require('./UNSAFE_BaseButton/themes/BaseButtonStyles.css.js');

require('./UNSAFE_BaseButton/themes/redwood/BaseButtonVariants.css.js');

require('./BareButton-e8a030d7.js');
require('./useBareButton-463210c9.js');
require('./usePressClick-7a34624a.js');
require('./buttonUtils-dff2ea3e.js');
require('./id-86356250.js');
require('./ButtonLabelLayout-c876923f.js');
require('./Text-056f1ec5.js');
require('./useButtonLabelLayoutTheme-d09db0c9.js');
require('./UNSAFE_ButtonLabelLayout/themes/redwood/ButtonLabelLayoutTheme.js');
require('./UNSAFE_ButtonLabelLayout/themes/ButtonLabelLayoutStyles.css.js');

require('./UNSAFE_ButtonLabelLayout/themes/redwood/ButtonLabelLayoutBaseTheme.css.js');

require('./UNSAFE_ButtonLabelLayout/themes/redwood/ButtonLabelLayoutVariants.css.js');


require('./MessageFormattingUtils-539d75b0.js');
require('./getLocale-8e17dc46.js');
require('./stringUtils-3e19c8af.js');
require('./TransitionGroup-b239d98f.js');
require('./MessagesContext-4e939750.js');


require('./UNSAFE_ComponentMessage/themes/redwood/ComponentMessageVariants.css.js');
require('./InputGroupContext-05f2a46f.js');
require('./Popup-de16d915.js');
require('./tabbableUtils-22a50472.js');
require('./head-68d0992f.js');
require('./_arity-c228159c.js');
require('./FocusTrap-4873a3f2.js');
require('./FocusTracker-e2ea43c3.js');
require('./PRIVATE_FocusTracker/themes/FocusTrackerStyles.css.js');

require('./Modal-b751b643.js');
require('./UNSAFE_Modal/themes/ModalStyles.css.js');

require('./popupUtils-1d123cbe.js');
require('./WindowOverlay-85d999fa.js');
require('./UNSAFE_WindowOverlay/themes/WindowOverlayStyles.css.js');

require('./UNSAFE_WindowOverlay/themes/WindowOverlayContract.css.js');
require('./UNSAFE_WindowOverlay/themes/redwood/WindowOverlayTheme.js');
require('./UNSAFE_WindowOverlay/themes/redwood/WindowOverlayBaseTheme.css.js');

require('./UNSAFE_WindowOverlay/themes/redwood/WindowOverlayVariants.css.js');
require('./usePopupAnimation-f3b34413.js');
require('./useAnimationStatus-8075dc19.js');
require('./popupAnimationUtils-dcace427.js');
require('./LiveRegion-62ce9ac7.js');

require('./UNSAFE_Popup/themes/PopupContract.css.js');
require('./UNSAFE_Popup/themes/redwood/PopupTheme.js');
require('./UNSAFE_Popup/themes/redwood/PopupBaseTheme.css.js');

require('./UNSAFE_Popup/themes/redwood/PopupVariants.css.js');

require('./UNSAFE_Separator/themes/SeparatorStyles.css.js');



require('./UNSAFE_Label/themes/redwood/LabelVariants.css.js');
require('./IconUserAssistance-89913ad1.js');
require('./Help-dbc23ccd.js');
require('./UNSAFE_UserAssistance/themes/UserAssistanceStyles.css.js');
require('./InlineHelp-2add2d34.js');
require('./UNSAFE_UserAssistance/themes/redwood/UserAssistanceTheme.js');
require('./UNSAFE_UserAssistance/themes/redwood/UserAssistanceBaseTheme.css.js');
require('./UNSAFE_RadioItem/themes/redwood/RadioTheme.js');
require('./UNSAFE_RadioItem/themes/redwood/RadioBaseTheme.css.js');

require('./UNSAFE_RadioItem/themes/redwood/RadioVariants.css.js');

require('./useFocusWithin-eb7f956a.js');
require('./PRIVATE_StyledCard/themes/StyledCardStyles.css.js');


// INFO: Label and SecondaryText Content handling
const RichSelectionTextContent = ({ secondaryTextId, isDisabled, isLayoutXL, label, labelId, secondaryText }) => {
    const labelContent = (jsxRuntime.jsx(TruncatingText.TruncatingText, { variant: isDisabled ? 'disabled' : 'primary', lineClamp: isLayoutXL ? 1 : 2, id: labelId, children: label }));
    const classes = classNames.classNames([
        isLayoutXL ? UNSAFE_RichSelectionItem_themes_RichSelectionItemStyles_css.styles.densityWrapper : UNSAFE_RichSelectionItem_themes_RichSelectionItemStyles_css.styles.growOneColumn,
        !isLayoutXL && !secondaryText && UNSAFE_RichSelectionItem_themes_RichSelectionItemStyles_css.styles.labelOnly
    ]);
    return (jsxRuntime.jsxs("div", { className: classes, children: [labelContent, secondaryText && (jsxRuntime.jsx(TruncatingText.TruncatingText, { size: "sm", variant: isDisabled ? 'disabled' : 'secondary', lineClamp: 3, id: secondaryTextId, children: secondaryText }))] }));
};

const ThumbnailXL = ({ thumbnailSrc, mediaAccessibleLabel, testId, isDisabled }) => {
    const testIdProps = useTestId.useTestId(testId ? `${testId}_thumbnail` : undefined);
    return (jsxRuntime.jsx(AspectRatio.AspectRatio, { ratio: "4/3", children: jsxRuntime.jsx("img", { alt: mediaAccessibleLabel, className: classNames.classNames([UNSAFE_RichSelectionItem_themes_RichSelectionItemStyles_css.styles.thumbnailImage, isDisabled && UNSAFE_RichSelectionItem_themes_RichSelectionItemStyles_css.styles.mediaDisabledOpacity]), role: !mediaAccessibleLabel ? 'presentation' : undefined, src: thumbnailSrc, ...testIdProps }) }));
};
const Thumbnail = ({ thumbnailSrc, mediaAccessibleLabel, layout, testId, isDisabled }) => {
    const thumbnailClass = classNames.classNames([
        layout === 'md' ? UNSAFE_RichSelectionItem_themes_RichSelectionItemStyles_css.styles.mdMediaVariant : UNSAFE_RichSelectionItem_themes_RichSelectionItemStyles_css.styles.smMediaVariant,
        isDisabled && UNSAFE_RichSelectionItem_themes_RichSelectionItemStyles_css.styles.mediaDisabledOpacity
    ]);
    const testIdProps = useTestId.useTestId(testId ? `${testId}_thumbnail` : undefined);
    return (jsxRuntime.jsx("img", { alt: mediaAccessibleLabel, className: thumbnailClass, role: !mediaAccessibleLabel ? 'presentation' : undefined, src: thumbnailSrc, ...testIdProps }));
};
const MediumRenderer = ({ mediaAccessibleLabel, thumbnailSrc, layout, avatar, testId, isDisabled }) => {
    const testIdProps = useTestId.useTestId(testId ? `${testId}_medium` : undefined);
    let mediumMediaItem;
    if (avatar) {
        const avatarElement = (jsxRuntime.jsx(Avatar.Avatar, { ...avatar, "aria-label": mediaAccessibleLabel, size: "md", ...testIdProps }));
        mediumMediaItem = isDisabled ? (jsxRuntime.jsx("span", { className: UNSAFE_RichSelectionItem_themes_RichSelectionItemStyles_css.styles.mediaDisabledOpacity, children: avatarElement })) : (avatarElement);
    }
    else {
        mediumMediaItem = (jsxRuntime.jsx(Thumbnail, { thumbnailSrc: thumbnailSrc, mediaAccessibleLabel: mediaAccessibleLabel, layout: layout, testId: testId, isDisabled: isDisabled }));
    }
    return jsxRuntime.jsx(Center.Center, { children: mediumMediaItem });
};
const SmallRenderer = ({ avatar, iconClass, isDisabled, layout, thumbnailSrc, mediaAccessibleLabel, testId }) => {
    const testIdProps = useTestId.useTestId(testId ? `${testId}_small` : undefined);
    const { iconWrapper, iconEnabledColor, iconDisabledColor, mediaDisabledOpacity } = UNSAFE_RichSelectionItem_themes_RichSelectionItemStyles_css.styles;
    let smallMediaItem;
    if (avatar) {
        const avatarElement = (jsxRuntime.jsx(Avatar.Avatar, { ...avatar, "aria-label": mediaAccessibleLabel, size: "xs", ...testIdProps }));
        smallMediaItem = isDisabled ? (jsxRuntime.jsx("span", { className: mediaDisabledOpacity, children: avatarElement })) : (avatarElement);
    }
    else if (thumbnailSrc) {
        smallMediaItem = (jsxRuntime.jsx(Thumbnail, { thumbnailSrc: thumbnailSrc, mediaAccessibleLabel: mediaAccessibleLabel, layout: layout, testId: testId, isDisabled: isDisabled, ...testIdProps }));
    }
    else if (iconClass) {
        const classes = classNames.classNames([
            iconClass,
            iconWrapper,
            isDisabled ? iconDisabledColor : iconEnabledColor
        ]);
        smallMediaItem = jsxRuntime.jsx("div", { class: classes, ...testIdProps });
    }
    return jsxRuntime.jsx(Center.Center, { children: smallMediaItem });
};
// INFO: Media Item handling
const RichSelectionMediaItem = ({ avatar, iconClass, isDisabled, hasSecondaryText, mediaAccessibleLabel, layout, thumbnailSrc, testId }) => {
    if (!layout) {
        return null;
    }
    switch (layout) {
        case 'xl':
            return (jsxRuntime.jsx(ThumbnailXL, { thumbnailSrc: thumbnailSrc, mediaAccessibleLabel: mediaAccessibleLabel, testId: testId, isDisabled: isDisabled }));
        case 'md':
            const mediumMedia = avatar ? { avatar } : thumbnailSrc ? { thumbnailSrc } : undefined;
            return (jsxRuntime.jsx(View.View, { alignSelf: hasSecondaryText ? 'start' : 'center', children: jsxRuntime.jsx(MediumRenderer, { mediaAccessibleLabel: mediaAccessibleLabel, layout: layout, testId: testId, isDisabled: isDisabled, ...mediumMedia }) }));
        case 'sm':
            const smallMedia = avatar
                ? { avatar, mediaAccessibleLabel }
                : thumbnailSrc
                    ? { thumbnailSrc, mediaAccessibleLabel }
                    : iconClass
                        ? { iconClass }
                        : undefined;
            return (jsxRuntime.jsx(View.View, { alignSelf: hasSecondaryText ? 'start' : 'center', children: jsxRuntime.jsx(SmallRenderer, { ...smallMedia, layout: layout, testId: testId, isDisabled: isDisabled }) }));
        default:
            return null;
    }
};

const RadioControl = ({ ariaLabel, value, inputRef, isFocus, isDisabled, isLayoutXL, isReadonly, isSelected, describedBy, labelledBy, onBlur, onFocus }) => {
    const { name, value: radioGroupValue, onCommit, setFocusIndex } = RadioItemContext.useRadioSetContext();
    const IconComp = isReadonly
        ? isSelected
            ? Check.SvgCheck
            : RadioButtonOff.SvgRadioButtonOff
        : isSelected
            ? RadioButtonOn.SvgRadioButtonOn
            : RadioButtonOff.SvgRadioButtonOff;
    const spanRef = hooks.useRef(null);
    const { interactionProps, applyActiveStyle } = useInteractionStyle.useInteractionStyle();
    const onChangeHandler = hooks.useCallback((event) => {
        event.stopPropagation();
        onCommit?.({ previousValue: radioGroupValue, value });
    }, [value, radioGroupValue, onCommit]);
    /**
     * When focused the view should have the respective RichSelectionItem.
     */
    hooks.useEffect(() => {
        const isActiveElement = document.activeElement === inputRef?.current;
        if (!isActiveElement && isFocus) {
            spanRef?.current?.scrollIntoView({ block: 'nearest' });
        }
    }, [inputRef, isFocus, spanRef]);
    const radioColor = isDisabled ? UNSAFE_RadioItem_themes_RadioStyles_css.radioDisabledColor : UNSAFE_RadioItem_themes_RadioStyles_css.radioDefaultColor;
    const { classes: radioIconClasses } = useComponentTheme.useComponentTheme(UNSAFE_RadioItem_themes_redwood_RadioIconTheme.RadioIconRedwoodTheme, {
        disabled: isDisabled ? 'isDisabled' : 'notDisabled',
        selected: isSelected ? 'isSelected' : 'notSelected',
        active: applyActiveStyle ? 'isActive' : 'notActive',
        richSelection: isLayoutXL ? 'isLayoutXL' : 'notLayoutXL'
    });
    const { isTabbable } = useTabbableMode.useTabbableMode();
    const { index } = RadioItemContext.useRadioItemContext();
    const tabIndex = isTabbable !== false && ((index === 0 && !radioGroupValue) || isSelected) ? 0 : -1;
    const translations = useTranslationBundle.useTranslationBundle('@oracle/oraclejet-preact');
    const readonlyLabel = `${ariaLabel} ${translations.formControl_readOnly()} ${RadioItem.getRadioSelectionText(translations, !!isSelected)}`;
    const hiddenInput = (jsxRuntime.jsx(HiddenAccessible.HiddenAccessible, { children: isReadonly ? (jsxRuntime.jsx("div", { "aria-describedby": describedBy, "data-checked": isSelected, ref: inputRef, tabIndex: tabIndex, onBlur: onBlur, onFocus: (event) => {
                setFocusIndex?.(index);
                onFocus?.(event);
            }, children: readonlyLabel })) : (jsxRuntime.jsx("input", { type: "radio", name: name, value: value, ...(!isDisabled && {
                onChange: onChangeHandler
            }), checked: isSelected, disabled: isDisabled, "aria-labelledby": labelledBy, "aria-describedby": describedBy, tabIndex: tabIndex, onBlur: onBlur, onFocus: onFocus, ref: inputRef })) }));
    // Radio control is visually hidden in readonly mode unless it is selected
    if (!isSelected && isReadonly) {
        return jsxRuntime.jsx("span", { class: UNSAFE_RichSelectionItem_themes_RichSelectionItemStyles_css.styles.inputElement, children: hiddenInput });
    }
    return (jsxRuntime.jsxs("span", { ref: spanRef, class: classNames.classNames([radioColor, isLayoutXL ? UNSAFE_RichSelectionItem_themes_RichSelectionItemStyles_css.styles.radioWrapper : undefined]), children: [hiddenInput, jsxRuntime.jsx("span", { "aria-hidden": "true", class: radioIconClasses, ...interactionProps, children: jsxRuntime.jsx(IconComp, { size: UNSAFE_RadioItem_themes_RadioIconContract_css.radioIconVars.iconSize, color: "currentColor" }) })] }));
};
const CheckboxControl = ({ ariaLabel, value, inputRef, inputReadonlyRef, isSelected, isLayoutXL, isDisabled, isReadonly, describedBy, labelledBy, onBlur, onFocus }) => {
    const { name } = CheckboxSetContext.useCheckboxSetContext();
    const { isTabbable } = useTabbableMode.useTabbableMode();
    const translations = useTranslationBundle.useTranslationBundle('@oracle/oraclejet-preact');
    const checkboxStyles = isLayoutXL ? UNSAFE_RichSelectionItem_themes_RichSelectionItemStyles_css.styles.checkboxWrapper : UNSAFE_RichSelectionItem_themes_RichSelectionItemStyles_css.styles.checkboxAlign;
    // JET-72374: We can't use the label content on its own for readonly because we also need to provide the state of the checkbox.
    const readonlyLabel = `${ariaLabel} ${translations.formControl_readOnly()} ${isSelected ? translations.checkbox_selected() : translations.checkbox_unselected()}`;
    // Checkbox control is visually hidden in readonly mode unless it is selected
    if (!isSelected && isReadonly) {
        return (jsxRuntime.jsx("span", { class: UNSAFE_RichSelectionItem_themes_RichSelectionItemStyles_css.styles.inputElement, children: jsxRuntime.jsx(HiddenAccessible.HiddenAccessible, { children: jsxRuntime.jsx("div", { onBlur: onBlur, onFocus: onFocus, ref: inputReadonlyRef, tabIndex: isTabbable ? 0 : -1, children: readonlyLabel }) }) }));
    }
    return (jsxRuntime.jsx("span", { class: checkboxStyles, children: jsxRuntime.jsx(StyledCheckbox.StyledCheckbox, { "aria-describedby": describedBy, isDisabled: isDisabled, isReadonly: isReadonly, isChecked: isSelected ? 'checked' : 'unchecked', name: name, value: value, onBlur: onBlur, onFocus: onFocus, ref: isReadonly ? inputReadonlyRef : inputRef, ...(isReadonly ? { 'aria-label': readonlyLabel } : { 'aria-labelledby': labelledBy }) }) }));
};

const onMouseDown = (event) => {
    // Prevent focus from going to the body when the user clicks something within
    // the card that is not focusable. This is important because we are detecting
    // focus loss on the core pack side and showing hidden messages.
    event.preventDefault();
};
/**
 * `RichSelectionItem` is a customizable input card element used in selection interfaces. It can display a label,
 * optional secondary text, and media (such as an avatar, thumbnail, or icon). It supports both radio and checkbox
 * controls, allowing users to make a single or multiple selections, respectively. The component also handles
 * different states such as disabled, readonly, and focus, and renders dynamically based on layout( 'xl', 'md', or 'sm').
 */
const RichSelectionItem = ({ avatar, iconClass, isReadonly: propsIsReadonly, mediaAccessibleLabel, secondaryText, thumbnailSrc, label, value, testId }) => {
    const { isReadonly: isFormReadonly, isDisabled: isFormDisabled } = useFormContext.useFormContext();
    const styledCardRef = hooks.useRef(null);
    const inputRef = hooks.useRef(null);
    const inputReadonlyRef = hooks.useRef(null);
    const { selection, layout, messageSeverity } = RichSelectionStartContext.useRichSelectionContext();
    const { isDisabled: isFieldDisabled, isReadonly: isFieldReadonly } = useFormFieldContext.useFormFieldContext();
    const { value: radioGroupValue, onCommit: onRadioCommit, setFocusIndex } = RadioItemContext.useRadioSetContext();
    const { value: checkboxGroupValue, onCommit: onCheckboxCommit } = CheckboxSetContext.useCheckboxSetContext();
    const { index } = RadioItemContext.useRadioItemContext();
    // Determining selection state for error
    const styledCardVariant = messageSeverity === 'error' ? 'danger' : undefined;
    const isDisabled = isFieldDisabled ?? isFormDisabled;
    const isReadonly = isFieldReadonly || isFormReadonly || propsIsReadonly;
    const isEnabled = !isDisabled && !isReadonly;
    const id = useId.useId();
    const isLayoutXL = layout === 'xl';
    const hasMedia = !!avatar || !!thumbnailSrc || !!iconClass;
    const labelId = `${id}-labelled-text`;
    const secondaryTextId = secondaryText ? `${id}-secondary-text` : undefined;
    const isSingleSelection = selection === 'single';
    const isSelected = isSingleSelection
        ? radioGroupValue === value
        : !!checkboxGroupValue?.has(value);
    const { isFocused: isFocus, focusProps: { onFocusIn, onFocusOut }, methods } = useFocusableTextField.useFocusableTextField({
        enabledElementRef: inputRef,
        readonlyElementRef: inputReadonlyRef,
        isReadonly,
        isDisabled
    });
    // Scroll into view when the RichSelectionItem is focused
    hooks.useEffect(() => {
        if (isFocus) {
            styledCardRef?.current?.scrollIntoView({ block: 'nearest' });
        }
    }, [isFocus, styledCardRef]);
    // Text content of label and secondaryText rendering
    const textContent = (jsxRuntime.jsx(RichSelectionTextContent, { secondaryTextId: secondaryTextId, labelId: labelId, isDisabled: isDisabled, isLayoutXL: isLayoutXL, label: label, secondaryText: secondaryText }));
    // Media content rendering based on available media props respective to layout
    const media = avatar
        ? { avatar, mediaAccessibleLabel, layout: layout }
        : thumbnailSrc
            ? { thumbnailSrc, mediaAccessibleLabel, layout: layout }
            : iconClass
                ? { iconClass, layout: layout }
                : undefined;
    const mediaContent = hasMedia ? (jsxRuntime.jsx(RichSelectionMediaItem, { ...media, isDisabled: isDisabled, hasSecondaryText: !!secondaryText, testId: testId ? `${testId}_media` : undefined })) : undefined;
    // Selection control rendering (radio or checkbox)
    const selectionControl = (jsxRuntime.jsx(Flex.Flex, { align: secondaryText ? 'start' : 'center', children: isSingleSelection ? (jsxRuntime.jsx(RadioControl, { ariaLabel: label, onFocus: onFocusIn, onBlur: onFocusOut, value: value, inputRef: inputRef, isFocus: isFocus, isLayoutXL: isLayoutXL, isDisabled: isDisabled, isReadonly: isReadonly, isSelected: isSelected, describedBy: secondaryTextId, labelledBy: `${id}-labelled-text` })) : (jsxRuntime.jsx(CheckboxControl, { ariaLabel: label, onFocus: onFocusIn, onBlur: onFocusOut, value: value, inputRef: inputRef, inputReadonlyRef: inputReadonlyRef, isFocus: isFocus, isDisabled: isDisabled, isLayoutXL: isLayoutXL, isReadonly: isReadonly, isSelected: isSelected, describedBy: secondaryTextId, labelledBy: `${id}-labelled-text` })) }));
    // Handler for checkbox selection changes
    const triggerCheckBoxAction = hooks.useCallback(() => {
        // if selected from a click, we need to manually set focus on the input to allow for keyboard navigation
        methods.focus();
        const newGroupValue = new Set(checkboxGroupValue);
        // If the value is already selected, remove it; otherwise, add it
        if (newGroupValue.has(value)) {
            newGroupValue.delete(value);
        }
        else {
            newGroupValue.add(value);
        }
        onCheckboxCommit?.({
            previousValue: checkboxGroupValue,
            value: newGroupValue.size > 0 ? newGroupValue : undefined
        });
    }, [checkboxGroupValue, methods, onCheckboxCommit, value]);
    // Handler for radio selection changes
    const triggerRadioAction = hooks.useCallback(() => {
        if (radioGroupValue !== value) {
            // if selected from a click, we need to manually set focus on the input to allow for keyboard navigation
            methods.focus();
            onRadioCommit?.({
                previousValue: radioGroupValue,
                value: value
            });
        }
    }, [radioGroupValue, value, methods, onRadioCommit]);
    // due to the placement of the start label in rich selection, we need to check if the first item contains media or secondary text.
    const startData = hooks.useMemo(() => ({ hasMedia, hasSecondaryText: !!secondaryText }), [hasMedia, secondaryText]);
    const richSelectionStartContext = RichSelectionStartContext.useRichSelectionStartContext();
    if (richSelectionStartContext?.isFirst) {
        richSelectionStartContext?.setStartData?.(startData);
    }
    // Event handlers for the interaction of the StyledCard
    const { pressProps } = usePress.usePress(isSingleSelection ? triggerRadioAction : triggerCheckBoxAction, {
        isDisabled: !isEnabled
    });
    // This is used to apply focus to the card while in readonly mode
    const onReadonlyClickHandler = hooks.useCallback(() => {
        // focus if radio
        inputRef?.current?.focus();
        setFocusIndex?.(index);
        // focus if checkbox
        inputReadonlyRef?.current?.focus();
    }, [index, setFocusIndex]);
    return (jsxRuntime.jsx(StyledCard.StyledCard, { cursor: isEnabled ? 'pointer' : 'arrow', display: "inlineFlex", height: "100%", isDisabled: !isEnabled, isFocusWithin: isFocus, isReadonly: isReadonly, isSelected: styledCardVariant !== 'danger' && isEnabled && isSelected, onMouseDown: onMouseDown, parentComponentVariant: "richSelectionItemCard", ref: styledCardRef, role: "listitem", styleVariant: styledCardVariant, testId: testId ? `${testId}_styled_card` : undefined, width: "100%", ...(isReadonly ? { onClick: onReadonlyClickHandler } : pressProps), children: isLayoutXL ? (jsxRuntime.jsxs(Flex.Flex, { direction: "column", height: "100%", width: "100%", children: [mediaContent, jsxRuntime.jsx("div", { class: UNSAFE_RichSelectionItem_themes_RichSelectionItemStyles_css.styles.selectionElement, children: selectionControl }), textContent] })) : (jsxRuntime.jsx("div", { className: classNames.classNames([
                UNSAFE_RichSelectionItem_themes_RichSelectionItemStyles_css.styles.densityWrapper,
                // if the item has a sm/md layout, without media or secondary text,
                // we need to set the min height of the card to match the start label
                !isLayoutXL &&
                    !hasMedia &&
                    !secondaryText &&
                    richSelectionStartContext !== null &&
                    UNSAFE_RichSelectionItem_themes_RichSelectionItemStyles_css.styles.startMinHeight
            ]), children: jsxRuntime.jsxs(Flex.Flex, { direction: "row", gap: "4x", height: "100%", width: "100%", children: [selectionControl, textContent, mediaContent] }) })) }));
};

exports.RichSelectionItem = RichSelectionItem;
