// @oracle/oraclejet-preact/cjs/UNSAFE_Switch.js
/* @oracle/oraclejet-preact: undefined */
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var jsxRuntime = require('preact/jsx-runtime');
var compat = require('preact/compat');
var hooks = require('preact/hooks');
var FormFieldContext = require('./FormFieldContext-69fe83d2.js');
var classNames = require('./classNames-c14c6ef3.js');
require('./FormContext-d1d05967.js');
var useFormContext = require('./useFormContext-a794f7b6.js');
var useFormFieldContext = require('./useFormFieldContext-4632eb73.js');
var useComponentTheme = require('./useComponentTheme-082fc8e4.js');
var UNSAFE_Label_themes_redwood_LabelTheme = require('./UNSAFE_Label/themes/redwood/LabelTheme.js');
var UNSAFE_TextField_themes_redwood_FormLayoutTheme = require('./UNSAFE_TextField/themes/redwood/FormLayoutTheme.js');
var StyledLabelValueLayout = require('./StyledLabelValueLayout-48da61b9.js');
var formControlUtils = require('./formControlUtils-31d8bc55.js');

var StyledSwitch = require('./StyledSwitch-17127a03.js');
var useTestId = require('./useTestId-4f53d561.js');
var stringUtils = require('./stringUtils-3e19c8af.js');
var useId = require('./useId-6c0eeb27.js');
var useFormFieldContextProps = require('./useFormFieldContextProps-0fa67f6c.js');
var useMessageSeverity = require('./useMessageSeverity-cfc4ae9d.js');
var Label = require('./Label-eb654b0e.js');
require('./logger-2b636482.js');
require('./TabbableModeContext-e99d527e.js');
var useTranslationBundle = require('./useTranslationBundle-12dd087d.js');


require('./UNSAFE_UserAssistance/themes/redwood/UserAssistanceVariants.css.js');
var InlineUserAssistance = require('./InlineUserAssistance-cfaa3d3e.js');
require('./Flex-3a257bc3.js');


require('./UNSAFE_MessageBanner/themes/redwood/MessageBannerVariants.css.js');
require('preact');
require('./LayerHost-3cd73603.js');


require('./hooks/UNSAFE_useTooltip/themes/redwood/TooltipContentVariants.css.js');
require('./tooltipUtils-2bc15136.js');

require('./IconButton-7503293b.js');

require('./MessageFormattingUtils-539d75b0.js');
require('./TransitionGroup-b239d98f.js');
require('./MessagesContext-4e939750.js');



require('./UNSAFE_ComponentMessage/themes/redwood/ComponentMessageVariants.css.js');
require('./Popup-de16d915.js');
require('./UNSAFE_Separator/themes/SeparatorStyles.css.js');


require('./UNSAFE_Label/themes/redwood/LabelVariants.css.js');
require('./InputGroupContext-05f2a46f.js');

require('./UNSAFE_Popup/themes/redwood/PopupVariants.css.js');
var UNSAFE_Switch_themes_redwood_SwitchTheme = require('./UNSAFE_Switch/themes/redwood/SwitchTheme.js');
require('./UNSAFE_Label/themes/LabelStyles.css.js');
require('./UNSAFE_Label/themes/redwood/LabelBaseTheme.css.js');
require('./vanilla-extract-recipes-createRuntimeFn.esm-d1301b2d.js');
require('./UNSAFE_TextField/themes/FormLayoutStyles.css.js');
require('./UNSAFE_TextField/themes/redwood/FormLayoutBaseTheme.css.js');


require('./size-1914356b.js');
require('./utils-c0efcd5e.js');
require('./themeContract.css-f3d431c6.js');
require('./colorUtils-01f3caa2.js');
require('./_curry1-e8f0d7ea.js');
require('./LayoutStyles.css-b4025421.js');

require('./PRIVATE_StyledSwitch/themes/StyledSwitchStyles.css.js');

require('./useTabbableMode-dbd254e8.js');
require('./PRIVATE_StyledSwitch/themes/redwood/StyledSwitchTheme.js');
require('./PRIVATE_StyledSwitch/themes/redwood/StyledSwitchVariants.css.js');

require('./HiddenAccessible-087eb49a.js');
require('./useActive-6770f917.js');
require('./useToggle-3ebba7d8.js');
require('./mergeProps-e3da7237.js');
require('./useTooltip-34accecc.js');
require('./useTooltipControlled-0a95856c.js');
require('./useHover-49b0430d.js');
require('./useFocus-1b288fb9.js');
require('./useTouch-4dec8729.js');
require('./useAnimation-65c0e455.js');
require('./useThemeInterpolations-a37b37d9.js');
require('./useColorScheme-f8b1447d.js');
require('./useScale-db16c2b8.js');
require('./theme-dfd6a9fe.js');
require('./Theme-f06687af.js');
require('./mergeInterpolations-6727b536.js');
require('./mergeDeepWithKey-210b024d.js');
require('./_curry3-18677bca.js');
require('./_curry2-c15d89cd.js');
require('./_isObject-28636267.js');
require('./useDensity-0b9fd9ca.js');
require('./unsafeDomAccess-c3a492f0.js');
require('./hooks/UNSAFE_useTooltip/themes/redwood/TooltipContentTheme.js');
require('./hooks/UNSAFE_useTooltip/themes/TooltipContentStyles.css.js');
require('./hooks/UNSAFE_useTooltip/themes/redwood/TooltipContentBaseTheme.css.js');

require('./clientHints-c40c5250.js');
require('./CompactLabelAssistance-e910bb56.js');
require('./Help-dbc23ccd.js');
require('./Icon-56789beb.js');
require('./UNSAFE_Icon/themes/IconStyle.css.js');
require('./useFocusWithin-eb7f956a.js');
require('./UNSAFE_UserAssistance/themes/UserAssistanceStyles.css.js');
require('./truncationUtils-2d44b04b.js');
require('./UNSAFE_TextField/themes/redwood/TextFieldTheme.js');
require('./UNSAFE_TextField/themes/TextFieldStyles.css.js');



require('./UNSAFE_TextField/themes/redwood/TextFieldBaseTheme.css.js');

require('./UNSAFE_TextField/themes/redwood/TextFieldVariants.css.js');

require('./ComponentMessageContainer-b692adf3.js');
require('./MessagesManager-cba4bc60.js');
require('./SuccessS-b6757c64.js');
require('./ErrorS-b027f7ad.js');
require('./InformationS-b0bb102b.js');
require('./WarningS-cdb0a7ad.js');
require('./PRIVATE_Message/themes/MessageStyles.css.js');
require('./soundUtils-ff3c9c0d.js');
require('./useMessagesContext-23d62331.js');
require('./Transition-942950ed.js');
require('./UNSAFE_ComponentMessage/themes/redwood/ComponentMessageTheme.js');
require('./UNSAFE_ComponentMessage/themes/ComponentMessageStyles.css.js');
require('./UNSAFE_ComponentMessage/themes/redwood/ComponentMessageBaseTheme.css.js');
require('./UNSAFE_UserAssistance/themes/redwood/UserAssistanceTheme.js');
require('./UNSAFE_UserAssistance/themes/redwood/UserAssistanceBaseTheme.css.js');
require('./InlineHelp-2add2d34.js');
require('./useInputGroupContext-4c10bd57.js');
require('./boxalignment-e287d370.js');
require('./arrayUtils-7d8dcfc3.js');

require('./vanilla-extract-sprinkles-createRuntimeSprinkles.esm-d68f3e0f.js');
require('./dimensions-6101ca1f.js');
require('./flexbox-2cae9a01.js');

require('./flexitem-fee13e26.js');



require('./index-0aa99b92.js');
require('./UNSAFE_Layer/themes/LayerHostStyles.css.js');

require('./EnvironmentProvider-5f2435e6.js');
require('./LayerManager-c535d963.js');
require('./Floating-79cbe5ad.js');
require('./useFloating-c753a7f9.js');
require('./useUser-1295b5cb.js');
require('./positionUtils-a780137e.js');
require('./refUtils-a9872e75.js');
require('./useOutsideClick-dbf3d604.js');
require('./Layer-b60deb1b.js');
require('./UNSAFE_Layer/themes/LayerStyles.css.js');

require('./useModal-0739a528.js');
require('./UNSAFE_Floating/themes/redwood/FloatingTheme.js');
require('./UNSAFE_Floating/themes/FloatingStyles.css.js');

require('./UNSAFE_Floating/themes/redwood/FloatingBaseTheme.css.js');

require('./UNSAFE_Floating/themes/redwood/FloatingVariants.css.js');
require('./vanilla-extract-dynamic.esm-2e1b7c25.js');
require('./UNSAFE_Floating/themes/FloatingContract.css.js');
require('./useUnsafeDomElementRef-f22a2f0c.js');
require('./BaseButton-d8f96bfa.js');
require('./UNSAFE_BaseButton/themes/redwood/BaseButtonTheme.js');
require('./UNSAFE_BaseButton/themes/BaseButtonStyles.css.js');

require('./UNSAFE_BaseButton/themes/redwood/BaseButtonVariants.css.js');

require('./BareButton-e8a030d7.js');
require('./useBareButton-463210c9.js');
require('./usePress-886180e4.js');
require('./usePressClick-7a34624a.js');
require('./buttonUtils-dff2ea3e.js');
require('./id-86356250.js');
require('./ButtonLabelLayout-c876923f.js');
require('./Text-056f1ec5.js');
require('./UNSAFE_Text/themes/TextStyles.css.js');

require('./useButtonLabelLayoutTheme-d09db0c9.js');
require('./UNSAFE_ButtonLabelLayout/themes/redwood/ButtonLabelLayoutTheme.js');
require('./UNSAFE_ButtonLabelLayout/themes/ButtonLabelLayoutStyles.css.js');

require('./UNSAFE_ButtonLabelLayout/themes/redwood/ButtonLabelLayoutBaseTheme.css.js');

require('./UNSAFE_ButtonLabelLayout/themes/redwood/ButtonLabelLayoutVariants.css.js');

require('./getLocale-8e17dc46.js');
require('./tabbableUtils-22a50472.js');
require('./head-68d0992f.js');
require('./_arity-c228159c.js');
require('./_isArray-73160ad5.js');
require('./_isString-f4443c9e.js');
require('./FocusTrap-4873a3f2.js');
require('./FocusTracker-e2ea43c3.js');
require('./PRIVATE_FocusTracker/themes/FocusTrackerStyles.css.js');

require('./Modal-b751b643.js');
require('./UNSAFE_Modal/themes/ModalStyles.css.js');

require('./popupUtils-1d123cbe.js');
require('./WindowOverlay-85d999fa.js');
require('./UNSAFE_WindowOverlay/themes/WindowOverlayStyles.css.js');

require('./UNSAFE_WindowOverlay/themes/WindowOverlayContract.css.js');
require('./UNSAFE_WindowOverlay/themes/redwood/WindowOverlayTheme.js');
require('./UNSAFE_WindowOverlay/themes/redwood/WindowOverlayBaseTheme.css.js');

require('./UNSAFE_WindowOverlay/themes/redwood/WindowOverlayVariants.css.js');
require('./usePopupAnimation-f3b34413.js');
require('./useAnimationStatus-8075dc19.js');
require('./popupAnimationUtils-dcace427.js');
require('./LiveRegion-62ce9ac7.js');

require('./UNSAFE_Popup/themes/PopupContract.css.js');
require('./UNSAFE_Popup/themes/redwood/PopupTheme.js');
require('./UNSAFE_Popup/themes/redwood/PopupBaseTheme.css.js');

require('./colors-f3ce979a.js');

require('./UNSAFE_Switch/themes/redwood/SwitchBaseTheme.css.js');


var switchLabelInnerStyles = {startSwitch:'SwitchLabelValueLayoutStyles_switchLabelInnerStyles_startSwitch__a7cetq0',insideSwitch:'SwitchLabelValueLayoutStyles_switchLabelInnerStyles_insideSwitch__a7cetq1'};

const SwitchLabelValueLayout = ({ columnSpan = 1, label, labelEdge = 'start', children, labelStartWidth = '33%' }) => {
    const { baseTheme: labelTheme } = useComponentTheme.useComponentTheme(UNSAFE_Label_themes_redwood_LabelTheme.LabelRedwoodTheme);
    const { baseTheme: formLayoutTheme } = useComponentTheme.useComponentTheme(UNSAFE_TextField_themes_redwood_FormLayoutTheme.FormLayoutRedwoodTheme);
    const isStart = labelEdge === 'start';
    const isTopOrInside = labelEdge === 'top' || labelEdge === 'inside';
    // labelStartWidth needs to take columnSpan into account
    const width = formControlUtils.computeStartLabelWidth(labelStartWidth, columnSpan);
    const valueWidth = `calc(100% - ${width})`;
    const { isFormLayout, isReadonly: isFormReadonly } = useFormContext.useFormContext();
    const { isReadonly } = useFormFieldContext.useFormFieldContext();
    const isPureReadonly = (isFormLayout && isFormReadonly) || (!isFormLayout && isReadonly);
    const labelStyleClasses = classNames.classNames([
        formLayoutTheme,
        StyledLabelValueLayout.labelSlotStyles.base,
        isStart && StyledLabelValueLayout.labelSlotStyles.start,
        isTopOrInside && StyledLabelValueLayout.labelSlotStyles.top,
        isTopOrInside && isPureReadonly && StyledLabelValueLayout.labelSlotStyles.topPureReadonly
    ]);
    const labelInnerClasses = classNames.classNames([
        labelTheme,
        formLayoutTheme,
        StyledLabelValueLayout.labelInnerStyles.base,
        isStart && StyledLabelValueLayout.labelInnerStyles.start,
        isStart && switchLabelInnerStyles.startSwitch
    ]);
    const labelStyles = isStart ? { flexBasis: width, width, maxWidth: width } : {};
    const valueStyles = isStart
        ? { flexBasis: valueWidth, width: valueWidth, maxWidth: valueWidth }
        : {};
    // Outer container and inner wrapper match CheckboxRadioLabelValueLayout's order exactly:
    return (jsxRuntime.jsx(StyledLabelValueLayout.StyledLabelValueLayout, { flexAlign: isReadonly && isStart ? 'stretch' : 'start', labelStyleClasses: labelStyleClasses, labelStyles: labelStyles, labelInnerClasses: labelInnerClasses, label: label, valueStyles: valueStyles, children: children }));
};

const SwitchField = ({ columnSpan, inlineUserAssistance, label, labelEdge, labelStartWidth, mainContent }) => {
    // Use SwitchLabelValueLayout for labelEdge 'start', 'top', 'inside'
    if (labelEdge && labelEdge !== 'none') {
        return (jsxRuntime.jsxs(SwitchLabelValueLayout, { columnSpan: columnSpan, label: label, labelEdge: labelEdge, labelStartWidth: labelStartWidth, children: [mainContent, inlineUserAssistance] }));
    }
    // Otherwise, for 'none', render children/assistance inline
    return (jsxRuntime.jsxs("div", { style: { display: 'flex', flexDirection: 'column' }, children: [mainContent, inlineUserAssistance] }));
};

/**
 * A custom hook to determine the props for a component that renders
 * a Switch. This is similar to useTextField, but unlike TextField
 * Switch does not render truncated labels nor does it render required.
 */
function useSwitch(props) {
    const { ariaDescribedBy, helpSourceLink, helpSourceText, isDisabled, isReadonly, label, labelEdge, messages, userAssistanceDensity, value } = props;
    const baseId = useId.useId();
    const labelId = labelEdge !== 'none' ? `${baseId}-label` : undefined;
    const uaId = !isDisabled ? `${baseId}-ua` : undefined;
    const hasValue = value ?? false;
    // Create the form field context (no truncation/required logic needed)
    const formFieldContext = useFormFieldContextProps.useFormFieldContextProps({
        hasValue,
        isDisabled,
        isReadonly
    });
    // Determine the highest severity from the messages
    const messageSeverity = useMessageSeverity.useMessageSeverity(messages);
    // Determine the label variant
    const labelVariant = labelEdge !== 'none' ? labelEdge : undefined;
    const ariaLabel = label && label.length > 0 && labelEdge === 'none' ? label : undefined;
    const ariaLabelledBy = ariaLabel ? undefined : labelId;
    // According to UX, Switch's compact user assistance density behaves like efficient.
    const resolvedUserAssistanceDensity = userAssistanceDensity === 'compact' ? 'efficient' : userAssistanceDensity;
    return {
        baseId,
        formFieldContext,
        styledSwitchProps: {
            'aria-describedby': stringUtils.l([uaId, ariaDescribedBy]),
            'aria-invalid': messageSeverity === 'error'
                ? 'true'
                : undefined,
            ...(ariaLabelledBy && { 'aria-labelledby': ariaLabelledBy }),
            ...(ariaLabel && { 'aria-label': ariaLabel })
        },
        labelProps: {
            helpSourceLink,
            helpSourceText,
            id: labelId,
            userAssistanceDensity: resolvedUserAssistanceDensity,
            variant: labelVariant
        },
        userAssistanceProps: {
            id: uaId
        }
    };
}

/**
 * The Switch component toggles between two mutually exclusive states, such as on and off.
 *
 */
const Switch = compat.forwardRef(({ 'aria-describedby': ariaDescribedBy, assistiveText, columnSpan, isDisabled: propIsDisabled, isReadonly: propIsReadonly, label, labelEdge: propLabelEdge, labelStartWidth: propLabelStartWidth, messages, onCommit, helpSourceLink, helpSourceText, testId, userAssistanceDensity: propUserAssistanceDensity, value: propValue }, ref) => {
    const { isDisabled: isFormDisabled, isReadonly: isFormReadonly, labelEdge: formLabelEdge, labelStartWidth: formLabelStartWidth, userAssistanceDensity: formUserAssistanceDensity } = useFormContext.useFormContext();
    // default to FormContext values if component properties are not specified
    const isDisabled = propIsDisabled ?? isFormDisabled;
    const isReadonly = propIsReadonly ?? isFormReadonly;
    const labelEdge = propLabelEdge ?? formLabelEdge;
    const labelStartWidth = propLabelStartWidth ?? formLabelStartWidth;
    const userAssistanceDensity = propUserAssistanceDensity ?? formUserAssistanceDensity;
    const { baseId, formFieldContext, styledSwitchProps, labelProps, userAssistanceProps } = useSwitch({
        ariaDescribedBy,
        helpSourceLink,
        helpSourceText,
        isDisabled,
        isReadonly,
        label,
        labelEdge,
        messages,
        userAssistanceDensity,
        value: propValue
    });
    // This ref is used for both imperative focus and enabling SwitchField to "click label to focus"
    const switchElementRef = hooks.useRef(null);
    // setRefs is passed as the ref prop to <StyledSwitch> below.
    // It synchronizes both:
    // - switchElementRef (internal, for label click/focus logic)
    // - the parent/test ref (for imperative .focus()/.blur() calls)
    // so that both internal and external focus logic work with the same DOM node.
    function setRefs(node) {
        switchElementRef.current = node;
        if (typeof ref === 'function') {
            ref(node);
        }
        else if (ref) {
            ref.current = node;
        }
    }
    // Label click focus logic
    const onClick = hooks.useCallback(() => {
        if (switchElementRef) {
            switchElementRef.current?.focus();
        }
    }, [switchElementRef]);
    const labelComp = labelEdge !== 'none' ? (jsxRuntime.jsx(Label.Label, { ...labelProps, ...(!isDisabled && !isReadonly ? { onClick } : {}), parentComponentVariant: "switch", children: label })) : undefined;
    const fieldLabelProps = {
        label: labelEdge !== 'none' ? labelComp : undefined,
        labelEdge: labelEdge !== 'none' ? labelEdge : undefined,
        labelStartWidth: labelEdge !== 'none' ? labelStartWidth : undefined
    };
    // According to UX, Switch's compact user assistance density behaves like efficient.
    const resolvedUserAssistanceDensity = userAssistanceDensity === 'compact' ? 'efficient' : userAssistanceDensity;
    const isInlineDensity = resolvedUserAssistanceDensity === 'efficient' || resolvedUserAssistanceDensity === 'reflow';
    const inlineUserAssistance = isInlineDensity ? (isDisabled ? (
    // save space for user assistance if density is 'efficient', even though we don't
    // render user assistance for disabled fields.
    // showInlineAssistanceOn is 'always' because for Switch we always show the user assistance.
    resolvedUserAssistanceDensity !== 'efficient' ? undefined : (jsxRuntime.jsx(InlineUserAssistance.InlineUserAssistance, { showInlineAssistanceOn: "always", userAssistanceDensity: resolvedUserAssistanceDensity, ...userAssistanceProps }))) : isReadonly ? (jsxRuntime.jsx(InlineUserAssistance.InlineUserAssistance, { fieldLabel: label, showInlineAssistanceOn: "always", messages: messages, userAssistanceDensity: resolvedUserAssistanceDensity, ...userAssistanceProps })) : (jsxRuntime.jsx(InlineUserAssistance.InlineUserAssistance, { assistiveText: assistiveText, fieldLabel: label, helpSourceLink: helpSourceLink, helpSourceText: helpSourceText, showInlineAssistanceOn: "always", messages: messages, userAssistanceDensity: resolvedUserAssistanceDensity, ...userAssistanceProps }))) : undefined;
    // toggle the value when we get the StyledSwitch's onChange handler is activated
    const onChangeHandler = hooks.useCallback(() => {
        const previousValue = !!propValue;
        const value = !previousValue;
        onCommit?.({
            previousValue,
            value
        });
    }, [onCommit, propValue]);
    const translations = useTranslationBundle.useTranslationBundle('@oracle/oraclejet-preact');
    const switchOn = translations.switch_on();
    const switchOff = translations.switch_off();
    const testIdProps = useTestId.useTestId(testId);
    const styledSwitchTestId = testId && `${testId}-styled-switch`;
    const styledSwitchComp = (jsxRuntime.jsx(StyledSwitch.StyledSwitch, { ...styledSwitchProps, baseId: baseId, isChecked: propValue, isDisabled: isDisabled, isReadonly: isReadonly, readonlyOnText: switchOn, readonlyOffText: switchOff, ref: setRefs, testId: styledSwitchTestId, onChange: onChangeHandler }));
    return (jsxRuntime.jsx("div", { ...testIdProps, className: UNSAFE_Switch_themes_redwood_SwitchTheme.SwitchRedwoodTheme.baseTheme, children: jsxRuntime.jsx(FormFieldContext.FormFieldContext.Provider, { value: formFieldContext, children: jsxRuntime.jsx(SwitchField, { columnSpan: columnSpan, inlineUserAssistance: inlineUserAssistance, mainContent: styledSwitchComp, ...fieldLabelProps }) }) }));
});

exports.Switch = Switch;
