// @oracle/oraclejet-preact/cjs/UNSAFE_BarChart.js
/* @oracle/oraclejet-preact: undefined */
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var jsxRuntime = require('preact/jsx-runtime');
var utils = require('./utils-ee4fc356.js');
var useUser = require('./useUser-1295b5cb.js');
var BarGroup = require('./BarGroup-62bf29eb.js');
var dataStyleUtils = require('./dataStyleUtils-4b0c5ce2.js');
var compat = require('preact/compat');
var Chart = require('./Chart-13a198f3.js');
require('./themeContract.css-f3d431c6.js');
require('./utils-c0efcd5e.js');
require('./colorUtils-01f3caa2.js');
require('./_curry1-e8f0d7ea.js');
require('./getLocale-8e17dc46.js');
require('./stringUtils-3e19c8af.js');
require('./labelUtils-45add12c.js');
require('./textUtils-e779c845.js');
require('./accUtils-4a359187.js');
require('./clientHints-c40c5250.js');
require('./utils-ca441171.js');
require('./datatipUtils-d1ca81a1.js');
require('./clientUtils-a4619fcd.js');
require('./dimensionUtils-3919753d.js');
require('./scale-9e43a5f6.js');
require('preact/hooks');
require('./LayerHost-3cd73603.js');
require('preact');
require('./index-0aa99b92.js');
require('./UNSAFE_Layer/themes/LayerHostStyles.css.js');

require('./PRIVATE_Chart/themes/ChartStyles.css.js');

require('./classNames-c14c6ef3.js');
require('./colorUtils-b2204778.js');
require('./TrackResizeContainer-a8631f0d.js');
require('./useSize-8115eae0.js');
require('./useResizeObserver-a9e97180.js');
require('./useAddBusyState-d19ae1fa.js');
require('./BusyStateContext-86f40d3c.js');
require('./useBusyStateContext-362eee2a.js');
require('./unsafeDomAccess-c3a492f0.js');
require('./size-1914356b.js');
require('./UNSAFE_visErrors.js');
require('./useTextDimensions-c8b41697.js');
require('./useCssVars-8e83d496.js');
require('./useTheme-2a034a21.js');
require('./useColorScheme-f8b1447d.js');
require('./useScale-db16c2b8.js');
require('./ReferenceLineArea-7bc34276.js');
require('./PRIVATE_PlotArea/themes/PlotAreaStyles.css.js');

require('./hooks/PRIVATE_useVisDnd/themes/VisDndStyles.css.js');

require('./lineUtils-222ae5af.js');
require('./useId-6c0eeb27.js');
require('./Axis-a3a56403.js');
require('./PRIVATE_Axis/themes/AxisStyles.css.js');

require('./MultilineText-837e9afd.js');
require('./PRIVATE_Axis/themes/redwood/AxisTheme.js');
require('./PRIVATE_Axis/themes/redwood/AxisBaseTheme.css.js');

require('./PRIVATE_Axis/themes/redwood/AxisVariants.css.js');
require('./vanilla-extract-recipes-createRuntimeFn.esm-d1301b2d.js');
require('./useComponentTheme-082fc8e4.js');
require('./logger-2b636482.js');
require('./useChartViewport-25f7d6db.js');
require('./eventsUtils-b87ea91f.js');
require('./usePinch-741e9343.js');
require('./usePointerGesture-0b9cc1b7.js');
require('./mergeProps-e3da7237.js');
require('./useChartNav-c69c817e.js');
require('./layoutUtils-af53a104.js');
require('./useVisHover-ad36112d.js');
require('./useSelectDrill-0b4b47ec.js');
require('./useChartDataCursor-67f7cc75.js');
require('./hooks/PRIVATE_useChartDataCursor/themes/DataCursorStyles.css.js');

require('./Marker-b40392cc.js');

require('./useChartDatatip-1bceba5d.js');
require('./useDatatip-b7b4f13f.js');
require('./Layer-b60deb1b.js');
require('./UNSAFE_Layer/themes/LayerStyles.css.js');

require('./useThemeInterpolations-a37b37d9.js');
require('./theme-dfd6a9fe.js');
require('./Theme-f06687af.js');
require('./mergeInterpolations-6727b536.js');
require('./mergeDeepWithKey-210b024d.js');
require('./_curry3-18677bca.js');
require('./_curry2-c15d89cd.js');
require('./_isObject-28636267.js');
require('./useDensity-0b9fd9ca.js');
require('./Floating-79cbe5ad.js');
require('./useFloating-c753a7f9.js');
require('./positionUtils-a780137e.js');
require('./refUtils-a9872e75.js');
require('./useOutsideClick-dbf3d604.js');
require('./arrayUtils-7d8dcfc3.js');
require('./useModal-0739a528.js');
require('./UNSAFE_Floating/themes/redwood/FloatingTheme.js');
require('./UNSAFE_Floating/themes/FloatingStyles.css.js');

require('./UNSAFE_Floating/themes/redwood/FloatingBaseTheme.css.js');

require('./UNSAFE_Floating/themes/redwood/FloatingVariants.css.js');
require('./vanilla-extract-dynamic.esm-2e1b7c25.js');
require('./UNSAFE_Floating/themes/FloatingContract.css.js');
require('./useUnsafeDomElementRef-f22a2f0c.js');
require('./hooks/PRIVATE_useDatatip/themes/useDatatip.css.js');

require('./VisTabularDatatip-abfef042.js');
require('./PRIVATE_VisTabularDatatip/themes/VisTabularDatatipStyles.css.js');

require('./useChartDrill-c052942f.js');
require('./useVisDrill-f76c2e9c.js');
require('./useDoubleTap-d3937fb6.js');
require('./clientUtils-ef5e20ce.js');
require('./useSelection-da64e28e.js');
require('./BaseButton-d8f96bfa.js');
require('./useHover-49b0430d.js');
require('./useToggle-3ebba7d8.js');
require('./useActive-6770f917.js');
require('./dimensions-6101ca1f.js');
require('./UNSAFE_BaseButton/themes/redwood/BaseButtonTheme.js');
require('./UNSAFE_BaseButton/themes/BaseButtonStyles.css.js');

require('./UNSAFE_BaseButton/themes/redwood/BaseButtonVariants.css.js');

require('./BareButton-e8a030d7.js');
require('./useBareButton-463210c9.js');
require('./usePress-886180e4.js');
require('./usePressClick-7a34624a.js');
require('./TabbableModeContext-e99d527e.js');
require('./useTabbableMode-dbd254e8.js');
require('./buttonUtils-dff2ea3e.js');
require('./id-86356250.js');
require('./useTestId-4f53d561.js');
require('./Text-056f1ec5.js');
require('./UNSAFE_Text/themes/TextStyles.css.js');

require('./useTranslationBundle-12dd087d.js');
require('./hooks/PRIVATE_useVisSelection/themes/ClearSelection.css.js');

require('./cssUtils-60d6bb40.js');
require('./useChartContextMenu-af989b4f.js');
require('./useContextMenu-2500df06.js');
require('./SelectMenuGroupContext-a349d259.js');
require('./Sheet-1d6e8cf8.js');
require('./Modal-b751b643.js');
require('./UNSAFE_Modal/themes/ModalStyles.css.js');

require('./tabbableUtils-22a50472.js');
require('./head-68d0992f.js');
require('./_arity-c228159c.js');
require('./_isArray-73160ad5.js');
require('./_isString-f4443c9e.js');
require('./popupUtils-1d123cbe.js');
require('./WindowOverlay-85d999fa.js');
require('./UNSAFE_WindowOverlay/themes/WindowOverlayStyles.css.js');

require('./UNSAFE_WindowOverlay/themes/WindowOverlayContract.css.js');
require('./UNSAFE_WindowOverlay/themes/redwood/WindowOverlayTheme.js');
require('./UNSAFE_WindowOverlay/themes/redwood/WindowOverlayBaseTheme.css.js');

require('./UNSAFE_WindowOverlay/themes/redwood/WindowOverlayVariants.css.js');
require('./UNSAFE_Sheet/themes/SheetStyles.css.js');

require('./useAnimationStatus-8075dc19.js');
require('./useAnimation-65c0e455.js');
require('./animationUtils-54852d03.js');
require('./Dropdown-1ba432a8.js');
require('./keyboardUtils-b42fe7d5.js');
require('./usePopupAnimation-f3b34413.js');
require('./popupAnimationUtils-dcace427.js');
require('./CollectionInteractionContext-f84ff7a1.js');
require('./UNSAFE_Dropdown/themes/dropdownStyles.css.js');

require('./useOutsideMousedown-70f36cca.js');
require('./useEffectEvent-832ffba5.js');
require('./UNSAFE_Menu/themes/MenuStyles.css.js');

require('./UNSAFE_Menu/themes/DropdownMenuStyles.css.js');

require('./Flex-3a257bc3.js');
require('./boxalignment-e287d370.js');

require('./vanilla-extract-sprinkles-createRuntimeSprinkles.esm-d68f3e0f.js');
require('./flexbox-2cae9a01.js');

require('./flexitem-fee13e26.js');


require('./Skeleton-8e35c7d1.js');
require('./borders-4b8488cb.js');
require('./UNSAFE_Skeleton/themes/redwood/SkeletonTheme.js');
require('./UNSAFE_Skeleton/themes/SkeletonStyles.css.js');

require('./UNSAFE_Skeleton/themes/redwood/SkeletonBaseTheme.css.js');

require('./UNSAFE_Skeleton/themes/redwood/SkeletonVariants.css.js');
require('./UNSAFE_Menu/themes/MenuSkeletonStyles.css.js');

require('./LiveRegion-62ce9ac7.js');

require('./UNSAFE_Menu/themes/redwood/MenuItemTheme.js');
require('./UNSAFE_Menu/themes/MenuItemStyles.css.js');

require('./UNSAFE_Menu/themes/redwood/MenuItemBaseTheme.css.js');

require('./UNSAFE_Menu/themes/redwood/MenuItemVariants.css.js');

require('./useInteractionStyle-c203a8a0.js');


require('./hooks/UNSAFE_useTooltip/themes/redwood/TooltipContentVariants.css.js');

require('./tooltipUtils-2bc15136.js');
require('./EnvironmentProvider-5f2435e6.js');
require('./LayerManager-c535d963.js');

require('./UNSAFE_Separator/themes/SeparatorStyles.css.js');



require('./UNSAFE_Menu/themes/redwood/MenuSeparatorVariants.css.js');
require('./useContextMenuGesture-530ef8dd.js');
require('./textSelectionUtils-57082239.js');
require('./useChartDnd-edaa3619.js');
require('./utils-da07b35e.js');
require('./useDnd-a7d9b60c.js');
require('./useTooltip-34accecc.js');
require('./useTooltipControlled-0a95856c.js');
require('./useFocus-1b288fb9.js');
require('./useTouch-4dec8729.js');
require('./hooks/UNSAFE_useTooltip/themes/redwood/TooltipContentTheme.js');
require('./hooks/UNSAFE_useTooltip/themes/TooltipContentStyles.css.js');
require('./hooks/UNSAFE_useTooltip/themes/redwood/TooltipContentBaseTheme.css.js');
require('./useChartMarquee-ede5cae4.js');
require('./useVisDragModeControls-2443a4e0.js');
require('./Button-4d923bed.js');
require('./Icon-56789beb.js');
require('./UNSAFE_Icon/themes/IconStyle.css.js');
require('./Move-b1978994.js');
require('./hooks/PRIVATE_useVisDragModeControls/themes/useVisDragModeControls.css.js');

require('./ZoomAlt-cca2384d.js');
require('./ButtonSetSingle-bfe91a67.js');
require('./useButtonSetContext-a2f31c8e.js');
require('./ButtonSet-ed013090.js');
require('./UNSAFE_ButtonSet/themes/ButtonSetStyles.css.js');

require('./ButtonSetPositionContext-6a358667.js');
require('./toggleButtonUtils-245111a7.js');
require('./useRovingTabIndexContainer-0ce4681d.js');
require('./ButtonSetItem-4322c957.js');
require('./ButtonSetButton-569e0efb.js');
require('./ToggleButtonLabel-c8f36317.js');
require('./ButtonLabelLayout-c876923f.js');
require('./useButtonLabelLayoutTheme-d09db0c9.js');
require('./UNSAFE_ButtonLabelLayout/themes/redwood/ButtonLabelLayoutTheme.js');
require('./UNSAFE_ButtonLabelLayout/themes/ButtonLabelLayoutStyles.css.js');

require('./UNSAFE_ButtonLabelLayout/themes/redwood/ButtonLabelLayoutBaseTheme.css.js');

require('./UNSAFE_ButtonLabelLayout/themes/redwood/ButtonLabelLayoutVariants.css.js');

require('./HiddenAccessible-087eb49a.js');

require('./ButtonSetIconButton-3778ac7b.js');
require('./useButtonSetPositionContext-409e4cd6.js');
require('./IconToggleButton-d2147159.js');
require('./useToggleAction-e4f15550.js');
require('./hooks/PRIVATE_useChartMarquee/themes/useChartMarquee.css.js');

require('./useVisTouchResponse-376e11d5.js');
require('./useOverviewContext-1732a3f1.js');
require('./OverviewSlidingWindow-0f5f00a3.js');
require('./PRIVATE_VisOverview/themes/VisOverview.css.js');

require('./DragH-ac9f879f.js');
require('./ScrollBar-76e58249.js');
require('./PRIVATE_ChartScrollBar/themes/ScrollBarStyles.css.js');


/**
 * @license
 * Copyright (c) %FIRST_YEAR% %CURRENT_YEAR%, Oracle and/or its affiliates.
 * Licensed under The Universal Permissive License (UPL), Version 1.0
 * as shown at https://oss.oracle.com/licenses/upl/
 * @ignore
 */
function createStackedGroupNavigator(getDataItem, series, numSeries, stackGroups, getStackGroup, isItemNavigable) {
    const hasMultipleStacks = Array.isArray(stackGroups) && stackGroups.length > 1;
    const seriesToStackIndex = series
        ? series.map((s) => {
            const stackName = getStackGroup(s) ?? 'default';
            const idx = stackGroups.indexOf(stackName);
            return idx >= 0 ? idx : 0;
        })
        : new Array(numSeries).fill(0);
    const stackToSeries = stackGroups.map(() => []);
    for (let i = 0; i < numSeries; i++) {
        const stIdx = seriesToStackIndex[i] ?? 0;
        if (stackToSeries[stIdx]) {
            stackToSeries[stIdx].push(i);
        }
    }
    const itemValue = (seriesIndex, groupIndex) => {
        const it = getDataItem(seriesIndex, groupIndex);
        return it?.value;
    };
    const valueSign = (value) => {
        if (value == null || value === 0)
            return 0;
        return value > 0 ? 1 : -1;
    };
    /**
     * Compute the stacked extreme coordinate for the given series at the given group:
     * - Positive values: cumulative positive sum up to and including this series within its stack (top edge).
     * - Negative values: cumulative negative sum up to and including this series within its stack (bottom edge).
     * - Zero/undefined: 0 (baseline).
     */
    const stackedEdge = (seriesIndex, groupIndex) => {
        const v = itemValue(seriesIndex, groupIndex);
        const sgn = valueSign(v);
        const stIdx = seriesToStackIndex[seriesIndex];
        const seriesInStack = stackToSeries[stIdx] || [];
        let sum = 0;
        for (let i = 0; i < seriesInStack.length; i++) {
            const s = seriesInStack[i];
            const val = itemValue(s, groupIndex);
            if (sgn > 0) {
                // Accumulate only positive contributions
                if ((val ?? 0) > 0)
                    sum += val;
            }
            else {
                // Accumulate only negative contributions
                if ((val ?? 0) < 0)
                    sum += val;
            }
            if (s === seriesIndex)
                break;
        }
        return sum;
    };
    /**
     * Returns all series indices ordered bottom->top by their stacked extreme coordinate
     * for a given group. Ties are broken by series index to provide a stable order.
     */
    const orderedSeriesByStackedEdge = (groupIndex) => Array.from({ length: numSeries }, (_, s) => ({ s, pos: stackedEdge(s, groupIndex) }))
        .sort((a, b) => (a.pos === b.pos ? a.s - b.s : a.pos - b.pos))
        .map((e) => e.s);
    /**
     * In-group navigation across all stacks:
     * - "next": move toward the top (larger stackedEdge)
     * - "prev": move toward the bottom (smaller stackedEdge)
     */
    const moveWithinStack = (info, direction) => {
        const group = info.groupIndex;
        const ordered = orderedSeriesByStackedEdge(group);
        const currIdx = ordered.indexOf(info.seriesIndex);
        if (currIdx === -1)
            return;
        const step = direction === 'next' ? 1 : -1;
        for (let k = currIdx + step; k >= 0 && k < ordered.length; k += step) {
            const nextSeries = ordered[k];
            if (isItemNavigable(nextSeries, group)) {
                return { seriesIndex: nextSeries, groupIndex: group };
            }
        }
        return;
    };
    return {
        hasMultipleStacks,
        moveWithinStack
    };
}

/**
 * @license
 * Copyright (c) %FIRST_YEAR% %CURRENT_YEAR%, Oracle and/or its affiliates.
 * Licensed under The Universal Permissive License (UPL), Version 1.0
 * as shown at https://oss.oracle.com/licenses/upl/
 * @ignore
 */
/**
 * Get next item in unstacked bar chart.
 * @param info
 * @param viewportEndIndex
 * @param numSeries
 * @param isNavigable
 * @returns
 */
function getNextItem(info, viewportEndIndex, numSeries, isNavigable) {
    const nextItemInGroup = dataStyleUtils.getNextItemInGroup(info, numSeries, isNavigable);
    if (nextItemInGroup)
        return nextItemInGroup;
    if (info.groupIndex < viewportEndIndex) {
        let nextItem = { seriesIndex: 0, groupIndex: info.groupIndex + 1 };
        if (isNavigable(nextItem.seriesIndex, nextItem.groupIndex))
            return nextItem;
        nextItem = getNextItem(nextItem, viewportEndIndex, numSeries, isNavigable);
        if (nextItem)
            return nextItem;
    }
    return;
}
/**
 * Get prev item in unstacked bar chart.
 * @param info
 * @param viewportStartIndex
 * @param numSeries
 * @param isNavigable
 * @returns
 */
function getPreviousItem(info, viewportStartIndex, numSeries, isNavigable) {
    const prevItemInGroup = dataStyleUtils.getPrevItemInGroup(info, isNavigable);
    if (prevItemInGroup)
        return prevItemInGroup;
    if (info.groupIndex > viewportStartIndex) {
        let prevItem = { groupIndex: info.groupIndex - 1, seriesIndex: numSeries - 1 };
        if (isNavigable(prevItem.seriesIndex, prevItem.groupIndex))
            return prevItem;
        prevItem = getPreviousItem(prevItem, viewportStartIndex, numSeries, isNavigable);
        if (prevItem)
            return prevItem;
    }
    return info;
}
function getBarChartNavUtil(getDataItem, series, numSeries, viewportStartIndex, viewportEndIndex, isStacked, isRtl, isHoriz, stackGroups, getStackGroup) {
    const isItemNavigable = (seriesIndex, groupIndex) => {
        const barItem = getDataItem(seriesIndex, groupIndex);
        return Boolean(barItem);
    };
    const getNextChartItem = (itemInfo, arrow) => {
        if (!isStacked &&
            (((arrow === 'ArrowDown' || arrow === 'ArrowUp') && !isHoriz) ||
                ((arrow === 'ArrowLeft' || arrow === 'ArrowRight') && isHoriz))) {
            return itemInfo;
        }
        let newInfo;
        // TODO: this is just to keep typescript happy.
        // figure out consistent typing while doing type cleanup.
        const info = {
            groupIndex: itemInfo.groupIndex,
            seriesIndex: itemInfo.seriesIndex
        };
        if (info.groupIndex < viewportStartIndex) {
            info.groupIndex = viewportStartIndex;
            return info;
        }
        else if (info.groupIndex > viewportEndIndex) {
            info.groupIndex = viewportEndIndex;
            return info;
        }
        if (isStacked) {
            // Stacked-group navigation helpers moved to stackedGroupNavUtils
            const { moveWithinStack } = createStackedGroupNavigator(getDataItem, series, numSeries, stackGroups, getStackGroup, isItemNavigable);
            if ((arrow === 'ArrowLeft' && !isHoriz) || (arrow === 'ArrowUp' && isHoriz)) {
                newInfo =
                    isRtl && !isHoriz
                        ? dataStyleUtils.getNextItemInSeries(info, viewportEndIndex, isItemNavigable)
                        : dataStyleUtils.getPrevItemInSeries(info, viewportStartIndex, isItemNavigable);
            }
            else if ((arrow === 'ArrowRight' && !isHoriz) || (arrow === 'ArrowDown' && isHoriz)) {
                newInfo =
                    isRtl && !isHoriz
                        ? dataStyleUtils.getPrevItemInSeries(info, viewportStartIndex, isItemNavigable)
                        : dataStyleUtils.getNextItemInSeries(info, viewportEndIndex, isItemNavigable);
            }
            else if ((arrow === 'ArrowDown' && !isHoriz) || (arrow === 'ArrowLeft' && isHoriz)) {
                const dir = isRtl && isHoriz ? 'next' : 'prev';
                newInfo = moveWithinStack(info, dir);
            }
            else if ((arrow === 'ArrowUp' && !isHoriz) || (arrow === 'ArrowRight' && isHoriz)) {
                const dir = isRtl && isHoriz ? 'prev' : 'next';
                newInfo = moveWithinStack(info, dir);
            }
        }
        else {
            if (!isHoriz) {
                if (arrow === 'ArrowRight') {
                    newInfo = isRtl
                        ? getPreviousItem(info, viewportStartIndex, numSeries, isItemNavigable)
                        : getNextItem(info, viewportEndIndex, numSeries, isItemNavigable);
                }
                else if (arrow === 'ArrowLeft') {
                    newInfo = isRtl
                        ? getNextItem(info, viewportEndIndex, numSeries, isItemNavigable)
                        : getPreviousItem(info, viewportStartIndex, numSeries, isItemNavigable);
                }
            }
            else {
                if (arrow === 'ArrowDown') {
                    newInfo = getNextItem(info, viewportEndIndex, numSeries, isItemNavigable);
                }
                else if (arrow === 'ArrowUp') {
                    newInfo = getPreviousItem(info, viewportStartIndex, numSeries, isItemNavigable);
                }
            }
        }
        return newInfo || info;
    };
    return getNextChartItem;
}

function OverviewBarChart(props) {
    return (jsxRuntime.jsx(BarChart, { datatip: () => ({ content: '' }), zoomAndScroll: "off", plotArea: {
            yMajorTick: {
                isRendered: false
            },
            xMajorTick: {}
        }, xAxis: {
            tickLabel: {
                autoRotate: false,
                // @ts-ignore
                isRendered: props.xAxis?.timeAxisType ? true : false
            }
        }, series: props.series, groups: props.groups, getDataItem: props.getDataItem, yAxis: {
            tickLabel: {
                isRendered: false
            }
        }, overview: "off" }));
}

/**
 * A Bar chart displays information graphically using rectangular bars with heights or lengths proportional to the values that they represent,
 * making relationships among the data easier to understand.
 */
const BarChart = compat.forwardRef(({ width, height, selectedIds = [], hiddenIds = [], highlightedIds, hideAndShowBehavior = 'none', orientation = 'vertical', yAxis, xAxis, y2Axis, series, groups, stacking = 'off', getDataItem, selectionMode = 'none', drilling = 'off', dataItemGaps = 0.5, ...props }, ref) => {
    const { direction } = useUser.useUser();
    const isRtl = direction === 'rtl';
    const hiddenSet = new Set(hiddenIds);
    const isHoriz = orientation === 'horizontal';
    const timeAxisType = xAxis?.timeAxisType;
    const selectedSet = new Set(selectedIds);
    const highlightedSet = new Set(highlightedIds);
    const isStacked = stacking != 'off';
    const stackLabelStyle = typeof stacking === 'string' ? undefined : stacking.stackLabelStyle;
    const getStackGroup = typeof stacking != 'string' ? stacking.getStackGroup : () => 'default';
    const getStackLabel = typeof stacking === 'object' && stacking.getStackLabel ? stacking.getStackLabel : undefined;
    const getGapRatio = (startIndex, endIndex) => {
        return utils.getBarGapRatio(isStacked, startIndex, endIndex);
    };
    const getNextChartItemFunc = (startIndex, endIndex) => {
        const stackGroups = isStacked
            ? Array.from(new Set(series.map((s) => getStackGroup(s) || 'default')))
            : [];
        return getBarChartNavUtil(getDataItem, series, series.length, startIndex, endIndex, isStacked, isRtl, isHoriz, stackGroups, getStackGroup);
    };
    const isSplitDualY = typeof y2Axis?.split === 'number';
    const getDataItemPos = (xScale, yScale, y2Scale, axisStepWidth, averageGroupZ, gapRatio) => {
        return BarGroup.getBarPosition(series, groups, getDataItem, xScale, yScale, y2Scale, hiddenSet, dataItemGaps, axisStepWidth, averageGroupZ, gapRatio, yAxis?.scale === 'log', isStacked, isRtl, isHoriz, isSplitDualY, getStackGroup, timeAxisType);
    };
    const findNearest = (xScale, yScale, y2Scale, axisStepWidth, averageGroupZ, gapRatio) => {
        const getSeriesDataInfo = BarGroup.getSeriesOffsetAndWidth(axisStepWidth, averageGroupZ, series, getDataItem, gapRatio, dataItemGaps, isStacked, isRtl, isHoriz);
        return utils.findNearestDataPoint(series, groups, xScale, yScale, getDataItem, isStacked, getSeriesDataInfo, BarGroup.getBarCenterCoordinate, yAxis?.scale === 'log', dataItemGaps, hiddenSet, timeAxisType, y2Scale);
    };
    const getMarkerColorAndType = (seriesIndex, groupIndex) => {
        return {
            color: dataStyleUtils.getItemColor(seriesIndex, groupIndex, series, getDataItem),
            type: 'circle'
        };
    };
    const getOffset = (groupCount) => {
        return utils.getStartAndEndAxesOffset(false, false, isHoriz, groupCount, true);
    };
    return (jsxRuntime.jsx(Chart.Chart, { ref: ref, width: width, height: height, yAxis: yAxis, xAxis: xAxis, y2Axis: y2Axis, selectedIds: selectedSet, hiddenIds: hiddenIds, hideAndShowBehavior: hideAndShowBehavior, orientation: orientation, series: series, groups: groups, isStacked: isStacked, getDataItem: getDataItem, stackLabelStyle: stackLabelStyle, selectionMode: selectionMode, getStackGroup: getStackGroup, drilling: drilling, renderGridLinesInFront: false, findNearest: findNearest, getDataItemPos: getDataItemPos, getNavUtil: getNextChartItemFunc, getGapRatio: getGapRatio, getMarkerInfo: getMarkerColorAndType, isRtl: isRtl, getStartAndEndAxesOffset: getOffset, defaultOverviewContentRenderer: () => {
            return jsxRuntime.jsx(OverviewBarChart, { series: series, groups: groups, getDataItem: getDataItem });
        }, ...props, children: ({ groupsInfo, xStartIndex, xEndIndex, xScale, yScale, y2Scale, gapRatio, activeId, focusedItemInfo, hoveredItemInfo, axisStepWidth, averageGroupZ, isPointInsideMarquee, getTextDimensions, defaultFontSize, resolvedColors, getItemAriaLabel, plotAreaClipPathId }) => {
            return (jsxRuntime.jsx("g", { clipPath: `url(#${plotAreaClipPathId})`, children: groupsInfo
                    .slice(xStartIndex, xEndIndex + 1)
                    .map(({ item: group }, index) => (jsxRuntime.jsx(BarGroup.BarGroup, { yScale: yScale, splitDualY: isSplitDualY, y2Scale: y2Scale, xScale: xScale, groupIndex: index + xStartIndex, selectedIds: selectedSet, hiddenIds: hiddenSet, highlightedIds: highlightedIds ? highlightedSet : highlightedIds, isDrillEnabled: drilling === 'on', series: series, barGapRatio: gapRatio, dataItemGaps: dataItemGaps, isPointInsideMarquee: isPointInsideMarquee, isStacked: isStacked, axisStepWidth: axisStepWidth, averageGroupZ: averageGroupZ, group: group, timeAxisType: xAxis?.timeAxisType, colors: resolvedColors, activeId: activeId, isLog: yAxis?.scale === 'log', orientation: orientation, getDataItem: getDataItem, isSelectionEnabled: selectionMode != 'none', isRtl: isRtl, focusedItemInfo: focusedItemInfo, hoveredItemInfo: hoveredItemInfo, getTextDimensions: getTextDimensions, defaultFontSize: defaultFontSize, getStackGroup: getStackGroup, getStackLabel: getStackLabel, getItemAriaLabel: getItemAriaLabel }))) }));
        } }));
});

exports.BarChart = BarChart;
