// @oracle/oraclejet-preact/cjs/UNSAFE_InputColor.js
/* @oracle/oraclejet-preact: undefined */
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var jsxRuntime = require('preact/jsx-runtime');
var compat = require('preact/compat');
var hooks = require('preact/hooks');
var swatchData = require('./swatchData-17c1d335.js');
var clientHints = require('./clientHints-c40c5250.js');
var FormFieldContext = require('./FormFieldContext-69fe83d2.js');
require('./logger-2b636482.js');
require('./TabbableModeContext-e99d527e.js');
var useTranslationBundle = require('./useTranslationBundle-12dd087d.js');


require('./UNSAFE_UserAssistance/themes/redwood/UserAssistanceVariants.css.js');
var InlineUserAssistance = require('./InlineUserAssistance-cfaa3d3e.js');
require('./Flex-3a257bc3.js');


require('./UNSAFE_MessageBanner/themes/redwood/MessageBannerVariants.css.js');
require('preact');
require('./LayerHost-3cd73603.js');


require('./hooks/UNSAFE_useTooltip/themes/redwood/TooltipContentVariants.css.js');
require('./tooltipUtils-2bc15136.js');

require('./IconButton-7503293b.js');

require('./MessageFormattingUtils-539d75b0.js');
require('./TransitionGroup-b239d98f.js');
require('./MessagesContext-4e939750.js');



require('./UNSAFE_ComponentMessage/themes/redwood/ComponentMessageVariants.css.js');
require('./Popup-de16d915.js');
require('./UNSAFE_Separator/themes/SeparatorStyles.css.js');


require('./UNSAFE_Label/themes/redwood/LabelVariants.css.js');
require('./InputGroupContext-05f2a46f.js');

require('./UNSAFE_Popup/themes/redwood/PopupVariants.css.js');
var UNSAFE_InputColor_themes_InputColorPickerDropdownStyles_css = require('./UNSAFE_InputColor/themes/InputColorPickerDropdownStyles.css.js');
var Sheet = require('./Sheet-1d6e8cf8.js');
var ColorPicker = require('./ColorPicker-9af6707c.js');
var SectionedContent = require('./SectionedContent-5eff5dce.js');
var useDownToggle = require('./useDownToggle-7b257d2b.js');
var useClearIcon = require('./useClearIcon-196c867e.js');
var useFocusableTextField = require('./useFocusableTextField-2c36840d.js');
require('./FormContext-d1d05967.js');
var useFormContext = require('./useFormContext-a794f7b6.js');
var useHover = require('./useHover-49b0430d.js');
var useTextField = require('./useTextField-6ecf97b7.js');
var Label = require('./Label-eb654b0e.js');
require('./keyboardUtils-b42fe7d5.js');
var StyledTextField = require('./StyledTextField-7839c90d.js');

var ReadonlyTextFieldInput = require('./ReadonlyTextFieldInput-60965d77.js');
var TextField = require('./TextField-5c164264.js');

require('./UNSAFE_TextField/themes/TextFieldStyles.css.js');




require('./UNSAFE_TextField/themes/redwood/TextFieldVariants.css.js');
var CompactUserAssistance = require('./CompactUserAssistance-a5f1bb2b.js');
var mergeProps = require('./mergeProps-e3da7237.js');
var ClearIcon = require('./ClearIcon-f61fc0b9.js');
var useCurrentValueReducer = require('./useCurrentValueReducer-f37ac024.js');
var useSelectableTextField = require('./useSelectableTextField-892b62db.js');
var ColorSwatch = require('./ColorSwatch-cddebd6b.js');
require('./colorUtils-01f3caa2.js');
require('./_curry1-e8f0d7ea.js');
require('./vanilla-extract-recipes-createRuntimeFn.esm-d1301b2d.js');
require('./ComponentMessageContainer-b692adf3.js');
require('./useTestId-4f53d561.js');
require('./MessagesManager-cba4bc60.js');
require('./classNames-c14c6ef3.js');
require('./SuccessS-b6757c64.js');
require('./Icon-56789beb.js');
require('./size-1914356b.js');
require('./utils-c0efcd5e.js');
require('./themeContract.css-f3d431c6.js');
require('./useTooltip-34accecc.js');
require('./useTooltipControlled-0a95856c.js');
require('./useId-6c0eeb27.js');
require('./useFocus-1b288fb9.js');
require('./useToggle-3ebba7d8.js');
require('./useTouch-4dec8729.js');
require('./useAnimation-65c0e455.js');
require('./useComponentTheme-082fc8e4.js');
require('./useThemeInterpolations-a37b37d9.js');
require('./useColorScheme-f8b1447d.js');
require('./useScale-db16c2b8.js');
require('./theme-dfd6a9fe.js');
require('./Theme-f06687af.js');
require('./mergeInterpolations-6727b536.js');
require('./mergeDeepWithKey-210b024d.js');
require('./_curry3-18677bca.js');
require('./_curry2-c15d89cd.js');
require('./_isObject-28636267.js');
require('./useDensity-0b9fd9ca.js');
require('./unsafeDomAccess-c3a492f0.js');
require('./hooks/UNSAFE_useTooltip/themes/redwood/TooltipContentTheme.js');
require('./hooks/UNSAFE_useTooltip/themes/TooltipContentStyles.css.js');
require('./hooks/UNSAFE_useTooltip/themes/redwood/TooltipContentBaseTheme.css.js');

require('./UNSAFE_Icon/themes/IconStyle.css.js');
require('./ErrorS-b027f7ad.js');
require('./InformationS-b0bb102b.js');
require('./WarningS-cdb0a7ad.js');
require('./PRIVATE_Message/themes/MessageStyles.css.js');
require('./soundUtils-ff3c9c0d.js');
require('./useMessagesContext-23d62331.js');
require('./Transition-942950ed.js');
require('./HiddenAccessible-087eb49a.js');
require('./UNSAFE_ComponentMessage/themes/redwood/ComponentMessageTheme.js');
require('./UNSAFE_ComponentMessage/themes/ComponentMessageStyles.css.js');
require('./UNSAFE_ComponentMessage/themes/redwood/ComponentMessageBaseTheme.css.js');
require('./UNSAFE_UserAssistance/themes/redwood/UserAssistanceTheme.js');
require('./UNSAFE_UserAssistance/themes/UserAssistanceStyles.css.js');
require('./UNSAFE_UserAssistance/themes/redwood/UserAssistanceBaseTheme.css.js');
require('./useFormFieldContext-4632eb73.js');
require('./InlineHelp-2add2d34.js');
require('./useTabbableMode-dbd254e8.js');
require('./useInputGroupContext-4c10bd57.js');
require('./boxalignment-e287d370.js');
require('./arrayUtils-7d8dcfc3.js');

require('./vanilla-extract-sprinkles-createRuntimeSprinkles.esm-d68f3e0f.js');
require('./dimensions-6101ca1f.js');
require('./flexbox-2cae9a01.js');

require('./flexitem-fee13e26.js');



require('./index-0aa99b92.js');
require('./UNSAFE_Layer/themes/LayerHostStyles.css.js');

require('./EnvironmentProvider-5f2435e6.js');
require('./LayerManager-c535d963.js');
require('./Floating-79cbe5ad.js');
require('./useFloating-c753a7f9.js');
require('./useUser-1295b5cb.js');
require('./positionUtils-a780137e.js');
require('./refUtils-a9872e75.js');
require('./useOutsideClick-dbf3d604.js');
require('./Layer-b60deb1b.js');
require('./UNSAFE_Layer/themes/LayerStyles.css.js');

require('./useModal-0739a528.js');
require('./UNSAFE_Floating/themes/redwood/FloatingTheme.js');
require('./UNSAFE_Floating/themes/FloatingStyles.css.js');

require('./UNSAFE_Floating/themes/redwood/FloatingBaseTheme.css.js');

require('./UNSAFE_Floating/themes/redwood/FloatingVariants.css.js');
require('./vanilla-extract-dynamic.esm-2e1b7c25.js');
require('./UNSAFE_Floating/themes/FloatingContract.css.js');
require('./useUnsafeDomElementRef-f22a2f0c.js');
require('./BaseButton-d8f96bfa.js');
require('./useActive-6770f917.js');
require('./UNSAFE_BaseButton/themes/redwood/BaseButtonTheme.js');
require('./UNSAFE_BaseButton/themes/BaseButtonStyles.css.js');

require('./UNSAFE_BaseButton/themes/redwood/BaseButtonVariants.css.js');

require('./BareButton-e8a030d7.js');
require('./useBareButton-463210c9.js');
require('./usePress-886180e4.js');
require('./usePressClick-7a34624a.js');
require('./buttonUtils-dff2ea3e.js');
require('./id-86356250.js');
require('./ButtonLabelLayout-c876923f.js');
require('./Text-056f1ec5.js');
require('./UNSAFE_Text/themes/TextStyles.css.js');

require('./useButtonLabelLayoutTheme-d09db0c9.js');
require('./UNSAFE_ButtonLabelLayout/themes/redwood/ButtonLabelLayoutTheme.js');
require('./UNSAFE_ButtonLabelLayout/themes/ButtonLabelLayoutStyles.css.js');

require('./UNSAFE_ButtonLabelLayout/themes/redwood/ButtonLabelLayoutBaseTheme.css.js');

require('./UNSAFE_ButtonLabelLayout/themes/redwood/ButtonLabelLayoutVariants.css.js');

require('./getLocale-8e17dc46.js');
require('./stringUtils-3e19c8af.js');
require('./tabbableUtils-22a50472.js');
require('./head-68d0992f.js');
require('./_arity-c228159c.js');
require('./_isArray-73160ad5.js');
require('./_isString-f4443c9e.js');
require('./FocusTrap-4873a3f2.js');
require('./FocusTracker-e2ea43c3.js');
require('./PRIVATE_FocusTracker/themes/FocusTrackerStyles.css.js');

require('./Modal-b751b643.js');
require('./UNSAFE_Modal/themes/ModalStyles.css.js');

require('./popupUtils-1d123cbe.js');
require('./WindowOverlay-85d999fa.js');
require('./UNSAFE_WindowOverlay/themes/WindowOverlayStyles.css.js');

require('./UNSAFE_WindowOverlay/themes/WindowOverlayContract.css.js');
require('./UNSAFE_WindowOverlay/themes/redwood/WindowOverlayTheme.js');
require('./UNSAFE_WindowOverlay/themes/redwood/WindowOverlayBaseTheme.css.js');

require('./UNSAFE_WindowOverlay/themes/redwood/WindowOverlayVariants.css.js');
require('./usePopupAnimation-f3b34413.js');
require('./useAnimationStatus-8075dc19.js');
require('./popupAnimationUtils-dcace427.js');
require('./LiveRegion-62ce9ac7.js');
require('./UNSAFE_Popup/themes/PopupContract.css.js');
require('./UNSAFE_Popup/themes/redwood/PopupTheme.js');
require('./UNSAFE_Popup/themes/redwood/PopupBaseTheme.css.js');

require('./colors-f3ce979a.js');


require('./UNSAFE_Sheet/themes/SheetStyles.css.js');

require('./animationUtils-54852d03.js');
require('./UNSAFE_ColorPicker/themes/ColorPickerContract.css.js');
require('./UNSAFE_ColorPicker/themes/redwood/ColorPickerTheme.js');
require('./UNSAFE_ColorPicker/themes/redwood/ColorPickerVariants.css.js');

require('./UNSAFE_ColorPicker/themes/redwood/ColorPickerBaseTheme.css.js');

require('./clientUtils-ef5e20ce.js');
require('./ColorPalette-9f041744.js');
require('./padding-ba041f6a.js');
require('./aria-889aedd0.js');
require('./resources/nls/en/bundle.js');
require('./SwatchFlexView-35c68cf1.js');
require('./useResizeObserver-a9e97180.js');
require('./useAddBusyState-d19ae1fa.js');
require('./BusyStateContext-86f40d3c.js');
require('./useBusyStateContext-362eee2a.js');
require('./BaseSwatchView-76da473f.js');
require('./keys-25710458.js');
require('./collectionUtils-f1d3183f.js');
require('./Selector-6550b8e0.js');
require('./BaseSelector-2dc69917.js');
require('./StyledCheckbox-9047ed7d.js');
require('./Check-1272bb86.js');
require('./CheckboxOff-a3ed334d.js');
require('./CheckboxOn-b5dda28b.js');
require('./CheckboxMixed-a17c8298.js');
require('./UNSAFE_Checkbox/themes/CheckboxIconContract.css.js');
require('./UNSAFE_Checkbox/themes/redwood/CheckboxIconTheme.js');
require('./UNSAFE_Checkbox/themes/CheckboxIconStyles.css.js');

require('./UNSAFE_Checkbox/themes/redwood/CheckboxIconBaseTheme.css.js');

require('./UNSAFE_Checkbox/themes/redwood/CheckboxIconVariants.css.js');


require('./PRIVATE_BaseSelector/themes/SelectorStyles.css.js');

require('./CollectionInteractionContext-f84ff7a1.js');
require('./useCollectionInteractionContext-6fc0200e.js');
require('./useCurrentKey-1b9e3382.js');
require('./useCollectionFocusRing-bed0aa16.js');
require('./useTabbableModeSet-5bce852a.js');
require('./useSwatchSelection-8ca16859.js');
require('./textSelectionUtils-57082239.js');
require('./LoadMoreCollection-eb1f2603.js');
require('./Collection-e27077ff.js');
require('./useViewportIntersect-2e802bac.js');
require('./PRIVATE_BaseSwatchView/themes/BaseSwatchViewStyles.css.js');

require('./Dropdown-1ba432a8.js');
require('./UNSAFE_Dropdown/themes/dropdownStyles.css.js');

require('./useOutsideMousedown-70f36cca.js');
require('./useEffectEvent-832ffba5.js');
require('./View-ed116d78.js');
require('./borders-4b8488cb.js');
require('./equals-098f7781.js');
require('./useFocusWithin-eb7f956a.js');
require('./truncationUtils-2d44b04b.js');
require('./useMessageSeverity-cfc4ae9d.js');
require('./useFormFieldContextProps-0fa67f6c.js');
require('./CompactLabelAssistance-e910bb56.js');
require('./Help-dbc23ccd.js');
require('./UNSAFE_Label/themes/redwood/LabelTheme.js');
require('./UNSAFE_Label/themes/LabelStyles.css.js');
require('./UNSAFE_Label/themes/redwood/LabelBaseTheme.css.js');
require('./UNSAFE_TextField/themes/redwood/TextFieldTheme.js');
require('./UNSAFE_TextField/themes/redwood/TextFieldBaseTheme.css.js');

require('./useAccessibleContext-c49d8d1b.js');
require('./useTextFieldInputHandlers-850d395c.js');
require('./formControlUtils-596c9dc3.js');
require('./textAlign-f41f49db.js');

require('./TextFieldUtils-96baac38.js');
require('./UNSAFE_TextField/themes/FormControlUtilsStyles.css.js');

require('./UNSAFE_TextField/themes/redwood/TextFieldInputTheme.js');
require('./UNSAFE_TextField/themes/TextFieldInputStyles.css.js');

require('./UNSAFE_TextField/themes/redwood/TextFieldInputVariants.css.js');

require('./LabelValueLayout-d67109b3.js');
require('./StyledLabelValueLayout-48da61b9.js');

require('./UNSAFE_TextField/themes/redwood/FormLayoutTheme.js');
require('./UNSAFE_TextField/themes/FormLayoutStyles.css.js');
require('./UNSAFE_TextField/themes/redwood/FormLayoutBaseTheme.css.js');

require('./formControlUtils-31d8bc55.js');
require('./LayoutStyles.css-b4025421.js');

require('./UNSAFE_Skeleton/themes/redwood/SkeletonTheme.js');
require('./UNSAFE_Skeleton/themes/SkeletonStyles.css.js');
require('./UNSAFE_Skeleton/themes/redwood/SkeletonBaseTheme.css.js');

require('./UNSAFE_Skeleton/themes/redwood/SkeletonVariants.css.js');
require('./UNSAFE_TextField/themes/redwood/ReadonlyTextFieldTheme.js');
require('./UNSAFE_TextField/themes/ReadonlyTextFieldStyles.css.js');

require('./UNSAFE_TextField/themes/redwood/ReadonlyTextFieldVariants.css.js');

require('./UNSAFE_TextField/themes/redwood/ReadonlyTextFieldInputTheme.js');
require('./UNSAFE_TextField/themes/ReadonlyTextFieldInputStyles.css.js');

require('./UNSAFE_TextField/themes/redwood/ReadonlyTextFieldInputVariants.css.js');
require('./Separator-85962472.js');
require('./DeleteCircleS-3567d884.js');
require('./useToggleAction-e4f15550.js');
require('./UNSAFE_ColorSwatch/themes/redwood/ColorSwatchTheme.js');
require('./UNSAFE_ColorSwatch/themes/ColorSwatchStyles.css.js');

require('./UNSAFE_ColorSwatch/themes/redwood/ColorSwatchBaseTheme.css.js');

require('./UNSAFE_ColorSwatch/themes/redwood/ColorSwatchVariants.css.js');
require('./CheckboxCrossed-40f3d0e6.js');

/**
 * Component that renders ColorPicker either inside a dropdown or sheet; in mobile,
 * it renders inside a sheet. Additionally, when a dropdown is being shown beneath
 * the component, we also show any user assistance text or messages below the picker
 * because otherwise that information would be obscured from view. When a sheet is being
 * shown, it will always display UA and messages at the bottom of the sheet.
 */
const InputColorPickerDropdown = ({ anchorRef, assistiveText, swatchSize, palette, helpSourceLink, helpSourceText, isDisabled, isOpen, isReadonly, label, onClose, onCommit, messages, testId, userAssistanceDensity, value }) => {
    // If the dropdown opens above the component, we don't want to show UA or messages.
    const [isDropdownAbove, setDropdownAbove] = hooks.useState(false);
    const [prevIsOpen, setPrevIsOpen] = hooks.useState(isOpen);
    const handleDropdownPosition = hooks.useCallback((data) => {
        setDropdownAbove(data.placement.startsWith('top'));
    }, []);
    // Reset the flag when the dropdown is closed.
    if (prevIsOpen !== isOpen) {
        setPrevIsOpen(isOpen);
        if (!isOpen) {
            setDropdownAbove(false);
        }
    }
    const sizeOptions = {
        isMaxWidthAdjusted: false,
        isMaxHeightAdjusted: true
    };
    // On mobile phones, the dropdown becomes a sheet.
    const shouldShowSheet = isPhone();
    const { colorPicker_label } = useTranslationBundle.useTranslationBundle('@oracle/oraclejet-preact');
    const colorPickerAriaLabel = colorPicker_label();
    const colorPickerMainContent = (jsxRuntime.jsx(ColorPicker.ColorPicker, { height: !shouldShowSheet ? 'auto' : undefined, onCommit: onCommit, 
        // for sheet, we want the width to grow to the width of the sheet, otherwise it will use the theme default.
        maxWidth: shouldShowSheet ? '100%' : undefined, width: "100%", swatchSize: swatchSize, palette: palette, testId: testId, value: value }));
    const inlineUserAssistanceFooter = isOpen &&
        (!isDropdownAbove || shouldShowSheet) &&
        !isDisabled &&
        !isReadonly &&
        (assistiveText || helpSourceLink || (messages && messages.length > 0)) &&
        (userAssistanceDensity === 'efficient' || userAssistanceDensity === 'reflow') ? (jsxRuntime.jsx(FormFieldContext.FormFieldContext.Provider, { value: { isFocused: true }, children: jsxRuntime.jsx("div", { class: UNSAFE_InputColor_themes_InputColorPickerDropdownStyles_css.userAssistanceStyles, children: jsxRuntime.jsx(InlineUserAssistance.InlineUserAssistance, { assistiveText: assistiveText, fieldLabel: label, helpSourceLink: helpSourceLink, helpSourceText: helpSourceText, messages: messages, userAssistanceDensity: userAssistanceDensity }) }) })) : undefined;
    const contentTestId = testId ? testId + (shouldShowSheet ? '_sheet' : '_dropdown') : undefined;
    // The same content is displayed either in a dropdown or sheet: color picker and inline UA.
    // Add styling to the dropdown to enforce a fixed width.
    const dropdownContent = (jsxRuntime.jsx("div", { class: UNSAFE_InputColor_themes_InputColorPickerDropdownStyles_css.dropdownStyles, children: jsxRuntime.jsx(SectionedContent.SectionedContent, { paddingBlockEnd: "3.5x", paddingBlockStart: "3.5x", paddingInlineEnd: "3.5x", paddingInlineStart: "3.5x", "aria-label": colorPickerAriaLabel, footer: inlineUserAssistanceFooter, hasFocusTrap: true, main: colorPickerMainContent, role: "dialog", testId: contentTestId }) }));
    const dropdown = shouldShowSheet ? (jsxRuntime.jsx(Sheet.Sheet, { isOpen: isOpen, onClose: onClose, children: jsxRuntime.jsx(SectionedContent.SectionedContent, { "aria-label": colorPickerAriaLabel, footer: inlineUserAssistanceFooter, hasFocusTrap: true, main: colorPickerMainContent, role: "dialog", testId: contentTestId }) })) : (jsxRuntime.jsx(SectionedContent.StickyPositioningDropdown, { anchorRef: anchorRef, isOpen: isOpen, onClose: onClose, onPosition: handleDropdownPosition, sizeOptions: sizeOptions, children: dropdownContent }));
    return dropdown;
};
const isPhone = () => {
    return clientHints.getClientHints().deviceType === 'phone';
};

/**
 * @license
 * Copyright (c) %FIRST_YEAR% %CURRENT_YEAR%, Oracle and/or its affiliates.
 * The Universal Permissive License (UPL), Version 1.0
 * as shown at https://oss.oracle.com/licenses/upl/
 * @ignore
 */
// Use a reducer function to manage InputColorState.
const inputColorReducer = (state, action) => {
    switch (action.type) {
        case 'colorChanged': {
            return {
                ...state,
                colorValue: action.data
            };
        }
        case 'colorSelected': {
            return { ...state, colorValue: action.data };
        }
        case 'reset': {
            return {
                colorValue: action.data
            };
        }
        default:
            return state;
    }
};
/**
 * useInputColorState calls useReducer to determine the initial state. Any changes to that state
 * are managed by dispatching actions to the reducer function.
 */
const useInputColorState = ({ value }) => {
    const [state, dispatch] = hooks.useReducer(inputColorReducer, {
        colorValue: value
    });
    return { state, dispatch };
};

/**
 * @license
 * Copyright (c) %FIRST_YEAR% %CURRENT_YEAR%, Oracle and/or its affiliates.
 * The Universal Permissive License (UPL), Version 1.0
 * as shown at https://oss.oracle.com/licenses/upl/
 * @ignore
 */
/**
 * Hook that provides handlers and other functionality for InputColor.
 */
const useInputColor = ({ onCommit, onInput, value: currentValue }) => {
    // Whether the dropdown is open.
    const [isOpen, setIsOpen] = hooks.useState(false);
    // Ref to the swatch icon button.
    const iconRef = hooks.useRef(null);
    // Ref to the text input which holds color value
    const colorRef = hooks.useRef(null);
    // Keep track of whether the dropdown was just dismissed from an outside mousedown event.
    const outsideMousedownRef = hooks.useRef(false);
    const outsideMousedownTimerRef = hooks.useRef();
    // Keep track of the last color values passed to onInput and onCommit.
    const lastCommittedColorRef = hooks.useRef(currentValue);
    const lastColorRef = hooks.useRef(currentValue);
    // Keep track of the last controlled value passed from the parent, so we can detect a programmatic change.
    const lastValueFromParentRef = hooks.useRef(currentValue);
    // Keep track of whether a color was selected in the color picker.
    const colorSelectedRef = hooks.useRef(false);
    const { state, dispatch } = useInputColorState({ value: currentValue });
    // If a new value is being pushed from the parent, and that value differs from
    // the current value of the reducer, then this is a programmatic update
    // and we need to call 'reset' to update the reducer state. Otherwise, the only
    // time the reducer state is updated from the value is on mount.
    if (swatchData.colorsAreDifferent(lastValueFromParentRef.current, currentValue)) {
        lastValueFromParentRef.current = currentValue;
        if (swatchData.colorsAreDifferent(currentValue, state.colorValue)) {
            dispatch({ type: 'reset', data: currentValue });
            lastColorRef.current = currentValue;
            lastCommittedColorRef.current = currentValue;
        }
    }
    // Reducer updates are async, so wait until the state updates and then call our callbacks.
    hooks.useEffect(() => {
        if (swatchData.colorsAreDifferent(state.colorValue, lastColorRef.current)) {
            onInput({ previousValue: lastColorRef.current, value: state.colorValue });
            lastColorRef.current = state.colorValue;
        }
        if (colorSelectedRef.current &&
            swatchData.colorsAreDifferent(state.colorValue, lastCommittedColorRef.current)) {
            onCommit?.({ previousValue: lastCommittedColorRef.current, value: state.colorValue });
            lastCommittedColorRef.current = state.colorValue;
            colorSelectedRef.current = false;
        }
    }, [onCommit, onInput, state.colorValue]);
    // Handle onInput from InputColor as values are entered in the input field.
    // We wait for the state to update before calling onInput.
    const colorInputHandler = hooks.useCallback((detail) => {
        dispatch({ type: 'colorChanged', data: detail.value });
    }, [dispatch]);
    // Handle onCommit from InputColor, which is called when the user presses Enter or the field is blurred.
    const colorCommitHandler = hooks.useCallback(() => {
        // If the current color value differs from the last committed value, then call onCommit.
        if (swatchData.colorsAreDifferent(lastCommittedColorRef.current, state.colorValue)) {
            onCommit?.({ value: state.colorValue, previousValue: lastCommittedColorRef.current });
            lastCommittedColorRef.current = state.colorValue;
        }
    }, [onCommit, state.colorValue]);
    // Respond to onAction event from the swatch icon button. Toggles the color picker dropdown.
    const colorIconActionHandler = hooks.useCallback(() => {
        // Only toggle the value of isOpen if we didn't just dismiss the dropdown.
        if (!outsideMousedownRef.current) {
            setIsOpen((prevIsOpen) => !prevIsOpen);
        }
        outsideMousedownRef.current = false;
    }, []);
    const dropdownCloseHandler = hooks.useCallback((detail) => {
        // Keep track of when the dropdown is closed via outside mousedown, including when the swatch icon
        // is clicked. We need this in order to toggle visibility of the picker and dropdown.
        if (detail.reason === 'outsideMousedown') {
            outsideMousedownRef.current = true;
            clearTimeout(outsideMousedownTimerRef.current);
            // Clear the flag after a short delay to ensure the dropdown can be opened again.
            outsideMousedownTimerRef.current = setTimeout(() => {
                outsideMousedownRef.current = false;
            }, 200);
        }
        else if (detail.reason === 'dismissed' || detail.reason === 'keyboardDismissed') {
            // Focus the swatch icon when the dropdown is closed.
            iconRef.current?.focus();
        }
        setIsOpen(false);
    }, []);
    // Handle color selection in the picker.
    //   - Updates the value in state.
    //   - Closes the dropdown.
    //   - Sets focus to the color field.
    const colorPickerCommitHandler = hooks.useCallback((detail) => {
        // Set focus to the swatch when a color is selected.
        iconRef.current?.focus();
        // Set a ref indicating a color was selected.
        colorSelectedRef.current = true;
        dispatch({ type: 'colorSelected', data: detail.value });
        setIsOpen(false);
    }, [dispatch]);
    // The Redwood UX design calls for the dropdown to open when the down arrow is pressed.
    // Handle opening the dropdown when the down arrow is pressed in the text edit field.
    const colorPickerDownHandler = hooks.useCallback(() => {
        setIsOpen(true);
    }, []);
    const { triggerProps: containerDropdownProps } = useDownToggle.useDownToggle({
        isOpen,
        onToggle: colorPickerDownHandler
    });
    return {
        colorIconActionHandler,
        colorPickerCommitHandler,
        dropdownCloseHandler,
        containerDropdownProps,
        colorValue: state.colorValue,
        iconRef,
        isOpen,
        colorCommitHandler,
        colorInputHandler,
        colorRef
    };
};

/**
 * An Input Color is a form field that allows users to select or input a color value,
 * using a color picker or by enterihg a HEX code.
 */
const InputColor = compat.forwardRef(({ 'aria-describedby': ariaDescribedBy, swatchSize, palette, value, isDisabled: propIsDisabled, isReadonly: propIsReadonly, labelEdge: propLabelEdge, labelStartWidth: propLabelStartWidth, isRequiredShown = false, isRequired = false, placeholder, hasClearIcon, columnSpan, textAlign: propTextAlign, autoComplete, messages, assistiveText, helpSourceLink, helpSourceText, label, onCommit, onInput, testId, variant, userAssistanceDensity: propUserAssistanceDensity }, ref) => {
    const { colorIconActionHandler, colorPickerCommitHandler, dropdownCloseHandler, colorValue, containerDropdownProps, iconRef, isOpen, colorCommitHandler, colorInputHandler, colorRef: enabledElementRef } = useInputColor({
        onCommit,
        onInput,
        value
    });
    const mainFieldRef = hooks.useRef(null);
    const { inputColor_selectColor } = useTranslationBundle.useTranslationBundle('@oracle/oraclejet-preact');
    const colorIconAriaLabel = inputColor_selectColor();
    const { onKeyDown, ...inputContainerAriaProps } = containerDropdownProps;
    const { currentCommitValue, dispatch } = useCurrentValueReducer.useCurrentValueReducer({ value: colorValue });
    const onInputAndDispatch = hooks.useCallback((detail) => {
        // Should dispatch happen first? This will queue up a re-render, ordering should not cause issues (this is async)
        dispatch({ type: 'input', payload: detail.value });
        colorInputHandler?.(detail);
    }, [colorInputHandler, dispatch]);
    const onCommitAndDispatch = hooks.useCallback((detail) => {
        // Should dispatch happen first? This will queue up a re-render, ordering should not cause issues (this is async)
        dispatch({ type: 'commit', payload: detail.value });
        colorCommitHandler?.(detail);
    }, [colorCommitHandler, dispatch]);
    const startContent = () => (jsxRuntime.jsx(ColorSwatch.ColorSwatch, { size: "xs", "aria-haspopup": palette ? 'grid' : undefined, "aria-label": palette ? colorIconAriaLabel : undefined, isDisabled: isDisabled, isButton: true, isTooltipDisabled: isOpen, ref: iconRef, testId: testId ? testId + '_colorIcon' : undefined, value: currentCommitValue, onToggle: isDisabled || !palette ? undefined : colorIconActionHandler }));
    const { isDisabled: isFormDisabled, isReadonly: isFormReadonly, labelEdge: formLabelEdge, labelStartWidth: formLabelStartWidth, textAlign: formTextAlign, userAssistanceDensity: formUserAssistanceDensity } = useFormContext.useFormContext();
    // default to FormContext values if component properties are not specified
    const isDisabled = propIsDisabled ?? isFormDisabled;
    const isReadonly = propIsReadonly ?? isFormReadonly;
    const labelEdge = propLabelEdge ?? formLabelEdge;
    const labelStartWidth = propLabelStartWidth ?? formLabelStartWidth;
    const textAlign = propTextAlign ?? formTextAlign;
    const userAssistanceDensity = propUserAssistanceDensity ?? formUserAssistanceDensity;
    const readonlyElementRef = hooks.useRef(null);
    const { focusProps, isFocused, methods: focusMethods } = useFocusableTextField.useFocusableTextField({
        isDisabled,
        isReadonly,
        enabledElementRef,
        readonlyElementRef
    });
    const { methods: selectMethods } = useSelectableTextField.useSelectableTextField(enabledElementRef);
    hooks.useImperativeHandle(ref, () => mergeProps.mergeProps(focusMethods, selectMethods), [
        focusMethods,
        selectMethods
    ]);
    const { hoverProps, isHover } = useHover.useHover({ isDisabled: isReadonly || isDisabled || false });
    const { formFieldContext, inputProps, labelProps, textFieldProps, userAssistanceProps } = useTextField.useTextField({
        ariaDescribedBy,
        helpSourceLink,
        helpSourceText,
        isDisabled,
        isFocused: isFocused || isOpen,
        isReadonly,
        isRequiredShown,
        labelEdge,
        messages,
        styleVariant: variant,
        userAssistanceDensity,
        value: colorValue
    });
    const onClickClearIcon = hooks.useCallback(() => {
        // Clicking the clear icon should put the focus on the input field
        enabledElementRef.current?.focus();
        // Send an event to clear the field's value if the value is not already cleared
        const nextValue = undefined;
        if (colorValue === nextValue)
            return;
        onInputAndDispatch?.({ previousValue: colorValue, value: nextValue });
    }, [onInputAndDispatch, colorValue, enabledElementRef]);
    const maybeClearIcon = useClearIcon.useClearIcon({
        clearIcon: jsxRuntime.jsx(ClearIcon.ClearIcon, { onClick: onClickClearIcon, testId: `${testId}_clearicon` }),
        display: hasClearIcon,
        hasValue: formFieldContext.hasValue,
        isFocused,
        isEnabled: !isReadonly && !isDisabled,
        isHover
    });
    const hasInsideLabel = label !== undefined && labelEdge === 'inside';
    const labelComp = labelEdge !== 'none' ? jsxRuntime.jsx(Label.Label, { ...labelProps, children: label }) : undefined;
    const fieldLabelProps = {
        label: labelEdge !== 'none' ? labelComp : undefined,
        labelEdge: labelEdge !== 'none' ? labelEdge : undefined,
        labelStartWidth: labelEdge !== 'none' ? labelStartWidth : undefined
    };
    const ariaLabel = labelEdge === 'none' ? label : undefined;
    const isInlineDensity = userAssistanceDensity === 'efficient' || userAssistanceDensity === 'reflow';
    const inlineUserAssistance = isInlineDensity ? (isDisabled || isReadonly ? (
    // save space for user assistance if density is 'efficient', even though we don't
    // render user assistance for disabled or readonly fields
    userAssistanceDensity !== 'efficient' ? undefined : (jsxRuntime.jsx(InlineUserAssistance.InlineUserAssistance, { userAssistanceDensity: userAssistanceDensity, ...userAssistanceProps }))) : (jsxRuntime.jsx(InlineUserAssistance.InlineUserAssistance, { assistiveText: assistiveText, fieldLabel: label, helpSourceLink: helpSourceLink, helpSourceText: helpSourceText, messages: messages, isRequiredShown: isRequiredShown, userAssistanceDensity: userAssistanceDensity, ...userAssistanceProps }))) : undefined;
    const aRef = hooks.useRef(null);
    const anchorRef = mainFieldRef ?? aRef;
    const compactUserAssistance = userAssistanceDensity === 'compact' ? (jsxRuntime.jsx(CompactUserAssistance.CompactUserAssistance, { anchorRef: anchorRef, messages: messages, assistiveText: assistiveText, ...userAssistanceProps })) : undefined;
    const mainContent = isReadonly ? (jsxRuntime.jsx(ReadonlyTextFieldInput.ReadonlyTextFieldInput, { "aria-describedby": inputProps['aria-describedby'], "aria-label": ariaLabel, "aria-labelledby": labelProps.id, as: "div", elementRef: readonlyElementRef, textAlign: textAlign, value: colorValue, hasEmptyLabel: label === '' && labelEdge === 'none', hasInsideLabel: hasInsideLabel })) : (jsxRuntime.jsx(StyledTextField.TextFieldInput, { "aria-labelledby": labelProps.id, "aria-label": ariaLabel, autoComplete: autoComplete, currentCommitValue: currentCommitValue, hasEmptyLabel: label === '' && labelEdge === 'none', hasEndContent: false, hasInsideLabel: labelComp !== undefined && labelEdge === 'inside', hasStartContent: true, inputRef: enabledElementRef, isRequired: isRequired, placeholder: placeholder, textAlign: textAlign, type: isMobile() ? 'text' : undefined, value: colorValue, onCommit: onCommitAndDispatch, onInput: onInputAndDispatch, onKeyDown: onKeyDown, ...inputProps }));
    return (jsxRuntime.jsxs(jsxRuntime.Fragment, { children: [jsxRuntime.jsx(FormFieldContext.FormFieldContext.Provider, { value: formFieldContext, children: isReadonly ? (jsxRuntime.jsx(StyledTextField.ReadonlyTextField, { role: "presentation", columnSpan: columnSpan, compactUserAssistance: compactUserAssistance, inlineUserAssistance: inlineUserAssistance, onBlur: focusProps.onFocusOut, onFocus: focusProps.onFocusIn, ref: anchorRef, mainContent: mainContent, testId: testId, ...fieldLabelProps })) : (jsxRuntime.jsx(TextField.TextField, { endContent: maybeClearIcon, columnSpan: columnSpan, compactUserAssistance: compactUserAssistance, inlineUserAssistance: inlineUserAssistance, mainContent: mainContent, onBlur: focusProps.onFocusOut, onFocus: focusProps.onFocusIn, startContent: startContent(), mainFieldRef: anchorRef, hasZeroStartMargin: true, testId: testId, ...textFieldProps, ...fieldLabelProps, ...hoverProps, ...inputContainerAriaProps })) }), isOpen && !isReadonly && !isDisabled && (jsxRuntime.jsx(InputColorPickerDropdown, { anchorRef: mainFieldRef, assistiveText: assistiveText, helpSourceLink: helpSourceLink, helpSourceText: helpSourceText, isDisabled: isDisabled, isReadonly: isReadonly, isOpen: isOpen, label: label, onClose: dropdownCloseHandler, onCommit: colorPickerCommitHandler, messages: messages, testId: testId ? testId + '_colorPicker' : undefined, userAssistanceDensity: userAssistanceDensity, 
                // memoize this for performance
                value: colorValue && swatchData.isValidColorValue(colorValue) ? colorValue : undefined, palette: palette, swatchSize: swatchSize }))] }));
});
/**
 * Helper function to determine whether the current device is a mobile device
 * @returns true if running on a mobile device, false otherwise
 */
function isMobile() {
    const deviceType = clientHints.getClientHints().deviceType;
    return deviceType === 'phone' || deviceType === 'tablet';
}
InputColor.displayName = 'InputColor';

exports.InputColor = InputColor;
