// @oracle/oraclejet-preact/cjs/UNSAFE_MessageToast.js
/* @oracle/oraclejet-preact: undefined */
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var jsxRuntime = require('preact/jsx-runtime');
var hooks = require('preact/hooks');
require('./LayerHost-3cd73603.js');
var EnvironmentProvider = require('./EnvironmentProvider-5f2435e6.js');
var useMessagesFocusManager = require('./useMessagesFocusManager-d10f7d1e.js');
var useBreakpointValues = require('./useBreakpointValues-66b40e00.js');
require('./MessagesContext-4e939750.js');
var useMessagesContext = require('./useMessagesContext-23d62331.js');
var useTranslationBundle = require('./useTranslationBundle-12dd087d.js');
var Message = require('./Message-258e59da.js');
require('preact');
var utils = require('./utils-c0efcd5e.js');
var useComponentTheme = require('./useComponentTheme-082fc8e4.js');
var useTestId = require('./useTestId-4f53d561.js');
var useColorScheme = require('./useColorScheme-f8b1447d.js');


require('./hooks/UNSAFE_useTooltip/themes/redwood/TooltipContentVariants.css.js');
require('./tooltipUtils-2bc15136.js');

var clientHints = require('./clientHints-c40c5250.js');
require('./IconButton-7503293b.js');

require('./MessageFormattingUtils-539d75b0.js');
var MessagesManager = require('./MessagesManager-cba4bc60.js');
var Flex = require('./Flex-3a257bc3.js');
var FocusTrap = require('./FocusTrap-4873a3f2.js');
var LiveRegion = require('./LiveRegion-62ce9ac7.js');
var UNSAFE_MessageToast_themes_redwood_MessageToastTheme = require('./UNSAFE_MessageToast/themes/redwood/MessageToastTheme.js');
var MessageLayer = require('./MessageLayer-8e6e3d98.js');
require('./index-0aa99b92.js');
require('preact/compat');
require('./UNSAFE_Layer/themes/LayerHostStyles.css.js');

require('./LayerManager-c535d963.js');
require('./useBreakpoints-9dbbd61b.js');
require('./UNSAFE_MessageBanner/themes/redwood/MessageBannerTheme.js');
require('./UNSAFE_MessageBanner/themes/MessageBannerStyles.css.js');

require('./UNSAFE_MessageBanner/themes/redwood/MessageBannerBaseTheme.css.js');

require('./UNSAFE_MessageBanner/themes/redwood/MessageBannerVariants.css.js');

require('./vanilla-extract-recipes-createRuntimeFn.esm-d1301b2d.js');
require('./timer-c32f7725.js');
require('./classNames-c14c6ef3.js');
require('./Close-d99c20a7.js');
require('./Icon-56789beb.js');
require('./size-1914356b.js');
require('./useTooltip-34accecc.js');
require('./useTooltipControlled-0a95856c.js');
require('./useId-6c0eeb27.js');
require('./useHover-49b0430d.js');
require('./useToggle-3ebba7d8.js');
require('./useFocus-1b288fb9.js');
require('./useTouch-4dec8729.js');
require('./mergeProps-e3da7237.js');
require('./useAnimation-65c0e455.js');
require('./useThemeInterpolations-a37b37d9.js');
require('./useScale-db16c2b8.js');
require('./theme-dfd6a9fe.js');
require('./Theme-f06687af.js');
require('./mergeInterpolations-6727b536.js');
require('./mergeDeepWithKey-210b024d.js');
require('./_curry3-18677bca.js');
require('./_curry1-e8f0d7ea.js');
require('./_curry2-c15d89cd.js');
require('./_isObject-28636267.js');
require('./useDensity-0b9fd9ca.js');
require('./unsafeDomAccess-c3a492f0.js');
require('./hooks/UNSAFE_useTooltip/themes/redwood/TooltipContentTheme.js');
require('./hooks/UNSAFE_useTooltip/themes/TooltipContentStyles.css.js');
require('./hooks/UNSAFE_useTooltip/themes/redwood/TooltipContentBaseTheme.css.js');

require('./logger-2b636482.js');
require('./UNSAFE_Icon/themes/IconStyle.css.js');
require('./PRIVATE_Message/themes/MessageStyles.css.js');
require('./themeContract.css-f3d431c6.js');
require('./colorUtils-01f3caa2.js');
require('./Floating-79cbe5ad.js');
require('./useFloating-c753a7f9.js');
require('./useUser-1295b5cb.js');
require('./positionUtils-a780137e.js');
require('./refUtils-a9872e75.js');
require('./useOutsideClick-dbf3d604.js');
require('./arrayUtils-7d8dcfc3.js');
require('./Layer-b60deb1b.js');
require('./UNSAFE_Layer/themes/LayerStyles.css.js');

require('./useModal-0739a528.js');
require('./UNSAFE_Floating/themes/redwood/FloatingTheme.js');
require('./UNSAFE_Floating/themes/FloatingStyles.css.js');

require('./UNSAFE_Floating/themes/redwood/FloatingBaseTheme.css.js');

require('./UNSAFE_Floating/themes/redwood/FloatingVariants.css.js');
require('./vanilla-extract-dynamic.esm-2e1b7c25.js');
require('./UNSAFE_Floating/themes/FloatingContract.css.js');
require('./useUnsafeDomElementRef-f22a2f0c.js');
require('./BaseButton-d8f96bfa.js');
require('./useActive-6770f917.js');
require('./dimensions-6101ca1f.js');
require('./UNSAFE_BaseButton/themes/redwood/BaseButtonTheme.js');
require('./UNSAFE_BaseButton/themes/BaseButtonStyles.css.js');

require('./UNSAFE_BaseButton/themes/redwood/BaseButtonVariants.css.js');

require('./BareButton-e8a030d7.js');
require('./useBareButton-463210c9.js');
require('./usePress-886180e4.js');
require('./usePressClick-7a34624a.js');
require('./TabbableModeContext-e99d527e.js');
require('./useTabbableMode-dbd254e8.js');
require('./buttonUtils-dff2ea3e.js');
require('./id-86356250.js');
require('./ButtonLabelLayout-c876923f.js');
require('./Text-056f1ec5.js');
require('./UNSAFE_Text/themes/TextStyles.css.js');

require('./useButtonLabelLayoutTheme-d09db0c9.js');
require('./UNSAFE_ButtonLabelLayout/themes/redwood/ButtonLabelLayoutTheme.js');
require('./UNSAFE_ButtonLabelLayout/themes/ButtonLabelLayoutStyles.css.js');

require('./UNSAFE_ButtonLabelLayout/themes/redwood/ButtonLabelLayoutBaseTheme.css.js');

require('./UNSAFE_ButtonLabelLayout/themes/redwood/ButtonLabelLayoutVariants.css.js');

require('./getLocale-8e17dc46.js');
require('./stringUtils-3e19c8af.js');
require('./SuccessS-b6757c64.js');
require('./ErrorS-b027f7ad.js');
require('./InformationS-b0bb102b.js');
require('./WarningS-cdb0a7ad.js');
require('./soundUtils-ff3c9c0d.js');
require('./TransitionGroup-b239d98f.js');
require('./Transition-942950ed.js');
require('./boxalignment-e287d370.js');

require('./vanilla-extract-sprinkles-createRuntimeSprinkles.esm-d68f3e0f.js');
require('./flexbox-2cae9a01.js');

require('./flexitem-fee13e26.js');


require('./tabbableUtils-22a50472.js');
require('./head-68d0992f.js');
require('./_arity-c228159c.js');
require('./_isArray-73160ad5.js');
require('./_isString-f4443c9e.js');
require('./FocusTracker-e2ea43c3.js');
require('./PRIVATE_FocusTracker/themes/FocusTrackerStyles.css.js');


require('./UNSAFE_MessageToast/themes/MessageToastStyles.css.js');

require('./UNSAFE_MessageToast/themes/redwood/MessageToastBaseTheme.css.js');

require('./UNSAFE_MessageToast/themes/redwood/MessageToastVariants.css.js');
require('./WindowOverlay-85d999fa.js');
require('./UNSAFE_WindowOverlay/themes/WindowOverlayStyles.css.js');

require('./UNSAFE_WindowOverlay/themes/WindowOverlayContract.css.js');
require('./UNSAFE_WindowOverlay/themes/redwood/WindowOverlayTheme.js');
require('./UNSAFE_WindowOverlay/themes/redwood/WindowOverlayBaseTheme.css.js');

require('./UNSAFE_WindowOverlay/themes/redwood/WindowOverlayVariants.css.js');

/**
 * Animation states for toast messages
 * TODO: Get this from theme context provider
 */
const animationStates = {
    entering: (node) => ({
        from: {
            transform: `translate(0, ${node.offsetHeight}px)`,
            opacity: 1
        },
        to: {
            transform: 'translate(0, 0)',
            opacity: 1
        },
        end: {
            transform: 'unset',
            opacity: 1
        },
        options: {
            duration: 250
        }
    }),
    exiting: {
        from: {
            opacity: 1
        },
        to: {
            opacity: 0
        },
        end: {
            opacity: 0
        },
        options: {
            duration: 250
        }
    }
};
const initialAnimationStyles = {
    opacity: 0
};
/**
 * Message toasts are short, non-critical, auto-dismissible messages that communicate non-disruptive contextual messages.
 */
function MessageToast({ data, detailRendererKey, iconRendererKey, renderers, onClose, offset, position, testId }) {
    const testIdProps = useTestId.useTestId(testId);
    const messagesRef = hooks.useRef(new Map());
    const autoFocusRef = hooks.useRef(null);
    const containerDivRef = hooks.useRef(null);
    const focusHandleRef = hooks.useRef(null);
    const actionRef = hooks.useRef();
    const [liveRegionText, setLiveRegionText] = hooks.useState();
    const [shouldRender, setShouldRender] = hooks.useState(data.length > 0);
    const [shouldTrapFocus, setShouldTrapFocus] = hooks.useState(false);
    const { deviceType } = clientHints.getClientHints();
    const translations = useTranslationBundle.useTranslationBundle('@oracle/oraclejet-preact');
    const { UNSAFE_messagesLayerId } = useMessagesContext.useMessagesContext();
    // We need a ref that holds the current & previous data length, as the exiting
    // node will always call handleNextFocus with previous data.
    // As in TransitionGroup, when an item is removed from the data, a new vnode
    // will not be created instead previous vnode will be used. So, the new handleNextFocus
    // will not be called when the old vnode exits. Thus, we will be using a ref
    // to always get the correct current data length.
    const dataLengthRef = hooks.useRef(data.length);
    const prevDataLengthRef = hooks.useRef(0);
    // Update the data length ref
    dataLengthRef.current = data.length;
    // Update the focusHandleRef
    hooks.useImperativeHandle(focusHandleRef, () => ({
        focus: () => {
            // Only trigger focus if the component is rendering messages
            if (data.length) {
                // set the autoFocus ref to the first message and enable the focus trap
                // this will automatically focus the first message
                autoFocusRef.current = messagesRef.current.get(data[0].key) ?? null;
                setShouldTrapFocus(true);
                return true;
            }
            return false;
        },
        contains: (element) => {
            // Only invoke method if the component is rendering messages
            if (data.length && element) {
                return containerDivRef.current?.contains(element) ?? false;
            }
            return false;
        }
    }), [data]);
    // Register handlers for focus management
    const { controller, handlers } = useMessagesFocusManager.useMessageFocusManager(focusHandleRef, {
        onFocus: hooks.useCallback(() => {
            setLiveRegionText(MessagesManager.getLiveRegionTextByDevice('from', deviceType, translations));
        }, [deviceType, translations]),
        onFocusLeave: hooks.useCallback(() => {
            // disable trap which should revert the focus back to the last focused item
            setShouldTrapFocus(false);
        }, [])
    }, {
        // we do not want the focus manager to handle Esc key as we handle it using the
        // FocusTrap here
        handleEscapeKey: false
    });
    /**
     * Handles focus when a message is closed and animated away from the DOM
     * @param key The key of the message
     * @param index The index of the message
     */
    const handleNextFocus = hooks.useCallback((_key, index, closedMessageNode) => {
        // we track the last action performed to close out the messages, we only want to
        // proceed if the action was performed using keyboard
        if (actionRef?.current?.reason !== 'keyboard') {
            return;
        }
        const isClosedMessageFocused = closedMessageNode?.contains(document.activeElement);
        // If there are no messages, do not render anything. As the old messages
        // are still in the DOM, use the data count to determine what to do next as it
        // represents the next state.
        if (dataLengthRef.current === 0) {
            setShouldRender(false);
            setShouldTrapFocus(false);
            // If the current message holds focus, then restore previous focus
            if (isClosedMessageFocused) {
                // As we have FocusTrap, trying to restore focus now will simply result in focus being pulled
                // back again. So, we restore the focus in the next tick so that the rendering will have been
                // done in the microtask.
                setTimeout(() => controller.restorePriorFocus(), 0);
            }
            return;
        }
        // In TransitionGroup, when an item is removed from the data, a new vnode
        // will not be created instead previous vnode will be used. So, the new handleNextFocus
        // will not be called when the old vnode exits. Thus, all the deps will not be pointing to
        // the current references, rather they will be pointing to the ones where this vnode
        // last existed in the data.
        // This means, the data will contain the closing message as well. But, all the ref objects
        // still point to the most up-to-date values. With that in mind, we will be evaluating the
        // following values.
        const renderedMessagesCount = data.length;
        // Now that this message is closed, focus the next message that will take this index. If no
        // message will take this message's index, then it means that this is the last message. If
        // that is the case, focus the message at the previous index.
        // Use the count of the messages that are currently shown in the UI (current state including
        // the message that will be removed). This way we can get the correct item from the data
        // as it will contain the closing message as well.
        const nextMessageIndexToFocus = index + 1 < renderedMessagesCount ? index + 1 : index - 1;
        // if next message is available then transfer the focus to the next element
        if (nextMessageIndexToFocus > -1 && isClosedMessageFocused) {
            const nextMessageKey = data[nextMessageIndexToFocus].key;
            messagesRef.current.get(nextMessageKey)?.focus();
        }
    }, [controller, data]);
    // We will be using a function based ref to store the refs in the Map
    const setHTMLDivElement = hooks.useCallback(
    // TODO: use ramda.curry here (currently it is throwing ts error when used with generics)
    (key) => {
        return (ref) => messagesRef.current.set(key, ref);
    }, []);
    const onCloseHandler = hooks.useCallback((item, actionDetail) => {
        actionRef.current = actionDetail;
        onClose?.(item);
    }, [onClose]);
    // Prioritize this component whenever the data changes and
    // the new data has at least one message
    hooks.useEffect(() => {
        const previousDataLength = prevDataLengthRef.current;
        prevDataLengthRef.current = data.length;
        // if there are messages that need to be rendered now, then do the following deeds:
        // 1. set the shouldRender flag
        // 2. set the live region
        // 3. prioritize this message component to be the latest
        if (data.length) {
            // set state to render content whenever the data is not empty
            setShouldRender(true);
            if (data.length > previousDataLength) {
                // Only when having a new message, update the aria-live area with the
                // text to indicate how to get the focus to the new message.
                setLiveRegionText(MessagesManager.getLiveRegionTextByDevice('to', deviceType, translations));
            }
            controller.prioritize();
            return;
        }
        // if there were messages and now removed, then set the live region to
        // announce this information
        if (previousDataLength) {
            setLiveRegionText(translations.messageToast_allMessagesClosed());
            return;
        }
        // When there are no messages, clear the live region so that
        // the navigation text will be read when a new message appear
        setLiveRegionText('');
    }, [controller, data.length, deviceType, translations]);
    // determine responsive max-width
    const maxWidth = useBreakpointValues.useBreakpointValues({
        sm: `calc(100vw - ${utils.xUnits(8)})`,
        md: `calc(100vw - ${utils.xUnits(12)})`,
        lg: `calc(150 * ${utils.xUnits(1)})`
    });
    // without offset, the messages will be hugging the vertical edges of the screen
    // so we will set a default offset.vertical value (any offset from the props will be
    // added to this).
    const paddedOffset = useBreakpointValues.useBreakpointValues({
        sm: {
            horizontal: typeof offset === 'number' ? offset : offset?.horizontal ?? 0,
            vertical: 16 + (typeof offset === 'number' ? offset : offset?.vertical ?? 0)
        },
        lg: {
            horizontal: typeof offset === 'number' ? offset : offset?.horizontal ?? 0,
            vertical: 24 + (typeof offset === 'number' ? offset : offset?.vertical ?? 0)
        }
    });
    const colorScheme = useColorScheme.useColorScheme();
    const newColorScheme = !colorScheme || colorScheme === 'light' ? 'dark' : 'light';
    const { classes } = useComponentTheme.useComponentTheme(UNSAFE_MessageToast_themes_redwood_MessageToastTheme.MessageToastRedwoodTheme);
    // When both shouldRender flag is false and no data to render, do not render
    // anything except the live region
    if (!shouldRender && data.length === 0) {
        return jsxRuntime.jsx(LiveRegion.LiveRegion, { children: liveRegionText });
    }
    return (jsxRuntime.jsxs(EnvironmentProvider.EnvironmentProvider, { environment: { colorScheme: newColorScheme }, children: [jsxRuntime.jsx(MessageLayer.MessageLayer, { offset: paddedOffset, position: position, children: jsxRuntime.jsx("div", { class: classes, ref: containerDivRef, tabIndex: -1, id: UNSAFE_messagesLayerId, ...handlers, ...testIdProps, children: jsxRuntime.jsx(FocusTrap.FocusTrap, { autoFocusRef: autoFocusRef, isDisabled: !shouldTrapFocus, children: jsxRuntime.jsx(Flex.Flex, { direction: "column", gap: "1x", align: "center", maxWidth: maxWidth, children: jsxRuntime.jsx(MessagesManager.MessagesManager, { animationStates: animationStates, initialAnimationStyles: initialAnimationStyles, data: data, onMessageWillRemove: handleNextFocus, children: ({ item }) => (jsxRuntime.jsx(Message.Message, { messageRef: setHTMLDivElement(item.key), item: item, detailRenderer: MessagesManager.getRenderer(item, detailRendererKey, renderers), iconRenderer: MessagesManager.getRenderer(item, iconRendererKey, renderers), onClose: onCloseHandler, variant: "toast" }, item.key)) }) }) }) }) }), jsxRuntime.jsx(LiveRegion.LiveRegion, { children: liveRegionText })] }));
}

exports.MessageToast = MessageToast;
