// @oracle/oraclejet-preact/cjs/UNSAFE_LineAreaChart.js
/* @oracle/oraclejet-preact: undefined */
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var jsxRuntime = require('preact/jsx-runtime');
var utils = require('./utils-ee4fc356.js');
var ReferenceLineArea = require('./ReferenceLineArea-7bc34276.js');
var MarkersWithLabel = require('./MarkersWithLabel-78bb1051.js');
var dataStyleUtils = require('./dataStyleUtils-4b0c5ce2.js');
var useUser = require('./useUser-1295b5cb.js');
var colorUtils = require('./colorUtils-b2204778.js');
require('./accUtils-4a359187.js');
var themeContract_css = require('./themeContract.css-f3d431c6.js');
var chartUtils = require('./chartUtils-0cd6f5ff.js');
var compat = require('preact/compat');
var Chart = require('./Chart-13a198f3.js');
require('./utils-c0efcd5e.js');
require('./colorUtils-01f3caa2.js');
require('./_curry1-e8f0d7ea.js');
require('./getLocale-8e17dc46.js');
require('./stringUtils-3e19c8af.js');
require('./labelUtils-45add12c.js');
require('./textUtils-e779c845.js');
require('./utils-ca441171.js');
require('./datatipUtils-d1ca81a1.js');
require('./clientUtils-a4619fcd.js');
require('./clientHints-c40c5250.js');
require('./dimensionUtils-3919753d.js');
require('./scale-9e43a5f6.js');
require('./PRIVATE_PlotArea/themes/PlotAreaStyles.css.js');

require('./hooks/PRIVATE_useVisDnd/themes/VisDndStyles.css.js');

require('./classNames-c14c6ef3.js');
require('./useColorScheme-f8b1447d.js');
require('preact/hooks');
require('./LayerHost-3cd73603.js');
require('preact');
require('./index-0aa99b92.js');
require('./UNSAFE_Layer/themes/LayerHostStyles.css.js');

require('./lineUtils-222ae5af.js');
require('./PRIVATE_Chart/themes/ChartStyles.css.js');

require('./Marker-b40392cc.js');

require('./TrackResizeContainer-a8631f0d.js');
require('./useSize-8115eae0.js');
require('./useResizeObserver-a9e97180.js');
require('./useAddBusyState-d19ae1fa.js');
require('./BusyStateContext-86f40d3c.js');
require('./useBusyStateContext-362eee2a.js');
require('./unsafeDomAccess-c3a492f0.js');
require('./size-1914356b.js');
require('./UNSAFE_visErrors.js');
require('./useTextDimensions-c8b41697.js');
require('./useCssVars-8e83d496.js');
require('./useTheme-2a034a21.js');
require('./useScale-db16c2b8.js');
require('./useId-6c0eeb27.js');
require('./Axis-a3a56403.js');
require('./PRIVATE_Axis/themes/AxisStyles.css.js');

require('./MultilineText-837e9afd.js');
require('./PRIVATE_Axis/themes/redwood/AxisTheme.js');
require('./PRIVATE_Axis/themes/redwood/AxisBaseTheme.css.js');

require('./PRIVATE_Axis/themes/redwood/AxisVariants.css.js');
require('./vanilla-extract-recipes-createRuntimeFn.esm-d1301b2d.js');
require('./useComponentTheme-082fc8e4.js');
require('./logger-2b636482.js');
require('./useChartViewport-25f7d6db.js');
require('./eventsUtils-b87ea91f.js');
require('./usePinch-741e9343.js');
require('./usePointerGesture-0b9cc1b7.js');
require('./mergeProps-e3da7237.js');
require('./useChartNav-c69c817e.js');
require('./layoutUtils-af53a104.js');
require('./useVisHover-ad36112d.js');
require('./useSelectDrill-0b4b47ec.js');
require('./useChartDataCursor-67f7cc75.js');
require('./hooks/PRIVATE_useChartDataCursor/themes/DataCursorStyles.css.js');

require('./useChartDatatip-1bceba5d.js');
require('./useDatatip-b7b4f13f.js');
require('./Layer-b60deb1b.js');
require('./UNSAFE_Layer/themes/LayerStyles.css.js');

require('./useThemeInterpolations-a37b37d9.js');
require('./theme-dfd6a9fe.js');
require('./Theme-f06687af.js');
require('./mergeInterpolations-6727b536.js');
require('./mergeDeepWithKey-210b024d.js');
require('./_curry3-18677bca.js');
require('./_curry2-c15d89cd.js');
require('./_isObject-28636267.js');
require('./useDensity-0b9fd9ca.js');
require('./Floating-79cbe5ad.js');
require('./useFloating-c753a7f9.js');
require('./positionUtils-a780137e.js');
require('./refUtils-a9872e75.js');
require('./useOutsideClick-dbf3d604.js');
require('./arrayUtils-7d8dcfc3.js');
require('./useModal-0739a528.js');
require('./UNSAFE_Floating/themes/redwood/FloatingTheme.js');
require('./UNSAFE_Floating/themes/FloatingStyles.css.js');

require('./UNSAFE_Floating/themes/redwood/FloatingBaseTheme.css.js');

require('./UNSAFE_Floating/themes/redwood/FloatingVariants.css.js');
require('./vanilla-extract-dynamic.esm-2e1b7c25.js');
require('./UNSAFE_Floating/themes/FloatingContract.css.js');
require('./useUnsafeDomElementRef-f22a2f0c.js');
require('./hooks/PRIVATE_useDatatip/themes/useDatatip.css.js');

require('./VisTabularDatatip-abfef042.js');
require('./PRIVATE_VisTabularDatatip/themes/VisTabularDatatipStyles.css.js');

require('./useChartDrill-c052942f.js');
require('./useVisDrill-f76c2e9c.js');
require('./useDoubleTap-d3937fb6.js');
require('./clientUtils-ef5e20ce.js');
require('./useSelection-da64e28e.js');
require('./BaseButton-d8f96bfa.js');
require('./useHover-49b0430d.js');
require('./useToggle-3ebba7d8.js');
require('./useActive-6770f917.js');
require('./dimensions-6101ca1f.js');
require('./UNSAFE_BaseButton/themes/redwood/BaseButtonTheme.js');
require('./UNSAFE_BaseButton/themes/BaseButtonStyles.css.js');

require('./UNSAFE_BaseButton/themes/redwood/BaseButtonVariants.css.js');

require('./BareButton-e8a030d7.js');
require('./useBareButton-463210c9.js');
require('./usePress-886180e4.js');
require('./usePressClick-7a34624a.js');
require('./TabbableModeContext-e99d527e.js');
require('./useTabbableMode-dbd254e8.js');
require('./buttonUtils-dff2ea3e.js');
require('./id-86356250.js');
require('./useTestId-4f53d561.js');
require('./Text-056f1ec5.js');
require('./UNSAFE_Text/themes/TextStyles.css.js');

require('./useTranslationBundle-12dd087d.js');
require('./hooks/PRIVATE_useVisSelection/themes/ClearSelection.css.js');

require('./cssUtils-60d6bb40.js');
require('./useChartContextMenu-af989b4f.js');
require('./useContextMenu-2500df06.js');
require('./SelectMenuGroupContext-a349d259.js');
require('./Sheet-1d6e8cf8.js');
require('./Modal-b751b643.js');
require('./UNSAFE_Modal/themes/ModalStyles.css.js');

require('./tabbableUtils-22a50472.js');
require('./head-68d0992f.js');
require('./_arity-c228159c.js');
require('./_isArray-73160ad5.js');
require('./_isString-f4443c9e.js');
require('./popupUtils-1d123cbe.js');
require('./WindowOverlay-85d999fa.js');
require('./UNSAFE_WindowOverlay/themes/WindowOverlayStyles.css.js');

require('./UNSAFE_WindowOverlay/themes/WindowOverlayContract.css.js');
require('./UNSAFE_WindowOverlay/themes/redwood/WindowOverlayTheme.js');
require('./UNSAFE_WindowOverlay/themes/redwood/WindowOverlayBaseTheme.css.js');

require('./UNSAFE_WindowOverlay/themes/redwood/WindowOverlayVariants.css.js');
require('./UNSAFE_Sheet/themes/SheetStyles.css.js');

require('./useAnimationStatus-8075dc19.js');
require('./useAnimation-65c0e455.js');
require('./animationUtils-54852d03.js');
require('./Dropdown-1ba432a8.js');
require('./keyboardUtils-b42fe7d5.js');
require('./usePopupAnimation-f3b34413.js');
require('./popupAnimationUtils-dcace427.js');
require('./CollectionInteractionContext-f84ff7a1.js');
require('./UNSAFE_Dropdown/themes/dropdownStyles.css.js');

require('./useOutsideMousedown-70f36cca.js');
require('./useEffectEvent-832ffba5.js');
require('./UNSAFE_Menu/themes/MenuStyles.css.js');

require('./UNSAFE_Menu/themes/DropdownMenuStyles.css.js');

require('./Flex-3a257bc3.js');
require('./boxalignment-e287d370.js');

require('./vanilla-extract-sprinkles-createRuntimeSprinkles.esm-d68f3e0f.js');
require('./flexbox-2cae9a01.js');

require('./flexitem-fee13e26.js');


require('./Skeleton-8e35c7d1.js');
require('./borders-4b8488cb.js');
require('./UNSAFE_Skeleton/themes/redwood/SkeletonTheme.js');
require('./UNSAFE_Skeleton/themes/SkeletonStyles.css.js');

require('./UNSAFE_Skeleton/themes/redwood/SkeletonBaseTheme.css.js');

require('./UNSAFE_Skeleton/themes/redwood/SkeletonVariants.css.js');
require('./UNSAFE_Menu/themes/MenuSkeletonStyles.css.js');

require('./LiveRegion-62ce9ac7.js');

require('./UNSAFE_Menu/themes/redwood/MenuItemTheme.js');
require('./UNSAFE_Menu/themes/MenuItemStyles.css.js');

require('./UNSAFE_Menu/themes/redwood/MenuItemBaseTheme.css.js');

require('./UNSAFE_Menu/themes/redwood/MenuItemVariants.css.js');

require('./useInteractionStyle-c203a8a0.js');


require('./hooks/UNSAFE_useTooltip/themes/redwood/TooltipContentVariants.css.js');

require('./tooltipUtils-2bc15136.js');
require('./EnvironmentProvider-5f2435e6.js');
require('./LayerManager-c535d963.js');

require('./UNSAFE_Separator/themes/SeparatorStyles.css.js');



require('./UNSAFE_Menu/themes/redwood/MenuSeparatorVariants.css.js');
require('./useContextMenuGesture-530ef8dd.js');
require('./textSelectionUtils-57082239.js');
require('./useChartDnd-edaa3619.js');
require('./utils-da07b35e.js');
require('./useDnd-a7d9b60c.js');
require('./useTooltip-34accecc.js');
require('./useTooltipControlled-0a95856c.js');
require('./useFocus-1b288fb9.js');
require('./useTouch-4dec8729.js');
require('./hooks/UNSAFE_useTooltip/themes/redwood/TooltipContentTheme.js');
require('./hooks/UNSAFE_useTooltip/themes/TooltipContentStyles.css.js');
require('./hooks/UNSAFE_useTooltip/themes/redwood/TooltipContentBaseTheme.css.js');
require('./useChartMarquee-ede5cae4.js');
require('./useVisDragModeControls-2443a4e0.js');
require('./Button-4d923bed.js');
require('./Icon-56789beb.js');
require('./UNSAFE_Icon/themes/IconStyle.css.js');
require('./Move-b1978994.js');
require('./hooks/PRIVATE_useVisDragModeControls/themes/useVisDragModeControls.css.js');

require('./ZoomAlt-cca2384d.js');
require('./ButtonSetSingle-bfe91a67.js');
require('./useButtonSetContext-a2f31c8e.js');
require('./ButtonSet-ed013090.js');
require('./UNSAFE_ButtonSet/themes/ButtonSetStyles.css.js');

require('./ButtonSetPositionContext-6a358667.js');
require('./toggleButtonUtils-245111a7.js');
require('./useRovingTabIndexContainer-0ce4681d.js');
require('./ButtonSetItem-4322c957.js');
require('./ButtonSetButton-569e0efb.js');
require('./ToggleButtonLabel-c8f36317.js');
require('./ButtonLabelLayout-c876923f.js');
require('./useButtonLabelLayoutTheme-d09db0c9.js');
require('./UNSAFE_ButtonLabelLayout/themes/redwood/ButtonLabelLayoutTheme.js');
require('./UNSAFE_ButtonLabelLayout/themes/ButtonLabelLayoutStyles.css.js');

require('./UNSAFE_ButtonLabelLayout/themes/redwood/ButtonLabelLayoutBaseTheme.css.js');

require('./UNSAFE_ButtonLabelLayout/themes/redwood/ButtonLabelLayoutVariants.css.js');

require('./HiddenAccessible-087eb49a.js');

require('./ButtonSetIconButton-3778ac7b.js');
require('./useButtonSetPositionContext-409e4cd6.js');
require('./IconToggleButton-d2147159.js');
require('./useToggleAction-e4f15550.js');
require('./hooks/PRIVATE_useChartMarquee/themes/useChartMarquee.css.js');

require('./useVisTouchResponse-376e11d5.js');
require('./useOverviewContext-1732a3f1.js');
require('./OverviewSlidingWindow-0f5f00a3.js');
require('./PRIVATE_VisOverview/themes/VisOverview.css.js');

require('./DragH-ac9f879f.js');
require('./ScrollBar-76e58249.js');
require('./PRIVATE_ChartScrollBar/themes/ScrollBarStyles.css.js');


const LINE_WITH_AREA_OPACITY = 0.2;
/**
 * The LineAreaChart displays information graphically using lines and filled areas, making relationships among the data easier to understand.
 */
const LineAreaChart = compat.forwardRef(({ width, height, selectedIds = [], hiddenIds = [], highlightedIds, hideAndShowBehavior = 'none', orientation = 'vertical', yAxis, y2Axis, series, groups, isStacked = false, getDataItem, selectionMode = 'none', drilling = 'off', type = 'line', overview = 'off', dataLabelEffect = 'outline', ...props }, ref) => {
    const { direction, forcedColors } = useUser.useUser();
    const isHighContrast = forcedColors === 'active';
    const isTestEnv = props.testId !== undefined;
    const isRtl = direction === 'rtl';
    const isLog = yAxis?.scale === 'log';
    const timeAxisType = props.xAxis?.timeAxisType;
    const hiddenSet = new Set(hiddenIds);
    const selectedSet = new Set(selectedIds);
    const highlightedSet = new Set(highlightedIds);
    const isHoriz = orientation === 'horizontal';
    const colors = colorUtils.getColorRamp();
    let renderedSeries = isStacked
        ? utils.getCumulativeSeries(series.length, groups.length, getDataItem, hiddenSet, isLog)
        : utils.getSeriesData(series?.length, groups?.length, getDataItem, hiddenSet, isLog);
    const renderedGetDataItem = (seriesIndex, groupIndex) => {
        if (!isStacked)
            return getDataItem(seriesIndex, groupIndex);
        return {
            ...getDataItem(seriesIndex, groupIndex),
            value: renderedSeries[seriesIndex][groupIndex]
        };
    };
    const getGapRatio = () => {
        return 1;
    };
    const getNextChartItemFunc = (startIndex, endIndex) => {
        // TODO: JET-71332
        //@ts-ignore
        return dataStyleUtils.getLineAreaChartNavUtil(getDataItem, series.length, startIndex, endIndex, isStacked, isRtl, isHoriz);
    };
    const getDataItemPos = (xScale, yScale, y2Scale) => {
        return (seriesIndex, groupIndex) => {
            const isAssociatedToY2 = series[seriesIndex].associatedYAxis === 'y2';
            return chartUtils.getLineAreaChartItemPosition(seriesIndex, groupIndex, renderedSeries, getDataItem, groups, isHoriz, xScale, isAssociatedToY2 ? y2Scale : yScale, timeAxisType);
        };
    };
    const getMarkerColorAndType = (seriesIndex, groupIndex) => {
        return {
            color: dataStyleUtils.getItemColor(seriesIndex, groupIndex, series, getDataItem),
            type: dataStyleUtils.getMarkerType(seriesIndex, groupIndex, getDataItem)
        };
    };
    const findNearest = (xScale, yScale, y2Scale) => {
        const getSeriesDataInfo = () => {
            return {
                offset: 0,
                dataWidth: 0
            };
        };
        const getYCoord = (value, yScale) => {
            return yScale.transform(value);
        };
        return utils.findNearestDataPoint(series, groups, xScale, yScale, renderedGetDataItem, isStacked, getSeriesDataInfo, getYCoord, isLog, 0, hiddenSet, timeAxisType, y2Scale);
    };
    const { hasCenteredSeries, hasUncenteredSeries } = series.reduce((acc, s) => {
        return {
            hasCenteredSeries: acc.hasCenteredSeries ||
                s.lineType === 'centeredSegmented' ||
                s.lineType === 'centeredStepped',
            hasUncenteredSeries: acc.hasUncenteredSeries || s.lineType === 'segmented' || s.lineType === 'stepped'
        };
    }, { hasCenteredSeries: false, hasUncenteredSeries: false });
    const getOffset = (groupCount) => {
        return utils.getStartAndEndAxesOffset(hasCenteredSeries, hasUncenteredSeries, isHoriz, groupCount);
    };
    const markers = utils.getMarkers();
    const areaOpacity = type === 'lineWithArea' ? LINE_WITH_AREA_OPACITY : 1;
    const isAreaRendered = type === 'area' || type === 'lineWithArea';
    const isContrastLineNeeded = type === 'area';
    const allLineSegmentCoords = [];
    return (jsxRuntime.jsx(Chart.Chart, { ref: ref, width: width, height: height, selectedIds: selectedSet, hiddenIds: hiddenIds, hideAndShowBehavior: hideAndShowBehavior, orientation: orientation, yAxis: yAxis, y2Axis: y2Axis, series: series, groups: groups, selectionMode: selectionMode, drilling: drilling, isStacked: isStacked, getDataItem: getDataItem, findNearest: findNearest, getDataItemPos: getDataItemPos, getMarkerInfo: getMarkerColorAndType, renderGridLinesInFront: isAreaRendered, getNavUtil: getNextChartItemFunc, getGapRatio: getGapRatio, getStartAndEndAxesOffset: getOffset, isRtl: isRtl, overview: overview, defaultOverviewContentRenderer: () => {
            return (jsxRuntime.jsx(OverviewLineAreaChart, { series: series, groups: groups, getDataItem: getDataItem, type: type }));
        }, ...props, children: ({ getItemAriaLabel, groupsInfo, xStartIndex, xEndIndex, xScale, yScale, y2Scale, activeId, focusedItemInfo, hoveredItemInfo, isPointInsideMarquee, defaultFontSize, getTextDimensions, plotAreaClipPathId, axisStepWidth }) => {
            const leafGroups = groupsInfo.map(({ item }) => item);
            renderedSeries = isStacked
                ? utils.getCumulativeSeries(series.length, leafGroups.length, getDataItem, hiddenSet, isLog)
                : utils.getSeriesData(series?.length, leafGroups?.length, getDataItem, hiddenSet, isLog);
            return (jsxRuntime.jsxs(jsxRuntime.Fragment, { children: [renderedSeries.map((lineAreaSeries, index) => {
                        if (lineAreaSeries.length === 0) {
                            /* TODO: need a better way to handle when a series is hidden. */
                            return;
                        }
                        const seriesItem = series[index];
                        const isAssociatedToY2 = seriesItem.associatedYAxis === 'y2';
                        const lineSegmentCoords = utils.getLineAreaSegments(isAssociatedToY2 ? y2Scale : yScale, xScale, lineAreaSeries, xStartIndex, xEndIndex, isLog, leafGroups, index, renderedGetDataItem, timeAxisType);
                        allLineSegmentCoords[index] = lineSegmentCoords[0];
                        const isBottomSegmentCurved = seriesItem.lineType === 'curved' &&
                            isStacked &&
                            renderedSeries.length > 1 &&
                            index > 0;
                        const isHighlighted = highlightedIds === undefined ||
                            highlightedIds.length === 0 ||
                            highlightedSet.has(seriesItem.id);
                        const prevCoords = isStacked && index != 0
                            ? allLineSegmentCoords.reduce((prevValue, newValue, j) => {
                                // find the index i such that i < index, and series i  has at least
                                // one or more data item unhidden. Only such series will can be
                                // used as bottomCoord for current series. Otherwise we have to use
                                // baseline coord
                                if (j >= index) {
                                    return prevValue;
                                }
                                if (newValue.length > 0) {
                                    return j;
                                }
                                return prevValue;
                            }, 0)
                            : undefined;
                        return (jsxRuntime.jsxs(jsxRuntime.Fragment, { children: [isAreaRendered && (jsxRuntime.jsx(ReferenceLineArea.AreaSeries, { yScale: isAssociatedToY2 ? y2Scale : yScale, lineType: seriesItem.lineType, color: seriesItem.areaColor || colors[index % colors.length], isHorizontal: isHoriz, areaColorOpacity: areaOpacity, isBottomSegmentCurved: isBottomSegmentCurved, bottomCoords: prevCoords !== undefined &&
                                        allLineSegmentCoords[prevCoords].some((value) => value != null)
                                        ? allLineSegmentCoords[prevCoords]
                                        : undefined, lineSegments: lineSegmentCoords, isLog: isLog, isHighlighted: isHighlighted, seriesIndex: index, groupWidth: axisStepWidth, isRtl: isRtl, timeAxisType: timeAxisType, plotAreaClipPathId: plotAreaClipPathId })), jsxRuntime.jsx(ReferenceLineArea.LineSeries, { lineType: seriesItem.lineType, 
                                    /* TODO: confirm with UX if this is contrastLine (changes depending on dark/light mode) or always white.
                                      It's always white in legacy.*/
                                    lineColor: isContrastLineNeeded
                                        ? themeContract_css.colorSchemeVars.dvt.contrastLine
                                        : seriesItem.lineColor, 
                                    /* The area generated in legacy are bit apart and not touching when stacked. Preact areas are a bit closer than legacy when curved.
                                    Increased the contrast linewidth (to 2.5 from legacy 1.25) to give legacy visual treatment.
                                    TODO: investivage why legacy areas are not touching. maybe we want to do the same in preact svg utils that
                                    generates this curve.
                                   */
                                    lineWidth: isContrastLineNeeded ? (isStacked ? 2.5 : 1.25) : seriesItem.lineWidth, lineStyle: seriesItem.lineStyle, color: seriesItem.lineColor || colors[index % colors.length], lineSegmentCoords: lineSegmentCoords, isHorizontal: isHoriz, isHighlighted: isHighlighted, seriesIndex: index, groupWidth: axisStepWidth, isRtl: isRtl, timeAxisType: timeAxisType, plotAreaClipPathId: plotAreaClipPathId }), 
                                /* For non stacked charts, markers should render in the order of series. Some markers might get hidden behind area but this preserves the legacy behavior.
                                JET-75357: Segmented and stepped lines may stack markers/lines incorrectly by series order, so excluded from non-stacked charts.
                               */
                                !isStacked &&
                                    seriesItem.lineType !== 'stepped' &&
                                    seriesItem.lineType !== 'segmented' && (jsxRuntime.jsx(MarkersWithLabel.MarkersWithLabel, { yScale: isAssociatedToY2 ? y2Scale : yScale, xScale: xScale, groups: leafGroups, series: seriesItem, timeAxisType: timeAxisType, color: seriesItem.lineColor ||
                                        seriesItem.areaColor ||
                                        colors[index % colors.length] /* TODO : make this use getItemColor function from dataStyleUtils. */, seriesIndex: index, startIndex: xStartIndex, endIndex: xEndIndex, markerType: markers[index % markers.length], orientation: orientation, isLog: isLog, activeId: activeId, getDataItem: getDataItem, isPointInsideMarquee: isPointInsideMarquee, selectedIds: selectedSet, highlightedIds: highlightedIds !== undefined ? highlightedSet : undefined, focusedItemIndex: focusedItemInfo.seriesIndex === index &&
                                        focusedItemInfo.isFocusVisible &&
                                        !(props.isDataCursorEnabled &&
                                            props.dataCursorStyle?.isMarkerDisplayed)
                                        ? focusedItemInfo.groupIndex
                                        : undefined, hoveredItemIndex: hoveredItemInfo?.isCurrent && hoveredItemInfo.seriesIndex === index
                                        ? hoveredItemInfo.groupIndex
                                        : undefined, isSelectionEnabled: selectionMode === 'single' || selectionMode === 'multiple', isDrillEnabled: drilling === 'on', defaultFontSize: defaultFontSize, dataLabelEffect: dataLabelEffect, isHighContrast: isHighContrast, isTestEnv: isTestEnv, getTextDimensions: getTextDimensions, getItemAriaLabel: getItemAriaLabel, isRtl: isRtl }))] }));
                    }), renderedSeries.map((lineAreaSeries, index) => {
                        if (lineAreaSeries.length === 0) {
                            /* TODO: need a better way to handle when a series is hidden.*/
                            return;
                        }
                        const seriesItem = series[index];
                        const isAssociatedToY2 = seriesItem.associatedYAxis === 'y2';
                        return isStacked ||
                            seriesItem.lineType === 'segmented' ||
                            seriesItem.lineType === 'stepped' ? (jsxRuntime.jsx(MarkersWithLabel.MarkersWithLabel, { yScale: isAssociatedToY2 ? y2Scale : yScale, series: series[index], xScale: xScale, groups: leafGroups, timeAxisType: timeAxisType, color: seriesItem.lineColor || seriesItem.areaColor || colors[index % colors.length], seriesIndex: index, isPointInsideMarquee: isPointInsideMarquee, startIndex: xStartIndex, endIndex: xEndIndex, markerType: markers[index % markers.length], orientation: orientation, isLog: isLog, activeId: activeId, getDataItem: renderedGetDataItem, selectedIds: selectedSet, focusedItemIndex: focusedItemInfo.seriesIndex === index && focusedItemInfo.isFocusVisible
                                ? focusedItemInfo.groupIndex
                                : undefined, hoveredItemIndex: hoveredItemInfo?.isCurrent && hoveredItemInfo.seriesIndex === index
                                ? hoveredItemInfo.groupIndex
                                : undefined, isSelectionEnabled: selectionMode === 'single' || selectionMode === 'multiple', isDrillEnabled: drilling === 'on', defaultFontSize: defaultFontSize, dataLabelEffect: dataLabelEffect, isHighContrast: isHighContrast, isTestEnv: isTestEnv, getTextDimensions: getTextDimensions, getItemAriaLabel: getItemAriaLabel, isRtl: isRtl })) : undefined;
                    })] }));
        } }));
});
function OverviewLineAreaChart(props) {
    return (jsxRuntime.jsx(LineAreaChart, { datatip: () => ({ content: '' }), zoomAndScroll: "off", overview: "off", plotArea: {
            yMajorTick: {
                isRendered: false
            }
        }, xAxis: {
            tickLabel: {
                autoRotate: false,
                // @ts-ignore
                isRendered: props.xAxis?.timeAxisType ? true : false
            }
        }, series: props.series, groups: props.groups, getDataItem: props.getDataItem, yAxis: {
            tickLabel: {
                isRendered: false
            }
        }, type: props.type }));
}

exports.LineAreaChart = LineAreaChart;
