// @oracle/oraclejet-preact/cjs/UNSAFE_SplitMenuButton.js
/* @oracle/oraclejet-preact: undefined */
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var jsxRuntime = require('preact/jsx-runtime');
var ButtonLabelLayout = require('./ButtonLabelLayout-c876923f.js');
var ChevronDown = require('./ChevronDown-55b2c0d4.js');
var SelectMenuGroupContext = require('./SelectMenuGroupContext-a349d259.js');
var hooks = require('preact/hooks');
require('preact');
var mergeProps = require('./mergeProps-e3da7237.js');
var useComponentTheme = require('./useComponentTheme-082fc8e4.js');
require('./LayerHost-3cd73603.js');
var compat = require('preact/compat');
var mergeInterpolations = require('./mergeInterpolations-6727b536.js');
var unsafeDomAccess = require('./unsafeDomAccess-c3a492f0.js');
var classNames = require('./classNames-c14c6ef3.js');


require('./hooks/UNSAFE_useTooltip/themes/redwood/TooltipContentVariants.css.js');
require('./tooltipUtils-2bc15136.js');

require('./UNSAFE_Menu/themes/MenuItemStyles.css.js');
require('./UNSAFE_Separator/themes/SeparatorStyles.css.js');


require('./UNSAFE_Menu/themes/redwood/MenuSeparatorVariants.css.js');

var Text = require('./Text-056f1ec5.js');
var BaseButton = require('./BaseButton-d8f96bfa.js');
var TabbableModeContext = require('./TabbableModeContext-e99d527e.js');
var dimensions = require('./dimensions-6101ca1f.js');
var useLongPress = require('./useLongPress-6de30c43.js');
var usePressClick = require('./usePressClick-7a34624a.js');
var UNSAFE_SplitMenuButton_themes_redwood_SplitMenuButtonTheme_css = require('./UNSAFE_SplitMenuButton/themes/redwood/SplitMenuButtonTheme.css.js');
require('./useButtonLabelLayoutTheme-d09db0c9.js');
require('./UNSAFE_ButtonLabelLayout/themes/redwood/ButtonLabelLayoutTheme.js');
require('./UNSAFE_ButtonLabelLayout/themes/ButtonLabelLayoutStyles.css.js');

require('./UNSAFE_ButtonLabelLayout/themes/redwood/ButtonLabelLayoutBaseTheme.css.js');

require('./UNSAFE_ButtonLabelLayout/themes/redwood/ButtonLabelLayoutVariants.css.js');

require('./vanilla-extract-recipes-createRuntimeFn.esm-d1301b2d.js');
require('./Icon-56789beb.js');
require('./size-1914356b.js');
require('./utils-c0efcd5e.js');
require('./themeContract.css-f3d431c6.js');
require('./colorUtils-01f3caa2.js');
require('./_curry1-e8f0d7ea.js');
require('./useTooltip-34accecc.js');
require('./useTooltipControlled-0a95856c.js');
require('./useId-6c0eeb27.js');
require('./useHover-49b0430d.js');
require('./useToggle-3ebba7d8.js');
require('./useFocus-1b288fb9.js');
require('./useTouch-4dec8729.js');
require('./useAnimation-65c0e455.js');
require('./useTestId-4f53d561.js');
require('./useThemeInterpolations-a37b37d9.js');
require('./useColorScheme-f8b1447d.js');
require('./useScale-db16c2b8.js');
require('./theme-dfd6a9fe.js');
require('./Theme-f06687af.js');
require('./useDensity-0b9fd9ca.js');
require('./hooks/UNSAFE_useTooltip/themes/redwood/TooltipContentTheme.js');
require('./hooks/UNSAFE_useTooltip/themes/TooltipContentStyles.css.js');
require('./hooks/UNSAFE_useTooltip/themes/redwood/TooltipContentBaseTheme.css.js');

require('./logger-2b636482.js');
require('./UNSAFE_Icon/themes/IconStyle.css.js');
require('./clientHints-c40c5250.js');
require('./Sheet-1d6e8cf8.js');
require('./Modal-b751b643.js');
require('./useModal-0739a528.js');
require('./Layer-b60deb1b.js');
require('./UNSAFE_Layer/themes/LayerStyles.css.js');

require('./UNSAFE_Modal/themes/ModalStyles.css.js');

require('./tabbableUtils-22a50472.js');
require('./head-68d0992f.js');
require('./_arity-c228159c.js');
require('./_curry3-18677bca.js');
require('./_curry2-c15d89cd.js');
require('./_isArray-73160ad5.js');
require('./_isString-f4443c9e.js');
require('./popupUtils-1d123cbe.js');
require('./WindowOverlay-85d999fa.js');
require('./vanilla-extract-dynamic.esm-2e1b7c25.js');
require('./UNSAFE_WindowOverlay/themes/WindowOverlayStyles.css.js');

require('./UNSAFE_WindowOverlay/themes/WindowOverlayContract.css.js');
require('./useUser-1295b5cb.js');
require('./UNSAFE_WindowOverlay/themes/redwood/WindowOverlayTheme.js');
require('./UNSAFE_WindowOverlay/themes/redwood/WindowOverlayBaseTheme.css.js');

require('./UNSAFE_WindowOverlay/themes/redwood/WindowOverlayVariants.css.js');
require('./UNSAFE_Sheet/themes/SheetStyles.css.js');

require('./refUtils-a9872e75.js');
require('./useAnimationStatus-8075dc19.js');
require('./animationUtils-54852d03.js');
require('./useTranslationBundle-12dd087d.js');
require('./Dropdown-1ba432a8.js');
require('./Floating-79cbe5ad.js');
require('./useFloating-c753a7f9.js');
require('./positionUtils-a780137e.js');
require('./useOutsideClick-dbf3d604.js');
require('./arrayUtils-7d8dcfc3.js');
require('./UNSAFE_Floating/themes/redwood/FloatingTheme.js');
require('./UNSAFE_Floating/themes/FloatingStyles.css.js');

require('./UNSAFE_Floating/themes/redwood/FloatingBaseTheme.css.js');

require('./UNSAFE_Floating/themes/redwood/FloatingVariants.css.js');
require('./UNSAFE_Floating/themes/FloatingContract.css.js');
require('./useUnsafeDomElementRef-f22a2f0c.js');
require('./keyboardUtils-b42fe7d5.js');
require('./usePopupAnimation-f3b34413.js');
require('./popupAnimationUtils-dcace427.js');
require('./CollectionInteractionContext-f84ff7a1.js');
require('./UNSAFE_Dropdown/themes/dropdownStyles.css.js');

require('./useOutsideMousedown-70f36cca.js');
require('./useEffectEvent-832ffba5.js');
require('./UNSAFE_Menu/themes/MenuStyles.css.js');
require('./UNSAFE_Menu/themes/DropdownMenuStyles.css.js');

require('./Flex-3a257bc3.js');
require('./boxalignment-e287d370.js');

require('./vanilla-extract-sprinkles-createRuntimeSprinkles.esm-d68f3e0f.js');
require('./flexbox-2cae9a01.js');

require('./flexitem-fee13e26.js');


require('./Skeleton-8e35c7d1.js');
require('./borders-4b8488cb.js');
require('./UNSAFE_Skeleton/themes/redwood/SkeletonTheme.js');
require('./UNSAFE_Skeleton/themes/SkeletonStyles.css.js');

require('./UNSAFE_Skeleton/themes/redwood/SkeletonBaseTheme.css.js');

require('./UNSAFE_Skeleton/themes/redwood/SkeletonVariants.css.js');
require('./UNSAFE_Menu/themes/MenuSkeletonStyles.css.js');

require('./LiveRegion-62ce9ac7.js');

require('./UNSAFE_Menu/themes/redwood/MenuItemTheme.js');
require('./UNSAFE_Menu/themes/redwood/MenuItemBaseTheme.css.js');

require('./UNSAFE_Menu/themes/redwood/MenuItemVariants.css.js');


require('./usePress-886180e4.js');
require('./useInteractionStyle-c203a8a0.js');
require('./useActive-6770f917.js');
require('./index-0aa99b92.js');
require('./UNSAFE_Layer/themes/LayerHostStyles.css.js');

require('./mergeDeepWithKey-210b024d.js');
require('./_isObject-28636267.js');
require('./EnvironmentProvider-5f2435e6.js');
require('./LayerManager-c535d963.js');

require('./UNSAFE_Text/themes/TextStyles.css.js');

require('./UNSAFE_BaseButton/themes/redwood/BaseButtonTheme.js');
require('./UNSAFE_BaseButton/themes/BaseButtonStyles.css.js');

require('./UNSAFE_BaseButton/themes/redwood/BaseButtonVariants.css.js');

require('./BareButton-e8a030d7.js');
require('./useBareButton-463210c9.js');
require('./useTabbableMode-dbd254e8.js');
require('./buttonUtils-dff2ea3e.js');
require('./id-86356250.js');



const menuButtonStyling = [
    'embedded',
    'min',
    'noBorderRadiusStart',
    'noBorderStart'
];
const menuButtonActiveStyling = [...menuButtonStyling, 'active'];
const actionbuttonStyling = [
    'embedded',
    'fill',
    'noBorderRadiusEnd',
    'noBorderEnd'
];
const actionButtonActiveStyling = [...actionbuttonStyling, 'active'];
const interpolations = [...Object.values(dimensions.dimensionInterpolations)];
const styleInterpolations = mergeInterpolations.mergeInterpolations(interpolations);
/**
 * A SplitMenuButton is a combined action and menu button.
 */
const SplitMenuButton = compat.forwardRef(({ children, label = '', variant = 'outlined', isDisabled = false, size = 'md', testId, onAction, 'aria-describedby': ariaDescribedBy, onBlur: propsOnBlur, onFocus, onMouseEnter, onMouseLeave, onTouchEnd, onTouchStart, ...props }, ref) => {
    const rootButtonRef = hooks.useRef(null);
    const actionButtonRef = hooks.useRef(null);
    const [initialFocus, setInitialFocus] = hooks.useState('menu');
    hooks.useImperativeHandle(ref, () => {
        return {
            focus: rootButtonRef.current.focus,
            blur: rootButtonRef.current.blur,
            [unsafeDomAccess.UNSAFE_DOM_ACCESS]: rootButtonRef.current[unsafeDomAccess.UNSAFE_DOM_ACCESS],
            // Need to delegate click to the inner action button since
            // clicking on the outer button does not trigger an action
            click: actionButtonRef.current.click
        };
    }, []);
    const [isMenuOpen, setIsMenuOpen] = hooks.useState(false);
    const [isActive, setIsActive] = hooks.useState(false);
    const { class: cls, ...styles } = styleInterpolations(props);
    // SPACE or ENTER on root area invokes action, as does CLICK/TAP on LEFT SIDE ACTION AREA
    // DOWN ARROW invokes menu, as does CLICK/TAP on RIGHT SIDE MENU AREA
    const onKeyDown = hooks.useCallback((event) => {
        if (event.key === ' ' || event.key === 'Enter') {
            setIsActive(true);
        }
        else if (event.key == 'ArrowDown') {
            setInitialFocus('firstItem');
            !isDisabled && setIsMenuOpen((menuOpen) => !menuOpen);
            // Prevent page scroll for down arrow, user definitely wanted to show menu in this case.
            // Don't prevent other keypress defaults, as user may want to hold SPACE down to scroll page.
            event.preventDefault();
        }
    }, [isDisabled, setIsMenuOpen, setIsActive, setInitialFocus]);
    const onKeyUp = hooks.useCallback((event) => {
        if (event.key === ' ' || event.key === 'Enter') {
            !isDisabled && onAction && onAction({ reason: 'keyboard' });
            setIsActive(false);
        }
    }, [isDisabled, onAction, setIsActive]);
    const { pressProps } = usePressClick.usePressClick(() => {
        setInitialFocus('firstItem');
        setIsMenuOpen((menuOpen) => !menuOpen);
    }, { isDisabled });
    const { longPressProps } = useLongPress.useLongPress(() => {
        setInitialFocus('firstItem');
        setIsMenuOpen((menuOpen) => !menuOpen);
    }, { isDisabled });
    const onBlur = hooks.useCallback(() => {
        setIsActive(false);
    }, [setIsActive]);
    const menuHandler = () => {
        setInitialFocus('menu');
        setIsMenuOpen(!isMenuOpen);
    };
    const handleClose = (e) => {
        //Focus has to be set on the launcher when menu is dismissed or
        //item is "selected"
        if (e.reason === 'dismissed' || e.reason === 'itemAction') {
            rootButtonRef.current?.focus();
        }
        setIsMenuOpen(false);
    };
    const { classes, styles: themeStyles } = useComponentTheme.useComponentTheme(UNSAFE_SplitMenuButton_themes_redwood_SplitMenuButtonTheme_css.SplitMenuButtonRedwoodTheme, {
        styleVariant: variant
    });
    const dividerStyles = classNames.classNames([
        themeStyles.divider,
        isDisabled && themeStyles.disabledDivider
    ]);
    const globalEventProps = {
        onBlur: propsOnBlur,
        onFocus,
        onMouseEnter,
        onMouseLeave,
        onTouchEnd,
        onTouchStart
    };
    const customGlobalEventProps = {
        ...mergeProps.mergeProps({ onBlur: onBlur }, longPressProps, pressProps, { onKeyDown: isDisabled ? undefined : onKeyDown }, { onKeyUp: onKeyUp })
    };
    const rootStyles = classNames.classNames([classes, cls]);
    // Using Span for top button allows us to avoid :hover state rendering by
    // taking advantage of :enabled not on spans
    const rootButton = () => {
        return (jsxRuntime.jsx(BaseButton.BaseButton, { elementDetails: { type: 'span' }, ref: rootButtonRef, isDisabled: isDisabled, styling: ['container', 'min'], variant: variant, size: size, testId: testId, "aria-label": label, "aria-roledescription": "split menu button, press down arrow to invoke menu", "aria-describedby": ariaDescribedBy, "aria-expanded": isMenuOpen, children: jsxRuntime.jsxs(TabbableModeContext.TabbableModeContext.Provider, { value: { isTabbable: false }, children: [jsxRuntime.jsx(BaseButton.BaseButton, { elementDetails: { type: 'span', isFocusable: false }, ref: actionButtonRef, variant: variant, styling: getActionButtonStyling(isActive), size: size, isDisabled: isDisabled, "aria-hidden": true, onAction: onAction, children: jsxRuntime.jsx(Text.Text, { truncation: 'ellipsis', weight: 'inherit', children: label }) }), jsxRuntime.jsx("div", { class: `${dividerStyles}` }), jsxRuntime.jsx(BaseButton.BaseButton, { elementDetails: { type: 'span', isFocusable: false }, variant: variant, styling: getMenuButtonStyling(isMenuOpen), size: size, isDisabled: isDisabled, "aria-hidden": true, onAction: menuHandler, children: jsxRuntime.jsx(ButtonLabelLayout.ButtonLabelLayout, { display: "icons", startIcon: jsxRuntime.jsx(ChevronDown.SvgChevronDown, {}), size: size, styling: "embedded" }) })] }) }));
    };
    if (isDisabled) {
        return (jsxRuntime.jsx("span", { role: "toolbar", class: rootStyles, style: styles, children: rootButton() }));
    }
    else {
        /*TODO: Remove this hardcoded offsetValue so we have dropdown offset only set on one single place.
          Could be done via menu props or having a dropdownMenu component. JET-60595*/
        /*TODO: Once menu expose a prop to set min width we would need to calculate offset width of button
         so we set the correct min width since dropdown menus are 200px but if triggered element width is
         bigger should be that size. JET-60745 */
        return (jsxRuntime.jsxs("span", { role: "toolbar", class: rootStyles, ...mergeProps.mergeProps(globalEventProps, customGlobalEventProps), style: styles, children: [rootButton(), jsxRuntime.jsx(SelectMenuGroupContext.Menu, { anchorRef: rootButtonRef, isOpen: isMenuOpen, onClose: handleClose, initialFocus: initialFocus, children: children })] }));
    }
});
function getMenuButtonStyling(isActive) {
    return isActive ? menuButtonActiveStyling : menuButtonStyling;
}
function getActionButtonStyling(isActive) {
    return isActive ? actionButtonActiveStyling : actionbuttonStyling;
}
SplitMenuButton.displayName = 'SplitMenuButton';

exports.SplitMenuButton = SplitMenuButton;
