// @oracle/oraclejet-preact/cjs/UNSAFE_ExpandableList.js
/* @oracle/oraclejet-preact: undefined */
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var jsxRuntime = require('preact/jsx-runtime');
var preact = require('preact');
var hooks = require('preact/hooks');
var compat = require('preact/compat');
var classNames = require('./classNames-c14c6ef3.js');
var useId = require('./useId-6c0eeb27.js');
var useTranslationBundle = require('./useTranslationBundle-12dd087d.js');
var keys = require('./keys-25710458.js');
var Button = require('./Button-020e0381.js');
var HiddenAccessible = require('./HiddenAccessible-087eb49a.js');
var ChevronDown = require('./ChevronDown-55b2c0d4.js');
var CollapseIcon = require('./CollapseIcon-9fbe13a9.js');
var LiveRegion = require('./LiveRegion-62ce9ac7.js');
var useUser = require('./useUser-1295b5cb.js');
var collectionUtils = require('./collectionUtils-f1d3183f.js');
var GroupLoadingIndicator = require('./GroupLoadingIndicator-7e035ba2.js');
var useNavigation = require('./useNavigation-fa3f5a50.js');
var List = require('./List-afcfa338.js');
require('./LayerHost-3cd73603.js');
require('./index-0aa99b92.js');
require('./UNSAFE_Layer/themes/LayerHostStyles.css.js');

require('./BaseButton-d8f96bfa.js');
require('./useHover-49b0430d.js');
require('./useToggle-3ebba7d8.js');
require('./useActive-6770f917.js');
require('./useColorScheme-f8b1447d.js');
require('./dimensions-6101ca1f.js');
require('./size-1914356b.js');
require('./utils-c0efcd5e.js');
require('./themeContract.css-f3d431c6.js');
require('./colorUtils-01f3caa2.js');
require('./_curry1-e8f0d7ea.js');
require('./mergeInterpolations-6727b536.js');
require('./mergeDeepWithKey-210b024d.js');
require('./_curry3-18677bca.js');
require('./_curry2-c15d89cd.js');
require('./_isObject-28636267.js');
require('./clientHints-c40c5250.js');
require('./mergeProps-e3da7237.js');
require('./UNSAFE_BaseButton/themes/redwood/BaseButtonTheme.js');
require('./UNSAFE_BaseButton/themes/BaseButtonStyles.css.js');

require('./UNSAFE_BaseButton/themes/redwood/BaseButtonVariants.css.js');
require('./vanilla-extract-recipes-createRuntimeFn.esm-d1301b2d.js');

require('./useComponentTheme-082fc8e4.js');
require('./logger-2b636482.js');
require('./BareButton-e8a030d7.js');
require('./useBareButton-463210c9.js');
require('./usePress-886180e4.js');
require('./usePressClick-7a34624a.js');
require('./TabbableModeContext-e99d527e.js');
require('./useTabbableMode-dbd254e8.js');
require('./buttonUtils-dff2ea3e.js');
require('./id-86356250.js');
require('./useTestId-4f53d561.js');
require('./unsafeDomAccess-c3a492f0.js');
require('./ButtonLabelLayout-c876923f.js');
require('./Text-056f1ec5.js');
require('./UNSAFE_Text/themes/TextStyles.css.js');

require('./useButtonLabelLayoutTheme-d09db0c9.js');
require('./UNSAFE_ButtonLabelLayout/themes/redwood/ButtonLabelLayoutTheme.js');
require('./UNSAFE_ButtonLabelLayout/themes/ButtonLabelLayoutStyles.css.js');

require('./UNSAFE_ButtonLabelLayout/themes/redwood/ButtonLabelLayoutBaseTheme.css.js');

require('./UNSAFE_ButtonLabelLayout/themes/redwood/ButtonLabelLayoutVariants.css.js');


require('./Icon-56789beb.js');
require('./useTooltip-34accecc.js');
require('./useTooltipControlled-0a95856c.js');
require('./useFocus-1b288fb9.js');
require('./useTouch-4dec8729.js');
require('./useAnimation-65c0e455.js');
require('./useThemeInterpolations-a37b37d9.js');
require('./useScale-db16c2b8.js');
require('./theme-dfd6a9fe.js');
require('./Theme-f06687af.js');
require('./useDensity-0b9fd9ca.js');
require('./hooks/UNSAFE_useTooltip/themes/redwood/TooltipContentTheme.js');
require('./hooks/UNSAFE_useTooltip/themes/TooltipContentStyles.css.js');

require('./hooks/UNSAFE_useTooltip/themes/redwood/TooltipContentBaseTheme.css.js');

require('./hooks/UNSAFE_useTooltip/themes/redwood/TooltipContentVariants.css.js');

require('./tooltipUtils-2bc15136.js');
require('./EnvironmentProvider-5f2435e6.js');
require('./LayerManager-c535d963.js');
require('./Floating-79cbe5ad.js');
require('./useFloating-c753a7f9.js');
require('./positionUtils-a780137e.js');
require('./refUtils-a9872e75.js');
require('./useOutsideClick-dbf3d604.js');
require('./arrayUtils-7d8dcfc3.js');
require('./Layer-b60deb1b.js');
require('./UNSAFE_Layer/themes/LayerStyles.css.js');

require('./useModal-0739a528.js');
require('./UNSAFE_Floating/themes/redwood/FloatingTheme.js');
require('./UNSAFE_Floating/themes/FloatingStyles.css.js');

require('./UNSAFE_Floating/themes/redwood/FloatingBaseTheme.css.js');

require('./UNSAFE_Floating/themes/redwood/FloatingVariants.css.js');
require('./vanilla-extract-dynamic.esm-2e1b7c25.js');
require('./UNSAFE_Floating/themes/FloatingContract.css.js');
require('./useUnsafeDomElementRef-f22a2f0c.js');
require('./UNSAFE_Icon/themes/IconStyle.css.js');

require('./ChevronRight-d1c46e18.js');
require('./ChevronLeft-2dcc3b2e.js');
require('./withDirectionIcon-4940a248.js');

require('./Selector-6550b8e0.js');
require('./BaseSelector-2dc69917.js');
require('./StyledCheckbox-9047ed7d.js');
require('./Check-1272bb86.js');
require('./CheckboxOff-a3ed334d.js');
require('./CheckboxOn-b5dda28b.js');
require('./CheckboxMixed-a17c8298.js');
require('./UNSAFE_Checkbox/themes/CheckboxIconContract.css.js');
require('./UNSAFE_Checkbox/themes/redwood/CheckboxIconTheme.js');
require('./UNSAFE_Checkbox/themes/CheckboxIconStyles.css.js');

require('./UNSAFE_Checkbox/themes/redwood/CheckboxIconBaseTheme.css.js');

require('./UNSAFE_Checkbox/themes/redwood/CheckboxIconVariants.css.js');


require('./PRIVATE_BaseSelector/themes/SelectorStyles.css.js');

require('./CollectionInteractionContext-f84ff7a1.js');
require('./useCollectionInteractionContext-6fc0200e.js');
require('./useInteractionStyle-c203a8a0.js');
require('./flexitem-fee13e26.js');

require('./vanilla-extract-sprinkles-createRuntimeSprinkles.esm-d68f3e0f.js');
require('./PRIVATE_List/themes/ListStyles.css.js');


require('./UNSAFE_GroupedList/themes/redwood/GroupedListTheme.js');
require('./UNSAFE_GroupedList/themes/GroupedListStyles.css.js');

require('./UNSAFE_GroupedList/themes/redwood/GroupedListBaseTheme.css.js');

require('./Flex-3a257bc3.js');
require('./boxalignment-e287d370.js');

require('./flexbox-2cae9a01.js');


require('./Skeleton-8e35c7d1.js');
require('./borders-4b8488cb.js');
require('./UNSAFE_Skeleton/themes/redwood/SkeletonTheme.js');
require('./UNSAFE_Skeleton/themes/SkeletonStyles.css.js');

require('./UNSAFE_Skeleton/themes/redwood/SkeletonBaseTheme.css.js');

require('./UNSAFE_Skeleton/themes/redwood/SkeletonVariants.css.js');
require('./useCurrentKey-1b9e3382.js');
require('./useReorderableContext-92514102.js');
require('./ReorderableContext-16396e17.js');
require('./useCollectionFocusRing-bed0aa16.js');
require('./useTabbableModeSet-5bce852a.js');
require('./LoadMoreCollection-eb1f2603.js');
require('./Collection-e27077ff.js');
require('./useViewportIntersect-2e802bac.js');
require('./VirtualizedCollection-a90800c7.js');
require('./FocusTrap-4873a3f2.js');
require('./tabbableUtils-22a50472.js');
require('./head-68d0992f.js');
require('./_arity-c228159c.js');
require('./_isArray-73160ad5.js');
require('./_isString-f4443c9e.js');
require('./FocusTracker-e2ea43c3.js');
require('./PRIVATE_FocusTracker/themes/FocusTrackerStyles.css.js');

require('./ImageVars.css-3fbb1c0b.js');
require('./useTheme-2a034a21.js');
require('./useSelection-37939f09.js');
require('./textSelectionUtils-57082239.js');
require('./useAdditiveSelection-bf845689.js');
require('./useItemAction-66c2387c.js');
require('./Inset-69f359cd.js');
require('./UNSAFE_Inset/themes/InsetStyles.css.js');



/**
 * @license
 * Copyright (c) %FIRST_YEAR% %CURRENT_YEAR%, Oracle and/or its affiliates.
 * The Universal Permissive License (UPL), Version 1.0
 * as shown at https://oss.oracle.com/licenses/upl/
 * @ignore
 */
/**
 * A hook that handles mouse and keyboard gesture that toggles the expanded state.
 * @param keyExtractor
 * @param expanded
 * @param onToggle
 * @returns
 */
function useExpandCollapse(keyExtractor, expanded, onToggle) {
    const { direction } = useUser.useUser();
    const onClick = hooks.useCallback((event) => {
        handleToggle(event, keyExtractor, onToggle);
    }, [keyExtractor, onToggle]);
    const onKeyUp = hooks.useCallback((event) => {
        if (event.key === 'Enter' || event.key === ' ') {
            handleToggle(event, keyExtractor, onToggle);
        }
        else if (event.key === 'ArrowLeft' || event.key === 'ArrowRight') {
            const isRtl = direction === 'rtl';
            const isLeftArrow = event.key === 'ArrowLeft' && !isRtl;
            handleArrowKey(isLeftArrow, event.target, expanded, keyExtractor, onToggle);
            // arrow key might cause it to scroll horizontally
            event.preventDefault();
            event.stopPropagation();
        }
    }, [keyExtractor, onToggle, expanded, direction]);
    return onToggle ? { onClick, onKeyDown, onKeyUp } : {};
}
// prevent default (propagation) for keyDown to prevent scrolling the container
const onKeyDown = (event) => {
    if (event.key === 'Enter' ||
        event.key === ' ' ||
        event.key === 'ArrowLeft' ||
        event.key === 'ArrowRight') {
        event.preventDefault();
    }
};
const handleArrowKey = (isLeftArrow, target, expanded, keyExtractor, onToggle) => {
    const itemKey = keyExtractor(target);
    if (onToggle && itemKey) {
        const contains = keys.containsKey(expanded, itemKey);
        if ((isLeftArrow && contains) || (!isLeftArrow && !contains)) {
            onToggle({ value: itemKey });
        }
    }
};
const handleToggle = (event, keyExtractor, onToggle) => {
    const itemKey = keyExtractor(event.target);
    if (onToggle && itemKey) {
        onToggle({ value: itemKey });
        if (event.type === 'click') {
            event.stopPropagation();
        }
    }
};

/**
 * An ExpandableList shows a two-level group of items where each group header can be expand or collapse.
 */
function ExpandableList({ children, data, onSelectionChange, onToggle, onPersistCurrentItem, ...props }) {
    const [statusText, setStatusText] = hooks.useState('');
    const translations = useTranslationBundle.useTranslationBundle('@oracle/oraclejet-preact');
    const currentKey = hooks.useRef();
    const handleOnPersistCurrentItem = hooks.useCallback((detail) => {
        currentKey.current = detail.value;
        onPersistCurrentItem?.(detail);
    }, [onPersistCurrentItem]);
    const updateStateText = hooks.useCallback((detail) => {
        onToggle?.(detail);
        if (data) {
            const previouslyExpanded = keys.containsKey(data?.expanded, detail.value);
            setStatusText(previouslyExpanded
                ? translations.expandableList_groupCollapse()
                : translations.expandableList_groupExpand());
            // we need to clear the live region text otherwise screenreader
            // won't read again if the content did not change the next time
            setTimeout(() => {
                setStatusText('');
            }, 1000);
        }
    }, [onToggle, data, translations]);
    const expandCollapseProps = useExpandCollapse((element) => {
        if (element.getAttribute('role') === 'treegrid') {
            return currentKey.current === undefined ? null : currentKey.current;
        }
        return collectionUtils.keyExtractor(element, GroupLoadingIndicator.GROUP_SELECTOR);
    }, data == null ? emptyKeys : data.expanded, onToggle && updateStateText);
    const descId = useId.useId();
    const stableDescIdRef = hooks.useRef(descId);
    const createHierarchicalContext = (context, options) => {
        const expanderRenderer = options !== undefined
            ? () => (jsxRuntime.jsx(Expander, { itemKey: context.metadata.key, expanded: options.expanded, onToggle: options.onToggle }))
            : undefined;
        return {
            ...context,
            parentKey: context.metadata.parentKey,
            leaf: context.metadata.isLeaf ?? true,
            depth: context.metadata.treeDepth ?? 1,
            expander: expanderRenderer
        };
    };
    const handleSelectionChange = hooks.useCallback((detail) => {
        if (onSelectionChange && data) {
            const keys$1 = useNavigation.excludeGroup(data, detail.value);
            if (keys.isSameKey(keys$1, detail.value)) {
                onSelectionChange(detail);
            }
            else if (keys$1.keys && keys$1.keys.size > 0) {
                onSelectionChange({ ...detail, value: keys$1 });
            }
        }
    }, [onSelectionChange, data]);
    const customItemRenderer = (context) => {
        const itemKey = context.listItemContext.metadata.key;
        const options = data && onToggle ? { expanded: keys.containsKey(data.expanded, itemKey), onToggle } : undefined;
        const itemContext = createHierarchicalContext(context.listItemContext, options);
        if (itemContext.leaf) {
            return context.defaultListItem(itemContext, { itemDepth: itemContext.depth + 1 });
        }
        else {
            return (jsxRuntime.jsx(GroupLoadingIndicator.ListGroupHeader, { ariaDescribedBy: stableDescIdRef.current, itemKey: itemContext.metadata.key, itemIndex: itemContext.index, itemDepth: itemContext.depth, isFocused: context.isFocused, isFocusRingVisible: context.isFocusRingVisible, isActive: context.isActive, isGridlineVisible: false, isExpandable: true, expandedKeys: data == null ? emptyKeys : data.expanded, onToggle: onToggle, children: children(itemContext) }, itemContext.metadata.key));
        }
    };
    return (jsxRuntime.jsxs(preact.Fragment, { children: [jsxRuntime.jsx(List.List, { data: data, role: "treegrid", loadingIndicator: jsxRuntime.jsx(GroupLoadingIndicator.GroupLoadingIndicator, {}), onSelectionChange: handleSelectionChange, customItemRenderer: customItemRenderer, onPersistCurrentItem: handleOnPersistCurrentItem, ...props, ...expandCollapseProps, children: (context) => {
                    return children(context);
                } }), onToggle && (jsxRuntime.jsxs(preact.Fragment, { children: [jsxRuntime.jsx(LiveRegion.LiveRegion, { children: statusText }), jsxRuntime.jsx("span", { id: stableDescIdRef.current, children: jsxRuntime.jsx(HiddenAccessible.HiddenAccessible, { children: translations.expandableList_expandCollapseInstructionText() }) })] }))] }));
}
const emptyKeys = { all: false, keys: new Set() };
const Expander = compat.memo((props) => {
    const classes = classNames.classNames(['oj-listview-expander']);
    const actionHandler = () => {
        props.onToggle({ value: props.itemKey });
    };
    return (jsxRuntime.jsx("span", { class: classes, role: "presentation", children: jsxRuntime.jsx(Button.Button, { onAction: actionHandler, variant: "ghost", startIcon: props.expanded ? jsxRuntime.jsx(ChevronDown.SvgChevronDown, { size: "6x" }) : jsxRuntime.jsx(CollapseIcon.CollapseIcon, { size: "6x" }) }) }));
}, (prev, next) => {
    return (prev.itemKey === next.itemKey &&
        prev.expanded === next.expanded &&
        prev.onToggle === next.onToggle);
});

exports.ExpandableList = ExpandableList;
