// @oracle/oraclejet-preact/cjs/UNSAFE_Train.js
/* @oracle/oraclejet-preact: undefined */
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var jsxRuntime = require('preact/jsx-runtime');
var hooks = require('preact/hooks');
var preact = require('preact');
var useComponentTheme = require('./useComponentTheme-082fc8e4.js');
var useTestId = require('./useTestId-4f53d561.js');
var UNSAFE_Train_themes_redwood_TrainTheme = require('./UNSAFE_Train/themes/redwood/TrainTheme.js');
var usePress = require('./usePress-886180e4.js');
var useInteractionStyle = require('./useInteractionStyle-c203a8a0.js');
var classNames = require('./classNames-c14c6ef3.js');
var mergeProps = require('./mergeProps-e3da7237.js');
var SuccessS = require('./SuccessS-b6757c64.js');
var ErrorS = require('./ErrorS-b027f7ad.js');
var InformationS = require('./InformationS-b0bb102b.js');
var WarningS = require('./WarningS-cdb0a7ad.js');
var useActive = require('./useActive-6770f917.js');
var useTranslationBundle = require('./useTranslationBundle-12dd087d.js');
var HiddenAccessible = require('./HiddenAccessible-087eb49a.js');
require('./logger-2b636482.js');
require('./LayerHost-3cd73603.js');
require('./index-0aa99b92.js');
require('preact/compat');
require('./UNSAFE_Layer/themes/LayerHostStyles.css.js');

require('./UNSAFE_Train/themes/TrainStyles.css.js');

require('./UNSAFE_Train/themes/redwood/TrainBaseTheme.css.js');

require('./UNSAFE_Train/themes/redwood/TrainVariants.css.js');

require('./vanilla-extract-recipes-createRuntimeFn.esm-d1301b2d.js');
require('./clientHints-c40c5250.js');
require('./useHover-49b0430d.js');
require('./useToggle-3ebba7d8.js');
require('./Icon-56789beb.js');
require('./size-1914356b.js');
require('./utils-c0efcd5e.js');
require('./themeContract.css-f3d431c6.js');
require('./colorUtils-01f3caa2.js');
require('./_curry1-e8f0d7ea.js');
require('./useTooltip-34accecc.js');
require('./useTooltipControlled-0a95856c.js');
require('./useId-6c0eeb27.js');
require('./useFocus-1b288fb9.js');
require('./useTouch-4dec8729.js');
require('./useAnimation-65c0e455.js');
require('./useThemeInterpolations-a37b37d9.js');
require('./useColorScheme-f8b1447d.js');
require('./useScale-db16c2b8.js');
require('./theme-dfd6a9fe.js');
require('./Theme-f06687af.js');
require('./mergeInterpolations-6727b536.js');
require('./mergeDeepWithKey-210b024d.js');
require('./_curry3-18677bca.js');
require('./_curry2-c15d89cd.js');
require('./_isObject-28636267.js');
require('./useDensity-0b9fd9ca.js');
require('./unsafeDomAccess-c3a492f0.js');
require('./hooks/UNSAFE_useTooltip/themes/redwood/TooltipContentTheme.js');
require('./hooks/UNSAFE_useTooltip/themes/TooltipContentStyles.css.js');

require('./hooks/UNSAFE_useTooltip/themes/redwood/TooltipContentBaseTheme.css.js');

require('./hooks/UNSAFE_useTooltip/themes/redwood/TooltipContentVariants.css.js');

require('./tooltipUtils-2bc15136.js');
require('./EnvironmentProvider-5f2435e6.js');
require('./LayerManager-c535d963.js');
require('./Floating-79cbe5ad.js');
require('./useFloating-c753a7f9.js');
require('./useUser-1295b5cb.js');
require('./positionUtils-a780137e.js');
require('./refUtils-a9872e75.js');
require('./useOutsideClick-dbf3d604.js');
require('./arrayUtils-7d8dcfc3.js');
require('./Layer-b60deb1b.js');
require('./UNSAFE_Layer/themes/LayerStyles.css.js');

require('./useModal-0739a528.js');
require('./UNSAFE_Floating/themes/redwood/FloatingTheme.js');
require('./UNSAFE_Floating/themes/FloatingStyles.css.js');

require('./UNSAFE_Floating/themes/redwood/FloatingBaseTheme.css.js');

require('./UNSAFE_Floating/themes/redwood/FloatingVariants.css.js');
require('./vanilla-extract-dynamic.esm-2e1b7c25.js');
require('./UNSAFE_Floating/themes/FloatingContract.css.js');
require('./useUnsafeDomElementRef-f22a2f0c.js');
require('./UNSAFE_Icon/themes/IconStyle.css.js');



const Context = preact.createContext(null);

/**
The JET Train component serves as a visual navigator enabling users to traverse through different 'steps'. Each 'step' is represented by the Step sub-component which can manifest its own state - 'visited', 'unvisited', or 'disabled', and is capable of communicating various message types including 'error', 'confirmation', 'warning', or 'info'.
 */
function Train({ onSelect, selectedStep, children, testId }) {
    const { classes, styles } = useComponentTheme.useComponentTheme(UNSAFE_Train_themes_redwood_TrainTheme.TrainRedwoodTheme);
    const testIdProps = useTestId.useTestId(testId);
    const selectedStepIndex = children.findIndex((step) => step.props.id === selectedStep);
    const onStepSelect = hooks.useCallback((event, stepId) => {
        onSelect?.({
            event: event,
            toStep: stepId
        });
    }, [onSelect]);
    const trainContext = {
        onSelect: onStepSelect,
        selectedStep: selectedStep,
        selectedStepIndex: selectedStepIndex,
        numberOfSteps: children.length,
        index: 0
    };
    const childrenArray = preact.toChildArray(children).filter(Boolean);
    return (jsxRuntime.jsx("div", { class: classes, ...testIdProps, children: childrenArray.map((step, index) => (jsxRuntime.jsx(Context.Provider, { value: { ...trainContext, index }, children: jsxRuntime.jsx("div", { class: styles.baseListStyle, children: step }) }))) }));
}

const Connector = ({ isConnected }) => {
    const { styles, variantClasses } = useComponentTheme.useComponentTheme(UNSAFE_Train_themes_redwood_TrainTheme.TrainRedwoodTheme, {
        connectorConnected: isConnected ? 'isConnected' : 'notConnected'
    });
    return jsxRuntime.jsx("span", { class: classNames.classNames([styles.connectorBaseStyle, variantClasses]) });
};

const messageTypeIcons = {
    confirmation: SuccessS.SvgSuccessS,
    error: ErrorS.SvgErrorS,
    fatal: ErrorS.SvgErrorS,
    info: InformationS.SvgInformationS,
    warning: WarningS.SvgWarningS
};
/**
 * UserChosenIcon Component for rendering the severity based icon in Message
 */
function UserChosenIcon({ messageType, isDisabled = false }) {
    const IconComponent = messageTypeIcons[messageType];
    const { styles, variantClasses } = useComponentTheme.useComponentTheme(UNSAFE_Train_themes_redwood_TrainTheme.TrainRedwoodTheme, {
        messageType: messageType,
        stepIconDisabled: isDisabled ? 'isDisabled' : 'notDisabled'
    });
    return (jsxRuntime.jsx("div", { class: classNames.classNames([styles.messageTypeIconBaseStyle, variantClasses]), role: "presentation", children: jsxRuntime.jsx(IconComponent, {}) }));
}

/**
 * Step icon
 */
const StepIcon = ({ isDisabled, isVisited, messageType }) => {
    const { index, selectedStepIndex } = hooks.useContext(Context);
    const isCurrent = selectedStepIndex === index;
    const { interactionProps, applyPseudoHoverStyle, applyHoverStyle } = useInteractionStyle.useInteractionStyle({
        isDisabled: isDisabled || isCurrent
    });
    const { isActive, activeProps } = useActive.useActive({ isDisabled: isDisabled || isCurrent });
    const mergedProps = mergeProps.mergeProps(interactionProps, activeProps);
    const { styles, variantClasses } = useComponentTheme.useComponentTheme(UNSAFE_Train_themes_redwood_TrainTheme.TrainRedwoodTheme, {
        stepIconDisabled: isDisabled ? 'isDisabled' : 'notDisabled',
        stepIconCurrent: isCurrent ? 'isCurrent' : 'notCurrent',
        stepIconVisited: isVisited ? 'isVisited' : 'notVisited',
        stepIconNeedsEventsHover: applyHoverStyle ? 'isNeedsEventsHover' : 'notNeedsEventsHover',
        stepIconPseudoHover: applyPseudoHoverStyle ? 'isPseudoHover' : 'notPseudoHover',
        stepIconActive: isActive ? 'isActive' : 'notActive'
    });
    const iconStyles = () => {
        if (messageType) {
            return jsxRuntime.jsx(UserChosenIcon, { messageType: messageType, isDisabled: isDisabled });
        }
        else {
            return (jsxRuntime.jsx("span", { ...mergedProps, class: classNames.classNames([styles.stepIconBaseStyle, variantClasses]), children: index + 1 }));
        }
    };
    return jsxRuntime.jsx("span", { class: styles.stepIconContainerStyle, children: iconStyles() });
};

/**
 * Step label component
 */
const StepLabel = ({ label, isDisabled, isVisited, messageType }) => {
    const { numberOfSteps, index, selectedStepIndex } = hooks.useContext(Context);
    const translations = useTranslationBundle.useTranslationBundle('@oracle/oraclejet-preact');
    const isCurrent = selectedStepIndex === index;
    const { interactionProps, applyPseudoHoverStyle, applyHoverStyle } = useInteractionStyle.useInteractionStyle({
        isDisabled: isDisabled || isCurrent
    });
    const currentStepStatus = isCurrent
        ? translations.train_current()
        : isVisited
            ? translations.train_visited()
            : translations.train_not_visited();
    const currentStep = index + 1;
    const statusTextForCurrentStep = messageType
        ? `${translations.train_status({
            currentStep: currentStep.toString(),
            numberOfSteps: numberOfSteps.toString()
        })}. ${currentStepStatus}, ${translations.train_message_type()}: ${messageType}`
        : `${translations.train_status({
            currentStep: currentStep.toString(),
            numberOfSteps: numberOfSteps.toString()
        })}. ${currentStepStatus}`;
    const { styles, variantClasses } = useComponentTheme.useComponentTheme(UNSAFE_Train_themes_redwood_TrainTheme.TrainRedwoodTheme, {
        stepLabelDisabled: isDisabled ? 'isDisabled' : 'notDisabled',
        stepLabelCurrent: isCurrent ? 'isCurrent' : 'notCurrent',
        stepLabelVisited: isVisited ? 'isVisited' : 'notVisited',
        stepLabelNeedsEventsHover: applyHoverStyle ? 'isNeedsEventsHover' : 'notNeedsEventsHover',
        stepLabelPseudoHover: applyPseudoHoverStyle ? 'isPseudoHover' : 'notPseudoHover'
    });
    const handleClick = (event) => {
        event.preventDefault();
    };
    return (jsxRuntime.jsx("span", { class: styles.stepLabelWrapperStyle, children: jsxRuntime.jsxs("a", { ...interactionProps, href: "#", class: classNames.classNames([styles.stepLabelBaseStyle, variantClasses]), onClick: handleClick, children: [label, jsxRuntime.jsx(HiddenAccessible.HiddenAccessible, { children: statusTextForCurrentStep })] }) }));
};

/**
 * Step component
 */
const Step = ({ id, isDisabled, isVisited, messageType, label }) => {
    const { onSelect, /* selectedStep ,*/ numberOfSteps, selectedStepIndex, index } = hooks.useContext(Context) || // get the context object or...
        (() => {
            // if there's no context object, throw an error
            throw new Error('Step component must be rendered within a Train component.');
        })();
    const isCurrent = selectedStepIndex === index;
    const [isActive, setIsActive] = hooks.useState(false);
    const { interactionProps, applyPseudoHoverStyle, applyHoverStyle } = useInteractionStyle.useInteractionStyle({
        isDisabled: isDisabled || isCurrent
    });
    const isConnectorConnected = selectedStepIndex > index;
    const onSelectHandler = hooks.useCallback((event) => {
        if (!isDisabled && !isCurrent) {
            onSelect(event, id);
        }
    }, [isDisabled, isCurrent, onSelect, id]);
    const { pressProps } = usePress.usePress(onSelectHandler, {});
    const isLast = index !== numberOfSteps - 1;
    const onMouseUpHandler = hooks.useCallback(() => {
        setIsActive(false);
    }, []);
    const onMouseDownHandler = hooks.useCallback(() => {
        setIsActive(true);
    }, []);
    const { styles, variantClasses } = useComponentTheme.useComponentTheme(UNSAFE_Train_themes_redwood_TrainTheme.TrainRedwoodTheme, {
        stepDisabled: isDisabled ? 'isDisabled' : 'notDisabled',
        needsEventsHover: applyHoverStyle ? 'isNeedsEventsHover' : 'notNeedsEventsHover',
        pseudoHover: applyPseudoHoverStyle ? 'isPseudoHover' : 'notPseudoHover',
        stepActive: isActive && !isCurrent ? 'isActive' : 'notActive'
    });
    const mergedProps = mergeProps.mergeProps(pressProps, interactionProps, {
        onMouseUp: onMouseUpHandler,
        onMouseDown: onMouseDownHandler
    });
    return (jsxRuntime.jsxs("span", { class: styles.stepContainerStyle, children: [isLast && jsxRuntime.jsx(Connector, { isConnected: isConnectorConnected }), jsxRuntime.jsxs("span", { ...mergedProps, class: classNames.classNames([styles.stepBaseStyle, variantClasses]), children: [jsxRuntime.jsx(StepIcon, { isVisited: isVisited, isDisabled: isDisabled, messageType: messageType }), jsxRuntime.jsx(StepLabel, { label: label, isVisited: isVisited, isDisabled: isDisabled, messageType: messageType })] })] }));
};

exports.Step = Step;
exports.Train = Train;
