// @oracle/oraclejet-preact/cjs/useBreakpointsExclusive-a35a1343.js
/* @oracle/oraclejet-preact: undefined */
'use strict';

var hooks = require('preact/hooks');

/*
 * This object defines default breakpoints for screens and containers.
 */
const defaultBreakpoints = {
    sm: '0px',
    md: '600px',
    lg: '1024px',
    xl: '1440px'
};
/*
 * Simple custom hook which returns the current breakpoint based on viewport width, "exclusive" meaning only the breakpoint matching the width will be returned.
 * The breakpoints are returned as an object that is a set of boolean values keyed to each breakpoint.
 * The window/viewport min-width is matched against a range of pixel values defined as the breakpoints.
 */
function useBreakpointsExclusive(breakpoints = defaultBreakpoints) {
    // this handler will return current breakpoints as a single object we can watch, rather than looping through each breakpoint key
    const queryMatches = hooks.useCallback(() => {
        let activeBreakpoint = null;
        const currentMatches = {};
        for (const [key, value] of Object.entries(breakpoints).reverse()) {
            if (window.matchMedia(`(min-width: ${value})`).matches) {
                activeBreakpoint = key;
                break; // Stop at the first match
            }
        }
        Object.keys(breakpoints).forEach((key) => {
            currentMatches[key] = key === activeBreakpoint;
        });
        return currentMatches;
    }, [breakpoints]);
    const [matches, setMatches] = hooks.useState(() => queryMatches());
    hooks.useEffect(() => {
        const mqls = Object.values(breakpoints).map((value) => window.matchMedia(`(min-width: ${value})`));
        const onChange = () => {
            setMatches(queryMatches());
        };
        mqls.forEach((mql) => mql.addEventListener('change', onChange));
        return () => {
            mqls.forEach((mql) => mql.removeEventListener('change', onChange));
        };
    }, [breakpoints, queryMatches]);
    return matches;
}

exports.defaultBreakpoints = defaultBreakpoints;
exports.useBreakpointsExclusive = useBreakpointsExclusive;
