// @oracle/oraclejet-preact/cjs/PRIVATE_BaseDiagram/themes/DiagramStyles.css.js
/* @oracle/oraclejet-preact: undefined */
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });



var buttonStyle = 'DiagramStyles_buttonStyle__4oac9p6';
var childContentStyle = 'DiagramStyles_childContentStyle__4oac9p8';
var dimensionStyle = 'DiagramStyles_dimensionStyle__4oac9pb';
var itemStyles = {nodeStyles:'DiagramStyles_nodeStyles__4oac9p2',linkStyles:'DiagramStyles_linkStyles__4oac9p3',dimmedItemStyle:'DiagramStyles_dimmedItemStyle__4oac9p9'};
var labelStyles = {labelStyle:'DiagramStyles_labelStyle__4oac9p4',labelBorderStyle:'DiagramStyles_labelBorderStyle__4oac9p7'};
var layerStyles = 'DiagramStyles_layerStyles__4oac9p1';
var linkLayerStyle = 'DiagramStyles_linkLayerStyle__4oac9p5';
var outerStyles = 'DiagramStyles_outerStyles__4oac9p0';
var panZoomStyles = 'DiagramStyles_panZoomStyles__4oac9pa';

exports.buttonStyle = buttonStyle;
exports.childContentStyle = childContentStyle;
exports.dimensionStyle = dimensionStyle;
exports.itemStyles = itemStyles;
exports.labelStyles = labelStyles;
exports.layerStyles = layerStyles;
exports.linkLayerStyle = linkLayerStyle;
exports.outerStyles = outerStyles;
exports.panZoomStyles = panZoomStyles;
